--- src/decompiled\Terraria\keyBoardInput.cs
+++ src/merged\Terraria\keyBoardInput.cs
@@ -1,11 +1,19 @@
+#if !WINDOWS
+using Microsoft.Xna.Framework.Input;
+#endif
 using System;
+#if WINDOWS
 using System.Runtime.InteropServices;
 using System.Windows.Forms;
+#else
+using System.Threading;
+#endif
 
 namespace Terraria
 {
 	public class keyBoardInput
 	{
+#if WINDOWS
 		public class inKey : IMessageFilter
 		{
 			public bool PreFilterMessage(ref Message m)
@@ -35,11 +43,28 @@
 
 		[DllImport("user32.dll", CallingConvention = CallingConvention.StdCall, CharSet = CharSet.Auto)]
 		public static extern bool TranslateMessage(IntPtr message);
+#else
+        public static event Action<char> newKeyEvent;
+#endif
 
-		static keyBoardInput()
+        static keyBoardInput()
 		{
+#if WINDOWS
 			keyBoardInput.slashToggle = true;
 			Application.AddMessageFilter(new keyBoardInput.inKey());
-		}
-	}
+#else
+            TextInputEXT.TextInput += delegate(char character)
+            {
+                keyBoardInput.onTextInput(character);
+            };
+#endif
+        }
+
+#if !WINDOWS
+        private static void onTextInput(char key)
+        {
+            keyBoardInput.newKeyEvent(key);
+        }
+#endif
+    }
 }

