--- src/decompiled\Terraria\Lang.cs
+++ src/merged\Terraria\Lang.cs
@@ -6,7 +_,9 @@
 using Terraria.GameInput;
 using Terraria.ID;
 using Terraria.Localization;
+#if CLIENT
 using Terraria.Map;
+#endif
 using Terraria.UI;
 
 namespace Terraria
@@ -457,6 +_,7 @@
 
 		public static void BuildMapAtlas()
 		{
+#if CLIENT
 			Lang._mapLegendCache = new LocalizedText[MapHelper.LookupCount()];
 			for (int i = 0; i < Lang._mapLegendCache.Length; i++)
 			{
@@ -742,6 +_,7 @@
 			Lang._mapLegendCache[MapHelper.TileToLookup(444, 0)] = Language.GetText("MapObject.BeeHive");
 			Lang._mapLegendCache[MapHelper.TileToLookup(466, 0)] = Lang._itemNameCache[3816];
 			Lang._mapLegendCache[MapHelper.TileToLookup(463, 0)] = Lang._itemNameCache[3813];
+#endif
 		}
 
 		public static NetworkText CreateDeathMessage(string deadPlayerName, int plr = -1, int npc = -1, int proj = -1, int other = -1, int projType = 0, int plrItemType = 0)

