--- src/decompiled\Terraria\Lang.cs
+++ src/merged\Terraria\Lang.cs
@@ -1,5 +1,7 @@
 using System;
+#if CLIENT
 using Terraria.Map;
+#endif
 
 namespace Terraria
 {
@@ -32686,6 +32688,7 @@
 				Main.buffTip[78] = "Melee attacks cause confetti to appear";
 				Main.buffName[79] = "Weapon Imbue: Poison";
 				Main.buffTip[79] = "Melee attacks poison enemies";
+#if CLIENT
 				if (Lang.mapLegend == null)
 				{
 					Lang.mapLegend = new string[MapHelper.LookupCount()];
@@ -32940,6 +32943,7 @@
 						Lang.mapLegend[j] = string.Empty;
 					}
 				}
+#endif
 				Lang.chestType[0] = "Chest";
 				Lang.chestType[1] = "Gold Chest";
 				Lang.chestType[2] = "Locked Gold Chest";
@@ -33626,6 +33630,7 @@
 				Main.buffTip[78] = "Nahkampfangriffe verursachen Konfetti zu erscheinen";
 				Main.buffName[79] = "Waffe verleihen: Gift";
 				Main.buffTip[79] = "Nahkampfangriffe Gift Feinde";
+#if CLIENT
 				Lang.mapLegend[MapHelper.TileToLookup(13, 0)] = "Flasche";
 				Lang.mapLegend[MapHelper.TileToLookup(14, 0)] = "Tabelle";
 				Lang.mapLegend[MapHelper.TileToLookup(15, 0)] = "Stuhl";
@@ -33642,6 +33647,7 @@
 				Lang.mapLegend[MapHelper.TileToLookup(114, 0)] = "Tueftler-Werkstatt";
 				Lang.mapLegend[MapHelper.TileToLookup(133, 0)] = "Adamantitschmiede";
 				Lang.mapLegend[MapHelper.TileToLookup(134, 0)] = "Mithrilamboss";
+#endif
 				Lang.prefix[1] = "Gross";
 				Lang.prefix[2] = "Massiv";
 				Lang.prefix[3] = "Gefährlich";
@@ -34248,6 +34254,7 @@
 				Main.buffTip[78] = "Attacchi corpo a corpo causano a comparire coriandoli";
 				Main.buffName[79] = "Arma Imbue: Poison";
 				Main.buffTip[79] = "Mischia attacca nemici veleno";
+#if CLIENT
 				Lang.mapLegend[MapHelper.TileToLookup(13, 0)] = "Bottiglia";
 				Lang.mapLegend[MapHelper.TileToLookup(14, 0)] = "Tavolo";
 				Lang.mapLegend[MapHelper.TileToLookup(15, 0)] = "Sedia";
@@ -34264,6 +34271,7 @@
 				Lang.mapLegend[MapHelper.TileToLookup(114, 0)] = "Laboratorio dell'inventore";
 				Lang.mapLegend[MapHelper.TileToLookup(133, 0)] = "Fornace di adamantio";
 				Lang.mapLegend[MapHelper.TileToLookup(134, 0)] = "Incudine di mitrilio";
+#endif
 				Lang.prefix[1] = "Grande";
 				Lang.prefix[2] = "Massiccio";
 				Lang.prefix[3] = "Pericoloso";
@@ -34870,6 +34878,7 @@
 				Main.buffTip[78] = "Les attaques de mêlée causent des confettis à comparaître";
 				Main.buffName[79] = "Arme Imbue: Poison";
 				Main.buffTip[79] = "Attaques de mêlée des ennemis de poison";
+#if CLIENT
 				Lang.mapLegend[MapHelper.TileToLookup(13, 0)] = "Bouteille";
 				Lang.mapLegend[MapHelper.TileToLookup(14, 0)] = "Table";
 				Lang.mapLegend[MapHelper.TileToLookup(15, 0)] = "Président";
@@ -34886,6 +34895,7 @@
 				Lang.mapLegend[MapHelper.TileToLookup(114, 0)] = "Atelier du bricoleur";
 				Lang.mapLegend[MapHelper.TileToLookup(133, 0)] = "Forge en adamantine";
 				Lang.mapLegend[MapHelper.TileToLookup(134, 0)] = "Enclume en mythril";
+#endif
 				Lang.prefix[1] = "Grand";
 				Lang.prefix[2] = "Massif";
 				Lang.prefix[3] = "Dangereuses";
@@ -35492,6 +35502,7 @@
 				Main.buffTip[78] = "Ataques cuerpo a cuerpo infligen confetti que aparezca";
 				Main.buffName[79] = "imbuir armas: tóxico";
 				Main.buffTip[79] = "Cuerpo a cuerpo ataca a los enemigos de veneno";
+#if CLIENT
 				Lang.mapLegend[MapHelper.TileToLookup(13, 0)] = "Botella";
 				Lang.mapLegend[MapHelper.TileToLookup(14, 0)] = "Mesa";
 				Lang.mapLegend[MapHelper.TileToLookup(15, 0)] = "Silla";
@@ -35508,6 +35519,7 @@
 				Lang.mapLegend[MapHelper.TileToLookup(114, 0)] = "Taller del reparador";
 				Lang.mapLegend[MapHelper.TileToLookup(133, 0)] = "Forja de adamantita";
 				Lang.mapLegend[MapHelper.TileToLookup(134, 0)] = "Yunque de mithril";
+#endif
 				Lang.prefix[1] = "Grande";
 				Lang.prefix[2] = "Masivo";
 				Lang.prefix[3] = "Peligroso";

