--- src/decompiled\Terraria\Minecart.cs
+++ src/merged\Terraria\Minecart.cs
@@ -90,10 +_,12 @@
 
 		public static void Initialize()
 		{
+#if CLIENT
 			if ((float)Main.minecartMountTexture.Width != 50f)
 			{
 				throw new Exception("Be sure to update Minecart.textureWidth to match the actual texture size of " + 50f + ".");
 			}
+#endif
 			Minecart._rightSideConnection = new int[36];
 			Minecart._leftSideConnection = new int[36];
 			Minecart._trackType = new int[36];

