--- src/decompiled\Terraria\Mount.cs
+++ src/merged\Terraria\Mount.cs
@@ -1077,6 +_,7 @@
 				mountData.textureHeight = mountData.frontTexture.Height;
 			}
 			Mount.drillTextureSize = new Vector2(80f, 80f);
+#if CLIENT
 			Vector2 value = new Vector2((float)mountData.textureWidth, (float)(mountData.textureHeight / mountData.totalFrames));
 			if (Mount.drillTextureSize != value)
 			{
@@ -1089,6 +_,7 @@
 					"."
 				}));
 			}
+#endif
 			mountData = new Mount.MountData();
 			Mount.mounts[9] = mountData;
 			mountData.spawnDust = 152;
@@ -1157,6 +_,7 @@
 			Mount.scutlixEyePositions[8] = new Vector2(70f, 34f);
 			Mount.scutlixEyePositions[9] = new Vector2(76f, 34f);
 			Mount.scutlixTextureSize = new Vector2(45f, 54f);
+#if CLIENT
 			Vector2 value2 = new Vector2((float)(mountData.textureWidth / 2), (float)(mountData.textureHeight / mountData.totalFrames));
 			if (Mount.scutlixTextureSize != value2)
 			{
@@ -1169,6 +_,7 @@
 					"."
 				}));
 			}
+#endif
 			for (int num5 = 0; num5 < Mount.scutlixEyePositions.Length; num5++)
 			{
 				Mount.scutlixEyePositions[num5] -= Mount.scutlixTextureSize;

