--- src/decompiled\Terraria\NPC.cs
+++ src/merged\Terraria\NPC.cs
@@ -28501,6 +_,7 @@
 																		31,
 																		229
 																	});
+#if CLIENT
 																	if (flag86)
 																	{
 																		MoonlordDeathDrama.AddExplosion(vector144);
@@ -28514,6 +_,7 @@
 																			dust.scale = num1122;
 																		}
 																	}
+#endif
 																	for (float num1125 = 0f; num1125 < this.ai[1] / 60f; num1125 += 1f)
 																	{
 																		Vector2 vector145 = Utils.RandomVector2(Main.rand, -1f, 1f);

