--- src/decompiled\Terraria\NPC.cs
+++ src/merged\Terraria\NPC.cs
@@ -24184,6 +_,7 @@
 						float num1171 = 1f + Main.rand.NextFloat();
 						float fadeIn = 0.4f + Main.rand.NextFloat();
 						int num1172 = Utils.SelectRandom<int>(Main.rand, 31, 229);
+#if CLIENT
 						if (flag76) {
 							MoonlordDeathDrama.AddExplosion(vector146);
 							for (float num1173 = 0f; num1173 < num1168 * 2f; num1173++) {
@@ -24195,6 +_,8 @@
 								dust2.scale = num1171;
 							}
 						}
+
+#endif
 
 						for (float num1174 = 0f; num1174 < this.ai[1] / 60f; num1174++) {
 							Vector2 vector147 = Utils.RandomVector2(Main.rand, -1f, 1f);
