--- src/decompiled\Terraria\Player.cs
+++ src/merged\Terraria\Player.cs
@@ -30881,9 +_,11 @@
 					int num4 = y;
 					bool num5 = Main.tile[x, num4].active();
 					WorldGen.KillTile(x, num4);
+#if CLIENT
 					if (num5 && !Main.tile[x, num4].active())
 						AchievementsHelper.HandleMining();
 
+#endif
 					if (Main.netMode == 1)
 						NetMessage.SendData(17, -1, -1, null, 0, x, num4);
 				}
