--- src/decompiled\Terraria\Player.cs
+++ src/merged\Terraria\Player.cs
@@ -38632,10 +_,12 @@
 					int num4 = y;
 					bool flag3 = Main.tile[x, num4].active();
 					WorldGen.KillTile(x, num4, false, false, false);
+#if CLIENT
 					if (flag3 && !Main.tile[x, num4].active())
 					{
 						AchievementsHelper.HandleMining();
 					}
+#endif
 					if (Main.netMode == 1)
 					{
 						NetMessage.SendData(17, -1, -1, null, 0, (float)x, (float)num4, 0f, 0, 0, 0);

