--- src/decompiled\Terraria\Program.cs
+++ src/merged\Terraria\Program.cs
@@ -4,7 +_,9 @@
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Threading;
+#if CLIENT
 using System.Windows.Forms;
+#endif
 using Terraria.Initializers;
 using Terraria.Social;
 
@@ -12,7 +_,11 @@
 {
 	public static class Program
 	{
+#if CLIENT
 		public const bool IsServer = false;
+#else
+        public const bool IsServer = true;
+#endif
 
 		public static Dictionary<string, string> LaunchParameters = new Dictionary<string, string>();
 
@@ -25,15 +_,23 @@
 			for (int i = 0; i < array.Length; i++)
 			{
 				Type type = array[i];
+#if WINDOWS
 				MethodInfo[] methods = type.GetMethods(BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
-				MethodInfo[] array2 = methods;
+#else
+                MethodInfo[] methods = type.GetMethods();
+#endif
+                MethodInfo[] array2 = methods;
 				for (int j = 0; j < array2.Length; j++)
 				{
 					MethodInfo methodInfo = array2[j];
 					if (!methodInfo.IsAbstract && !methodInfo.ContainsGenericParameters && methodInfo.GetMethodBody() != null)
 					{
+#if WINDOWS
 						RuntimeHelpers.PrepareMethod(methodInfo.MethodHandle);
-					}
+#else
+                        Program.JitForcedMethodCache = methodInfo.MethodHandle.GetFunctionPointer();
+#endif
+                    }
 				}
 			}
 		}
@@ -94,7 +_,11 @@
 					{
 						Program.ForceLoadAssembly(Assembly.GetExecutingAssembly(), true);
 					};
+#if CLIENT
 					main.Run();
+#else
+                    main.DedServ();
+#endif
 				}
 				catch (Exception e)
 				{
@@ -113,7 +_,14 @@
 					streamWriter.WriteLine(e);
 					streamWriter.WriteLine("");
 				}
+#if CLIENT
 				MessageBox.Show(e.ToString(), "Terraria: Error");
+#else
+				Console.WriteLine("Server crash: " + DateTime.Now);
+				Console.WriteLine(e);
+				Console.WriteLine("");
+				Console.WriteLine("Please send crashlog.txt to support@terraria.org");
+#endif
 			}
 			catch
 			{

