--- src/decompiled\Terraria\WindowsLaunch.cs
+++ src/merged\Terraria\WindowsLaunch.cs
@@ -1,3 +_,4 @@
+#if WINDOWS
 using System;
 using System.IO;
 using System.Reflection;
@@ -89,6 +_,10 @@
 				arg_24_1 = (WindowsLaunch.<>c.<>9__5_0 = new ResolveEventHandler(WindowsLaunch.<>c.<>9.<Main>b__5_0));
 			}
 			arg_24_0.AssemblyResolve += arg_24_1;
+#if SERVER
+			WindowsLaunch._handleRoutine = new WindowsLaunch.HandlerRoutine(WindowsLaunch.ConsoleCtrlCheck);
+			WindowsLaunch.SetConsoleCtrlHandler(WindowsLaunch._handleRoutine, true);
+#endif
 			Program.LaunchGame(args, false);
 		}
 	}

