--- src/decompiled\Terraria\WorldGen.cs
+++ src/merged\Terraria\WorldGen.cs
@@ -15,10 +_,12 @@
 using Terraria.ID;
 using Terraria.IO;
 using Terraria.Localization;
-using Terraria.Map;
 using Terraria.ObjectData;
 using Terraria.Utilities;
 using Terraria.World.Generation;
+#if CLIENT
+using Terraria.Map;
+#endif
 
 namespace Terraria
 {
@@ -47796,6 +_,7 @@
 		public static bool UpdateMapTile(int i, int j, bool addToList = true)
 		{
 			bool result = false;
+#if CLIENT
 			if (Main.mapEnabled && !WorldGen.noMapUpdate && !WorldGen.gen && Main.Map[i, j].Light > 0 && Main.Map.UpdateType(i, j) && addToList)
 			{
 				result = true;
@@ -47810,6 +_,7 @@
 					Main.refreshMap = true;
 				}
 			}
+#endif
 			return result;
 		}
 

