--- src/decompiled\Terraria\WorldGen.cs
+++ src/merged\Terraria\WorldGen.cs
@@ -16,7 +_,9 @@
 using Terraria.ID;
 using Terraria.IO;
 using Terraria.Localization;
+#if CLIENT
 using Terraria.Map;
+#endif
 using Terraria.ObjectData;
 using Terraria.Utilities;
 using Terraria.World.Generation;
@@ -48151,6 +_,7 @@
 		public static bool UpdateMapTile(int i, int j, bool addToList = true)
 		{
 			bool result = false;
+#if CLIENT
 			if (Main.mapEnabled && !WorldGen.noMapUpdate && !WorldGen.gen && ((Main.Map[i, j].Light > 0 && Main.Map.UpdateType(i, j)) & addToList))
 			{
 				result = true;
@@ -48165,6 +_,7 @@
 					Main.refreshMap = true;
 				}
 			}
+#endif
 			return result;
 		}
 

