--- src/TerrariaNetCore/Terraria/ID/AchievementHelperID.cs
+++ src/tModLoader/Terraria/ID/AchievementHelperID.cs
@@ -1,8 +_,8 @@
 namespace Terraria.ID;
 
-public class AchievementHelperID
+public partial class AchievementHelperID
 {
-	public class Events
+	public partial class Events
 	{
 		public const int NightStart = 0;
 		public const int DayStart = 1;
@@ -35,7 +_,7 @@
 		public const int DroneDiedInSpace = 28;
 	}
 
-	public class Special
+	public partial class Special
 	{
 		public const int ConsumeHeart = 0;
 		public const int ConsumeStar = 1;
