--- src/TerrariaNetCore/Terraria/ID/AmmoID.cs
+++ src/tModLoader/Terraria/ID/AmmoID.cs
@@ -2,10 +_,14 @@
 
 namespace Terraria.ID;
 
-public static class AmmoID
+public static partial class AmmoID
 {
-	public class Sets
+	public partial class Sets
 	{
+		/// <summary>
+		/// Associates a launcher's item type (<see cref="Item.type"/>) and an ammo's item type (<see cref="Item.type"/>) to the projectile type (<see cref="Projectile.type"/>) they will shoot when used together.
+		/// <br/> For example, a <see cref="ItemID.SnowmanCannon"/> used with a <see cref="ItemID.MiniNukeI"/> will fire the <see cref="ProjectileID.MiniNukeSnowmanRocketI"/>.
+		/// </summary>
 		public static Dictionary<int, Dictionary<int, int>> SpecificLauncherAmmoProjectileMatches = new Dictionary<int, Dictionary<int, int>> {
 			{
 				759,
