--- src/TerrariaNetCore/Terraria/ID/AmmoID.cs
+++ src/tModLoader/Terraria/ID/AmmoID.cs
@@ -1,11 +_,18 @@
+#if !TMLANALYZER
 using System.Collections.Generic;
+#endif
 
 namespace Terraria.ID;
 
-public static class AmmoID
+public static partial class AmmoID
 {
-	public class Sets
+#if !TMLANALYZER
+	public partial class Sets
 	{
+		/// <summary>
+		/// Associates a launcher's item type (<see cref="Item.type"/>) and an ammo's item type (<see cref="Item.type"/>) to the projectile type (<see cref="Projectile.type"/>) they will shoot when used together.
+		/// <br/> For example, a <see cref="ItemID.SnowmanCannon"/> used with a <see cref="ItemID.MiniNukeI"/> will fire the <see cref="ProjectileID.MiniNukeSnowmanRocketI"/>.
+		/// </summary>
 		public static Dictionary<int, Dictionary<int, int>> SpecificLauncherAmmoProjectileMatches = new Dictionary<int, Dictionary<int, int>> {
 			{
 				759,
@@ -94,6 +_,7 @@
 			}
 		};
 	}
+#endif
 
 	public static int None = 0;
 	public static int Gel = 23;
@@ -101,6 +_,7 @@
 	public static int Coin = 71;
 	public static int FallenStar = 75;
 	public static int Bullet = 97;
+	/// <summary> Items using this as <see cref="Item.ammo"/> can use <see cref="ItemID.Sets.SandgunAmmoProjectileData"/> to customize the projectile and bonus damage of this ammo. </summary>
 	public static int Sand = 169;
 	public static int Dart = 283;
 	public static int Rocket = 771;
