--- src/TerrariaNetCore/Terraria/ID/AmmoID.cs
+++ src/tModLoader/Terraria/ID/AmmoID.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using Terraria.ModLoader;
 
 namespace Terraria.ID;
 
@@ -6,6 +_,7 @@
 {
 	public class Sets
 	{
+		public static SetFactory Factory = new SetFactory(ItemLoader.ItemCount);
 		public static Dictionary<int, Dictionary<int, int>> SpecificLauncherAmmoProjectileMatches = new Dictionary<int, Dictionary<int, int>> {
 				{ 759, new Dictionary<int, int> {
 						{
@@ -257,6 +_,10 @@
 							717
 						}
 					} }		};
+
+		public static bool[] IsArrow = Factory.CreateBoolSet(false, Arrow, Stake);
+		public static bool[] IsBullet = Factory.CreateBoolSet(false, Bullet, CandyCorn);
+		public static bool[] IsRocket = Factory.CreateBoolSet(false, Rocket, StyngerBolt, JackOLantern, NailFriendly);
 	}
 
 	public static int None = 0;
