--- src/Terraria/Terraria/ID/BiomeID.cs
+++ src/tModLoader/Terraria/ID/BiomeID.cs
@@ -1,7 +_,15 @@
+using Terraria.GameContent;
+
 namespace Terraria.ID
 {
-	internal class BiomeID
+	public class BiomeID
 	{
+		/// <summary>
+		/// Equivalent to <see cref="ShopHelper.BiomeNameByKey"/>
+		/// </summary>
+		/// <param name="biomeID"></param>
+		/// <returns></returns>
+		public static string GetLocalizedName(int biomeID) => ShopHelper.BiomeNameByKey(biomeID);
 		public const int Forest = 0;
 		public const int NormalUnderground = 1;
 		public const int Snow = 2;
@@ -13,5 +_,6 @@
 		public const int Dungeon = 8;
 		public const int Corruption = 9;
 		public const int Crimson = 10;
+		public const int Count = 11;
 	}
 }
