--- src/TerrariaNetCore/Terraria/ID/BiomeConversionID.cs
+++ src/tModLoader/Terraria/ID/BiomeConversionID.cs
@@ -1,5 +_,10 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
+#if TMLCODEASSIST
+[tModCodeAssist.IDType.Sets.AssociatedName(ModLoader.Annotations.IDTypeAttribute.BiomeConversion)]
+#endif
 public static class BiomeConversionID
 {
 	public const int Purity = 0;
@@ -10,4 +_,5 @@
 	public const int Sand = 5;
 	public const int Snow = 6;
 	public const int Dirt = 7;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(BiomeConversionID), typeof(int)); // TML
 }
