--- src/Terraria/Terraria/ID/BiomeID.cs
+++ src/tModLoader/Terraria/ID/BiomeID.cs
@@ -1,7 +_,10 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID
 {
-	internal class BiomeID
+	public class BiomeID
 	{
+		public static readonly IdDictionary Search = IdDictionary.Create<BiomeID, int>();
 		public const int Forest = 0;
 		public const int NormalUnderground = 1;
 		public const int Snow = 2;
@@ -13,5 +_,6 @@
 		public const int Dungeon = 8;
 		public const int Corruption = 9;
 		public const int Crimson = 10;
+		public const int Count = 11;
 	}
 }
