--- src/Terraria/Terraria/ID/BuffID.cs
+++ src/tModLoader/Terraria/ID/BuffID.cs
@@ -1,5 +_,8 @@
 using ReLogic.Reflection;
 
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID
 {
 	public class BuffID
@@ -12,7 +_,7 @@
 				public bool faceLeft;
 			}
 
-			public static SetFactory Factory = new SetFactory(327);
+			public static SetFactory Factory = new SetFactory(BuffLoader.BuffCount);
 			public static bool[] IsWellFed = Factory.CreateBoolSet(26, 206, 207);
 			public static bool[] IsAnNPCWhipDebuff = Factory.CreateBoolSet(307, 313, 319, 316, 310, 309, 315, 326);
 			public static bool[] TimeLeftDoesNotDecrease = Factory.CreateBoolSet(28);
