--- src/TerrariaNetCore/Terraria/ID/BuffID.cs
+++ src/tModLoader/Terraria/ID/BuffID.cs
@@ -1,10 +_,11 @@
 using ReLogic.Reflection;
+using Terraria.ModLoader;
 
 namespace Terraria.ID;
 
-public class BuffID
+public partial class BuffID
 {
-	public class Sets
+	public partial class Sets
 	{
 		public class BuffMountData
 		{
@@ -12,9 +_,12 @@
 			public bool faceLeft;
 		}
 
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(BuffLoader.BuffCount);
 		public static bool[] IsWellFed = Factory.CreateBoolSet(26, 206, 207);
 		public static bool[] IsFedState = Factory.CreateBoolSet(26, 206, 207, 332, 333, 334);
+		/// <summary>
+		/// Indicates that this buff is a whip debuff. Some enemies are immune to all non-whip debuffs, and some are additionally immune to whip debuffs. Setting this to true for a whip debuff properly accounts for that interaction. (see <see cref="Terraria.DataStructures.NPCDebuffImmunityData.ImmuneToAllBuffsThatAreNotWhips"/>)
+		/// </summary>
 		public static bool[] IsAnNPCWhipDebuff = Factory.CreateBoolSet(307, 313, 319, 316, 310, 309, 315, 326, 340);
 		public static bool[] TimeLeftDoesNotDecrease = Factory.CreateBoolSet(28, 334, 29, 159, 150, 93, 348);
 		public static bool[] CanBeRemovedByNetMessage = Factory.CreateBoolSet(313);
