--- src/Terraria/Terraria/ID/BuffID.cs
+++ src/tModLoader/Terraria/ID/BuffID.cs
@@ -3,3 +_,6 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID
 {
 	public class BuffID
@@ -12,8 +_,8 @@
 				public bool faceLeft;
 			}
 
-			public static SetFactory Factory = new SetFactory(327);
+			public static SetFactory Factory = new SetFactory(BuffLoader.BuffCount);
 			public static bool[] IsWellFed = Factory.CreateBoolSet(26, 206, 207);
 			public static bool[] IsAnNPCWhipDebuff = Factory.CreateBoolSet(307, 313, 319, 316, 310, 309, 315, 326);
 			public static bool[] TimeLeftDoesNotDecrease = Factory.CreateBoolSet(28);
 			public static bool[] CanBeRemovedByNetMessage = Factory.CreateBoolSet(313);
@@ -600,5 +_,6 @@
 		public const int FlinxMinion = 325;
 		public const int BoneWhipNPCDebuff = 326;
 		public const int Count = 327;
+		public static readonly IdDictionary Search = IdDictionary.Create<BuffID, int>();
 	}
 }
