--- src/TerrariaNetCore/Terraria/ID/ChainID.cs
+++ src/tModLoader/Terraria/ID/ChainID.cs
@@ -1,6 +_,8 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-public static class ChainID
+public class ChainID
 {
 	public const short TendonHook = 0;
 	public const short ThornHook = 1;
@@ -20,4 +_,5 @@
 	public const short LunarStardustGlow = 15;
 	public const short StaticHook = 16;
 	public static readonly short Count = 17;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(ChainID), typeof(short)); // TML
 }
