--- src/Terraria/Terraria/ID/ContentSamples.cs
+++ src/tModLoader/Terraria/ID/ContentSamples.cs
@@ -1,7 +_,10 @@
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.GameContent.Bestiary;
 using Terraria.Graphics.Shaders;
+using Terraria.ModLoader;
+using Terraria.Utilities;
 
 namespace Terraria.ID
 {
@@ -92,6 +_,7 @@
 				public ItemGroupAndOrderInGroup(Item item) {
 					ItemType = item.type;
 					Group = GetItemGroup(item, out OrderInGroup);
+					ItemLoader.ModifyResearchSorting(item, ref Group);
 				}
 			}
 
@@ -633,7 +_,7 @@
 
 			public static void SetCreativeMenuOrder() {
 				List<Item> list = new List<Item>();
-				for (int i = 1; i < 5088; i++) {
+				for (int i = 1; i < ItemLoader.ItemCount; i++) {
 					Item item = new Item();
 					item.SetDefaults(i);
 					list.Add(item);
@@ -721,11 +_,13 @@
 		public static Dictionary<int, CreativeHelper.ItemGroupAndOrderInGroup> ItemCreativeSortingId = new Dictionary<int, CreativeHelper.ItemGroupAndOrderInGroup>();
 
 		public static void Initialize() {
+			Main.rand ??= new UnifiedRandom((int)DateTime.Now.Ticks);
+		
 			NpcsByNetId.Clear();
 			NpcNetIdsByPersistentIds.Clear();
 			NpcPersistentIdsByNetIds.Clear();
 			NpcBestiarySortingId.Clear();
-			for (int i = -65; i < 668; i++) {
+			for (int i = -65; i < NPCLoader.NPCCount; i++) {
 				NPC nPC = new NPC();
 				nPC.SetDefaults(i);
 				NpcsByNetId[i] = nPC;
@@ -737,14 +_,14 @@
 
 			ModifyNPCIds();
 			ProjectilesByType.Clear();
-			for (int j = 0; j < 956; j++) {
+			for (int j = 0; j < ProjectileLoader.ProjectileCount; j++) {
 				Projectile projectile = new Projectile();
 				projectile.SetDefaults(j);
 				ProjectilesByType[j] = projectile;
 			}
 
 			ItemsByType.Clear();
-			for (int k = 0; k < 5088; k++) {
+			for (int k = 0; k < ItemLoader.ItemCount; k++) {
 				Item item = new Item();
 				item.SetDefaults(k);
 				ItemsByType[k] = item;
@@ -903,7 +_,7 @@
 			NPCSpawnParams nPCSpawnParams = default(NPCSpawnParams);
 			nPCSpawnParams.gameModeData = Main.RegisterdGameModes[0];
 			NPCSpawnParams spawnparams = nPCSpawnParams;
-			for (int i = -65; i < 668; i++) {
+			for (int i = -65; i < NPCLoader.NPCCount; i++) {
 				NPC nPC = new NPC();
 				nPC.SetDefaults(i, spawnparams);
 				NpcBestiaryRarityStars[i] = GetNPCBestiaryRarityStarsCount(nPC);
