--- src/Terraria/Terraria/ID/ContentSamples.cs
+++ src/tModLoader/Terraria/ID/ContentSamples.cs
@@ -2,6 +_,7 @@
 using System.Linq;
 using Terraria.GameContent.Bestiary;
 using Terraria.Graphics.Shaders;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -680,7 +_,7 @@
 			NpcNetIdsByPersistentIds.Clear();
 			NpcPersistentIdsByNetIds.Clear();
 			NpcBestiarySortingId.Clear();
-			for (int i = -65; i < 663; i++) {
+			for (int i = -65; i < NPCLoader.NPCCount; i++) {
 				NPC nPC = new NPC();
 				nPC.SetDefaults(i);
 				NpcsByNetId[i] = nPC;
@@ -692,14 +_,14 @@
 
 			ModifyNPCIds();
 			ProjectilesByType.Clear();
-			for (int j = 0; j < 950; j++) {
+			for (int j = 0; j < ProjectileLoader.ProjectileCount; j++) {
 				Projectile projectile = new Projectile();
 				projectile.SetDefaults(j);
 				ProjectilesByType[j] = projectile;
 			}
 
 			ItemsByType.Clear();
-			for (int k = 0; k < 5045; k++) {
+			for (int k = 0; k < ItemLoader.ItemCount; k++) {
 				Item item = new Item();
 				item.SetDefaults(k);
 				ItemsByType[k] = item;
@@ -858,7 +_,7 @@
 			NPCSpawnParams nPCSpawnParams = default(NPCSpawnParams);
 			nPCSpawnParams.gameModeData = Main.RegisterdGameModes[0];
 			NPCSpawnParams spawnparams = nPCSpawnParams;
-			for (int i = -65; i < 663; i++) {
+			for (int i = -65; i < NPCLoader.NPCCount; i++) {
 				NPC nPC = new NPC();
 				nPC.SetDefaults(i, spawnparams);
 				NpcBestiaryRarityStars[i] = GetNPCBestiaryRarityStarsCount(nPC);
