--- src/TerrariaNetCore/Terraria/ID/GameModeID.cs
+++ src/tModLoader/Terraria/ID/GameModeID.cs
@@ -1,9 +_,15 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-internal class GameModeID
+#if TMLCODEASSIST
+[tModCodeAssist.IDType.Sets.AssociatedName(ModLoader.Annotations.IDTypeAttribute.GameMode)]
+#endif
+public static class GameModeID
 {
 	public const short Normal = 0;
 	public const short Expert = 1;
 	public const short Master = 2;
 	public const short Creative = 3;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(GameModeID), typeof(int));
 }
