--- src/TerrariaNetCore/Terraria/ID/InvasionID.cs
+++ src/tModLoader/Terraria/ID/InvasionID.cs
@@ -1,6 +_,8 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-public static class InvasionID
+public class InvasionID
 {
 	public const short CachedInvasions = 3;
 	public const short CachedFrostMoon = 1;
@@ -12,4 +_,5 @@
 	public const short PirateInvasion = 3;
 	public const short MartianMadness = 4;
 	public static readonly short Count = 5;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(InvasionID), typeof(short)); // TML
 }
