--- src/TerrariaNetCore/Terraria/ID/ItemUseStyleID.cs
+++ src/tModLoader/Terraria/ID/ItemUseStyleID.cs
@@ -1,20 +_,65 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
+/// <summary>
+/// Assign <see cref="Item.useStyle"/> to one of these to give your item an animation while in use. The <see href="https://terraria.wiki.gg/wiki/Use_Style_IDs">Use Style IDs wiki page</see> has examples and animations of each of these use styles.
+/// <br/> If none of these animations match what you want, consider making the animation a behavior of a held projectile or use the <see cref="ModLoader.ModItem.UseStyle"/> hook to implement a custom animation.
+/// </summary>
 public class ItemUseStyleID
 {
 	public const int None = 0;
+	/// <summary>Item is swung in an overhead arch.<br/>Used for many sword weapons, block placement, etc.</summary>
 	public const int Swing = 1;
+	/// <summary>
+	/// Unused
+	/// </summary>
 	public const int DrinkOld = 7;
+	/// <summary>Item is thrust horizontally.<br/>Use by <see cref="ItemID.Umbrella"/> and <see cref="ItemID.TragicUmbrella"/></summary>
 	public const int Thrust = 3;
+	/// <summary>Item is held up.<br/>Used for items such as mana/life crystals, life fruit, magic mirrors, and summoning items</summary>
 	public const int HoldUp = 4;
+	/// <summary>
+	/// Item is held in front of player pointing towards mouse. Item sprite should be oriented horizontally with handle at left and muzzle end at right. Set <c><see cref="Item.staff"/>[Type] = true;</c> in SetStaticDefaults if the weapon sprite is oriented like a staff (handle at bottom left, end at top right).
+	/// <br/> Used by items such as guns, spell books, flails, yoyos, and spears
+	/// </summary>
 	public const int Shoot = 5;
+	/// <summary>
+	/// Item is brought in front of player and rotated towards player.
+	/// <br/> Used by Recall Potion and Potion of Return
+	/// </summary>
 	public const int DrinkLong = 6;
+	/// <summary>Eating or using<br/>Used by food</summary>
 	public const int EatFood = 2;
+	/// <summary>
+	/// Item is swung like a golf club.
+	/// <br/> Used by golf clubs.
+	/// </summary>
 	public const int GolfPlay = 8;
+	/// <summary>
+	/// Item is brought in front of player and rotated towards player.
+	/// <br/> Used by potions, drinks, flasks, and hair dyes.
+	/// </summary>
 	public const int DrinkLiquid = 9;
 	public const int HiddenAnimation = 10;
+	/// <summary>
+	/// Used by Lawn Mower
+	/// </summary>
 	public const int MowTheLawn = 11;
+	/// <summary>
+	/// Used by guitar-shaped instruments
+	/// </summary>
 	public const int Guitar = 12;
+	/// <summary>
+	/// Item is thrust in any direction towards the mouse.
+	/// Used by Shortswords
+	/// </summary>
 	public const int Rapier = 13;
+	/// <summary>
+	/// Item is raised high by off hand.
+	/// <br/> Used by Nightglow.
+	/// </summary>
 	public const int RaiseLamp = 14;
+	public const int Count = 15;
+	public static readonly IdDictionary Search = IdDictionary.Create<ItemUseStyleID, int>(); // Added by TML
 }
