--- src/Terraria/Terraria/ID/ItemID.cs
+++ src/tModLoader/Terraria/ID/ItemID.cs
@@ -2,10 +_,11 @@
 using ReLogic.Reflection;
 using System.Collections.Generic;
 using Terraria.DataStructures;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
-	public class ItemID
+	public partial class ItemID
 	{
 		public struct BannerEffect
 		{
@@ -33,9 +_,9 @@
 			}
 		}
 
-		public class Sets
+		public partial class Sets
 		{
-			public static SetFactory Factory = new SetFactory(5125);
+			public static SetFactory Factory = new SetFactory(ItemLoader.ItemCount);
 			public static List<int> ItemsThatAreProcessedAfterNormalContentSample = new List<int> {
 				1533,
 				1534,
@@ -918,10 +_,14 @@
 			public static bool[] CommonCoin = Factory.CreateBoolSet(71, 72, 73, 74);
 			public static bool[] ItemIconPulse = Factory.CreateBoolSet(520, 521, 575, 549, 548, 547, 3456, 3457, 3458, 3459, 3580, 3581);
 			public static bool[] ItemNoGravity = Factory.CreateBoolSet(520, 521, 575, 549, 548, 547, 3453, 3454, 3455, 3456, 3457, 3458, 3459, 3580, 3581, 4143);
+			// Paired with a change in Player.ExtractinatorUse, non default extractType goes by itemId so DesertFossil is 3347 rather than 1 now
-			public static int[] ExtractinatorMode = Factory.CreateIntSet(-1, 424, 0, 1103, 0, 3347, 1);
+			public static int[] ExtractinatorMode = Factory.CreateIntSet(-1, 424, 0, 1103, 0, 3347, 3347);
 			public static int[] StaffMinionSlotsRequired = Factory.CreateIntSet(1);
 			public static bool[] ExoticPlantsForDyeTrade = Factory.CreateBoolSet(3385, 3386, 3387, 3388);
 			public static bool[] NebulaPickup = Factory.CreateBoolSet(3453, 3454, 3455);
+			/// <summary>
+			/// Use in conjunction with Main.RegisterItemAnimation to enable its animation in the world
+			/// </summary>
 			public static bool[] AnimatesAsSoul = Factory.CreateBoolSet(575, 547, 520, 548, 521, 549, 3580, 3581);
 			public static bool[] gunProj = Factory.CreateBoolSet(3475, 3540, 3854, 3930);
 			public static int[] SortingPriorityBossSpawns = Factory.CreateIntSet(-1, 43, 1, 560, 2, 70, 3, 1331, 3, 361, 4, 5120, 5, 1133, 5, 4988, 6, 544, 7, 556, 8, 557, 9, 2495, 10, 2673, 11, 602, 12, 1844, 13, 1958, 14, 1293, 15, 2767, 16, 4271, 16, 3601, 17, 1291, 18, 109, 19, 29, 20, 50, 21, 3199, 21, 3124, 22);
