--- src/TerrariaNetCore/Terraria/ID/ItemSourceID.cs
+++ src/tModLoader/Terraria/ID/ItemSourceID.cs
@@ -1,15 +_,34 @@
+using System;
+
+#nullable enable
+
 namespace Terraria.ID;
 
-public static class ItemSourceID
+//TML: Made internal in favor of string contexts on every entity source.
+internal static class ItemSourceID
 {
 	public static readonly int None = 0;
 	public static readonly int SetBonus_Nebula = 1;
 	public static readonly int LuckyCoin = 2;
 	public static readonly int PlayerDeath = 3;
-	public static readonly int PlayerDrop = 4;
+	public static readonly int ThrowItem = 4;
 	public static readonly int GrandDesignOrMultiColorWrench = 5;
 	public static readonly int TorchGod = 6;
 	public static readonly int Shimmer = 8;
 	public static readonly int Digesting = 9;
 	public static readonly int InventoryOverflow = 10;
+
+	public static string? ToContextString(int itemSourceId) => itemSourceId switch {
+		SetBonus_Nebula => nameof(SetBonus_Nebula),
+		LuckyCoin => nameof(LuckyCoin),
+		ThrowItem => nameof(ThrowItem),
+		GrandDesignOrMultiColorWrench => nameof(GrandDesignOrMultiColorWrench),
+		SortingWithNoSpace => nameof(SortingWithNoSpace),
+		Shimmer => nameof(Shimmer),
+		Digesting => nameof(Digesting),
+		TorchGod => nameof(TorchGod),
+		PlayerDeath => throw new ArgumentException($"Unexpected or invalid value: '{itemSourceId}'."), // Due to 'EntitySource_Death being added.
+		InventoryOverflow => throw new ArgumentException($"Unexpected or invalid value: '{itemSourceId}'."), // Due to 'EntitySource_OverfullInventory being added.
+		_ => null,
+	};
 }
