--- src/TerrariaNetCore/Terraria/ID/ItemSourceID.cs
+++ src/tModLoader/Terraria/ID/ItemSourceID.cs
@@ -1,16 +_,34 @@
+using System;
+
+#nullable enable
+
 namespace Terraria.ID;
 
+//TML: Made internal in favor of string contexts on every entity source.
-public static class ItemSourceID
+internal static class ItemSourceID
 {
 	public const int None = 0;
 	public const int SetBonus_Nebula = 1;
 	public const int LuckyCoin = 2;
 	public const int PlayerDeath = 3;
-	public const int PlayerDropItemCheck = 4;
+	public const int ThrowItem = 4;
 	public const int GrandDesignOrMultiColorWrench = 5;
 	public const int TorchGod = 6;
 	public const int SortingWithNoSpace = 7;
 	public const int Shimmer = 8;
 	public const int Digesting = 9;
 	public static readonly int Count = 10;
+
+	public static string? ToContextString(int itemSourceId) => itemSourceId switch {
+		SetBonus_Nebula => nameof(SetBonus_Nebula),
+		LuckyCoin => nameof(LuckyCoin),
+		ThrowItem => nameof(ThrowItem),
+		GrandDesignOrMultiColorWrench => nameof(GrandDesignOrMultiColorWrench),
+		SortingWithNoSpace => nameof(SortingWithNoSpace),
+		Shimmer => nameof(Shimmer),
+		Digesting => nameof(Digesting),
+		TorchGod => nameof(TorchGod),
+		PlayerDeath => throw new ArgumentException($"Unexpected or invalid value: '{itemSourceId}'."), // Due to 'EntitySource_Death being added.
+		_ => null,
+	};
 }
