--- src/TerrariaNetCore/Terraria/ID/MessageID.cs
+++ src/tModLoader/Terraria/ID/MessageID.cs
@@ -1,163 +_,673 @@
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.ID;
 
-public class MessageID
+// The comments here may be quite outdated, and constants need to be made sense of.
+
+/// <summary>
+/// Vanilla netcode messages. These are used in <see cref="NetMessage.SendData(int, int, int, Terraria.Localization.NetworkText, int, float, float, float, int, int, int)"/> to send data between the server and clients. The documentation of each <see cref="MessageID"/> value explains the the purpose of the message and most importantly the meaning of each of the SendData parameters for that message.
+/// </summary>
+public partial class MessageID
 {
 	public const byte NeverCalled = 0;
+
+	// Set by the client when the socket connects, client connection state 0->1
+	// Sends a string "Terraria{Main.curRelease}"
+	// Checks if a client is banned, and kicks if so
+	// Does a version check and kicks if the version is not equal
+	// If the server has a password, sends RequestPassword, otherwise sends LoadPlayer
+	// TModLoader changes the version string to ModLoader.versionedName
+	// If the server accepts vanilla clients and the client sent a vanilla version tModLoader flags them as a vanilla client and communicates using an unmodified protocol
+	// Otherwise sends SyncMods instead of LoadPlayer
 	public const byte Hello = 1;
+
+	// Kick with reason, displays the text "{ip} was booted: {reason}" in the server console
 	public const byte Kick = 2;
+
+	// Sent once the server has accepted the player and assigned an id
+	// Client connection state 1->2
+	// Client sends SyncPlayer, ClientUUID, PlayerHealth, PlayerMana, PlayerBuffs and SyncEquipment
+	// TModLoader calls PlayerLoader.SyncPlayer
+	// Client sends RequestWorldInfo
+	// Client connection state 2->3
 	public const byte PlayerInfo = 3;
+
+	// Sends a player's name, visual configuration, accessory visibility, difficulty and extra accessory flag
+	// Kicks if the playername is empty, too long, or the same as another player on the server
+	// Forwarded to other clients
 	public const byte SyncPlayer = 4;
+
+	// Sends item slot, type, quantity, prefix, and then mod data with ItemIO.SendModData
+	// Item slot is an index into a contiguous array of the player's (inventory, armor, miscEquips, miscDyes, bank, safe and trash)
+	// Packet will be ignored on the client unless they are quick-stacking to chests because clients have authority over their inventory
+	// Forwarded to other clients if it is not a bank, safe or trash slot
 	public const byte SyncEquipment = 5;
+
+	// Sent by the client in response to LoadPlayer, after all player data has been sent, and the client state is transitioning from 2->3
+	// Server calls RemoteClient.ResetSections and state 1->2
+	// Server responds with WorldInfo and InvasionProgressReport
 	public const byte RequestWorldData = 6;
+
+	// Client connection state 3->4
+	/// <summary>
+	/// Sends all of the world state data, such as time of day, weather, events, world size and name, biome info, killed bosses, etc from the server to the clients. This includes modded world data from <see cref="ModLoader.ModSystem.NetSend(System.IO.BinaryWriter)"/>.
+	/// <br/><br/>Sent whenever any of the properties in the packet change (except time of day, which gets synced whenever other properties change anyway)
+	/// <br/><br/> Mods should send this message when manually manipulating world state data, such as a value synced in <see cref="ModLoader.ModSystem.NetSend(System.IO.BinaryWriter)"/>, to keep each client in sync with the server.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters: None
+	/// <br/><br/> <b>Server->Clients only</b>
+	/// </summary>
 	public const byte WorldData = 7;
+
+	// Client state 5->6
+	// Client sends requested spawn location
+	// Server sends RequestTileData, TileSection, FrameSection for 5x3 section blocks around the requested and default spawns,
+	//    and 3x3 section blocks around the exits of any portals in the spawn areas
+	// SyncItem and ItemOwner for every active item
+	// SyncNPC for every active npc
+	// SyncProjectile for every active projectile with the netImportant property or listed in Main.projPet
+	// NPCKillCountDeathTally for all NPCs (technically BannerIDs)
+	// StartPlaying, TileCounts, WorldInfo, MoonlordCountdown, UpdateTowerShieldStrengths
 	public const byte SpawnTileData = 8;
+
+	// Sets the client status text
+	// Only use is before NetMessage.SendSection with message "Receiving tile data"
 	public const byte StatusTextSize = 9;
+
+	// Sends a compressed 200x150 block section of the map
+	// Sent from NetMessage.SendSection which also sends NPCInfo for all npcs in the section
 	public const byte TileSection = 10;
+
+	// Frames a rectangle of tile sections
+	// TModLoader changes walls from byte to short
+	// Sent after a NetMessage.SendSection block
 	[Old("Deprecated. Framing happens as needed after TileSection is sent.")]
 	public const byte TileFrameSection = 11;
+
+	// Sent from Player.Spawn
+	// If this is the first spawn for a connected client
+	//    RemoteClient state 3->10
+	//    NetMessage.syncPlayers()
+	//    responds with AnglerQuest
+	// Forwarded to other clients
 	public const byte PlayerSpawn = 12;
+
+	/// <summary>
+	/// Syncs the player's movement keystates, item actions, grapple, gravity, stealth and velocity. Primarily sent from sync with clientClone.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The player.whoAmI of the player
+	/// <br/><br/> <b>Forwarded to other clients</b>
+	/// </summary>
 	public const byte PlayerControls = 13;
+
+	// Sets player[i].active
+	// Sent from NetMessage.syncPlayers
 	public const byte PlayerActive = 14;
+
 	[Old("Deprecated.")]
-	public const byte Unknown15 = 15;
+	public const byte Unused15 = 15;
+
+	// Player's current and maximum life
+	// Sent on change, and as part of various syncs
+	// Dead flag is set if life <= 0
+	// Forwarded to other clients
 	public const byte PlayerLifeMana = 16;
+
+	/// <summary>
+	/// Sends changes made to a specific tile coordinate.
+	/// <br/><br/> Use this whenever manipulating tiles using the <see cref="WorldGen"/> methods mentioned below.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The change type. These correspond to methods in <see cref="WorldGen"/>:
+	/// <br/> 0->KillTile, 1->PlaceTile, 2->KillWall, 3->PlaceWall, 4->KillTile (No Item), 5->PlaceWire, 6->KillWire, 7->PoundTile, 8->PlaceActuator, 9->KillActuator, 10->PlaceWire2, 11->KillWire2, 10->PlaceWire3, 13->KillWire3, 14->SlopeTile, 15->Minecart.FrameTrack, 16->PlaceWire4, 17->KillWire4, 18->Wiring.PokeLogicGate, 19->Wiring.Actuate, 20->KillTile (Determine Fail On Server), 21->ReplaceTile, 22->ReplaceWall, 23->SlopeTile+PoundTile
+	/// <br/> <b>number2:</b> x tile coordinate
+	/// <br/> <b>number3:</b> y tile coordinate
+	/// <br/> <b>number4:</b> Changes meaning based on change type:
+	/// <br/> KillTile/KillWall/KillTile (NoItem)/KillTile (Determine Fail On Server)->Fail if 1, PlaceTile->Tile type, PlaceWall->Wall type, SlopeTile->slope value, ReplaceTile->target tile type, ReplaceWall->target wall type, SlopeTile+PoundTile->slope value,
+	/// <br/> <b>number5:</b> Tile style, only used with PlaceTile and ReplaceTile change types
+	/// <br/><br/> <b>Client->Server and Server->Clients. Automatically forwarded to other clients</b>
+	/// </summary>
 	public const byte TileManipulation = 17;
+
+	// Doesn't seem like this is actually sent anywhere
 	public const byte SetTime = 18;
+
+	// ChangeType, x, y, direction
+	// ChangeType WorldGen.{
+	//    OpenDoor = 0, CloseDoor = 1,
+	//    ShiftTrapdoor(onlyCloseOrOpen = 1) = 2, ShiftTrapdoor(onlyCloseOrOpen = 0) = 3,
+	//    ShiftTallGate(closing = false) = 4, ShiftTallGate(closing = true) = 5 }
+	// Forwarded to other clients
 	public const byte ToggleDoorState = 19;
+
+	// TModLoader changes walls from byte to short
+	/// <summary>
+	/// Sends a square area of tile data (with upper left x, y) and given size (minimum 1)
+	/// <br/> Forwarded to other clients
+	/// <br/><br/> This is usually used through the <see cref="NetMessage.SendTileSquare(int, int, int, TileChangeType)"/> helper methods.
+	/// </summary>
-	public const byte Unknown20 = 20;
+	public const byte TileSquare = 20;
+
+	/// <summary>
+	/// Syncs all info about an item in the world, includding modded data.
+	/// <br/><br/> Use this whenever an item in the world is manipulated. <c>Item.NewItem</c> calls this automatically on the server to sync the item with other clients, but on clients this needs to be used or <c>Player.QuickSpawnItem</c> should be used instead.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The index of the item within <see cref="Main.item"/>
+	/// <br/> <b>number2:</b> Set to 1 to ignore the delay for the spawning client to pickup the item. The <c>noGrabDelay</c> parameter of <c>Item.NewItem</c> sets this.
+	/// <br/><br/> <b>Forwarded to other clients</b>
+	/// </summary>
 	public const byte SyncItem = 21;
+
+	// Sets the owner of an item
+	// If set to a player, sets Item.keepTime to 15
+	// Server rejects packets that don't try and set owner to the sender's playerID
+	// Server always sets owner to 255
+	// Forwarded to other clients
 	public const byte ItemOwner = 22;
+
+	/// <summary>
+	/// Sends all info about an NPC, position, velocity, and AI. This includes modded data from <see cref="ModNPC.SendExtraAI(System.IO.BinaryWriter)"/> and <see cref="GlobalNPC.SendExtraAI(NPC, ModLoader.IO.BitWriter, System.IO.BinaryWriter)"/>.
+	/// <br/><br/> Use this when manually spawning an NPC on the server using a <c>NPC.NewNPC</c>.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The index of the spawned NPC within the <see cref="Main.npc"/> array
+	/// <br/><br/> <b>Server->Clients only</b>
+	/// </summary>
 	public const byte SyncNPC = 23;
+
+	// Hit an NPC, presumably unused because it uses the item damage/knockback directly without the random variance
 	public const byte UnusedMeleeStrike = 24;
+
+	// A line of chat text, with sender and color information
 	[Old("Deprecated. Use NetTextModule instead.")]
-	public const byte Unused25 = 25;
+	public const byte ChatText = 25;
+
+	// Damage, direction, pvp, deathText, crit, cooldown
+	// Forwarded to other clients
 	[Old("Deprecated.")]
-	public const byte Unused26 = 26;
+	public const byte HurtPlayer = 26;
+
+	// Sends all info about a projectile, position, velocity, and AI
+	// ProjectileLoader.SendExtraAI appends data to this packet
+	// Forwarded to other clients
 	public const byte SyncProjectile = 27;
+
+	// NpcID, damage, knockback, direction, crit
+	// Forwarded to other clients
 	public const byte DamageNPC = 28;
+
+	// Sent from Projectile.Kill
+	// Forwarded to other clients
 	public const byte KillProjectile = 29;
+
+	// Forwarded to other clients
+	// Forwarding includes the chat message "{player} has {en/dis}abled PvP"
 	public const byte TogglePVP = 30;
+
+	/// <summary>
+	/// Sent from a client to request access to a <see cref="Chest"/>. Clients have to request access to chests for the server to send the current chest contents and to avoid networking bugs that could arise from network lag and multiple clients interacting with the same chest inventory.
+	/// <br/><br/> If not in use by another client, the server will reply with <see cref="SyncChestItem"/> for each item in the chest and then <see cref="SyncPlayerChest"/>.  Other clients are sent the <see cref="SyncPlayerChestIndex"/> message.
+	/// <br/><br/> Modders should use this message exactly as shown in <see href="https://github.com/tModLoader/tModLoader/blob/stable/ExampleMod/Content/Tiles/Furniture/ExampleChest.cs">ExampleChest</see>.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> x tile coordinate of the chest (top left corner)
+	/// <br/> <b>number2:</b> y tile coordinate of the chest (top left corner)
+	/// <br/><br/> <b>Client->Server only</b>
+	/// </summary>
 	public const byte RequestChestOpen = 31;
+
+	// ChestIndex, slot, stack, prefix, itemType
+	// ItemIO.ReceiveModData appends to this packet
+	/// <summary>
+	/// Syncs the item in a specific chest slot.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The index of the chest within the <see cref="Main.chest"/> array
+	/// <br/> <b>number2:</b> The index of the item within that chest's <see cref="Chest.item"/> array
+	/// </summary>
 	public const byte SyncChestItem = 32;
+
+	// Name is only sent if nameLen <= Chest.MaxNameLength
+	// On client, does the open and close sounds (if changed)
+	// On server, sends ChestName and SyncPlayerChestIndex to other clients
+	/// <summary>
+	/// Syncs basic information about a chest. Used on clients to send chest name changes to the server (the server then syncs the name change to other clients via <see cref="ChestName"/>). Used on the server to allow a specific client to open the chest the client previously requested to open via a <see cref="RequestChestOpen"/> message. Other clients are sent the <see cref="SyncPlayerChestIndex"/> message.
+	/// <br/><br/> Modders should use this message exactly as shown in <see href="https://github.com/tModLoader/tModLoader/blob/stable/ExampleMod/Content/Tiles/Furniture/ExampleChest.cs">ExampleChest</see>.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>text:</b> The name of the chest. <c>NetworkText.FromLiteral(Main.chest[chestIndex].name)</c> should be used. Only relevant if <c>number2</c> is set to 1.
+	/// <br/> <b>number:</b> The index of the chest within the <see cref="Main.chest"/> array. This is typically <see cref="Player.chest"/>.
+	/// <br/> <b>number2:</b> If 1, then the <c>text</c> parameter (chest name) is being synced.
+	/// </summary>
 	public const byte SyncPlayerChest = 33;
+
+	// ChangeType, x, y, chestType, [chestIndex]
+	// ChangeType { PlaceChest, KillChest, PlaceDresser, KillDresser, PlaceChest2, KillChest2 }
+	// TMod adds 100 to first 4 changeTypes and sends modType
+	// On server {
+	//    if a place fails, the packet is sent back to the client, and the item dropped
+	//    if a place succeeds the packet is forwarded to all clients with chestIndex = WorldGen.PlaceChest
+	//    if a destroy succeeds, the packet is forwarded to all clients with chestIndex = -1
+	// }
+	// On client {
+	//    a place with chestIndex = -1 calls WorldGen.KillTile
+	//    a place with valid chestIndex calls WorldGen.PlaceChestDirect or PlaceDresserDirect
+	//    a destroy calls Chest.DestroyChestDirect and WorldGen.KillTile
+	// }
 	public const byte ChestUpdates = 34;
+
+	// Restores health
+	// Forwarded to other clients
 	public const byte PlayerHeal = 35;
+
+	// Player.zone1 and zone2, containing the biome information
+	// Forwarded to other clients
+	// Sent on change
 	public const byte SyncPlayerZone = 36;
+
+	// Sent by the server to display the password input menu.
+	// Sets RemoteClient.State = -1
 	public const byte RequestPassword = 37;
+
+	// Checks the supplied password and responds with either KickClient or Login
 	public const byte SendPassword = 38;
+
+	// Sets Item.owner = 255
+	// Sent by the server when the item release count hits 300 (every 5 seconds)
+	// Client responds with ItemOwner
 	public const byte ReleaseItemOwnership = 39;
+
+	// Player.talkNPC, sent on change
+	// Forwarded to other clients
 	public const byte SyncTalkNPC = 40;
+
+	// Player.itemRotation, itemAnimation and channel
+	// Forwarded to other clients
 	public const byte ShotAnimationAndSound = 41;
-	public const byte Unknown42 = 42;
-	public const byte Unknown43 = 43;
+
+	// Player's current and maximum mana
+	// Sent on change, and the major client and server sync
+	public const byte PlayerMana = 42;
+
+	// Restores mana
+	// Forwarded to other clients
+	public const byte ManaEffect = 43;
+
+	// Player.KillMe
+	// Forwarded to other clients
 	[Old("Deprecated.")]
-	public const byte Unknown44 = 44;
-	public const byte Unknown45 = 45;
-	public const byte Unknown46 = 46;
-	public const byte Unknown47 = 47;
+	public const byte KillPlayer = 44;
+
+	// Forwarded to clients
+	// Forwarding includes the chat message "{player} has joined the {team} party"
+	public const byte PlayerTeam = 45;
+
+	// If the server finds a sign at the location, it replies with ReadSign
+	public const byte RequestReadSign = 46;
+
+	// SignIndex, x, y, text, player
+	// If the text changes on the server, the packet is forwarded to other clients
+	// Opens the sign gui and plays the sound on the client
+	public const byte ReadSign = 47;
+
+	// X, y, liquid, liquidType
+	// If NetPlayer.spamCheck will kick the player if the placement position is more than 10 tiles from the player
 	[Old("Deprecated. Use NetLiquidModule instead.")]
 	public const byte LiquidUpdate = 48;
+
+	// Spawns a newly joined player
+	// Client state 6->10
+	// Calls Player.Spawn which sends SpawnPlayer to the server
+	// Sent from RequestTileData once the world is ready for the player to spawn
 	public const byte InitialSpawn = 49;
+
+	// Contents of Player.buffType
+	// On receipt, the buffTime for active buffs is set to 60 ticks (1 second)
+	// BuffTime is only ever decremented on the owner's client, and the server relies on the client resending this packet when buffs are generated or expire
+	// Forwarded to other clients
 	public const byte PlayerBuffs = 50;
+
+	// Param, actionType
+	// ActionType {
+	// 1 - NPC.SpawnSkeletron
+	// 2 - Play grapple sound at the position of Main.player[param]. forwarded to other clients
+	// 3 - Main.Sundialing
+	// 4 - Main.npc[param].BigMimicSpawnSmoke
+	// }
 	public const byte MiscDataSync = 51;
+
+	// Sends TileSquare to all clients
+	/// <summary>
+	/// Unlocks or locks the chest or door at the provided tile coordinates. The server will also sync the changed tiles to all clients.
+	/// <br/><br/> Modders should use this message exactly as shown in <see href="https://github.com/tModLoader/tModLoader/blob/stable/ExampleMod/Content/Tiles/Furniture/ExampleChest.cs">ExampleChest</see>.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number2:</b> UnlockType. 1->Chest.Unlock, 2->WorldGen.UnlockDoor, 3->Chest.Lock
+	/// <br/> <b>number3:</b> x tile coordinate of the chest (top left corner)
+	/// <br/> <b>number4:</b> y tile coordinate of the chest (top left corner)
+	/// <br/><br/> <b>Forwarded to other clients</b>
+	/// </summary>
 	public const byte LockAndUnlock = 52;
+
+	// NpcID, buffType, duration
+	// NPC.AddBuff sends this if netMode is client
+	// Adds a buff to an NPC
+	// SyncNPCBuff is sent to all clients
 	public const byte AddNPCBuff = 53;
+
+	// Sends all buffs on an npc
 	public const byte NPCBuffs = 54;
+
+	// Server does nothing except forward the packet to the buffed player IF the buff is a pvpBuff
+	// Client rejects packets for other players
 	public const byte AddPlayerBuff = 55;
+
+	// NPC.displayName for town NPCs and Skeleton Merchant
+	// Client sends this after SetDefaults (npcNameLookup = 0) 'help I don't know what their name is'
+	// Server replies with name
 	public const byte UniqueTownNPCInfoSyncRequest = 56;
+
+	// Hallowed/corrupted/crimson tile counts (out of 100)
-	public const byte Unknown57 = 57;
+	public const byte TileCounts = 57;
+
+	// Main.harpNote = pitch
+	// Plays harp or bell sound based on held item at player position
+	// Forwarded to other clients
 	public const byte InstrumentSound = 58;
+
+	// Wiring.HitSwitch(x, y)
+	// Forwarded to other clients
+	/// <summary>
+	/// Calls <see cref="Wiring.HitSwitch(int, int)"/> at the provided tile coordinates. This is used to sync effects from tile interactions, like playing sounds.
+	/// <br/><br/> Modders should use this message whenever they call <see cref="Wiring.HitSwitch(int, int)"/> to sync the effects.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> x tile coordinate to use with Wiring.HitSwitch
+	/// <br/> <b>number2:</b> y tile coordinate to use with Wiring.HitSwitch
+	/// <br/><br/> <b>Forwarded to other clients</b>
+	/// </summary>
 	public const byte HitSwitch = 59;
+
+	// NpcID, homeX, homeY, changeType
+	// ChangeType WorldGen.{ kickOut = 0, moveRoom = 1 }
+	// Sent by client on WorldGen.kickOut/moveRoom, propogated by server in UpdateNPC
-	public const byte Unknown60 = 60;
+	public const byte NPCHome = 60;
+
+	/// <summary>
+	/// Attempts to spawn an NPC on the player, start a specific event, or use a pet license. Spawned NPC must be in <see cref="NPCID.Sets.MPAllowedEnemies"/>, this will not allow multiple to spawn.
+	/// <br/><br/> Mods should use this in boss spawner items to spawn a boss.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> The player.whoAmI of the player to spawn on.
+	/// <br/> <b>number2:</b> If positive, this is the NPC type to spawn. If negative, this is a special event value from one of the following:
+	/// GoblinArmy: -1, FrostLegion: -2, Pirates: -3, PumpkinMoon: -4, SnowMoon: -5, Eclipse: -6, MartianInvation: -7, MoonLordTimer: -8, BloodMoon: -10, <see cref="NPC.combatBookWasUsed"/>: -11, Pet License: -12 through -15, MechQueen: 16, <see cref="NPC.combatBookVolumeTwoWasUsed"/>: -17, <see cref="NPC.peddlersSatchelWasUsed"/>: -18,
+	/// <br/><br/> <b>Client->Server only</b>
+	/// </summary>
 	public const byte SpawnBossUseLicenseStartEvent = 61;
-	public const byte Unknown62 = 62;
-	public const byte Unknown63 = 63;
-	public const byte Unknown64 = 64;
+
+	// Player, dodgeType { NinjaDodge = 1, ShadowDodge = 2 }
+	// Forwarded to other clients
+	public const byte Dodge = 62;
+
+	// WorldGen.paintTile(x, y, color)
+	// Forwarded to other clients
+	public const byte PaintTile = 63;
+
+	// WorldGen.paintWall(x, y, color)
+	// Forwarded to other clients
+	public const byte PaintWall = 64;
+
+	// If PlayerToPlayer, the nearest player in 9999 units is found, and a chat message sent "{player} has teleported to {player2}"
+	// Index is set to whoAmI on server, so NPC teleports cannot be sent from client
+	// Player teleports are forwarded to other clients
+	/// <summary>
+	/// Teleports the player (or an NPC) to a provided world coordinate. (Sets position, not Center)
+	/// <br/><br/> When using any of these teleport methods in multiplayer, make sure to use this instead of calling the methods directly on the client.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> teleportType: 0->Player, 1->NPC, 2->PlayerToPlayer
+	/// <br/> <b>number2:</b> The player.whoAmI of the player to teleport (or NPC.whoAmI for NPC teleports)
+	/// <br/> <b>number3:</b> The destination x world coordinate
+	/// <br/> <b>number4:</b> The destination y world coordinate
+	/// <br/> <b>number5:</b> The <c>Style</c> parameter of <see cref="Player.Teleport(Microsoft.Xna.Framework.Vector2, int, int)"/>, controls the visuals and sounds of the teleport. Equivalent to <see cref="TeleportationStyleID"/> values.
+	/// <br/> <b>number6:</b> Unknown
+	/// <br/> <b>number7:</b> The <c>extraInfo</c> parameter of <see cref="Player.Teleport(Microsoft.Xna.Framework.Vector2, int, int)"/>, only relevant with pylon teleporting.
+	/// <br/><br/> Using this message for long range player teleports won't work correctly because the client might not have the destination section loaded. You might need to use a <see cref="ModPacket"/> to call <see cref="RemoteClient.CheckSection(int, Microsoft.Xna.Framework.Vector2, int)"/> on the server and then call Player.Teleport and relay this message manually.
+	/// <br/><br/> <b>Client->Server for player teleports, Server->Clients for NPC teleports.</b> The player teleport message is relayed to other clients to sync the visuals.
+	/// </summary>
 	public const byte TeleportEntity = 65;
+
-	public const byte Unknown66 = 66;
+	public const byte SpiritHeal = 66;
+
 	public const byte Unknown67 = 67;
+
+	// Client sends this in response to LoadPlayer, server ignores it
-	public const byte Unknown68 = 68;
+	public const byte ClientUUID = 68;
+
+	// ChestIndex, x, y, name
+	// Forwarded to other clients
 	public const byte ChestName = 69;
+
 	public const byte BugCatching = 70;
+
 	public const byte BugReleasing = 71;
+
 	public const byte TravelMerchantItems = 72;
+
+	/// <summary>
+	/// Teleports the client to set locations using one of 4 teleport methods. Use <see cref="TeleportEntity"/> instead to teleport to arbitrary coordinates.
+	/// <br/><br/> When using any of these teleport methods in multiplayer, make sure to use this instead of calling the methods directly on the client.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number: </b> The teleportation method to call:
+	/// <br/> 0-><see cref="Player.TeleportationPotion"/> (Random Location), 1-><see cref="Player.MagicConch"/> (Ocean), 2-><see cref="Player.DemonConch"/> (Underworld), 3-><see cref="Player.Shellphone_Spawn"/> (World spawn point).
+	/// <br/><br/> <b>Client->Server only</b>
+	/// </summary>
 	public const byte RequestTeleportationByServer = 73;
+
 	public const byte AnglerQuest = 74;
+
 	public const byte AnglerQuestFinished = 75;
+
 	public const byte QuestsCountSync = 76;
+
+	/// <summary>
+	/// Sends a temporary tile animation created in <see cref="Animation.NewTemporaryAnimation"/> on the server to clients.
+	/// <br/><br/> No need to manually use this message.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/> <b>number:</b> Temporary tile animation ID (<see cref="Animation.RegisterTemporaryAnimation(int, int[])"/>)
+	/// <br/> <b>number2:</b> Tile Type
+	/// <br/> <b>number3:</b> x tile coordinate
+	/// <br/> <b>number4:</b> y tile coordinate
+	/// </summary>
 	public const byte TemporaryAnimation = 77;
+
 	public const byte InvasionProgressReport = 78;
+
 	public const byte PlaceObject = 79;
+
 	public const byte SyncPlayerChestIndex = 80;
+	/// <summary>
+	/// Sends a <see cref="CombatText"/> from the server to clients. Sends a number, use <see cref="CombatTextString"/> instead for the string version. Used automatically by <see cref="Player.HealEffect(int, bool)"/>, but custom <see cref="CombatText"/> would need to manually sync using this message.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters, correlating to <see cref="CombatText.NewText(Microsoft.Xna.Framework.Rectangle, Microsoft.Xna.Framework.Color, int, bool, bool)"/> parameters:
+	/// <br/><b>number:</b> The text color: <see cref="Microsoft.Xna.Framework.Color.PackedValue"/>
+	/// <br/><b>number2:</b> The CombatText center X coordinate
+	/// <br/><b>number3:</b> The CombatText center Y coordinate
+	/// <br/><b>number4:</b> The number shown by the CombatText
+	/// </summary>
 	public const byte CombatTextInt = 81;
+
+	// Used by NetModule, NetMaager and NetPacket for a more extensible packet system.
+	// Sent by NetManager.SendData
+	// Handled by NetManager.Read
+	// Currently the only module (id 0) is the NetLiquidModule
 	public const byte NetModules = 82;
+
+	// BannerID, killCount
+	// Supports modded banners now that check for bannerID > Main.MaxBannerTypes removed
+	// Sent to all clients whenever an npc with a banner is killed
 	public const byte NPCKillCountDeathTally = 83;
+
 	public const byte PlayerStealth = 84;
+
+	// Requests chest quick-stacking of a single item
+	// The client Player.inventoryChestStack flag is set
+	// The server performs the quickstack and sends a SyncEquipment packet to all players
 	public const byte QuickStackChests = 85;
+
+	/// <summary>
+	/// Syncs the data of a TileEntity from server to clients. Will cause TileEntities to send and receive their data to sync their values. (<see cref="ModLoader.ModTileEntity.NetSend(System.IO.BinaryWriter)"/>, <see cref="ModLoader.ModTileEntity.NetReceive(System.IO.BinaryReader)"/>)
+	/// <para/> <c>NetMessage.SendData</c> parameters:
+	/// <br/><b>number:</b> The <see cref="DataStructures.TileEntity.ID"/>
+	/// <br/><br/> <b>Server->Clients only</b>
+	/// </summary>
 	public const byte TileEntitySharing = 86;
+
+	/// <summary>
+	/// Syncs a TileEntity placement attempt from a client to the server. Will cause the server to place the TileEntity and then sync it using <see cref="TileEntitySharing"/>.
+	/// <br/><br/> This is typically used in the method contained in the <see cref="ObjectData.TileObjectData.HookPostPlaceMyPlayer"/> of the Tile placing the Tile Entity, but modders can use <see cref="ModLoader.ModTileEntity.Generic_HookPostPlaceMyPlayer"/> instead of writing custom netcode using this message.
+	/// <br/><br/> <c>NetMessage.SendData</c> parameters:
+	/// <br/><b>number:</b> The x tile coordinate of the TileEntity, not necessarily the top left corner of the multitile.
+	/// <br/><b>number2:</b> The y tile coordinate of the TileEntity, not necessarily the top left corner of the multitile.
+	/// <br/><b>number3:</b> The <see cref="TileEntityID"/> or <see cref="ModLoader.ModTileEntity.Type"/>
+	/// <br/><br/> <b>Client->Server only</b>
+	/// </summary>
 	public const byte TileEntityPlacement = 87;
+
 	public const byte ItemTweaker = 88;
+
 	public const byte ItemFrameTryPlacing = 89;
+
+	/// <summary>
+	/// Exactly the same as <see cref="SyncItem"/>, but sets the instanced flag and is only sent server -> client. Used by loot that is dropped per player, like boss bags.
+	/// <br/><br/> <see cref="SyncItem"/> docs:
+	/// <br/><br/> <inheritdoc cref="SyncItem"/>
+	/// </summary>
-	public const byte InstancedItem = 90;
+	public const byte InstancedItem = 90
+		;
 	public const byte SyncEmoteBubble = 91;
+
 	public const byte SyncExtraValue = 92;
+
 	public const byte SocialHandshake = 93;
+
 	public const byte Deprecated1 = 94;
+
 	public const byte MurderSomeoneElsesPortal = 95;
+
 	public const byte TeleportPlayerThroughPortal = 96;
+
 	public const byte AchievementMessageNPCKilled = 97;
+
 	public const byte AchievementMessageEventHappened = 98;
+
 	public const byte MinionRestTargetUpdate = 99;
+
 	public const byte TeleportNPCThroughPortal = 100;
+
 	public const byte UpdateTowerShieldStrengths = 101;
+
 	public const byte NebulaLevelupRequest = 102;
+
 	public const byte MoonlordHorror = 103;
+
 	public const byte ShopOverride = 104;
+
 	public const byte GemLockToggle = 105;
+
 	public const byte PoofOfSmoke = 106;
+
 	public const byte SmartTextMessage = 107;
+
 	public const byte WiredCannonShot = 108;
+
 	public const byte MassWireOperation = 109;
+
 	public const byte MassWireOperationPay = 110;
+
 	public const byte ToggleParty = 111;
+
 	public const byte SpecialFX = 112;
+
 	public const byte CrystalInvasionStart = 113;
+
 	public const byte CrystalInvasionWipeAllTheThingsss = 114;
+
 	public const byte MinionAttackTargetUpdate = 115;
+
 	public const byte CrystalInvasionSendWaitTime = 116;
+
 	public const byte PlayerHurtV2 = 117;
+
 	public const byte PlayerDeathV2 = 118;
+
+	/// <summary>
+	/// Sends a <see cref="CombatText"/> from the server to clients. Sends a string, use <see cref="CombatTextInt"/> instead for the int version. Used automatically by <see cref="Player.HealEffect(int, bool)"/>, but custom <see cref="CombatText"/> would need to manually sync using this message.
+	/// <para/> <c>NetMessage.SendData</c> parameters, correlating to <see cref="CombatText.NewText(Microsoft.Xna.Framework.Rectangle, Microsoft.Xna.Framework.Color, string, bool, bool)"/> parameters:
+	/// <br/><b>number:</b> The text color: <see cref="Microsoft.Xna.Framework.Color.PackedValue"/>
+	/// <br/><b>number2:</b> The CombatText center X coordinate
+	/// <br/><b>number3:</b> The CombatText center Y coordinate
+	/// <br/><b>text:</b> The text shown by the CombatText
+	/// </summary>
 	public const byte CombatTextString = 119;
+
 	public const byte Emoji = 120;
+
 	public const byte TEDisplayDollItemSync = 121;
+
 	public const byte RequestTileEntityInteraction = 122;
+
 	public const byte WeaponsRackTryPlacing = 123;
+
 	public const byte TEHatRackItemSync = 124;
+
 	public const byte SyncTilePicking = 125;
+
 	public const byte SyncRevengeMarker = 126;
+
 	public const byte RemoveRevengeMarker = 127;
+
 	public const byte LandGolfBallInCup = 128;
+
 	public const byte FinishedConnectingToServer = 129;
+
 	public const byte FishOutNPC = 130;
+
 	public const byte TamperWithNPC = 131;
+
 	public const byte PlayLegacySound = 132;
+
 	public const byte FoodPlatterTryPlacing = 133;
+
 	public const byte UpdatePlayerLuckFactors = 134;
+
 	public const byte DeadPlayer = 135;
+
 	public const byte SyncCavernMonsterType = 136;
+
 	public const byte RequestNPCBuffRemoval = 137;
+
 	public const byte ClientSyncedInventory = 138;
+
 	public const byte SetCountsAsHostForGameplay = 139;
+
 	public const byte SetMiscEventValues = 140;
+
 	public const byte RequestLucyPopup = 141;
+
 	public const byte SyncProjectileTrackers = 142;
+
 	public const byte CrystalInvasionRequestedToSkipWaitTime = 143;
+
 	public const byte RequestQuestEffect = 144;
+
+	/// <summary>
+	/// Exactly the same as <see cref="SyncItem"/>, but also syncs Item.shimmered and Item.shimmerTime. Used when an item is transformed in shimmer.
+	/// <br/><br/> <see cref="SyncItem"/> docs:
+	/// <br/><br/> <inheritdoc cref="SyncItem"/>
+	/// </summary>
 	public const byte SyncItemsWithShimmer = 145;
+
 	public const byte ShimmerActions = 146;
+
 	public const byte SyncLoadout = 147;
+
+	/// <summary>
+	/// Exactly the same as <see cref="SyncItem"/>, but also syncs Item.timeLeftInWhichTheItemCannotBeTakenByEnemies. Used by the Lucky Coin OnHit effect.
+	/// <br/><br/> <see cref="SyncItem"/> docs:
+	/// <br/><br/> <inheritdoc cref="SyncItem"/>
+	/// </summary>
 	public const byte SyncItemCannotBeTakenByEnemies = 148;
+
 	public static readonly byte Count = 149;
 }
