--- src/TerrariaNetCore/Terraria/ID/MountID.cs
+++ src/tModLoader/Terraria/ID/MountID.cs
@@ -1,13 +_,30 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID;
 
-public static class MountID
+public class MountID
 {
+#if !TMLANALYZER
 	public static class Sets
 	{
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(MountLoader.MountCount, nameof(MountID), Search);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given <see cref="MountID"/>, then that mount is categorized as a minecart.
+		/// <br/> Minecarts have a dedicated equipment slot, will be auto-used if the player interacts with a <see cref="TileID.MinecartTrack"/>, and are buffed if <see cref="Player.UsingSuperCart"/> is <see langword="true"/>.
+		/// <br/> Defaults to <see langword="false"/>.
+		/// </summary>
 		public static bool[] Cart = Factory.CreateBoolSet(6, 11, 13, 15, 16, 18, 19, 20, 21, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 51, 53);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given <see cref="MountID"/>, then that mount will face the player's velocity instead of their <see cref="Entity.direction"/>.
+		/// <br/> Vanilla uses this set for many minecarts.
+		/// <br/> Defaults to <see langword="false"/>.
+		/// </summary>
 		public static bool[] FacePlayersVelocity = Factory.CreateBoolSet(15, 16, 11, 18, 19, 20, 21, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 51, 53);
 	}
+#endif
 
 	public const int None = -1;
 	public const int Rudolph = 0;
@@ -65,4 +_,6 @@
 	public const int Wolf = 52;
 	public const int TerraFartMinecart = 53;
 	public static int Count = 54;
+
+	public static IdDictionary Search = IdDictionary.Create<MountID, int>();
 }
