--- src/TerrariaNetCore/Terraria/ID/MountID.cs
+++ src/tModLoader/Terraria/ID/MountID.cs
@@ -1,19 +_,31 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID;
 
-public static class MountID
+public class MountID
 {
+#if !TMLANALYZER
 	public static class Sets
 	{
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(MountLoader.MountCount, nameof(MountID), Search);
+
 		public static bool[] CanUseHooks = Factory.CreateBoolSet(54, 57, 58, 59, 60);
 		public static bool[] CanDash = Factory.CreateBoolSet(56, 57, 58, 59, 60, 61, 62, 63);
 		public static bool[] DoesNotOverrideBodyFrames = Factory.CreateBoolSet(57, 58, 59, 60);
 		public static bool[] DoesNotOverrideLegFrames = Factory.CreateBoolSet(57, 58, 59, 60);
 		public static bool[] DoesNotOverrideBackpackDraw = Factory.CreateBoolSet(57, 58, 59, 60);
 		public static bool[] IsRollerSkates = Factory.CreateBoolSet(57, 58, 59, 60);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given <see cref="MountID"/>, then that mount is categorized as a minecart.
+		/// <br/> Minecarts have a dedicated equipment slot, will be auto-used if the player interacts with a <see cref="TileID.MinecartTrack"/>, and are buffed if <see cref="Player.UsingSuperCart"/> is <see langword="true"/>.
+		/// <br/> Defaults to <see langword="false"/>.
+		/// </summary>
 		public static bool[] Cart = Factory.CreateBoolSet(6, 13, 11, 15, 16, 18, 19, 20, 21, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 38, 39, 51, 53);
 		public static bool[] IsTransformationMount = Factory.CreateBoolSet(52, 54, 55, 56, 61);
 		public static bool[] PlayerIsHidden = Factory.CreateBoolSet(52, 54, 55, 56, 61);
 		public static bool[] DontHoldItems = Factory.CreateBoolSet(55, 56, 61);
 		public static bool[] DontDismountWhenCCed = Factory.CreateBoolSet(55, 56, 61);
 	}
+#endif
@@ -84,4 +_,6 @@
 	public const int Chillet = 62;
 	public const int ChilletIgnis = 63;
 	public static int Count = 64;
+
+	public static IdDictionary Search = IdDictionary.Create<MountID, int>();
 }
