--- src/TerrariaNetCore/Terraria/ID/MusicID.cs
+++ src/tModLoader/Terraria/ID/MusicID.cs
@@ -1,11 +_,20 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID;
 
-public static class MusicID
+public class MusicID
 {
 	public static class Sets
 	{
-		public static SetFactory Factory = new SetFactory(105);
+		public static SetFactory Factory = new SetFactory(MusicLoader.MusicCount, nameof(MusicID), Search);
 		public static bool[] CanBeRecorded = Factory.CreateBoolSet(true, 0, 28, 45, 102);
+
+		/// <summary>
+		/// Skips Terraria's <see cref="Terraria.Audio.ASoundEffectBasedAudioTrack.ReMapVolumeToMatchXact(float)"/> function to make music play at its intended volume.
+		/// <para/>This should be set in <see cref="ModType.SetStaticDefaults()"/>, preferably <see cref="ModSystem"/>
+		/// </summary>
+		public static bool[] SkipsVolumeRemap = Factory.CreateBoolSet(false);
 	}
 
 	public const int None = 0;
@@ -114,4 +_,6 @@
 	public const int RainbowBoulderLoop = 103;
 	public const int Skeletron = 104;
 	public const int Count = 105;
+
+	public static IdDictionary Search = IdDictionary.Create<MusicID, int>();
 }
