--- src/TerrariaNetCore/Terraria/ID/NPCHeadID.cs
+++ src/tModLoader/Terraria/ID/NPCHeadID.cs
@@ -1,10 +_,21 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID;
 
 public class NPCHeadID
 {
 	public static class Sets
 	{
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(NPCHeadLoader.NPCHeadCount, nameof(NPCHeadID), Search);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given NPC head ID, then that NPC head will never be drawn in the housing UI.
+		/// </summary>
+		/// <remarks>
+		/// <see cref="ModNPC"/>s with <see cref="ModNPC.TownNPCStayingHomeless"/> set to <see langword="true"/> will also not draw their heads in the housing UI.
+		/// <br/> You can get an NPC's head ID using <see cref="GameContent.TownNPCProfiles.GetHeadIndexSafe(NPC)"/> and <see cref="ContentSamples.NpcsByNetId"/>.
+		/// </remarks>
 		public static bool[] CannotBeDrawnInHousingUI = Factory.CreateBoolSet(21, 80);
 	}
 
@@ -90,4 +_,6 @@
 	public const int TavernkeepShimmered = 79;
 	public const int TravelingMerchantShimmered = 80;
 	public static readonly int Count = 81;
+
+	public static IdDictionary Search = IdDictionary.Create<NPCHeadID, int>();
 }
