--- src/Terraria/Terraria/ID/NPCID.cs
+++ src/tModLoader/Terraria/ID/NPCID.cs
@@ -3,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using Terraria.DataStructures;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -61,7 +_,7 @@
 				public const int BoneJavelin = 169;
 			}
 
-			public static SetFactory Factory = new SetFactory(665);
+			public static SetFactory Factory = new SetFactory(NPCLoader.NPCCount);
 			public static Dictionary<int, int> SpecialSpawningRules = new Dictionary<int, int> {
 				{ 259, 0 },
 				{ 260, 0 },
@@ -4104,7 +_,7 @@
 
 				dictionary.Add(6, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_7",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_7", // TODO: This and all subsequent filepath patches are temporary until we overhaul the asset system. Very hacky.
 					Position = new Vector2(20f, 29f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 10f
@@ -4122,7 +_,7 @@
 
 				dictionary.Add(9, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_10",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_10",
 					Position = new Vector2(2f, 24f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 10f
@@ -4140,7 +_,7 @@
 
 				dictionary.Add(12, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_13",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_13",
 					Position = new Vector2(40f, 22f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 10f
@@ -4253,7 +_,7 @@
 
 				dictionary.Add(37, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_39",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_39",
 					Position = new Vector2(40f, 23f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 10f
@@ -4473,7 +_,7 @@
 
 				dictionary.Add(86, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_87",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_87",
 					Position = new Vector2(55f, 15f),
 					PortraitPositionXOverride = 4f,
 					PortraitPositionYOverride = 10f
@@ -4518,7 +_,7 @@
 
 				dictionary.Add(94, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_95",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_95",
 					Position = new Vector2(20f, 28f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 10f
@@ -4536,7 +_,7 @@
 
 				dictionary.Add(97, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_98",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_98",
 					Position = new Vector2(40f, 24f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 12f
@@ -4614,7 +_,7 @@
 
 				dictionary.Add(112, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_113",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_113",
 					Position = new Vector2(56f, 5f),
 					PortraitPositionXOverride = 10f,
 					PortraitPositionYOverride = 0f
@@ -4627,7 +_,7 @@
 
 				dictionary.Add(114, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_115",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_115",
 					Position = new Vector2(56f, 3f),
 					PortraitPositionXOverride = 55f,
 					PortraitPositionYOverride = 0f
@@ -4642,7 +_,7 @@
 
 				dictionary.Add(116, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_117",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_117",
 					Position = new Vector2(10f, 20f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 0f
@@ -4699,7 +_,7 @@
 
 				dictionary.Add(126, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_127",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_127",
 					Position = new Vector2(0f, 0f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 1f
@@ -4746,7 +_,7 @@
 
 				dictionary.Add(133, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_134",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_134",
 					Position = new Vector2(60f, 8f),
 					PortraitPositionXOverride = 3f,
 					PortraitPositionYOverride = 0f
@@ -5240,7 +_,7 @@
 
 				dictionary.Add(243, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_245",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_245",
 					Position = new Vector2(2f, 48f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 24f
@@ -5316,7 +_,7 @@
 
 				dictionary.Add(258, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_259",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_259",
 					Position = new Vector2(0f, 25f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 8f
@@ -5324,7 +_,7 @@
 
 				dictionary.Add(259, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_260",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_260",
 					Position = new Vector2(0f, 25f),
 					PortraitPositionXOverride = 1f,
 					PortraitPositionYOverride = 4f
@@ -5972,7 +_,7 @@
 
 				dictionary.Add(392, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_395",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_395",
 					Position = new Vector2(-1f, 18f),
 					PortraitPositionXOverride = 1f,
 					PortraitPositionYOverride = 0f
@@ -6000,7 +_,7 @@
 
 				dictionary.Add(397, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_398",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_398",
 					Position = new Vector2(0f, 5f),
 					Scale = 0.4f,
 					PortraitScale = 0.7f,
@@ -6019,7 +_,7 @@
 
 				dictionary.Add(401, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_402",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_402",
 					Position = new Vector2(42f, 15f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 0f
@@ -6054,7 +_,7 @@
 
 				dictionary.Add(411, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_412",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_412",
 					Position = new Vector2(50f, 28f),
 					PortraitPositionXOverride = 0f,
 					PortraitPositionYOverride = 4f
@@ -6274,7 +_,7 @@
 
 				dictionary.Add(453, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_454",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_454",
 					Position = new Vector2(57f, 10f),
 					PortraitPositionXOverride = 5f,
 					PortraitPositionYOverride = 0f
@@ -6441,7 +_,7 @@
 
 				dictionary.Add(489, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_491",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_491",
 					Position = new Vector2(30f, -5f),
 					Scale = 0.8f,
 					PortraitPositionXOverride = 1f,
@@ -6551,7 +_,7 @@
 
 				dictionary.Add(509, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_510",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_510",
 					Position = new Vector2(55f, 18f),
 					PortraitPositionXOverride = 10f,
 					PortraitPositionYOverride = 12f
@@ -6569,7 +_,7 @@
 
 				dictionary.Add(511, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_513",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_513",
 					Position = new Vector2(37f, 24f),
 					PortraitPositionXOverride = 10f,
 					PortraitPositionYOverride = 17f
@@ -7001,7 +_,7 @@
 
 				dictionary.Add(593, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_594",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_594",
 					Scale = 0.8f
 				};
 
@@ -7117,7 +_,7 @@
 
 				dictionary.Add(620, value);
 				value = new NPCBestiaryDrawModifiers(0) {
-					CustomTexturePath = "Images/UI/Bestiary/NPCs/NPC_621",
+					CustomTexturePath = "Terraria/UI/Bestiary/NPCs/NPC_621",
 					Position = new Vector2(46f, 20f),
 					PortraitPositionXOverride = 10f,
 					PortraitPositionYOverride = 17f
