--- src/Terraria/Terraria/ID/NPCID.cs
+++ src/tModLoader/Terraria/ID/NPCID.cs
@@ -3,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using Terraria.DataStructures;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -61,7 +_,7 @@
 				public const int BoneJavelin = 169;
 			}
 
-			public static SetFactory Factory = new SetFactory(668);
+			public static SetFactory Factory = new SetFactory(NPCLoader.NPCCount);
 			public static Dictionary<int, int> SpecialSpawningRules = new Dictionary<int, int> {
 				{ 259, 0 },
 				{ 260, 0 },
@@ -3476,16 +_,26 @@
 				Dictionary<int, NPCBestiaryDrawModifiers> leinforsEntries = GetLeinforsEntries();
 				Dictionary<int, NPCBestiaryDrawModifiers> groxEntries = GetGroxEntries();
 				Dictionary<int, NPCBestiaryDrawModifiers> dictionary = new Dictionary<int, NPCBestiaryDrawModifiers>();
+
+				//Because we can't modify the dictionary while enumerating it.
+				NPCBestiaryDrawModifiers FilterPaths(NPCBestiaryDrawModifiers modifiers) {
+					if (modifiers.CustomTexturePath != null) {
+						modifiers.CustomTexturePath = $"Terraria/{modifiers.CustomTexturePath}";
+					}
+
+					return modifiers;
+				}
+
 				foreach (KeyValuePair<int, NPCBestiaryDrawModifiers> item in groxEntries) {
-					dictionary[item.Key] = item.Value;
+					dictionary[item.Key] = FilterPaths(item.Value);
 				}
 
 				foreach (KeyValuePair<int, NPCBestiaryDrawModifiers> item2 in leinforsEntries) {
-					dictionary[item2.Key] = item2.Value;
+					dictionary[item2.Key] = FilterPaths(item2.Value);
 				}
 
 				foreach (KeyValuePair<int, NPCBestiaryDrawModifiers> item3 in redigitEntries) {
-					dictionary[item3.Key] = item3.Value;
+					dictionary[item3.Key] = FilterPaths(item3.Value);
 				}
 
 				return dictionary;
