--- src/TerrariaNetCore/Terraria/ID/PlayerVariantID.cs
+++ src/tModLoader/Terraria/ID/PlayerVariantID.cs
@@ -1,12 +_,27 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-public static class PlayerVariantID
+public class PlayerVariantID
 {
 	public class Sets
 	{
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(Count, nameof(PlayerVariantID), Search);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given skin variant (<see cref="Player.skinVariant"/>), then that variant is for a male (<see cref="Player.Male"/>) player.
+		/// <br/> Defaults to <see langword="false"/>.
+		/// </summary>
 		public static bool[] Male = Factory.CreateBoolSet(0, 1, 2, 3, 8, 10);
+
+		/// <summary>
+		/// Links a skin variant (<see cref="Player.skinVariant"/>) to the corresponding skin variant of the opposite gender.
+		/// </summary>
 		public static int[] AltGenderReference = Factory.CreateIntSet(0, 0, 4, 4, 0, 1, 5, 5, 1, 2, 6, 6, 2, 3, 7, 7, 3, 8, 9, 9, 8, 10, 11, 11, 10);
+
+		/// <summary>
+		/// The order of skin variants (<see cref="Player.skinVariant"/>) for male players.
+		/// </summary>
 		public static int[] VariantOrderMale = new int[6] {
 			0,
 			1,
@@ -15,6 +_,10 @@
 			8,
 			10
 		};
+
+		/// <summary>
+		/// The order of skin variants (<see cref="Player.skinVariant"/>) for female players.
+		/// </summary>
 		public static int[] VariantOrderFemale = new int[6] {
 			4,
 			5,
@@ -38,4 +_,6 @@
 	public const int MaleDisplayDoll = 10;
 	public const int FemaleDisplayDoll = 11;
 	public static readonly int Count = 12;
+
+	public static IdDictionary Search = IdDictionary.Create<PlayerVariantID, int>();
 }
