--- src/TerrariaNetCore/Terraria/ID/PlayerTextureID.cs
+++ src/tModLoader/Terraria/ID/PlayerTextureID.cs
@@ -1,6 +_,8 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-public static class PlayerTextureID
+public class PlayerTextureID
 {
 	public const int Head = 0;
 	public const int EyeWhites = 1;
@@ -19,4 +_,5 @@
 	public const int Extra = 14;
 	public const int EyeBlink = 15;
 	public static readonly int Count = 16;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(PlayerTextureID), typeof(int)); // TML
 }
