--- src/TerrariaNetCore/Terraria/ID/PlayerDifficultyID.cs
+++ src/tModLoader/Terraria/ID/PlayerDifficultyID.cs
@@ -1,9 +_,15 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
+#if TMLCODEASSIST
+[tModCodeAssist.IDType.Sets.AssociatedName(ModLoader.Annotations.IDTypeAttribute.PlayerDifficulty)]
+#endif
 public class PlayerDifficultyID
 {
 	public const byte SoftCore = 0;
 	public const byte MediumCore = 1;
 	public const byte Hardcore = 2;
 	public const byte Creative = 3;
+	public static readonly IdDictionary Search = IdDictionary.Create<PlayerDifficultyID, byte>(); // TML
 }
