--- src/TerrariaNetCore/Terraria/ID/PrefixID.cs
+++ src/tModLoader/Terraria/ID/PrefixID.cs
@@ -1,10 +_,18 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID;
 
 public class PrefixID
 {
 	public static class Sets
 	{
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(PrefixLoader.PrefixCount, nameof(PrefixID), Search);
+
+		/// <summary>
+		/// If <see langword="true"/> for a given prefix type (<see cref="Item.prefix"/>), then that prefix has a 66% to not be applied when randomly rolled (using <c>-1</c>) when calling <see cref="Item.Prefix(int)"/>.
+		/// <br/> Defaults to <see langword="false"/>.
+		/// </summary>
 		public static bool[] ReducedNaturalChance = Factory.CreateBoolSet(7, 8, 9, 10, 11, 22, 23, 24, 29, 30, 31, 39, 40, 56, 41, 47, 48, 49);
 	}
 
@@ -93,4 +_,7 @@
 	public const int Mythical = 83;
 	public const int Legendary2 = 84;
 	public static readonly int Count = 85;
+
+	// Added by TML.
+	public static readonly IdDictionary Search = IdDictionary.Create<PrefixID, int>();
 }
