--- src/TerrariaNetCore/Terraria/ID/ProjectileSourceID.cs
+++ src/tModLoader/Terraria/ID/ProjectileSourceID.cs
@@ -1,6 +_,9 @@
+using System;
+
 namespace Terraria.ID
 {
+	//TML: Internalized. See EntitySourceID.cs
-	public static class ProjectileSourceID
+	internal static class ProjectileSourceID
 	{
 		public const int None = 0;
 		public const int SetBonus_SolarExplosion_WhenTakingDamage = 1;
@@ -21,5 +_,19 @@
 		public const int SetBonus_GhostHurt = 16;
 		public const int VampireKnives = 18;
 		public const int Count = 19;
+
+		// Reminder: this is temporary.
+		public static int ToEntitySourceID(int projectileSourceID) => projectileSourceID switch {
+			WeaponEnchantment_Confetti => EntitySourceID.WeaponEnchantment_Confetti,
+			PlayerDeath_TombStone => EntitySourceID.Death_TombStone,
+			TorchGod => EntitySourceID.TorchGod,
+			FallingStar => EntitySourceID.FallingStar,
+			PlayerHurt_DropFootball => EntitySourceID.PlayerHurt_DropFootball,
+			StormTigerTierSwap => EntitySourceID.StormTigerTierSwap,
+			AbigailTierSwap => EntitySourceID.AbigailTierSwap,
+			VampireKnives => EntitySourceID.VampireKnives,
+			None => EntitySourceID.None,
+			_ => throw new ArgumentException($"Invalid value: '{projectileSourceID}'.")
+		};
 	}
 }
