--- src/TerrariaNetCore/Terraria/ID/RecipeGroupID.cs
+++ src/tModLoader/Terraria/ID/RecipeGroupID.cs
@@ -15,5 +_,10 @@
 		public static int Dragonflies = 10;
 		public static int Turtles = 11;
 		public static int Fruit = 12;
+		public static int Wood = 13;
+		public static int IronBar = 14;
+		public static int PressurePlate = 15;
+		public static int Sand = 16;
+		public static int Fragment = 17;
 	}
 }
