--- src/TerrariaNetCore/Terraria/ID/RecipeGroupID.cs
+++ src/tModLoader/Terraria/ID/RecipeGroupID.cs
@@ -1,5 +_,9 @@
 namespace Terraria.ID;
 
+/// <summary>
+/// Contains recipe group ids suitable for use in <see cref="Recipe.AddRecipeGroup(int, int)"/>.
+/// <br/> Note that FishForDinner and GoldenCritter are actually unused and won't work. Also note that the ID values shown are incorrect.
+/// </summary>
 public class RecipeGroupID
 {
 	public static int Birds = 0;
@@ -10,7 +_,9 @@
 	public static int Butterflies = 5;
 	public static int Fireflies = 6;
 	public static int Snails = 7;
+	// Unused.
 	public static int FishForDinner = 8;
+	// Unused.
 	public static int GoldenCritter = 9;
 	public static int Dragonflies = 10;
 	public static int Turtles = 11;
