--- src/TerrariaNetCore/Terraria/ID/SurfaceBackgroundID.cs
+++ src/tModLoader/Terraria/ID/SurfaceBackgroundID.cs
@@ -1,3 +_,5 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
 public static class SurfaceBackgroundID
@@ -16,4 +_,5 @@
 	public const int Forest3 = 11;
 	public const int Forest4 = 12;
 	public const int Empty = 13;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(SurfaceBackgroundID), typeof(int)); // TML
 }
