--- src/TerrariaNetCore/Terraria/ID/TeleportationStyleID.cs
+++ src/tModLoader/Terraria/ID/TeleportationStyleID.cs
@@ -1,6 +_,8 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
-public static class TeleportationStyleID
+public class TeleportationStyleID
 {
 	public const int TeleporterTile = 0;
 	public const int RodOfDiscord = 1;
@@ -16,4 +_,5 @@
 	public const int ShellphoneSpawn = 11;
 	public const int ShimmerTownNPCTransform = 12;
 	public const int MysticFrog = 13;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(TeleportationStyleID), typeof(int)); // TML
 }
