--- src/TerrariaNetCore/Terraria/ID/TileID.cs
+++ src/tModLoader/Terraria/ID/TileID.cs
@@ -1,11 +_,12 @@
 using System.Collections.Generic;
 using ReLogic.Reflection;
+using Terraria.ModLoader;
 
 namespace Terraria.ID;
 
-public class TileID
+public partial class TileID
 {
-	public static class Sets
+	public static partial class Sets
 	{
 		public static class Conversion
 		{
@@ -106,7 +_,7 @@
 			};
 		}
 
-		public static SetFactory Factory = new SetFactory(Count);
+		public static SetFactory Factory = new SetFactory(TileLoader.TileCount);
 		public static bool[] IceSkateSlippery = Factory.CreateBoolSet(161, 162, 127, 163, 164, 200, 659);
 		public static bool[] DontDrawTileSliced = Factory.CreateBoolSet(false, 137, 235, 388, 476, 160, 138, 664, 665, 630, 631);
 		public static bool[] AllowsSaveCompressionBatching = Factory.CreateBoolSet(true, 520, 423);
@@ -248,6 +_,10 @@
 		public static bool[] TouchDamageBleeding = Factory.CreateBoolSet(48, 232);
 		public static int[] TouchDamageImmediate = Factory.CreateIntSet(0, 32, 10, 69, 17, 80, 6, 352, 10, 655, 100, 48, 60, 232, 80, 484, 25);
 		public static bool[] Falling = Factory.CreateBoolSet(53, 234, 112, 116, 224, 123, 330, 331, 332, 333, 495);
+		/// <summary>
+		/// Set to true automatically for all tiles that are <see cref="Main.tileSolid"/> and not <see cref="Main.tileSolidTop"/>. These tiles will merge with other tiles that are <see cref="Main.tileBlendAll"/>.
+		/// <br/>You can modify this in <see cref="ModTile.PostSetupTileMerge"/>.
+		/// </summary>
 		public static bool[] BlockMergesWithMergeAllBlock = Factory.CreateBoolSet();
 		public static bool[] OreMergesWithMud = Factory.CreateBoolSet(7, 166, 6, 167, 9, 168, 8, 169, 22, 204, 37, 58, 107, 221, 108, 222, 111, 223);
 		public static bool[] Ore = Factory.CreateBoolSet(7, 166, 6, 167, 9, 168, 8, 169, 22, 204, 37, 58, 107, 221, 108, 222, 111, 223, 211);
