--- src/TerrariaNetCore/Terraria/ID/TorchID.cs
+++ src/tModLoader/Terraria/ID/TorchID.cs
@@ -1,10 +_,14 @@
+#if !TMLCODEASSIST
 using System.Runtime.InteropServices;
 using Microsoft.Xna.Framework;
+#endif
+
 
 namespace Terraria.ID;
 
-public static class TorchID
+public static partial class TorchID
 {
+#if !TMLCODEASSIST
 	private interface ITorchLightProvider
 	{
 		void GetRGB(out float r, out float g, out float b);
@@ -97,6 +_,7 @@
 		310
 	};
 	private static ITorchLightProvider[] _lights;
+#endif
 	public const short Torch = 0;
 	public const short Blue = 1;
 	public const short Red = 2;
@@ -123,6 +_,7 @@
 	public const short Shimmer = 23;
 	public static readonly short Count = 24;
 
+#if !TMLCODEASSIST
 	public static void Initialize()
 	{
 		ITorchLightProvider[] array = new ITorchLightProvider[Count];
@@ -160,4 +_,5 @@
 		else
 			_lights[torchID].GetRGB(out R, out G, out B);
 	}
+#endif
 }
