--- src/TerrariaNetCore/Terraria/ID/WaterStyleID.cs
+++ src/tModLoader/Terraria/ID/WaterStyleID.cs
@@ -1,3 +_,5 @@
+using ReLogic.Reflection;
+
 namespace Terraria.ID;
 
 public static class WaterStyleID
@@ -16,4 +_,5 @@
 	public const int Honey = 11;
 	public const int UndergroundDesert = 12;
 	public static readonly int Count = 13;
+	public static readonly IdDictionary Search = IdDictionary.Create(typeof(WaterStyleID), typeof(int));
 }
