--- src/TerrariaNetCore/Terraria/Graphics/Light/LegacyLighting.cs
+++ src/tModLoader/Terraria/Graphics/Light/LegacyLighting.cs
@@ -4,6 +_,7 @@
 using ReLogic.Threading;
 using Terraria.DataStructures;
 using Terraria.Graphics.Capture;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 
 namespace Terraria.Graphics.Light;
@@ -144,6 +_,9 @@
 
 	public Vector3 GetColor(int x, int y)
 	{
+		if (Testing.DebugOptions.devLightTilesCheat)
+			return Vector3.One;
+
 		if (x < _expandedRectLeft || x >= _expandedRectRight || y < _expandedRectTop || y >= _expandedRectBottom)
 			return Vector3.Zero;
 
@@ -581,6 +_,7 @@
 		num3 = Utils.Clamp(_expandedRectTop, 5, Main.maxTilesY - 1);
 		num4 = Utils.Clamp(_expandedRectBottom, 5, Main.maxTilesY - 1);
 		Main.UpdateSceneMetrics();
+		// Extra patch context.
 		_tileScanner.Update();
 		_tileScanner.ExportTo(new Rectangle(num, num3, num2 - num, num4 - num3), _lightMap, new TileLightScannerOptions {
 			DrawInvisibleWalls = Main.ShouldShowInvisibleBlocksAndWalls()
@@ -588,6 +_,7 @@
 
 		for (int m = num; m < num2; m++) {
 			LightingState[] array3 = _states[m - _expandedRectLeft];
+			//patch file: n, num17
 			for (int n = num3; n < num4; n++) {
 				LightingState lightingState2 = array3[n - _expandedRectTop];
 				Tile tile = Main.tile[m, n];
@@ -671,6 +_,12 @@
 		if (!Main.gamePaused && !CaptureManager.Instance.IsCapturing)
 			_tempLights.Clear();
 
+		if (Testing.DebugOptions.devLightTilesCheat) {
+			Main.renderCount = 1;
+			TimeLogger.LightingInit.AddTime(fromTimestamp);
+			return;
+		}
+
 		_minX += _expandedRectLeft;
 		_maxX += _expandedRectLeft;
 		_minY += _expandedRectTop;
@@ -688,6 +_,9 @@
 
 	private void DoColors()
 	{
+		if (Testing.DebugOptions.devLightTilesCheat)
+			return;
+
 		if (IsColorOrWhiteMode) {
 			_blueWave += (float)_blueDir * 0.0001f;
 			if (_blueWave > 1f) {
@@ -761,6 +_,8 @@
 						_wetLightB = 0f;
 						break;
 				}
+
+				LoaderManager.Get<WaterStylesLoader>().LightColorMultiplier(Main.waterStyle, _negLight * _blueWave, ref _wetLightR, ref _wetLightG, ref _wetLightB);
 			}
 			else {
 				_negLight = 0.9f;
@@ -792,6 +_,8 @@
 				_negLight2 *= 0.85f;
 				_negLight3 *= 0.85f;
 			}
+
+			SystemLoader.ModifyLightingBrightness(ref _negLight, ref _negLight2); // TODO: _negLight3
 		}
 		else {
 			_negLight = 0.04f;
@@ -820,6 +_,8 @@
 				_negLight2 += 0.18f;
 				_negLight3 += 0.18f;
 			}
+
+			SystemLoader.ModifyLightingBrightness(ref _negLight, ref _negLight2);
 
 			_wetLightR = _negLight * 1.2f;
 			_wetLightG = _negLight * 1.1f;
