--- src/Terraria/Terraria/Graphics/Light/LegacyLighting.cs
+++ src/tModLoader/Terraria/Graphics/Light/LegacyLighting.cs
@@ -5,6 +_,7 @@
 using System.Diagnostics;
 using Terraria.DataStructures;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light
 {
@@ -106,7 +_,7 @@
 		private int _expandedRectTop;
 		private int _expandedRectRight;
 		private int _expandedRectBottom;
-		private float _negLight = 0.04f;
+		internal float _negLight = 0.04f;
 		private float _negLight2 = 0.16f;
 		private float _wetLightR = 0.16f;
 		private float _wetLightG = 0.16f;
@@ -114,7 +_,7 @@
 		private float _honeyLightR = 0.16f;
 		private float _honeyLightG = 0.16f;
 		private float _honeyLightB = 0.16f;
-		private float _blueWave = 1f;
+		internal float _blueWave = 1f;
 		private int _blueDir = 1;
 		private RectArea _minBoundArea;
 		private RectArea _requestedArea;
@@ -578,6 +_,7 @@
 			_tileScanner.ExportTo(new Rectangle(num, num3, num2 - num, num4 - num3), _lightMap);
 			for (int m = num; m < num2; m++) {
 				LightingState[] array3 = _states[m - _expandedRectLeft];
+				//patch file: n, num17
 				for (int n = num3; n < num4; n++) {
 					LightingState lightingState2 = array3[n - _expandedRectTop];
 					Tile tile = Main.tile[m, n];
@@ -746,6 +_,8 @@
 							_wetLightB = 0f;
 							break;
 					}
+
+					WaterStyleLoader.LightColorMultiplier(Main.waterStyle, ref _wetLightR, ref _wetLightG, ref _wetLightB);
 				}
 				else {
 					_negLight = 0.9f;
@@ -772,6 +_,8 @@
 					_negLight *= 0.85f;
 					_negLight2 *= 0.85f;
 				}
+
+				ModHooks.ModifyLightingBrightness(ref _negLight, ref _negLight2);
 			}
 			else {
 				_negLight = 0.04f;
@@ -796,6 +_,7 @@
 					_negLight2 += 0.18f;
 				}
 
+				ModHooks.ModifyLightingBrightness(ref _negLight, ref _negLight2);
 				_wetLightR = _negLight * 1.2f;
 				_wetLightG = _negLight * 1.1f;
 			}
