--- src/TerrariaNetCore/Terraria/Graphics/Light/LightingEngine.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightingEngine.cs
@@ -3,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light
 {
@@ -175,6 +_,12 @@
 				workingLightMap.LightDecayThroughAir *= 0.85f;
 				workingLightMap.LightDecayThroughSolid *= 0.85f;
 			}
+
+			float throughAir = 1f;
+			float throughSolid = 1f;
+			SystemLoader.ModifyLightingBrightness(ref throughAir, ref throughSolid);
+			workingLightMap.LightDecayThroughAir *= throughAir;
+			workingLightMap.LightDecayThroughSolid *= throughSolid;
 		}
 
 		private void ApplyPerFrameLights() {
