--- src/TerrariaNetCore/Terraria/Graphics/Light/LightMap.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightMap.cs
@@ -182,9 +_,21 @@
 
 	public void SetSize(int width, int height)
 	{
+		//TML:
+		// The width and height parameters here are max Y and max X.
+		// We need to add one to each dimension to get the correct Array Length on resize
+		// - Solxan
+		int neededSize = (width + 1) * (height + 1);
+
+		/*
 		if (width * height > _colors.Length) {
 			_colors = new Vector3[width * height];
 			_mask = new LightMaskMode[width * height];
+		}
+		*/
+		if (neededSize > _colors.Length) {
+			_colors = new Vector3[neededSize];
+			_mask = new LightMaskMode[neededSize];
 		}
 
 		Width = width;
