--- src/Terraria/Terraria/Graphics/Light/LightingEngine.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightingEngine.cs
@@ -31,7 +_,7 @@
 		public const int AREA_PADDING = 28;
 		private const int NON_VISIBLE_PADDING = 18;
 		private readonly List<PerFrameLight> _perFrameLights = new List<PerFrameLight>();
-		private TileLightScanner _tileScanner;
+		private TileLightScanner _tileScanner = new();
 		private LightMap _activeLightMap = new LightMap();
 		private Rectangle _activeProcessedArea;
 		private LightMap _workingLightMap = new LightMap();
@@ -39,10 +_,6 @@
 		private readonly Stopwatch _timer = new Stopwatch();
 		private EngineState _state;
 
-		public LightingEngine(World world) {
-			SetWorld(world);
-		}
-
 		public void AddLight(int x, int y, Vector3 color) {
 			_perFrameLights.Add(new PerFrameLight(new Point(x, y), color));
 		}
@@ -201,10 +_,6 @@
 			_state = EngineState.MinimapUpdate;
 			_activeLightMap = new LightMap();
 			_workingLightMap = new LightMap();
-		}
-
-		private void SetWorld(World world) {
-			_tileScanner = new TileLightScanner(world);
 		}
 
 		private void ExportToMiniMap() {
