--- src/Terraria/Terraria/Graphics/Light/LightMap.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightMap.cs
@@ -195,9 +_,11 @@
 		private int IndexOf(int x, int y) => x * Height + y;
 
 		public void SetSize(int width, int height) {
-			if (width * height > _colors.Length) {
-				_colors = new Vector3[width * height];
-				_mask = new LightMaskMode[width * height];
+			int neededSize = (width + 1) * (height + 1);
+
+			if (neededSize > _colors.Length) {
+				_colors = new Vector3[neededSize];
+				_mask = new LightMaskMode[neededSize];
 			}
 
 			Width = width;
