--- src/TerrariaNetCore/Terraria/Graphics/Light/LightingEngine.cs
+++ src/tModLoader/Terraria/Graphics/Light/LightingEngine.cs
@@ -3,6 +_,7 @@
 using System.Diagnostics;
 using Microsoft.Xna.Framework;
 using ReLogic.Threading;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light;
 
@@ -171,6 +_,13 @@
 				break;
 		}
 
+		float factor = 0.91f; // Taken from the constant at the start of the method.
+		float throughWaterR = workingLightMap.LightDecayThroughWater.X;
+		float throughWaterG = workingLightMap.LightDecayThroughWater.Y;
+		float throughWaterB = workingLightMap.LightDecayThroughWater.Z;
+		LoaderManager.Get<WaterStylesLoader>().LightColorMultiplier(Main.waterStyle, factor, ref throughWaterR, ref throughWaterG, ref throughWaterB);
+		workingLightMap.LightDecayThroughWater = new Vector3(throughWaterR, throughWaterG, throughWaterB);
+
 		if (Main.player[Main.myPlayer].nightVision) {
 			workingLightMap.LightDecayThroughAir *= 1.03f;
 			workingLightMap.LightDecayThroughSolid *= 1.03f;
@@ -193,6 +_,15 @@
 
 		workingLightMap.LightDecayThroughAir *= Player.airLightDecay;
 		workingLightMap.LightDecayThroughSolid *= Player.solidLightDecay;
+
+		//TML: Shouldn't this just ref to LightMap values instead?
+		float throughAir = 1f;
+		float throughSolid = 1f;
+
+		SystemLoader.ModifyLightingBrightness(ref throughAir, ref throughSolid);
+
+		workingLightMap.LightDecayThroughAir *= throughAir;
+		workingLightMap.LightDecayThroughSolid *= throughSolid;
 	}
 
 	private void ApplyPerFrameLights()
