--- src/TerrariaNetCore/Terraria/Graphics/Light/TileLightScanner.cs
+++ src/tModLoader/Terraria/Graphics/Light/TileLightScanner.cs
@@ -5,6 +_,7 @@
 using Terraria.GameContent.Liquid;
 using Terraria.ID;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light;
 
@@ -280,9 +_,12 @@
 			num3 = (float)(int)color.B / 765f;
 		}
 
+		WallLoader.ModifyLight(x, y, tile.wall, ref num, ref num2, ref num3);
+
 		if (lightColor.X < num)
 			lightColor.X = num;
 
+		// Extra patch context.
 		if (lightColor.Y < num2)
 			lightColor.Y = num2;
 
@@ -2952,8 +_,11 @@
 					B = 0.15f;
 					break;
 			}
+
+			TileLoader.ModifyLight(x, y, tile.type, ref R, ref G, ref B);
 		}
 
+		// TODO: flag parameter in ModifyLight
 		if (flag && tile.color() != 0) {
 			Color color2 = WorldGen.paintColor(tile.color());
 			R = (float)(int)color2.R / 255f;
@@ -2964,6 +_,7 @@
 		if (lightColor.X < R)
 			lightColor.X = R;
 
+		// Extra patch context.
 		if (lightColor.Y < G)
 			lightColor.Y = G;
 
