--- src/TerrariaNetCore/Terraria/Graphics/Light/TileLightScanner.cs
+++ src/tModLoader/Terraria/Graphics/Light/TileLightScanner.cs
@@ -5,6 +_,7 @@
 using Terraria.GameContent.Liquid;
 using Terraria.ID;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light;
 
@@ -255,9 +_,12 @@
 				break;
 		}
 
+		WallLoader.ModifyLight(x, y, tile.wall, ref num, ref num2, ref num3);
+
 		if (lightColor.X < num)
 			lightColor.X = num;
 
+		// Extra patch context.
 		if (lightColor.Y < num2)
 			lightColor.Y = num2;
 
@@ -2214,11 +_,14 @@
 					B = 0.15f;
 					break;
 			}
+
+			TileLoader.ModifyLight(x, y, tile.type, ref R, ref G, ref B);
 		}
 
 		if (lightColor.X < R)
 			lightColor.X = R;
 
+		// Extra patch context.
 		if (lightColor.Y < G)
 			lightColor.Y = G;
 
