--- src/TerrariaNetCore/Terraria/Graphics/Light/TileLightScanner.cs
+++ src/tModLoader/Terraria/Graphics/Light/TileLightScanner.cs
@@ -5,6 +_,7 @@
 using Terraria.GameContent.Liquid;
 using Terraria.ID;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Light;
 
@@ -273,16 +_,18 @@
 				break;
 		}
 
-		if (flag && tile.wallColor() != 0) {
+			if (flag && tile.wallColor() != 0) {
-			Color color = WorldGen.paintColor(tile.wallColor());
+				Color color = WorldGen.paintColor(tile.wallColor());
-			num = (float)(int)color.R / 765f;
+				num = (float)(int)color.R / 765f;
-			num2 = (float)(int)color.G / 765f;
+				num2 = (float)(int)color.G / 765f;
-			num3 = (float)(int)color.B / 765f;
+				num3 = (float)(int)color.B / 765f;
-		}
+			}
+			WallLoader.ModifyLight(x, y, tile.wall, ref num, ref num2, ref num3);
 
-		if (lightColor.X < num)
+			if (lightColor.X < num)
 			lightColor.X = num;
 
+		// Extra patch context.
 		if (lightColor.Y < num2)
 			lightColor.Y = num2;
 
@@ -2952,6 +_,8 @@
 					B = 0.15f;
 					break;
 			}
+
+			TileLoader.ModifyLight(x, y, tile.type, ref R, ref G, ref B);
 		}
 
 		if (flag && tile.color() != 0) {
@@ -2964,6 +_,7 @@
 		if (lightColor.X < R)
 			lightColor.X = R;
 
+		// Extra patch context.
 		if (lightColor.Y < G)
 			lightColor.Y = G;
 
