--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/TerrariaNetCore/ReLogic/ReLogic.csproj
@@ -1,40 +_,36 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
-	<Import Project="..\Configuration.targets" />
+	<Import Project="../Configuration.targets" />
 
+	<!-- Build properties -->
 	<PropertyGroup>
 		<OutputType>Library</OutputType>
 		<Version>1.0.0.0</Version>
 		<Company>Re-Logic</Company>
 		<Copyright>Copyright © Re-Logic 2017</Copyright>
 		<RootNamespace>ReLogic</RootNamespace>
-		<!-- Because project name is "ReLogic_LegacyNetFramework" -->
-		<AssemblyName>ReLogic</AssemblyName>
 	</PropertyGroup>
 
-	<ItemGroup>
-	</ItemGroup>
-
 	<!-- References -->
 	<ItemGroup>
+		<!-- Projects -->
+		<ProjectReference Include="..\..\..\FNA\FNA.Core.csproj" />
+		<!-- NuGet Packages -->
+		<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
 		<!-- DLLs -->
-		<Reference Include="..\Terraria\Libraries\DotNetZip\Ionic.Zip.CF.dll" />
+		<Reference Include="..\Terraria\Libraries\Common\Ionic.Zip.CF.dll" />
-		<Reference Include="..\Terraria\Libraries\FNA\FNA.dll" Condition="$(DefineConstants.Contains('FNA'))" />
-		<Reference Include="..\Terraria\Libraries\XNA\Microsoft.Xna.Framework.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="..\Terraria\Libraries\XNA\Microsoft.Xna.Framework.Game.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="..\Terraria\Libraries\XNA\Microsoft.Xna.Framework.Graphics.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="..\Terraria\Libraries\JSON.NET\Newtonsoft.Json.dll" />
 		<Reference Include="..\Terraria\Libraries\Common\SteelSeriesEngineWrapper.dll" />
-		<Reference Include="System" />
-		<Reference Include="System.Core" />
-		<Reference Include="System.Drawing" />
-		<Reference Include="System.Windows.Forms" />
 	</ItemGroup>
 
 	<!-- Resources -->
 	<ItemGroup>
 		<!-- Embedded resources -->
-		<EmbeddedResource Include="Microsoft\**" />
+		<EmbeddedResource Include="Microsoft/**" />
+
+		<!-- Excluded code files -->
+		<Compile Remove="Localization\IME\Windows/*" />
+		<Compile Remove="Localization\IME\WindowsIme.cs" />
+		<Compile Remove="OS\Windows\WindowsMessageHook.cs" />
 	</ItemGroup>
 
 </Project>
