--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/TerrariaNetCore/ReLogic/ReLogic.csproj
@@ -8,17 +_,10 @@
 		<RootNamespace>ReLogic</RootNamespace>
 	</PropertyGroup>
 	<ItemGroup>
-		<Reference Include="../Terraria/Libraries/DotNetZip/Ionic.Zip.CF.dll" />
-		<Reference Include="../Terraria/Libraries/FNA/FNA.dll" Condition="$(DefineConstants.Contains('FNA'))" />
-		<Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.Game.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.Graphics.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-		<Reference Include="../Terraria/Libraries/JSON.NET/Newtonsoft.Json.dll" />
+		<Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+		<ProjectReference Include="../../../FNA/FNA.Core.csproj" />
+		<Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
 		<Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
-		<Reference Include="System" />
-		<Reference Include="System.Core" />
-		<Reference Include="System.Drawing" />
-		<Reference Include="System.Windows.Forms" />
 	</ItemGroup>
 	<ItemGroup>
 		<EmbeddedResource Include="Microsoft/**" />
