--- src/decompiled\Terraria.Graphics.Capture\CaptureInterface.cs
+++ src/Terraria\Terraria.Graphics.Capture\CaptureInterface.cs
@@ -917,8 +_,19 @@
 						ToggleCamera();
 					}
 
-					if (i == num2++ && flag2 && flag)
+					if (i == num2++ && flag2 && flag) {
+#if LINUX
+						string fileName = Path.Combine(Main.SavePath, "Captures");
+						Process.Start(new ProcessStartInfo(fileName) {
+							FileName = "open-folder",
+							Arguments = fileName,
+							UseShellExecute = true,
+							CreateNoWindow = true
+						});
+#else
 						Process.Start(Path.Combine(Main.SavePath, "Captures"));
+#endif
+					}
 
 					if (i == num2++ && flag2) {
 						ToggleCamera(On: false);
