--- src/decompiled\Terraria.Graphics.Capture\CaptureInterface.cs
+++ src/Terraria\Terraria.Graphics.Capture\CaptureInterface.cs
@@ -918,7 +_,17 @@
 					}
 
 					if (i == num2++ && flag2 && flag)
+#if LINUX
+						Process.Start(new ProcessStartInfo(fileName) {
+							FileName = "open-folder",
+							Arguments = fileName,
+							UseShellExecute = true,
+							CreateNoWindow = true
+						});
+#else
 						Process.Start(Path.Combine(Main.SavePath, "Captures"));
+
+#endif
 
 					if (i == num2++ && flag2) {
 						ToggleCamera(On: false);
