--- src/decompiled/Terraria.Graphics.Capture/CaptureInterface.cs
+++ src/Terraria/Terraria.Graphics.Capture/CaptureInterface.cs
@@ -889,7 +_,17 @@
 
 					if (i == num2++ && flag2 && flag) {
 						SoundEngine.PlaySound(12);
+#if LINUX
+						string fileName = Path.Combine(Main.SavePath, "Captures");
+						Process.Start(new ProcessStartInfo(fileName) {
+							FileName = "open-folder",
+							Arguments = fileName,
+							UseShellExecute = true,
+							CreateNoWindow = true
+						});
+#else
 						Utils.OpenFolder(Path.Combine(Main.SavePath, "Captures"));
+#endif
 					}
 
 					if (i == num2++ && flag2) {
