--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/AWorldListItem.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/AWorldListItem.cs
@@ -4,12 +_,17 @@
 using ReLogic.Content;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent.UI.Elements;
 
 public abstract class AWorldListItem : UIPanel
 {
+	// Added by TML
+	public WorldFileData Data => _data;
+	protected bool _showVanillaSeedIcons = true;
+
 	protected WorldFileData _data;
 	protected int _glitchFrameCounter;
 	protected int _glitchFrame;
@@ -136,10 +_,13 @@
 		if (_data.SkyblockWorld)
 			list.Add(GetSeedIcon("Skyblock", hardmodeVariants: false, evilVariants: false));
 
+		list.RemoveAll(icon => icon == null);
+
 		if (list.Count > 0)
 			return list;
 
 		list.Add(GetSeedIcon(""));
+		_showVanillaSeedIcons = false; // This won't do anything unless the world file actually has loaded modded seeds
 		return list;
 	}
 
@@ -160,6 +_,8 @@
 
 	private Asset<Texture2D> GetSeedIcon(string seed, bool hardmodeVariants = true, bool evilVariants = true)
 	{
+		if (SeedLoader.IsVanillaSeedDependency(seed, _data))
+			return null;
 		string text = "Images/UI/Icon";
 		if (hardmodeVariants)
 			text += (_data.IsHardMode ? "Hallow" : "");
