--- src/Terraria/Terraria/GameContent/UI/Elements/UIHeader.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIHeader.cs
@@ -16,11 +_,13 @@
 			set {
 				if (_text != value) {
 					_text = value;
+
-#if CLIENT
+					if (!Main.dedServ) {
-					Vector2 vector = FontAssets.DeathText.Value.MeasureString(Text);
+						Vector2 vector = FontAssets.DeathText.Value.MeasureString(Text);
-					Width.Pixels = vector.X;
+						Width.Pixels = vector.X;
-					Height.Pixels = vector.Y;
+						Height.Pixels = vector.Y;
-#endif
+					}
+					
 					Width.Precent = 0f;
 					Height.Precent = 0f;
 					Recalculate();
