--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIKeybindingListItem.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIKeybindingListItem.cs
@@ -4,6 +_,7 @@
 using Terraria.GameContent.UI.Chat;
 using Terraria.GameInput;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 using Terraria.UI.Chat;
 
@@ -99,11 +_,19 @@
 		private string GetFriendlyName() {
 			switch (_keybind) {
 				default:
+					if (KeybindLoader.modKeybinds.TryGetValue(_keybind, out ModKeybind modKeybind))
+						return modKeybind.DisplayName.GetTranslation(Language.ActiveCulture);
 					return _keybind;
 				case "MouseLeft":
 					return Lang.menu[162].Value;
 				case "MouseRight":
 					return Lang.menu[163].Value;
+				case "MouseMiddle":
+					return Language.GetTextValue("tModLoader.MouseMiddle");
+				case "MouseXButton1":
+					return Language.GetTextValue("tModLoader.MouseXButton1");
+				case "MouseXButton2":
+					return Language.GetTextValue("tModLoader.MouseXButton2");
 				case "Up":
 					return Lang.menu[148].Value;
 				case "Down":
