--- src/Terraria/Terraria/GameContent/UI/Elements/UIList.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIList.cs
@@ -7,7 +_,7 @@
 
 namespace Terraria.GameContent.UI.Elements
 {
-	public class UIList : UIElement, IEnumerable<UIElement>, IEnumerable
+	public partial class UIList : UIElement, IEnumerable<UIElement>, IEnumerable
 	{
 		public delegate bool ElementSearchMethod(UIElement element);
 
@@ -29,9 +_,9 @@
 			public override Rectangle GetViewCullingArea() => base.Parent.GetDimensions().ToRectangle();
 		}
 
-		protected List<UIElement> _items = new List<UIElement>();
+		internal List<UIElement> _items = new List<UIElement>();
 		protected UIScrollbar _scrollbar;
-		private UIElement _innerList = new UIInnerList();
+		internal UIElement _innerList = new UIInnerList();
 		private float _innerListHeight;
 		public float ListPadding = 5f;
 		public Action<List<UIElement>> ManualSortMethod;
