--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIList.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIList.cs
@@ -7,7 +_,7 @@
 
 namespace Terraria.GameContent.UI.Elements;
 
-public class UIList : UIElement, IEnumerable<UIElement>, IEnumerable
+public partial class UIList : UIElement, IEnumerable<UIElement>, IEnumerable
 {
 	public delegate bool ElementSearchMethod(UIElement element);
 
@@ -30,9 +_,11 @@
 		public override Rectangle GetViewCullingArea() => base.Parent.GetDimensions().ToRectangle();
 	}
 
+	//TML: Made public instead of protected.
-	protected List<UIElement> _items = new List<UIElement>();
+	public List<UIElement> _items = new List<UIElement>();
 	protected UIScrollbar _scrollbar;
+	//TML: Made internal instead of private.
-	private UIElement _innerList = new UIInnerList();
+	internal UIElement _innerList = new UIInnerList();
 	private float _innerListHeight;
 	public float ListPadding = 5f;
 	public Action<List<UIElement>> ManualSortMethod;
@@ -71,6 +_,7 @@
 	public virtual bool Remove(UIElement item)
 	{
 		_innerList.RemoveChild(item);
+		// If order is stable doesn't make sense to reorder, left because it's in vanilla
 		UpdateOrder();
 		return _items.Remove(item);
 	}
