--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIPanel.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIPanel.cs
@@ -14,15 +_,24 @@
 	public Color BorderColor = Color.Black;
 	public Color BackgroundColor = new Color(63, 82, 151) * 0.7f;
 
-	public UIPanel()
+	// Added by TML.
+	private bool _needsTextureLoading;
+
+	private void LoadTextures()
 	{
+		// These used to be moved to OnActivate in order to avoid texture loading on JIT thread.
+		// Doing so caused issues with missing backgrounds and borders because Activate wasn't always being called.
 		if (_borderTexture == null)
 			_borderTexture = Main.Assets.Request<Texture2D>("Images/UI/PanelBorder");
 
 		if (_backgroundTexture == null)
 			_backgroundTexture = Main.Assets.Request<Texture2D>("Images/UI/PanelBackground");
+	}
 
+	public UIPanel()
+	{
 		SetPadding(_cornerSize);
+		_needsTextureLoading = true;
 	}
 
 	public UIPanel(Asset<Texture2D> customBackground, Asset<Texture2D> customborder, int customCornerSize = 12, int customBarSize = 4)
@@ -58,6 +_,11 @@
 
 	protected override void DrawSelf(SpriteBatch spriteBatch)
 	{
+		if (_needsTextureLoading) {
+			_needsTextureLoading = false;
+			LoadTextures();
+		}
+
 		if (_backgroundTexture != null)
 			DrawPanel(spriteBatch, _backgroundTexture.Value, BackgroundColor);
 
