--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UITextPanel.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UITextPanel.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using ReLogic.Graphics;
 using Terraria.UI;
+using Terraria.UI.Chat;
 
 namespace Terraria.GameContent.UI.Elements;
 
@@ -56,6 +_,20 @@
 		}
 	}
 
+	protected virtual Vector2 TextDrawPosition {
+		get {
+			CalculatedStyle innerDimensions = GetInnerDimensions();
+			Vector2 result = innerDimensions.Position();
+			result.X += (innerDimensions.Width - _textSize.X) * TextHAlign;
+			if (_isLarge)
+				result.Y -= 10f * _textScale * _textScale;
+			else
+				result.Y -= 2f * _textScale;
+
+			return result;
+		}
+	}
+
 	public UITextPanel(T text, float textScale = 1f, bool large = false)
 	{
 		SetText(text, textScale, large);
@@ -74,7 +_,14 @@
 
 	public virtual void SetText(T text, float textScale, bool large)
 	{
+		/*
 		Vector2 textSize = new Vector2((large ? FontAssets.DeathText.Value : FontAssets.MouseText.Value).MeasureString(text.ToString()).X, large ? 32f : 16f) * textScale;
+		*/
+
+		DynamicSpriteFont dynamicSpriteFont = large ? FontAssets.DeathText.Value : FontAssets.MouseText.Value;
+		Vector2 textSize = ChatManager.GetStringSize(dynamicSpriteFont, text.ToString(), new Vector2(textScale));
+		textSize.Y = (large ? 32f : 16f) * textScale;
+
 		_text = text;
 		_textScale = textScale;
 		_textSize = textSize;
@@ -93,6 +_,8 @@
 
 	protected void DrawText(SpriteBatch spriteBatch)
 	{
+		// 1.4.5 Backport: Draw IME composition strings (#5293)
+		/*
 		CalculatedStyle innerDimensions = GetInnerDimensions();
 		Vector2 pos = innerDimensions.Position();
 		if (_isLarge)
@@ -101,6 +_,7 @@
 			pos.Y -= 2f * _textScale;
 
 		pos.X += (innerDimensions.Width - _textSize.X) * TextHAlign;
+		*/
 		string text = Text;
 		if (HideContents) {
 			if (_asterisks == null || _asterisks.Length != text.Length)
@@ -109,9 +_,20 @@
 			text = _asterisks;
 		}
 
+		/*
 		if (_isLarge)
 			Utils.DrawBorderStringBig(spriteBatch, text, pos, _color, _textScale);
 		else
 			Utils.DrawBorderString(spriteBatch, text, pos, _color, _textScale);
+		*/
+		DrawText(spriteBatch, text, TextDrawPosition, _color);
+	}
+
+	protected void DrawText(SpriteBatch spriteBatch, string text, Vector2 position, Color color)
+	{
+		if (_isLarge)
+			Utils.DrawBorderStringBig(spriteBatch, text, position, color, _textScale);
+		else
+			Utils.DrawBorderString(spriteBatch, text, position, color, _textScale);
 	}
 }
