--- src/Terraria/Terraria/GameContent/UI/Elements/UIText.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIText.cs
@@ -122,6 +_,8 @@
 		}
 
 		private void InternalSetText(object text, float textScale, bool large) {
+			if (Main.dedServ)
+				return;
 			DynamicSpriteFont dynamicSpriteFont = large ? FontAssets.DeathText.Value : FontAssets.MouseText.Value;
 			_text = text;
 			_isLarge = large;
