--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UITextBox.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UITextBox.cs
@@ -1,11 +_,13 @@
 using System;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using ReLogic.Localization.IME;
+using ReLogic.OS;
 using Terraria.UI;
 
 namespace Terraria.GameContent.UI.Elements;
 
-internal class UITextBox : UITextPanel<string>
+public class UITextBox : UITextPanel<string>
 {
 	private int _cursor;
 	private int _frameCount;
@@ -13,6 +_,19 @@
 	public bool ShowInputTicker = true;
 	public bool HideSelf;
 
+	protected override Vector2 TextDrawPosition {
+		get {
+			Vector2 textDrawPosition = base.TextDrawPosition;
+			if (ShowInputTicker) {
+				string compositionString = Platform.Get<IImeService>().CompositionString;
+				if (!string.IsNullOrEmpty(compositionString))
+					textDrawPosition.X -= (base.IsLarge ? FontAssets.DeathText.Value : FontAssets.MouseText.Value).MeasureString(compositionString).X * base.TextScale * TextHAlign;
+			}
+
+			return textDrawPosition;
+		}
+	}
+
 	public UITextBox(string text, float textScale = 1f, bool large = false)
 		: base(text, textScale, large)
 	{
@@ -66,7 +_,19 @@
 
 		_cursor = base.Text.Length;
 		base.DrawSelf(spriteBatch);
+		if (!ShowInputTicker)
+			return;
+
+		Vector2 textDrawPosition = TextDrawPosition;
+		string compositionString = Platform.Get<IImeService>().CompositionString;
+		if (!string.IsNullOrEmpty(compositionString)) {
+			textDrawPosition.X += (base.IsLarge ? FontAssets.DeathText.Value : FontAssets.MouseText.Value).MeasureString(compositionString).X * base.TextScale;
+			DrawText(spriteBatch, compositionString, TextDrawPosition + new Vector2(base.TextSize.X, 0f), Main.imeCompositionStringColor);
+		}
+
 		_frameCount++;
+		// 1.4.5 Backport: Draw IME composition strings (#5293)
+		/*
 		if ((_frameCount %= 40) <= 20 && ShowInputTicker) {
 			CalculatedStyle innerDimensions = GetInnerDimensions();
 			Vector2 pos = innerDimensions.Position();
@@ -81,6 +_,17 @@
 				Utils.DrawBorderStringBig(spriteBatch, "|", pos, base.TextColor, base.TextScale);
 			else
 				Utils.DrawBorderString(spriteBatch, "|", pos, base.TextColor, base.TextScale);
+		}
+		*/
+		if ((_frameCount %= 40) <= 20) {
+			textDrawPosition.X +=
+				(base.IsLarge ? FontAssets.DeathText.Value : FontAssets.MouseText.Value)
+				.MeasureString(base.Text.Substring(0, _cursor)).X * base.TextScale;
+			textDrawPosition.X += 6f - (base.IsLarge ? 8f : 4f) * base.TextScale;
+			if (base.IsLarge)
+				textDrawPosition.Y += 2f * base.TextScale;
+
+			DrawText(spriteBatch, "|", textDrawPosition, base.TextColor);
 		}
 	}
 }
