--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIText.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIText.cs
@@ -37,6 +_,8 @@
 		}
 		set {
 			_isWrapped = value;
+			if(value)
+				MinWidth.Set(0, 0); // TML: IsWrapped when true should prevent changing MinWidth, otherwise can't shrink in width due to CreateWrappedText+GetInnerDimensions logic. IsWrapped is false in ctor, so need to undo changes.
 			InternalSetText(_text, _textScale, _isLarge);
 		}
 	}
@@ -160,9 +_,9 @@
 		else
 			_textSize.Y = (large ? 32f : 16f) * _textScale;
 
-		MinWidth.Set((IsWrapped || DynamicallyScaleDownToWidth) ? 0f : (_textSize.X + PaddingLeft + PaddingRight), 0f);
+			MinWidth.Set((IsWrapped || DynamicallyScaleDownToWidth) ? 0f : (_textSize.X + PaddingLeft + PaddingRight), 0f);
-		MinHeight.Set(_textSize.Y + PaddingTop + PaddingBottom, 0f);
+			MinHeight.Set(_textSize.Y + PaddingTop + PaddingBottom, 0f);
-		if (this.OnInternalTextChange != null)
+			if (this.OnInternalTextChange != null)
-			this.OnInternalTextChange();
+				this.OnInternalTextChange();
-	}
+		}
 }
