--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UIText.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIText.cs
@@ -37,6 +_,8 @@
 		}
 		set {
 			_isWrapped = value;
+			if(value)
+				MinWidth.Set(0, 0); // TML: IsWrapped when true should prevent changing MinWidth, otherwise can't shrink in width due to CreateWrappedText+GetInnerDimensions logic. IsWrapped is false in ctor, so need to undo changes.
 			InternalSetText(_text, _textScale, _isLarge);
 		}
 	}
