--- src/TerrariaNetCore/Terraria/GameContent/UI/Elements/UITextPanel.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UITextPanel.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using ReLogic.Graphics;
 using Terraria.UI;
+using Terraria.UI.Chat;
 
 namespace Terraria.GameContent.UI.Elements;
 
@@ -74,7 +_,14 @@
 
 	public virtual void SetText(T text, float textScale, bool large)
 	{
+		/*
 		Vector2 textSize = new Vector2((large ? FontAssets.DeathText.Value : FontAssets.MouseText.Value).MeasureString(text.ToString()).X, large ? 32f : 16f) * textScale;
+		*/
+
+		DynamicSpriteFont dynamicSpriteFont = large ? FontAssets.DeathText.Value : FontAssets.MouseText.Value;
+		Vector2 textSize = ChatManager.GetStringSize(dynamicSpriteFont, text.ToString(), new Vector2(textScale));
+		textSize.Y = (large ? 32f : 16f) * textScale;
+
 		_text = text;
 		_textScale = textScale;
 		_textSize = textSize;
