--- src/Terraria/Terraria/GameContent/UI/Elements/UIWorldListItem.cs
+++ src/tModLoader/Terraria/GameContent/UI/Elements/UIWorldListItem.cs
@@ -2,10 +_,13 @@
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
 using ReLogic.OS;
+using System;
 using Terraria.Audio;
 using Terraria.DataStructures;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Config;
 using Terraria.Social;
 using Terraria.UI;
 
@@ -28,6 +_,7 @@
 		private Asset<Texture2D> _buttonDeleteTexture;
 		private UIImageButton _deleteButton;
 		private int _orderInList;
+		private Asset<Texture2D> _configTexture;
 		private bool _canBePlayed;
 
 		public bool IsFavorite => _data.IsFavorite;
@@ -85,6 +_,17 @@
 				num += 24f;
 			}
 
+			//if (ConfigManager.Configs.Count > 0)
+			//{
+			//	UIImageButton configButton = new UIImageButton(this._configTexture);
+			//	configButton.VAlign = 1f;
+			//	configButton.Left.Set(num, 0f);
+			//	configButton.OnClick += new UIElement.MouseEvent(this.ConfigButtonClick);
+			//	configButton.OnMouseOver += new UIElement.MouseEvent(this.ConfigMouseOver);
+			//	configButton.OnMouseOut += new UIElement.MouseEvent(this.ButtonMouseOut);
+			//	Append(configButton);
+			//	num += 24f;
+			//}
 			UIImageButton uIImageButton5 = new UIImageButton(_buttonDeleteTexture) {
 				VAlign = 1f,
 				HAlign = 1f
@@ -124,6 +_,7 @@
 			_buttonPlayTexture = Main.Assets.Request<Texture2D>("Images/UI/ButtonPlay");
 			_buttonSeedTexture = Main.Assets.Request<Texture2D>("Images/UI/ButtonSeed");
 			_buttonDeleteTexture = Main.Assets.Request<Texture2D>("Images/UI/ButtonDelete");
+			_configTexture = ModLoader.UI.UICommon.ButtonConfigTexture;
 		}
 
 		private void InitializeAppearance() {
@@ -172,6 +_,10 @@
 				_buttonLabel.SetText(Language.GetTextValue("UI.MoveToCloud"));
 		}
 
+		private void ConfigMouseOver(UIMouseEvent evt, UIElement listeningElement) {
+			_buttonLabel.SetText("Edit World Config");
+		}
+
 		private void PlayMouseOver(UIMouseEvent evt, UIElement listeningElement) {
 			_buttonLabel.SetText(Language.GetTextValue("UI.Play"));
 		}
@@ -239,9 +_,29 @@
 				else
 					Main.menuMode = 10;
 
-				if (!Main.menuMultiplayer)
+				if (!Main.menuMultiplayer) {
+					ConfigManager.LoadAll(); // Makes sure MP configs are cleared.
+					if (ConfigManager.AnyModNeedsReload()) {
+						ModLoader.ModLoader.OnSuccessfulLoad += PlayReload();
+						ModLoader.ModLoader.Reload();
+						return;
+					}
+	
+					ConfigManager.OnChangedAll();
 					WorldGen.playWorld();
+				}
 			}
+		}
+
+		internal static Action PlayReload() {
+			// Main.ActivePlayerFileData gets cleared during reload
+			var path = Main.ActivePlayerFileData.Path;
+			var isCloudSave = Main.ActivePlayerFileData.IsCloudSave;
+			return () => {
+				// re-select the current player
+				Player.GetFileData(path, isCloudSave).SetAsActive();
+				WorldGen.playWorld();
+			};
 		}
 
 		private bool TryMovingToRejectionMenuIfNeeded(int worldGameMode) {
@@ -290,6 +_,9 @@
 		private void SeedButtonClick(UIMouseEvent evt, UIElement listeningElement) {
 			Platform.Get<IClipboard>().Value = _data.GetFullSeedText();
 			_buttonLabel.SetText(Language.GetTextValue("UI.SeedCopied"));
+		}
+
+		private void ConfigButtonClick(UIMouseEvent evt, UIElement listeningElement) {
 		}
 
 		public override int CompareTo(object obj) {
