--- src/decompiled/Terraria/Social/WeGame/CoreSocialModule.cs
+++ src/Terraria/Terraria/Social/WeGame/CoreSocialModule.cs
@@ -20,10 +_,14 @@
 		public void Initialize() {
 			RailGameID railGameID = new RailGameID();
 			railGameID.id_ = 2000328uL;
-			string[] array = new string[1] {
-				" "
-			};
-
+			string[] array = new string[0];
+			if (!Main.dedServ) {
+				array = new string[1] {
+					" "
+				};
+			} else {
+				array = Environment.GetCommandLineArgs();
+			}
 			if (rail_api.RailNeedRestartAppForCheckingEnvironment(railGameID, array.Length, array))
 				Environment.Exit(1);
 
@@ -57,7 +_,12 @@
 		}
 
 		public void Shutdown() {
+#if !WINDOWS
+			isRailValid = false;
+			AppDomain.CurrentDomain.ProcessExit += delegate {
+#else
 			Application.ApplicationExit += delegate {
+#endif
 				isRailValid = false;
 			};
 
