--- src/decompiled/ReLogic/OS/Windows/WindowsPlatform.cs
+++ src/Terraria/ReLogic/OS/Windows/WindowsPlatform.cs
@@ -17,10 +_,8 @@
 		}
 
 		public override void InitializeClientServices(IntPtr windowHandle) {
-			if (_wndProcHook == null)
-				_wndProcHook = new WindowsMessageHook(windowHandle);
-
-			RegisterService((IImeService)new WindowsIme(_wndProcHook, windowHandle));
+			//RegisterService((IImeService)new WindowsIme(_wndProcHook, windowHandle));
+			RegisterService((IImeService)new FnaIme());
 		}
 
 		protected override void Dispose(bool disposing) {
