--- src/decompiled/Terraria/Social/SocialAPI.cs
+++ src/Terraria/Terraria/Social/SocialAPI.cs
@@ -22,7 +_,10 @@
 		public static void Initialize(SocialMode? mode = null) {
 			if (!mode.HasValue) {
 				mode = SocialMode.None;
+				if(!Main.dedServ)
 				mode = SocialMode.Steam;
+				if (Main.dedServ && Program.LaunchParameters.ContainsKey("-steam"))
+					mode = SocialMode.Steam;
 			}
 
 			_mode = mode.Value;
@@ -71,8 +_,11 @@
 			Cloud = LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 			Overlay = LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
 			Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
+			if (!Main.dedServ) {
 			Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
 			WeGameHelper.WriteDebugString("LoadSteam modules");
+			} else
+			Network = LoadModule<Terraria.Social.Steam.NetServerSocialModule>();
 		}
 
 		private static void LoadWeGame() {
@@ -80,8 +_,11 @@
 			Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
 			Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
 			Overlay = LoadModule<Terraria.Social.WeGame.OverlaySocialModule>();
+			if (!Main.dedServ) {
 			Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
 			WeGameHelper.WriteDebugString("LoadWeGame modules");
+			} else
+			Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
 		}
 	}
 }
