--- src/TerrariaNetCore/Terraria/IO/PlayerFileData.cs
+++ src/tModLoader/Terraria/IO/PlayerFileData.cs
@@ -2,6 +_,8 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -14,6 +_,9 @@
 	private readonly Stopwatch _timer = new Stopwatch();
 	private bool _isTimerActive;
 
+	// Added by TML.
+	public CustomModDataException customDataFail = null;
+
 	public Player Player {
 		get {
 			return _player;
@@ -69,6 +_,8 @@
 		if (!FileUtilities.MoveToCloud(base.Path, playerPathFromName))
 			return;
 
+		PlayerIO.MoveToCloud(Path, playerPathFromName);
+
 		string fileName = GetFileName(includeExtension: false);
 		string path = Main.PlayerPath + System.IO.Path.DirectorySeparatorChar + fileName + System.IO.Path.DirectorySeparatorChar;
 		if (Directory.Exists(path)) {
@@ -94,6 +_,8 @@
 		if (!FileUtilities.MoveToLocal(base.Path, playerPathFromName))
 			return;
 
+		PlayerIO.MoveToLocal(Path, playerPathFromName);
+
 		string fileName = GetFileName(includeExtension: false);
 		string mapPath = System.IO.Path.Combine(Main.CloudPlayerPath, fileName);
 		foreach (string item in (from path in SocialAPI.Cloud.GetFiles().ToList()
@@ -111,7 +_,9 @@
 
 	private bool MapBelongsToPath(string mapPath, string filePath)
 	{
+		//TML: Added a check for .tmap too.
-		if (!filePath.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase))
+		if (!filePath.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase)
+		&& !filePath.EndsWith(".tmap", StringComparison.CurrentCultureIgnoreCase))
 			return false;
 
 		string value = mapPath.Replace('\\', '/');
