--- src/TerrariaNetCore/Terraria/IO/ResourcePack.cs
+++ src/tModLoader/Terraria/IO/ResourcePack.cs
@@ -94,7 +_,11 @@
 			return _services.Get<IAssetRepository>().Request<Texture2D>("Images/UI/DefaultResourcePackIcon").Value;
 
 		using Stream stream = OpenStream("icon.png");
+
+		/*
 		return _services.Get<AssetReaderCollection>().Read<Texture2D>(stream, ".png");
+		*/
+		return _services.Get<IAssetRepository>().CreateUntracked<Texture2D>(stream, ".png").Value;
 	}
 
 	private void LoadManifest()
