--- src/Terraria/Terraria/IO/WorldFileData.cs
+++ src/tModLoader/Terraria/IO/WorldFileData.cs
@@ -4,6 +_,7 @@
 using System.IO;
 using Terraria.Audio;
 using Terraria.Localization;
+using Terraria.ModLoader.IO;
 using Terraria.Utilities;
 
 namespace Terraria.IO
@@ -138,6 +_,7 @@
 			if (!base.IsCloudSave) {
 				string worldPathFromName = Main.GetWorldPathFromName(Name, cloudSave: true);
 				if (FileUtilities.MoveToCloud(base.Path, worldPathFromName)) {
+					WorldIO.MoveToCloud(Path, worldPathFromName);
 					Main.LocalFavoriteData.ClearEntry(this);
 					_isCloudSave = true;
 					_path = worldPathFromName;
@@ -150,6 +_,7 @@
 			if (base.IsCloudSave) {
 				string worldPathFromName = Main.GetWorldPathFromName(Name, cloudSave: false);
 				if (FileUtilities.MoveToLocal(base.Path, worldPathFromName)) {
+					WorldIO.MoveToLocal(Path, worldPathFromName);
 					Main.CloudFavoritesData.ClearEntry(this);
 					_isCloudSave = false;
 					_path = worldPathFromName;
