--- src/TerrariaNetCore/ReLogic/OS/IWindowService.cs
+++ src/tModLoader/ReLogic/OS/IWindowService.cs
@@ -1,3 +_,4 @@
+using System;
 using Microsoft.Xna.Framework;
 
 namespace ReLogic.OS;
@@ -13,4 +_,23 @@
 	float GetScaling();
 
 	void SetQuickEditEnabled(bool enabled);
+
+	// Added by TML.
+	void HideConsole();
+
+	// Added by TML.
+	void ShowConsole();
+
+	// Added by TML.
+	void ReleaseConsole();
+
+	// Added by TML.
+	void CreateAndRedirectConsole();
+
+	// Added by TML.
+	void SetIcon(GameWindow window)
+	{
+		IntPtr surface = SDL2.SDL.SDL_LoadBMP("Libraries/Native/tModLoader.bmp");
+		SDL2.SDL.SDL_SetWindowIcon(window.Handle, surface);
+	}
 }
