--- src/TerrariaNetCore/Terraria/Social/SocialAPI.cs
+++ src/tModLoader/Terraria/Social/SocialAPI.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using Terraria.ModLoader.Engine;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 using Terraria.Social.WeGame;
@@ -21,21 +_,34 @@
 		public static SocialMode Mode => _mode;
 
 		public static void Initialize(SocialMode? mode = null) {
+			_modules = new List<ISocialModule>();
 			if (!mode.HasValue) {
 				mode = SocialMode.None;
 
 				/*
 				mode = SocialMode.Steam;
 				*/
-				if (!Main.dedServ) // #if !SERVER
-					mode = SocialMode.Steam;
+
+				// Added FamilyShared check as workaround for #881
+				if (!Main.dedServ && !SteamedWraps.FamilyShared) { // #if !SERVER
+#if DEBUG
+					if (!Program.LaunchParameters.ContainsKey("-nosteam")) {
+#endif
+						if (InstallVerifier.DistributionPlatform == DistributionPlatform.Steam) {
+							mode = SocialMode.Steam;
+						}
+#if DEBUG
+					}
+#endif
+				}
+
+				// ^^^ Confusing
 
 				if (Main.dedServ && Program.LaunchParameters.ContainsKey("-steam"))
 					mode = SocialMode.Steam;
 			}
 
 			_mode = mode.Value;
-			_modules = new List<ISocialModule>();
 			JoinRequests = new ServerJoinRequestsManager();
 			Main.OnTickForInternalCodeOnly += JoinRequests.Update;
 			switch (Mode) {
@@ -46,10 +_,9 @@
 					LoadWeGame();
 					break;
 			}
+			//TODO: Allow both APIs to be used at once. Currently, it would reinitiallize some of them.
 
-			foreach (ISocialModule module in _modules) {
-				module.Initialize();
-			}
+			SteamedWraps.Initialize();
 		}
 
 		public static void Shutdown() {
@@ -73,25 +_,45 @@
 		private static void LoadDiscord() {
 		}
 
+		private static bool _steamAPILoaded = false;
-		private static void LoadSteam() {
+		internal static void LoadSteam() {
+			if (_steamAPILoaded) return;
 			LoadModule<Terraria.Social.Steam.CoreSocialModule>();
 			Friends = LoadModule<Terraria.Social.Steam.FriendsSocialModule>();
+			// This seems to freeze the game when steam_appid.txt not present, and steam_appid.txt isn't supposed to be used in release
+			//if (!ModLoader.Engine.Steam.IsSteamApp) {
-			Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+				//Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+			//}
 			Cloud = LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 			Overlay = LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
 			Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
 			Platform = LoadModule<Terraria.Social.Steam.PlatformSocialModule>();
 
-			if (Main.dedServ) {
+			if (Main.dedServ) { // #if SERVER
 				Network = LoadModule<Terraria.Social.Steam.NetServerSocialModule>();
-				return;
+				goto SkipClientCode;
 			}
 
 			Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
+
 			WeGameHelper.WriteDebugString("LoadSteam modules");
+
+			SkipClientCode:
+
+			foreach (ISocialModule module in _modules) {
+				module.Initialize();
+			}
+
+			_steamAPILoaded = true;
+			ModLoader.Engine.Steam.PostInit();
 		}
+
+		private static bool _wegameAPILoaded = false;
 
 		private static void LoadWeGame() {
+			if (_wegameAPILoaded)
+				return;
+
 			LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
 			Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
 			Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
@@ -99,11 +_,19 @@
 
 			if (Main.dedServ) { // #if SERVER
 				Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
-				return;
+				goto SkipClientCode;
 			}
-
+			
 			Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
+
 			WeGameHelper.WriteDebugString("LoadWeGame modules");
+
+			SkipClientCode:
+
+			foreach (ISocialModule module in _modules) {
+				module.Initialize();
+			}
+			_wegameAPILoaded = true;
 		}
 	}
 }
