--- src/TerrariaNetCore/Terraria/Social/SocialAPI.cs
+++ src/tModLoader/Terraria/Social/SocialAPI.cs
@@ -1,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using ReLogic.OS;
+using Terraria.ModLoader.Engine;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 using Terraria.Social.WeGame;
@@ -11,6 +_,7 @@
 {
 	private static SocialMode _mode;
 	public static Terraria.Social.Base.FriendsSocialModule Friends;
+	/// <summary> Unused in tModLoader. Will always be null. Achievements for tModLoader are only stored locally. </summary>
 	public static Terraria.Social.Base.AchievementsSocialModule Achievements;
 	public static Terraria.Social.Base.CloudSocialModule Cloud;
 	public static Terraria.Social.Base.NetSocialModule Network;
@@ -24,8 +_,11 @@
 
 	public static void Initialize(SocialMode? mode = null)
 	{
+
 		if (!mode.HasValue) {
 			mode = SocialMode.None;
+
+			/*
 			if (Main.dedServ) {
 				if (Program.LaunchParameters.ContainsKey("-steam"))
 					mode = SocialMode.Steam;
@@ -33,6 +_,23 @@
 			else {
 				mode = SocialMode.Steam;
 			}
+			*/
+
+			// Added FamilyShared check as workaround for #881
+			if (!Main.dedServ && !SteamedWraps.FamilyShared) {
+#if DEBUG
+				if (!Program.LaunchParameters.ContainsKey("-nosteam")) {
+#endif
+					if (InstallVerifier.DistributionPlatform == DistributionPlatform.Steam) {
+						mode = SocialMode.Steam;
+					}
+#if DEBUG
+				}
+#endif
+			}
+			else if (Program.LaunchParameters.ContainsKey("-steam")) {
+				mode = SocialMode.Steam;
+			}
 		}
 
 		_mode = mode.Value;
@@ -48,9 +_,16 @@
 				break;
 		}
 
+		// Moved to LoadSteam
+		/*
 		foreach (ISocialModule module in _modules) {
 			module.Initialize();
 		}
+		*/
+
+		//TODO: Allow both APIs to be used at once. Currently, it would reinitiallize some of them.
+		SteamedWraps.Initialize();
+		ModLoader.Logging.tML.Info($"Current 1281930 Workshop Folder Directory: {WorkshopHelper.GetWorkshopFolder(ModLoader.Engine.Steam.TMLAppID_t)}");
 	}
 
 	public static void Shutdown()
@@ -80,11 +_,26 @@
 			_ = Environment.Is64BitProcess;
 	}
 
+	private static bool _steamAPILoaded;
+
-	private static void LoadSteam()
+	internal static void LoadSteam()
 	{
+		if (_steamAPILoaded)
+			return;
+
 		LoadModule<Terraria.Social.Steam.CoreSocialModule>();
 		Friends = LoadModule<Terraria.Social.Steam.FriendsSocialModule>();
+
+		// This seems to freeze the game when steam_appid.txt not present,
+		// and that file isn't supposed to be used in release.
+		/*
+		if (ModLoader.Engine.Steam.IsSteamApp)
+			goto SkipAchievements;
+
 		Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+		SkipAchievements:
+		*/
+
 		Cloud = LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 		Overlay = LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
 		Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
@@ -94,11 +_,27 @@
 		else
 			Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
 
+		/*
 		WeGameHelper.WriteDebugString("LoadSteam modules");
+		*/
+
+		//TML: Moved here from Initialize()
+		foreach (ISocialModule module in _modules) {
+			module.Initialize();
+		}
+
+		_steamAPILoaded = true;
+		ModLoader.Engine.Steam.PostInit();
 	}
+
+	// Added by TML.
+	private static bool _wegameAPILoaded;
 
 	private static void LoadWeGame()
 	{
+		if (_wegameAPILoaded)
+			return;
+
 		LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
 		Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
 		Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
@@ -109,5 +_,12 @@
 			Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
 
 		WeGameHelper.WriteDebugString("LoadWeGame modules");
+
+		//TML: Moved here from Initialize()
+		foreach (ISocialModule module in _modules) {
+			module.Initialize();
+		}
+
+		_wegameAPILoaded = true;
 	}
 }
