--- src/TerrariaNetCore/Terraria/Social/SocialAPI.cs
+++ src/tModLoader/Terraria/Social/SocialAPI.cs
@@ -1,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using ReLogic.OS;
+using Terraria.ModLoader.Engine;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 using Terraria.Social.WeGame;
@@ -24,19 +_,27 @@
 
 	public static void Initialize(SocialMode? mode = null)
 	{
+		_modules = new List<ISocialModule>();
 		if (!mode.HasValue) {
 			mode = SocialMode.None;
-			if (Main.dedServ) {
-				if (Program.LaunchParameters.ContainsKey("-steam"))
-					mode = SocialMode.Steam;
+
+			// Added FamilyShared check as workaround for #881
+			if (!Main.dedServ && !SteamedWraps.FamilyShared) { // #if !SERVER
+#if DEBUG
+				if (!Program.LaunchParameters.ContainsKey("-nosteam")) {
+#endif
+					if (InstallVerifier.DistributionPlatform == DistributionPlatform.Steam) {
+						mode = SocialMode.Steam;
+					}
+#if DEBUG
+				}
+#endif
 			}
-			else {
+			else if (Program.LaunchParameters.ContainsKey("-steam"))
 				mode = SocialMode.Steam;
-			}
 		}
 
 		_mode = mode.Value;
-		_modules = new List<ISocialModule>();
 		JoinRequests = new ServerJoinRequestsManager();
 		Main.OnTickForInternalCodeOnly += JoinRequests.Update;
 		switch (Mode) {
@@ -48,9 +_,8 @@
 				break;
 		}
 
-		foreach (ISocialModule module in _modules) {
-			module.Initialize();
-		}
+		//TODO: Allow both APIs to be used at once. Currently, it would reinitiallize some of them.
+		SteamedWraps.Initialize();
 	}
 
 	public static void Shutdown()
@@ -80,11 +_,19 @@
 			_ = Environment.Is64BitProcess;
 	}
 
+	private static bool _steamAPILoaded;
+
-	private static void LoadSteam()
+	internal static void LoadSteam()
 	{
+		if (_steamAPILoaded)
+			return;
+
 		LoadModule<Terraria.Social.Steam.CoreSocialModule>();
 		Friends = LoadModule<Terraria.Social.Steam.FriendsSocialModule>();
+		// This seems to freeze the game when steam_appid.txt not present, and steam_appid.txt isn't supposed to be used in release
+		//if (!ModLoader.Engine.Steam.IsSteamApp) {
-		Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+			//Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+		//}
 		Cloud = LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 		Overlay = LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
 		Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
@@ -94,11 +_,26 @@
 		else
 			Network = LoadModule<Terraria.Social.Steam.NetClientSocialModule>();
 
+		/*
 		WeGameHelper.WriteDebugString("LoadSteam modules");
+		*/
+
+		foreach (ISocialModule module in _modules) {
+			module.Initialize();
+		}
+
+		_steamAPILoaded = true;
+		ModLoader.Engine.Steam.PostInit();
 	}
+
+	// Added by TML.
+	private static bool _wegameAPILoaded;
 
 	private static void LoadWeGame()
 	{
+		if (_wegameAPILoaded)
+			return;
+
 		LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
 		Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
 		Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
@@ -109,5 +_,11 @@
 			Network = LoadModule<Terraria.Social.WeGame.NetClientSocialModule>();
 
 		WeGameHelper.WriteDebugString("LoadWeGame modules");
+
+		foreach (ISocialModule module in _modules) {
+			module.Initialize();
+		}
+
+		_wegameAPILoaded = true;
 	}
 }
