--- src/Terraria/Terraria/Social/SocialAPI.cs
+++ src/tModLoader/Terraria/Social/SocialAPI.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using Terraria.ModLoader.Engine;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 using Terraria.Social.WeGame;
@@ -20,16 +_,21 @@
 		public static SocialMode Mode => _mode;
 
 		public static void Initialize(SocialMode? mode = null) {
+			_modules = new List<ISocialModule>();
 			if (!mode.HasValue) {
 				mode = SocialMode.None;
 				if(!Main.dedServ)
+#if DEBUG
+					if (!Program.LaunchParameters.ContainsKey("-nosteam"))
+#endif
+						if (InstallVerifier.IsValid && InstallVerifier.IsSteam)
-				mode = SocialMode.Steam;
+							mode = SocialMode.Steam;
+
 				if (Main.dedServ && Program.LaunchParameters.ContainsKey("-steam"))
 					mode = SocialMode.Steam;
 			}
 
 			_mode = mode.Value;
-			_modules = new List<ISocialModule>();
 			JoinRequests = new ServerJoinRequestsManager();
 			Main.OnTickForInternalCodeOnly += JoinRequests.Update;
 			switch (Mode) {
@@ -40,10 +_,7 @@
 					LoadWeGame();
 					break;
 			}
-
-			foreach (ISocialModule module in _modules) {
-				module.Initialize();
-			}
+			//TODO: Allow both APIs to be used at once. Currently, it would reinitiallize some of them.
 		}
 
 		public static void Shutdown() {
@@ -67,10 +_,15 @@
 		private static void LoadDiscord() {
 		}
 
+		private static bool _steamAPILoaded = false;
-		private static void LoadSteam() {
+		internal static void LoadSteam() {
+			if (_steamAPILoaded) return;
 			LoadModule<Terraria.Social.Steam.CoreSocialModule>();
 			Friends = LoadModule<Terraria.Social.Steam.FriendsSocialModule>();
+			// This seems to freeze the game when steam_appid.txt not present, and steam_appid.txt isn't supposed to be used in release
+			//if (!ModLoader.Engine.Steam.IsSteamApp) {
-			Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+				//Achievements = LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
+			//}
 			Cloud = LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 			Overlay = LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
 			Workshop = LoadModule<Terraria.Social.Steam.WorkshopSocialModule>();
@@ -79,9 +_,15 @@
 			WeGameHelper.WriteDebugString("LoadSteam modules");
 			} else
 			Network = LoadModule<Terraria.Social.Steam.NetServerSocialModule>();
+			foreach (ISocialModule module in _modules) {
+				module.Initialize();
+			}
+			_steamAPILoaded = true;
 		}
 
+		private static bool _wegameAPILoaded = false;
 		private static void LoadWeGame() {
+			if(_wegameAPILoaded) return;
 			LoadModule<Terraria.Social.WeGame.CoreSocialModule>();
 			Cloud = LoadModule<Terraria.Social.WeGame.CloudSocialModule>();
 			Friends = LoadModule<Terraria.Social.WeGame.FriendsSocialModule>();
@@ -91,6 +_,10 @@
 			WeGameHelper.WriteDebugString("LoadWeGame modules");
 			} else
 			Network = LoadModule<Terraria.Social.WeGame.NetServerSocialModule>();
+			foreach (ISocialModule module in _modules) {
+				module.Initialize();
+			}
+			_wegameAPILoaded = true;
 		}
 	}
 }
