--- src/TerrariaNetCore/Terraria/GameContent/LootSimulation/LootSimulator.cs
+++ src/tModLoader/Terraria/GameContent/LootSimulation/LootSimulator.cs
@@ -3,6 +_,7 @@
 using System.Linq;
 using ReLogic.OS;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.LootSimulation;
 
@@ -59,7 +_,7 @@
 		string text = "";
 		Stopwatch stopwatch = new Stopwatch();
 		stopwatch.Start();
-		for (int i = -65; i < NPCID.Count; i++) {
+		for (int i = -65; i < NPCLoader.NPCCount; i++) {
 			if (TryGettingLootFor(i, timesMultiplier, out var outputText))
 				text = text + outputText + "\n\n";
 		}
