--- src/Terraria/Terraria/Net/Sockets/SocialSocket.cs
+++ src/TerrariaNetCore/Terraria/Net/Sockets/SocialSocket.cs
@@ -1,4 +_,5 @@
 using System.Threading;
+using System.Threading.Tasks;
 using Terraria.Social;
 using Terraria.Testing;
 
@@ -40,7 +_,11 @@
 		Invariant.Assert(_remoteAddress != null, "Send cannot be called on a null address");
 		Invariant.Assert(offset == 0, "Non-Zero offsets are not yet supported for Social Network Sends");
 		SocialAPI.Network.Send(_remoteAddress, data, size);
+#if NETCORE
+		Task.Run(() => callback.Invoke(state));
+#else
 		callback.BeginInvoke(state, null, null);
+#endif
 	}
 
 	private void ReadCallback(byte[] data, int offset, int size, SocketReceiveCallback callback, object state)
@@ -56,7 +_,11 @@
 	void ISocket.AsyncReceive(byte[] data, int offset, int size, SocketReceiveCallback callback, object state)
 	{
 		Invariant.Assert(_remoteAddress != null, "Receive cannot be called on a null address");
+#if NETCORE
+		Task.Run(() => new InternalReadCallback(ReadCallback).Invoke(data, offset, size, callback, state));
+#else
 		new InternalReadCallback(ReadCallback).BeginInvoke(data, offset, size, callback, state, null, null);
+#endif
 	}
 
 	bool ISocket.IsDataAvailable() => SocialAPI.Network.IsDataAvailable(_remoteAddress);
