--- src/Terraria/Configuration.targets
+++ src/tModLoader/Configuration.targets
@@ -1,6 +_,9 @@
 <Project>
   <PropertyGroup>
     <TargetFramework>net5.0</TargetFramework>
+    <!-- Defines the Minimum patch version that is required to play. pre v5.0.5 has bugs in release for windows runtime-->
+    <!-- If changing this value, also need to update Main.cs: tServer.StartInfo.FileName = Path.Combine("dotnet", "5.0.5", "dotnet");-->
+    <RuntimeFrameworkVersion>5.0.5</RuntimeFrameworkVersion>
     <Configurations>Debug;Release</Configurations>
     <PlatformTarget>AnyCPU</PlatformTarget>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
