--- src/Terraria/Configuration.targets
+++ src/tModLoader/Configuration.targets
@@ -1,6 +_,6 @@
 <Project>
   <PropertyGroup>
-    <TargetFramework>net40</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
     <Configurations>WindowsDebug;WindowsRelease;WindowsServerDebug;WindowsServerRelease;LinuxRelease;LinuxServerRelease;MacRelease;MacServerRelease</Configurations>
     <PlatformTarget>x86</PlatformTarget>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
