--- src/Terraria\TerrariaMac.csproj
+++ src/tModLoader\TerrariaMac.csproj
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <ProjectGuid>{E3BD64F8-9FAE-4778-9C5D-692E041180D0}</ProjectGuid>
+    <ProjectGuid>{9B1110E5-84CB-43D9-88FC-A5ABA3E09D61}</ProjectGuid>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
     <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
     <OutputType>Exe</OutputType>
@@ -28,6 +28,7 @@
     <DefineConstants>CLIENT; MAC</DefineConstants>
     <PlatformTarget>x86</PlatformTarget>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
     <DefineConstants>CLIENT; MAC</DefineConstants>
@@ -40,6 +41,10 @@
     <Reference Include="Ionic.Zip.CF, Version=1.9.1.8, Culture=neutral, PublicKeyToken=edbe51ad942a3f5c, processorArchitecture=MSIL">
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\references\Ionic.Zip.CF.dll</HintPath>
+    </Reference>
+    <Reference Include="MP3Sharp, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\references\MP3Sharp.dll</HintPath>
     </Reference>
     <Reference Include="Newtonsoft.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
       <SpecificVersion>False</SpecificVersion>
@@ -53,6 +58,7 @@
     <Reference Include="System" />
     <Reference Include="System.Drawing" />
     <Reference Include="System.Windows.Forms" />
+    <Reference Include="System.XML" />
     <Reference Include="WindowsBase" />
   </ItemGroup>
   <ItemGroup>
@@ -211,6 +217,7 @@
     <Compile Include="Terraria.ID\ItemID.cs" />
     <Compile Include="Terraria.ID\MessageID.cs" />
     <Compile Include="Terraria.ID\MountID.cs" />
+    <Compile Include="Terraria.ID\MusicID.cs" />
     <Compile Include="Terraria.ID\NPCID.cs" />
     <Compile Include="Terraria.ID\PlayerTextureID.cs" />
     <Compile Include="Terraria.ID\PlayerVariantID.cs" />
@@ -238,6 +245,80 @@
     <Compile Include="Terraria.Map\MapHelper.cs" />
     <Compile Include="Terraria.Map\MapTile.cs" />
     <Compile Include="Terraria.Map\WorldMap.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModLoaderMod.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryPlayer.cs" />
+    <Compile Include="Terraria.ModLoader.Default\StartBag.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\DuplicateNameException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\MissingResourceException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\RecipeException.cs" />
+    <Compile Include="Terraria.ModLoader.IO\ItemIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\MapIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\PlayerIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TileIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TmodFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\UploadFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\WorldIO.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIBuildMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIColorTextPanel.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIDownloadMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIErrorMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMessageBox.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModBrowser.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModControls.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModDownloadItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModInfo.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSourceItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSources.cs" />
+    <Compile Include="Terraria.ModLoader\BuffLoader.cs" />
+    <Compile Include="Terraria.ModLoader\BuildProperties.cs" />
+    <Compile Include="Terraria.ModLoader\CraftGroup.cs" />
+    <Compile Include="Terraria.ModLoader\EquipLoader.cs" />
+    <Compile Include="Terraria.ModLoader\EquipTexture.cs" />
+    <Compile Include="Terraria.ModLoader\EquipType.cs" />
+    <Compile Include="Terraria.ModLoader\ErrorLogger.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalNPC.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalTile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalWall.cs" />
+    <Compile Include="Terraria.ModLoader\Interface.cs" />
+    <Compile Include="Terraria.ModLoader\ItemLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MapEntry.cs" />
+    <Compile Include="Terraria.ModLoader\MapLegend.cs" />
+    <Compile Include="Terraria.ModLoader\MapLoader.cs" />
+    <Compile Include="Terraria.ModLoader\Mod.cs" />
+    <Compile Include="Terraria.ModLoader\ModBuff.cs" />
+    <Compile Include="Terraria.ModLoader\ModDust.cs" />
+    <Compile Include="Terraria.ModLoader\ModGore.cs" />
+    <Compile Include="Terraria.ModLoader\ModHooks.cs" />
+    <Compile Include="Terraria.ModLoader\ModItem.cs" />
+    <Compile Include="Terraria.ModLoader\ModLoader.cs" />
+    <Compile Include="Terraria.ModLoader\ModMountData.cs" />
+    <Compile Include="Terraria.ModLoader\ModNPC.cs" />
+    <Compile Include="Terraria.ModLoader\ModPlayer.cs" />
+    <Compile Include="Terraria.ModLoader\ModProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\ModProperties.cs" />
+    <Compile Include="Terraria.ModLoader\ModRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\ModSound.cs" />
+    <Compile Include="Terraria.ModLoader\ModTile.cs" />
+    <Compile Include="Terraria.ModLoader\ModWall.cs" />
+    <Compile Include="Terraria.ModLoader\MountLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MountTextureType.cs" />
+    <Compile Include="Terraria.ModLoader\MusicWrapper.cs" />
+    <Compile Include="Terraria.ModLoader\NPCHeadLoader.cs" />
+    <Compile Include="Terraria.ModLoader\NPCLoader.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnInfo.cs" />
+    <Compile Include="Terraria.ModLoader\PlayerHooks.cs" />
+    <Compile Include="Terraria.ModLoader\ProjectileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundType.cs" />
+    <Compile Include="Terraria.ModLoader\TileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WallLoader.cs" />
     <Compile Include="Terraria.Modules\AnchorDataModule.cs" />
     <Compile Include="Terraria.Modules\AnchorTypesModule.cs" />
     <Compile Include="Terraria.Modules\LiquidDeathModule.cs" />
@@ -377,4 +458,7 @@
     <Compile Include="Terraria\WorldSections.cs" />
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <PropertyGroup>
+    <PostBuildEvent>if "$(TerrariaSteamFolder)" == "" (copy "$(TargetPath)" "C:\Program Files (x86)\Steam\steamapps\common\terraria") else (copy "$(TargetPath)" "$(TerrariaSteamFolder)")</PostBuildEvent>
+  </PropertyGroup>
+</Project>
