--- src/Terraria\Terraria.csproj
+++ src/tModLoader\Terraria.csproj
@@ -28,12 +28,18 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
     <DefineConstants>CLIENT; WINDOWS</DefineConstants>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
     <DefineConstants>CLIENT; WINDOWS</DefineConstants>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+  <PropertyGroup>
+    <ApplicationIcon>Icon.ico</ApplicationIcon>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.CF">
+    <Reference Include="Ionic.Zip.CF, Version=1.9.1.8, Culture=neutral, PublicKeyToken=edbe51ad942a3f5c, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
       <HintPath>Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.Xna.Framework.Game" />
@@ -41,7 +47,8 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll</HintPath>
     </Reference>
-    <Reference Include="Steamworks.NET">
+    <Reference Include="Steamworks.NET, Version=7.0.0.0, Culture=neutral, processorArchitecture=x86">
+      <SpecificVersion>False</SpecificVersion>
       <HintPath>Terraria.Libraries\Steamworks.NET\Steamworks.NET.dll</HintPath>
     </Reference>
     <Reference Include="System.Windows.Forms" />
@@ -51,11 +58,17 @@
     <Reference Include="Microsoft.Xna.Framework.Xact" />
     <Reference Include="System.Core" />
     <Reference Include="System.Drawing" />
+    <Reference Include="System.XML" />
     <Reference Include="WindowsBase" />
   </ItemGroup>
   <ItemGroup>
     <Compile Include="Extensions\EnumerationExtensions.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="Properties\Resources.Designer.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>Resources.resx</DependentUpon>
+    </Compile>
     <Compile Include="Terraria.Achievements\Achievement.cs" />
     <Compile Include="Terraria.Achievements\AchievementCategory.cs" />
     <Compile Include="Terraria.Achievements\AchievementCondition.cs" />
@@ -236,6 +249,72 @@
     <Compile Include="Terraria.Map\MapHelper.cs" />
     <Compile Include="Terraria.Map\MapTile.cs" />
     <Compile Include="Terraria.Map\WorldMap.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModLoaderMod.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryItem.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\DuplicateNameException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\MissingResourceException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\RecipeException.cs" />
+    <Compile Include="Terraria.ModLoader.IO\ItemIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\MapIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TileIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TmodFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\UploadFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\WorldIO.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIBuildMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIColorTextPanel.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIErrorMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMessageBox.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModBrowser.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModDownloadItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModInfo.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSourceItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSources.cs" />
+    <Compile Include="Terraria.ModLoader\BuffLoader.cs" />
+    <Compile Include="Terraria.ModLoader\BuildProperties.cs" />
+    <Compile Include="Terraria.ModLoader\CraftGroup.cs" />
+    <Compile Include="Terraria.ModLoader\EquipLoader.cs" />
+    <Compile Include="Terraria.ModLoader\EquipType.cs" />
+    <Compile Include="Terraria.ModLoader\ErrorLogger.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalNPC.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalTile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalWall.cs" />
+    <Compile Include="Terraria.ModLoader\Interface.cs" />
+    <Compile Include="Terraria.ModLoader\ItemLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MapEntry.cs" />
+    <Compile Include="Terraria.ModLoader\MapLegend.cs" />
+    <Compile Include="Terraria.ModLoader\MapLoader.cs" />
+    <Compile Include="Terraria.ModLoader\Mod.cs" />
+    <Compile Include="Terraria.ModLoader\ModBuff.cs" />
+    <Compile Include="Terraria.ModLoader\ModDust.cs" />
+    <Compile Include="Terraria.ModLoader\ModGore.cs" />
+    <Compile Include="Terraria.ModLoader\ModHooks.cs" />
+    <Compile Include="Terraria.ModLoader\ModItem.cs" />
+    <Compile Include="Terraria.ModLoader\ModLoader.cs" />
+    <Compile Include="Terraria.ModLoader\ModMountData.cs" />
+    <Compile Include="Terraria.ModLoader\ModNPC.cs" />
+    <Compile Include="Terraria.ModLoader\ModProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\ModProperties.cs" />
+    <Compile Include="Terraria.ModLoader\ModRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\ModSound.cs" />
+    <Compile Include="Terraria.ModLoader\ModTile.cs" />
+    <Compile Include="Terraria.ModLoader\ModWall.cs" />
+    <Compile Include="Terraria.ModLoader\MountLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MountTextureType.cs" />
+    <Compile Include="Terraria.ModLoader\NPCHeadLoader.cs" />
+    <Compile Include="Terraria.ModLoader\NPCLoader.cs" />
+    <Compile Include="Terraria.ModLoader.IO\PlayerIO.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnInfo.cs" />
+    <Compile Include="Terraria.ModLoader\ProjectileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundType.cs" />
+    <Compile Include="Terraria.ModLoader\TileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WallLoader.cs" />
     <Compile Include="Terraria.Modules\AnchorDataModule.cs" />
     <Compile Include="Terraria.Modules\AnchorTypesModule.cs" />
     <Compile Include="Terraria.Modules\LiquidDeathModule.cs" />
@@ -379,9 +458,20 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft.Xna.Framework.RuntimeProfile" />
+    <EmbeddedResource Include="Properties\Resources.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
     <EmbeddedResource Include="Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll" />
     <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll" />
     <EmbeddedResource Include="Terraria.Libraries\Steamworks.NET\Steamworks.NET.dll" />
   </ItemGroup>
+  <ItemGroup>
+    <Content Include="Icon.ico" />
+    <None Include="Terraria.ModLoader.Default\MysteryItem.png" />
+  </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <PropertyGroup>
+    <PostBuildEvent>if "$(TerrariaSteamFolder)" == "" (copy "$(TargetPath)" "C:\Program Files (x86)\Steam\steamapps\common\terraria") else (copy "$(TargetPath)" "$(TerrariaSteamFolder)")</PostBuildEvent>
+  </PropertyGroup>
 </Project>
