--- src/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria.csproj
@@ -1,7 +_,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <Import Project="..\TerrariaSteamPath.targets" />
   <PropertyGroup>
-    <TargetFramework>net40</TargetFramework>
+    <TargetFramework>net45</TargetFramework>
     <OutputType>WinExe</OutputType>
     <Version>1.3.5.3</Version>
     <Company>Re-Logic</Company>
@@ -10,9 +_,10 @@
     <Configurations>WindowsDebug;WindowsServerDebug;WindowsRelease;WindowsServerRelease;MacRelease;MacServerRelease;LinuxRelease;LinuxServerRelease</Configurations>
     <AssemblySearchPaths>$(AssemblySearchPaths);{GAC}</AssemblySearchPaths>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <LangVersion>7.3</LangVersion>
     <PlatformTarget>AnyCPU</PlatformTarget>
     <Optimize>true</Optimize>
-	<OutputName>Terraria</OutputName>
+    <OutputName>tModLoader</OutputName>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Server'))">
     <OutputType>Exe</OutputType>
@@ -40,13 +_,16 @@
   </PropertyGroup>
   <PropertyGroup>
     <PdbFile>bin\$(OutputName)</PdbFile>
-    <ApplicationIcon>Icon.ico</ApplicationIcon>
+    <ApplicationIcon>tModLoader.ico</ApplicationIcon>
   </PropertyGroup>
   <ItemGroup>
     <Reference Include="FNA" Condition="$(DefineConstants.Contains('FNA'))">
       <HintPath>..\..\references\FNA.dll</HintPath>
     </Reference>
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.Reduced">
+      <HintPath>..\..\references\Ionic.Zip.Reduced.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.CSharp" />
     <Reference Include="Microsoft.Xna.Framework" Condition="$(DefineConstants.Contains('XNA'))">
       <Private>true</Private>
     </Reference>
@@ -59,6 +_,30 @@
     <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('XNA'))">
       <Private>true</Private>
     </Reference>
+    <Reference Include="Mono.Cecil">
+      <HintPath>..\..\references\Mono.Cecil.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil.Pdb">
+      <HintPath>..\..\references\Mono.Cecil.Pdb.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil.Mdb">
+      <HintPath>..\..\references\Mono.Cecil.Mdb.dll</HintPath>
+    </Reference>
+    <Reference Include="MP3Sharp">
+      <HintPath>..\..\references\MP3Sharp.dll</HintPath>
+    </Reference>
+    <Reference Include="NVorbis">
+      <HintPath>..\..\references\NVorbis.dll</HintPath>
+    </Reference>
+    <Reference Include="log4net">
+      <HintPath>..\..\references\log4net.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.RuntimeDetour">
+      <HintPath>..\..\references\MonoMod.RuntimeDetour.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.Utils">
+      <HintPath>..\..\references\MonoMod.Utils.dll</HintPath>
+    </Reference>
     <Reference Include="Newtonsoft.Json" />
     <Reference Include="ReLogic">
       <HintPath Condition="$(DefineConstants.Contains('WINDOWS'))">..\..\references\ReLogicWindows.dll</HintPath>
@@ -73,6 +_,7 @@
     <Reference Include="System" />
     <Reference Include="System.Core" />
     <Reference Include="System.Drawing" />
+    <Reference Include="System.XML" />
     <Reference Include="System.Windows.Forms">
       <HintPath Condition="!$(DefineConstants.Contains('WINDOWS'))">..\..\references\System.Windows.Forms.Mono.dll</HintPath>
     </Reference>
@@ -80,8 +_,11 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft.Xna.Framework.RuntimeProfile" />
-    <EmbeddedResource Include="Terraria.Libraries/**" />
+    <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll" >
+		<LogicalName>Newtonsoft.Json.dll</LogicalName>
+	</EmbeddedResource>
     <EmbeddedResource Include="Terraria.Localization.Content/*" />
+    <EmbeddedResource Include="**\*.png" />
     <EmbeddedResource Include="..\..\references\ReLogicWindows.dll" Condition="$(DefineConstants.Contains('WINDOWS'))">
       <LogicalName>ReLogic.dll</LogicalName>
     </EmbeddedResource>
@@ -100,6 +_,27 @@
     <EmbeddedResource Include="..\..\references\Steamworks.NET.Linux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
       <LogicalName>Steamworks.NET.dll</LogicalName>
     </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\TerrariaHooks.XNA.dll" Condition="$(DefineConstants.Contains('XNA'))">
+      <LogicalName>TerrariaHooks.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\TerrariaHooks.FNA.dll" Condition="$(DefineConstants.Contains('FNA'))">
+      <LogicalName>TerrariaHooks.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\Ionic.Zip.Reduced.dll" />
+    <EmbeddedResource Include="..\..\references\MP3Sharp.dll" />
+    <EmbeddedResource Include="..\..\references\NVorbis.dll" />
+    <EmbeddedResource Include="..\..\references\log4net.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.Pdb.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.Mdb.dll" />
+    <EmbeddedResource Include="..\..\references\MonoMod.RuntimeDetour.dll" />
+    <EmbeddedResource Include="..\..\references\MonoMod.Utils.dll" />
+    <EmbeddedResource Include="..\..\references\System.ValueTuple.Windows.dll" Condition="$(DefineConstants.Contains('WINDOWS'))">
+      <LogicalName>System.ValueTuple.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\System.ValueTuple.Mono.dll" Condition="!$(DefineConstants.Contains('WINDOWS'))">
+      <LogicalName>System.ValueTuple.dll</LogicalName>
+    </EmbeddedResource>
   </ItemGroup>
   <Target Name="EditBin" AfterTargets="Build">
     <PropertyGroup>
@@ -116,7 +_,11 @@
   </Target>
   <Target Name="CopyToStreamDir" AfterTargets="EditBin" Condition="$(Configuration.Contains('Windows'))">
     <Message Text="Copying $(OutputName) to Steam Dir..." Importance="high" />
-    <Copy SourceFiles="$(TargetPath)" DestinationFiles="$(TerrariaSteamPath)\$(OutputName).exe" />
+    <Copy SourceFiles="$(TargetPath)" DestinationFiles="$(tModLoaderSteamPath)\$(OutputName).exe" />
-    <Copy SourceFiles="$(TargetDir)$(OutputName).pdb" DestinationFolder="$(TerrariaSteamPath)" />
+    <Copy SourceFiles="$(TargetDir)$(OutputName).pdb" DestinationFolder="$(tModLoaderSteamPath)" />
+  </Target>
+  <Target Name="CopyToModCompile" AfterTargets="EditBin" Condition="$(Configuration)==MacRelease">
+    <Message Text="Copying tModLoaderMac to ModCompile..." Importance="high" />
+    <Copy SourceFiles="$(TargetPath)" DestinationFiles="$(tModLoaderSteamPath)\ModCompile\tModLoader.FNA.exe" />
   </Target>
 </Project>
