--- src/Terraria\Terraria.csproj
+++ src/tModLoader\Terraria.csproj
@@ -3,22 +_,24 @@
   <PropertyGroup>
     <ProjectGuid>{3996D5FA-6E59-4FE4-9F2B-40EEEF9645D5}</ProjectGuid>
     <AssemblyName>Terraria</AssemblyName>
-    <TargetFramework>net40</TargetFramework>
+    <TargetFramework>net45</TargetFramework>
     <Version>1.3.5.1</Version>
     <Company>Re-Logic</Company>
     <Copyright>Copyright © Re-Logic 2017</Copyright>
     <RootNamespace />
     <Configurations>WindowsDebug;WindowsServerDebug;WindowsRelease;WindowsServerRelease;MacRelease;MacServerRelease;LinuxRelease;LinuxServerRelease</Configurations>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <LangVersion>latest</LangVersion>
+    <LangVersion>7.3</LangVersion>
     <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
     <OutputType>WinExe</OutputType>
     <PlatformTarget>AnyCPU</PlatformTarget>
     <Optimize>true</Optimize>
+    <OutputName>tModLoader</OutputName>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Server'))">
     <OutputType>Exe</OutputType>
     <DefineConstants>$(DefineConstants);SERVER</DefineConstants>
+    <OutputName>$(OutputName)Server</OutputName>
   </PropertyGroup>
   <PropertyGroup Condition="!$(Configuration.Contains('Server'))">
     <DefineConstants>$(DefineConstants);CLIENT</DefineConstants>
@@ -26,6 +_,7 @@
   <PropertyGroup Condition="$(Configuration.Contains('Debug'))">
     <Optimize>false</Optimize>
     <DefineConstants>$(DefineConstants);DEBUG</DefineConstants>
+    <OutputName>$(OutputName)Debug</OutputName>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Windows'))">
     <PlatformTarget>x86</PlatformTarget>
@@ -38,13 +_,31 @@
   <PropertyGroup Condition="$(Configuration.Contains('Linux'))">
     <DefineConstants>$(DefineConstants);LINUX;FNA</DefineConstants>
   </PropertyGroup>
+  <PropertyGroup>
+    <PdbFile>bin\$(OutputName)</PdbFile>
+    <ApplicationIcon>Icon.ico</ApplicationIcon>
+  </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Xna.Framework.Game" Condition="$(DefineConstants.Contains('XNA'))" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" Condition="$(DefineConstants.Contains('XNA'))" />
-    <Reference Include="Microsoft.Xna.Framework" Condition="$(DefineConstants.Contains('XNA'))" />
-    <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('XNA'))" />
+    <None Remove="Terraria.ModLoader.UI\ButtonDowngrade.png" />
+    <None Remove="Terraria.ModLoader.UI\ButtonModConfig.png" />
+    <None Remove="Terraria.ModLoader.UI\ButtonModInfo.png" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="Microsoft.Xna.Framework.Game" Condition="$(DefineConstants.Contains('XNA'))">
+      <Private>true</Private>
+    </Reference>
+    <Reference Include="Microsoft.Xna.Framework.Graphics" Condition="$(DefineConstants.Contains('XNA'))">
+      <Private>true</Private>
+    </Reference>
+    <Reference Include="Microsoft.Xna.Framework" Condition="$(DefineConstants.Contains('XNA'))">
+      <Private>true</Private>
+    </Reference>
+    <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('XNA'))">
+      <Private>true</Private>
+    </Reference>
     <Reference Include="Newtonsoft.Json" />
     <Reference Include="System" />
+    <Reference Include="System.XML" />
     <Reference Include="Steamworks.NET">
       <HintPath Condition="$(DefineConstants.Contains('WINDOWS'))">..\..\references\Steamworks.NET.Windows.dll</HintPath>
       <HintPath Condition="$(DefineConstants.Contains('MAC'))">..\..\references\Steamworks.NET.OSX.dll</HintPath>
@@ -57,11 +_,35 @@
     </Reference>
     <Reference Include="System.Core" />
     <Reference Include="System.Drawing" />
-    <Reference Include="Ionic.Zip.CF">
-      <HintPath>Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll</HintPath>
+    <Reference Include="Ionic.Zip.Reduced">
+      <HintPath>..\..\references\Ionic.Zip.Reduced.dll</HintPath>
     </Reference>
     <Reference Include="FNA" Condition="$(DefineConstants.Contains('FNA'))">
       <HintPath>..\..\references\FNA.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil">
+      <HintPath>..\..\references\Mono.Cecil.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil.Pdb">
+      <HintPath>..\..\references\Mono.Cecil.Pdb.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil.Mdb">
+      <HintPath>..\..\references\Mono.Cecil.Mdb.dll</HintPath>
+    </Reference>
+    <Reference Include="MP3Sharp">
+      <HintPath>..\..\references\MP3Sharp.dll</HintPath>
+    </Reference>
+    <Reference Include="NVorbis">
+      <HintPath>..\..\references\NVorbis.dll</HintPath>
+    </Reference>
+    <Reference Include="log4net">
+      <HintPath>..\..\references\log4net.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.RuntimeDetour">
+      <HintPath>..\..\references\MonoMod.RuntimeDetour.dll</HintPath>
+    </Reference>
+    <Reference Include="MonoMod.Utils">
+      <HintPath>..\..\references\MonoMod.Utils.dll</HintPath>
     </Reference>
     <Reference Include="System.Windows.Forms">
       <HintPath Condition="!$(DefineConstants.Contains('WINDOWS'))">..\..\references\System.Windows.Forms.Mono.dll</HintPath>
@@ -297,9 +_,15 @@
     <Compile Include="Terraria.ID\GoreID.cs" />
     <Compile Include="Terraria.ID\InvasionID.cs" />
     <Compile Include="Terraria.ID\ItemAlternativeFunctionID.cs" />
+    <Compile Include="Terraria.ID\ItemHoldStyleID.cs" />
     <Compile Include="Terraria.ID\ItemID.cs" />
+    <Compile Include="Terraria.ID\ItemRarityID.cs" />
+    <Compile Include="Terraria.ID\ItemUseStyleID.cs" />
+    <Compile Include="Terraria.ID\LangID.cs" />
     <Compile Include="Terraria.ID\MessageID.cs" />
     <Compile Include="Terraria.ID\MountID.cs" />
+    <Compile Include="Terraria.ID\MusicID.cs" />
+    <Compile Include="Terraria.ID\NetmodeID.cs" />
     <Compile Include="Terraria.ID\NPCID.cs" />
     <Compile Include="Terraria.ID\PlayerTextureID.cs" />
     <Compile Include="Terraria.ID\PlayerVariantID.cs" />
@@ -339,6 +_,230 @@
     <Compile Include="Terraria.Map\MapHelper.cs" />
     <Compile Include="Terraria.Map\MapTile.cs" />
     <Compile Include="Terraria.Map\WorldMap.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ArrayElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\BooleanElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\CollectionElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ColorElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ConfigElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\DictionaryElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\EnumElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\FloatElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\IntegerElements.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ItemDefinitionElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ListElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\ObjectElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\PropertyFieldWrapper.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\RangeElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\SetElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\StringInputElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\StringOptionElement.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\UIModConfig.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\UIModConfigList.cs" />
+    <Compile Include="Terraria.ModLoader.Config.UI\Vector2Element.cs" />
+    <Compile Include="Terraria.ModLoader.Config\ItemDefinition.cs" />
+    <Compile Include="Terraria.ModLoader.Config\ConfigAttributes.cs" />
+    <Compile Include="Terraria.ModLoader.Config\ModConfig.cs" />
+    <Compile Include="Terraria.ModLoader.Config\ConfigManager.cs" />
+    <Compile Include="Terraria.ModLoader.Core\AssemblyResolving.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Developer\DeveloperItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Developer\DeveloperPlayer.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Developer\Jofairden.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Dinidini.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Orian.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\PatreonItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\PatronModPlayer.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Polyblank.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Glory.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Squid.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Remeus.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\Saethar.cs" />
+    <Compile Include="Terraria.ModLoader.Default.Patreon\toplayz.cs" />
+    <Compile Include="Terraria.ModLoader.Default\AprilFools.cs" />
+    <Compile Include="Terraria.ModLoader.Default\HelpCommand.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModlistCommand.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModLoaderMod.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryGlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTile.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTileEntity.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTilesWorld.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryWorld.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryPlayer.cs" />
+    <Compile Include="Terraria.ModLoader.Default\NetHandler.cs" />
+    <Compile Include="Terraria.ModLoader.Default\StartBag.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\BuildException.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\GoGVerifier.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\HiDefGraphicsIssues.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\CustomModDataException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\MissingResourceException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\ModSortingException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\OldHookException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\RecipeException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\ResourceLoadException.cs" />
+    <Compile Include="Terraria.ModLoader.IO\BigEndian.cs" />
+    <Compile Include="Terraria.ModLoader.IO\BinaryIO.cs" />
+    <Compile Include="Terraria.ModLoader.Core\ContentConverters.cs" />
+    <Compile Include="Terraria.ModLoader.IO\ImageIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\ItemIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\MapIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagSerializer.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagPrinter.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagSerializable.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TileIO.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\TMLContentManager.cs" />
+    <Compile Include="Terraria.ModLoader.Core\TmodFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\UploadFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\WorldIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagCompound.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagIO.cs" />
+    <Compile Include="Terraria.ModLoader.Core\EntryReadStream.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\DownloadFile.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\DownloadModFile.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UILoadModsProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIDownloadProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIProgressBar.cs" />
+    <Compile Include="Terraria.ModLoader.UI.Elements\FixedUIScrollbar.cs" />
+    <Compile Include="Terraria.ModLoader.UI.Elements\UIGrid.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\ModBrowserSortMode.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\ModBrowserSortModesExtensions.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\ModIconStatus.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\ModSideFilter.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\ModSideFilterModesExtensions.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\SearchFilter.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\SearchFilterModesExtensions.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\TimeHelper.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UIBrowserFilterToggle.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UIModBrowserElements.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UIModDownloadItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UpdateFilter.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UpdateFilterModesExtensions.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIAutoScaleTextTextPanel.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIBuildModProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UICommon.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UICreateMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIDeveloperModeHelp.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIEnterPassphraseMenu.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UICycleImage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIEnterSteamIDMenu.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIErrorMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIHoverImage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIInfoMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIInputTextField.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoaderAnimatedImage.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIExtractModProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIManagePublished.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMemoryBar.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMessageBox.cs" />
+    <Compile Include="Terraria.ModLoader.UI.ModBrowser\UIModBrowser.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModControls.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModInfo.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModPackItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModPacks.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModManageItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSourceItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSources.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIUpdateMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI.DownloadManager\UIUploadModProgress.cs" />
+    <Compile Include="Terraria.ModLoader.Core\AssemblyManager.cs" />
+    <Compile Include="Terraria.ModLoader\AutoloadEquip.cs" />
+    <Compile Include="Terraria.ModLoader\AutoloadHead.cs" />
+    <Compile Include="Terraria.ModLoader\BackgroundStyleLoaders.cs" />
+    <Compile Include="Terraria.ModLoader\BackgroundTextureLoader.cs" />
+    <Compile Include="Terraria.ModLoader\BuffLoader.cs" />
+    <Compile Include="Terraria.ModLoader.Core\BuildProperties.cs" />
+    <Compile Include="Terraria.ModLoader\CombinedHooks.cs" />
+    <Compile Include="Terraria.ModLoader\CommandManager.cs" />
+    <Compile Include="Terraria.ModLoader.Core\CecilDebugHeaderFile.cs" />
+    <Compile Include="Terraria.ModLoader\DrawInfo.cs" />
+    <Compile Include="Terraria.ModLoader\EffectsTracker.cs" />
+    <Compile Include="Terraria.ModLoader\EquipLoader.cs" />
+    <Compile Include="Terraria.ModLoader\EquipTexture.cs" />
+    <Compile Include="Terraria.ModLoader\EquipType.cs" />
+    <Compile Include="Terraria.ModLoader\ErrorLogger.cs" />
+    <Compile Include="Terraria.ModLoader.Core\FileAssociation.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\FallbackSyncContext.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\GLCallLocker.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalBuff.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalNPC.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalTile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalWall.cs" />
+    <Compile Include="Terraria.ModLoader\Interface.cs" />
+    <Compile Include="Terraria.ModLoader\ItemLoader.cs" />
+    <Compile Include="Terraria.ModLoader.Core\LocalMod.cs" />
+    <Compile Include="Terraria.ModLoader\Logging.cs" />
+    <Compile Include="Terraria.ModLoader\MapEntry.cs" />
+    <Compile Include="Terraria.ModLoader\MapLegend.cs" />
+    <Compile Include="Terraria.ModLoader\MapLoader.cs" />
+    <Compile Include="Terraria.ModLoader.Core\MdbManager.cs" />
+    <Compile Include="Terraria.ModLoader.Core\MemoryTracking.cs" />
+    <Compile Include="Terraria.ModLoader\Mod.cs" />
+    <Compile Include="Terraria.ModLoader.IO\BackupIO.cs" />
+    <Compile Include="Terraria.ModLoader\ModInternals.cs" />
+    <Compile Include="Terraria.ModLoader\ModBackgroundStyle.cs" />
+    <Compile Include="Terraria.ModLoader\ModBuff.cs" />
+    <Compile Include="Terraria.ModLoader\ModCommand.cs" />
+    <Compile Include="Terraria.ModLoader.Core\ModCompile.cs" />
+    <Compile Include="Terraria.ModLoader\ConvertiblePlants.cs" />
+    <Compile Include="Terraria.ModLoader\ModContent.cs" />
+    <Compile Include="Terraria.ModLoader\ModDust.cs" />
+    <Compile Include="Terraria.ModLoader\ModGore.cs" />
+    <Compile Include="Terraria.ModLoader\ModHooks.cs" />
+    <Compile Include="Terraria.ModLoader\ModHotkey.cs" />
+    <Compile Include="Terraria.ModLoader\ModItem.cs" />
+    <Compile Include="Terraria.ModLoader\ModLoader.cs" />
+    <Compile Include="Terraria.ModLoader\ModMountData.cs" />
+    <Compile Include="Terraria.ModLoader\ModNet.cs" />
+    <Compile Include="Terraria.ModLoader\ModNPC.cs" />
+    <Compile Include="Terraria.ModLoader.Core\ModOrganizer.cs" />
+    <Compile Include="Terraria.ModLoader\ModPacket.cs" />
+    <Compile Include="Terraria.ModLoader\ModPlayer.cs" />
+    <Compile Include="Terraria.ModLoader\ModPrefix.cs" />
+    <Compile Include="Terraria.ModLoader\ModProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\ModProperties.cs" />
+    <Compile Include="Terraria.ModLoader\ModRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\ModSound.cs" />
+    <Compile Include="Terraria.ModLoader\ModSide.cs" />
+    <Compile Include="Terraria.ModLoader\ModTile.cs" />
+    <Compile Include="Terraria.ModLoader\ModTileEntity.cs" />
+    <Compile Include="Terraria.ModLoader\ModTranslation.cs" />
+    <Compile Include="Terraria.ModLoader\ModWall.cs" />
+    <Compile Include="Terraria.ModLoader\ModWaterStyle.cs" />
+    <Compile Include="Terraria.ModLoader\ModWorld.cs" />
+    <Compile Include="Terraria.ModLoader\MonoModHooks.cs" />
+    <Compile Include="Terraria.ModLoader\MusicPriority.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnHelper.cs" />
+    <Compile Include="Terraria.ModLoader\FrameworkVersion.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeEditor.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeFinder.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeGroupHelper.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeHooks.cs" />
+    <Compile Include="Terraria.ModLoader\DrawableTooltipLine.cs" />
+    <Compile Include="Terraria.ModLoader.Engine\ServerHangWatchdog.cs" />
+    <Compile Include="Terraria.ModLoader\TooltipLine.cs" />
+    <Compile Include="Terraria.ModLoader\TopoSort.cs" />
+    <Compile Include="Terraria.ModLoader\WaterStyleLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WorldHooks.cs" />
+    <Compile Include="Terraria.ModLoader\MountLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MountTextureType.cs" />
+    <Compile Include="Terraria.ModLoader\MusicWrappers.cs" />
+    <Compile Include="Terraria.ModLoader\NPCHeadLoader.cs" />
+    <Compile Include="Terraria.ModLoader\NPCLoader.cs" />
+    <Compile Include="Terraria.ModLoader.IO\PlayerIO.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnInfo.cs" />
+    <Compile Include="Terraria.ModLoader\DrawLayer.cs" />
+    <Compile Include="Terraria.ModLoader\ProjectileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundType.cs" />
+    <Compile Include="Terraria.ModLoader\TileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WallLoader.cs" />
+    <Compile Include="Terraria.ModLoader\PlayerHooks.cs" />
     <Compile Include="Terraria.Modules\AnchorDataModule.cs" />
     <Compile Include="Terraria.Modules\AnchorTypesModule.cs" />
     <Compile Include="Terraria.Modules\LiquidDeathModule.cs" />
@@ -466,6 +_,7 @@
     <Compile Include="Terraria\IngameOptions.cs" />
     <Compile Include="Terraria\Item.cs" />
     <Compile Include="Terraria\ItemText.cs" />
+    <Compile Include="Terraria\Item_tML.cs" />
     <Compile Include="Terraria\Lang.cs" />
     <Compile Include="Terraria\Lighting.cs" />
     <Compile Include="Terraria\LinuxLaunch.cs" Condition="$(DefineConstants.Contains('LINUX'))" />
@@ -509,8 +_,10 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft.Xna.Framework.RuntimeProfile" />
-    <EmbeddedResource Include="Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll" />
-    <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll" />
+    <EmbeddedResource Include="..\..\references\Ionic.Zip.Reduced.dll" />
+    <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll">
+      <LogicalName>Newtonsoft.Json.dll</LogicalName>
+    </EmbeddedResource>
     <EmbeddedResource Include="..\..\references\ReLogicWindows.dll" Condition="$(DefineConstants.Contains('WINDOWS'))">
       <LogicalName>ReLogic.dll</LogicalName>
     </EmbeddedResource>
@@ -529,6 +_,21 @@
     <EmbeddedResource Include="..\..\references\Steamworks.NET.Linux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
       <LogicalName>Steamworks.NET.dll</LogicalName>
     </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\TerrariaHooks.XNA.dll" Condition="$(DefineConstants.Contains('XNA'))">
+      <LogicalName>TerrariaHooks.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\TerrariaHooks.FNA.dll" Condition="$(DefineConstants.Contains('FNA'))">
+      <LogicalName>TerrariaHooks.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\MP3Sharp.dll" />
+    <EmbeddedResource Include="..\..\references\NVorbis.dll" />
+    <EmbeddedResource Include="..\..\references\log4net.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.Pdb.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.Mdb.dll" />
+    <EmbeddedResource Include="..\..\references\MonoMod.RuntimeDetour.dll" />
+    <EmbeddedResource Include="..\..\references\MonoMod.Utils.dll" />
+    <EmbeddedResource Include="..\..\references\System.ValueTuple.dll" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Game.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Items.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Main.json" />
@@ -593,4 +_,147 @@
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Projectiles.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Town.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Content Include="Icon.ico" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonDowngrade.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Config.UI\ButtonUpDown.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Config.UI\ButtonCollapsed.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Config.UI\ButtonExpanded.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Config.UI\ButtonPlus.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Head_Head_Shader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Arms_Shader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Body_Shader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Legs_Legs_Shader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Arms_Glow.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Body_Body_Glow.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Head_Head_Glow.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Developer\PowerRanger_Legs_Legs_Glow.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Glory_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Squid_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonDownload.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonDownloadMultiple.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonModConfig.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonModInfo.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\CopyCodeButton.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Wings.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Saethar_Wings_Wings.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Orian_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Remeus_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\Polyblank_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Wings.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Wings_Wings.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\dinidini_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\KittyKitCatCat_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Body_Arms.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Body_Body.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Head_Head.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default.Patreon\toplayz_Legs_Legs.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Config.UI\ButtonConfig.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonExclamation.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\LoaderBG.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\Loader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonError.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\MysteryTile.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\MysteryItem.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\StartBag.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\UIModBrowserIcons.png" />
+  </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Terraria.Localization.Content.en-US.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.ru-RU.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.pt-BR.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.pl-PL.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.es-ES.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.fr-FR.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.de-DE.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.it-IT.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.tModLoader.json" />
+  </ItemGroup>
+  <Target Name="EditBin" AfterTargets="Build">
+    <PropertyGroup>
+        <EditBinOpts>/largeaddressaware</EditBinOpts>
+		<DumpBin>/headers</DumpBin>
+    </PropertyGroup>
+    <PropertyGroup Condition="$(DefineConstants.Contains('DEBUG'))">
+        <EditBinOpts>$(EditBinOpts) /stack:0x800000</EditBinOpts>
+    </PropertyGroup>
+    <Message Text="Applying editbin $(EditBinOpts)" Importance="high" />
+    <Exec Command="&quot;$(SolutionDir)largeaddresstool\editbin&quot; $(EditBinOpts) &quot;$(TargetPath)&quot;" />
+	<Message Text="Checking headers with dumpbin." Importance="high" />
+	<Exec Command="&quot;$(SolutionDir)largeaddresstool\dumpbin&quot; $(DumpBin) &quot;$(TargetPath)&quot; | findstr &quot;large&quot;" />
+  </Target>
+  <Target Name="CopyToStreamDir" AfterTargets="EditBin" Condition="$(Configuration.Contains('Windows'))">
+    <Message Text="Copying $(OutputName) to Steam Dir..." Importance="high" />
+    <Exec Command="&quot;$(ProjectDir)..\..\setup\bin\setup&quot; --steamdir" ConsoleToMsBuild="true">
+      <Output TaskParameter="ConsoleOutput" PropertyName="SteamDir" />
+    </Exec>
+    <Copy SourceFiles="$(TargetPath)" DestinationFiles="$(SteamDir)\$(OutputName).exe" />
+    <Copy SourceFiles="$(TargetDir)$(OutputName).pdb" DestinationFolder="$(SteamDir)" />
+  </Target>
+  <Target Name="CopyToModCompile" AfterTargets="EditBin" Condition="$(Configuration)==MacRelease">
+    <Message Text="Copying tModLoaderMac to ModCompile..." Importance="high" />
+    <Exec Command="&quot;$(ProjectDir)..\..\setup\bin\setup&quot; --steamdir" ConsoleToMsBuild="true">
+      <Output TaskParameter="ConsoleOutput" PropertyName="SteamDir" />
+    </Exec>
+    <Copy SourceFiles="$(TargetPath)" DestinationFiles="$(SteamDir)\ModCompile\tModLoader.FNA.exe" />
+  </Target>
 </Project>

