--- src/Terraria\Terraria.csproj
+++ src/tModLoader\Terraria.csproj
@@ -17,40 +_,52 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'WindowsRelease|x86'">
     <OutputPath>bin\x86\WindowsRelease\</OutputPath>
     <DefineConstants>CLIENT; WINDOWS</DefineConstants>
+    <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'WindowsServerRelease|x86'">
     <OutputPath>bin\x86\WindowsServerRelease\</OutputPath>
     <DefineConstants>SERVER; WINDOWS</DefineConstants>
+    <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'WindowsDebug|x86'">
     <OutputPath>bin\x86\WindowsDebug\</OutputPath>
     <DefineConstants>DEBUG;CLIENT; WINDOWS</DefineConstants>
+    <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'WindowsServerDebug|x86' ">
     <DefineConstants>DEBUG;SERVER; WINDOWS</DefineConstants>
     <OutputPath>bin\x86\WindowsServerDebug\</OutputPath>
+    <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'MacRelease|x86'">
     <OutputPath>bin\x86\MacRelease\</OutputPath>
     <DefineConstants>CLIENT; MAC</DefineConstants>
+    <PlatformTarget>x64</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'MacServerRelease|x86'">
     <OutputPath>bin\x86\MacServerRelease\</OutputPath>
     <DefineConstants>SERVER; MAC</DefineConstants>
+    <PlatformTarget>x64</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'LinuxRelease|x86'">
     <OutputPath>bin\x86\LinuxRelease\</OutputPath>
     <DefineConstants>CLIENT; LINUX</DefineConstants>
+    <PlatformTarget>x64</PlatformTarget>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'LinuxServerRelease|x86'">
     <OutputPath>bin\x86\LinuxServerRelease\</OutputPath>
     <DefineConstants>SERVER; LINUX</DefineConstants>
+    <PlatformTarget>x64</PlatformTarget>
+  </PropertyGroup>
+  <PropertyGroup>
+    <ApplicationIcon>Icon.ico</ApplicationIcon>
   </PropertyGroup>
   <Choose>
     <When Condition="$(DefineConstants.Contains('DEBUG'))">
       <PropertyGroup>
         <Optimize>false</Optimize>
         <DebugType>full</DebugType>
+        <OutputType>Exe</OutputType>
       </PropertyGroup>
     </When>
     <Otherwise>
@@ -81,19 +_,47 @@
       <HintPath>Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll</HintPath>
     </Reference>
     <Reference Include="System" />
-    <Reference Include="Steamworks.NET">
+    <Reference Include="System.XML" />
+    <Reference Include="Steamworks.NET, Version=7.0.0.0, Culture=neutral, processorArchitecture=x86" Condition="$(DefineConstants.Contains('WINDOWS'))">
+      <SpecificVersion>False</SpecificVersion>
       <HintPath>Terraria.Libraries\Steamworks.NET.Windows\Steamworks.NET.dll</HintPath>
     </Reference>
-    <Reference Include="ReLogic">
+    <Reference Include="Steamworks.NET, Version=7.0.0.0, Culture=neutral, processorArchitecture=x86" Condition="$(DefineConstants.Contains('MAC'))">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\references\Steamworks.NET.OSX.dll</HintPath>
+    </Reference>
+    <Reference Include="Steamworks.NET, Version=7.0.0.0, Culture=neutral, processorArchitecture=x86" Condition="$(DefineConstants.Contains('LINUX'))">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\references\Steamworks.NET.Linux.dll</HintPath>
+    </Reference>
+    <Reference Include="ReLogic" Condition="$(DefineConstants.Contains('WINDOWS'))">
       <HintPath>Terraria.Libraries\ReLogic\ReLogic.dll</HintPath>
+    </Reference>
+    <Reference Include="ReLogic" Condition="$(DefineConstants.Contains('MAC'))">
+      <HintPath>..\..\references\ReLogicMac.dll</HintPath>
+    </Reference>
+    <Reference Include="ReLogic" Condition="$(DefineConstants.Contains('LINUX'))">
+      <HintPath>..\..\references\ReLogicLinux.dll</HintPath>
     </Reference>
     <Reference Include="System.Core" />
     <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('WINDOWS'))" />
     <Reference Include="System.Drawing" />
-    <Reference Include="Ionic.Zip.CF">
-      <HintPath>Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll</HintPath>
-    </Reference>
-    <Reference Include="System.Windows.Forms" />
+    <Reference Include="Ionic.Zip.Reduced">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\references\Ionic.Zip.Reduced.dll</HintPath>
+    </Reference>
+    <Reference Include="FNA" Condition="!$(DefineConstants.Contains('WINDOWS'))">
+      <HintPath>..\..\references\FNA.dll</HintPath>
+    </Reference>
+    <Reference Include="Mono.Cecil">
+      <HintPath>..\..\references\Mono.Cecil.dll</HintPath>
+    </Reference>
+    <Reference Include="MP3Sharp">
+      <HintPath>..\..\references\MP3Sharp.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Windows.Forms">
+      <HintPath Condition="!$(DefineConstants.Contains('WINDOWS'))">..\..\references\System.Windows.Forms.Mono.dll</HintPath>
+    </Reference>
     <Reference Include="WindowsBase" />
   </ItemGroup>
   <ItemGroup>
@@ -327,8 +_,11 @@
     <Compile Include="Terraria.ID\InvasionID.cs" />
     <Compile Include="Terraria.ID\ItemAlternativeFunctionID.cs" />
     <Compile Include="Terraria.ID\ItemID.cs" />
+    <Compile Include="Terraria.ID\LangID.cs" />
     <Compile Include="Terraria.ID\MessageID.cs" />
     <Compile Include="Terraria.ID\MountID.cs" />
+    <Compile Include="Terraria.ID\MusicID.cs" />
+    <Compile Include="Terraria.ID\NetmodeID.cs" />
     <Compile Include="Terraria.ID\NPCID.cs" />
     <Compile Include="Terraria.ID\PlayerTextureID.cs" />
     <Compile Include="Terraria.ID\PlayerVariantID.cs" />
@@ -368,6 +_,150 @@
     <Compile Include="Terraria.Map\MapHelper.cs" />
     <Compile Include="Terraria.Map\MapTile.cs" />
     <Compile Include="Terraria.Map\WorldMap.cs" />
+    <Compile Include="Terraria.ModLoader.Default\AprilFools.cs" />
+    <Compile Include="Terraria.ModLoader.Default\HelpCommand.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModlistCommand.cs" />
+    <Compile Include="Terraria.ModLoader.Default\ModLoaderMod.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryGlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTile.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTileEntity.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryTilesWorld.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryWorld.cs" />
+    <Compile Include="Terraria.ModLoader.Default\MysteryPlayer.cs" />
+    <Compile Include="Terraria.ModLoader.Default\StartBag.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\AddRecipesException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\CustomModDataException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\ModNameException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\MissingResourceException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\ModSortingException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\OldHookException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\RecipeException.cs" />
+    <Compile Include="Terraria.ModLoader.Exceptions\ResourceLoadException.cs" />
+    <Compile Include="Terraria.ModLoader.IO\BigEndian.cs" />
+    <Compile Include="Terraria.ModLoader.IO\BinaryIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\FontCacheIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\ItemIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\MapIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagSerializer.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagPrinter.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagSerializable.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TileIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TmodFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\UploadFile.cs" />
+    <Compile Include="Terraria.ModLoader.IO\WAVCacheIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\WorldIO.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagCompound.cs" />
+    <Compile Include="Terraria.ModLoader.IO\TagIO.cs" />
+    <Compile Include="Terraria.ModLoader.UI.Elements\FixedUIScrollbar.cs" />
+    <Compile Include="Terraria.ModLoader.UI.Elements\UIGrid.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIBuildMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIColorTextPanel.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UICommon.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIDownloadMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIEnterPassphraseMenu.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UICycleImage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIDownloadMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIEnterSteamIDMenu.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIErrorMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIHoverImage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIInfoMessage.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIInputTextField.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoader.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIExtractMod.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UILoadProgress.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIManagePublished.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMessageBox.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModBrowser.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModControls.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModDownloadItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModInfo.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModPackItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModPacks.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModManageItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIMods.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSourceItem.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIModSources.cs" />
+    <Compile Include="Terraria.ModLoader.UI\UIUpdateMessage.cs" />
+    <Compile Include="Terraria.ModLoader\AssemblyManager.cs" />
+    <Compile Include="Terraria.ModLoader\AutoloadEquip.cs" />
+    <Compile Include="Terraria.ModLoader\AutoloadHead.cs" />
+    <Compile Include="Terraria.ModLoader\BackgroundStyleLoaders.cs" />
+    <Compile Include="Terraria.ModLoader\BackgroundTextureLoader.cs" />
+    <Compile Include="Terraria.ModLoader\BuffLoader.cs" />
+    <Compile Include="Terraria.ModLoader\BuildProperties.cs" />
+    <Compile Include="Terraria.ModLoader\CommandManager.cs" />
+    <Compile Include="Terraria.ModLoader\DrawInfo.cs" />
+    <Compile Include="Terraria.ModLoader\EquipLoader.cs" />
+    <Compile Include="Terraria.ModLoader\EquipTexture.cs" />
+    <Compile Include="Terraria.ModLoader\EquipType.cs" />
+    <Compile Include="Terraria.ModLoader\ErrorLogger.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalBuff.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalItem.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalNPC.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalTile.cs" />
+    <Compile Include="Terraria.ModLoader\GlobalWall.cs" />
+    <Compile Include="Terraria.ModLoader\Interface.cs" />
+    <Compile Include="Terraria.ModLoader\ItemLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MapEntry.cs" />
+    <Compile Include="Terraria.ModLoader\MapLegend.cs" />
+    <Compile Include="Terraria.ModLoader\MapLoader.cs" />
+    <Compile Include="Terraria.ModLoader\Mod.cs" />
+    <Compile Include="Terraria.ModLoader\ModBackgroundStyle.cs" />
+    <Compile Include="Terraria.ModLoader\ModBuff.cs" />
+    <Compile Include="Terraria.ModLoader\ModCommand.cs" />
+    <Compile Include="Terraria.ModLoader\ModCompile.cs" />
+    <Compile Include="Terraria.ModLoader\ConvertiblePlants.cs" />
+    <Compile Include="Terraria.ModLoader\ModDust.cs" />
+    <Compile Include="Terraria.ModLoader\ModGore.cs" />
+    <Compile Include="Terraria.ModLoader\ModHooks.cs" />
+    <Compile Include="Terraria.ModLoader\ModHotkey.cs" />
+    <Compile Include="Terraria.ModLoader\ModItem.cs" />
+    <Compile Include="Terraria.ModLoader\ModLoader.cs" />
+    <Compile Include="Terraria.ModLoader\ModMountData.cs" />
+    <Compile Include="Terraria.ModLoader\ModNet.cs" />
+    <Compile Include="Terraria.ModLoader\ModNPC.cs" />
+    <Compile Include="Terraria.ModLoader\ModPacket.cs" />
+    <Compile Include="Terraria.ModLoader\ModPlayer.cs" />
+    <Compile Include="Terraria.ModLoader\ModProjectile.cs" />
+    <Compile Include="Terraria.ModLoader\ModProperties.cs" />
+    <Compile Include="Terraria.ModLoader\ModRecipe.cs" />
+    <Compile Include="Terraria.ModLoader\ModSound.cs" />
+    <Compile Include="Terraria.ModLoader\ModSide.cs" />
+    <Compile Include="Terraria.ModLoader\ModTile.cs" />
+    <Compile Include="Terraria.ModLoader\ModTileEntity.cs" />
+    <Compile Include="Terraria.ModLoader\ModTranslation.cs" />
+    <Compile Include="Terraria.ModLoader\ModWall.cs" />
+    <Compile Include="Terraria.ModLoader\ModWaterStyle.cs" />
+    <Compile Include="Terraria.ModLoader\ModWorld.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnHelper.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeEditor.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeFinder.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeGroupHelper.cs" />
+    <Compile Include="Terraria.ModLoader\RecipeHooks.cs" />
+    <Compile Include="Terraria.ModLoader\DrawableTooltipLine.cs" />
+    <Compile Include="Terraria.ModLoader\TooltipLine.cs" />
+    <Compile Include="Terraria.ModLoader\TopoSort.cs" />
+    <Compile Include="Terraria.ModLoader\WaterStyleLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WorldHooks.cs" />
+    <Compile Include="Terraria.ModLoader\MountLoader.cs" />
+    <Compile Include="Terraria.ModLoader\MountTextureType.cs" />
+    <Compile Include="Terraria.ModLoader.Audio\MusicWrappers.cs" />
+    <Compile Include="Terraria.ModLoader\NPCHeadLoader.cs" />
+    <Compile Include="Terraria.ModLoader\NPCLoader.cs" />
+    <Compile Include="Terraria.ModLoader.IO\PlayerIO.cs" />
+    <Compile Include="Terraria.ModLoader\NPCSpawnInfo.cs" />
+    <Compile Include="Terraria.ModLoader\DrawLayer.cs" />
+    <Compile Include="Terraria.ModLoader\ProjectileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundLoader.cs" />
+    <Compile Include="Terraria.ModLoader\SoundType.cs" />
+    <Compile Include="Terraria.ModLoader\TileLoader.cs" />
+    <Compile Include="Terraria.ModLoader\WallLoader.cs" />
+    <Compile Include="Terraria.ModLoader\PlayerHooks.cs" />
     <Compile Include="Terraria.Modules\AnchorDataModule.cs" />
     <Compile Include="Terraria.Modules\AnchorTypesModule.cs" />
     <Compile Include="Terraria.Modules\LiquidDeathModule.cs" />
@@ -495,6 +_,7 @@
     <Compile Include="Terraria\IngameOptions.cs" />
     <Compile Include="Terraria\Item.cs" />
     <Compile Include="Terraria\ItemText.cs" />
+    <Compile Include="Terraria\Item_tML.cs" />
     <Compile Include="Terraria\Lang.cs" />
     <Compile Include="Terraria\Lighting.cs" />
     <Compile Include="Terraria\LinuxLaunch.cs" Condition="$(DefineConstants.Contains('LINUX'))" />
@@ -538,10 +_,24 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft.Xna.Framework.RuntimeProfile" />
-    <EmbeddedResource Include="Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll" />
+    <EmbeddedResource Include="..\..\references\Ionic.Zip.Reduced.dll" />
     <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll" />
-    <EmbeddedResource Include="Terraria.Libraries\ReLogic\ReLogic.dll" />
-    <EmbeddedResource Include="Terraria.Libraries\Steamworks.NET.Windows\Steamworks.NET.dll" />
+    <EmbeddedResource Include="Terraria.Libraries\ReLogic\ReLogic.dll" Condition="$(DefineConstants.Contains('WINDOWS'))" />
+    <EmbeddedResource Include="..\..\references\ReLogicMac.dll" Condition="$(DefineConstants.Contains('MAC'))">
+      <LogicalName>Terraria.Libraries.ReLogic.ReLogic.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\ReLogicLinux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
+      <LogicalName>Terraria.Libraries.ReLogic.ReLogic.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="Terraria.Libraries\Steamworks.NET.Windows\Steamworks.NET.dll" Condition="$(DefineConstants.Contains('WINDOWS'))" />
+    <EmbeddedResource Include="..\..\references\Steamworks.NET.OSX.dll" Condition="$(DefineConstants.Contains('MAC'))">
+      <LogicalName>Terraria.Libraries.Steamworks.NET.OSX.Steamworks.NET.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\Steamworks.NET.Linux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
+      <LogicalName>Terraria.Libraries.Steamworks.NET.Linux.Steamworks.NET.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\MP3Sharp.dll" />
+    <EmbeddedResource Include="..\..\references\Mono.Cecil.dll" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Game.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Items.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Main.json" />
@@ -606,5 +_,57 @@
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Projectiles.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Town.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Content Include="Icon.ico" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\LoaderBG.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\Loader.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\ButtonError.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\MysteryTile.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\MysteryItem.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.Default\StartBag.png" />
+    <EmbeddedResource Include="Terraria.ModLoader.UI\UIModBrowserIcons.png" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="App.config">
+      <SubType>Designer</SubType>
+    </None>
+    <EmbeddedResource Include="Terraria.Localization.Content.en-US.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.ru-RU.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.pt-BR.tModLoader.json" />
+    <EmbeddedResource Include="Terraria.Localization.Content.pl-PL.tModLoader.json" />
+  </ItemGroup>
+  <ItemGroup />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-</Project>
+  <PropertyGroup>
+    <PostBuildEvent>SET CONFIGNAME=$(ConfigurationName)
+
+SET EDITBINOPTS=/largeaddressaware
+IF NOT %25CONFIGNAME:Debug=%25==%25CONFIGNAME%25 SET EDITBINOPTS=%25EDITBINOPTS%25 /stack:0x800000
+"$(SolutionDir)"largeaddresstool\editbin %25EDITBINOPTS%25 "$(TargetPath)"
+
+SET RELEASEFILE=""
+
+IF $(ConfigurationName) == MacRelease SET RELEASEFILE=ModCompile/tModLoaderMac
+IF $(ConfigurationName) == WindowsServerDebug SET RELEASEFILE=tModLoaderServerDebug
+IF $(ConfigurationName) == WindowsServerRelease SET RELEASEFILE=tModLoaderServer
+IF $(ConfigurationName) == WindowsDebug SET RELEASEFILE=tModLoaderDebug
+IF $(ConfigurationName) == WindowsRelease SET RELEASEFILE=tModLoader
+
+for /f "tokens=* delims=" %25%25i in ('"$(ProjectDir)..\..\setup\bin\setup" --steamdir') do set TERRARIASTEAMDIR=%25%25i
+
+IF NOT %25RELEASEFILE%25=="" (
+
+mkdir "%25TERRARIASTEAMDIR%25\ModCompile\"
+copy "$(TargetDir)$(TargetName)$(TargetExt)" "%25TERRARIASTEAMDIR%25\%25RELEASEFILE%25$(TargetExt)"
+
+IF NOT %25RELEASEFILE:Debug=%25==%25RELEASEFILE%25 (
+	copy "$(TargetDir)$(TargetName).pdb" "%25TERRARIASTEAMDIR%25\%25RELEASEFILE%25.pdb"
+)
+
+)</PostBuildEvent>
+  </PropertyGroup>
+  <PropertyGroup>
+    <PreBuildEvent>
+    </PreBuildEvent>
+  </PropertyGroup>
+</Project>
