--- src/TerrariaNetCore/Terraria/Social/SocialAPI.cs
+++ src/Unified/Terraria/Social/SocialAPI.cs
@@ -1,9 +_,11 @@
 using System;
 using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
 using ReLogic.OS;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 using Terraria.Social.WeGame;
+using Terraria.Unified.Startup;
 
 namespace Terraria.Social;
 
@@ -26,12 +_,25 @@
 	{
 		if (!mode.HasValue) {
 			mode = SocialMode.None;
+
+			/*
 			if (Main.dedServ) {
 				if (Program.LaunchParameters.ContainsKey("-steam"))
 					mode = SocialMode.Steam;
 			}
 			else {
 				mode = SocialMode.Steam;
+			}
+			*/
+
+			var logger = GameLaunch.Instance.GameLogger;
+			logger.LogDebug("Attempting to resolve social mode...");
+			if (LaunchHelper.TryGetSocialMode(out SocialMode resolvedMode)) {
+				mode = resolvedMode;
+				logger.LogInformation("Resolved social mode: {0}", resolvedMode);
+			}
+			else {
+				logger.LogInformation("Failed to automatically resolve social mode; defaulting to None");
 			}
 		}
 
