--- src/TerrariaNetCore/Terraria/Achievements/AchievementManager.cs
+++ src/tModLoader/Terraria/Achievements/AchievementManager.cs
@@ -7,6 +_,8 @@
 using System.Linq;
 using System.Security.Cryptography;
 using System.Text;
+using Steamworks;
+using Terraria.ModLoader.Engine;
 using Terraria.Social;
 using Terraria.UI;
 using Terraria.Utilities;
@@ -51,6 +_,7 @@
 		}
 
 		private void Save(string path, bool cloud) {
+			if (Steam.IsSteamApp) return;
 			lock (_ioLock) {
 				if (SocialAPI.Achievements != null)
 					SocialAPI.Achievements.StoreStats();
@@ -76,10 +_,13 @@
 		public List<Achievement> CreateAchievementsList() => _achievements.Values.ToList();
 
 		public void Load() {
+			if (Steam.IsSteamApp) return;
 			Load(_savePath, _isCloudSave);
 		}
 
-		private void Load(string path, bool cloud) {
+		private void Load(string path, bool cloud)
+		{
+			if (Steam.IsSteamApp) return;
 			bool flag = false;
 			lock (_ioLock) {
 				if (!FileUtilities.Exists(path, cloud))
