--- src/TerrariaNetCore/Terraria/Achievements/AchievementCondition.cs
+++ src/tModLoader/Terraria/Achievements/AchievementCondition.cs
@@ -1,5 +_,6 @@
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using Terraria.ModLoader.Engine;
 
 namespace Terraria.Achievements
 {
@@ -30,6 +_,8 @@
 		}
 
 		public virtual void Complete() {
+			if (Steam.IsSteamApp) return;
+
 			if (!_isCompleted) {
 				_isCompleted = true;
 				if (this.OnComplete != null)
