--- src/TerrariaNetCore/Terraria/Achievements/AchievementManager.cs
+++ src/tModLoader/Terraria/Achievements/AchievementManager.cs
@@ -54,6 +_,10 @@
 
 	private void Save(string path, bool cloud)
 	{
+		// Check added by TML.
+		if (SocialAPI.Mode == SocialMode.Steam)
+			return;
+
 		lock (_ioLock) {
 			if (SocialAPI.Achievements != null)
 				SocialAPI.Achievements.StoreStats();
@@ -77,11 +_,19 @@
 
 	public void Load()
 	{
+		// Check added by TML.
+		if (SocialAPI.Mode == SocialMode.Steam)
+			return;
+
 		Load(_savePath, _isCloudSave);
 	}
 
 	private void Load(string path, bool cloud)
 	{
+		// Check added by TML.
+		if (SocialAPI.Mode == SocialMode.Steam)
+			return;
+
 		bool flag = false;
 		lock (_ioLock) {
 			if (!FileUtilities.Exists(path, cloud))
