--- src/TerrariaNetCore/Terraria/Achievements/AchievementManager.cs
+++ src/tModLoader/Terraria/Achievements/AchievementManager.cs
@@ -7,6 +_,7 @@
 using Newtonsoft.Json;
 using Newtonsoft.Json.Bson;
 using Newtonsoft.Json.Linq;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.UI;
 using Terraria.Utilities;
@@ -33,7 +_,8 @@
 
 	public AchievementManager()
 	{
-		if (SocialAPI.Achievements != null) {
+		//Solxan: Ensures achievements are saved
+		if (false /*SocialAPI.Achievements != null*/) {
 			_savePath = SocialAPI.Achievements.GetSavePath();
 			_isCloudSave = true;
 			_cryptoKey = SocialAPI.Achievements.GetEncryptionKey();
@@ -59,10 +_,19 @@
 				SocialAPI.Achievements.StoreStats();
 
 			try {
+				Dictionary<string, object> existingData = LoadExistingAchievements(path, cloud);
+
+				foreach (var kvp in _achievements) {
+					existingData[kvp.Key] = kvp.Value;
+				}
+
 				using MemoryStream memoryStream = new MemoryStream();
 				using CryptoStream cryptoStream = new CryptoStream(memoryStream, new RijndaelManaged().CreateEncryptor(_cryptoKey, _cryptoKey), CryptoStreamMode.Write);
 				using BsonWriter bsonWriter = new BsonWriter(cryptoStream);
+				/*
 				JsonSerializer.Create(_serializerSettings).Serialize(bsonWriter, _achievements);
+				*/
+				JsonSerializer.Create(_serializerSettings).Serialize(bsonWriter, existingData);
 				bsonWriter.Flush();
 				cryptoStream.FlushFinalBlock();
 				FileUtilities.WriteAllBytes(path, memoryStream.ToArray(), cloud);
@@ -75,6 +_,24 @@
 
 	public List<Achievement> CreateAchievementsList() => _achievements.Values.ToList();
 
+	private Dictionary<string, object> LoadExistingAchievements(string path, bool cloud)
+	{
+		try {
+			if (FileUtilities.Exists(path, cloud)) {
+				byte[] fileData = FileUtilities.ReadAllBytes(path, cloud);
+				using MemoryStream memoryStream = new MemoryStream(fileData);
+				using CryptoStream cryptoStream = new CryptoStream(memoryStream,
+					new RijndaelManaged().CreateDecryptor(_cryptoKey, _cryptoKey), CryptoStreamMode.Read);
+				using BsonReader bsonReader = new BsonReader(cryptoStream);
+				return JsonSerializer.Create(_serializerSettings).Deserialize<Dictionary<string, object>>(bsonReader) ??
+				       new Dictionary<string, object>();
+			}
+		}
+		catch (Exception) {
+			FileUtilities.Delete(path, cloud);
+		}
+		return new Dictionary<string, object>();
+	}
+
 	public void Load()
 	{
 		Load(_savePath, _isCloudSave);
@@ -131,6 +_,7 @@
 			achievement.Value.ClearProgress();
 		}
 
+		// TODO: Make a backup instead of losing them permanently?
 		Save();
 	}
 
@@ -145,6 +_,12 @@
 	{
 		_achievements.Add(achievement.Name, achievement);
 		achievement.OnCompleted += AchievementCompleted;
+	}
+
+	public void Unregister(Achievement achievement)
+	{
+		_achievements.Remove(achievement.Name);
+		achievement.OnCompleted -= AchievementCompleted;
 	}
 
 	public void RegisterIconIndex(string achievementName, int iconIndex)
