--- src/TerrariaNetCore/Terraria/Net/NetManager.cs
+++ src/tModLoader/Terraria/Net/NetManager.cs
@@ -1,6 +_,9 @@
 using System.Collections.Generic;
 using System.IO;
+using log4net.Core;
+using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Net.Sockets;
 using Terraria.Testing;
 
 namespace Terraria.Net;
@@ -52,7 +_,9 @@
 
 	public void Broadcast(NetPacket packet, int ignoreClient = -1)
 	{
+		LogSend(packet, ignoreClient: ignoreClient);
+
 		Invariant.Assert(Main.dedServ, "Packets can only be broadcast from a server.");
 		for (int i = 0; i < 256; i++) {
 			if (i != ignoreClient && Netplay.Clients[i].IsConnected())
 				SendData(Netplay.Clients[i].Socket, packet);
@@ -63,7 +_,9 @@
 
 	public void Broadcast(NetPacket packet, BroadcastCondition conditionToBroadcast, int ignoreClient = -1)
 	{
+		LogSend(packet, ignoreClient: ignoreClient);
+
 		Invariant.Assert(Main.dedServ, "Packets can only be broadcast from a server.");
 		for (int i = 0; i < 256; i++) {
 			if (i != ignoreClient && Netplay.Clients[i].IsConnected() && conditionToBroadcast(i))
 				SendData(Netplay.Clients[i].Socket, packet);
@@ -111,16 +_,18 @@
 
 	public void SendToServer(NetPacket packet)
 	{
+		LogSend(packet);
 		Invariant.Assert(!Main.dedServ, "Packets can only be sent to the server from a client.");
 		SendData(Netplay.Connection.Socket, packet);
 		packet.Recycle();
 	}
 
 	public void SendToClient(NetPacket packet, int playerId)
 	{
+		LogSend(packet, toClient: playerId);
 		Invariant.Assert(Main.dedServ, "Packets can only be sent to a client from a server.");
 		if (Netplay.Clients[playerId].IsConnected())
 			SendData(Netplay.Clients[playerId].Socket, packet);
 
 		packet.Recycle();
 	}
@@ -144,10 +_,18 @@
 			socket.AsyncSend(packet.Buffer.Data, 0, packet.Length, EmptyCallback);
 		}
 		catch {
+			Logging.ServerConsoleLine(Language.GetTextValue("Error.ExceptionNormal", Language.GetTextValue("Error.DataSentAfterConnectionLost")), Level.Warn);
 		}
 	}
 
 	private static void EmptyCallback(object state)
 	{
+	}
+
+	// Added by TML.
+	private void LogSend(NetPacket packet, int toClient = -1, int ignoreClient = -1)
+	{
+		if (ModNet.DetailedLogging)
+			ModNet.LogSend(toClient, ignoreClient, $"NetPacket {_modules[packet.Id].GetType().Name}", packet.Length);
 	}
 }
