--- src/TerrariaNetCore/Terraria/GameContent/UI/CustomCurrencyManager.cs
+++ src/tModLoader/Terraria/GameContent/UI/CustomCurrencyManager.cs
@@ -13,6 +_,7 @@
 	public static void Initialize()
 	{
 		_nextCurrencyIndex = 0;
+		_currencies.Clear(); // Added by TML.
 		CustomCurrencyID.DefenderMedals = RegisterCurrency(new CustomCurrencySingleCoin(3817, 999L));
 	}
 
@@ -138,4 +_,13 @@
 	{
 		_currencies[item.shopSpecialCurrency].GetItemExpectedPrice(item, out calcForSelling, out calcForBuying);
 	}
+
+	/// <summary>
+	/// Attempts to retrieve a CustomCurrencySystem object with the specified id from the _currencies dictionary.
+	/// </summary>
+	/// <param name="id">The id of the currency system to retrieve.</param>
+	/// <param name="system">When this method returns, contains the retrieved CustomCurrencySystem object, or null if the retrieval failed.</param>
+	/// <returns>true if the retrieval was successful; otherwise, false.</returns>
+	public static bool TryGetCurrencySystem(int id, out CustomCurrencySystem system) => _currencies.TryGetValue(id, out system); // Resolves #3133
+	
 }
