--- src/Terraria/Terraria/GameContent/UI/EmoteBubble.cs
+++ src/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs
@@ -9,7 +_,7 @@
 {
 	public class EmoteBubble
 	{
-		private static int[] CountNPCs = new int[668];
+		internal static int[] CountNPCs = new int[NPCID.Count];
 		public static Dictionary<int, EmoteBubble> byID = new Dictionary<int, EmoteBubble>();
 		private static List<int> toClean = new List<int>();
 		public static int NextID;
