--- src/Terraria/Terraria/GameContent/UI/FancyClassicPlayerResourcesDisplaySet.cs
+++ src/tModLoader/Terraria/GameContent/UI/FancyClassicPlayerResourcesDisplaySet.cs
@@ -5,7 +_,7 @@
 
 namespace Terraria.GameContent.UI
 {
-	public class FancyClassicPlayerResourcesDisplaySet : IPlayerResourcesDisplaySet
+	public partial class FancyClassicPlayerResourcesDisplaySet : IPlayerResourcesDisplaySet
 	{
 		private float _currentPlayerLife;
 		private float _lifePerHeart;
