--- src/TerrariaNetCore/Terraria/GameContent/UI/GameTipsDisplay.cs
+++ src/tModLoader/Terraria/GameContent/UI/GameTipsDisplay.cs
@@ -3,11 +_,12 @@
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Graphics;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 
 namespace Terraria.GameContent.UI;
 
-public class GameTipsDisplay
+public partial class GameTipsDisplay
 {
 	private class GameTip
 	{
@@ -62,7 +_,12 @@
 
 	public GameTipsDisplay(ITipProvider tipProvider)
 	{
-		_tipProvider = tipProvider;
+		_tipsDefault = Language.FindAll(Lang.CreateDialogFilter("LoadingTips_Default."));
+		_tipsGamepad = Language.FindAll(Lang.CreateDialogFilter("LoadingTips_GamePad."));
+		_tipsKeyboard = Language.FindAll(Lang.CreateDialogFilter("LoadingTips_Keyboard."));
+		_lastTip = null;
+
+		Initialize();
 	}
 
 	public void Update()
@@ -142,7 +_,31 @@
 	private void AddNewTip(double currentTime)
 	{
 		string textKey = "UI.Back";
-		string key = _tipProvider.RollAvailableTip().Key;
+		/*
+		List<LocalizedText> list = new List<LocalizedText>();
+		list.AddRange(_tipsDefault);
+		if (PlayerInput.UsingGamepad)
+			list.AddRange(_tipsGamepad);
+		else
+			list.AddRange(_tipsKeyboard);
+
+		if (_lastTip != null)
+			list.Remove(_lastTip);
+		*/
+
+		var list = new List<GameTipData>(allTips);
+
+		if (PlayerInput.UsingGamepad)
+			list.RemoveAll(tip => tip.Mod is null && tip.TipText.Key.StartsWith("LoadingTips_Keyboard"));
+		else
+			list.RemoveAll(tip => tip.Mod is null && tip.TipText.Key.StartsWith("LoadingTips_GamePad"));
+
+		if (_lastTip != null)
+			list.RemoveAll(tip => tip.TipText == _lastTip);
+
+		list.RemoveAll(tip => !tip.isVisible);
+
+		string key = (_lastTip = ((list.Count != 0) ? list[Main.rand.Next(list.Count)].TipText : LocalizedText.Empty)).Key;
 		if (Language.Exists(key))
 			textKey = key;
 
