--- src/TerrariaNetCore/Terraria/GameContent/UI/GameTipsDisplay.cs
+++ src/tModLoader/Terraria/GameContent/UI/GameTipsDisplay.cs
@@ -3,11 +_,12 @@
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Graphics;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 
 namespace Terraria.GameContent.UI;
 
-public class GameTipsDisplay
+public partial class GameTipsDisplay
 {
 	private class GameTip
 	{
