--- src/TerrariaNetCore/Terraria/GameContent/UI/ItemRarity.cs
+++ src/tModLoader/Terraria/GameContent/UI/ItemRarity.cs
@@ -1,6 +_,7 @@
 using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.UI;
 
@@ -22,6 +_,13 @@
 		_rarities.Add(7, Colors.RarityLime);
 		_rarities.Add(8, Colors.RarityYellow);
 		_rarities.Add(9, Colors.RarityCyan);
+
+		// TML-added:
+		_rarities.Add(10, Colors.RarityDarkRed);
+		_rarities.Add(11, Colors.RarityDarkPurple);
+
+		for (int i = ItemRarityID.Count; i < RarityLoader.RarityCount; i++)
+			_rarities.Add(i, RarityLoader.GetRarity(i).RarityColor);
 	}
 
 	public static Color GetColor(int rarity)
