--- src/TerrariaNetCore/Terraria/GameContent/Liquid/LiquidRenderer.cs
+++ src/Unified/Terraria/GameContent/Liquid/LiquidRenderer.cs
@@ -98,7 +_,7 @@
 	private float _frameState;
 	private float _waterfallFrameState;
 
-	private static Tile[,] Tiles => Main.tile;
+	private static Tilemap Tiles => Main.tile;
 
 	public event Action<Color[], Rectangle> WaveFilters;
 
@@ -133,7 +_,7 @@
 		if (_waveMask.Length < drawArea.Width * drawArea.Height)
 			_waveMask = new Color[drawArea.Width * drawArea.Height];
 
-		Tile tile = null;
+		Tile tile = default;
 		fixed (LiquidCache* ptr = &_cache[1]) {
 			LiquidCache* ptr2 = ptr;
 			int num = rectangle.Height * 2 + 2;
