--- src/TerrariaNetCore/ReLogic/OS/Windows/WindowService.cs
+++ src/Unified/ReLogic/OS/Windows/WindowService.cs
@@ -81,10 +_,10 @@
 		((Form)Control.FromHandle(window.Handle)).Location = new System.Drawing.Point(x, y);
 	}
 
-	public Microsoft.Xna.Framework.Rectangle GetBounds(GameWindow window)
+	public System.Drawing.Rectangle GetBounds(GameWindow window)
 	{
 		Form form = (Form)Control.FromHandle(window.Handle);
-		return new Microsoft.Xna.Framework.Rectangle(form.Bounds.X, form.Bounds.Y, form.Bounds.Width, form.Bounds.Height);
+		return new System.Drawing.Rectangle(form.Bounds.X, form.Bounds.Y, form.Bounds.Width, form.Bounds.Height);
 	}
 	*/
 }
