--- src/decompiled/Terraria/GameContent/UI/Elements/UIHeader.cs
+++ src/Terraria/Terraria/GameContent/UI/Elements/UIHeader.cs
@@ -16,9 +_,11 @@
 			set {
 				if (_text != value) {
 					_text = value;
+					if (!Main.dedServ) {
 					Vector2 vector = FontAssets.DeathText.Value.MeasureString(Text);
 					Width.Pixels = vector.X;
 					Height.Pixels = vector.Y;
+					}
 					Width.Precent = 0f;
 					Height.Precent = 0f;
 					Recalculate();
