--- src/decompiled/Terraria/IO/WorldFile.cs
+++ src/Terraria/Terraria/IO/WorldFile.cs
@@ -1280,7 +_,7 @@
 				try {
 					Main.WorldFileMetadata = FileMetadata.Read(reader, FileType.World);
 				}
-				catch (FileFormatException value) {
+				catch (FormatException value) {
 					Console.WriteLine(Language.GetTextValue("Error.UnableToLoadWorld"));
 					Console.WriteLine(value);
 					return false;
