--- src/Terraria/Terraria/Utilities/CrashDump.cs
+++ src/tModLoader/Terraria/Utilities/CrashDump.cs
@@ -55,11 +_,13 @@
 
 		private static string CreateDumpName() {
 			DateTime dateTime = DateTime.Now.ToLocalTime();
-#if CLIENT
-			const string Name = "Terraria";
-#else
-			const string Name = "TerrariaServer";
-#endif
+			string Name;
+			if (!Main.dedServ) {
+				Name = "Terraria";
+			}
+			else {
+				Name = "TerrariaServer";
+			}
 			return string.Format("{0}_{1}_{2}_{3}.dmp", Name, Main.versionNumber, dateTime.ToString("MM-dd-yy_HH-mm-ss-ffff", CultureInfo.InvariantCulture), Thread.CurrentThread.ManagedThreadId);
 		}
 
