--- src/TerrariaNetCore/Terraria/Utilities/FileOperationAPIWrapper.cs
+++ src/tModLoader/Terraria/Utilities/FileOperationAPIWrapper.cs
@@ -1,9 +_,10 @@
-#if WINDOWS
 using System;
 using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
 
 namespace Terraria.Utilities
 {
+	[SupportedOSPlatform("windows")]
 	public static class FileOperationAPIWrapper
 	{
 		[Flags]
@@ -25,7 +_,7 @@
 			FO_RENAME
 		}
 
-		[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto, Pack = 1)]
+		[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto)]
 		private struct SHFILEOPSTRUCT
 		{
 			public IntPtr hwnd;
@@ -79,4 +_,3 @@
 		private static bool DeleteCompletelySilent(string path) => DeleteFile(path, FileOperationFlags.FOF_SILENT | FileOperationFlags.FOF_NOCONFIRMATION | FileOperationFlags.FOF_NOERRORUI);
 	}
 }
-#endif
