--- src/Terraria/Terraria/Utilities/PlatformUtilities.cs
+++ src/tModLoader/Terraria/Utilities/PlatformUtilities.cs
@@ -61,25 +_,26 @@
 #if !WINDOWS
 		public unsafe static void SavePng(Stream stream, int width, int height, int imgWidth, int imgHeight, byte[] data) {
 #if CLIENT
+			throw new NotImplementedException("TODO");
-			IntPtr intPtr = SDL.SDL_CreateRGBSurface(0u, imgWidth, imgHeight, 32, 255u, 65280u, 16711680u, 4278190080u);
+			//IntPtr intPtr = SDL.SDL_CreateRGBSurface(0u, imgWidth, imgHeight, 32, 255u, 65280u, 16711680u, 4278190080u);
-			SDL.SDL_LockSurface(intPtr);
+			//SDL.SDL_LockSurface(intPtr);
-			PlatformUtilities.SDL_Surface* ptr = (PlatformUtilities.SDL_Surface*)((void*)intPtr);
+			//PlatformUtilities.SDL_Surface* ptr = (PlatformUtilities.SDL_Surface*)((void*)intPtr);
-			Marshal.Copy(data, 0, ptr->pixels, width * height * 4);
+			//Marshal.Copy(data, 0, ptr->pixels, width * height * 4);
-			SDL.SDL_UnlockSurface(intPtr);
+			//SDL.SDL_UnlockSurface(intPtr);
-			data = null;
+			//data = null;
-			if (width != imgWidth || height != imgHeight) {
+			//if (width != imgWidth || height != imgHeight) {
-				IntPtr intPtr2 = SDL.SDL_CreateRGBSurface(0u, width, height, 32, 255u, 65280u, 16711680u, 4278190080u);
+			//	IntPtr intPtr2 = SDL.SDL_CreateRGBSurface(0u, width, height, 32, 255u, 65280u, 16711680u, 4278190080u);
-				SDL.SDL_BlitScaled(intPtr, IntPtr.Zero, intPtr2, IntPtr.Zero);
+			//	SDL.SDL_BlitScaled(intPtr, IntPtr.Zero, intPtr2, IntPtr.Zero);
-				SDL.SDL_FreeSurface(intPtr);
+			//	SDL.SDL_FreeSurface(intPtr);
-				intPtr = intPtr2;
+			//	intPtr = intPtr2;
-			}
+			//}
 
-			byte[] array = new byte[width * height * 4 + 41 + 57 + 256];
+			//byte[] array = new byte[width * height * 4 + 41 + 57 + 256];
-			IntPtr intPtr3 = SDL.SDL_RWFromMem(array, array.Length);
+			//IntPtr intPtr3 = SDL.SDL_RWFromMem(array, array.Length);
-			SDL_image.IMG_SavePNG_RW(intPtr, intPtr3, 1);
+			//SDL_image.IMG_SavePNG_RW(intPtr, intPtr3, 1);
-			SDL.SDL_FreeSurface(intPtr);
+			//SDL.SDL_FreeSurface(intPtr);
-			int count = ((int)array[33] << 24 | (int)array[34] << 16 | (int)array[35] << 8 | (int)array[36]) + 41 + 57;
+			//int count = ((int)array[33] << 24 | (int)array[34] << 16 | (int)array[35] << 8 | (int)array[36]) + 41 + 57;
-			stream.Write(array, 0, count);
+			//stream.Write(array, 0, count);
 #endif
 		}
 #endif
