--- src/TerrariaNetCore/Terraria/Physics/BallCollision.cs
+++ src/tModLoader/Terraria/Physics/BallCollision.cs
@@ -156,7 +_,7 @@
 		tileEdges = ((!(velocity.Y > velocity.X)) ? (tileEdges | TileEdges.TopRightSlope) : (tileEdges | TileEdges.BottomLeftSlope));
 		tileEdges = ((!(velocity.Y > 0f - velocity.X)) ? (tileEdges | TileEdges.TopLeftSlope) : (tileEdges | TileEdges.BottomRightSlope));
 		collisionPoint = Vector2.Zero;
-		collisionTile = null;
+		collisionTile = default;
 		float num = float.MaxValue;
 		Vector2 closestPointOut = default(Vector2);
 		float distanceSquaredOut = 0f;
