--- src/Terraria\Terraria.ID\BuffID.cs
+++ src/tModLoader\Terraria.ID\BuffID.cs
@@ -1,3 +_,6 @@
+using ReLogic.Reflection;
+using Terraria.ModLoader;
+
 namespace Terraria.ID
 {
 	public class BuffID
@@ -208,5 +_,32 @@
 		public const int Oiled = 204;
 		public const int BallistaPanic = 205;
 		public const int Count = 206;
+		public static readonly IdDictionary Search = IdDictionary.Create<BuffID, int>();
+
+		public static string GetUniqueKey(int type) {
+			if (type < 1 || type >= BuffLoader.BuffCount)
+				throw new ArgumentOutOfRangeException("Invalid type: " + type);
+
+			if (type < Count)
+				return "Terraria " + Search.GetName(type);
+
+			var modBuff = BuffLoader.GetBuff(type);
+			return $"{modBuff.mod.Name} {modBuff.Name}";
+		}
+
+		public static int TypeFromUniqueKey(string key) {
+			string[] parts = key.Split(new char[] { ' ' }, 2);
+			if (parts.Length != 2)
+				return 0;
+
+			if (parts[0] == "Terraria") {
+				if (!Search.ContainsName(parts[1]))
+					return 0;
+
+				return Search.GetId(parts[1]);
+			}
+
+			return ModLoader.ModLoader.GetMod(parts[0])?.BuffType(parts[1]) ?? 0;
+		}
 	}
 }
