--- src/Terraria\Terraria.ID\ItemID.cs
+++ src/tModLoader\Terraria.ID\ItemID.cs
@@ -306,7 +_,7 @@
 					1103,
 					0,
 					3347,
-					1
+					3347
 				});
 			public static int[] StaffMinionSlotsRequired = ItemID.Sets.Factory.CreateIntSet(1, new int[0]);
 			public static bool[] ExoticPlantsForDyeTrade = ItemID.Sets.Factory.CreateBoolSet(new int[]
@@ -889,6 +_,7 @@
 					431,
 					432
 				});
+			public static bool[] IsAMaterial = ItemID.Sets.Factory.CreateBoolSet();
 		}
 
 		public const short YellowPhasesaberOld = -24;

