--- src/Terraria\Terraria.ID\NPCID.cs
+++ src/tModLoader\Terraria.ID\NPCID.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
+using ReLogic.Reflection;
 using System;
 using System.Collections.Generic;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -3759,5 +_,34 @@
 			}
 			return id;
 		}
+
+		public static readonly IdDictionary Search = IdDictionary.Create<NPCID, short>();
+
+		public static string GetUniqueKey(int type) {
+			if (type < -65 || type >= NPCLoader.NPCCount)
+				throw new ArgumentOutOfRangeException("Invalid type: " + type);
+			if (type < Count) {
+				return "Terraria " + NPCID.Search.GetName(type);
+			}
+			var modNPC = NPCLoader.GetNPC(type);
+			return $"{modNPC.mod.Name} {modNPC.Name}";
+		}
+
+		public static string GetUniqueKey(NPC npc) => GetUniqueKey(npc.type);
+
+		public static int TypeFromUniqueKey(string key) {
+			string[] parts = key.Split(new char[] { ' ' }, 2);
+			if (parts.Length != 2)
+				return 0;
+			return TypeFromUniqueKey(parts[0], parts[1]);
+		}
+		public static int TypeFromUniqueKey(string mod, string name) {
+			if (mod == "Terraria") {
+				if (!NPCID.Search.ContainsName(name))
+					return 0;
+				return NPCID.Search.GetId(name);
+			}
+			return ModLoader.ModLoader.GetMod(mod)?.NPCType(name) ?? 0;
+		}
 	}
 }

