--- src/Terraria\Terraria.ID\NPCID.cs
+++ src/tModLoader\Terraria.ID\NPCID.cs
@@ -1,4 +_,5 @@
 using Microsoft.Xna.Framework;
+using ReLogic.Reflection;
 using System;
 using System.Collections.Generic;
 
@@ -1777,6 +_,7 @@
 		public const short DD2LightningBugT3 = 578;
 		public const short BartenderUnconscious = 579;
 		public const short Count = 580;
+		public static readonly IdDictionary Search = IdDictionary.Create<NPCID, short>();
 		private static readonly int[] NetIdMap = new int[]
 		{
 			81,
@@ -3759,5 +_,30 @@
 			}
 			return id;
 		}
+
+		public static string GetUniqueId(int npcType) {
+			if ( NPCID.Search.ContainsId(npcType) ) {
+				return "Terraria " + NPCID.Search.GetName(npcType);
+			}
+			else {
+				var npc = new NPC();
+				npc.SetDefaults(npcType);
+
+				if ( npc.modNPC != null ) {
+					return npc.modNPC.mod.Name + " " + npc.modNPC.Name;
+				}
+			}
+
+			return "" + npcType;
+		}
+
+		public static string GetUniqueId(NPC npc) {
+			if ( npc.modNPC == null ) {
+				return "Terraria " + NPCID.Search.GetName(npc.type);
+			}
+			else {
+				return npc.modNPC.mod.Name + " " + npc.modNPC.Name;
+			}
+		}
 	}
 }

