--- src/Terraria\Terraria.ID\ProjectileID.cs
+++ src/tModLoader\Terraria.ID\ProjectileID.cs
@@ -1,4 +_,6 @@
+using ReLogic.Reflection;
 using System;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -1199,5 +_,38 @@
 		public const short BookStaffShot = 712;
 		public const short DD2Win = 713;
 		public const short Count = 714;
+
+		public static readonly IdDictionary Search = IdDictionary.Create<ProjectileID, short>();
+
+		public static string GetReadableUniqueKey(int projType) {
+			if ( ProjectileID.Search.ContainsId(projType) ) {
+				return "Terraria " + ProjectileID.Search.GetName(projType);
+			}
+
+			var proj = new Projectile();
+			proj.SetDefaults(projType);
+
+			if ( proj.modProjectile != null ) {
+				return ProjectileID.GetReadableUniqueKey(proj.modProjectile);
+			}
+
+			throw new ArgumentOutOfRangeException("Invalid `Projectile.type` value.");
+		}
+
+		public static string GetReadableUniqueKey(Projectile proj) {
+			if ( ProjectileID.Search.ContainsId(proj.type) ) {
+				return "Terraria " + ProjectileID.Search.GetName(proj.type);
+			}
+
+			if ( proj.modProjectile != null ) {
+				return ProjectileID.GetReadableUniqueKey(proj.modProjectile);
+			}
+
+			throw new ArgumentException("Invalid `Projectile`.");
+		}
+
+		public static string GetReadableUniqueKey(ModProjectile modProj) {
+			return modProj.mod.Name + " " + modProj.Name;
+		}
 	}
 }

