--- src/Terraria\Terraria.ID\TileID.cs
+++ src/tModLoader\Terraria.ID\TileID.cs
@@ -294,6 +_,20 @@
 					468
 				});
 			public static bool[] AllTiles = TileID.Sets.Factory.CreateBoolSet(true, new int[0]);
+			public static bool[] Stone = TileID.Sets.Factory.CreateBoolSet(new int[]
+				{
+					1,
+					117,
+					25,
+					203
+				});
+			public static bool[] Grass = TileID.Sets.Factory.CreateBoolSet(new int[]
+				{
+					2,
+					23,
+					109,
+					199
+				});
 			public static bool[] Mud = TileID.Sets.Factory.CreateBoolSet(new int[]
 				{
 					59
@@ -385,6 +_,31 @@
 					404,
 					368,
 					367
+				});
+			public static bool[] CanBeClearedDuringOreRunner = TileID.Sets.Factory.CreateBoolSet(new int[]
+				{
+					0,
+					1,
+					23,
+					25,
+					40,
+					53,
+					57,
+					59,
+					60,
+					70,
+					109,
+					112,
+					116,
+					117,
+					147,
+					161,
+					163,
+					164,
+					199,
+					200,
+					203,
+					234
 				});
 			public static bool[] Corrupt = TileID.Sets.Factory.CreateBoolSet(new int[]
 				{

