--- src/Terraria\Terraria.ID\WallID.cs
+++ src/tModLoader\Terraria.ID\WallID.cs
@@ -1,4 +_,6 @@
+using ReLogic.Reflection;
 using System;
+using Terraria.ModLoader;
 
 namespace Terraria.ID
 {
@@ -307,5 +_,24 @@
 		public const byte SillyBalloonPurpleWall = 229;
 		public const byte SillyBalloonGreenWall = 230;
 		public const byte Count = 231;
+
+		public static readonly IdDictionary Search = IdDictionary.Create<WallID, short>();
+
+		public static string GetReadableUniqueKey(int wallType) {
+			if( WallID.Search.ContainsId(wallType) ) {
+				return "Terraria " + WallID.Search.GetName(wallType);
+			}
+
+			ModWall modWall = WallLoader.GetWall(wallType);
+			if( modWall != null ) {
+				return WallID.GetReadableUniqueKey(modWall);
+			}
+
+			throw new ArgumentOutOfRangeException("Invalid wall type value.");
+		}
+
+		public static string GetReadableUniqueKey(ModWall wall) {
+			return wall.mod.Name + " " + wall.Name;
+		}
 	}
 }

