--- src/merged\Terraria.Chat\ChatCommandProcessor.cs
+++ src/Terraria\Terraria.Chat\ChatCommandProcessor.cs
@@ -17,20 +_,18 @@
 
 		public ChatCommandProcessor AddCommand<T>() where T : IChatCommand, new()
 		{
-			ChatCommandProcessor.<>c__DisplayClass2<T> <>c__DisplayClass = new ChatCommandProcessor.<>c__DisplayClass2<T>();
 			ChatCommandAttribute cacheableAttribute = AttributeUtilities.GetCacheableAttribute<T, ChatCommandAttribute>();
-			<>c__DisplayClass.commandKey = "ChatCommand." + cacheableAttribute.Name;
+			string commandKey = "ChatCommand." + cacheableAttribute.Name;
 			ChatCommandId chatCommandId = ChatCommandId.FromType<T>();
 			this._commands[chatCommandId] = ((default(T) == null) ? Activator.CreateInstance<T>() : default(T));
-			if (Language.Exists(<>c__DisplayClass.commandKey))
+			if (Language.Exists(commandKey))
 			{
-				this._localizedCommands.Add(Language.GetText(<>c__DisplayClass.commandKey), chatCommandId);
+				this._localizedCommands.Add(Language.GetText(commandKey), chatCommandId);
 			}
 			else
 			{
-				ChatCommandProcessor.<>c__DisplayClass2<T> expr_8A = <>c__DisplayClass;
-				expr_8A.commandKey += "_";
-				LocalizedText[] array = Language.FindAll((string key, LocalizedText text) => key.StartsWith(<>c__DisplayClass.commandKey));
+				commandKey += "_";
+				LocalizedText[] array = Language.FindAll((string key, LocalizedText text) => key.StartsWith(commandKey));
 				LocalizedText[] array2 = array;
 				for (int i = 0; i < array2.Length; i++)
 				{

