--- src/decompiled/ReLogic/ReLogic.csproj
+++ src/Terraria/ReLogic/ReLogic.csproj
@@ -8,11 +_,15 @@
     <RootNamespace>ReLogic</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.CF">
+      <HintPath>../Terraria/Libraries/DotNetZip/Ionic.Zip.CF.dll</HintPath>
+    </Reference>
     <Reference Include="Microsoft.Xna.Framework" />
     <Reference Include="Microsoft.Xna.Framework.Game" />
     <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Newtonsoft.Json" />
+    <Reference Include="Newtonsoft.Json">
+      <HintPath>../Terraria/Libraries/JSON.NET/Newtonsoft.Json.dll</HintPath>
+    </Reference>
     <Reference Include="System" />
     <Reference Include="System.Core" />
     <Reference Include="System.Drawing" />
