--- src/decompiled/ReLogic/ReLogic.csproj
+++ src/Terraria/ReLogic/ReLogic.csproj
@@ -6,20 +_,18 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © Re-Logic 2017</Copyright>
     <RootNamespace>ReLogic</RootNamespace>
+    <TargetFramework>net5.0</TargetFramework>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="Microsoft.Xna.Framework" />
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Newtonsoft.Json" />
-    <Reference Include="SteelSeriesEngineWrapper" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Windows.Forms" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/FNA/FNA.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
+    <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
   </ItemGroup>
   <ItemGroup>
+	<Compile Remove="Localization/IME/Windows/*" />
+	<Compile Remove="Localization/IME/WindowsIme.cs" />
+	<Compile Remove="OS/Windows/WindowsMessageHook.cs" />
     <EmbeddedResource Include="Microsoft/**" />
   </ItemGroup>
 </Project>
