--- src/decompiled/ReLogic/ReLogic.csproj
+++ src/Terraria/ReLogic/ReLogic.csproj
@@ -8,18 +_,18 @@
     <RootNamespace>ReLogic</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="Microsoft.Xna.Framework" />
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Newtonsoft.Json" />
-    <Reference Include="SteelSeriesEngineWrapper" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Windows.Forms" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/FNA/FNA.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
+    <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Remove="Localization/IME/Windows/*" />
+    <Compile Remove="Localization/IME/WindowsIme.cs" />
+    <Compile Remove="OS/Windows/WindowsMessageHook.cs" />
     <EmbeddedResource Include="Microsoft/**" />
+  </ItemGroup>
+  <ItemGroup>
+      <PackageReference Include="SixLabors.ImageSharp" Version="1.0.3" />
   </ItemGroup>
 </Project>
