--- src/decompiled\Terraria.Social.Steam\CoreSocialModule.cs
+++ src/merged\Terraria.Social.Steam\CoreSocialModule.cs
@@ -25,11 +_,13 @@
 		public void Initialize()
 		{
 			CoreSocialModule._instance = this;
+#if CLIENT
 			if (SteamAPI.RestartAppIfNecessary(new AppId_t(105600u)))
 			{
 				Environment.Exit(1);
 				return;
 			}
+#endif
 			if (!SteamAPI.Init())
 			{
 				MessageBox.Show(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
@@ -40,6 +_,9 @@
 			ThreadPool.QueueUserWorkItem(new WaitCallback(this.SteamTickLoop), null);
 			Main.OnTick += new Action(this.PulseSteamTick);
 			Main.OnTick += new Action(this.PulseSteamCallback);
+#if MAC
+			this._onOverlayActivated = Callback<GameOverlayActivated_t>.Create(new Callback<GameOverlayActivated_t>.DispatchDelegate(this.OnOverlayActivated));
+#endif
 		}
 
 		public void PulseSteamTick()
@@ -94,7 +_,12 @@
 
 		public void Shutdown()
 		{
+#if WINDOWS
 			Application.ApplicationExit += delegate(object obj, EventArgs evt)
+#else
+			this.IsSteamValid = false;
+			AppDomain.CurrentDomain.ProcessExit += delegate(object obj, EventArgs evt)
+#endif
 			{
 				this.IsSteamValid = false;
 			};

