--- src/Terraria/Terraria/Social/WeGame/CoreSocialModule.cs
+++ src/TerrariaNetCore/Terraria/Social/WeGame/CoreSocialModule.cs
@@ -1,7 +_,6 @@
 using System;
 using System.Runtime.InteropServices;
 using System.Threading;
-using System.Windows.Forms;
 using rail;
 using ReLogic.OS;
 using Terraria.Utilities;
@@ -63,12 +_,15 @@
 
 	public void Shutdown()
 	{
+#if !NETCORE
 		if (Platform.IsWindows) {
 			Application.ApplicationExit += delegate {
 				isRailValid = false;
 			};
 		}
-		else {
+		else
+#endif
+		{
 			isRailValid = false;
 			AppDomain.CurrentDomain.ProcessExit += delegate {
 				isRailValid = false;
