--- src/merged\Terraria.ID\SetFactory.cs
+++ src/Terraria\Terraria.ID\SetFactory.cs
@@ -52,18 +52,19 @@
 			return result;
 		}
 
-		public void Recycle<T>(T[] buffer)
+		public void Recycle(bool[] buffer)
 		{
 			lock (this._queueLock)
 			{
-				if (typeof(T).Equals(typeof(bool)))
-				{
-					this._boolBufferCache.Enqueue((bool[])buffer);
+				this._boolBufferCache.Enqueue(buffer);
 				}
-				else if (typeof(T).Equals(typeof(int)))
-				{
-					this._intBufferCache.Enqueue((int[])buffer);
 				}
+
+		public void Recycle(int[] buffer)
+		{
+			lock (this._queueLock)
+			{
+				this._intBufferCache.Enqueue(buffer);
 			}
 		}
 
@@ -117,7 +118,7 @@
 			}
 			for (int j = 0; j < inputs.Length; j += 2)
 			{
-				array[(int)((short)inputs[j])] = (T)((object)inputs[j + 1]);
+				array[(int)inputs[j]] = (T)((object)inputs[j + 1]);
 			}
 			return array;
 		}

