--- src/Terraria\Terraria\Minecart.cs
+++ src/tModLoader\Terraria\Minecart.cs
@@ -55,12 +55,10 @@
 
 		public static void Initialize()
 		{
-#if CLIENT
 			if ((float)Main.minecartMountTexture.Width != 50f)
 			{
 				throw new Exception("Be sure to update Minecart.textureWidth to match the actual texture size of " + Main.minecartMountTexture.Width + ".");
 			}
-#endif
 			Minecart._rightSideConnection = new int[36];
 			Minecart._leftSideConnection = new int[36];
 			Minecart._trackType = new int[36];

