--- src/Terraria\Terraria\Mount.cs
+++ src/tModLoader\Terraria\Mount.cs
@@ -946,20 +946,18 @@
 				mountData.textureHeight = mountData.frontTexture.Height;
 			}
 			Mount.drillTextureSize = new Vector2(80f, 80f);
-#if CLIENT
 			Vector2 value = new Vector2((float)mountData.textureWidth, (float)(mountData.textureHeight / mountData.totalFrames));
 			if (Mount.drillTextureSize != value)
 			{
 				throw new Exception(string.Concat(new object[]
-				{
-					"Be sure to update the Drill texture origin to match the actual texture size of ",
-					mountData.textureWidth,
-					", ",
-					mountData.textureHeight,
-					"."
-				}));
-			}
-#endif
+						{
+							"Be sure to update the Drill texture origin to match the actual texture size of ",
+							mountData.textureWidth,
+							", ",
+							mountData.textureHeight,
+							"."
+						}));
+			}
 			mountData = new Mount.MountData();
 			Mount.mounts[9] = mountData;
 			mountData.spawnDust = 152;
@@ -1028,20 +1026,18 @@
 			Mount.scutlixEyePositions[8] = new Vector2(70f, 34f);
 			Mount.scutlixEyePositions[9] = new Vector2(76f, 34f);
 			Mount.scutlixTextureSize = new Vector2(45f, 54f);
-#if CLIENT
 			Vector2 value2 = new Vector2((float)(mountData.textureWidth / 2), (float)(mountData.textureHeight / mountData.totalFrames));
 			if (Mount.scutlixTextureSize != value2)
 			{
 				throw new Exception(string.Concat(new object[]
-				{
-					"Be sure to update the Scutlix texture origin to match the actual texture size of ",
-					mountData.textureWidth,
-					", ",
-					mountData.textureHeight,
-					"."
-				}));
-			}
-#endif
+						{
+							"Be sure to update the Scutlix texture origin to match the actual texture size of ",
+							mountData.textureWidth,
+							", ",
+							mountData.textureHeight,
+							"."
+						}));
+			}
 			for (int num5 = 0; num5 < Mount.scutlixEyePositions.Length; num5++)
 			{
 				Mount.scutlixEyePositions[num5] -= Mount.scutlixTextureSize;

