--- src/Terraria\Terraria\NetMessage.cs
+++ src/tModLoader\Terraria\NetMessage.cs
@@ -1385,7 +1385,7 @@
 					{
 						memoryStream.Position = 0L;
 						MemoryStream memoryStream2 = new MemoryStream();
-						using (DeflateStream deflateStream = new DeflateStream(memoryStream2, CompressionMode.Compress, true))
+						using (DeflateStream deflateStream = new DeflateStream(memoryStream2, 0, true))
 						{
 							memoryStream.CopyTo(deflateStream);
 							deflateStream.Flush();
@@ -1557,7 +1557,7 @@
 						if (tile2.wall != 0)
 						{
 							b |= 4;
-							array4[num5] = tile2.wall;
+							array4[num5] = (byte)tile2.wall;
 							num5++;
 							if (tile2.wallColor() != 0)
 							{
@@ -1959,12 +1959,6 @@
 				}
 				catch
 				{
-#if SERVER
-					if (num < NetMessage.buffer.Length - 100)
-					{
-						Console.WriteLine("Error on message " + NetMessage.buffer[num + 2]);
-					}
-#endif
 					i = 0;
 					num = 0;
 				}

