--- src/TerrariaNetCore/Terraria/GameContent/ObjectInteractions/BlockBecauseYouAreOverAnImportantTile.cs
+++ src/tModLoader/Terraria/GameContent/ObjectInteractions/BlockBecauseYouAreOverAnImportantTile.cs
@@ -1,3 +_,5 @@
+using Terraria.ID;
+
 namespace Terraria.GameContent.ObjectInteractions;
 
 public class BlockBecauseYouAreOverAnImportantTile : ISmartInteractBlockReasonProvider
@@ -6,6 +_,7 @@
 	{
 		int tileTargetX = Player.tileTargetX;
 		int tileTargetY = Player.tileTargetY;
+		// Extra patch context.
 		if (!WorldGen.InWorld(tileTargetX, tileTargetY, 10))
 			return true;
 
@@ -15,12 +_,14 @@
 
 		if (tile.active()) {
 			switch (tile.type) {
+				/*
 				case 4:
 				case 33:
 				case 334:
 				case 395:
 				case 410:
 				case 455:
+				// Extra patch context.
 				case 471:
 				case 480:
 				case 509:
@@ -28,6 +_,9 @@
 				case 657:
 				case 658:
 					return true;
+				*/
+				default:
+					return TileID.Sets.DisableSmartInteract[tile.type];
 			}
 		}
 
