--- src/TerrariaNetCore/Terraria/GameContent/ObjectInteractions/BlockBecauseYouAreOverAnImportantTile.cs
+++ src/tModLoader/Terraria/GameContent/ObjectInteractions/BlockBecauseYouAreOverAnImportantTile.cs
@@ -3,3 +_,5 @@
+using Terraria.ID;
+
 namespace Terraria.GameContent.ObjectInteractions;
 
 public class BlockBecauseYouAreOverAnImportantTile : ISmartInteractBlockReasonProvider
@@ -8,6 +_,7 @@
 	{
 		int tileTargetX = Player.tileTargetX;
 		int tileTargetY = Player.tileTargetY;
+		// Extra patch context.
 		if (!WorldGen.InWorld(tileTargetX, tileTargetY, 10))
 			return true;
 
@@ -17,11 +_,13 @@
 
 		if (tile.active()) {
 			switch (tile.type) {
+				/*
 				case 33:
 				case 334:
 				case 395:
 				case 410:
 				case 455:
+				// Extra patch context.
 				case 471:
 				case 480:
 				case 509:
@@ -34,5 +_,8 @@
 				case 725:
 				case 733:
 					return true;
+				*/
+				default:
+					return TileID.Sets.DisableSmartInteract[tile.type];
 			}
 
