--- src/Terraria/Terraria/GameContent/ObjectInteractions/TileSmartInteractCandidateProvider.cs
+++ src/tModLoader/Terraria/GameContent/ObjectInteractions/TileSmartInteractCandidateProvider.cs
@@ -1,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using System;
 using System.Collections.Generic;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.ObjectInteractions
 {
@@ -350,6 +_,9 @@
 						case 356:
 							if (!Main.fastForwardTime && (Main.netMode == 1 || Main.sundialCooldown == 0))
 								targets.Add(new Tuple<int, int>(i, j));
+							break;
+						default:
+							TileLoader.HasSmartInteract(tile.type);
 							break;
 					}
 				}
