--- src/merged\Terraria.UI.Gamepad\UILinkPointNavigator.cs
+++ src/Terraria\Terraria.UI.Gamepad\UILinkPointNavigator.cs
@@ -294,7 +_,7 @@
 			}
 			UILinkPointNavigator.Pages[UILinkPointNavigator.CurrentPage].Update();
 			int num = 12;
-			navigatorDirections != Vector2.Zero;
+			bool _ = navigatorDirections != Vector2.Zero;
 			if (navigatorDirections.X == -1f && UILinkPointNavigator.XCooldown == 0)
 			{
 				UILinkPointNavigator.XCooldown = num;

