--- src/tModLoader/Terraria.ModLoader.IO/WorldIO.cs
+++ src/64bit/Terraria.ModLoader.IO/WorldIO.cs
@@ -5,6 +_,7 @@
 using Terraria.ID;
 using Terraria.ModLoader.Default;
 using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.x64bit;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -76,7 +_,7 @@
 
 		internal static List<TagCompound> SaveChests() {
 			var list = new List<TagCompound>();
-			for (int k = 0; k < 1000; k++) {
+			for (int k = 0; k < Core64.maxChest; k++) {
 				var chest = Main.chest[k];
 				if (chest == null)
 					continue;
@@ -91,6 +_,7 @@
 					["y"] = chest.y
 				});
 			}
+
 			return list;
 		}
 
@@ -514,13 +_,16 @@
 		internal static void EraseWorld(string path, bool cloudSave) {
 			path = Path.ChangeExtension(path, ".twld");
 			if (!cloudSave) {
-#if WINDOWS
-				FileOperationAPIWrapper.MoveToRecycleBin(path);
-				FileOperationAPIWrapper.MoveToRecycleBin(path + ".bak");
-#else
-				File.Delete(path);
-				File.Delete(path + ".bak");
-#endif
+				if (PlatformUtilities.IsWindows) {
+					FileOperationAPIWrapper.MoveToRecycleBin(path);
+					FileOperationAPIWrapper.MoveToRecycleBin(path + ".bak");
+				}
+				else {
+					File.Delete(path);
+					File.Delete(path + ".bak");
+				}
+
+
 			}
 			else if (SocialAPI.Cloud != null) {
 				SocialAPI.Cloud.Delete(path);
