--- src/TerrariaNetCore/Terraria/GameContent/Drawing/WallDrawing.cs
+++ src/tModLoader/Terraria/GameContent/Drawing/WallDrawing.cs
@@ -4,6 +_,7 @@
 using Terraria.Graphics;
 using Terraria.ID;
 using Terraria.Testing;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.Drawing;
 
@@ -11,7 +_,7 @@
 {
 	public static bool QuickPaintLookup = true;
 	private static VertexColors _glowPaintColors = new VertexColors(Color.White);
-	private Tile[,] _tileArray;
+	private Tilemap _tileArray;
 	private TilePaintSystemV2 _paintSystem;
 	private bool _shouldShowInvisibleWalls;
 	private DrawBlackHelper drawBlackHelper;
@@ -66,8 +_,12 @@
 		int underworldLayer = Main.UnderworldLayer;
 		drawBlackHelper = new DrawBlackHelper(0u, drawOffSet);
 		_lastPaintLookupKey = default(TilePaintSystemV2.WallVariationKey);
+
+		// TML:
+		// Swapped loop iteration order to improve memory access linearity.
+		// Minor perf improvement, may cause slight visual discrepancies with tiles that overdraw
-		for (int i = firstTileY; i < lastTileY; i++) {
+		for (int j = firstTileX; j < lastTileX; j++) {
-			for (int j = firstTileX; j < lastTileX; j++) {
+			for (int i = firstTileY; i < lastTileY; i++) {
 				Tile tile = _tileArray[j, i];
 				if (tile == null) {
 					tile = new Tile();
@@ -75,9 +_,15 @@
 				}
 
 				ushort wall = tile.wall;
+				//TML: Prior to 1.4.4, the check was just the line below. Should PreDraw be called for "invisible walls"? Tiles do call PreDraw for invisible tiles, so it would make sense, but this behavior has been here since 1.4.4.
+				//if (wall <= 0 || FullTile(j, i))
 				if (wall <= 0 || FullTile(j, i) || (wall == 318 && !_shouldShowInvisibleWalls) || (tile.invisibleWall() && !_shouldShowInvisibleWalls))
 					continue;
 
+				if (!WallLoader.PreDraw(j, i, wall, Main.spriteBatch)) {
+					goto PostDraw;
+				}
+
 				Color color = Lighting.GetColor(j, i);
 				if (tile.fullbrightWall())
 					color = Color.White;
@@ -164,6 +_,9 @@
 					if (flag3)
 						DrawOutline(TextureAssets.WallOutline.Value, new Vector2(j * 16 - (int)screenPosition.X, i * 16 - (int)screenPosition.Y + 14) + drawOffSet, new Rectangle(0, 14, 16, 2), color);
 				}
+
+				PostDraw:
+				WallLoader.PostDraw(j, i, wall, Main.spriteBatch);
 			}
 		}
 
