--- src/Terraria\Terraria\Chest.cs
+++ src/tModLoader\Terraria\Chest.cs
@@ -4,7 +_,10 @@
 using Terraria.GameContent.Achievements;
 using Terraria.GameContent.Events;
 using Terraria.ID;
+using Terraria.Localization;
 using Terraria.ObjectData;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 
 namespace Terraria
 {
@@ -172,8 +_,10 @@
 
 		public static void ServerPlaceItem(int plr, int slot)
 		{
+			//@todo: need to debug server here
 			Main.player[plr].inventory[slot] = Chest.PutItemInNearbyChest(Main.player[plr].inventory[slot], Main.player[plr].Center);
-			NetMessage.SendData(5, -1, -1, null, plr, (float)slot, (float)Main.player[plr].inventory[slot].prefix, 0f, 0, 0, 0);
+			NetworkText txt = NetworkText.FromLiteral(ItemIO.ToBase64(Main.player[plr].inventory[slot]));
+			NetMessage.SendData(5, -1, -1, txt, plr, (float)slot, (float)Main.player[plr].inventory[slot].prefix, 0f, 0, 0, 0);
 		}
 
 		public static Item PutItemInNearbyChest(Item item, Vector2 position)
@@ -418,17 +_,25 @@
 				}
 				Main.chest[num] = chest;
 			}
-			else if (type == 21)
+			else if (type == TileID.Containers)
 			{
 				NetMessage.SendData(34, -1, -1, null, 0, (float)x, (float)y, (float)style, 0, 0, 0);
 			}
-			else if (type == 467)
+			else if (type == TileID.Containers2)
 			{
 				NetMessage.SendData(34, -1, -1, null, 4, (float)x, (float)y, (float)style, 0, 0, 0);
 			}
-			else
+			else if (type == TileID.Dressers)
 			{
 				NetMessage.SendData(34, -1, -1, null, 2, (float)x, (float)y, (float)style, 0, 0, 0);
+			}
+			else if (TileID.Sets.BasicChest[type])
+			{
+				NetMessage.SendData(34, -1, -1, null, 100, (float)x, (float)y, (float)style, 0, type, 0);
+			}
+			else if (TileLoader.IsDresser(type))
+			{
+				NetMessage.SendData(34, -1, -1, null, 102, (float)x, (float)y, (float)style, 0, type, 0);
 			}
 			return num;
 		}
@@ -882,9 +_,11 @@
 								Main.travelShop[num2++] = 3641;
 								break;
 						}
-					}
-				}
-			}
+						//patch file: num2
+					}
+				}
+			}
+			NPCLoader.SetupTravelShop(Main.travelShop, ref num2);
 		}
 
 		public void SetupShop(int type)
@@ -2229,6 +_,7 @@
 					num++;
 				}
 			}
+			NPCLoader.SetupShop(type, this, ref num);
 			if (Main.player[Main.myPlayer].discount)
 			{
 				for (int num7 = 0; num7 < num; num7++)

