--- src/Terraria\Terraria\Chest.cs
+++ src/tModLoader\Terraria\Chest.cs
@@ -5,6 +_,7 @@
 using Terraria.GameContent.Events;
 using Terraria.ID;
 using Terraria.ObjectData;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -393,13 +_,21 @@
 				}
 				Main.chest[num] = chest;
 			}
-			else if (type == 21)
+			else if (type == TileID.Containers)
 			{
 				NetMessage.SendData(34, -1, -1, "", 0, (float)x, (float)y, (float)style, 0, 0, 0);
 			}
-			else
+			else if (type == TileID.Dressers)
 			{
 				NetMessage.SendData(34, -1, -1, "", 2, (float)x, (float)y, (float)style, 0, 0, 0);
+			}
+			else if (TileLoader.IsChest(type))
+			{
+				NetMessage.SendData(34, -1, -1, "", 100, (float)x, (float)y, (float)style, 0, type, 0);
+			}
+			else if (TileLoader.IsDresser(type))
+			{
+				NetMessage.SendData(34, -1, -1, "", 102, (float)x, (float)y, (float)style, 0, type, 0);
 			}
 			return num;
 		}
@@ -853,9 +_,11 @@
 								Main.travelShop[num2++] = 3641;
 								break;
 						}
-					}
-				}
-			}
+						//patch file: num2
+					}
+				}
+			}
+			NPCLoader.SetupTravelShop(Main.travelShop, ref num2);
 		}
 
 		public void SetupShop(int type)
@@ -2200,6 +_,7 @@
 					num++;
 				}
 			}
+			NPCLoader.SetupShop(type, this, ref num);
 			if (Main.player[Main.myPlayer].discount)
 			{
 				for (int num7 = 0; num7 < num; num7++)

