--- src/Terraria\Terraria\Chest.cs
+++ src/tModLoader\Terraria\Chest.cs
@@ -2,6 +_,7 @@
 using System;
 using Terraria.DataStructures;
 using Terraria.GameContent.Achievements;
+using Terraria.ModLoader;
 using Terraria.ObjectData;
 
 namespace Terraria
@@ -851,9 +_,11 @@
 								Main.travelShop[num2++] = 3641;
 								break;
 						}
-					}
-				}
-			}
+						//patch file: num2
+					}
+				}
+			}
+			NPCLoader.SetupTravelShop(Main.travelShop, ref num2);
 		}
 
 		public void SetupShop(int type)
@@ -1982,6 +_,7 @@
 					num++;
 				}
 			}
+			NPCLoader.SetupShop(type, this, ref num);
 			if (Main.player[Main.myPlayer].discount)
 			{
 				for (int num7 = 0; num7 < num; num7++)

