--- src/TerrariaNetCore/Terraria/Cloud.cs
+++ src/tModLoader/Terraria/Cloud.cs
@@ -1,11 +_,12 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.GameContent;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 
 namespace Terraria;
 
-public class Cloud
+public partial class Cloud
 {
 	public Vector2 position;
 	public float scale;
@@ -87,29 +_,37 @@
 		if (rand.Next(3) == 0)
 			Main.cloud[num].position.Y -= rand.Next((int)((float)Main.screenHeight * 0.1f));
 
+		int? cloudChoice = CloudLoader.ChooseCloud(num);
+
+		if (!CloudLoader.cloudLoaded || !cloudChoice.HasValue || cloudChoice.Value == 0 || cloudChoice.Value >= CloudLoader.CloudCount || cloudChoice.Value >= TextureAssets.Cloud.Length || !TextureAssets.Cloud[cloudChoice.Value].IsLoaded) {
-		Main.cloud[num].type = rand.Next(4);
+			Main.cloud[num].type = rand.Next(4);
-		if ((Main.cloudAlpha > 0f && rand.Next(4) != 0) || (Main.cloudBGActive >= 1f && rand.Next(2) == 0)) {
+			if ((Main.cloudAlpha > 0f && rand.Next(4) != 0) || (Main.cloudBGActive >= 1f && rand.Next(2) == 0)) {
-			Main.cloud[num].type = rand.Next(18, 22);
+				Main.cloud[num].type = rand.Next(18, 22);
-			if ((double)Main.cloud[num].scale >= 1.15)
+				if ((double)Main.cloud[num].scale >= 1.15)
-				Main.cloud[num].position.Y -= 150f;
+					Main.cloud[num].position.Y -= 150f;
 
-			if (Main.cloud[num].scale >= 1f)
+				if (Main.cloud[num].scale >= 1f)
-				Main.cloud[num].position.Y -= 150f;
+					Main.cloud[num].position.Y -= 150f;
-		}
+			}
-		else if (Main.cloudBGActive <= 0f && Main.cloudAlpha == 0f && Main.cloud[num].scale < 1f && Main.cloud[num].position.Y < (float)(-Main.screenHeight) * 0.15f && (double)Main.numClouds <= 80.0) {
+			else if (Main.cloudBGActive <= 0f && Main.cloudAlpha == 0f && Main.cloud[num].scale < 1f && Main.cloud[num].position.Y < (float)(-Main.screenHeight) * 0.15f && (double)Main.numClouds <= 80.0) {
-			Main.cloud[num].type = rand.Next(9, 14);
+				Main.cloud[num].type = rand.Next(9, 14);
-		}
+			}
-		else if ((((double)Main.cloud[num].scale < 1.15 && Main.cloud[num].position.Y < (float)(-Main.screenHeight) * 0.3f) || ((double)Main.cloud[num].scale < 0.85 && Main.cloud[num].position.Y < (float)Main.screenHeight * 0.15f)) && ((double)Main.numClouds > 70.0 || Main.cloudBGActive >= 1f)) {
+			else if ((((double)Main.cloud[num].scale < 1.15 && Main.cloud[num].position.Y < (float)(-Main.screenHeight) * 0.3f) || ((double)Main.cloud[num].scale < 0.85 && Main.cloud[num].position.Y < (float)Main.screenHeight * 0.15f)) && ((double)Main.numClouds > 70.0 || Main.cloudBGActive >= 1f)) {
-			Main.cloud[num].type = rand.Next(4, 9);
+				Main.cloud[num].type = rand.Next(4, 9);
-		}
+			}
-		else if (Main.cloud[num].position.Y > (float)(-Main.screenHeight) * 0.15f && rand.Next(2) == 0 && (double)Main.numClouds > 20.0) {
+			else if (Main.cloud[num].position.Y > (float)(-Main.screenHeight) * 0.15f && rand.Next(2) == 0 && (double)Main.numClouds > 20.0) {
-			Main.cloud[num].type = rand.Next(14, 18);
+				Main.cloud[num].type = rand.Next(14, 18);
-		}
+			}
 
-		if (rand.Next((Main.dontStarveWorld || Main.tenthAnniversaryWorld) ? 25 : 150) == 0)
+			if (rand.Next((Main.dontStarveWorld || Main.tenthAnniversaryWorld) ? 25 : 150) == 0)
-			Main.cloud[num].type = RollRareCloud();
+				Main.cloud[num].type = RollRareCloud();
-		else if (Main.tenthAnniversaryWorld && rand.Next(3) == 0)
+			else if (Main.tenthAnniversaryWorld && rand.Next(3) == 0)
-			Main.cloud[num].type = RollRareCloud();
+				Main.cloud[num].type = RollRareCloud();
+		}
+		else {
+			Main.cloud[num].type = cloudChoice.Value;
+		}
+			
 
 		if ((double)Main.cloud[num].scale > 1.2)
 			Main.cloud[num].position.Y += 100f;
