--- src/Terraria\Terraria\Collision.cs
+++ src/tModLoader\Terraria\Collision.cs
@@ -2548,6 +_,7 @@
 							return new Vector2((float)num9, (float)num5);
 						}
 					}
+					//TODO - New Hook - ModTile.HurtPlayer
 				}
 			}
 			return default(Vector2);
@@ -3035,7 +_,7 @@
 					if (Main.tileSolid[(int)tile.type])
 					{
 						int num4 = tile.blockType();
-						if (tile.type == 19)
+						if (TileID.Sets.Platforms[tile.type])
 						{
 							int num5 = (int)(tile.frameX / 18);
 							if (((num5 >= 0 && num5 <= 7) || (num5 >= 12 && num5 <= 16)) && (num == 0f || flag))

