--- src/TerrariaNetCore/Terraria/Collision.cs
+++ src/tModLoader/Terraria/Collision.cs
@@ -3,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria;
 
@@ -1030,7 +_,7 @@
 		if (tile == null || !tile.active() || tile.inActive() || !Main.tileSolid[tile.type])
 			return false;
 
-		if (treatPlatformsAsNonSolid && tile.type > 0 && tile.type <= TileID.Count && (TileID.Sets.Platforms[tile.type] || tile.type == 380))
+		if (treatPlatformsAsNonSolid && tile.type > 0 && (TileID.Sets.Platforms[tile.type] || tile.type == 380))
 			return false;
 
 		int num = tile.blockType();
@@ -2614,7 +_,10 @@
 			tile = Main.tile[num2, num3];
 			tile2 = Main.tile[num2, num3 - 1];
 			if (specialChecksMode == 1)
+				/*
 				flag5 = tile.type != 16 && tile.type != 18 && tile.type != 14 && tile.type != 469 && tile.type != 134;
+				*/
+				flag5 = !TileID.Sets.IgnoredByNpcStepUp[tile.type];
 
 			flag4 = flag4 && ((tile.nactive() && (!tile.topSlope() || (tile.slope() == 1 && position.X + (float)(width / 2) < (float)(num2 * 16)) || (tile.slope() == 2 && position.X + (float)(width / 2) > (float)(num2 * 16 + 16))) && (!tile.topSlope() || position.Y + (float)height > (float)(num3 * 16)) && ((Main.tileSolid[tile.type] && !Main.tileSolidTop[tile.type]) || (holdsMatching && ((Main.tileSolidTop[tile.type] && tile.frameY == 0) || TileID.Sets.Platforms[tile.type]) && (!Main.tileSolid[tile2.type] || !tile2.nactive()) && flag5))) || (tile2.halfBrick() && tile2.nactive()));
 			flag4 &= !Main.tileSolidTop[tile.type] || !Main.tileSolidTop[tile2.type];
@@ -2694,7 +_,7 @@
 			if (tile.active()) {
 				if (Main.tileSolid[tile.type]) {
 					int num5 = tile.blockType();
-					if (tile.type != 19) {
+					if (!TileID.Sets.Platforms[tile.type]) {
 						switch (num5) {
 							case 1:
 								return 0f;
