--- src/Terraria/Terraria/Collision.cs
+++ src/tModLoader/Terraria/Collision.cs
@@ -2100,7 +_,7 @@
 			Vector2 vector2 = default(Vector2);
 			for (int i = num; i < num2; i++) {
 				for (int j = num3; j < num4; j++) {
-					if (Main.tile[i, j] == null || Main.tile[i, j].slope() != 0 || Main.tile[i, j].inActive() || !Main.tile[i, j].active() || (Main.tile[i, j].type != 32 && Main.tile[i, j].type != 37 && Main.tile[i, j].type != 48 && Main.tile[i, j].type != 232 && Main.tile[i, j].type != 53 && Main.tile[i, j].type != 57 && Main.tile[i, j].type != 58 && Main.tile[i, j].type != 69 && Main.tile[i, j].type != 76 && Main.tile[i, j].type != 112 && Main.tile[i, j].type != 116 && Main.tile[i, j].type != 123 && Main.tile[i, j].type != 224 && Main.tile[i, j].type != 234 && Main.tile[i, j].type != 352 && Main.tile[i, j].type != 484 && (!Main.dontStarveWorld || Main.tile[i, j].type != 80)))
+					if (Main.tile[i, j] == null || Main.tile[i, j].slope() != 0 || Main.tile[i, j].inActive() || !Main.tile[i, j].active() || (TileID.Sets.TouchDamageVines[Main.tile[i, j].type] <= 0 || Main.tile[i, j].type == 80) && TileID.Sets.TouchDamageHot[Main.tile[i, j].type] != 20 && TileID.Sets.TouchDamageOther[Main.tile[i, j].type] <= 0 && (TileID.Sets.TouchDamageSands[Main.tile[i, j].type] != 15 || Main.tile[i, j].type == 69) && Main.tile[i, j].type != 484 && (!Main.dontStarveWorld || Main.tile[i, j].type != 80))
 						continue;
 
 					vector2.X = i * 16;
@@ -2113,11 +_,15 @@
 						num6 -= 8;
 					}
 
+
 					switch (type) {
+						/*
 						case 32:
 						case 69:
 						case 80:
 						case 352: {
+						*/
+						case int _ when TileID.Sets.TouchDamageVines[type] > 0: {
 								if (!(vector.X + (float)Width > vector2.X) || !(vector.X < vector2.X + 16f) || !(vector.Y + (float)Height > vector2.Y) || !(vector.Y < vector2.Y + (float)num6 + 0.011f))
 									continue;
 
@@ -2125,6 +_,7 @@
 								if (vector.X + (float)(Width / 2) < vector2.X + 8f)
 									num8 = -1;
 
+								/*
 								num5 = 10;
 								switch (type) {
 									case 69:
@@ -2134,6 +_,8 @@
 										num5 = 6;
 										break;
 								}
+								*/
+								num5 = TileID.Sets.TouchDamageVines[type];
 
 								if (type == 32 || type == 69 || type == 352) {
 									WorldGen.KillTile(i, j);
@@ -2143,12 +_,15 @@
 
 								return new Vector2(num8, num5);
 							}
+						/*
 						case 53:
 						case 112:
 						case 116:
 						case 123:
 						case 224:
 						case 234:
+						*/
+						case int _ when TileID.Sets.TouchDamageSands[type] > 0 && type != 57 && type != 69:
 							if (vector.X + (float)Width - 2f >= vector2.X && vector.X + 2f <= vector2.X + 16f && vector.Y + (float)Height - 2f >= vector2.Y && vector.Y + 2f <= vector2.Y + (float)num6) {
 								int num7 = 1;
 								if (vector.X + (float)(Width / 2) < vector2.X + 8f)
@@ -2165,14 +_,11 @@
 						if (vector.X + (float)(Width / 2) < vector2.X + 8f)
 							num9 = -1;
 
-						if (!fireImmune && (type == 37 || type == 58 || type == 76))
+						if (!fireImmune && TileID.Sets.TouchDamageHot[type] == 20)
 							num5 = 20;
 
-						if (type == 48)
-							num5 = 60;
-
-						if (type == 232)
-							num5 = 80;
+						if (TileID.Sets.TouchDamageOther[type] > 0)
+							num5 = TileID.Sets.TouchDamageOther[type];
 
 						if (type == 484)
 							num5 = 25;
@@ -2180,6 +_,7 @@
 						if (num5 > 0)
 							return new Vector2(num9, num5);
 					}
+					//TODO - New Hook - ModTile.HurtPlayer
 				}
 			}
 
@@ -2564,7 +_,8 @@
 				tile = Main.tile[num2, num3];
 				tile2 = Main.tile[num2, num3 - 1];
 				if (specialChecksMode == 1)
-					flag5 = (tile.type != 16 && tile.type != 18 && tile.type != 14 && tile.type != 469 && tile.type != 134);
+					//flag5 = (tile.type != 16 && tile.type != 18 && tile.type != 14 && tile.type != 469 && tile.type != 134);
+					flag5 = !TileID.Sets.IgnoredByNpcStepUp[tile.type];
 
 				flag4 = (flag4 && ((tile.nactive() && (!tile.topSlope() || (tile.slope() == 1 && position.X + (float)(width / 2) < (float)(num2 * 16)) || (tile.slope() == 2 && position.X + (float)(width / 2) > (float)(num2 * 16 + 16))) && (!tile.topSlope() || position.Y + (float)height > (float)(num3 * 16)) && ((Main.tileSolid[tile.type] && !Main.tileSolidTop[tile.type]) || (holdsMatching && ((Main.tileSolidTop[tile.type] && tile.frameY == 0) || TileID.Sets.Platforms[tile.type]) && (!Main.tileSolid[tile2.type] || !tile2.nactive()) && flag5))) || (tile2.halfBrick() && tile2.nactive())));
 				flag4 &= (!Main.tileSolidTop[tile.type] || !Main.tileSolidTop[tile2.type]);
@@ -2642,7 +_,7 @@
 				if (tile.active()) {
 					if (Main.tileSolid[tile.type]) {
 						int num5 = tile.blockType();
-						if (tile.type != 19) {
+						if (!TileID.Sets.Platforms[tile.type]) {
 							switch (num5) {
 								case 1:
 									return 0f;
