--- src/Terraria/Terraria/Collision.cs
+++ src/tModLoader/Terraria/Collision.cs
@@ -2062,7 +_,7 @@
 			Vector2 vector2 = default(Vector2);
 			for (int i = num; i < num2; i++) {
 				for (int j = num3; j < num4; j++) {
-					if (Main.tile[i, j] == null || Main.tile[i, j].slope() != 0 || Main.tile[i, j].inActive() || !Main.tile[i, j].active() || (Main.tile[i, j].type != 32 && Main.tile[i, j].type != 37 && Main.tile[i, j].type != 48 && Main.tile[i, j].type != 232 && Main.tile[i, j].type != 53 && Main.tile[i, j].type != 57 && Main.tile[i, j].type != 58 && Main.tile[i, j].type != 69 && Main.tile[i, j].type != 76 && Main.tile[i, j].type != 112 && Main.tile[i, j].type != 116 && Main.tile[i, j].type != 123 && Main.tile[i, j].type != 224 && Main.tile[i, j].type != 234 && Main.tile[i, j].type != 352 && Main.tile[i, j].type != 484))
+					if (Main.tile[i, j] == null || Main.tile[i, j].slope() != 0 || Main.tile[i, j].inActive() || !Main.tile[i, j].active() || (TileID.Sets.TouchDamageVines[Main.tile[i, j].type] <= 0 || Main.tile[i, j].type == 80) && TileID.Sets.TouchDamageHot[Main.tile[i, j].type] != 20 && TileID.Sets.TouchDamageOther[Main.tile[i, j].type] <= 0 && (TileID.Sets.TouchDamageSands[Main.tile[i, j].type] != 15 || Main.tile[i, j].type == 69) && Main.tile[i, j].type != 484)
 						continue;
 
 					vector2.X = i * 16;
@@ -2075,7 +_,7 @@
 						num6 -= 8;
 					}
 
-					if (type == 32 || type == 69 || type == 80 || type == 352 || (type == 80 && Main.expertMode)) {
+					if (TileID.Sets.TouchDamageVines[type] > 0 || type == 80 && Main.expertMode) {
 						if (!(vector.X + (float)Width > vector2.X) || !(vector.X < vector2.X + 16f) || !(vector.Y + (float)Height > vector2.Y) || !(vector.Y < vector2.Y + (float)num6 + 0.011f))
 							continue;
 
@@ -2083,15 +_,7 @@
 						if (vector.X + (float)(Width / 2) < vector2.X + 8f)
 							num7 = -1;
 
-						num5 = 10;
-						switch (type) {
-							case 69:
-								num5 = 17;
-								break;
-							case 80:
-								num5 = 6;
-								break;
-						}
+						num5 = TileID.Sets.TouchDamageVines[type];
 
 						if (type == 32 || type == 69 || type == 352) {
 							WorldGen.KillTile(i, j);
@@ -2102,7 +_,7 @@
 						return new Vector2(num7, num5);
 					}
 
-					if (type == 53 || type == 112 || type == 116 || type == 123 || type == 224 || type == 234) {
+					else if (TileID.Sets.TouchDamageSands[type] == 15 && type != 57 && type != 69) {
 						if (vector.X + (float)Width - 2f >= vector2.X && vector.X + 2f <= vector2.X + 16f && vector.Y + (float)Height - 2f >= vector2.Y && vector.Y + 2f <= vector2.Y + (float)num6) {
 							int num8 = 1;
 							if (vector.X + (float)(Width / 2) < vector2.X + 8f)
@@ -2117,14 +_,11 @@
 						if (vector.X + (float)(Width / 2) < vector2.X + 8f)
 							num9 = -1;
 
-						if (!fireImmune && (type == 37 || type == 58 || type == 76))
+						if (!fireImmune && TileID.Sets.TouchDamageHot[type] == 20)
 							num5 = 20;
 
-						if (type == 48)
-							num5 = 60;
-
-						if (type == 232)
-							num5 = 80;
+						if (TileID.Sets.TouchDamageOther[type] > 0)
+							num5 = TileID.Sets.TouchDamageOther[type];
 
 						if (type == 484)
 							num5 = 25;
@@ -2132,6 +_,7 @@
 						if (num5 > 0)
 							return new Vector2(num9, num5);
 					}
+					//TODO - New Hook - ModTile.HurtPlayer
 				}
 			}
 
@@ -2558,7 +_,7 @@
 				if (tile.active()) {
 					if (Main.tileSolid[tile.type]) {
 						int num5 = tile.blockType();
-						if (tile.type != 19) {
+						if (!TileID.Sets.Platforms[tile.type]) {
 							switch (num5) {
 								case 1:
 									return 0f;
