--- src/TerrariaNetCore/Terraria/Collision.cs
+++ src/tModLoader/Terraria/Collision.cs
@@ -1,8 +_,10 @@
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria;
 
@@ -1003,7 +_,16 @@
 		for (int i = num3; i < value2; i++) {
 			for (int j = value3; j < value4; j++) {
 				Tile tile = Main.tile[i, j];
+				bool? canLiquidTypeDrown = LiquidLoader.ChecksForDrowning(tile.liquidType());
+				if (canLiquidTypeDrown == null) {
+					canLiquidTypeDrown = true;
+					if (tile.lava() || tile.shimmer())
+						canLiquidTypeDrown = false;
+
+					if (tile.liquidType() >= LiquidID.Count)
+						canLiquidTypeDrown = LiquidLoader.GetLiquid(tile.liquidType()).ChecksForDrowning;
+				}
-				if (tile != null && tile.liquid > 0 && !tile.lava() && !tile.shimmer() && (j != num4 || !tile.active() || !Main.tileSolid[tile.type] || Main.tileSolidTop[tile.type] || (includeSlopes && tile.blockType() != 0))) {
+				if (tile != null && tile.liquid > 0 && (bool)canLiquidTypeDrown && (j != num4 || !tile.active() || !Main.tileSolid[tile.type] || Main.tileSolidTop[tile.type] || (includeSlopes && tile.blockType() != 0))) {
 					vector2.X = i * 16;
 					vector2.Y = j * 16;
 					int num5 = 16;
@@ -1030,7 +_,7 @@
 		if (tile == null || !tile.active() || tile.inActive() || !Main.tileSolid[tile.type])
 			return false;
 
-		if (treatPlatformsAsNonSolid && tile.type > 0 && tile.type <= TileID.Count && (TileID.Sets.Platforms[tile.type] || tile.type == 380))
+		if (treatPlatformsAsNonSolid && tile.type > 0 && (TileID.Sets.Platforms[tile.type] || tile.type == 380))
 			return false;
 
 		int num = tile.blockType();
@@ -1135,6 +_,12 @@
 
 	public static bool WetCollision(Vector2 Position, int Width, int Height)
 	{
+		return WetCollision(Position, Width, Height, out bool[] _);
+	}
+	public static bool WetCollision(Vector2 Position, int Width, int Height, out bool[] liquidsIn)
+	{
+		liquidsIn = new bool[LiquidLoader.LiquidCount];
+		bool flag = false;
 		honey = false;
 		shimmer = false;
 		Vector2 vector = new Vector2(Position.X + (float)(Width / 2), Position.Y + (float)(Height / 2));
@@ -1176,7 +_,8 @@
 						if (Main.tile[i, j].shimmer())
 							shimmer = true;
 
-						return true;
+						liquidsIn[Main.tile[i, j].liquidType()] = true;
+						flag = true;
 					}
 				}
 				else {
@@ -1192,17 +_,25 @@
 						else if (Main.tile[i, j - 1].shimmer())
 							shimmer = true;
 
-						return true;
+						liquidsIn[Main.tile[i, j - 1].liquidType()] = true;
+						flag = true;
 					}
 				}
 			}
 		}
-
-		return false;
+		return flag;
 	}
 
 	public static bool LavaCollision(Vector2 Position, int Width, int Height)
 	{
+		LavaCollision(Position, Width, Height, out bool[] liquidsIn);
+		return liquidsIn[LiquidID.Lava];
+	}
+
+	public static bool LavaCollision(Vector2 Position, int Width, int Height, out bool[] liquidsIn)
+	{
+		liquidsIn = new bool[LiquidLoader.LiquidCount];
+		bool flag = false;
 		int value = (int)(Position.X / 16f) - 1;
 		int value2 = (int)((Position.X + (float)Width) / 16f) + 2;
 		int value3 = (int)(Position.Y / 16f) - 1;
@@ -1214,7 +_,7 @@
 		Vector2 vector = default(Vector2);
 		for (int i = num; i < value2; i++) {
 			for (int j = value3; j < value4; j++) {
-				if (Main.tile[i, j] != null && Main.tile[i, j].liquid > 0 && Main.tile[i, j].lava()) {
+				if (Main.tile[i, j] != null && Main.tile[i, j].liquid > 0) {
 					vector.X = i * 16;
 					vector.Y = j * 16;
 					int num2 = 16;
@@ -1222,13 +_,28 @@
 					num3 /= 32f;
 					vector.Y += num3 * 2f;
 					num2 -= (int)(num3 * 2f);
-					if (Position.X + (float)Width > vector.X && Position.X < vector.X + 16f && Position.Y + (float)Height > vector.Y && Position.Y < vector.Y + (float)num2)
-						return true;
+					if (Position.X + (float)Width > vector.X && Position.X < vector.X + 16f && Position.Y + (float)Height > vector.Y && Position.Y < vector.Y + (float)num2) {
+						liquidsIn[Main.tile[i, j].liquidType()] = true;
+						flag = true;
+					}
 				}
 			}
 		}
+		return flag;
+	}
 
-		return false;
+	public static bool GetModdedWets(Vector2 Position, int Width, int Height, out bool[] liquidsIn)
+	{
+		liquidsIn = new bool[LiquidLoader.LiquidCount];
+		WetCollision(Position, Width, Height, out bool[] liquidsInNorm);
+		LavaCollision(Position, Width, Height, out bool[] liquidsInAlt);
+		Array.Copy(liquidsInNorm, liquidsIn, LiquidLoader.LiquidCount);
+		for (int i = LiquidID.Count; i < LiquidLoader.LiquidCount; i++) {
+			if (LiquidLoader.GetLiquid(i).UsesLavaCollisionForWet) {
+				liquidsIn[i] = liquidsInAlt[i];
+			}
+		}
+		return liquidsIn.Contains(true);
 	}
 
 	public static Vector4 WalkDownSlope(Vector2 Position, Vector2 Velocity, int Width, int Height, float gravity = 0f)
@@ -1906,7 +_,8 @@
 		return false;
 	}
 
+	// Added walkableLiquids parameter, removed lavaWalk parameters.
-	public static Vector2 WaterCollision(Vector2 Position, Vector2 Velocity, int Width, int Height, bool fallThrough = false, bool fall2 = false, bool lavaWalk = true)
+	public static Vector2 WaterCollision(Vector2 Position, Vector2 Velocity, int Width, int Height, bool[] walkableLiquids, bool fallThrough = false, bool fall2 = false)
 	{
 		Vector2 result = Velocity;
 		Vector2 vector = Position + Velocity;
@@ -1922,7 +_,7 @@
 		Vector2 vector3 = default(Vector2);
 		for (int i = num; i < value2; i++) {
 			for (int j = value3; j < value4; j++) {
-				if (Main.tile[i, j] != null && Main.tile[i, j].liquid > 0 && Main.tile[i, j - 1].liquid == 0 && (!Main.tile[i, j].lava() || lavaWalk)) {
+				if (Main.tile[i, j] != null && Main.tile[i, j].liquid > 0 && Main.tile[i, j - 1].liquid == 0 && walkableLiquids[Main.tile[i, j].liquidType()]) {
 					int num2 = (int)Main.tile[i, j].liquid / 32 * 2 + 2;
 					vector3.X = i * 16;
 					vector3.Y = j * 16 + 16 - num2;
@@ -2228,7 +_,25 @@
 		return false;
 	}
 
+	// legacy overload for abi compat, remove in 1.4.5
+	/// <inheritdoc cref="SwitchTiles(Entity, Vector2, int, int, Vector2, int)"/>
+	[OriginalOverload]
+	[Obsolete("Use SwitchTiles(Entity, Vector2, int, int, Vector2, int) instead")]
 	public static bool SwitchTiles(Vector2 Position, int Width, int Height, Vector2 oldPosition, int objType)
+	=> SwitchTiles(null, Position, Width, Height, oldPosition, objType);
+
+	/// <summary>
+	/// Calls collision related tile actions. Acts on <see cref="TileID.Sets.PressurePlate"/>, ProjectilePressurePad, LandMine, GeyserTrap, as well as modded content using <see cref="ModTile.SwitchTiles"/>
+	/// <br/><br/> Successful collisions usually call <see cref="Wiring.HitSwitch(int, int)"/> and sync it.
+	/// <include file = 'CommonDocs.xml' path='Common/SwitchTilesObjType' />
+	/// </summary>
+	/// <param name="entity"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='entity']"/></param>
+	/// <param name="Position"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='position']"/></param>
+	/// <param name="Width"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='width']"/></param>
+	/// <param name="Height"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='height']"/></param>
+	/// <param name="oldPosition"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='oldPosition']"/></param>
+	/// <param name="objType"><inheritdoc cref="ModTile.SwitchTiles" path="/param[@name='objType']"/></param>
+	public static bool SwitchTiles(Entity entity, Vector2 Position, int Width, int Height, Vector2 oldPosition, int objType)
 	{
 		int num = (int)(Position.X / 16f) - 1;
 		int num2 = (int)((Position.X + (float)Width) / 16f) + 2;
@@ -2253,13 +_,18 @@
 					continue;
 
 				int type = Main.tile[i, j].type;
+				if (TileLoader.SwitchTiles(i, j, Main.tile[i, j].TileType, entity, Position, Width, Height, oldPosition, objType))
+					return true;
+				/*
 				if (!Main.tile[i, j].active() || (type != 135 && type != 210 && type != 443 && type != 442))
+				*/
+				if (!Main.tile[i, j].active() || ((TileID.Sets.PressurePlate[type] == -1 || TileID.Sets.PressurePlate[type] == -3) && type != 210 && type != 443))
 					continue;
 
 				vector.X = i * 16;
 				vector.Y = j * 16 + 12;
 				bool flag = false;
-				if (type == 442) {
+				if (TileID.Sets.PressurePlate[type] == -4) {
 					if (objType == 4) {
 						float r1StartX = 0f;
 						float r1StartY = 0f;
@@ -2322,6 +_,7 @@
 					}
 
 					int num5 = Main.tile[i, j].frameY / 18;
+					if (TileID.Sets.PressurePlate[type] >= 0) num5 = TileID.Sets.PressurePlate[type];
 					bool flag2 = true;
 					if ((num5 == 4 || num5 == 2 || num5 == 3 || num5 == 6 || num5 == 7) && objType != 1)
 						flag2 = false;
@@ -2348,6 +_,7 @@
 		return false;
 	}
 
+	/// <summary><strong>Unused.</strong></summary>
 	public bool SwitchTilesNew(Vector2 Position, int Width, int Height, Vector2 oldPosition, int objType)
 	{
 		Point point = Position.ToTileCoordinates();
@@ -2614,7 +_,10 @@
 			tile = Main.tile[num2, num3];
 			tile2 = Main.tile[num2, num3 - 1];
 			if (specialChecksMode == 1)
+				/*
 				flag5 = tile.type != 16 && tile.type != 18 && tile.type != 14 && tile.type != 469 && tile.type != 134;
+				*/
+				flag5 = !TileID.Sets.IgnoredByNpcStepUp[tile.type];
 
 			flag4 = flag4 && ((tile.nactive() && (!tile.topSlope() || (tile.slope() == 1 && position.X + (float)(width / 2) < (float)(num2 * 16)) || (tile.slope() == 2 && position.X + (float)(width / 2) > (float)(num2 * 16 + 16))) && (!tile.topSlope() || position.Y + (float)height > (float)(num3 * 16)) && ((Main.tileSolid[tile.type] && !Main.tileSolidTop[tile.type]) || (holdsMatching && ((Main.tileSolidTop[tile.type] && tile.frameY == 0) || TileID.Sets.Platforms[tile.type]) && (!Main.tileSolid[tile2.type] || !tile2.nactive()) && flag5))) || (tile2.halfBrick() && tile2.nactive()));
 			flag4 &= !Main.tileSolidTop[tile.type] || !Main.tileSolidTop[tile2.type];
@@ -2694,7 +_,7 @@
 			if (tile.active()) {
 				if (Main.tileSolid[tile.type]) {
 					int num5 = tile.blockType();
-					if (tile.type != 19) {
+					if (!TileID.Sets.Platforms[tile.type]) {
 						switch (num5) {
 							case 1:
 								return 0f;
