--- src/Terraria\Terraria\Collision.cs
+++ src/tModLoader\Terraria\Collision.cs
@@ -2471,7 +_,7 @@
 			{
 				for (int j = num3; j < num4; j++)
 				{
-					if (Main.tile[i, j] != null && Main.tile[i, j].slope() == 0 && !Main.tile[i, j].inActive() && Main.tile[i, j].active() && (Main.tile[i, j].type == 32 || Main.tile[i, j].type == 37 || Main.tile[i, j].type == 48 || Main.tile[i, j].type == 232 || Main.tile[i, j].type == 53 || Main.tile[i, j].type == 57 || Main.tile[i, j].type == 58 || Main.tile[i, j].type == 69 || Main.tile[i, j].type == 76 || Main.tile[i, j].type == 112 || Main.tile[i, j].type == 116 || Main.tile[i, j].type == 123 || Main.tile[i, j].type == 224 || Main.tile[i, j].type == 234 || Main.tile[i, j].type == 352))
+					if (Main.tile[i, j] != null && Main.tile[i, j].slope() == 0 && !Main.tile[i, j].inActive() && Main.tile[i, j].active() && ((TileID.Sets.TouchDamageVines[Main.tile[i, j].type] > 0 && Main.tile[i, j].type != 80) || TileID.Sets.TouchDamageHot[Main.tile[i, j].type] == 20 || TileID.Sets.TouchDamageOther[Main.tile[i, j].type] > 0 || (TileID.Sets.TouchDamageSands[Main.tile[i, j].type] == 15 && Main.tile[i, j].type != 69)))
 					{
 						Vector2 vector2;
 						vector2.X = (float)(i * 16);
@@ -2484,7 +_,7 @@
 							vector2.Y += 8f;
 							num6 -= 8;
 						}
-						if (type == 32 || type == 69 || type == 80 || type == 352 || (type == 80 && Main.expertMode))
+						if (TileID.Sets.TouchDamageVines[type] > 0 || (type == 80 && Main.expertMode))
 						{
 							if (vector.X + (float)Width > vector2.X && vector.X < vector2.X + 16f && vector.Y + (float)Height > vector2.Y && vector.Y < vector2.Y + (float)num6 + 0.011f)
 							{
@@ -2493,15 +_,7 @@
 								{
 									num7 = -1;
 								}
-								num5 = 10;
-								if (type == 69)
-								{
-									num5 = 17;
-								}
-								else if (type == 80)
-								{
-									num5 = 6;
-								}
+								num5 = TileID.Sets.TouchDamageVines[type];
 								if (type == 32 || type == 69 || type == 352)
 								{
 									WorldGen.KillTile(i, j, false, false, false);
@@ -2513,7 +_,7 @@
 								return new Vector2((float)num7, (float)num5);
 							}
 						}
-						else if (type == 53 || type == 112 || type == 116 || type == 123 || type == 224 || type == 234)
+						else if (TileID.Sets.TouchDamageSands[type] == 15 && type != 57 && type != 69)
 						{
 							if (vector.X + (float)Width - 2f >= vector2.X && vector.X + 2f <= vector2.X + 16f && vector.Y + (float)Height - 2f >= vector2.Y && vector.Y + 2f <= vector2.Y + (float)num6)
 							{
@@ -2533,21 +_,18 @@
 							{
 								num9 = -1;
 							}
-							if (!fireImmune && (type == 37 || type == 58 || type == 76))
+							if (!fireImmune && TileID.Sets.TouchDamageHot[type] == 20)
 							{
 								num5 = 20;
 							}
-							if (type == 48)
-							{
-								num5 = 40;
-							}
-							if (type == 232)
-							{
-								num5 = 60;
+							if (TileID.Sets.TouchDamageOther[type] > 0)
+							{
+								num5 = TileID.Sets.TouchDamageOther[type];
 							}
 							return new Vector2((float)num9, (float)num5);
 						}
 					}
+					//TODO - New Hook - ModTile.HurtPlayer
 				}
 			}
 			return default(Vector2);
@@ -3035,7 +_,7 @@
 					if (Main.tileSolid[(int)tile.type])
 					{
 						int num4 = tile.blockType();
-						if (tile.type == 19)
+						if (TileID.Sets.Platforms[tile.type])
 						{
 							int num5 = (int)(tile.frameX / 18);
 							if (((num5 >= 0 && num5 <= 7) || (num5 >= 12 && num5 <= 16)) && (num == 0f || flag))

