--- src/Terraria\Terraria\Dust.cs
+++ src/tModLoader\Terraria\Dust.cs
@@ -4,6 +_,8 @@
 using Terraria.GameContent.Events;
 using Terraria.Graphics.Shaders;
 using Terraria.Utilities;
+using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -28,6 +_,7 @@
 		public ArmorShaderData shader;
 		public object customData;
 		public bool firstFrame;
+		internal int realType = -1;
 
 		public static Dust NewDustPerfect(Vector2 Position, int Type, Vector2? Velocity = null, int Alpha = 0, Color newColor = default(Color), float Scale = 1f)
 		{
@@ -193,6 +_,7 @@
 					{
 						dust.alpha = 50;
 					}
+					ModDust.SetupDust(dust);
 					if (dust.type != 34 && dust.type != 35 && dust.type != 152)
 					{
 						break;
@@ -303,6 +_,10 @@
 
 		public static int dustWater()
 		{
+			if (Main.waterStyle >= WaterStyleLoader.vanillaWaterCount)
+			{
+				return WaterStyleLoader.GetWaterStyle(Main.waterStyle).GetSplashDust();
+			}
 			switch (Main.waterStyle)
 			{
 				case 2:
@@ -343,6 +_,13 @@
 					if (dust.active)
 					{
 						Dust.dCount += 1f;
+						ModDust.SetupUpdateType(dust);
+						ModDust modDust = ModDust.GetDust(dust.type);
+						if (modDust != null && !modDust.Update(dust))
+						{
+							ModDust.TakeDownUpdateType(dust);
+							continue;
+						}
 						if (dust.scale > 10f)
 						{
 							dust.active = false;
@@ -1893,7 +_,7 @@
 							}
 							Lighting.AddLight((int)(dust.position.X / 16f), (int)(dust.position.Y / 16f), num106 * 0.7f, num106, num106 * 0.8f);
 						}
-						else
+						else if (modDust == null || !modDust.MidUpdate(dust))
 						{
 							Dust expr_446A_cp_0 = dust;
 							expr_446A_cp_0.velocity.X = expr_446A_cp_0.velocity.X * 0.99f;
@@ -2150,6 +_,7 @@
 						{
 							dust.active = false;
 						}
+						ModDust.TakeDownUpdateType(dust);
 					}
 				}
 				else
@@ -2188,6 +_,15 @@
 
 		public Color GetAlpha(Color newColor)
 		{
+			ModDust modDust = ModDust.GetDust(this.type);
+			if (modDust != null)
+			{
+				Color? modColor = modDust.GetAlpha(this, newColor);
+				if (modColor.HasValue)
+				{
+					return modColor.Value;
+				}
+			}
 			float num = (float)(255 - this.alpha) / 255f;
 			if (this.type == 259)
 			{

