--- src/Terraria\Terraria\Dust.cs
+++ src/tModLoader\Terraria\Dust.cs
@@ -2,6 +_,8 @@
 using System;
 using Terraria.GameContent;
 using Terraria.Graphics.Shaders;
+using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -24,6 +_,7 @@
 		public ArmorShaderData shader;
 		public object customData;
 		public bool firstFrame;
+		internal int realType = -1;
 
 		public static int NewDust(Vector2 Position, int Width, int Height, int Type, float SpeedX = 0f, float SpeedY = 0f, int Alpha = 0, Color newColor = default(Color), float Scale = 1f)
 		{
@@ -168,6 +_,7 @@
 					{
 						dust.alpha = 50;
 					}
+					ModDust.SetupDust(dust);
 					if (dust.type != 34 && dust.type != 35 && dust.type != 152)
 					{
 						break;
@@ -290,6 +_,13 @@
 					if (dust.active)
 					{
 						Dust.dCount += 1f;
+						ModDust.SetupUpdateType(dust);
+						ModDust modDust = ModDust.GetDust(dust.type);
+						if (modDust != null && !modDust.Update(dust))
+						{
+							ModDust.TakeDownUpdateType(dust);
+							continue;
+						}
 						if (dust.scale > 10f)
 						{
 							dust.active = false;
@@ -1729,7 +_,7 @@
 							}
 							Lighting.AddLight((int)(dust.position.X / 16f), (int)(dust.position.Y / 16f), num105 * 0.7f, num105, num105 * 0.8f);
 						}
-						else
+						else if (modDust == null || !modDust.MidUpdate(dust))
 						{
 							Dust expr_3E5B_cp_0 = dust;
 							expr_3E5B_cp_0.velocity.X = expr_3E5B_cp_0.velocity.X * 0.99f;
@@ -1954,6 +_,7 @@
 						{
 							dust.active = false;
 						}
+						ModDust.TakeDownUpdateType(dust);
 					}
 				}
 				else
@@ -1992,6 +_,15 @@
 
 		public Color GetAlpha(Color newColor)
 		{
+			ModDust modDust = ModDust.GetDust(this.type);
+			if (modDust != null)
+			{
+				Color? modColor = modDust.GetAlpha(this, newColor);
+				if (modColor.HasValue)
+				{
+					return modColor.Value;
+				}
+			}
 			float num = (float)(255 - this.alpha) / 255f;
 			if (this.type == 259)
 			{

