--- src/TerrariaNetCore/Terraria/Framing.cs
+++ src/tModLoader/Terraria/Framing.cs
@@ -1,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria;
 
@@ -321,6 +_,6 @@
 		if (i <= 0 || j <= 0 || i >= Main.maxTilesX - 1 || j >= Main.maxTilesY - 1 || Main.tile[i, j] == null)
 			return;
 
-		if (Main.tile[i, j].wall >= WallID.Count)
+		if (Main.tile[i, j].wall >= WallLoader.WallCount)
 			Main.tile[i, j].wall = 0;
 
@@ -379,6 +_,12 @@
 
 		if (num == 15)
 			num += centerWallFrameLookup[i % 3][j % 3];
+
+		if (!WallLoader.WallFrame(i, j, tile.wall, resetFrame, ref num, ref num2))
+			return;
+
+		// Moved from if-block contents above.
+		tile.wallFrameNumber((byte)num2);
 
 		Point16 point = wallFrameLookup[num][num2];
 		tile.wallFrameX(point.X);
