--- src/Terraria/Terraria/Gore.cs
+++ src/tModLoader/Terraria/Gore.cs
@@ -7,10 +_,11 @@
 using Terraria.Graphics.Effects;
 using Terraria.ID;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
-	public class Gore
+	public partial class Gore
 	{
 		public static int goreTime = 600;
 		public Vector2 position;
@@ -239,6 +_,10 @@
 			if (Main.netMode == 2 || !active)
 				return;
 
+			if (ModGore != null && !ModGore.Update(this))
+				return;
+
+			GoreLoader.SetupUpdateType(this);
 			switch (GoreID.Sets.SpecialAI[type]) {
 				case 4:
 					UpdateAmbientFloorCloud();
@@ -830,6 +_,7 @@
 				else
 					Lighting.AddLight((int)((position.X + 32f * scale / 2f) / 16f), (int)((position.Y + 32f * scale / 2f) / 16f), num37, num38, num39);
 			}
+			GoreLoader.TakeDownUpdateType(this);
 		}
 
 		private void Gore_UpdateLeaf() {
@@ -1035,6 +_,13 @@
 			Main.gore[num].alpha = 0;
 			Main.gore[num].rotation = 0f;
 			Main.gore[num].scale = Scale;
+			
+			var gore = Main.gore[num];
+			gore.ResetNewFields();
+			
+			gore.ModGore = GoreLoader.GetModGore(gore.type);
+			gore.ModGore?.OnSpawn(gore);
+			
 			if (!ChildSafety.Disabled && ChildSafety.DangerousGore(Type)) {
 				Main.gore[num].type = Main.rand.Next(11, 14);
 				Main.gore[num].scale = Main.rand.NextFloat() * 0.5f + 0.5f;
@@ -1149,6 +_,12 @@
 		}
 
 		public Color GetAlpha(Color newColor) {
+			if (ModGore != null) {
+				Color? modColor = ModGore.GetAlpha(this, newColor);
+				if (modColor.HasValue)
+					return modColor.Value;
+			}
+
 			float num = (float)(255 - alpha) / 255f;
 			int r;
 			int g;
