--- src/TerrariaNetCore/Terraria/HitTile.cs
+++ src/tModLoader/Terraria/HitTile.cs
@@ -2,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.GameContent;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 
 namespace Terraria;
@@ -307,7 +_,7 @@
 
 			bool flag3 = false;
 			bool flag4 = false;
-			if (tile.type == 10) {
+			if (TileLoader.IsClosedDoor(tile)) {
 				flag3 = false;
 			}
 			else if (Main.tileSolid[tile.type] && !Main.tileSolidTop[tile.type]) {
@@ -346,6 +_,14 @@
 
 			int animationTimeElapsed = data[j].animationTimeElapsed;
 			if (!((float)animationTimeElapsed >= 10f)) {
+				/* Check if the current tile is a modded tile. */
+				if (TileLoader.GetTile(tile.type) is { } modTile) {
+					/* If it is a modtile, custom drawing is handled and we should skip the rest of this loop for this tile. */
+					if (modTile.DrawCracks(spriteBatch, x, y, damage, data[j].crackStyle, true, animationTimeElapsed)) {
+						continue;
+					}
+				}
+
 				float num5 = (float)animationTimeElapsed / 10f;
 				Color color = Lighting.GetColor(x, y);
 				float rotation = 0f;
