--- src/Terraria/Terraria/HitTile.cs
+++ src/tModLoader/Terraria/HitTile.cs
@@ -2,11 +_,12 @@
 using Microsoft.Xna.Framework.Graphics;
 using System;
 using Terraria.GameContent;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 
 namespace Terraria
 {
-	public class HitTile
+	public partial class HitTile
 	{
 		public class HitTileObject
 		{
@@ -134,6 +_,16 @@
 		}
 
 		public int AddDamage(int tileId, int damageAmount, bool updateAmount = true) {
+			int damage = AddDamage_Inner(tileId, damageAmount, updateAmount);
+			if (!updateAmount || !decayPaused || damageAmount == 0)
+				return damage;
+			if (neededPrune < 2 && damage >= 100)
+				neededPrune = 2;
+			else if (neededPrune < 1)
+				neededPrune = 1;
+			return damage;
+		}
+		private int AddDamage_Inner(int tileId, int damageAmount, bool updateAmount) {
 			if (tileId < 0 || tileId > 500)
 				return 0;
 
@@ -196,6 +_,8 @@
 		}
 
 		public void Prune() {
+			if (decayPaused)
+				return;
 			bool flag = false;
 			for (int i = 0; i <= 500; i++) {
 				HitTileObject hitTileObject = data[i];
@@ -291,7 +_,7 @@
 
 				bool flag2 = false;
 				bool flag3 = false;
-				if (Main.tile[x, y].type == 10) {
+				if (TileLoader.IsClosedDoor(Main.tile[x, y])) {
 					flag2 = false;
 				}
 				else if (Main.tileSolid[Main.tile[x, y].type] && !Main.tileSolidTop[Main.tile[x, y].type]) {
