--- src/Terraria/Terraria/IngameOptions.cs
+++ src/tModLoader/Terraria/IngameOptions.cs
@@ -4,13 +_,17 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.Audio;
+using Steamworks;
 using Terraria.GameContent;
 using Terraria.GameContent.UI;
 using Terraria.GameInput;
 using Terraria.Localization;
+using Terraria.ModLoader.Engine;
+using Terraria.ModLoader.UI;
 using Terraria.Social;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -18,7 +_,8 @@
 	{
 		public const int width = 670;
 		public const int height = 480;
-		public static float[] leftScale = new float[10] {
+		public static float[] leftScale = new float[] {
+			0.7f,
 			0.7f,
 			0.7f,
 			0.7f,
@@ -174,7 +_,7 @@
 			noSound = false;
 			bool flag5 = SocialAPI.Network != null && SocialAPI.Network.CanInvite();
 			int num7 = flag5 ? 1 : 0;
-			int num8 = 5 + num7 + 2;
+			int num8 = 5 + num7 + 2 + 1; // Add 1 for ModConfig menu
 			Vector2 vector = new Vector2(value3.X + value2.X / 4f, value3.Y + (float)(num3 * 5 / 2));
 			Vector2 vector2 = new Vector2(0f, value2.Y - (float)(num3 * 5)) / (num8 + 1);
 			if (flag)
@@ -246,6 +_,20 @@
 			}
 
 			num10++;
+			if (DrawLeftSide(sb, "Mod Configuration", num10, vector, vector2, leftScale, 0.7f, 0.8f, 0.01f)) {
+				leftHover = num10;
+				if (flag4) {
+					Close();
+					IngameFancyUI.CoverNextFrame();
+					Main.playerInventory = false;
+					Main.editChest = false;
+					Main.npcChatText = "";
+					Main.inFancyUI = true;
+					Main.InGameUI.SetState(Interface.modConfigList);
+				}
+			}
+
+			num10++;
 			if (flag5 && DrawLeftSide(sb, Lang.menu[147].Value, num10, vector, vector2, leftScale)) {
 				leftHover = num10;
 				if (flag4) {
@@ -257,6 +_,7 @@
 			if (flag5)
 				num10++;
 
+			if(!Steam.IsSteamApp) {
 			if (DrawLeftSide(sb, Lang.menu[131].Value, num10, vector, vector2, leftScale)) {
 				leftHover = num10;
 				if (flag4) {
@@ -264,8 +_,9 @@
 					IngameFancyUI.OpenAchievements();
 				}
 			}
-
 			num10++;
+			}
+
 			if (DrawLeftSide(sb, Lang.menu[118].Value, num10, vector, vector2, leftScale)) {
 				leftHover = num10;
 				if (flag4)
@@ -276,6 +_,8 @@
 			if (DrawLeftSide(sb, Lang.inter[35].Value, num10, vector, vector2, leftScale)) {
 				leftHover = num10;
 				if (flag4) {
+					SystemLoader.PreSaveAndQuit();
+					// Calls before anything else is
 					Close();
 					Main.menuMode = 10;
 					Main.gameMenu = true;
@@ -301,6 +_,7 @@
 					break;
 				case 1:
 					num12 = 10;
+					num12++; // Room for tModLoader BossBarStyle option.
 					num4 = 1f;
 					num5 = 1.001f;
 					num6 = 0.001f;
@@ -730,6 +_,14 @@
 					rightHover = num16;
 					if (flag4)
 						Main.ActivePlayerResourcesSet = playerResourcesDisplaySet;
+				}
+
+				num16++;
+				string text = BossBarLoader.InsertMenu(out Action onClick);
+				if (DrawRightSide(sb, text, num16, vector3, vector4, rightScale[num16], (rightScale[num16] - num4) / (num5 - num4))) {
+					rightHover = num16;
+					if (flag4)
+						onClick();
 				}
 
 				num16++;
