--- src/Terraria\Terraria\Lang.cs
+++ src/tModLoader\Terraria\Lang.cs
@@ -6,6 +_,7 @@
 using Terraria.GameInput;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 #if CLIENT
 using Terraria.Map;
@@ -35,7 +_,7 @@
 		[Obsolete("Lang arrays have been replaced with the new Language.GetText system.")]
 		public static LocalizedText[] chestType2 = new LocalizedText[2];
 		public static LocalizedText[] prefix = new LocalizedText[84];
-		public static LocalizedText[] _mapLegendCache;
+		public static MapLegend _mapLegendCache;
 		private static LocalizedText[] _itemNameCache = new LocalizedText[3930];
 		private static LocalizedText[] _projectileNameCache = new LocalizedText[714];
 		private static LocalizedText[] _npcNameCache = new LocalizedText[580];
@@ -304,6 +_,11 @@
 			}
 			int num = Main.anglerQuestItemNetIDs[Main.anglerQuest];
 			Main.npcChatCornerItem = num;
+			string modText = ItemLoader.AnglerChat(num);
+			if (modText != null)
+			{
+				return modText;
+			}
 			return Language.GetTextValueWith("AnglerQuestText.Quest_" + ItemID.Search.GetName(num), obj);
 		}
 

