--- src/Terraria\Terraria\Lang.cs
+++ src/tModLoader\Terraria\Lang.cs
@@ -1,5 +_,6 @@
 using System;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 #if CLIENT
 using Terraria.Map;
@@ -18,7 +_,7 @@
 		public static string[] dt = new string[4];
 		public static string[] chestType = new string[52];
 		public static string[] dresserType = new string[28];
-		public static string[] mapLegend;
+		public static MapLegend mapLegend;
 		public static string[] prefix = new string[84];
 		public static string the;
 
@@ -3628,6 +_,10 @@
 
 		public static string npcName(int l, bool english = false)
 		{
+			if (l >= Main.maxNPCTypes)
+			{
+				return NPCLoader.DisplayName(l);
+			}
 			if (Lang.lang <= 1 || english)
 			{
 				switch (l)
@@ -32723,10 +_,21 @@
 							array = array2;
 						}
 						else
+						//patch file: array, array2, flag, text
 						{
 							array[num2++] = "Piranhas and sharks are ugly! Soooo ugly! Did you know there's a fish that looks very pretty and still can eat your face off? I would pay 2 platinum to see that happen, by the way... To the point, though, you catchy for me. Just make sure I have it before you lose your face!";
 							flag = true;
 							text = "\n(Caught Jungle Surface)";
+						}
+						break;
+					default:
+						if (gotFish)
+						{
+							array = array2;
+						}
+						else
+						{
+							flag = true;
 						}
 						break;
 				}
@@ -32739,6 +_,7 @@
 					text2 = text2 + " " + array[i];
 				}
 			}
+			ItemLoader.AnglerChat(gotFish, Main.anglerQuestFinished, Main.anglerQuestItemNetIDs[Main.anglerQuest], ref text2, ref text);
 			if (text.Length > 0)
 			{
 				text2 = text2 + "\n" + text;
@@ -33792,7 +_,7 @@
 #if CLIENT
 				if (Lang.mapLegend == null)
 				{
-					Lang.mapLegend = new string[MapHelper.LookupCount()];
+					Lang.mapLegend = new MapLegend(MapHelper.LookupCount());
 				}
 				Lang.mapLegend[MapHelper.TileToLookup(4, 0)] = "Torch";
 				Lang.mapLegend[MapHelper.TileToLookup(4, 1)] = "Torch";

