--- src/Terraria\Terraria\Lighting.cs
+++ src/tModLoader\Terraria\Lighting.cs
@@ -10,6 +_,7 @@
 using Terraria.DataStructures;
 using Terraria.GameContent;
 using Terraria.Graphics;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -610,6 +_,7 @@
 				Main.player[Main.myPlayer].NPCBannerBuff[m] = false;
 			}
 			Main.player[Main.myPlayer].hasBanner = false;
+			WorldHooks.ResetNearbyTileEffects();
 			int[] tileCounts = WorldGen.tileCounts;
 			num4 = Lighting.firstToLightX;
 			num5 = Lighting.lastToLightX;
@@ -622,6 +_,7 @@
 			for (int n = num4; n < num5; n++)
 			{
 				Lighting.LightingState[] array3 = Lighting.states[n - Lighting.firstToLightX];
+				//patch file: n, num17
 				for (int num17 = num6; num17 < num7; num17++)
 				{
 					Lighting.LightingState lightingState3 = array3[num17 - Lighting.firstToLightY];
@@ -773,7 +_,7 @@
 							lightingState3.b2 = num20;
 						}
 					}
-					byte wall = tile.wall;
+					ushort wall = tile.wall;
 					if (wall <= 137)
 					{
 						if (wall != 33)
@@ -880,8 +_,10 @@
 								break;
 						}
 					}
+					WallLoader.ModifyLight(n, num17, wall, ref num18, ref num19, ref num20);
 					if (tile.active())
 					{
+						bool closer = false;
 						if (n > num4 + num15 && n < num5 - num15 && num17 > num6 + num16 && num17 < num7 - num16)
 						{
 							tileCounts[(int)tile.type]++;
@@ -912,6 +_,7 @@
 								Main.player[Main.myPlayer].NPCBannerBuff[num29] = true;
 								Main.player[Main.myPlayer].hasBanner = true;
 							}
+							closer = true;
 						}
 						ushort type = tile.type;
 						if (type != 139)
@@ -965,6 +_,11 @@
 						{
 							Main.musicBox = (int)(tile.frameY / 36);
 						}
+						if (TileLoader.IsModMusicBox(tile) && tile.frameX >= 36)
+						{
+							Main.musicBox = SoundLoader.tileToMusic[tile.type][tile.frameY / 36 * 36];
+						}
+						TileLoader.NearbyEffects(n, num17, type, closer);
 						if (Main.tileBlockLight[(int)tile.type] && (Lighting.lightMode >= 2 || (tile.type != 131 && !tile.inActive() && tile.slope() == 0)))
 						{
 							lightingState3.stopLight = true;
@@ -2442,16 +_,19 @@
 						}
 					}
 					IL_31E1:
+					TileLoader.ModifyLight(n, num17, tile.type, ref num18, ref num19, ref num20);
 					if (Lighting.RGB)
 					{
 						if (lightingState3.r2 < num18)
 						{
 							lightingState3.r2 = num18;
 						}
+						//patch file: num18, num19
 						if (lightingState3.g2 < num19)
 						{
 							lightingState3.g2 = num19;
 						}
+						//patch file: num19, num20
 						if (lightingState3.b2 < num20)
 						{
 							lightingState3.b2 = num20;
@@ -2585,7 +_,7 @@
 			if (Main.player[Main.myPlayer].accOreFinder)
 			{
 				Main.player[Main.myPlayer].bestOre = -1;
-				for (int num61 = 0; num61 < 446; num61++)
+				for (int num61 = 0; num61 < Main.tileValue.Length; num61++)
 				{
 					if (tileCounts[num61] > 0 && Main.tileValue[num61] > 0 && (Main.player[Main.myPlayer].bestOre < 0 || Main.tileValue[num61] > Main.tileValue[Main.player[Main.myPlayer].bestOre]))
 					{
@@ -2593,6 +_,7 @@
 					}
 				}
 			}
+			WorldHooks.TileCountsAvailable(tileCounts);
 			Array.Clear(tileCounts, 0, tileCounts.Length);
 			if (Main.holyTiles < 0)
 			{

