--- src/TerrariaNetCore/Terraria/Lighting.cs
+++ src/tModLoader/Terraria/Lighting.cs
@@ -14,8 +_,11 @@
 	[Old]
 	public static int OffScreenTiles = 45;
 	private static LightMode _mode = LightMode.Color;
+
+	//TML: Made both of these public.
-	private static readonly LightingEngine NewEngine = new LightingEngine();
+	public static readonly LightingEngine NewEngine = new LightingEngine();
-	private static readonly LegacyLighting LegacyEngine = new LegacyLighting(Main.Camera);
+	public static readonly LegacyLighting LegacyEngine = new LegacyLighting(Main.Camera);
+
 	private static ILightingEngine _activeEngine;
 
 	public static float GlobalBrightness { get; set; }
