--- src/Terraria\Terraria\Lighting.cs
+++ src/tModLoader\Terraria\Lighting.cs
@@ -9,6 +_,7 @@
 using System.Threading;
 using Terraria.DataStructures;
 using Terraria.Graphics;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -609,6 +_,7 @@
 				Main.player[Main.myPlayer].NPCBannerBuff[m] = false;
 			}
 			Main.player[Main.myPlayer].hasBanner = false;
+			WorldHooks.ResetNearbyTileEffects();
 			int[] tileCounts = WorldGen.tileCounts;
 			num4 = Lighting.firstToLightX;
 			num5 = Lighting.lastToLightX;
@@ -772,7 +_,7 @@
 							lightingState3.b2 = num20;
 						}
 					}
-					byte wall = tile.wall;
+					ushort wall = tile.wall;
 					if (wall <= 137)
 					{
 						if (wall != 33)
@@ -879,8 +_,10 @@
 								break;
 						}
 					}
+					WallLoader.ModifyLight(n, num17, wall, ref num18, ref num19, ref num20);
 					if (tile.active())
 					{
+						bool closer = false;
 						if (n > num4 + num15 && n < num5 - num15 && num17 > num6 + num16 && num17 < num7 - num16)
 						{
 							tileCounts[(int)tile.type]++;
@@ -911,6 +_,7 @@
 								Main.player[Main.myPlayer].NPCBannerBuff[num29] = true;
 								Main.player[Main.myPlayer].hasBanner = true;
 							}
+							closer = true;
 						}
 						ushort type = tile.type;
 						if (type != 139)
@@ -964,6 +_,11 @@
 						{
 							Main.musicBox = (int)(tile.frameY / 36);
 						}
+						if (TileLoader.IsModMusicBox(tile) && tile.frameX >= 36)
+						{
+							Main.musicBox = SoundLoader.tileToMusic[tile.type][tile.frameY / 36 * 36];
+						}
+						TileLoader.NearbyEffects(n, num17, type, closer);
 						if (Main.tileBlockLight[(int)tile.type] && (Lighting.lightMode >= 2 || (tile.type != 131 && !tile.inActive() && tile.slope() == 0)))
 						{
 							lightingState3.stopLight = true;
@@ -2379,6 +_,7 @@
 						}
 					}
 					IL_304B:
+					TileLoader.ModifyLight(n, num17, tile.type, ref num18, ref num19, ref num20);
 					if (Lighting.RGB)
 					{
 						if (lightingState3.r2 < num18)
@@ -2522,7 +_,7 @@
 			if (Main.player[Main.myPlayer].accOreFinder)
 			{
 				Main.player[Main.myPlayer].bestOre = -1;
-				for (int num60 = 0; num60 < 419; num60++)
+				for (int num60 = 0; num60 < Main.tileValue.Length; num60++)
 				{
 					if (tileCounts[num60] > 0 && Main.tileValue[num60] > 0 && (Main.player[Main.myPlayer].bestOre < 0 || Main.tileValue[num60] > Main.tileValue[Main.player[Main.myPlayer].bestOre]))
 					{
@@ -2530,6 +_,7 @@
 					}
 				}
 			}
+			WorldHooks.TileCountsAvailable(tileCounts);
 			Array.Clear(tileCounts, 0, tileCounts.Length);
 			if (Main.holyTiles < 0)
 			{

