--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -19,6 +_,9 @@
 using Terraria.Net;
 using Terraria.Testing;
 using Terraria.UI;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 
 namespace Terraria
 {
@@ -83,10 +_,8 @@
 			int num = 0;
 			num = start + 1;
 			b = (byte)(messageType = readBuffer[start]);
-			if (b >= 141)
-				return;
-
-			Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
+			if (b < LegacyNetDiagnosticsUI.maxMsg)
+				Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 			if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
 				Netplay.Connection.StatusCount++;
 
@@ -105,7 +_,7 @@
 			}
 
 			if (Main.netMode == 2) {
-				if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68)
+				if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b < 250)
 					NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 				if (Netplay.Clients[whoAmI].State == 0 && b != 1)
@@ -117,6 +_,9 @@
 
 			reader.BaseStream.Position = num;
 			NPCSpawnParams spawnparams;
+			if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+				return;
+
 			switch (b) {
 				case 15:
 				case 25:
@@ -135,9 +_,15 @@
 						if (Netplay.Clients[whoAmI].State != 0)
 							break;
 
-						if (reader.ReadString() == "Terraria" + 238) {
+						string clientVersion = reader.ReadString();
+						ModNet.isModdedClient[whoAmI] = clientVersion == BuildInfo.versionedName;
+						if (ModNet.isModdedClient[whoAmI] || clientVersion == "Terraria" + Main.curRelease && ModNet.AllowVanillaClients) {
 							if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 								Netplay.Clients[whoAmI].State = 1;
+								if (ModNet.isModdedClient[whoAmI]) {
+									ModNet.SyncMods(whoAmI);
+									return;
+								}
 								NetMessage.TrySendData(3, whoAmI);
 							}
 							else {
@@ -146,7 +_,10 @@
 							}
 						}
 						else {
-							NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+							if (clientVersion.StartsWith("tModLoader"))
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n(You are on {clientVersion}, server is on {BuildInfo.versionedName})"));
+							else
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + "\n(You cannot connect to a tModLoader Server with an unmodded client.)"));
 						}
 					}
 					break;
@@ -214,6 +_,7 @@
 							NetMessage.TrySendData(5, -1, -1, null, num44, 220 + num53, (int)player5.bank4.item[num53].prefix);
 						}
 
+						PlayerLoader.SyncPlayer(player5, -1, -1, true);
 						NetMessage.TrySendData(6);
 						if (Netplay.Connection.State == 2)
 							Netplay.Connection.State = 3;
@@ -236,7 +_,7 @@
 							player4.hair = 0;
 
 						player4.name = reader.ReadString().Trim().Trim();
-						player4.hairDye = reader.ReadByte();
+						player4.hairDye = ModNet.AllowVanillaClients ? reader.ReadByte() : reader.ReadVarInt();
 						BitsByte bitsByte4 = reader.ReadByte();
 						for (int num41 = 0; num41 < 8; num41++) {
 							player4.hideVisibleAccessory[num41] = bitsByte4[num41];
@@ -319,9 +_,9 @@
 						Player player12 = Main.player[num232];
 						lock (player12) {
 							int num233 = reader.ReadInt16();
-							int stack6 = reader.ReadInt16();
-							int num234 = reader.ReadByte();
-							int type14 = reader.ReadInt16();
+							int stack6 = ModNet.AllowVanillaClients ? reader.ReadInt16() : -1;
+							int num234 = ModNet.AllowVanillaClients ? reader.ReadByte() : -1;
+							int type14 = ModNet.AllowVanillaClients ? reader.ReadInt16() : -1;
 							Item[] array3 = null;
 							Item[] array4 = null;
 							int num235 = 0;
@@ -376,20 +_,37 @@
 							}
 
 							if (flag13) {
+								if (!ModNet.AllowVanillaClients) {
+									player12.trashItem = ItemIO.Receive(reader, readStack: true);
+									goto SkipVanillaRead;
+								}
+
 								player12.trashItem = new Item();
 								player12.trashItem.netDefaults(type14);
 								player12.trashItem.stack = stack6;
 								player12.trashItem.Prefix(num234);
+
+								SkipVanillaRead:
+
 								if (num232 == Main.myPlayer && !Main.ServerSideCharacter)
 									Main.clientPlayer.trashItem = player12.trashItem.Clone();
 							}
 							else if (num233 <= 58) {
 								int type15 = array3[num235].type;
 								int stack7 = array3[num235].stack;
+
+								if (!ModNet.AllowVanillaClients) {
+									array3[num235] = ItemIO.Receive(reader, readStack: true);
+									goto SkipVanillaRead;
+								}
+
 								array3[num235] = new Item();
 								array3[num235].netDefaults(type14);
 								array3[num235].stack = stack6;
 								array3[num235].Prefix(num234);
+
+								SkipVanillaRead:
+
 								if (num232 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num235] = array3[num235].Clone();
 
@@ -405,10 +_,18 @@
 								}
 							}
 							else {
+								if (!ModNet.AllowVanillaClients) {
+									array3[num235] = ItemIO.Receive(reader, readStack: true);
+									goto SkipVanillaRead;
+								}
+
 								array3[num235] = new Item();
 								array3[num235].netDefaults(type14);
 								array3[num235].stack = stack6;
 								array3[num235].Prefix(num234);
+
+								SkipVanillaRead:
+
 								if (num232 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num235] = array3[num235].Clone();
 							}
@@ -565,6 +_,8 @@
 						Main.invasionType = reader.ReadSByte();
 						Main.LobbyId = reader.ReadUInt64();
 						Sandstorm.IntendedSeverity = reader.ReadSingle();
+						if (!ModNet.AllowVanillaClients && Netplay.Connection.State > 4)
+							WorldIO.ReceiveModData(reader);
 						if (Netplay.Connection.State == 3) {
 							Main.windSpeedCurrent = Main.windSpeedTarget;
 							Netplay.Connection.State = 4;
@@ -696,6 +_,9 @@
 							NetMessage.TrySendData(83, whoAmI, -1, null, num176);
 						}
 
+						for (int type = NPCID.Count; type < NPCLoader.NPCCount; type++)
+							NetMessage.SendData(83, whoAmI, -1, null, type, 0f, 0f, 0f, 0, 0, 0);
+
 						NetMessage.TrySendData(49, whoAmI);
 						NetMessage.TrySendData(57, whoAmI);
 						NetMessage.TrySendData(7, whoAmI);
@@ -1137,8 +_,8 @@
 						int num177 = reader.ReadInt16();
 						Vector2 position3 = reader.ReadVector2();
 						Vector2 velocity5 = reader.ReadVector2();
-						int stack4 = reader.ReadInt16();
-						int pre2 = reader.ReadByte();
+						int stack4 = ModNet.AllowVanillaClients ? reader.ReadInt16() : reader.ReadVarInt();
+						int pre2 = ModNet.AllowVanillaClients ? reader.ReadByte() : reader.ReadVarInt();
 						int num178 = reader.ReadByte();
 						int num179 = reader.ReadInt16();
 						if (Main.netMode == 1) {
@@ -1151,7 +_,7 @@
 							Item item2 = Main.item[num180];
 							ItemSyncPersistentStats itemSyncPersistentStats = default(ItemSyncPersistentStats);
 							itemSyncPersistentStats.CopyFrom(item2);
-							bool newAndShiny = (item2.newAndShiny || item2.netID != num179) && ItemSlot.Options.HighlightNewItems && (num179 < 0 || num179 >= 5088 || !ItemID.Sets.NeverAppearsAsNewInInventory[num179]);
+							bool newAndShiny = (item2.newAndShiny || item2.netID != num179) && ItemSlot.Options.HighlightNewItems && (num179 < 0 || !ItemID.Sets.NeverAppearsAsNewInInventory[num179]);
 							item2.netDefaults(num179);
 							item2.newAndShiny = newAndShiny;
 							item2.Prefix(pre2);
@@ -1159,6 +_,7 @@
 							item2.position = position3;
 							item2.velocity = velocity5;
 							item2.active = true;
+							ItemIO.ReceiveModData(item2, reader);
 							if (b == 90) {
 								item2.instanced = true;
 								item2.playerIndexTheItemIsReservedFor = Main.myPlayer;
@@ -1198,6 +_,7 @@
 							obj5.position = position3;
 							obj5.velocity = velocity5;
 							obj5.active = true;
+							ItemIO.ReceiveModData(obj5, reader);
 							obj5.playerIndexTheItemIsReservedFor = Main.myPlayer;
 							if (flag8) {
 								NetMessage.TrySendData(21, -1, -1, null, num177);
@@ -1329,9 +_,10 @@
 						if (num115 == 245)
 							NPC.golemBoss = num112;
 
-						if (nPC4.type >= 0 && nPC4.type < 668 && Main.npcCatchable[nPC4.type])
+						if (nPC4.type >= 0 && nPC4.type < NPCLoader.NPCCount && Main.npcCatchable[nPC4.type])
 							nPC4.releaseOwner = reader.ReadByte();
 
+						NPCLoader.ReceiveExtraAI(nPC4, reader);
 						break;
 					}
 				case 24: {
@@ -1371,7 +_,7 @@
 						int num133 = bitsByte12[7] ? reader.ReadInt16() : (-1);
 						if (num133 >= 1000)
 							num133 = -1;
-
+						byte[] extraAI = ProjectileLoader.ReadExtraAI(reader, bitsByte12);
 						if (Main.netMode == 2) {
 							if (num131 == 949) {
 								num130 = 255;
@@ -1428,6 +_,7 @@
 							Main.projectileIdentity[num130, num133] = num134;
 						}
 
+						ProjectileLoader.ReceiveExtraAI(projectile, extraAI);
 						projectile.ProjectileFixDesperation();
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(27, -1, whoAmI, null, num134);
@@ -1436,7 +_,7 @@
 					}
 				case 28: {
 						int num4 = reader.ReadInt16();
-						int num5 = reader.ReadInt16();
+						int num5 = ModNet.AllowVanillaClients ? reader.ReadInt16() : reader.ReadInt32();
 						float num6 = reader.ReadSingle();
 						int num7 = reader.ReadByte() - 1;
 						byte b2 = reader.ReadByte();
@@ -1538,9 +_,10 @@
 				case 32: {
 						int num259 = reader.ReadInt16();
 						int num260 = reader.ReadByte();
-						int stack8 = reader.ReadInt16();
-						int pre3 = reader.ReadByte();
-						int type17 = reader.ReadInt16();
+						int stack8 = ModNet.AllowVanillaClients ? reader.ReadInt16() : -1;
+						int pre3 = ModNet.AllowVanillaClients ? reader.ReadByte() : -1;
+						int type17 = ModNet.AllowVanillaClients ? reader.ReadInt16() : -1;
+
 						if (num259 >= 0 && num259 < 8000) {
 							if (Main.chest[num259] == null)
 								Main.chest[num259] = new Chest();
@@ -1548,9 +_,17 @@
 							if (Main.chest[num259].item[num260] == null)
 								Main.chest[num259].item[num260] = new Item();
 
+							if (!ModNet.AllowVanillaClients) {
+								ItemIO.Receive(Main.chest[num259].item[num260], reader, readStack: true);
+								goto SkipVanillaItemSetup;
+							}
+
 							Main.chest[num259].item[num260].netDefaults(type17);
 							Main.chest[num259].item[num260].Prefix(pre3);
 							Main.chest[num259].item[num260].stack = stack8;
+
+							SkipVanillaItemSetup:
+
 							Recipe.FindRecipes(canDelayCheck: true);
 						}
 
@@ -1608,6 +_,7 @@
 						break;
 					}
 				case 34: {
+						// TODO, check that this didn't get messed up...why % 100?
 						byte b3 = reader.ReadByte();
 						int num16 = reader.ReadInt16();
 						int num17 = reader.ReadInt16();
@@ -1616,22 +_,32 @@
 						if (Main.netMode == 2)
 							num19 = 0;
 
+
+						ushort modType = 0;
+						if (b3 >= 100)
+							modType = reader.ReadUInt16();
+
 						if (Main.netMode == 2) {
 							switch (b3) {
-								case 0: {
-										int num20 = WorldGen.PlaceChest(num16, num17, 21, notNearOtherChests: false, num18);
+								case byte _ when b3 % 100 == 0: {
+										if (modType == 0)
+											modType = TileID.Containers;
+
+										int num20 = WorldGen.PlaceChest(num16, num17, modType, notNearOtherChests: false, num18);
 										if (num20 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num16, num17, num18, num20);
-											Item.NewItem(num16 * 16, num17 * 16, 32, 32, Chest.chestItemSpawn[num18], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num16, num17, num18, num20, modType);
+											int itemSpawn = b3 < 100 ? Chest.chestItemSpawn[num18] : TileLoader.GetTile(modType).ChestDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num16 * 16, num17 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num16, num17, num18, num20);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num16, num17, num18, num20, modType);
 										}
 
 										return;
 									}
-								case 1:
-									if (Main.tile[num16, num17].type == 21) {
+								case byte _ when b3 % 100 == 1:
+									if (Main.tile[num16, num17].type == 21 || b3 == 101 && TileID.Sets.BasicChest[Main.tile[num16, num17].type]) {
 										Tile tile = Main.tile[num16, num17];
 										if (tile.frameX % 36 != 0)
 											num16--;
@@ -1650,20 +_,25 @@
 							}
 
 							switch (b3) {
-								case 2: {
-										int num21 = WorldGen.PlaceChest(num16, num17, 88, notNearOtherChests: false, num18);
+								case byte _ when b3 % 100 == 2: {
+										if (modType == 0)
+											modType = TileID.Dressers;
+
+										int num21 = WorldGen.PlaceChest(num16, num17, modType, notNearOtherChests: false, num18);
 										if (num21 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num16, num17, num18, num21);
-											Item.NewItem(num16 * 16, num17 * 16, 32, 32, Chest.dresserItemSpawn[num18], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num16, num17, num18, num21, modType);
+											int itemSpawn = b3 < 100 ? Chest.dresserItemSpawn[num18] : TileLoader.GetTile(modType).DresserDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num16 * 16, num17 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num16, num17, num18, num21);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num16, num17, num18, num21, modType);
 										}
 
 										return;
 									}
-								case 3:
-									if (Main.tile[num16, num17].type == 88) {
+								case byte _ when b3 % 100 == 3:
+									if (Main.tile[num16, num17].type == 88 || b3 == 103 && TileID.Sets.BasicDresser[Main.tile[num16, num17].type]) {
 										Tile tile2 = Main.tile[num16, num17];
 										num16 -= tile2.frameX % 54 / 18;
 										if (tile2.frameY % 36 != 0)
@@ -1713,21 +_,25 @@
 						}
 
 						switch (b3) {
-							case 0:
+							case byte _ when b3 % 100 == 0:
 								if (num19 == -1) {
 									WorldGen.KillTile(num16, num17);
 									break;
 								}
 								SoundEngine.PlaySound(0, num16 * 16, num17 * 16);
+								if (modType == 0)
+									modType = TileID.Containers;
-								WorldGen.PlaceChestDirect(num16, num17, 21, num18, num19);
+								WorldGen.PlaceChestDirect(num16, num17, modType, num18, num19);
 								break;
-							case 2:
+							case byte _ when b3 % 100 == 2:
 								if (num19 == -1) {
 									WorldGen.KillTile(num16, num17);
 									break;
 								}
 								SoundEngine.PlaySound(0, num16 * 16, num17 * 16);
+								if (modType == 0)
+									modType = TileID.Dressers;
-								WorldGen.PlaceDresserDirect(num16, num17, 88, num18, num19);
+								WorldGen.PlaceDresserDirect(num16, num17, modType, num18, num19);
 								break;
 							case 4:
 								if (num19 == -1) {
@@ -1769,6 +_,9 @@
 						obj7.zone2 = reader.ReadByte();
 						obj7.zone3 = reader.ReadByte();
 						obj7.zone4 = reader.ReadByte();
+						if (!ModNet.AllowVanillaClients)
+							BiomeLoader.ReceiveCustomBiomes(obj7, reader);
+
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(36, -1, whoAmI, null, num224);
 
@@ -1790,6 +_,11 @@
 					if (Main.netMode == 2) {
 						if (reader.ReadString() == Netplay.ServerPassword) {
 							Netplay.Clients[whoAmI].State = 1;
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -1825,7 +_,8 @@
 						float itemRotation = reader.ReadSingle();
 						int itemAnimation = reader.ReadInt16();
 						player9.itemRotation = itemRotation;
+						// TML changes how item synchronization works, making it based on remote players' already-synchronized input. Setting itemAnimation in this packet would interfere with that.
-						player9.itemAnimation = itemAnimation;
+						//player9.itemAnimation = itemAnimation;
 						player9.channel = player9.inventory[player9.selectedItem].channel;
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(41, -1, whoAmI, null, num104);
@@ -1973,7 +_,7 @@
 							break;
 
 						Player player6 = Main.player[num55];
-						for (int num56 = 0; num56 < 22; num56++) {
+						for (int num56 = 0; num56 < Player.MaxBuffs; num56++) {
 							player6.buffType[num56] = reader.ReadUInt16();
 							if (player6.buffType[num56] > 0)
 								player6.buffTime[num56] = 60;
@@ -2181,7 +_,7 @@
 						if (Main.netMode != 2)
 							break;
 
-						if (num191 >= 0 && num191 < 668 && NPCID.Sets.MPAllowedEnemies[num191]) {
+						if (num191 >= 0 && num191 < NPCID.Sets.MPAllowedEnemies.Length && NPCID.Sets.MPAllowedEnemies[num191]) {
 							if (!NPC.AnyNPCs(num191))
 								NPC.SpawnOnPlayer(plr, num191);
 						}
@@ -2469,6 +_,7 @@
 					break;
 				case 74:
 					if (Main.netMode == 1) {
+						Netplay.syncingWorld = false;
 						Main.anglerQuest = reader.ReadByte();
 						Main.anglerQuestFinished = reader.ReadBoolean();
 					}
@@ -2558,7 +_,7 @@
 					if (Main.netMode == 1) {
 						int num83 = reader.ReadInt16();
 						int num84 = reader.ReadInt32();
-						if (num83 >= 0 && num83 < 289)
+						if (num83 >= 0) // && num83 < 289) //This prevented mod BannerIds from syncing.
 							NPC.killCount[num83] = num84;
 					}
 					break;
@@ -2594,7 +_,7 @@
 							}
 						}
 						else {
-							TileEntity tileEntity = TileEntity.Read(reader, networkSend: true);
+							TileEntity tileEntity = TileEntity.Read(reader, networkSend: true, lightSend: true); //TML: Made lightSend be true.
 							tileEntity.ID = num54;
 							TileEntity.ByID[tileEntity.ID] = tileEntity;
 							TileEntity.ByPosition[tileEntity.Position] = tileEntity;
@@ -2669,10 +_,22 @@
 					if (Main.netMode == 2) {
 						short x11 = reader.ReadInt16();
 						int y10 = reader.ReadInt16();
+
+						Item item;
+
+						if (ModNet.AllowVanillaClients) {
-						int netid3 = reader.ReadInt16();
+							int netId = reader.ReadInt16();
-						int prefix3 = reader.ReadByte();
+							int prefix = reader.ReadByte();
-						int stack5 = reader.ReadInt16();
+							int stack = reader.ReadInt16();
-						TEItemFrame.TryPlacing(x11, y10, netid3, prefix3, stack5);
+
+							item = new Item(netId, prefix, stack);
+						}
+						else {
+							item = ItemIO.Receive(reader);
+							item.stack = reader.ReadVarInt();
+						}
+
+						TEItemFrame.TryPlacing(x11, y10, item, item.stack);
 					}
 					break;
 				case 91: {
@@ -3148,10 +_,22 @@
 					if (Main.netMode == 2) {
 						short x8 = reader.ReadInt16();
 						int y7 = reader.ReadInt16();
+
+						Item item;
+
+						if (ModNet.AllowVanillaClients) {
-						int netid2 = reader.ReadInt16();
+							int netId = reader.ReadInt16();
-						int prefix2 = reader.ReadByte();
+							int prefix = reader.ReadByte();
-						int stack3 = reader.ReadInt16();
+							int stack = reader.ReadInt16();
-						TEWeaponsRack.TryPlacing(x8, y7, netid2, prefix2, stack3);
+
+							item = new Item(netId, stack, prefix);
+						}
+						else {
+							item = ItemIO.Receive(reader);
+							item.stack = reader.ReadVarInt();
+						}
+
+						TEWeaponsRack.TryPlacing(x8, y7, item, item.stack);
 					}
 					break;
 				case 124: {
@@ -3285,10 +_,22 @@
 					if (Main.netMode == 2) {
 						short x3 = reader.ReadInt16();
 						int y3 = reader.ReadInt16();
+
+						Item item;
+
+						if (ModNet.AllowVanillaClients) {
-						int netid = reader.ReadInt16();
+							int netId = reader.ReadInt16();
-						int prefix = reader.ReadByte();
+							int prefix = reader.ReadByte();
-						int stack2 = reader.ReadInt16();
+							int stack = reader.ReadInt16();
-						TEFoodPlatter.TryPlacing(x3, y3, netid, prefix, stack2);
+
+							item = new Item(netId, stack, prefix);
+						}
+						else {
+							item = ItemIO.Receive(reader);
+							item.stack = reader.ReadVarInt();
+						}
+
+						TEWeaponsRack.TryPlacing(x3, y3, item, item.stack);
 					}
 					break;
 				case 134: {
@@ -3348,6 +_,27 @@
 						CreditsRollEvent.SetRemainingTimeDirect(reader.ReadInt32());
 					}
 					break;
+				case MessageID.InGameChangeConfig:
+					ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+					return;
+				case MessageID.ModPacket:
+					ModNet.HandleModPacket(reader, whoAmI, length);
+					return;
+				case MessageID.SyncMods:
+					if (Main.netMode == 1) {
+						ModNet.SyncClientMods(reader);
+					}
+					else {
+						ModNet.SendNetIDs(whoAmI);
+						NetMessage.SendData(3, whoAmI);
+					}
+					return;
+				case MessageID.ModFile:
+					if (Main.netMode == 1)
+						ModNet.ReceiveMod(reader);
+					else
+						ModNet.SendMod(reader.ReadString(), whoAmI);
+					return;
 				default:
 					if (Netplay.Clients[whoAmI].State == 0)
 						NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
