--- src/TerrariaNetCore/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -1,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
 using Terraria.Audio;
@@ -17,12 +_,16 @@
 using Terraria.ID;
 using Terraria.Localization;
 using Terraria.Map;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Net;
 using Terraria.Net.Sockets;
 using Terraria.Social;
 using Terraria.Social.Steam;
 using Terraria.Testing;
 using Terraria.UI;
+using LegacySoundStyle = Terraria.Audio.SoundStyle;
 
 namespace Terraria;
 
@@ -113,8 +_,14 @@
 		int num = 0;
 		num = start + 1;
 		b = (byte)(messageType = readBuffer[start]);
+
+		if (ModNet.DetailedLogging)
+			ModNet.Debug(whoAmI, $"GetData {MessageID.GetName(b)}({b}), {length+2}");
+
+		/*
 		if (b >= MessageID.Count)
 			return;
+		*/
 
 		Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 		if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
@@ -135,7 +_,8 @@
 		}
 
 		if (Main.netMode == 2) {
+			//TML: Added 'b < 250' check.
-			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161)
+			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161 && b < 250)
 				NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 			if (Netplay.Clients[whoAmI].State == 0 && b != 1)
@@ -146,6 +_,10 @@
 			ResetReader();
 
 		reader.BaseStream.Position = num;
+
+		if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+			return;
+
 		switch (b) {
 			case 1: {
 				if (Main.netMode != 2)
@@ -163,9 +_,20 @@
 					if (Netplay.Clients[whoAmI].State != 0)
 						break;
 
+					string clientVersion = reader.ReadString();
+
+					/*
 					if (reader.ReadString() == "Terraria" + 326) {
+					*/
+					if (ModNet.IsClientCompatible(clientVersion, out ModNet.isModdedClient[whoAmI], out var kickMsg)) {
 						if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 							Netplay.Clients[whoAmI].State = 1;
+
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -174,7 +_,10 @@
 						}
 					}
 					else {
+						/*
 						NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+						*/
+						NetMessage.TrySendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n({kickMsg})"));
 					}
 				}
 
@@ -184,6 +_,9 @@
 				if (Main.netMode == 1) {
 					Netplay.Disconnect = true;
 					Main.statusText = NetworkText.Deserialize(reader).ToString();
+
+					// Added by TML.
+					Main.menuMode = MenuID.MultiplayerJoining;
 				}
 				break;
 			case 3:
@@ -230,6 +_,9 @@
 					if (!string.IsNullOrWhiteSpace(Netplay.HostToken))
 						NetMessage.TrySendData(161, -1, -1, NetworkText.FromLiteral(Netplay.HostToken));
 
+
+					PlayerLoader.SyncPlayer(player17, -1, -1, true);
+
 					NetMessage.TrySendData(6);
 					if (Netplay.Connection.State == 2)
 						Netplay.Connection.State = 3;
@@ -255,11 +_,11 @@
 
 				player18.voicePitchOffset = Utils.Clamp(player18.voicePitchOffset, -1f, 1f);
 				player18.hair = reader.ReadByte();
-				if (player18.hair >= 228)
+				if (player18.hair >= HairLoader.Count)
 					player18.hair = 0;
 
 				player18.name = reader.ReadString().Trim().Trim();
-				player18.hairDye = reader.ReadByte();
+				player18.hairDye = reader.Read7BitEncodedInt();
 				ReadAccessoryVisibility(reader, player18.hideVisibleAccessory);
 				player18.hideMisc = reader.ReadByte();
 				player18.hairColor = reader.ReadRGB();
@@ -335,9 +_,12 @@
 			case 5: {
 				int num25 = reader.ReadByte();
 				int num26 = reader.ReadInt16();
+				/*
 				int stack2 = reader.ReadInt16();
 				int prefixWeWant = reader.ReadByte();
 				int type2 = reader.ReadInt16();
+				*/
+				Item item = ItemIO.Receive(reader, readStack: true);
 				BitsByte bitsByte2 = reader.ReadByte();
 				bool favorited = bitsByte2[0];
 				bool flag = bitsByte2[1];
@@ -351,10 +_,12 @@
 				lock (player2) {
 					PlayerItemSlotID.SlotReference slot = new PlayerItemSlotID.SlotReference(player2, num26);
 					PlayerItemSlotID.SlotReference slotReference = new PlayerItemSlotID.SlotReference(Main.clientPlayer, num26);
+					/*
 					Item item = new Item();
 					item.SetDefaults(type2);
 					item.stack = stack2;
 					item.Prefix(prefixWeWant);
+					*/
 					item.favorited = favorited;
 					slot.Item = item;
 					if (num25 == Main.myPlayer && !Main.ServerSideCharacter)
@@ -568,6 +_,10 @@
 					ExtraSpawnPointManager.Read(reader, networking: true);
 					Main.dungeonX = reader.ReadInt16();
 					Main.dungeonY = reader.ReadInt16();
+
+					if (Netplay.Connection.State > 4)
+						WorldIO.ReceiveModData(reader);
+
 					if (Netplay.Connection.State == 3) {
 						Main.windSpeedCurrent = Main.windSpeedTarget;
 						Netplay.Connection.State = 4;
@@ -802,6 +_,10 @@
 					NetMessage.TrySendData(12, -1, whoAmI, null, whoAmI, (int)(byte)playerSpawnContext);
 					NetMessage.TrySendData(129, whoAmI);
 					NetMessage.greetPlayer(whoAmI);
+
+					if (Main.ActiveWorldFileData.ModSaveErrors.Any()) {
+						ChatHelper.SendChatMessageToClient(Utils.CreateSaveErrorMessage("tModLoader.WorldCustomDataSaveFail", Main.ActiveWorldFileData.ModSaveErrors), Color.OrangeRed, whoAmI);
+					}
 					if (Main.player[num104].unlockedBiomeTorches) {
 						NPC nPC = new NPC();
 						nPC.SetDefaults(664);
@@ -945,8 +_,11 @@
 					Player player9 = Main.player[num73];
 					player9.statLife = reader.ReadInt16();
 					player9.statLifeMax = reader.ReadInt16();
+					// Allow health stat to exceed vanilla bounds (#HealthManaAPI)
+					/*
 					if (player9.statLifeMax < 20)
 						player9.statLifeMax = 20;
+					*/
 
 					player9.dead = player9.statLife <= 0;
 					if (Main.netMode == 2)
@@ -1151,7 +_,7 @@
 				BitsByte bitsByte19 = (byte)0;
 				BitsByte bitsByte20 = (byte)0;
 				BitsByte bitsByte21 = (byte)0;
-				Tile tile4 = null;
+				Tile tile4 = default;
 				for (int num95 = num91; num95 < num91 + num93; num95++) {
 					for (int num96 = num92; num96 < num92 + num94; num96++) {
 						if (Main.tile[num95, num96] == null)
@@ -1231,8 +_,12 @@
 				int num41 = reader.ReadInt16();
 				Vector2 vector = reader.ReadVector2();
 				Vector2 velocity = reader.ReadVector2();
+				/*
 				int stack3 = reader.ReadInt16();
 				int num42 = reader.ReadByte();
+				*/
+				int stack3 = reader.Read7BitEncodedInt();
+				int num42 = reader.Read7BitEncodedInt();
 				BitsByte bitsByte4 = reader.ReadByte();
 				bool num43 = bitsByte4[2];
 				bool flag3 = bitsByte4[3];
@@ -1260,6 +_,7 @@
 						worldItem.Prefix(num42);
 
 					worldItem.stack = stack3;
+					ItemIO.ReceiveModData(worldItem.inner, reader);
 					worldItem.position = vector;
 					worldItem.velocity = velocity;
 					worldItem.shimmered = shimmered;
@@ -1272,6 +_,7 @@
 					}
 				}
 				else {
+					if (num41 == 400) Main.timeItemSlotCannotBeReusedFor[num41] = 0; // Fix #4372
 					if (Main.timeItemSlotCannotBeReusedFor[num41] > 0)
 						break;
 
@@ -1293,6 +_,7 @@
 						worldItem.Prefix(num42);
 
 					worldItem.stack = stack3;
+					ItemIO.ReceiveModData(worldItem.inner, reader);
 					worldItem.position = vector;
 					worldItem.velocity = velocity;
 					worldItem.shimmered = shimmered;
@@ -1443,9 +_,13 @@
 				if (num179 > -1)
 					nPC5.TransformVisuals(num179, nPC5.type);
 
-				if (nPC5.type >= 0 && nPC5.type < NPCID.Count && Main.npcCatchable[nPC5.type])
+				if (nPC5.type >= 0 && Main.npcCatchable[nPC5.type])
 					nPC5.releaseOwner = reader.ReadByte();
 
+				// Extra AI is read into a temporary buffer for parity with ProjectileLoader code, and to avoid exceptions causing underreads.
+				if (bitsByte31[5])
+					NPCLoader.ReceiveExtraAI(nPC5, NPCLoader.ReadExtraAI(reader));
+
 				if (flag16)
 					nPC5.OnSpawn(new EntitySource_Sync());
 
@@ -1469,6 +_,8 @@
 				float knockBack2 = (bitsByte28[5] ? reader.ReadSingle() : 0f);
 				int originalDamage = (bitsByte28[6] ? reader.ReadInt16() : 0);
 				array[2] = (bitsByte29[0] ? reader.ReadSingle() : 0f);
+				byte[] extraAI = bitsByte29[1] ? ProjectileLoader.ReadExtraAI(reader) : null;
+
 				if (Main.netMode == 2 && (Main.projHostile[num158] || key2.Spawner != whoAmI))
 					break;
 
@@ -1496,6 +_,10 @@
 					proj2.ai[num159] = array[num159];
 				}
 
+				if (extraAI != null) {
+					ProjectileLoader.ReceiveExtraAI(proj2, extraAI);
+				}
+
 				if (flag15)
 					proj2.FinalizeProjectile();
 
@@ -1510,10 +_,29 @@
 
 				int num240 = reader.ReadByte();
 				int num241 = reader.ReadByte();
+				/*
 				int num242 = reader.ReadInt16();
 				float num243 = reader.ReadSingle();
 				int num244 = reader.ReadByte() - 1;
 				byte b18 = reader.ReadByte();
+				*/
+
+				var num242 = reader.Read7BitEncodedInt();
+				NPC.HitInfo hit = default;
+				if (num212 >= 0) {
+					hit = new NPC.HitInfo {
+						Damage = num212,
+						SourceDamage = reader.Read7BitEncodedInt(),
+						DamageType = DamageClassLoader.DamageClasses[reader.Read7BitEncodedInt()],
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle()
+					};
+					BitsByte flags = reader.ReadByte();
+					hit.Crit = flags[0];
+					hit.InstantKill = flags[1];
+					hit.HideCombatText = flags[2];
+				}
+
 				NPC nPC7 = Main.npc[num240];
 				if (Main.netMode == 2) {
 					if (nPC7.generation != num241)
@@ -1529,7 +_,10 @@
 				}
 
 				if (num242 >= 0) {
+					/*
 					nPC7.StrikeNPC(num242, num243, num244, b18 == 1, fromNet: true, (Main.netMode == 2) ? whoAmI : 255);
+					*/
+					nPC7.StrikeNPC(hit, fromNet: true);
 				}
 				else {
 					nPC7.life = 0;
@@ -1538,7 +_,11 @@
 				}
 
 				if (Main.netMode == 2) {
+					/*
 					NetMessage.TrySendData(28, -1, whoAmI, null, num240, num242, num243, num244, b18);
+					*/
+					NetMessage.SendStrikeNPC(Main.npc[num240], hit, ignoreClient: whoAmI);
+
 					if (nPC7.life <= 0)
 						NetMessage.TrySendData(23, -1, -1, null, num240);
 
@@ -1616,16 +_,26 @@
 			case 32: {
 				int num149 = reader.ReadInt16();
 				int num150 = reader.ReadByte();
+				/*
 				int stack7 = reader.ReadInt16();
 				int prefixWeWant3 = reader.ReadByte();
 				int type15 = reader.ReadInt16();
+				*/
+				int stack7 = -1;
+				int prefixWeWant3 = -1;
+				int type15 = -1;
 				if (num149 >= 0 && num149 < 8000 && Main.chest[num149] != null) {
 					if (Main.chest[num149].item[num150] == null)
 						Main.chest[num149].item[num150] = new Item();
 
+					ItemIO.Receive(Main.chest[num149].item[num150], reader, readStack: true);
+					goto SkipVanillaItemSetup;
+
 					Main.chest[num149].item[num150].SetDefaults(type15);
 					Main.chest[num149].item[num150].Prefix(prefixWeWant3);
 					Main.chest[num149].item[num150].stack = stack7;
+					SkipVanillaItemSetup:
+
 					if (Main.netMode == 2)
 						NetMessage.TrySendData(32, -1, whoAmI, null, num149, num150);
 				}
@@ -1639,7 +_,7 @@
 				int num24 = reader.ReadByte();
 				string name = string.Empty;
 				if (num24 != 0) {
-					if (num24 <= 20)
+					if (num24 <= Chest.MaxNameLength)
 						name = reader.ReadString();
 					else if (num24 != 255)
 						num24 = 0;
@@ -1693,25 +_,49 @@
 				if (Main.netMode == 2)
 					num35 = 0;
 
+
+				ushort modType = 0;
+				if (b2 >= 100)
+					modType = reader.ReadUInt16();
+
 				if (Main.netMode == 2) {
 					using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
 						switch (b4) {
+							/*
 							case 0: {
+							*/
+							case byte _ when b4 % 100 == 0: {
+								if (modType == 0)
+									modType = TileID.Containers;
+	
+								//TML: '21' -> 'modType'
-								int num38 = WorldGen.PlaceChest(num32, num33, 21, notNearOtherChests: false, num34);
+								int num38 = WorldGen.PlaceChest(num32, num33, modType, notNearOtherChests: false, num34);
+
 								if (num38 == -1) {
+									//TML: Added 'modType' at the end.
-									NetMessage.TrySendData(34, whoAmI, -1, null, b4, num32, num33, num34, num38);
+									NetMessage.TrySendData(34, whoAmI, -1, null, b4, num32, num33, num34, num38, modType);
+									/*
 									int itemDrop_Chests2 = WorldGen.GetItemDrop_Chests(num34, secondType: false);
+									*/
+									int itemDrop_Chests2 = b4 < 100 ? WorldGen.GetItemDrop_Chests(num34, secondType: false) : TileLoader.GetItemDropFromTypeAndStyle(modType, num34);
 									if (itemDrop_Chests2 > 0)
 										Item.NewItem(new EntitySource_TileBreak(num32, num33), num32 * 16, num33 * 16, 32, 32, itemDrop_Chests2, 1, noBroadcast: true);
 								}
 								else {
+									//TML: Added 'modType' at the end.
-									NetMessage.TrySendData(34, -1, -1, null, b4, num32, num33, num34, num38);
+									NetMessage.TrySendData(34, -1, -1, null, b4, num32, num33, num34, num38, modType);
 								}
 
 								break;
 							}
+							/*
 							case 1:
+							*/
+							case byte _ when b4 % 100 == 1:
+								/*
 								if (Main.tile[num32, num33].type == 21) {
+								*/
+								if (Main.tile[num32, num33].type == 21 || b4 == 101 && TileID.Sets.BasicChest[Main.tile[num32, num33].type]) {
 									Tile tile = Main.tile[num32, num33];
 									if (tile.frameX % 36 != 0)
 										num32--;
@@ -1729,20 +_,42 @@
 								goto default;
 							default:
 								switch (b4) {
+									/*
 									case 2: {
+									*/
+									case byte _ when b4 % 100 == 2: {
+										if (modType == 0)
+											modType = TileID.Dressers;
+	
+										//TML: '88' -> 'modType'
-										int num36 = WorldGen.PlaceChest(num32, num33, 88, notNearOtherChests: false, num34);
+										int num36 = WorldGen.PlaceChest(num32, num33, modType, notNearOtherChests: false, num34);
+
 										if (num36 == -1) {
+											//TML: Added 'modType' at the end.
-											NetMessage.TrySendData(34, whoAmI, -1, null, b4, num32, num33, num34, num36);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b4, num32, num33, num34, num36, modType);
+
+											/*
 											Item.NewItem(new EntitySource_TileBreak(num32, num33), num32 * 16, num33 * 16, 32, 32, WorldGen.GetItemDrop_Dressers(num34), 1, noBroadcast: true);
+											*/
+											int itemSpawn = b4 < 100 ? WorldGen.GetItemDrop_Dressers(num34) : TileLoader.GetItemDropFromTypeAndStyle(modType, num34);
+											if (itemSpawn > 0)
+												Item.NewItem(new EntitySource_TileBreak(num32, num33), num32 * 16, num33 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
+											//TML: Added 'modType' at the end.
-											NetMessage.TrySendData(34, -1, -1, null, b4, num32, num33, num34, num36);
+											NetMessage.TrySendData(34, -1, -1, null, b4, num32, num33, num34, num36, modType);
 										}
 
 										break;
 									}
+									/*
 									case 3:
+									*/
+									case byte _ when b4 % 100 == 3:
+										/*
 										if (Main.tile[num32, num33].type == 88) {
+										*/
+										if (Main.tile[num32, num33].type == 88 || (b4 == 103 && TileID.Sets.BasicDresser[Main.tile[num32, num33].type])) {
 											Tile tile2 = Main.tile[num32, num33];
 											num32 -= tile2.frameX % 54 / 18;
 											if (tile2.frameY % 36 != 0)
@@ -1798,21 +_,37 @@
 				}
 
 				switch (b4) {
+					/*
 					case 0:
+					*/
+					case byte _ when b4 % 100 == 0:
 						if (num35 == -1) {
 							WorldGen.KillTile(num32, num33);
 							break;
 						}
 						SoundEngine.PlaySound(0, num32 * 16, num33 * 16);
+
+						if (modType == 0)
+							modType = TileID.Containers;
+
+						//TML: '21' -> 'modType'.
-						WorldGen.PlaceChestDirect(num32, num33, 21, num34, num35);
+						WorldGen.PlaceChestDirect(num32, num33, modType, num34, num35);
 						break;
+					/*
 					case 2:
+					*/
+					case byte _ when b4 % 100 == 2:
 						if (num35 == -1) {
 							WorldGen.KillTile(num32, num33);
 							break;
 						}
 						SoundEngine.PlaySound(0, num32 * 16, num33 * 16);
+
+						if (modType == 0)
+							modType = TileID.Dressers;
+
+						//TML: '88' -> 'modType'.
-						WorldGen.PlaceDresserDirect(num32, num33, 88, num34, num35);
+						WorldGen.PlaceDresserDirect(num32, num33, modType, num34, num35);
 						break;
 					case 4:
 						if (num35 == -1) {
@@ -1857,6 +_,12 @@
 				player16.zone4 = reader.ReadByte();
 				player16.zone5 = reader.ReadByte();
 				player16.townNPCs = reader.ReadByte();
+
+				BiomeLoader.ReceiveCustomBiomes(player16, reader);
+
+				//TML: Recalculate the 'not in any biome' value.
+				player16.ZonePurity = player16.InZonePurity();
+
 				if (Main.netMode == 2) {
 					if (!flag12 && player16.zone5[0])
 						NPC.Spawner.SpawnFaelings(player16);
@@ -1882,6 +_,12 @@
 				if (Main.netMode == 2) {
 					if (reader.ReadString() == Netplay.ServerPassword) {
 						Netplay.Clients[whoAmI].State = 1;
+
+						if (ModNet.isModdedClient[whoAmI]) {
+							ModNet.SyncMods(whoAmI);
+							return;
+						}
+
 						NetMessage.TrySendData(3, whoAmI);
 					}
 					else {
@@ -1928,7 +_,12 @@
 				float itemRotation = reader.ReadSingle();
 				int itemAnimation = reader.ReadInt16();
 				player3.itemRotation = itemRotation;
+				// TML changes how item synchronization works, making it based on remote players' already-synchronized input.
+				// Setting itemAnimation in this packet would interfere with that. #ItemTimeOnAllClients
+				/*
 				player3.itemAnimation = itemAnimation;
+				*/
+				
 				player3.channel = player3.inventory[player3.selectedItem].channel;
 				if (Main.netMode == 2)
 					NetMessage.TrySendData(41, -1, whoAmI, null, num30);
@@ -2053,7 +_,11 @@
 				if (Main.tile[num3, num4] == null)
 					Main.tile[num3, num4] = new Tile();
 
+				// Can't lock on struct. Lock does nothing anyway as this is the only place it's acquired
+				/*
 				lock (Main.tile[num3, num4]) {
+				*/
+				{
 					Main.tile[num3, num4].liquid = b2;
 					Main.tile[num3, num4].liquidType(liquidType);
 					if (Main.netMode == 2) {
@@ -2319,7 +_,7 @@
 				if (Main.netMode != 2)
 					break;
 
-				if (num238 >= 0 && num238 < NPCID.Count && NPCID.Sets.MPAllowedEnemies[num238]) {
+				if (num238 >= 0 && NPCID.Sets.MPAllowedEnemies[num238]) {
 					if (!NPC.AnyNPCs(num238))
 						NPC.SpawnOnPlayer(num237, num238);
 				}
@@ -2757,7 +_,8 @@
 						TileEntity.Remove(result3);
 				}
 				else {
-					TileEntity tileEntity = TileEntity.Read(reader, 326, networkSend: true);
+					//TML: Made 'lightSend' true.
+					TileEntity tileEntity = TileEntity.Read(reader, Main.curRelease, networkSend: true, lightSend: true);
 					tileEntity.ID = num116;
 					TileEntity.Add(tileEntity);
 				}
@@ -2833,11 +_,21 @@
 
 				int x = reader.ReadInt16();
 				int y = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader);
+				item.stack = reader.Read7BitEncodedInt();
+				/*
 				int type = reader.ReadInt16();
 				int prefix = reader.ReadByte();
 				int stack = reader.ReadInt16();
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEItemFrame.TryPlacing(x, y, type, prefix, stack);
+					*/
+					TEItemFrame.TryPlacing(x, y, item, 1);
 					break;
 				}
 			}
@@ -2847,6 +_,13 @@
 
 				int num246 = reader.ReadInt32();
 				int num247 = reader.ReadByte();
+
+				// If type corresponds to projectile (magic number 2), read another byte for owner #WorldUIAnchorProjectileSyncFix
+				if (num247 == 2) {
+					int owner = reader.ReadByte();
+					num247 |= owner << 8; // Reassign num247 for use in DeserializeNetAnchor - this is now "packedOwnerType"
+				}
+
 				if (num247 == 255) {
 					if (EmoteBubble.byID.ContainsKey(num246))
 						EmoteBubble.byID.Remove(num246);
@@ -2879,6 +_,7 @@
 					EmoteBubble.byID[num246].ID = num246;
 					EmoteBubble.byID[num246].metadata = metadata;
 					EmoteBubble.OnBubbleChange(num246);
+					EmoteBubbleLoader.OnSpawn(EmoteBubble.byID[num246]);
 					break;
 				}
 			}
@@ -3250,7 +_,16 @@
 				break;
 			case 117: {
 				int num27 = reader.ReadByte();
+				bool directHurt = false;
+				if (num27 == 255) {
+					num27 = reader.ReadByte();
+					directHurt = true;
+				}
+
 				if (Main.netMode != 2 || whoAmI == num27 || (Main.player[num27].hostile && Main.player[whoAmI].hostile)) {
+					if (directHurt)
+						goto ReadDirectHurt;
+
 					PlayerDeathReason playerDeathReason2 = PlayerDeathReason.FromReader(reader);
 					int damage = reader.ReadInt16();
 					int num28 = reader.ReadByte() - 1;
@@ -3261,6 +_,26 @@
 					Main.player[num27].Hurt(playerDeathReason2, damage, num28, pvp2, quiet: true, flag2, num29);
 					if (Main.netMode == 2)
 						NetMessage.SendPlayerHurt(num27, playerDeathReason2, damage, num28, flag2, pvp2, num29, -1, whoAmI);
+
+					break;
+
+					ReadDirectHurt:
+					BitsByte pack = reader.ReadByte();
+					var args = new Player.HurtInfo {
+						DamageSource = PlayerDeathReason.FromReader(reader),
+						PvP = pack[0],
+						CooldownCounter = reader.ReadSByte(),
+						Dodgeable = pack[1],
+						SourceDamage = reader.Read7BitEncodedInt(),
+						Damage = reader.Read7BitEncodedInt(),
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle(),
+						DustDisabled = pack[2],
+						SoundDisabled = pack[3]
+					};
+					Main.player[num27].Hurt(args, quiet: true);
+					if (Main.netMode == 2)
+						NetMessage.SendPlayerHurt(num27, args, ignoreClient: whoAmI);
 				}
 
 				break;
@@ -3286,7 +_,7 @@
 					num259 = whoAmI;
 
 				int num260 = reader.ReadByte();
-				if (num260 >= 0 && num260 < EmoteID.Count && Main.netMode == 2) {
+				if (num260 >= 0 && num260 < EmoteBubbleLoader.EmoteBubbleCount && Main.netMode == 2) {
 					EmoteBubble.NewBubble(num260, new WorldUIAnchor(Main.player[num259]), 360);
 					EmoteBubble.CheckForNPCsToReactToEmoteBubble(num260, Main.player[num259]);
 				}
@@ -3347,11 +_,21 @@
 
 				int x14 = reader.ReadInt16();
 				int y14 = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader, readStack: true);
+				/*
 				int type19 = reader.ReadInt16();
 				int prefix4 = reader.ReadByte();
 				int stack8 = reader.ReadInt16();
+				item = new Item(netid2, stack4, prefix2);
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEWeaponsRack.TryPlacing(x14, y14, type19, prefix4, stack8);
+					*/
+					TEWeaponsRack.TryPlacing(x14, y14, item, item.stack);
 					break;
 				}
 			}
@@ -3480,10 +_,23 @@
 					ushort key = reader.ReadUInt16();
 					LegacySoundStyle legacySoundStyle = SoundID.SoundByIndex[key];
 					BitsByte bitsByte23 = reader.ReadByte();
+					/*
 					int num117 = -1;
 					float num118 = 1f;
 					float num119 = 0f;
 					SoundEngine.PlaySound(Style: (!bitsByte23[0]) ? legacySoundStyle.Style : reader.ReadInt32(), volumeScale: (!bitsByte23[1]) ? legacySoundStyle.Volume : MathHelper.Clamp(reader.ReadSingle(), 0f, 1f), pitchOffset: (!bitsByte23[2]) ? legacySoundStyle.GetRandomPitch() : MathHelper.Clamp(reader.ReadSingle(), -1f, 1f), type: legacySoundStyle.SoundId, x: point2.X, y: point2.Y);
+					*/
+
+					if (bitsByte23[0])
+						legacySoundStyle.Variants = new int[] { reader.ReadInt32() };
+
+					if (bitsByte23[1])
+						legacySoundStyle.Volume = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					if (bitsByte23[2])
+						legacySoundStyle.Pitch = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					SoundEngine.PlaySound(legacySoundStyle, point2.ToVector2());
 				}
 				break;
 			case 133: {
@@ -3492,17 +_,27 @@
 
 				int x9 = reader.ReadInt16();
 				int y9 = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader, readStack: true);
+				/*
 				int type12 = reader.ReadInt16();
 				int prefix3 = reader.ReadByte();
 				int stack6 = reader.ReadInt16();
+				item = new Item(netid, stack, prefix);
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEFoodPlatter.TryPlacing(x9, y9, type12, prefix3, stack6);
+					*/
+					TEFoodPlatter.TryPlacing(x9, y9, item, item.stack);
 					break;
 				}
 			}
 			case 134: {
 				int num106 = reader.ReadByte();
-				int ladyBugLuckTimeLeft = reader.ReadInt32();
+				double ladyBugLuckTimeLeft = reader.ReadDouble();
 				float torchLuck = reader.ReadSingle();
 				byte luckPotion = reader.ReadByte();
 				bool hasGardenGnomeNearby = reader.ReadBoolean();
@@ -3650,11 +_,18 @@
 
 				int x4 = reader.ReadInt16();
 				int y4 = reader.ReadInt16();
+				Item item = ItemIO.Receive(reader);
+				int stack4 = reader.Read7BitEncodedInt();
+				/*
 				int type6 = reader.ReadInt16();
 				int prefix2 = reader.ReadByte();
 				int stack4 = reader.ReadInt16();
+				*/
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEDeadCellsDisplayJar.TryPlacing(x4, y4, type6, prefix2, stack4);
+					*/
+					TEDeadCellsDisplayJar.TryPlacing(x4, y4, item, stack4);
 					break;
 				}
 			}
@@ -3692,7 +_,7 @@
 					Player player5 = Main.player[num51];
 					Item item3 = player5.inventory[player5.selectedItem];
 					if (item3.UseSound != null)
-						SoundEngine.PlaySound(item3.UseSound, player5.Center, item3.useSoundPitch);
+						SoundEngine.PlaySound(item3.UseSound, player5.Center /*, item3.useSoundPitch*/);
 				}
 
 				break;
@@ -3753,6 +_,35 @@
 				Main.player[whoAmI].host = !string.IsNullOrWhiteSpace(Netplay.HostToken) && Netplay.HostToken == text;
 				break;
 			}
+			case MessageID.InGameChangeConfig:
+				ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+				return;
+			case MessageID.ModPacket:
+				ModNet.HandleModPacket(reader, whoAmI, length);
+				return;
+			case MessageID.SyncMods:
+				if (Main.netMode == 1) {
+					ModNet.SyncClientMods(reader);
+				}
+				else {
+					ModNet.SendNetIDs(whoAmI);
+					NetMessage.SendData(3, whoAmI);
+				}
+				return;
+			case MessageID.ModFile:
+				if (Main.netMode == 1)
+					ModNet.ReceiveMod(reader);
+				else
+					ModNet.SendMod(reader.ReadString(), whoAmI);
+				return;
+			case MessageID.KeepAliveDuringModReload:
+				string keepAlive = "Keep Alive During Mod Reload";
+				ModNet.Log(whoAmI, keepAlive);
+				// Logging.ServerConsoleLine doesn't log client info and ModNet.Log doesn't output to console, so manually replicate here.
+				var client = Netplay.Clients[whoAmI];
+				Console.WriteLine($"[{whoAmI}][{client.Socket?.GetRemoteAddress()?.GetFriendlyName()} ({client.Name})]: {keepAlive}");
+				// Does nothing. "Netplay.Clients[whoAmI].TimeOutTimer = 0;" at top of method should keep the connection alive
+				return;
 			default:
 				if (Main.netMode == 2 && Netplay.Clients[whoAmI].State == 0)
 					NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
