--- src/TerrariaNetCore/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -1,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
 using Terraria.Audio;
@@ -17,10 +_,14 @@
 using Terraria.ID;
 using Terraria.Localization;
 using Terraria.Map;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Net;
 using Terraria.Net.Sockets;
 using Terraria.Testing;
 using Terraria.UI;
+using LegacySoundStyle = Terraria.Audio.SoundStyle;
 
 namespace Terraria;
 
@@ -111,8 +_,14 @@
 		int num = 0;
 		num = start + 1;
 		b = (byte)(messageType = readBuffer[start]);
+
+		if (ModNet.DetailedLogging)
+			ModNet.Debug(whoAmI, $"GetData {MessageID.GetName(b)}({b}), {length+2}");
+
+		/*
 		if (b >= MessageID.Count)
 			return;
+		*/
 
 		Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 		if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
@@ -133,7 +_,8 @@
 		}
 
 		if (Main.netMode == 2) {
+			//TML: Added 'b < 250' check.
-			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161)
+			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161 && b < 250)
 				NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 			if (Netplay.Clients[whoAmI].State == 0 && b != 1)
@@ -144,6 +_,10 @@
 			ResetReader();
 
 		reader.BaseStream.Position = num;
+
+		if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+			return;
+
 		switch (b) {
 			case 1:
 				if (Main.netMode != 2)
@@ -155,9 +_,20 @@
 					if (Netplay.Clients[whoAmI].State != 0)
 						break;
 
+					string clientVersion = reader.ReadString();
+
+					/*
 					if (reader.ReadString() == "Terraria" + 319) {
+					*/
+					if (ModNet.IsClientCompatible(clientVersion, out ModNet.isModdedClient[whoAmI], out var kickMsg)) {
 						if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 							Netplay.Clients[whoAmI].State = 1;
+
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -166,7 +_,10 @@
 						}
 					}
 					else {
+						/*
 						NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+						*/
+						NetMessage.TrySendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n({kickMsg})"));
 					}
 				}
 				break;
@@ -174,6 +_,9 @@
 				if (Main.netMode == 1) {
 					Netplay.Disconnect = true;
 					Main.statusText = NetworkText.Deserialize(reader).ToString();
+
+					// Added by TML.
+					Main.menuMode = MenuID.MultiplayerJoining;
 				}
 				break;
 			case 3:
@@ -220,6 +_,9 @@
 					if (!string.IsNullOrWhiteSpace(Netplay.HostToken))
 						NetMessage.TrySendData(161, -1, -1, NetworkText.FromLiteral(Netplay.HostToken));
 
+
+					PlayerLoader.SyncPlayer(player9, -1, -1, true);
+
 					NetMessage.TrySendData(6);
 					if (Netplay.Connection.State == 2)
 						Netplay.Connection.State = 3;
@@ -245,11 +_,11 @@
 
 				player18.voicePitchOffset = Utils.Clamp(player18.voicePitchOffset, -1f, 1f);
 				player18.hair = reader.ReadByte();
-				if (player18.hair >= 228)
+				if (player18.hair >= HairLoader.Count)
 					player18.hair = 0;
 
 				player18.name = reader.ReadString().Trim().Trim();
-				player18.hairDye = reader.ReadByte();
+				player18.hairDye = reader.Read7BitEncodedInt();
 				ReadAccessoryVisibility(reader, player18.hideVisibleAccessory);
 				player18.hideMisc = reader.ReadByte();
 				player18.hairColor = reader.ReadRGB();
@@ -325,9 +_,12 @@
 			case 5: {
 				int num35 = reader.ReadByte();
 				int num36 = reader.ReadInt16();
+				/*
 				int stack3 = reader.ReadInt16();
 				int prefixWeWant2 = reader.ReadByte();
 				int type4 = reader.ReadInt16();
+				*/
+				Item item = ItemIO.Receive(reader, readStack: true);
 				BitsByte bitsByte3 = reader.ReadByte();
 				bool favorited = bitsByte3[0];
 				bool flag2 = bitsByte3[1];
@@ -341,10 +_,12 @@
 				lock (player2) {
 					PlayerItemSlotID.SlotReference slot = new PlayerItemSlotID.SlotReference(player2, num36);
 					PlayerItemSlotID.SlotReference slotReference = new PlayerItemSlotID.SlotReference(Main.clientPlayer, num36);
+					/*
 					Item item = new Item();
 					item.SetDefaults(type4);
 					item.stack = stack3;
 					item.Prefix(prefixWeWant2);
+					*/
 					item.favorited = favorited;
 					slot.Item = item;
 					if (num35 == Main.myPlayer && !Main.ServerSideCharacter)
@@ -551,6 +_,10 @@
 					Main.LobbyId = reader.ReadUInt64();
 					Sandstorm.IntendedSeverity = reader.ReadSingle();
 					ExtraSpawnPointManager.Read(reader, networking: true);
+
+					if (Netplay.Connection.State > 4)
+						WorldIO.ReceiveModData(reader);
+
 					if (Netplay.Connection.State == 3) {
 						Main.windSpeedCurrent = Main.windSpeedTarget;
 						Netplay.Connection.State = 4;
@@ -787,6 +_,10 @@
 					NetMessage.TrySendData(12, -1, whoAmI, null, whoAmI, (int)(byte)playerSpawnContext);
 					NetMessage.TrySendData(129, whoAmI);
 					NetMessage.greetPlayer(whoAmI);
+
+					if (Main.ActiveWorldFileData.ModSaveErrors.Any()) {
+						ChatHelper.SendChatMessageToClient(Utils.CreateSaveErrorMessage("tModLoader.WorldCustomDataSaveFail", Main.ActiveWorldFileData.ModSaveErrors), Color.OrangeRed, whoAmI);
+					}
 					if (Main.player[num144].unlockedBiomeTorches) {
 						NPC nPC = new NPC();
 						nPC.SetDefaults(664);
@@ -924,8 +_,11 @@
 					Player player15 = Main.player[num164];
 					player15.statLife = reader.ReadInt16();
 					player15.statLifeMax = reader.ReadInt16();
+					// Allow health stat to exceed vanilla bounds (#HealthManaAPI)
+					/*
 					if (player15.statLifeMax < 20)
 						player15.statLifeMax = 20;
+					*/
 
 					player15.dead = player15.statLife <= 0;
 					if (Main.netMode == 2)
@@ -1128,7 +_,7 @@
 				BitsByte bitsByte5 = (byte)0;
 				BitsByte bitsByte6 = (byte)0;
 				BitsByte bitsByte7 = (byte)0;
-				Tile tile4 = null;
+				Tile tile4 = default;
 				for (int l = num62; l < num62 + num64; l++) {
 					for (int m = num63; m < num63 + num65; m++) {
 						if (Main.tile[l, m] == null)
@@ -1210,8 +_,12 @@
 				int num201 = reader.ReadInt16();
 				Vector2 position4 = reader.ReadVector2();
 				Vector2 velocity3 = reader.ReadVector2();
+/*
 				int stack7 = reader.ReadInt16();
 				int prefix4 = reader.ReadByte();
+*/
+				int stack7 = reader.Read7BitEncodedInt();
+				int prefix4 = reader.Read7BitEncodedInt();
 				BitsByte bitsByte15 = reader.ReadByte();
 				bool flag16 = bitsByte15[0];
 				bool flag17 = bitsByte15[1];
@@ -1231,11 +_,12 @@
 				if (Main.netMode == 1) {
 					ItemSyncPersistentStats itemSyncPersistentStats = default(ItemSyncPersistentStats);
 					itemSyncPersistentStats.CopyFrom(worldItem4);
-					bool newAndShiny = (worldItem4.newAndShiny || worldItem4.type != num202) && ItemSlot.Options.HighlightNewItems && (num202 < 0 || num202 >= ItemID.Count || !ItemID.Sets.NeverAppearsAsNewInInventory[num202]);
+					bool newAndShiny = (worldItem4.newAndShiny || worldItem4.type != num202) && ItemSlot.Options.HighlightNewItems && (num202 < 0 || !ItemID.Sets.NeverAppearsAsNewInInventory[num202]);
 					worldItem4.SetDefaults(num202);
 					worldItem4.newAndShiny = newAndShiny;
 					worldItem4.Prefix(prefix4);
 					worldItem4.stack = stack7;
+					ItemIO.ReceiveModData(worldItem4.inner, reader);
 					worldItem4.position = position4;
 					worldItem4.velocity = velocity3;
 					worldItem4.shimmered = shimmered;
@@ -1254,6 +_,7 @@
 					itemSyncPersistentStats.PasteInto(worldItem4);
 				}
 				else {
+					if (num201 == 400) Main.timeItemSlotCannotBeReusedFor[num201] = 0; // Fix #4372
 					if (Main.timeItemSlotCannotBeReusedFor[num201] > 0)
 						break;
 
@@ -1278,6 +_,7 @@
 
 					worldItem4.Prefix(prefix4);
 					worldItem4.stack = stack7;
+					ItemIO.ReceiveModData(worldItem4.inner, reader);
 					worldItem4.position = position4;
 					worldItem4.velocity = velocity3;
 					worldItem4.timeLeftInWhichTheItemCannotBeTakenByEnemies = timeLeftInWhichTheItemCannotBeTakenByEnemies;
@@ -1434,11 +_,16 @@
 				if (num218 == 668)
 					NPC.deerclopsBoss = num215;
 
-				if (nPC5.type >= 0 && nPC5.type < NPCID.Count && Main.npcCatchable[nPC5.type])
+				if (nPC5.type >= 0 && Main.npcCatchable[nPC5.type])
 					nPC5.releaseOwner = reader.ReadByte();
 
+				// Extra AI is read into a temporary buffer for parity with ProjectileLoader code, and to avoid exceptions causing underreads.
+				if (bitsByte21[5])
+					NPCLoader.ReceiveExtraAI(nPC5, NPCLoader.ReadExtraAI(reader));
+
 				break;
 			}
+			/*
 			case 24: {
 				int num154 = reader.ReadInt16();
 				int num155 = reader.ReadByte();
@@ -1454,6 +_,7 @@
 
 				break;
 			}
+			*/
 			case 27: {
 				int num226 = reader.ReadInt16();
 				Vector2 position5 = reader.ReadVector2();
@@ -1474,6 +_,8 @@
 					num229 = -1;
 
 				array2[2] = (bitsByte23[0] ? reader.ReadSingle() : 0f);
+				byte[] extraAI = bitsByte23[1] ? ProjectileLoader.ReadExtraAI(reader) : null;
+
 				if (Main.netMode == 2) {
 					if (num228 == 949) {
 						num227 = 255;
@@ -1530,6 +_,10 @@
 					Main.projectileIdentity[num227, num229] = num230;
 				}
 
+				if (extraAI != null) {
+					ProjectileLoader.ReceiveExtraAI(projectile, extraAI);
+				}
+
 				projectile.ProjectileFixDesperation();
 				if (Main.netMode == 2)
 					NetMessage.TrySendData(27, -1, whoAmI, null, num230);
@@ -1538,10 +_,29 @@
 			}
 			case 28: {
 				int num211 = reader.ReadInt16();
+				/*
 				int num212 = reader.ReadInt16();
 				float num213 = reader.ReadSingle();
 				int num214 = reader.ReadByte() - 1;
 				byte b14 = reader.ReadByte();
+				*/
+
+				var num212 = reader.Read7BitEncodedInt();
+				NPC.HitInfo hit = default;
+				if (num212 >= 0) {
+					hit = new NPC.HitInfo {
+						Damage = num212,
+						SourceDamage = reader.Read7BitEncodedInt(),
+						DamageType = DamageClassLoader.DamageClasses[reader.Read7BitEncodedInt()],
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle()
+					};
+					BitsByte flags = reader.ReadByte();
+					hit.Crit = flags[0];
+					hit.InstantKill = flags[1];
+					hit.HideCombatText = flags[2];
+				}
+
 				if (Main.netMode == 2) {
 					if (num212 < 0)
 						num212 = 0;
@@ -1550,7 +_,10 @@
 				}
 
 				if (num212 >= 0) {
+					/*
 					Main.npc[num211].StrikeNPC(num212, num213, num214, b14 == 1, noEffect: false, fromNet: true, (Main.netMode == 2) ? whoAmI : 255);
+					*/
+					Main.npc[num211].StrikeNPC(hit, fromNet: true);
 				}
 				else {
 					Main.npc[num211].life = 0;
@@ -1559,7 +_,11 @@
 				}
 
 				if (Main.netMode == 2) {
+					/*
 					NetMessage.TrySendData(28, -1, whoAmI, null, num211, num212, num213, num214, b14);
+					*/
+					NetMessage.SendStrikeNPC(Main.npc[num211], hit, ignoreClient: whoAmI);
+
 					if (Main.npc[num211].life <= 0)
 						NetMessage.TrySendData(23, -1, -1, null, num211);
 
@@ -1630,16 +_,26 @@
 			case 32: {
 				int num27 = reader.ReadInt16();
 				int num28 = reader.ReadByte();
+				/*
 				int stack2 = reader.ReadInt16();
 				int prefixWeWant = reader.ReadByte();
 				int type3 = reader.ReadInt16();
+				*/
+				int stack2 = -1;
+				int prefixWeWant = -1;
+				int type3 = -1;
 				if (num27 >= 0 && num27 < 8000 && Main.chest[num27] != null) {
 					if (Main.chest[num27].item[num28] == null)
 						Main.chest[num27].item[num28] = new Item();
 
+					ItemIO.Receive(Main.chest[num27].item[num28], reader, readStack: true);
+					goto SkipVanillaItemSetup;
+
 					Main.chest[num27].item[num28].SetDefaults(type3);
 					Main.chest[num27].item[num28].Prefix(prefixWeWant);
 					Main.chest[num27].item[num28].stack = stack2;
+					SkipVanillaItemSetup:
+
 					if (Main.netMode == 2)
 						NetMessage.TrySendData(32, -1, whoAmI, null, num27, num28);
 				}
@@ -1653,7 +_,7 @@
 				int num43 = reader.ReadByte();
 				string name = string.Empty;
 				if (num43 != 0) {
-					if (num43 <= 20)
+					if (num43 <= Chest.MaxNameLength)
 						name = reader.ReadString();
 					else if (num43 != 255)
 						num43 = 0;
@@ -1707,24 +_,48 @@
 				if (Main.netMode == 2)
 					num5 = 0;
 
+
+				ushort modType = 0;
+				if (b2 >= 100)
+					modType = reader.ReadUInt16();
+
 				if (Main.netMode == 2) {
 					switch (b2) {
+						/*
 						case 0: {
+						*/
+						case byte _ when b2 % 100 == 0: {
+							if (modType == 0)
+								modType = TileID.Containers;
+
+							//TML: '21' -> 'modType'
-							int num8 = WorldGen.PlaceChest(num2, num3, 21, notNearOtherChests: false, num4);
+							int num8 = WorldGen.PlaceChest(num2, num3, modType, notNearOtherChests: false, num4);
+
 							if (num8 == -1) {
+								//TML: Added 'modType' at the end.
-								NetMessage.TrySendData(34, whoAmI, -1, null, b2, num2, num3, num4, num8);
+								NetMessage.TrySendData(34, whoAmI, -1, null, b2, num2, num3, num4, num8, modType);
+								/*
 								int itemDrop_Chests2 = WorldGen.GetItemDrop_Chests(num4, secondType: false);
+								*/
+								int itemDrop_Chests2 = b2 < 100 ? WorldGen.GetItemDrop_Chests(num4, secondType: false) : TileLoader.GetItemDropFromTypeAndStyle(modType, num4);
 								if (itemDrop_Chests2 > 0)
 									Item.NewItem(new EntitySource_TileBreak(num2, num3), num2 * 16, num3 * 16, 32, 32, itemDrop_Chests2, 1, noBroadcast: true);
 							}
 							else {
+								//TML: Added 'modType' at the end.
-								NetMessage.TrySendData(34, -1, -1, null, b2, num2, num3, num4, num8);
+								NetMessage.TrySendData(34, -1, -1, null, b2, num2, num3, num4, num8, modType);
 							}
 
 							break;
 						}
+						/*
 						case 1:
+						*/
+						case byte _ when b2 % 100 == 1:
+							/*
 							if (Main.tile[num2, num3].type == 21) {
+							*/
+							if (Main.tile[num2, num3].type == 21 || b2 == 101 && TileID.Sets.BasicChest[Main.tile[num2, num3].type]) {
 								Tile tile = Main.tile[num2, num3];
 								if (tile.frameX % 36 != 0)
 									num2--;
@@ -1742,20 +_,42 @@
 							goto default;
 						default:
 							switch (b2) {
+								/*
 								case 2: {
+								*/
+								case byte _ when b2 % 100 == 2: {
+									if (modType == 0)
+										modType = TileID.Dressers;
+
+									//TML: '88' -> 'modType'
-									int num6 = WorldGen.PlaceChest(num2, num3, 88, notNearOtherChests: false, num4);
+									int num6 = WorldGen.PlaceChest(num2, num3, modType, notNearOtherChests: false, num4);
+
 									if (num6 == -1) {
+										//TML: Added 'modType' at the end.
-										NetMessage.TrySendData(34, whoAmI, -1, null, b2, num2, num3, num4, num6);
+										NetMessage.TrySendData(34, whoAmI, -1, null, b2, num2, num3, num4, num6, modType);
+
+										/*
 										Item.NewItem(new EntitySource_TileBreak(num2, num3), num2 * 16, num3 * 16, 32, 32, WorldGen.GetItemDrop_Dressers(num4), 1, noBroadcast: true);
+										*/
+										int itemSpawn = b2 < 100 ? WorldGen.GetItemDrop_Dressers(num4) : TileLoader.GetItemDropFromTypeAndStyle(modType, num4);
+										if (itemSpawn > 0)
+											Item.NewItem(new EntitySource_TileBreak(num2, num3), num2 * 16, num3 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 									}
 									else {
+										//TML: Added 'modType' at the end.
-										NetMessage.TrySendData(34, -1, -1, null, b2, num2, num3, num4, num6);
+										NetMessage.TrySendData(34, -1, -1, null, b2, num2, num3, num4, num6, modType);
 									}
 
 									break;
 								}
+								/*
 								case 3:
+								*/
+								case byte _ when b2 % 100 == 3:
+									/*
 									if (Main.tile[num2, num3].type == 88) {
+									*/
+									if (Main.tile[num2, num3].type == 88 || (b2 == 103 && TileID.Sets.BasicDresser[Main.tile[num2, num3].type])) {
 										Tile tile2 = Main.tile[num2, num3];
 										num2 -= tile2.frameX % 54 / 18;
 										if (tile2.frameY % 36 != 0)
@@ -1810,21 +_,37 @@
 				}
 
 				switch (b2) {
+					/*
 					case 0:
+					*/
+					case byte _ when b2 % 100 == 0:
 						if (num5 == -1) {
 							WorldGen.KillTile(num2, num3);
 							break;
 						}
 						SoundEngine.PlaySound(0, num2 * 16, num3 * 16);
+
+						if (modType == 0)
+							modType = TileID.Containers;
+
+						//TML: '21' -> 'modType'.
-						WorldGen.PlaceChestDirect(num2, num3, 21, num4, num5);
+						WorldGen.PlaceChestDirect(num2, num3, modType, num4, num5);
 						break;
+					/*
 					case 2:
+					*/
+					case byte _ when b2 % 100 == 2:
 						if (num5 == -1) {
 							WorldGen.KillTile(num2, num3);
 							break;
 						}
 						SoundEngine.PlaySound(0, num2 * 16, num3 * 16);
+
+						if (modType == 0)
+							modType = TileID.Dressers;
+
+						//TML: '88' -> 'modType'.
-						WorldGen.PlaceDresserDirect(num2, num3, 88, num4, num5);
+						WorldGen.PlaceDresserDirect(num2, num3, modType, num4, num5);
 						break;
 					case 4:
 						if (num5 == -1) {
@@ -1869,6 +_,12 @@
 				player17.zone4 = reader.ReadByte();
 				player17.zone5 = reader.ReadByte();
 				player17.townNPCs = reader.ReadByte();
+
+				BiomeLoader.ReceiveCustomBiomes(player17, reader);
+
+				//TML: Recalculate the 'not in any biome' value.
+				player17.ZonePurity = player17.InZonePurity();
+
 				if (Main.netMode == 2) {
 					if (!flag14 && player17.zone5[0])
 						NPC.Spawner.SpawnFaelings(player17);
@@ -1894,6 +_,12 @@
 				if (Main.netMode == 2) {
 					if (reader.ReadString() == Netplay.ServerPassword) {
 						Netplay.Clients[whoAmI].State = 1;
+
+						if (ModNet.isModdedClient[whoAmI]) {
+							ModNet.SyncMods(whoAmI);
+							return;
+						}
+
 						NetMessage.TrySendData(3, whoAmI);
 					}
 					else {
@@ -1940,7 +_,12 @@
 				float itemRotation = reader.ReadSingle();
 				int itemAnimation = reader.ReadInt16();
 				player.itemRotation = itemRotation;
+				// TML changes how item synchronization works, making it based on remote players' already-synchronized input.
+				// Setting itemAnimation in this packet would interfere with that. #ItemTimeOnAllClients
+				/*
 				player.itemAnimation = itemAnimation;
+				*/
+				
 				player.channel = player.inventory[player.selectedItem].channel;
 				if (Main.netMode == 2)
 					NetMessage.TrySendData(41, -1, whoAmI, null, num33);
@@ -2065,7 +_,11 @@
 				if (Main.tile[num14, num15] == null)
 					Main.tile[num14, num15] = new Tile();
 
+				// Can't lock on struct. Lock does nothing anyway as this is the only place it's acquired
+				/*
 				lock (Main.tile[num14, num15]) {
+				*/
+				{
 					Main.tile[num14, num15].liquid = b3;
 					Main.tile[num14, num15].liquidType(liquidType);
 					if (Main.netMode == 2) {
@@ -2331,7 +_,7 @@
 				if (Main.netMode != 2)
 					break;
 
-				if (num260 >= 0 && num260 < NPCID.Count && NPCID.Sets.MPAllowedEnemies[num260]) {
+				if (num260 >= 0 && NPCID.Sets.MPAllowedEnemies[num260]) {
 					if (!NPC.AnyNPCs(num260))
 						NPC.SpawnOnPlayer(num259, num260);
 				}
@@ -2769,7 +_,8 @@
 						TileEntity.Remove(result3);
 				}
 				else {
-					TileEntity tileEntity = TileEntity.Read(reader, 319, networkSend: true);
+					//TML: Made 'lightSend' true.
+					TileEntity tileEntity = TileEntity.Read(reader, Main.curRelease, networkSend: true, lightSend: true);
 					tileEntity.ID = num150;
 					TileEntity.Add(tileEntity);
 				}
@@ -2843,10 +_,21 @@
 				if (Main.netMode == 2) {
 					short x = reader.ReadInt16();
 					int y = reader.ReadInt16();
+
+					Item item;
+					item = ItemIO.Receive(reader);
+					item.stack = reader.Read7BitEncodedInt();
+					/*
 					int type = reader.ReadInt16();
 					int prefix = reader.ReadByte();
 					int stack = reader.ReadInt16();
+					item = new Item(netid3, stack6, prefix3);
+					*/
+
+					/*
 					TEItemFrame.TryPlacing(x, y, type, prefix, stack);
+					*/
+					TEItemFrame.TryPlacing(x, y, item, 1);
 				}
 				break;
 			case 91: {
@@ -2855,6 +_,13 @@
 
 				int num262 = reader.ReadInt32();
 				int num263 = reader.ReadByte();
+
+				// If type corresponds to projectile (magic number 2), read another byte for owner #WorldUIAnchorProjectileSyncFix
+				if (num263 == 2) {
+					int owner = reader.ReadByte();
+					num263 |= owner << 8; // Reassign num263 for use in DeserializeNetAnchor - this is now "packedOwnerType"
+				}
+
 				if (num263 == 255) {
 					if (EmoteBubble.byID.ContainsKey(num262))
 						EmoteBubble.byID.Remove(num262);
@@ -2887,6 +_,7 @@
 					EmoteBubble.byID[num262].ID = num262;
 					EmoteBubble.byID[num262].metadata = metadata;
 					EmoteBubble.OnBubbleChange(num262);
+					EmoteBubbleLoader.OnSpawn(EmoteBubble.byID[num262]);
 					break;
 				}
 			}
@@ -2910,7 +_,7 @@
 			}
 			case 94: {
 				string text6 = reader.ReadString();
-				reader.ReadInt32();
+				int idincommand = reader.ReadInt32();
 				int num254 = (int)reader.ReadSingle();
 				reader.ReadSingle();
 				if (DebugOptions.enableDebugCommands) {
@@ -2921,6 +_,31 @@
 						DebugOptions.Shared_ServerPing = num254;
 						DebugNetworkStream.Latency = (uint)(num254 / 2);
 					}
+					if (text6 == "/quickload-clientprobe") {
+						if (Main.netMode == 2)
+							NetMessage.SendData(MessageID.DevCommands, -1, whoAmI, NetworkText.FromLiteral(text6), whoAmI);
+						else if (Netplay.ServerIPText == "127.0.0.1")
+							NetMessage.SendData(MessageID.DevCommands, -1, -1, NetworkText.FromLiteral("/quickload-clientresp " + QuickLoad.Serialize(new QuickLoad.JoinWorld().WithCurrentState())), idincommand);
+
+						break;
+					}
+					if (text6.StartsWith("/quickload-clientresp ")) {
+						if (Main.netMode == 2) {
+							NetMessage.SendData(MessageID.DevCommands, idincommand, -1, NetworkText.FromLiteral(text6));
+							break;
+						}
+
+						var json = text6.Substring("/quickload-clientresp ".Length);
+						var config = (QuickLoad.JoinWorld)QuickLoad.Deserialize(json);
+						QuickLoad hostConfig;
+						if (QuickLoad.TryRead(out hostConfig) && hostConfig is QuickLoad.JoinWorld) {
+							var hostJoinConfig = (QuickLoad.JoinWorld)hostConfig;
+							hostJoinConfig.ExtraClients.Add(config);
+							QuickLoad.Set(hostJoinConfig);
+							ChatHelper.DisplayMessage(NetworkText.FromLiteral("/quickload added " + Path.GetFileName(config.PlayerPath)), new Color(250, 250, 0), Main.maxPlayers);
+						}
+						break;
+					}
 				}
 
 				break;
@@ -3227,7 +_,16 @@
 				break;
 			case 117: {
 				int num24 = reader.ReadByte();
+				bool directHurt = false;
+				if (num24 == 255) {
+					num24 = reader.ReadByte();
+					directHurt = true;
+				}
+
 				if (Main.netMode != 2 || whoAmI == num24 || (Main.player[num24].hostile && Main.player[whoAmI].hostile)) {
+					if (directHurt)
+						goto ReadDirectHurt;
+
 					PlayerDeathReason playerDeathReason2 = PlayerDeathReason.FromReader(reader);
 					int damage = reader.ReadInt16();
 					int num25 = reader.ReadByte() - 1;
@@ -3238,6 +_,26 @@
 					Main.player[num24].Hurt(playerDeathReason2, damage, num25, pvp2, quiet: true, flag, num26);
 					if (Main.netMode == 2)
 						NetMessage.SendPlayerHurt(num24, playerDeathReason2, damage, num25, flag, pvp2, num26, -1, whoAmI);
+
+					break;
+
+					ReadDirectHurt:
+					BitsByte pack = reader.ReadByte();
+					var args = new Player.HurtInfo {
+						DamageSource = PlayerDeathReason.FromReader(reader),
+						PvP = pack[0],
+						CooldownCounter = reader.ReadSByte(),
+						Dodgeable = pack[1],
+						SourceDamage = reader.Read7BitEncodedInt(),
+						Damage = reader.Read7BitEncodedInt(),
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle(),
+						DustDisabled = pack[2],
+						SoundDisabled = pack[3]
+					};
+					Main.player[num24].Hurt(args, quiet: true);
+					if (Main.netMode == 2)
+						NetMessage.SendPlayerHurt(num24, args, ignoreClient: whoAmI);
 				}
 
 				break;
@@ -3263,7 +_,7 @@
 					num271 = whoAmI;
 
 				int num272 = reader.ReadByte();
-				if (num272 >= 0 && num272 < EmoteID.Count && Main.netMode == 2) {
+				if (num272 >= 0 && num272 < EmoteBubbleLoader.EmoteBubbleCount && Main.netMode == 2) {
 					EmoteBubble.NewBubble(num272, new WorldUIAnchor(Main.player[num271]), 360);
 					EmoteBubble.CheckForNPCsToReactToEmoteBubble(num272, Main.player[num271]);
 				}
@@ -3322,10 +_,20 @@
 				if (Main.netMode == 2) {
 					short x14 = reader.ReadInt16();
 					int y14 = reader.ReadInt16();
+
+					Item item;
+					item = ItemIO.Receive(reader, readStack: true);
+					/*
 					int type19 = reader.ReadInt16();
 					int prefix5 = reader.ReadByte();
 					int stack8 = reader.ReadInt16();
+					item = new Item(netid2, stack4, prefix2);
+					*/
+
+					/*
 					TEWeaponsRack.TryPlacing(x14, y14, type19, prefix5, stack8);
+					*/
+					TEWeaponsRack.TryPlacing(x14, y14, item, item.stack);
 				}
 				break;
 			case 124: {
@@ -3453,25 +_,48 @@
 					ushort key = reader.ReadUInt16();
 					LegacySoundStyle legacySoundStyle = SoundID.SoundByIndex[key];
 					BitsByte bitsByte11 = reader.ReadByte();
+					/*
 					int num141 = -1;
 					float num142 = 1f;
 					float num143 = 0f;
 					SoundEngine.PlaySound(Style: (!bitsByte11[0]) ? legacySoundStyle.Style : reader.ReadInt32(), volumeScale: (!bitsByte11[1]) ? legacySoundStyle.Volume : MathHelper.Clamp(reader.ReadSingle(), 0f, 1f), pitchOffset: (!bitsByte11[2]) ? legacySoundStyle.GetRandomPitch() : MathHelper.Clamp(reader.ReadSingle(), -1f, 1f), type: legacySoundStyle.SoundId, x: point2.X, y: point2.Y);
+					*/
+
+					if (bitsByte11[0])
+						legacySoundStyle.Variants = new int[] { reader.ReadInt32() };
+
+					if (bitsByte11[1])
+						legacySoundStyle.Volume = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					if (bitsByte11[2])
+						legacySoundStyle.Pitch = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					SoundEngine.PlaySound(legacySoundStyle, point2.ToVector2());
 				}
 				break;
 			case 133:
 				if (Main.netMode == 2) {
 					short x9 = reader.ReadInt16();
 					int y9 = reader.ReadInt16();
+
+					Item item;
+					item = ItemIO.Receive(reader, readStack: true);
+					/*
 					int type13 = reader.ReadInt16();
 					int prefix3 = reader.ReadByte();
 					int stack6 = reader.ReadInt16();
+					item = new Item(netid, stack, prefix);
+					*/
+
+					/*
 					TEFoodPlatter.TryPlacing(x9, y9, type13, prefix3, stack6);
+					*/
+					TEFoodPlatter.TryPlacing(x9, y9, item, item.stack);
 				}
 				break;
 			case 134: {
 				int num94 = reader.ReadByte();
-				int ladyBugLuckTimeLeft = reader.ReadInt32();
+				double ladyBugLuckTimeLeft = reader.ReadDouble();
 				float torchLuck = reader.ReadSingle();
 				byte luckPotion = reader.ReadByte();
 				bool hasGardenGnomeNearby = reader.ReadBoolean();
@@ -3612,10 +_,15 @@
 				if (Main.netMode == 2) {
 					short x4 = reader.ReadInt16();
 					int y4 = reader.ReadInt16();
+					Item item = ItemIO.Receive(reader);
+					int stack = reader.Read7BitEncodedInt();
+					/*
 					int type10 = reader.ReadInt16();
 					int prefix2 = reader.ReadByte();
 					int stack4 = reader.ReadInt16();
 					TEDeadCellsDisplayJar.TryPlacing(x4, y4, type10, prefix2, stack4);
+					*/
+					TEDeadCellsDisplayJar.TryPlacing(x4, y4, item, stack);
 				}
 				break;
 			case 150: {
@@ -3652,7 +_,10 @@
 					Player player3 = Main.player[num39];
 					Item item2 = player3.inventory[player3.selectedItem];
 					if (item2.UseSound != null)
+						/*
 						SoundEngine.PlaySound(item2.UseSound, player3.Center, item2.useSoundPitch);
+						*/
+						SoundEngine.PlaySound(item2.UseSound, player3.Center);
 				}
 
 				break;
@@ -3713,6 +_,35 @@
 				Main.player[whoAmI].host = !string.IsNullOrWhiteSpace(Netplay.HostToken) && Netplay.HostToken == text;
 				break;
 			}
+			case MessageID.InGameChangeConfig:
+				ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+				return;
+			case MessageID.ModPacket:
+				ModNet.HandleModPacket(reader, whoAmI, length);
+				return;
+			case MessageID.SyncMods:
+				if (Main.netMode == 1) {
+					ModNet.SyncClientMods(reader);
+				}
+				else {
+					ModNet.SendNetIDs(whoAmI);
+					NetMessage.SendData(3, whoAmI);
+				}
+				return;
+			case MessageID.ModFile:
+				if (Main.netMode == 1)
+					ModNet.ReceiveMod(reader);
+				else
+					ModNet.SendMod(reader.ReadString(), whoAmI);
+				return;
+			case MessageID.KeepAliveDuringModReload:
+				string keepAlive = "Keep Alive During Mod Reload";
+				ModNet.Log(whoAmI, keepAlive);
+				// Logging.ServerConsoleLine doesn't log client info and ModNet.Log doesn't output to console, so manually replicate here.
+				var client = Netplay.Clients[whoAmI];
+				Console.WriteLine($"[{whoAmI}][{client.Socket?.GetRemoteAddress()?.GetFriendlyName()} ({client.Name})]: {keepAlive}");
+				// Does nothing. "Netplay.Clients[whoAmI].TimeOutTimer = 0;" at top of method should keep the connection alive
+				return;
 			default:
 				if (Main.netMode == 2 && Netplay.Clients[whoAmI].State == 0)
 					NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
