--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -18,7 +_,10 @@
 using Terraria.Localization;
 using Terraria.Net;
 using Terraria.Testing;
 using Terraria.UI;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 
 namespace Terraria
 {
@@ -83,10 +_,8 @@
 			int num = 0;
 			num = start + 1;
 			b = (byte)(messageType = readBuffer[start]);
-			if (b >= 140)
-				return;
-
-			Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
+			if (b < LegacyNetDiagnosticsUI.maxMsg)
+				Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 			if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
 				Netplay.Connection.StatusCount++;
 
@@ -104,14 +_,17 @@
 				return;
 			}
 
-			if (Main.netMode == 2 && Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68)
+			if (Main.netMode == 2 && Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b < 250)
 				NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 			if (reader == null)
 				ResetReader();
 
 			reader.BaseStream.Position = num;
 			NPCSpawnParams spawnparams;
+			if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+				return;
+
 			switch (b) {
 				case 15:
 				case 25:
@@ -132,9 +_,15 @@
 						if (Netplay.Clients[whoAmI].State != 0)
 							break;
 
-						if (reader.ReadString() == "Terraria" + 227) {
+						string clientVersion = reader.ReadString();
+						ModNet.isModdedClient[whoAmI] = clientVersion == ModLoader.ModLoader.versionedName;
+						if (ModNet.isModdedClient[whoAmI] || clientVersion == "Terraria" + 194 && ModNet.AllowVanillaClients) {
 							if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 								Netplay.Clients[whoAmI].State = 1;
+								if (ModNet.isModdedClient[whoAmI]) {
+									ModNet.SyncMods(whoAmI);
+									return;
+								}
 								NetMessage.TrySendData(3, whoAmI);
 							}
 							else {
@@ -143,7 +_,10 @@
 							}
 						}
 						else {
-							NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+							if (clientVersion.StartsWith("tModLoader"))
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n(You are on {clientVersion}, server is on {ModLoader.ModLoader.versionedName})"));
+							else
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + "\n(You cannot connect to a tModLoader Server with an unmodded client.)"));
 						}
 					}
 					break;
@@ -209,6 +_,7 @@
 							NetMessage.TrySendData(5, -1, -1, null, num127, 58 + player7.armor.Length + player7.dye.Length + player7.miscEquips.Length + player7.miscDyes.Length + player7.bank.item.Length + player7.bank2.item.Length + player7.bank3.item.Length + 2 + num136, (int)player7.bank4.item[num136].prefix);
 						}
 
+						PlayerHooks.SyncPlayer(player7, -1, -1, true);
 						NetMessage.TrySendData(6);
 						if (Netplay.Connection.State == 2)
 							Netplay.Connection.State = 3;
@@ -371,8 +_,9 @@
 								player14.trashItem.netDefaults(type14);
 								player14.trashItem.stack = stack6;
 								player14.trashItem.Prefix(num245);
 								if (num243 == Main.myPlayer && !Main.ServerSideCharacter)
 									Main.clientPlayer.trashItem = player14.trashItem.Clone();
+								ItemIO.ReceiveModData(player14.trashItem, reader);
 							}
 							else if (num244 <= 58) {
 								int type15 = array3[num246].type;
@@ -381,6 +_,7 @@
 								array3[num246].netDefaults(type14);
 								array3[num246].stack = stack6;
 								array3[num246].Prefix(num245);
+								ItemIO.ReceiveModData(array3[num246], reader);
 								if (num243 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num246] = array3[num246].Clone();
 
@@ -400,8 +_,9 @@
 								array3[num246].netDefaults(type14);
 								array3[num246].stack = stack6;
 								array3[num246].Prefix(num245);
 								if (num243 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num246] = array3[num246].Clone();
+								ItemIO.ReceiveModData(array3[num246], reader);
 							}
 
 							if (Main.netMode == 2 && num243 == whoAmI && num244 <= 58 + player14.armor.Length + player14.dye.Length + player14.miscEquips.Length + player14.miscDyes.Length)
@@ -553,6 +_,9 @@
 							Main.StopSlimeRain();
 
 						Main.invasionType = reader.ReadSByte();
+						if (!ModNet.AllowVanillaClients)
+							WorldIO.ReceiveModData(reader);
+
 						Main.LobbyId = reader.ReadUInt64();
 						Sandstorm.IntendedSeverity = reader.ReadSingle();
 						if (Netplay.Connection.State == 3) {
@@ -686,6 +_,9 @@
 							NetMessage.TrySendData(83, whoAmI, -1, null, num122);
 						}
 
+						for (int type = NPCID.Count; type < NPCLoader.NPCCount; type++)
+							NetMessage.SendData(83, whoAmI, -1, null, type, 0f, 0f, 0f, 0, 0, 0);
+
 						NetMessage.TrySendData(49, whoAmI);
 						NetMessage.TrySendData(57, whoAmI);
 						NetMessage.TrySendData(7, whoAmI);
@@ -1133,9 +_,9 @@
 
 							int num239 = num236;
 							Item item3 = Main.item[num239];
 							ItemSyncPersistentStats itemSyncPersistentStats = default(ItemSyncPersistentStats);
 							itemSyncPersistentStats.CopyFrom(item3);
-							bool newAndShiny = (item3.newAndShiny || item3.netID != num238) && ItemSlot.Options.HighlightNewItems && (num238 < 0 || num238 >= 5043 || !ItemID.Sets.NeverAppearsAsNewInInventory[num238]);
+							bool newAndShiny = (item3.newAndShiny || item3.netID != num238) && ItemSlot.Options.HighlightNewItems && (num238 < 0 || !ItemID.Sets.NeverAppearsAsNewInInventory[num238]);
 							item3.netDefaults(num238);
 							item3.newAndShiny = newAndShiny;
 							item3.Prefix(pre2);
@@ -1143,6 +_,7 @@
 							item3.position = position3;
 							item3.velocity = velocity5;
 							item3.active = true;
+							ItemIO.ReceiveModData(item3, reader);
 							if (b == 90) {
 								item3.instanced = true;
 								item3.playerIndexTheItemIsReservedFor = Main.myPlayer;
@@ -1182,6 +_,7 @@
 							obj8.position = position3;
 							obj8.velocity = velocity5;
 							obj8.active = true;
+							ItemIO.ReceiveModData(obj8, reader);
 							obj8.playerIndexTheItemIsReservedFor = Main.myPlayer;
 							if (flag13) {
 								NetMessage.TrySendData(21, -1, -1, null, num236);
@@ -1313,9 +_,10 @@
 						if (num55 == 245)
 							NPC.golemBoss = num52;
 
-						if (nPC2.type >= 0 && nPC2.type < 663 && Main.npcCatchable[nPC2.type])
+						if (nPC2.type >= 0 && nPC2.type < NPCLoader.NPCCount && Main.npcCatchable[nPC2.type])
 							nPC2.releaseOwner = reader.ReadByte();
 
+						NPCLoader.ReceiveExtraAI(nPC2, reader);
 						break;
 					}
 				case 24: {
@@ -1354,7 +_,7 @@
 						int num153 = bitsByte8[7] ? reader.ReadInt16() : (-1);
 						if (num153 >= 1000)
 							num153 = -1;
-
+						byte[] extraAI = ProjectileLoader.ReadExtraAI(reader, bitsByte8);
 						if (Main.netMode == 2) {
 							num150 = whoAmI;
 							if (Main.projHostile[num151])
@@ -1405,6 +_,7 @@
 							Main.projectileIdentity[num150, num153] = num154;
 						}
 
+						ProjectileLoader.ReceiveExtraAI(projectile, extraAI);
 						projectile.ProjectileFixDesperation();
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(27, -1, whoAmI, null, num154);
@@ -1413,7 +_,7 @@
 					}
 				case 28: {
 						int num3 = reader.ReadInt16();
-						int num4 = reader.ReadInt16();
+						int num4 = ModNet.AllowVanillaClients ? reader.ReadInt16() : reader.ReadInt32();
 						float num5 = reader.ReadSingle();
 						int num6 = reader.ReadByte() - 1;
 						byte b2 = reader.ReadByte();
@@ -1528,8 +_,9 @@
 							Main.chest[num260].item[num261].netDefaults(type17);
 							Main.chest[num260].item[num261].Prefix(pre3);
 							Main.chest[num260].item[num261].stack = stack8;
+							ItemIO.ReceiveModData(Main.chest[num260].item[num261], reader);
 							Recipe.FindRecipes(canDelayCheck: true);
 						}
 
 						break;
 					}
@@ -1585,6 +_,7 @@
 						break;
 					}
 				case 34: {
+						// TODO, check that this didn't get messed up...why % 100?
 						byte b3 = reader.ReadByte();
 						int num15 = reader.ReadInt16();
 						int num16 = reader.ReadInt16();
@@ -1593,22 +_,32 @@
 						if (Main.netMode == 2)
 							num18 = 0;
 
+
+						ushort modType = 0;
+						if (b3 >= 100)
+							modType = reader.ReadUInt16();
+
 						if (Main.netMode == 2) {
 							switch (b3) {
-								case 0: {
-										int num19 = WorldGen.PlaceChest(num15, num16, 21, notNearOtherChests: false, num17);
+								case byte _ when b3 % 100 == 0: {
+										if (modType == 0)
+											modType = TileID.Containers;
+
+										int num19 = WorldGen.PlaceChest(num15, num26, modType, notNearOtherChests: false, num17);
 										if (num19 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num19);
-											Item.NewItem(num15 * 16, num16 * 16, 32, 32, Chest.chestItemSpawn[num17], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num19, modType);
+											int itemSpawn = b3 < 100 ? Chest.chestItemSpawn[num17] : TileLoader.GetTile(modType).chestDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num15 * 16, num16 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num19);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num19, modType);
 										}
 
 										return;
 									}
 								case 1:
-									if (Main.tile[num15, num16].type == 21) {
+									if (Main.tile[num15, num16].type == 21 || b3 == 101 && TileID.Sets.BasicChest[Main.tile[num15, num16].type]) {
 										Tile tile = Main.tile[num15, num16];
 										if (tile.frameX % 36 != 0)
 											num15--;
@@ -1627,20 +_,25 @@
 							}
 
 							switch (b3) {
-								case 2: {
-										int num20 = WorldGen.PlaceChest(num15, num16, 88, notNearOtherChests: false, num17);
+								case byte _ when b3 % 100 == 2: {
+										if (modType == 0)
+											modType = TileID.Dressers;
+
+										int num20 = WorldGen.PlaceChest(num15, num26, modType, notNearOtherChests: false, num17);
 										if (num20 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num20);
-											Item.NewItem(num15 * 16, num16 * 16, 32, 32, Chest.dresserItemSpawn[num17], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num20, modType);
+											int itemSpawn = b3 < 100 ? Chest.dresserItemSpawn[num17] : TileLoader.GetTile(modType).dresserDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num15 * 16, num16 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num20);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num20, modType);
 										}
 
 										return;
 									}
 								case 3:
-									if (Main.tile[num15, num16].type == 88) {
+									if (Main.tile[num15, num16].type == 88 || b3 == 103 && TileLoader.IsDresser(Main.tile[num15, num16].type)) {
 										Tile tile2 = Main.tile[num15, num16];
 										num15 -= tile2.frameX % 54 / 18;
 										if (tile2.frameY % 36 != 0)
@@ -1690,21 +_,25 @@
 						}
 
 						switch (b3) {
-							case 0:
+							case byte _ when b3 % 100 == 0:
 								if (num18 == -1) {
 									WorldGen.KillTile(num15, num16);
 									break;
 								}
 								SoundEngine.PlaySound(0, num15 * 16, num16 * 16);
+								if (modType == 0)
+									modType = TileID.Containers;
-								WorldGen.PlaceChestDirect(num15, num16, 21, num17, num18);
+								WorldGen.PlaceChestDirect(num15, num16, modType, num17, num18);
 								break;
-							case 2:
+							case byte _ when b3 % 100 == 2:
 								if (num18 == -1) {
 									WorldGen.KillTile(num15, num16);
 									break;
 								}
 								SoundEngine.PlaySound(0, num15 * 16, num16 * 16);
+								if (modType == 0)
+									modType = TileID.Dressers;
-								WorldGen.PlaceDresserDirect(num15, num16, 88, num17, num18);
+								WorldGen.PlaceDresserDirect(num15, num16, modType, num17, num18);
 								break;
 							case 4:
 								if (num18 == -1) {
@@ -1746,6 +_,9 @@
 						obj6.zone2 = reader.ReadByte();
 						obj6.zone3 = reader.ReadByte();
 						obj6.zone4 = reader.ReadByte();
+						if (!ModNet.AllowVanillaClients)
+							PlayerHooks.ReceiveCustomBiomes(obj6, reader);
+
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(36, -1, whoAmI, null, num225);
 
@@ -1767,6 +_,11 @@
 					if (Main.netMode == 2) {
 						if (reader.ReadString() == Netplay.ServerPassword) {
 							Netplay.Clients[whoAmI].State = 1;
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -1952,7 +_,7 @@
 							break;
 
 						Player player4 = Main.player[num47];
-						for (int num48 = 0; num48 < 22; num48++) {
+						for (int num48 = 0; num48 < Player.MaxBuffs; num48++) {
 							player4.buffType[num48] = reader.ReadUInt16();
 							if (player4.buffType[num48] > 0)
 								player4.buffTime[num48] = 60;
@@ -2440,6 +_,7 @@
 					break;
 				case 74:
 					if (Main.netMode == 1) {
+						Netplay.syncingWorld = false;
 						Main.anglerQuest = reader.ReadByte();
 						Main.anglerQuestFinished = reader.ReadBoolean();
 					}
@@ -2529,7 +_,7 @@
 					if (Main.netMode == 1) {
 						int num74 = reader.ReadInt16();
 						int num75 = reader.ReadInt32();
-						if (num74 >= 0 && num74 < 289)
+						if (num74 >= 0)// && num74 < 289) This prevented mod BannerIds from syncing.
 							NPC.killCount[num74] = num75;
 					}
 					break;
@@ -3312,7 +_,28 @@
 						bool flag = reader.ReadBoolean();
 						Main.countsAsHostForGameplay[num2] = flag;
 					}
 					break;
+				case MessageID.InGameChangeConfig:
+					ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+					return;
+				case MessageID.ModPacket:
+					ModNet.HandleModPacket(reader, whoAmI, length);
+					return;
+				case MessageID.SyncMods:
+					if (Main.netMode == 1) {
+						ModNet.SyncClientMods(reader);
+					}
+					else {
+						ModNet.SendNetIDs(whoAmI);
+						NetMessage.SendData(3, whoAmI);
+					}
+					return;
+				case MessageID.ModFile:
+					if (Main.netMode == 1)
+						ModNet.ReceiveMod(reader);
+					else
+						ModNet.SendMod(reader.ReadString(), whoAmI);
+					return;
 			}
 		}
 	}
