--- src/Terraria/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -19,6 +_,9 @@
 using Terraria.Net;
 using Terraria.Testing;
 using Terraria.UI;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 
 namespace Terraria
 {
@@ -83,10 +_,8 @@
 			int num = 0;
 			num = start + 1;
 			b = (byte)(messageType = readBuffer[start]);
-			if (b >= 140)
-				return;
-
-			Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
+			if (b < LegacyNetDiagnosticsUI.maxMsg)
+				Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 			if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
 				Netplay.Connection.StatusCount++;
 
@@ -105,7 +_,7 @@
 			}
 
 			if (Main.netMode == 2) {
-				if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68)
+				if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b < 250)
 					NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 				if (Netplay.Clients[whoAmI].State == 0 && b != 1)
@@ -117,6 +_,9 @@
 
 			reader.BaseStream.Position = num;
 			NPCSpawnParams spawnparams;
+			if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+				return;
+
 			switch (b) {
 				case 15:
 				case 25:
@@ -135,9 +_,15 @@
 						if (Netplay.Clients[whoAmI].State != 0)
 							break;
 
-						if (reader.ReadString() == "Terraria" + 230) {
+						string clientVersion = reader.ReadString();
+						ModNet.isModdedClient[whoAmI] = clientVersion == ModLoader.ModLoader.versionedName;
+						if (ModNet.isModdedClient[whoAmI] || clientVersion == "Terraria" + 230 && ModNet.AllowVanillaClients) {
 							if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 								Netplay.Clients[whoAmI].State = 1;
+								if (ModNet.isModdedClient[whoAmI]) {
+									ModNet.SyncMods(whoAmI);
+									return;
+								}
 								NetMessage.TrySendData(3, whoAmI);
 							}
 							else {
@@ -146,7 +_,10 @@
 							}
 						}
 						else {
-							NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+							if (clientVersion.StartsWith("tModLoader"))
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n(You are on {clientVersion}, server is on {ModLoader.ModLoader.versionedName})"));
+							else
+								NetMessage.SendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + "\n(You cannot connect to a tModLoader Server with an unmodded client.)"));
 						}
 					}
 					break;
@@ -212,6 +_,7 @@
 							NetMessage.TrySendData(5, -1, -1, null, num222, 58 + player11.armor.Length + player11.dye.Length + player11.miscEquips.Length + player11.miscDyes.Length + player11.bank.item.Length + player11.bank2.item.Length + player11.bank3.item.Length + 2 + num231, (int)player11.bank4.item[num231].prefix);
 						}
 
+						PlayerHooks.SyncPlayer(player11, -1, -1, true);
 						NetMessage.TrySendData(6);
 						if (Netplay.Connection.State == 2)
 							Netplay.Connection.State = 3;
@@ -379,6 +_,7 @@
 								player13.trashItem.Prefix(num241);
 								if (num239 == Main.myPlayer && !Main.ServerSideCharacter)
 									Main.clientPlayer.trashItem = player13.trashItem.Clone();
+								ItemIO.ReceiveModData(player13.trashItem, reader);
 							}
 							else if (num240 <= 58) {
 								int type15 = array3[num242].type;
@@ -387,6 +_,7 @@
 								array3[num242].netDefaults(type14);
 								array3[num242].stack = stack6;
 								array3[num242].Prefix(num241);
+								ItemIO.ReceiveModData(array3[num242], reader);
 								if (num239 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num242] = array3[num242].Clone();
 
@@ -408,6 +_,7 @@
 								array3[num242].Prefix(num241);
 								if (num239 == Main.myPlayer && !Main.ServerSideCharacter)
 									array4[num242] = array3[num242].Clone();
+								ItemIO.ReceiveModData(array3[num242], reader);
 							}
 
 							if (Main.netMode == 2 && num239 == whoAmI && num240 <= 58 + player13.armor.Length + player13.dye.Length + player13.miscEquips.Length + player13.miscDyes.Length)
@@ -561,6 +_,8 @@
 						Main.invasionType = reader.ReadSByte();
 						Main.LobbyId = reader.ReadUInt64();
 						Sandstorm.IntendedSeverity = reader.ReadSingle();
+						if (!ModNet.AllowVanillaClients && Netplay.Connection.State > 4)
+							WorldIO.ReceiveModData(reader);
 						if (Netplay.Connection.State == 3) {
 							Main.windSpeedCurrent = Main.windSpeedTarget;
 							Netplay.Connection.State = 4;
@@ -692,6 +_,9 @@
 							NetMessage.TrySendData(83, whoAmI, -1, null, num168);
 						}
 
+						for (int type = NPCID.Count; type < NPCLoader.NPCCount; type++)
+							NetMessage.SendData(83, whoAmI, -1, null, type, 0f, 0f, 0f, 0, 0, 0);
+
 						NetMessage.TrySendData(49, whoAmI);
 						NetMessage.TrySendData(57, whoAmI);
 						NetMessage.TrySendData(7, whoAmI);
@@ -1141,7 +_,7 @@
 							Item item2 = Main.item[num72];
 							ItemSyncPersistentStats itemSyncPersistentStats = default(ItemSyncPersistentStats);
 							itemSyncPersistentStats.CopyFrom(item2);
-							bool newAndShiny = (item2.newAndShiny || item2.netID != num71) && ItemSlot.Options.HighlightNewItems && (num71 < 0 || num71 >= 5045 || !ItemID.Sets.NeverAppearsAsNewInInventory[num71]);
+							bool newAndShiny = (item2.newAndShiny || item2.netID != num71) && ItemSlot.Options.HighlightNewItems && (num71 < 0 || !ItemID.Sets.NeverAppearsAsNewInInventory[num71]);
 							item2.netDefaults(num71);
 							item2.newAndShiny = newAndShiny;
 							item2.Prefix(pre2);
@@ -1149,6 +_,7 @@
 							item2.position = position2;
 							item2.velocity = velocity;
 							item2.active = true;
+							ItemIO.ReceiveModData(item2, reader);
 							if (b == 90) {
 								item2.instanced = true;
 								item2.playerIndexTheItemIsReservedFor = Main.myPlayer;
@@ -1188,6 +_,7 @@
 							obj3.position = position2;
 							obj3.velocity = velocity;
 							obj3.active = true;
+							ItemIO.ReceiveModData(obj3, reader);
 							obj3.playerIndexTheItemIsReservedFor = Main.myPlayer;
 							if (flag5) {
 								NetMessage.TrySendData(21, -1, -1, null, num69);
@@ -1319,9 +_,10 @@
 						if (num83 == 245)
 							NPC.golemBoss = num80;
 
-						if (nPC3.type >= 0 && nPC3.type < 663 && Main.npcCatchable[nPC3.type])
+						if (nPC3.type >= 0 && nPC3.type < NPCLoader.NPCCount && Main.npcCatchable[nPC3.type])
 							nPC3.releaseOwner = reader.ReadByte();
 
+						NPCLoader.ReceiveExtraAI(nPC3, reader);
 						break;
 					}
 				case 24: {
@@ -1360,7 +_,7 @@
 						int num125 = bitsByte8[7] ? reader.ReadInt16() : (-1);
 						if (num125 >= 1000)
 							num125 = -1;
-
+						byte[] extraAI = ProjectileLoader.ReadExtraAI(reader, bitsByte8);
 						if (Main.netMode == 2) {
 							if (num123 == 949) {
 								num122 = 255;
@@ -1416,6 +_,7 @@
 							Main.projectileIdentity[num122, num125] = num126;
 						}
 
+						ProjectileLoader.ReceiveExtraAI(projectile, extraAI);
 						projectile.ProjectileFixDesperation();
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(27, -1, whoAmI, null, num126);
@@ -1424,7 +_,7 @@
 					}
 				case 28: {
 						int num3 = reader.ReadInt16();
-						int num4 = reader.ReadInt16();
+						int num4 = ModNet.AllowVanillaClients ? reader.ReadInt16() : reader.ReadInt32();
 						float num5 = reader.ReadSingle();
 						int num6 = reader.ReadByte() - 1;
 						byte b2 = reader.ReadByte();
@@ -1539,6 +_,7 @@
 							Main.chest[num260].item[num261].netDefaults(type17);
 							Main.chest[num260].item[num261].Prefix(pre3);
 							Main.chest[num260].item[num261].stack = stack8;
+							ItemIO.ReceiveModData(Main.chest[num260].item[num261], reader);
 							Recipe.FindRecipes(canDelayCheck: true);
 						}
 
@@ -1596,6 +_,7 @@
 						break;
 					}
 				case 34: {
+						// TODO, check that this didn't get messed up...why % 100?
 						byte b3 = reader.ReadByte();
 						int num15 = reader.ReadInt16();
 						int num16 = reader.ReadInt16();
@@ -1604,22 +_,32 @@
 						if (Main.netMode == 2)
 							num18 = 0;
 
+
+						ushort modType = 0;
+						if (b3 >= 100)
+							modType = reader.ReadUInt16();
+
 						if (Main.netMode == 2) {
 							switch (b3) {
-								case 0: {
-										int num19 = WorldGen.PlaceChest(num15, num16, 21, notNearOtherChests: false, num17);
+								case byte _ when b3 % 100 == 0: {
+										if (modType == 0)
+											modType = TileID.Containers;
+
+										int num19 = WorldGen.PlaceChest(num15, num16, modType, notNearOtherChests: false, num17);
 										if (num19 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num19);
-											Item.NewItem(num15 * 16, num16 * 16, 32, 32, Chest.chestItemSpawn[num17], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num19, modType);
+											int itemSpawn = b3 < 100 ? Chest.chestItemSpawn[num17] : TileLoader.GetTile(modType).chestDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num15 * 16, num16 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num19);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num19, modType);
 										}
 
 										return;
 									}
 								case 1:
-									if (Main.tile[num15, num16].type == 21) {
+									if (Main.tile[num15, num16].type == 21 || b3 == 101 && TileID.Sets.BasicChest[Main.tile[num15, num16].type]) {
 										Tile tile = Main.tile[num15, num16];
 										if (tile.frameX % 36 != 0)
 											num15--;
@@ -1638,20 +_,25 @@
 							}
 
 							switch (b3) {
-								case 2: {
-										int num20 = WorldGen.PlaceChest(num15, num16, 88, notNearOtherChests: false, num17);
+								case byte _ when b3 % 100 == 2: {
+										if (modType == 0)
+											modType = TileID.Dressers;
+
+										int num20 = WorldGen.PlaceChest(num15, num16, modType, notNearOtherChests: false, num17);
 										if (num20 == -1) {
-											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num20);
-											Item.NewItem(num15 * 16, num16 * 16, 32, 32, Chest.dresserItemSpawn[num17], 1, noBroadcast: true);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b3, num15, num16, num17, num20, modType);
+											int itemSpawn = b3 < 100 ? Chest.dresserItemSpawn[num17] : TileLoader.GetTile(modType).dresserDrop;
+											if (itemSpawn > 0)
+												Item.NewItem(num15 * 16, num16 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
-											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num20);
+											NetMessage.TrySendData(34, -1, -1, null, b3, num15, num16, num17, num20, modType);
 										}
 
 										return;
 									}
 								case 3:
-									if (Main.tile[num15, num16].type == 88) {
+									if (Main.tile[num15, num16].type == 88 || b3 == 103 && TileID.Sets.BasicDresser[Main.tile[num15, num16].type]) {
 										Tile tile2 = Main.tile[num15, num16];
 										num15 -= tile2.frameX % 54 / 18;
 										if (tile2.frameY % 36 != 0)
@@ -1701,21 +_,25 @@
 						}
 
 						switch (b3) {
-							case 0:
+							case byte _ when b3 % 100 == 0:
 								if (num18 == -1) {
 									WorldGen.KillTile(num15, num16);
 									break;
 								}
 								SoundEngine.PlaySound(0, num15 * 16, num16 * 16);
+								if (modType == 0)
+									modType = TileID.Containers;
-								WorldGen.PlaceChestDirect(num15, num16, 21, num17, num18);
+								WorldGen.PlaceChestDirect(num15, num16, modType, num17, num18);
 								break;
-							case 2:
+							case byte _ when b3 % 100 == 2:
 								if (num18 == -1) {
 									WorldGen.KillTile(num15, num16);
 									break;
 								}
 								SoundEngine.PlaySound(0, num15 * 16, num16 * 16);
+								if (modType == 0)
+									modType = TileID.Dressers;
-								WorldGen.PlaceDresserDirect(num15, num16, 88, num17, num18);
+								WorldGen.PlaceDresserDirect(num15, num16, modType, num17, num18);
 								break;
 							case 4:
 								if (num18 == -1) {
@@ -1757,6 +_,9 @@
 						obj7.zone2 = reader.ReadByte();
 						obj7.zone3 = reader.ReadByte();
 						obj7.zone4 = reader.ReadByte();
+						if (!ModNet.AllowVanillaClients)
+							PlayerHooks.ReceiveCustomBiomes(obj7, reader);
+
 						if (Main.netMode == 2)
 							NetMessage.TrySendData(36, -1, whoAmI, null, num214);
 
@@ -1778,6 +_,11 @@
 					if (Main.netMode == 2) {
 						if (reader.ReadString() == Netplay.ServerPassword) {
 							Netplay.Clients[whoAmI].State = 1;
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -1963,7 +_,7 @@
 							break;
 
 						Player player5 = Main.player[num51];
-						for (int num52 = 0; num52 < 22; num52++) {
+						for (int num52 = 0; num52 < Player.MaxBuffs; num52++) {
 							player5.buffType[num52] = reader.ReadUInt16();
 							if (player5.buffType[num52] > 0)
 								player5.buffTime[num52] = 60;
@@ -2451,6 +_,7 @@
 					break;
 				case 74:
 					if (Main.netMode == 1) {
+						Netplay.syncingWorld = false;
 						Main.anglerQuest = reader.ReadByte();
 						Main.anglerQuestFinished = reader.ReadBoolean();
 					}
@@ -2540,7 +_,7 @@
 					if (Main.netMode == 1) {
 						int num73 = reader.ReadInt16();
 						int num74 = reader.ReadInt32();
-						if (num73 >= 0 && num73 < 289)
+						if (num73 >= 0) // && num73 < 289) //This prevented mod BannerIds from syncing.
 							NPC.killCount[num73] = num74;
 					}
 					break;
@@ -3325,6 +_,27 @@
 						Main.countsAsHostForGameplay[num2] = flag;
 					}
 					break;
+				case MessageID.InGameChangeConfig:
+					ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+					return;
+				case MessageID.ModPacket:
+					ModNet.HandleModPacket(reader, whoAmI, length);
+					return;
+				case MessageID.SyncMods:
+					if (Main.netMode == 1) {
+						ModNet.SyncClientMods(reader);
+					}
+					else {
+						ModNet.SendNetIDs(whoAmI);
+						NetMessage.SendData(3, whoAmI);
+					}
+					return;
+				case MessageID.ModFile:
+					if (Main.netMode == 1)
+						ModNet.ReceiveMod(reader);
+					else
+						ModNet.SendMod(reader.ReadString(), whoAmI);
+					return;
 				default:
 					if (Netplay.Clients[whoAmI].State == 0)
 						NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
