--- src/Terraria/Terraria/Minecart.cs
+++ src/tModLoader/Terraria/Minecart.cs
@@ -1,8 +_,6 @@
 using Microsoft.Xna.Framework;
 using System;
-#if CLIENT
 using Terraria.GameContent;
-#endif
 
 namespace Terraria
 {
@@ -57,10 +_,8 @@
 		private static int[][] _tileHeight;
 
 		public static void Initialize() {
-#if CLIENT
-			if ((float)TextureAssets.MinecartMount.Width() != 50f)
+			if (!Main.dedServ && (float)TextureAssets.MinecartMount.Width() != 50f)
 				throw new Exception("Be sure to update Minecart.textureWidth to match the actual texture size of " + 50f + ".");
-#endif
 
 			_rightSideConnection = new int[36];
 			_leftSideConnection = new int[36];
