--- src/Terraria/Terraria/MonoLaunch.cs
+++ src/tModLoader/Terraria/MonoLaunch.cs
@@ -7,6 +_,7 @@
 using System.Runtime.InteropServices;
 using System.Runtime.Loader;
 using Terraria;
+using Terraria.ModLoader;
 
 internal static class MonoLaunch
 {
@@ -58,16 +_,16 @@
 					return handle;
 				}
 				
-				Console.WriteLine($"Native Resolve: {assembly.FullName} -> {name}");
+				Logging.tML.Debug($"Native Resolve: {assembly.FullName} -> {name}");
 
 				var dir = Path.Combine(Environment.CurrentDirectory, "Libraries", "Native", getNativeDir(name));
 				var files = Directory.GetFiles(dir, $"*{name}*", SearchOption.AllDirectories);
 				var match = files.FirstOrDefault();
 				
-				Console.WriteLine(match == null ? "\tnot found in Libraries/Native" : $"\tattempting load {match}");
+				Logging.tML.Debug(match == null ? "\tnot found in Libraries/Native" : $"\tattempting load {match}");
 				
 				if (match != null && NativeLibrary.TryLoad(match, out handle)) {
-					Console.WriteLine("\tsuccess");
+					Logging.tML.Debug("\tsuccess");
 					return assemblies[name] = handle;
 				}
 				else {
