--- src/TerrariaNetCore/Terraria/Mount.cs
+++ src/tModLoader/Terraria/Mount.cs
@@ -12,6 +_,8 @@
 using Terraria.GameInput;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
+using static Terraria.ModLoader.BackupIO;
 
 namespace Terraria;
 
@@ -111,10 +_,14 @@
 
 	public class MountData
 	{
+		public ModMount ModMount { get; set; }
+
+		/// <summary> Texture drawn behind the player while in use. </summary>
 		public Asset<Texture2D> backTexture = Asset<Texture2D>.Empty;
 		public Asset<Texture2D> backTextureGlow = Asset<Texture2D>.Empty;
 		public Asset<Texture2D> backTextureExtra = Asset<Texture2D>.Empty;
 		public Asset<Texture2D> backTextureExtraGlow = Asset<Texture2D>.Empty;
+		/// <summary> Texture drawn in front of the player while in use. </summary>
 		public Asset<Texture2D> frontTexture = Asset<Texture2D>.Empty;
 		public Asset<Texture2D> frontTextureGlow = Asset<Texture2D>.Empty;
 		public Asset<Texture2D> frontTextureExtra = Asset<Texture2D>.Empty;
@@ -145,6 +_,10 @@
 		public int abilityDuration;
 		public int abilityCooldown;
 		public int walkingGraceTimeMax;
+		/// <summary>
+		/// The player will attempt to be dismounted if they attempt to use an item.
+		/// <para/> Some examples include <see cref="MountID.Rat"/>, <see cref="MountID.Bat"/>, and <see cref="MountID.Pixie"/>.
+		/// </summary>
 		public bool dismountsOnItemUse;
 		public int spawnDust;
 		public bool spawnDustNoGravity;
@@ -178,6 +_,16 @@
 		public bool emitsLight;
 		public MountDelegatesData delegations = new MountDelegatesData();
 		public int playerXOffset;
+		/// <summary> Alternate <see cref="runSpeed"/> for this minecart while the <see href="https://terraria.wiki.gg/wiki/Minecart_Upgrade_Kit">Minecart Upgrade Kit</see> is active. Defaults to <c>20f</c>.</summary>
+		public float MinecartUpgradeRunSpeed = 20f;
+		/// <summary> Alternate <see cref="acceleration"/> for this minecart while the <see href="https://terraria.wiki.gg/wiki/Minecart_Upgrade_Kit">Minecart Upgrade Kit</see> is active. Defaults to <c>0.1f</c>.</summary>
+		public float MinecartUpgradeAcceleration = 0.1f;
+		/// <summary> Alternate <see cref="jumpSpeed"/> for this minecart while the <see href="https://terraria.wiki.gg/wiki/Minecart_Upgrade_Kit">Minecart Upgrade Kit</see> is active. Defaults to <c>5.15f</c>.</summary>
+		public float MinecartUpgradeJumpSpeed = 5.15f;
+		/// <summary> Alternate <see cref="dashSpeed"/> for this minecart while the <see href="https://terraria.wiki.gg/wiki/Minecart_Upgrade_Kit">Minecart Upgrade Kit</see> is active. Defaults to <c>20f</c>.</summary>
+		public float MinecartUpgradeDashSpeed = 20f;
+		/// <summary> Alternate <see cref="jumpHeight"/> for this minecart while the <see href="https://terraria.wiki.gg/wiki/Minecart_Upgrade_Kit">Minecart Upgrade Kit</see> is active. Defaults to <c>15</c>.</summary>
+		public int MinecartUpgradeJumpHeight = 15;
 	}
 
 	public enum DismountCheckResult
@@ -198,7 +_,7 @@
 	public const int DrawBackExtra = 1;
 	public const int DrawFront = 2;
 	public const int DrawFrontExtra = 3;
-	private static MountData[] mounts;
+	public static MountData[] mounts;
 	private static Vector2[] scutlixEyePositions;
 	private static Vector2 scutlixTextureSize;
 	public const int scutlixBaseDamage = 50;
@@ -212,34 +_,39 @@
 	public static int amountOfBeamsAtOnce = 2;
 	public const float maxDrillLength = 48f;
 	private static Vector2 santankTextureSize;
-	private MountData _data;
+	public MountData _data;
-	private int _type;
+	public int _type;
-	private bool _flipDraw;
+	public bool _flipDraw;
-	private int _frame;
+	public int _frame;
-	private float _frameCounter;
+	public float _frameCounter;
-	private int _frameExtra;
+	public int _frameExtra;
-	private float _frameExtraCounter;
+	public float _frameExtraCounter;
-	private int _frameState;
+	public int _frameState;
-	private int _flyTime;
+	public int _flyTime;
-	private int _idleTime;
+	public int _idleTime;
-	private int _idleTimeNext;
+	public int _idleTimeNext;
-	private float _fatigue;
+	public float _fatigue;
-	private float _fatigueMax;
+	public float _fatigueMax;
-	private bool _abilityCharging;
+	public bool _abilityCharging;
-	private int _abilityCharge;
+	public int _abilityCharge;
-	private int _abilityCooldown;
+	public int _abilityCooldown;
-	private int _abilityDuration;
+	public int _abilityDuration;
-	private bool _abilityActive;
+	public bool _abilityActive;
-	private bool _aiming;
+	public bool _aiming;
-	private bool _shouldSuperCart;
+	public bool _shouldSuperCart;
-	private int _walkingGraceTimeLeft;
+	public int _walkingGraceTimeLeft;
 	public List<DrillDebugDraw> _debugDraw;
-	private object _mountSpecificData;
+	public object _mountSpecificData;
-	private bool _active;
+	public bool _active;
+	[Obsolete("Unused, use MountData.MinecartUpgradeRunSpeed instead")]
 	public static float SuperCartRunSpeed = 20f;
+	[Obsolete("Unused, use MountData.MinecartUpgradeDashSpeed instead")]
 	public static float SuperCartDashSpeed = 20f;
+	[Obsolete("Unused, use MountData.MinecartUpgradeAcceleration instead")]
 	public static float SuperCartAcceleration = 0.1f;
+	[Obsolete("Unused, use MountData.MinecartUpgradeJumpHeight instead")]
 	public static int SuperCartJumpHeight = 15;
+	[Obsolete("Unused, use MountData.MinecartUpgradeJumpSpeed instead")]
 	public static float SuperCartJumpSpeed = 5.15f;
 	private MountDelegatesData _defaultDelegatesData = new MountDelegatesData();
 	public static int[] idleFrames_Rat = new int[11] {
@@ -256,6 +_,7 @@
 		0
 	};
 
+	/// <summary> True if the player is currently using a mount. </summary>
 	public bool Active => _active;
 
 	public int Type => _type;
@@ -331,7 +_,7 @@
 				return _data.runSpeed + 2f;
 
 			if (_shouldSuperCart)
-				return SuperCartRunSpeed;
+				return _data.MinecartUpgradeRunSpeed;
 
 			return _data.runSpeed;
 		}
@@ -340,7 +_,7 @@
 	public float DashSpeed {
 		get {
 			if (_shouldSuperCart)
-				return SuperCartDashSpeed;
+				return _data.MinecartUpgradeDashSpeed;
 
 			return _data.dashSpeed;
 		}
@@ -349,7 +_,7 @@
 	public float Acceleration {
 		get {
 			if (_shouldSuperCart)
-				return SuperCartAcceleration;
+				return _data.MinecartUpgradeAcceleration;
 
 			return _data.acceleration;
 		}
@@ -2511,6 +_,16 @@
 		}
 	}
 
+	/// <summary>
+	/// This method sets a variety of MountData values common to most minecarts.
+	/// <para/>Notably:<code>
+	/// runSpeed = 13f;
+	/// dashSpeed = 13f;
+	/// acceleration = 0.04f;
+	/// jumpHeight = 15;
+	/// jumpSpeed = 5.15f;
+	/// </code>
+	/// </summary>
 	public static void SetAsMinecart(MountData newMount, int buff, Asset<Texture2D> texture, int verticalOffset = 0, int playerVerticalOffset = 0)
 	{
 		newMount.Minecart = true;
@@ -2566,13 +_,14 @@
 
 	public static int GetHeightBoost(int MountType)
 	{
-		if (MountType <= -1 || MountType >= MountID.Count)
+		if (MountType <= -1 || MountType >= mounts.Length)
 			return 0;
 
 		return mounts[MountType].heightBoost;
 	}
 
-	public int JumpHeight(float xVelocity)
+	//TML: Added 'player' parameter.
+	public int JumpHeight(Player player, float xVelocity)
 	{
 		int num = _data.jumpHeight;
 		switch (_type) {
@@ -2587,15 +_,18 @@
 				if (_frameState == 4)
 					num += 5;
 				break;
+			// Extra patch context.
 		}
 
 		if (_shouldSuperCart)
-			num = SuperCartJumpHeight;
+			num = _data.MinecartUpgradeJumpHeight;
 
+		MountLoader.JumpHeight(player, _data, ref num, xVelocity);
 		return num;
 	}
 
-	public float JumpSpeed(float xVelocity)
+	//TML: Added 'player' parameter.
+	public float JumpSpeed(Player player, float xVelocity)
 	{
 		float num = _data.jumpSpeed;
 		switch (_type) {
@@ -2608,10 +_,13 @@
 				if (_frameState == 4)
 					num += 2.5f;
 				break;
+			// Extra patch context.
 		}
 
 		if (_shouldSuperCart)
-			num = SuperCartJumpSpeed;
+			num = _data.MinecartUpgradeJumpSpeed;
+
+		MountLoader.JumpSpeed(player, _data, ref num, xVelocity);
 
 		return num;
 	}
@@ -2975,7 +_,10 @@
 						Vector2 vector = mountedPlayer.Center + new Vector2(mountedPlayer.width * -mountedPlayer.direction, 26f);
 						Vector2 vector2 = new Vector2(0f, -4f).RotatedByRandom(0.10000000149011612);
 						Projectile.NewProjectile(GetProjectileSpawnSource(mountedPlayer), vector.X, vector.Y, vector2.X, vector2.Y, 930, damage, 0f, Main.myPlayer);
+						/*
 						SoundEngine.PlaySound(SoundID.Item89.SoundId, (int)vector.X, (int)vector.Y, SoundID.Item89.Style, 0.2f);
+						*/
+						SoundEngine.PlaySound(SoundID.Item89 with { Volume = SoundID.Item89.Volume * 0.2f }, vector);
 					}
 
 					int type = 14;
@@ -2990,7 +_,10 @@
 					v *= 12f;
 					v = v.RotatedByRandom(0.20000000298023224);
 					Projectile.NewProjectile(GetProjectileSpawnSource(mountedPlayer), vector4.X, vector4.Y, v.X, v.Y, type, damage2, 0f, Main.myPlayer);
+					/*
 					SoundEngine.PlaySound(SoundID.Item11.SoundId, (int)vector4.X, (int)vector4.Y, SoundID.Item11.Style, 0.2f);
+					*/
+					SoundEngine.PlaySound(SoundID.Item11 with { Volume = SoundID.Item11.Volume * 0.2f }, vector4);
 				}
 				break;
 			case 8:
@@ -3221,6 +_,9 @@
 
 	public void UpdateFrame(Player mountedPlayer, int state, Vector2 velocity)
 	{
+		if (!MountLoader.UpdateFrame(mountedPlayer, state, velocity))
+			return;
+
 		if (_frameState != state) {
 			if (_type == 56 && ((_frameState == 5 && (mountedPlayer.dash <= 0 || mountedPlayer.dashDelay >= 0)) || (_frameState != 5 && mountedPlayer.dash > 0 && mountedPlayer.dashDelay < 0)))
 				_frameCounter = 0f;
@@ -4114,7 +_,7 @@
 
 		Tile tileSafely = Framing.GetTileSafely(p.X, p.Y);
 		if (WorldGen.SolidTile(p)) {
-			int num = WorldGen.KillTile_GetTileDustAmount(fail: true, tileSafely);
+			int num = WorldGen.KillTile_GetTileDustAmount(true, tileSafely, p.X, p.Y);
 			if (num > 1)
 				num = 1;
 
@@ -4229,6 +_,9 @@
 	{
 		mountedPlayer.autoJump = AutoJump;
 		_shouldSuperCart = MountID.Sets.Cart[_type] && mountedPlayer.UsingSuperCart;
+
+		MountLoader.UpdateEffects(mountedPlayer);
+
 		if (_shouldSuperCart) {
 			CastSuperCartLaser(mountedPlayer);
 			float num = 1f + Math.Abs(mountedPlayer.velocity.X) / RunSpeed * 2.5f;
@@ -4238,10 +_,13 @@
 		switch (_type) {
 			case 62:
 			case 63:
+				mountedPlayer.allDamage += 0.1f;
+				/*
 				mountedPlayer.meleeDamage += 0.1f;
 				mountedPlayer.rangedDamage += 0.1f;
 				mountedPlayer.magicDamage += 0.1f;
 				mountedPlayer.minionDamage += 0.1f;
+				*/
 				break;
 			case 23: {
 				Vector2 pos3 = mountedPlayer.Center + GetWitchBroomTrinketOriginOffset(mountedPlayer) + (GetWitchBroomTrinketRotation(mountedPlayer) + (float)Math.PI / 2f).ToRotationVector2() * 11f;
@@ -4395,10 +_,13 @@
 					else
 						Lighting.AddLight(mountedPlayer.Center + Vector2.UnitX * mountedPlayer.direction * 20f, vector10.X, vector10.Y, vector10.Z);
 
+					mountedPlayer.allDamage += 0.15f;
+					/*
 					mountedPlayer.meleeDamage += 0.15f;
 					mountedPlayer.rangedDamage += 0.15f;
 					mountedPlayer.magicDamage += 0.15f;
 					mountedPlayer.minionDamage += 0.15f;
+					*/
 				}
 				if (mountedPlayer.statLife <= mountedPlayer.statLifeMax2 / 2)
 					mountedPlayer.MountFishronSpecialCounter = 60f;
@@ -4698,6 +_,9 @@
 	public bool AimAbility(Player mountedPlayer, Vector2 mousePosition)
 	{
 		_aiming = true;
+
+		MountLoader.AimAbility(this, mountedPlayer, mousePosition);
+
 		switch (_type) {
 			case 9: {
 				int frameExtra = _frameExtra;
@@ -4860,7 +_,11 @@
 		}
 
 		int type = _type;
+
+		/*
 		if (type == 50 && texture2D != null) {
+		*/
+		if (type == 50 && texture2D != null && texture2D != Asset<Texture2D>.DefaultValue) {
 			PlayerQueenSlimeMountTextureContent queenSlimeMount = TextureAssets.RenderTargets.QueenSlimeMount;
 			queenSlimeMount.Request();
 			if (queenSlimeMount.IsReady)
@@ -5213,10 +_,13 @@
 				break;
 		}
 
+		flag2 = !MountLoader.Draw(this, playerDrawData, drawType, drawPlayer, ref texture2D, ref texture2D2, ref Position, ref value2, ref drawColor, ref color, ref num9, ref spriteEffects, ref origin, ref scale, shadow);
+
 		if (!flag2) {
 			DrawData item = new DrawData(texture2D, Position, value2, drawColor, num9, origin, scale, spriteEffects);
 			item.shader = currentShader;
 			playerDrawData.Add(item);
+			// Patch note: num9 is used above.
 			if (texture2D2 != null) {
 				item = new DrawData(texture2D2, Position, value2, color * ((float)(int)drawColor.A / 255f), num9, origin, scale, spriteEffects);
 				item.shader = currentShader;
@@ -5402,6 +_,10 @@
 		bool cart = Cart;
 		_active = false;
 		mountedPlayer.ClearBuff(_data.buff);
+
+		bool skipDust = false;
+		MountLoader.Dismount(this, mountedPlayer, ref skipDust);
+
 		_mountSpecificData = null;
 		int type = _type;
 		if ((uint)(type - 55) <= 1u && mountedPlayer.noItems && !mountedPlayer.cursed)
@@ -5417,7 +_,7 @@
 		if (mountedPlayer.petting.isPetting && mountedPlayer.petting.mount)
 			mountedPlayer.StopPettingAnimal();
 
-		if (!ignoreEffect)
+		if (!ignoreEffect && !skipDust)
 			DoSpawnDust(mountedPlayer, isDismounting: true);
 
 		Reset();
@@ -5443,7 +_,7 @@
 
 	public void SetMount(int m, Player mountedPlayer)
 	{
-		if (_type == m || m <= -1 || m >= MountID.Count || (m == 5 && mountedPlayer.wet))
+		if (_type == m || m <= -1 || m >= mounts.Length || (m == 5 && mountedPlayer.wet))
 			return;
 
 		if (_active) {
@@ -5533,7 +_,15 @@
 			_mountSpecificData = new SelectiveFlyingMountData();
 
 		FinalizeMountData(m, mountedPlayer);
+		bool skipDust = false;
+		MountLoader.SetMount(this, mountedPlayer, ref skipDust);
+
+		if (skipDust)
+			goto PostDustSpawn;
+
 		DoSpawnDust(mountedPlayer, isDismounting: false);
+		PostDustSpawn:
+
 		if (_type == 38 && mountedPlayer.whoAmI == Main.myPlayer)
 			AchievementsHelper.NotifyProgressionEvent(32);
 	}
