--- src/Terraria/Terraria/NetMessage.cs
+++ src/tModLoader/Terraria/NetMessage.cs
@@ -1,4 +_,5 @@
 using Ionic.Zlib;
+using log4net.Core;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
 using System;
@@ -11,6 +_,8 @@
 using Terraria.ID;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 
 namespace Terraria
@@ -50,7 +_,7 @@
 		}
 
 		public static MessageBuffer[] buffer = new MessageBuffer[257];
-		private static PlayerDeathReason _currentPlayerDeathReason;
+		public static PlayerDeathReason _currentPlayerDeathReason;
 		private static NetSoundInfo _currentNetSoundInfo;
 		private static CoinLossRevengeSystem.RevengeMarker _currentRevengeMarker;
 
@@ -76,6 +_,9 @@
 			if (Main.netMode == 2 && remoteClient >= 0)
 				num = remoteClient;
 
+			if (ModNet.HijackSendData(num, msgType, remoteClient, ignoreClient, text, number, number2, number3, number4, number5, number6, number7))
+				return;
+
 			lock (buffer[num]) {
 				BinaryWriter writer = buffer[num].writer;
 				if (writer == null) {
@@ -89,12 +_,12 @@
 				writer.Write((byte)msgType);
 				switch (msgType) {
 					case 1:
-						writer.Write("Terraria" + 230);
+						writer.Write(ModLoader.ModLoader.versionedName);
 						break;
 					case 2:
 						text.Serialize(writer);
 						if (Main.dedServ)
-							Console.WriteLine(Language.GetTextValue("CLI.ClientWasBooted", Netplay.Clients[num].Socket.GetRemoteAddress().ToString(), text));
+							Logging.ServerConsoleLine(Language.GetTextValue("CLI.ClientWasBooted", Netplay.Clients[num].Socket.GetRemoteAddress().ToString(), text));
 						break;
 					case 3:
 						writer.Write((byte)remoteClient);
@@ -161,6 +_,7 @@
 							writer.Write((short)num23);
 							writer.Write((byte)number3);
 							writer.Write((short)num24);
+							ItemIO.SendModData(item7, writer);
 							break;
 						}
 					case 7: {
@@ -298,12 +_,15 @@
 							writer.Write((short)WorldGen.SavedOreTiers.Mythril);
 							writer.Write((short)WorldGen.SavedOreTiers.Adamantite);
 							writer.Write((sbyte)Main.invasionType);
+
 							if (SocialAPI.Network != null)
 								writer.Write(SocialAPI.Network.GetLobbyId());
 							else
 								writer.Write(0uL);
 
 							writer.Write(Sandstorm.IntendedSeverity);
+							if (!ModNet.AllowVanillaClients)
+								WorldIO.SendModData(writer);
 							break;
 						}
 					case 8:
@@ -510,6 +_,7 @@
 								value4 = (short)item6.netID;
 
 							writer.Write(value4);
+							ItemIO.SendModData(item6, writer);
 							break;
 						}
 					case 22:
@@ -579,9 +_,10 @@
 								}
 							}
 
-							if (nPC2.type >= 0 && nPC2.type < 663 && Main.npcCatchable[nPC2.type])
+							if (nPC2.type >= 0 && Main.npcCatchable[nPC2.type])
 								writer.Write((byte)nPC2.releaseOwner);
 
+							NPCLoader.SendExtraAI(nPC2, writer);
 							break;
 						}
 					case 24:
@@ -614,18 +_,24 @@
 							if (projectile.knockBack != 0f)
 								bb21[5] = true;
 
-							if (projectile.type > 0 && projectile.type < 950 && ProjectileID.Sets.NeedsUUID[projectile.type])
+							if (projectile.type > 0 && ProjectileID.Sets.NeedsUUID[projectile.type])
 								bb21[7] = true;
 
 							if (projectile.originalDamage != 0)
 								bb21[6] = true;
 
+							byte[] extraAI = ProjectileLoader.SendExtraAI(projectile, ref bb21);
 							writer.Write(bb21);
 							for (int num18 = 0; num18 < Projectile.maxAI; num18++) {
 								if (bb21[num18])
 									writer.Write(projectile.ai[num18]);
 							}
 
+							if (bb21[Projectile.maxAI + 1]) {
+								writer.Write((byte)extraAI.Length);
+								writer.Write(extraAI);
+							}
+
 							if (bb21[4])
 								writer.Write((short)projectile.damage);
 
@@ -642,7 +_,10 @@
 						}
 					case 28:
 						writer.Write((short)number);
+						if (ModNet.AllowVanillaClients)
-						writer.Write((short)number2);
+							writer.Write((short)number2);
+						else
+							writer.Write((int)number2);
 						writer.Write(number3);
 						writer.Write((byte)(number4 + 1f));
 						writer.Write((byte)number5);
@@ -670,6 +_,7 @@
 							writer.Write((short)item5.stack);
 							writer.Write(item5.prefix);
 							writer.Write(value2);
+							ItemIO.SendModData(item5, writer);
 							break;
 						}
 					case 33: {
@@ -713,6 +_,8 @@
 						else {
 							writer.Write((short)0);
 						}
+						if (number >= 100)
+							writer.Write((ushort)number6);
 						break;
 					case 35:
 						writer.Write((byte)number);
@@ -725,6 +_,9 @@
 							writer.Write(player3.zone2);
 							writer.Write(player3.zone3);
 							writer.Write(player3.zone4);
+							if (!ModNet.AllowVanillaClients)
+								PlayerHooks.SendCustomBiomes(player3, writer);
+
 							break;
 						}
 					case 38:
@@ -777,7 +_,7 @@
 						}
 					case 50: {
 							writer.Write((byte)number);
-							for (int n = 0; n < 22; n++) {
+							for (int n = 0; n < Player.MaxBuffs; n++) {
 								writer.Write((ushort)Main.player[number].buffType[n]);
 							}
 
@@ -1325,7 +_,10 @@
 
 				int num25 = (int)writer.BaseStream.Position;
 				writer.BaseStream.Position = position;
+				// TODO: Make this int or some other solution. This used to be cast to short. NetMessage.CheckBytes reads ushort though, so changing this is fine. We want to allow for more data primarily because writing tile
+				// chunks can go more than 32767 (and even 65535) bytes. This is because ModTileEntities can have large amounts of data especially when they sync Items and instanced GlobalItem classes are in play.
+				// Complications include AllowVanillaClients, message length byte length negotiation, and MessageBuffer.readBuffer having a fixed size.
-				writer.Write((short)num25);
+				writer.Write((ushort)num25);
 				writer.BaseStream.Position = num25;
 				if (Main.netMode == 1) {
 					if (Netplay.Connection.Socket.IsConnected()) {
@@ -1646,7 +_,7 @@
 							}
 						}
 
-						if (tile2.type == 88 && tile2.frameX % 54 == 0 && tile2.frameY % 36 == 0) {
+						if (TileID.Sets.BasicDresser[tile2.type] && tile2.frameX % 54 == 0 && tile2.frameY % 36 == 0) {
 							short num8 = (short)Chest.FindChest(j, i);
 							if (num8 != -1) {
 								array[num] = num8;
@@ -1666,7 +_,7 @@
 								array2[num2++] = num10;
 						}
 
-						if (tile2.type == 425 && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0) {
+						if ((tile2.type == 425 || tile2.type >= TileID.Count && Main.tileSign[tile2.type]) && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0) {
 							short num11 = (short)Sign.ReadSign(j, i);
 							if (num11 != -1)
 								array2[num2++] = num11;
@@ -1831,6 +_,14 @@
 				writer.Write(sign.text);
 			}
 
+			foreach (var item in TileEntity.ByPosition) {
+				Point16 pos = item.Key;
+				if (pos.X >= xStart && pos.X < xStart + width && pos.Y >= yStart && pos.Y < yStart + height) {
+					if (item.Value.type > 2)
+						array3[num3++] = (short)item.Value.ID;
+				}
+			}
+
 			writer.Write(num3);
 			for (int m = 0; m < num3; m++) {
 				TileEntity.Write(writer, TileEntity.ByID[array3[m]]);
@@ -2042,10 +_,11 @@
 					buffer[i].totalData += streamLength;
 					buffer[i].checkBytes = true;
 				}
-				catch {
+				catch (Exception e) {
 					if (Main.netMode == 1) {
 						Main.menuMode = 15;
 						Main.statusText = Language.GetTextValue("Error.BadHeaderBufferOverflow");
+						Logging.Terraria.Error(Main.statusText, e);
 						Netplay.Disconnect = true;
 					}
 					else {
@@ -2091,7 +_,7 @@
 				catch (Exception) {
 #if SERVER
 					if (num < NetMessage.buffer.Length - 100)
-						Console.WriteLine("Error on message " + NetMessage.buffer[num + 2]);
+						Logging.ServerConsoleLine("Error on message " + NetMessage.buffer[num + 2], Level.Warn);
 #endif
 					num2 = 0;
 					num = 0;
@@ -2295,7 +_,7 @@
 			}
 
 			if (!flag) {
-				Console.WriteLine(Language.GetTextValue("Net.ServerAutoShutdown"));
+				Logging.ServerConsoleLine(Language.GetTextValue("Net.ServerAutoShutdown"));
 				WorldFile.SaveWorld();
 				Netplay.Disconnect = true;
 			}
@@ -2346,11 +_,12 @@
 					SendData(5, toWho, fromWho, null, plr, 58 + Main.player[plr].armor.Length + Main.player[plr].dye.Length + Main.player[plr].miscEquips.Length + 1 + m, (int)Main.player[plr].miscDyes[m].prefix);
 				}
 
+				PlayerHooks.SyncPlayer(Main.player[plr], toWho, fromWho, false);
 				if (!Netplay.Clients[plr].IsAnnouncementCompleted) {
 					Netplay.Clients[plr].IsAnnouncementCompleted = true;
 					ChatHelper.BroadcastChatMessage(NetworkText.FromKey(Lang.mp[19].Key, Main.player[plr].name), new Color(255, 240, 20), plr);
 					if (Main.dedServ)
-						Console.WriteLine(Lang.mp[19].Format(Main.player[plr].name));
+						Logging.ServerConsoleLine(Lang.mp[19].Format(Main.player[plr].name));
 				}
 
 				return;
@@ -2362,7 +_,7 @@
 				Netplay.Clients[plr].IsAnnouncementCompleted = false;
 				ChatHelper.BroadcastChatMessage(NetworkText.FromKey(Lang.mp[20].Key, Netplay.Clients[plr].Name), new Color(255, 240, 20), plr);
 				if (Main.dedServ)
-					Console.WriteLine(Lang.mp[20].Format(Netplay.Clients[plr].Name));
+					Logging.ServerConsoleLine(Lang.mp[20].Format(Netplay.Clients[plr].Name));
 
 				Netplay.Clients[plr].Name = "Anonymous";
 			}
