--- src/Terraria\Terraria\NetMessage.cs
+++ src/tModLoader\Terraria\NetMessage.cs
@@ -7,6 +_,8 @@
 using Terraria.GameContent.Tile_Entities;
 using Terraria.ID;
 using Terraria.IO;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.Net.Sockets;
 using Terraria.Social;
 
@@ -42,7 +_,7 @@
 				switch (msgType)
 				{
 					case 1:
-						writer.Write("Terraria" + Main.curRelease);
+						writer.Write(ModLoader.ModLoader.versionedName);
 						break;
 					case 2:
 						writer.Write(text);
@@ -404,7 +_,10 @@
 									}
 									if (tile.wall > 0)
 									{
-										writer.Write(tile.wall);
+										if (ModNet.AllowVanillaClients)
+											writer.Write((byte)tile.wall);
+										else
+											writer.Write(tile.wall);
 									}
 									if (tile.liquid > 0 && Main.netMode == 2)
 									{
@@ -431,6 +_,7 @@
 								value = (short)item2.netID;
 							}
 							writer.Write(value);
+							ItemIO.SaveCustomData(item2, writer);
 							break;
 						}
 					case 22:
@@ -498,6 +_,7 @@
 							{
 								writer.Write((byte)nPC.releaseOwner);
 							}
+							NPCLoader.SendExtraAI(nPC, writer);
 							break;
 						}
 					case 24:
@@ -544,10 +_,11 @@
 									bb14[num10] = true;
 								}
 							}
-							if (projectile.type > 0 && projectile.type < 656 && ProjectileID.Sets.NeedsUUID[projectile.type])
+							if (projectile.type > 0 && ProjectileID.Sets.NeedsUUID[projectile.type])
 							{
 								bb14[Projectile.maxAI] = true;
 							}
+							byte[] extraAI = ProjectileLoader.SendExtraAI(projectile, ref bb14);
 							writer.Write(bb14);
 							for (int num11 = 0; num11 < Projectile.maxAI; num11++)
 							{
@@ -555,6 +_,11 @@
 								{
 									writer.Write(projectile.ai[num11]);
 								}
+							}
+							if (bb14[Projectile.maxAI + 1])
+							{
+								writer.Write((byte)extraAI.Length);
+								writer.Write(extraAI);
 							}
 							if (bb14[Projectile.maxAI])
 							{
@@ -594,6 +_,7 @@
 							writer.Write((short)item3.stack);
 							writer.Write(item3.prefix);
 							writer.Write(value3);
+							ItemIO.SaveCustomData(item3, writer);
 							break;
 						}
 					case 33:
@@ -634,6 +_,10 @@
 						writer.Write((short)number2);
 						writer.Write((short)number3);
 						writer.Write((short)number4);
+						if (number >= 100)
+						{
+							writer.Write((ushort)number6);
+						}
 						if (Main.netMode == 2)
 						{
 							Netplay.GetSectionX((int)number2);
@@ -1461,7 +_,7 @@
 			int num5 = 0;
 			int num6 = 0;
 			byte b = 0;
-			byte[] array4 = new byte[13];
+			byte[] array4 = new byte[ModNet.AllowVanillaClients ? 13 : 14];
 			Tile tile = null;
 			for (int i = yStart; i < yStart + height; i++)
 			{
@@ -1509,7 +_,7 @@
 								num5++;
 								b |= 32;
 							}
-							if (tile2.type == 21 && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0)
+							if (TileLoader.IsChest(tile2.type) && tile2.frameX % 36 == 0 && tile2.frameY % 36 == 0)
 							{
 								short num7 = (short)Chest.FindChest(j, i);
 								if (num7 != -1)
@@ -1518,7 +_,7 @@
 									num += 1;
 								}
 							}
-							if (tile2.type == 88 && tile2.frameX % 54 == 0 && tile2.frameY % 36 == 0)
+							if (TileLoader.IsDresser(tile2.type) && tile2.frameX % 54 == 0 && tile2.frameY % 36 == 0)
 							{
 								short num8 = (short)Chest.FindChest(j, i);
 								if (num8 != -1)
@@ -1603,7 +_,17 @@
 						if (tile2.wall != 0)
 						{
 							b |= 4;
-							array4[num5] = tile2.wall;
+							if (ModNet.AllowVanillaClients)
+							{
+								array4[num5] = (byte) tile2.wall;
+							}
+							else
+							{
+								byte[] wallBytes = BitConverter.GetBytes(tile2.wall);
+								array4[num5] = wallBytes[0];
+								num5++;
+								array4[num5] = wallBytes[1];
+							}
 							num5++;
 							if (tile2.wallColor() != 0)
 							{
@@ -1839,7 +_,7 @@
 						}
 						if ((b3 & 4) == 4)
 						{
-							tile.wall = reader.ReadByte();
+							tile.wall = ModNet.AllowVanillaClients ? reader.ReadByte() : reader.ReadUInt16();
 							if ((b & 16) == 16)
 							{
 								tile.wallColor(reader.ReadByte());

