--- src/TerrariaNetCore/Terraria/Netplay.cs
+++ src/tModLoader/Terraria/Netplay.cs
@@ -7,8 +_,11 @@
 using NATUPNPLib;
 using ReLogic.OS;
 using Terraria.Audio;
+using Terraria.ID;
 using Terraria.Localization;
 using Terraria.Map;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
 using Terraria.Net;
 using Terraria.Net.Sockets;
 using Terraria.Social;
@@ -78,6 +_,12 @@
 		Main.ActiveNetDiagnosticsUI.Reset();
 	}
 
+	/// <summary>
+	/// Sets all world sections as unloaded for all clients. This results in all clients being sent the world sections again.
+	/// <br/><br/> This is used at the end of Hardmode world generation to force all clients to receive the tiles from the server again.
+	/// <br/><br/> If you are doing massive world changes in-game outside of the Hardmode transformation, this is a simple way to ensure that all clients see those tile changes without sending countless <see cref="NetMessage.SendTileSquare(int, int, int, TileChangeType)"/> messages.
+	/// <br/><br/> <b>Call on Server only!</b>
+	/// </summary>
 	public static void ResetSections()
 	{
 		for (int i = 0; i < 256; i++) {
@@ -165,6 +_,8 @@
 
 	private static void OnConnectionAccepted(ISocket client)
 	{
+		Logging.ServerConsoleLine(Language.GetTextValue("Net.ClientConnecting", client.GetRemoteAddress().GetFriendlyName()));
+
 		int num = FindNextOpenClientSlot();
 		if (num != -1) {
 			Clients[num].Reset();
@@ -298,8 +_,13 @@
 		for (int i = 0; i < 256; i++) {
 			if (Clients[i].PendingTermination) {
 				if (Clients[i].PendingTerminationApproved) {
-					Clients[i].Reset();
+					// TML: Re-order to prevent RemoteClient.Reset from re-instantiating the player before
+					//      execution ends up in PlayerLoader.PlayerDisconnect, which would invoke the hook
+					//      with useless Player/ModPlayer data.
+					//      Set the RemoteClient.State to 0 so SyncOnePlayer will still treat them as disconnected.
+					Clients[i].State = 0;
 					NetMessage.SyncDisconnectedPlayer(i);
+					Clients[i].Reset();
 				}
 
 				continue;
@@ -312,7 +_,7 @@
 			}
 
 			if (Clients[i].IsActive) {
-				Clients[i].PendingTermination = true;
+				Clients[i].SetPendingTermination("Connection lost");
 				Clients[i].PendingTerminationApproved = true;
 				continue;
 			}
@@ -413,6 +_,12 @@
 
 	private static void ClientLoopSetup(RemoteAddress address)
 	{
+		string friendlyName = address.GetFriendlyName();
+		if (string.IsNullOrEmpty(friendlyName))
+			Utils.ShowFancyErrorMessage("Unable to find Steam Hosted Multiplayer server.", 0);
+
+		Logging.Terraria.InfoFormat("Connecting to {0}", friendlyName);
+
 		Connection.ResetSpecialFlags();
 		ResetNetDiag();
 		Main.ServerSideCharacter = false;
@@ -429,11 +_,11 @@
 		Main.netMode = 1;
 		Main.menuMode = 14;
 		if (!Main.autoPass)
-			Main.statusText = Language.GetTextValue("Net.ConnectingTo", address.GetFriendlyName());
+			Main.statusText = Language.GetTextValue("Net.ConnectingTo", friendlyName);
 
 		Disconnect = false;
 		Connection = new RemoteServer();
-		Connection.ReadBuffer = new byte[1024];
+		Connection.ReadBuffer = new byte[ushort.MaxValue]; // TML: 1024 -> ushort.MaxValue
 	}
 
 	private static void InnerClientLoop()
@@ -453,6 +_,24 @@
 						NetMessage.SendData(1);
 					}
 
+					if (Connection.State == 1 && ModNet.NetReloadActive && Main.netMode == 0) {
+						Main.netMode = 1;
+						SendKeepAliveDuringModReloadMessage();
+						Main.netMode = 0;
+					}
+					if (Connection.State == 1 && Main.menuMode == 888 && Main.MenuUI.CurrentState == Interface.serverModsDifferMessage && Main.netMode == 1) {
+						// Same logic as above, but not currently reloading, so netMode is still 1, so don't change it.
+						SendKeepAliveDuringModReloadMessage();
+					}
+					static void SendKeepAliveDuringModReloadMessage()
+					{
+						// Every 30 seconds. Need StopWatch as timer because this code called more frequently than game loop.
+						if (ModNet.NetReloadKeepAliveTimer.ElapsedMilliseconds > 30000) {
+							ModNet.NetReloadKeepAliveTimer.Restart();
+							new ModPacket(MessageID.KeepAliveDuringModReload).Send();
+						}
+					}
+
 					if (Connection.State == 2 && num != Connection.State)
 						Main.statusText = Language.GetTextValue("Net.SendingPlayerData");
 
@@ -472,6 +_,9 @@
 						Main.cloudAlpha = Main.maxRaining;
 						Main.ToggleGameplayUpdates(state: false);
 						WorldGen.clearWorld();
+
+						SystemLoader.OnWorldLoad();
+
 						if (Main.mapEnabled)
 							Main.Map.Load();
 					}
@@ -523,6 +_,11 @@
 			}
 
 			if (!Main.gameMenu) {
+				// This is only called in case of kicks and other sudden disconnects.
+				// The more common (client) call to this is in Main.SaveAndQuitCallback.
+				string originalStatusText = Main.statusText;
+				SystemLoader.OnWorldUnload();
+
 				Main.gameMenu = true;
 				Main.SwitchNetMode(0);
 				MapHelper.noStatusText = true;
@@ -532,6 +_,7 @@
 				SoundEngine.StopTrackedSounds();
 				MapHelper.noStatusText = false;
 				Main.menuMode = 14;
+				Main.statusText = originalStatusText;
 			}
 
 			NetMessage.buffer[256].Reset();
@@ -589,7 +_,29 @@
 		thread.Start();
 	}
 
+	/*
 	public static bool SetRemoteIP(string remoteAddress) => SetRemoteIPOld(remoteAddress);
+	*/
+
+	public static bool SetRemoteIP(string remoteAddress)
+	{
+		if (remoteAddress.Contains(':')) {
+			int colonInd = remoteAddress.LastIndexOf(':');
+
+			if (int.TryParse(remoteAddress.AsSpan(colonInd + 1), out int port)) {
+				remoteAddress = remoteAddress.Substring(0, colonInd);
+
+				bool set = SetRemoteIPOld(remoteAddress);
+
+				if (set)
+					ListenPort = port;
+
+				return set;
+			}
+		}
+
+		return SetRemoteIPOld(remoteAddress);
+	}
 
 	public static bool SetRemoteIPOld(string remoteAddress)
 	{
