--- src/Terraria/Terraria/NPC.cs
+++ src/tModLoader/Terraria/NPC.cs
@@ -21,10 +_,11 @@
 using Terraria.Localization;
 using Terraria.Utilities;
 using Terraria.WorldBuilding;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
-	public class NPC : Entity
+	public partial class NPC : Entity
 	{
 		private const int NPC_TARGETS_START = 300;
 		public bool IsABestiaryIconDummy;
@@ -103,14 +_,14 @@
 		private static int townRangeY = sHeight;
 		public float npcSlots = 1f;
 		private static bool noSpawnCycle = false;
-		private static int activeTime = 750;
+		public static int activeTime = 750;
 		private static int defaultSpawnRate = 600;
 		private static int defaultMaxSpawns = 5;
 		public bool dontCountMe;
 		public const int maxBuffs = 5;
 		public int[] buffType = new int[5];
 		public int[] buffTime = new int[5];
-		public bool[] buffImmune = new bool[323];
+		public bool[] buffImmune = new bool[BuffLoader.BuffCount];
 		public bool midas;
 		public bool ichor;
 		public bool onFire;
@@ -649,7 +_,7 @@
 		public static void UpdateFoundActiveNPCs() {
 			for (int i = 0; i < 200; i++) {
 				NPC nPC = Main.npc[i];
-				if (nPC.active && nPC.type >= 0 && nPC.type < 663)
+				if (nPC.active && nPC.type >= 0 && nPC.type < NPCLoader.NPCCount)
 					npcsFoundForCheckActive[nPC.type] = true;
 			}
 
@@ -1759,7 +_,7 @@
 				case 656:
 					return Language.RandomFromCategory("BunnyNames_White", WorldGen.genRand).Value;
 				default:
-					return "";
+					return NPCLoader.TownNPCName(npcType);
 			}
 		}
 
@@ -1955,7 +_,7 @@
 				case 656:
 					return 39;
 				default:
-					return -1;
+					return NPCHeadLoader.GetNPCHeadSlot(type);
 			}
 		}
 
@@ -2020,7 +_,7 @@
 				case 39:
 					return 656;
 				default:
-					return -1;
+					return NPCHeadLoader.GetNPCFromHeadSlot(headIndex);
 			}
 		}
 
@@ -2074,6 +_,7 @@
 					break;
 			}
 
+			NPCLoader.BossHeadRotation(this, ref result);
 			return result;
 		}
 
@@ -2083,6 +_,7 @@
 			if (num == 491 && spriteDirection == 1)
 				result = SpriteEffects.FlipHorizontally;
 
+			NPCLoader.BossHeadSpriteEffects(this, ref result);
 			return result;
 		}
 
@@ -2756,6 +_,16 @@
 			}
 		}
 
+		public void CloneDefaults(int Type) {
+			int originalType = type;
+			int originalNetID = netID;
+			ModNPC originalModEntity = modNPC;
+			SetDefaultsKeepPlayerInteraction(Type);
+			type = originalType;
+			netID = originalNetID;
+			modNPC = originalModEntity;
+		}
+
 		public void SetDefaultsKeepPlayerInteraction(int Type) {
 			bool[] array = new bool[playerInteraction.Length];
 			for (int i = 0; i < playerInteraction.Length; i++) {
@@ -2792,6 +_,8 @@
 				return;
 			}
 
+			modNPC = null;
+			globalNPCs = new GlobalNPC[0];
 			waterMovementSpeed = (lavaMovementSpeed = 0.5f);
 			honeyMovementSpeed = 0.25f;
 			netOffset *= 0f;
@@ -2837,7 +_,10 @@
 				buffType[k] = 0;
 			}
 
+			if (buffImmune.Length != BuffLoader.BuffCount)
+				Array.Resize(ref buffImmune, BuffLoader.BuffCount);
+
-			for (int l = 0; l < 323; l++) {
+			for (int l = 0; l < buffImmune.Length; l++) {
 				buffImmune[l] = false;
 			}
 
@@ -11611,8 +_,9 @@
 				noGravity = true;
 			}
 
+			NPCLoader.SetDefaults(this);
 			if (flag) {
-				for (int num4 = 0; num4 < 323; num4++) {
+				for (int num4 = 0; num4 < buffImmune.Length; num4++) {
 					buffImmune[num4] = true;
 				}
 			}
@@ -12155,6 +_,7 @@
 					break;
 			}
 
+			NPCLoader.ScaleExpertStats(this, numPlayers, balance);
 			defDefense = defense;
 			defDamage = damage;
 			life = lifeMax;
@@ -12358,6 +_,10 @@
 		}
 
 		public void AI() {
+			NPCLoader.NPCAI(this);
+		}
+
+		public void VanillaAI() {
 			if (aiStyle == 0) {
 				if (Main.netMode != 1) {
 					for (int i = 0; i < 255; i++) {
@@ -37736,6 +_,7 @@
 			if (type == 638 || type == 656)
 				num = 0;
 
+			//patch file: flag
 			bool flag = Main.raining;
 			if (!Main.dayTime)
 				flag = true;
@@ -37824,6 +_,8 @@
 					num2 += 0.15f;
 					defense += 8;
 				}
+
+				NPCLoader.BuffTownNPC(ref num2, ref defense);
 			}
 
 			if (type == 142 && Main.netMode != 1 && !Main.xMas) {
@@ -37883,7 +_,7 @@
 					break;
 			}
 
-			if (type >= 0 && type < 663 && NPCID.Sets.TownCritter[type] && target == 255) {
+			if (type >= 0 && NPCID.Sets.TownCritter[type] && target == 255) {
 				TargetClosest();
 				if (position.X < Main.player[target].position.X) {
 					base.direction = 1;
@@ -38035,7 +_,14 @@
 			int num11 = -1;
 			if (!flag6 && Main.netMode != 1 && !flag3) {
 				for (int l = 0; l < 200; l++) {
+					if (!Main.npc[l].active) continue;
+					bool? modCanHit = NPCLoader.CanHitNPC(Main.npc[l], this);
+					if (modCanHit.HasValue && !modCanHit.Value)
+						continue;
+
+					bool canHitVal = modCanHit.HasValue && modCanHit.Value;
+					// TODO: NPCLoader.CanHitNPC should return a plain bool. The only use of "override true" below is to force the skeleton merchant to be hit by skeletons
-					if (!Main.npc[l].active || Main.npc[l].friendly || Main.npc[l].damage <= 0 || !(Main.npc[l].Distance(base.Center) < num6) || (type == 453 && NPCID.Sets.Skeletons[Main.npc[l].type]) || (!Main.npc[l].noTileCollide && !Collision.CanHit(base.Center, 0, 0, Main.npc[l].Center, 0, 0)))
+					if (!Main.npc[l].active || Main.npc[l].friendly || Main.npc[l].damage <= 0 || !(Main.npc[l].Distance(base.Center) < num6) || (type == 453 && NPCID.Sets.Skeletons[Main.npc[l].type] && canHitVal) || (!Main.npc[l].noTileCollide && !Collision.CanHit(base.Center, 0, 0, Main.npc[l].Center, 0, 0)))
 						continue;
 
 					flag10 = true;
@@ -38271,7 +_,7 @@
 
 					if (closeDoor && ((position.X + (float)(width / 2)) / 16f > (float)(doorX + 2) || (position.X + (float)(width / 2)) / 16f < (float)(doorX - 2))) {
 						Tile tileSafely = Framing.GetTileSafely(doorX, doorY);
-						if (tileSafely.type == 11) {
+						if (TileLoader.CloseDoorID(tileSafely) >= 0) {
 							if (WorldGen.CloseDoor(doorX, doorY)) {
 								closeDoor = false;
 								NetMessage.SendData(19, -1, -1, null, 1, doorX, doorY, base.direction);
@@ -38416,7 +_,7 @@
 						Tile tileSafely4 = Framing.GetTileSafely(num18, num19 - 1);
 						Tile tileSafely5 = Framing.GetTileSafely(num18, num19 - 2);
 						bool flag15 = height / 16 < 3;
-						if (townNPC && tileSafely5.nactive() && (tileSafely5.type == 10 || tileSafely5.type == 388) && (Main.rand.Next(10) == 0 || flag)) {
+						if (townNPC && tileSafely5.nactive() && (TileLoader.OpenDoorID(tileSafely5) >= 0 || tileSafely5.type == 388) && (Main.rand.Next(10) == 0 || flag)) {
 							if (Main.netMode != 1) {
 								if (WorldGen.OpenDoor(num18, num19 - 2, base.direction)) {
 									closeDoor = true;
@@ -38738,6 +_,7 @@
 					num24 = 589;
 					num26 = 7f;
 					num25 = 22;
+					//patch file: num22, num23, num25, num26, num24, maxValue2
 					num27 = 1;
 					num28 = 10;
 					maxValue = 1;
@@ -38745,6 +_,10 @@
 					num29 = 10f;
 				}
 
+				NPCLoader.TownNPCAttackStrength(this, ref num25, ref knockBack);
+				NPCLoader.TownNPCAttackCooldown(this, ref num28, ref maxValue);
+				NPCLoader.TownNPCAttackProj(this, ref num24, ref num27);
+				NPCLoader.TownNPCAttackProjSpeed(this, ref num26, ref num29, ref num31);
 				if (Main.expertMode)
 					num25 = (int)((float)num25 * Main.GameModeInfo.TownNPCDamageMultiplier);
 
@@ -38789,7 +_,7 @@
 				int num37 = 0;
 				int maxValue2 = 0;
 				float knockBack2 = 0f;
-				int num38 = 0;
+				float num38 = 0;
 				bool flag18 = false;
 				float num39 = 0f;
 				if ((float)NPCID.Sets.AttackTime[type] == ai[1]) {
@@ -38894,12 +_,14 @@
 				}
 				else if (type == 228) {
 					num33 = 267;
+					//patch file: num31, num33, num32, num34
 					num35 = 14f;
 					num34 = 20;
 					num36 = 1;
 					num37 = 10;
 					maxValue2 = 1;
 					knockBack2 = 3f;
+					//patch file: num35, maxValue3, knockBack2, num36, num37
 					num38 = 6;
 					num39 = 0.4f;
 				}
@@ -38996,6 +_,12 @@
 					}
 				}
 
+				NPCLoader.TownNPCAttackStrength(this, ref num34, ref knockBack2);
+				NPCLoader.TownNPCAttackCooldown(this, ref num37, ref maxValue2);
+				NPCLoader.TownNPCAttackProj(this, ref num33, ref num36);
+				NPCLoader.TownNPCAttackProjSpeed(this, ref num35, ref num38, ref num39);
+				NPCLoader.TownNPCAttackShoot(this, ref flag18);
+
 				if (Main.expertMode)
 					num34 = (int)((float)num34 * Main.GameModeInfo.TownNPCDamageMultiplier);
 
@@ -39085,20 +_,24 @@
 
 				if (type == 54) {
 					num43 = 585;
+					//patch file: num45, num44, n
 					num45 = 10f;
 					num44 = 16;
 					n = 30;
+					//patch file: num46, maxValue3, knockBack3, num50
 					num46 = 20;
 					maxValue3 = 15;
 					knockBack3 = 2f;
 					num50 = 1f;
 				}
 				else if (type == 108) {
+					//patch file: num49
 					num43 = 15;
 					num45 = 6f;
 					num44 = 18;
 					n = 15;
 					num46 = 15;
+					//patch file: num46
 					maxValue3 = 5;
 					knockBack3 = 3f;
 					num47 = 20f;
@@ -39121,6 +_,11 @@
 					knockBack3 = 3f;
 				}
 
+				NPCLoader.TownNPCAttackStrength(this, ref num44, ref knockBack3);
+				NPCLoader.TownNPCAttackCooldown(this, ref num46, ref maxValue3);
+				NPCLoader.TownNPCAttackProj(this, ref num43, ref n);
+				NPCLoader.TownNPCAttackProjSpeed(this, ref num45, ref num47, ref num50);
+				NPCLoader.TownNPCAttackMagic(this, ref num49);
 				if (Main.expertMode)
 					num44 = (int)((float)num44 * Main.GameModeInfo.TownNPCDamageMultiplier);
 
@@ -39231,12 +_,16 @@
 				}
 				else if (type == 637 || type == 638 || type == 656) {
 					num60 = 10;
+					//patch file: num59, num61, num62, num58
 					num62 = (num63 = 32);
 					num59 = 15;
 					maxValue4 = 8;
 					num61 = 3f;
 				}
 
+				NPCLoader.TownNPCAttackStrength(this, ref num60, ref num61);
+				NPCLoader.TownNPCAttackCooldown(this, ref num59, ref maxValue4);
+				NPCLoader.TownNPCAttackSwing(this, ref num63, ref num62);
 				if (Main.expertMode)
 					num60 = (int)((float)num60 * Main.GameModeInfo.TownNPCDamageMultiplier);
 
@@ -42808,7 +_,7 @@
 					Main.tile[num177 - direction, num178 + 1] = new Tile();
 
 				Main.tile[num177, num178 + 1].halfBrick();
-				if (Main.tile[num177, num178 - 1].nactive() && (Main.tile[num177, num178 - 1].type == 10 || Main.tile[num177, num178 - 1].type == 388) && flag8) {
+				if (Main.tile[num177, num178 - 1].nactive() && (TileLoader.IsClosedDoor(Main.tile[num177, num178 - 1]) || Main.tile[num177, num178 - 1].type == 388) && flag8) {
 					ai[2] += 1f;
 					ai[3] = 0f;
 					if (ai[2] >= 60f) {
@@ -42847,7 +_,7 @@
 									NetMessage.SendData(17, -1, -1, null, 0, num177, num178 - 1);
 							}
 							else {
-								if (Main.tile[num177, num178 - 1].type == 10) {
+								if (TileLoader.OpenDoorID(Main.tile[num177, num178 - 1]) >= 0) {
 									bool flag24 = WorldGen.OpenDoor(num177, num178 - 1, direction);
 									if (!flag24) {
 										ai[3] = num52;
@@ -43973,7 +_,7 @@
 								Tile tileSafely = Framing.GetTileSafely(i, j);
 								bool flag2 = tileSafely.active() && Main.tileSolid[tileSafely.type] && !Main.tileFrameImportant[tileSafely.type];
 								if (flag && flag2) {
-									int num38 = WorldGen.KillTile_GetTileDustAmount(fail: true, tileSafely);
+									int num38 = WorldGen.KillTile_GetTileDustAmount(true, tileSafely, i, j);
 									for (int k = 0; k < num38; k++) {
 										Dust obj = Main.dust[WorldGen.KillTile_MakeTileDust(i, j, tileSafely)];
 										obj.velocity.Y -= 3f + (float)num37 * 1.5f;
@@ -45597,7 +_,7 @@
 				Tile tileSafely4 = Framing.GetTileSafely(num64, num65 - 1);
 				Tile tileSafely5 = Framing.GetTileSafely(num64, num65 - 2);
 				Tile tileSafely6 = Framing.GetTileSafely(num64, num65 - 3);
-				if (flag8 && tileSafely4.nactive() && (tileSafely4.type == 10 || tileSafely4.type == 388)) {
+				if (flag8 && tileSafely4.nactive() && (TileLoader.IsClosedDoor(tileSafely4) || tileSafely4.type == 388)) {
 					ai[0] += 1f;
 					ai[3] = 0f;
 					if (ai[0] >= 60f) {
@@ -45627,7 +_,7 @@
 									NetMessage.SendData(17, -1, -1, null, 0, num64, num65 - 1);
 							}
 							else {
-								if (tileSafely4.type == 10) {
+								if (TileLoader.IsClosedDoor(tileSafely4)) {
 									bool flag38 = WorldGen.OpenDoor(num64, num65 - 1, direction);
 									if (!flag38) {
 										ai[3] = num19;
@@ -47608,6 +_,10 @@
 		public bool UsesPartyHat() {
 			if (type == 441 || type == 37 || type == 633)
 				return false;
+			
+			if (!NPCLoader.UsesPartyHat(this)) {
+				return false;
+			}
 
 			if (ForcePartyHatOn)
 				return true;
@@ -47660,6 +_,10 @@
 				num = TextureAssets.Npc[type].Height() / Main.npcFrameCount[type];
 			}
 
+			NPCLoader.FindFrame(this, num);
+		}
+
+		public void VanillaFindFrame(int num) {
 			int num2 = 0;
 			if (aiAction == 0)
 				num2 = ((velocity.Y < 0f) ? 2 : ((velocity.Y > 0f) ? 3 : ((velocity.X != 0f) ? 1 : 0)));
@@ -55512,7 +_,10 @@
 					if (Main.tile[x, y].type == 467)
 						number2 = 5;
 
-					NetMessage.SendData(34, -1, -1, null, number2, x, y, 0f, number);
+					if (Main.tile[x, y].type >= TileID.Count)
+						number2 = 101;
+
+					NetMessage.SendData(34, -1, -1, null, number2, x, y, 0f, number, Main.tile[x, y].type, 0);
 					NetMessage.SendTileSquare(-1, x, y, 3);
 				}
 
@@ -55835,6 +_,9 @@
 			if (!active || DoesntDespawnToInactivity())
 				return;
 
+			if (!NPCLoader.CheckActive(this))
+				return;
+
 			if (townNPC) {
 				AddIntoPlayersTownNPCSlots();
 				return;
@@ -55994,6 +_,9 @@
 				return;
 			}
 
+			if (!NPCLoader.CheckDead(this))
+				return;
+
 			noSpawnCycle = true;
 			if (townNPC && type != 37 && type != 453) {
 				if (Main.netMode != 2)
@@ -56042,7 +_,9 @@
 			if (DeathSound != null)
 				SoundEngine.PlaySound(DeathSound, base.position);
 
+			if (NPCLoader.SpecialNPCLoot(this)) {
+			}
-			if (type == 13 || type == 14 || type == 15) {
+			else if (type == 13 || type == 14 || type == 15) {
 				DropEoWLoot();
 			}
 			else if (type == 134) {
@@ -56555,7 +_,7 @@
 		}
 
 		public static void ResetKillCount() {
-			for (int i = 0; i < 663; i++) {
+			for (int i = 0; i < killCount.Length; i++) {
 				killCount[i] = 0;
 			}
 		}
@@ -56648,7 +_,7 @@
 
 			if (type == 1 && ai[1] > 0f) {
 				int num3 = (int)ai[1];
-				if (num3 > 0 && num3 < 5045) {
+				if (num3 > 0) {
 					int stackForSlimeItemDrop = GetStackForSlimeItemDrop(num3);
 					Item.NewItem((int)position.X, (int)position.Y, width, height, num3, stackForSlimeItemDrop);
 				}
@@ -59122,7 +_,7 @@
 		}
 
 		public void NPCLoot() {
-			if (Main.netMode == 1 || type >= 663)
+			if (Main.netMode == 1)
 				return;
 
 			Player closestPlayer = Main.player[Player.FindClosest(position, width, height)];
@@ -59139,6 +_,9 @@
 			if ((type == 23 && Main.hardMode) || (SpawnedFromStatue && NPCID.Sets.NoEarlymodeLootWhenSpawnedFromStatue[type] && !Main.hardMode) || (SpawnedFromStatue && NPCID.Sets.StatueSpawnedDropRarity[type] != -1f && (Main.rand.NextFloat() >= NPCID.Sets.StatueSpawnedDropRarity[type] || !AnyInteractions())))
 				return;
 
+			if (!NPCLoader.PreNPCLoot(this))
+				return;
+
 			bool num = downedMechBoss1 && downedMechBoss2 && downedMechBoss3;
 			DoDeathEvents_BeforeLoot(closestPlayer);
 			NPCLoot_DropItems(closestPlayer);
@@ -59150,6 +_,7 @@
 					ChatHelper.BroadcastChatMessage(NetworkText.FromKey(Lang.misc[32].Key), new Color(50, 255, 130));
 			}
 
+			NPCLoader.NPCLoot(this);
 			NPCLoot_DropMoney(closestPlayer);
 			NPCLoot_DropHeals(closestPlayer);
 		}
@@ -59187,7 +_,7 @@
 			WoFKilledToday = false;
 		}
 
-		private void DoDeathEvents_DropBossPotionsAndHearts() {
+		private void DoDeathEvents_DropBossPotionsAndHearts(ref string typeName) {
 			int stack = Main.rand.Next(5, 16);
 			int num = 28;
 			if (type == 113)
@@ -59211,6 +_,7 @@
 			else if (type == 398)
 				num = 3544;
 
+			NPCLoader.BossLoot(this, ref typeName, ref num);
 			Item.NewItem((int)position.X, (int)position.Y, width, height, num, stack);
 			int num2 = Main.rand.Next(5) + 5;
 			for (int i = 0; i < num2; i++) {
@@ -59228,7 +_,7 @@
 			}
 		}
 
-		private void DoDeathEvents_CelebrateBossDeath() {
+		private void DoDeathEvents_CelebrateBossDeath(string typeName) {
 			if (type == 125 || type == 126) {
 				if (Main.netMode == 0)
 					Main.NewText(Language.GetTextValue("Announcement.HasBeenDefeated_Plural", Language.GetTextValue("Enemies.TheTwins")), 175, 75);
@@ -59242,7 +_,7 @@
 					ChatHelper.BroadcastChatMessage(NetworkText.FromKey("Announcement.HasBeenDefeated_Single", NetworkText.FromKey("Enemies.MoonLord")), new Color(175, 75, 255));
 			}
 			else if (Main.netMode == 0) {
-				Main.NewText(Language.GetTextValue("Announcement.HasBeenDefeated_Single", TypeName), 175, 75);
+				Main.NewText(Language.GetTextValue("Announcement.HasBeenDefeated_Single", typeName), 175, 75);
 			}
 			else if (Main.netMode == 2) {
 				ChatHelper.BroadcastChatMessage(NetworkText.FromKey("Announcement.HasBeenDefeated_Single", GetTypeNetName()), new Color(175, 75, 255));
@@ -59624,8 +_,10 @@
 			}
 
 			if (boss) {
+				string typeName = TypeName;
+
-				DoDeathEvents_DropBossPotionsAndHearts();
+				DoDeathEvents_DropBossPotionsAndHearts(ref typeName);
-				DoDeathEvents_CelebrateBossDeath();
+				DoDeathEvents_CelebrateBossDeath(typeName);
 				if (Main.netMode == 2)
 					NetMessage.SendData(7);
 			}
@@ -60017,8 +_,10 @@
 					NetMessage.SendData(106, -1, -1, null, (int)position.X, position.Y);
 				}
 				else {
-					new Item().SetDefaults(Main.npc[i].catchItem);
+					// new Item().SetDefaults(Main.npc[i].catchItem);
-					Item.NewItem((int)Main.player[who].Center.X, (int)Main.player[who].Center.Y, 0, 0, Main.npc[i].catchItem, 1, noBroadcast: false, 0, noGrabDelay: true);
+					int itemWhoAmI = Item.NewItem((int)Main.player[who].Center.X, (int)Main.player[who].Center.Y, 0, 0, Main.npc[i].catchItem, 1, true, 0, true, false);
+					NPCLoader.OnCatchNPC(Main.npc[i], Main.player[who], Main.item[itemWhoAmI]);
+					NetMessage.SendData(MessageID.SyncItem, -1, -1, null, itemWhoAmI, 1f); // NewItem above changed to noBroadcast, number2 is 1 for noGrabDelay effect.
 					Main.npc[i].active = false;
 					NetMessage.SendData(23, -1, -1, null, i);
 				}
@@ -60075,6 +_,7 @@
 			if (type == 551)
 				itemType = 3860;
 
+			NPCLoader.BossBag(this, ref itemType);
 			DropItemInstanced(position, base.Size, itemType);
 		}
 
@@ -60205,7 +_,7 @@
 				NetMessage.SendData(71, -1, -1, null, x, y, Type, Style);
 			}
 			else {
-				if (Type < 0 || Type >= 663 || !Main.npcCatchable[Type] || !CanReleaseNPCs(who))
+				if (Type < 0 || !Main.npcCatchable[Type] || !CanReleaseNPCs(who))
 					return;
 
 				switch (Type) {
@@ -60422,6 +_,7 @@
 			int num9;
 			int num10;
 			bool flag15;
+			NPCSpawnInfo spawnInfo = new NPCSpawnInfo();
 			while (true) {
 				if (num7 >= 255)
 					return;
@@ -60442,11 +_,12 @@
 						flag11 = false;
 						flag12 = false;
 						flag13 = false;
-						flag14 = (downedPlantBoss && Main.hardMode);
+						flag14 = (downedPlantBoss && Main.hardMode); //patch file: flag13
 						isItAHappyWindyDay = Main.IsItAHappyWindyDay;
 						if (Main.player[num7].active && Main.invasionType > 0 && Main.invasionDelay == 0 && Main.invasionSize > 0 && (double)Main.player[num7].position.Y < Main.worldSurface * 16.0 + (double)sHeight) {
 							int num8 = 3000;
 							if ((double)Main.player[num7].position.X > Main.invasionX * 16.0 - (double)num8 && (double)Main.player[num7].position.X < Main.invasionX * 16.0 + (double)num8) {
+							//patch file: flag5
 								flag6 = true;
 							}
 							else if (Main.invasionX >= (double)(Main.maxTilesX / 2 - 5) && Main.invasionX <= (double)(Main.maxTilesX / 2 + 5)) {
@@ -60470,6 +_,7 @@
 							flag5 = true;
 
 						if (Main.tile[num9, num10].wall == 87)
+							//patch file: flag3
 							flag4 = true;
 
 						flag2 = false;
@@ -60747,6 +_,7 @@
 								}
 							}
 						}
+						NPCLoader.EditSpawnRate(Main.player[num7], ref spawnRate, ref maxSpawns);
 
 						flag15 = false;
 						if (Main.player[num7].active && !Main.player[num7].dead && Main.player[num7].nearbyActiveNPCs < (float)maxSpawns && Main.rand.Next(spawnRate) == 0) {
@@ -60772,6 +_,7 @@
 								safeRangeY += (int)((double)(sHeight / 16) * 0.5 / (double)num12);
 							}
 
+							NPCLoader.EditSpawnRange(Main.player[num7], ref spawnRangeX, ref spawnRangeY, ref safeRangeX, ref safeRangeY);
 							int num13 = (int)(Main.player[num7].position.X / 16f) - spawnRangeX;
 							int num14 = (int)(Main.player[num7].position.X / 16f) + spawnRangeX;
 							int num15 = (int)(Main.player[num7].position.Y / 16f) - spawnRangeY;
@@ -60890,18 +_,23 @@
 								flag2 = false;
 
 							if (Main.tile[num, num2 - 1].liquid > 0 && Main.tile[num, num2 - 2].liquid > 0 && !Main.tile[num, num2 - 1].lava()) {
+								//patch file: flag7
 								if (Main.tile[num, num2 - 1].honey())
 									flag8 = true;
+								//patch file: flag6
 								else
 									flag7 = true;
 							}
 
 							int num29 = (int)Main.player[num7].Center.X / 16;
 							int num30 = (int)(Main.player[num7].Bottom.Y + 8f) / 16;
+							spawnInfo.playerFloorX = num29;
+							spawnInfo.playerFloorY = num30;
 							if (Main.tile[num, num2].type == 367) {
 								flag10 = true;
 							}
 							else if (Main.tile[num, num2].type == 368) {
+								//patch file: flag8
 								flag9 = true;
 							}
 							else if (Main.tile[num29, num30].type == 367) {
@@ -60986,6 +_,7 @@
 						for (int num40 = num - num39; num40 < num + num39; num40++) {
 							for (int num41 = num2 - num39; num41 < num2 + num39; num41++) {
 								if (Main.tile[num40, num41].wall == 62)
+									//patch file: flag10
 									flag11 = true;
 							}
 						}
@@ -61006,6 +_,7 @@
 						for (int num45 = num - num44; num45 < num + num44; num45++) {
 							for (int num46 = num2 - num44; num46 < num2 + num44; num46++) {
 								if (WallID.Sets.Conversion.Sandstone[Main.tile[num45, num46].wall] || WallID.Sets.Conversion.HardenedSand[Main.tile[num45, num46].wall])
+									//patch file: flag12
 									flag13 = true;
 							}
 						}
@@ -61027,9 +_,33 @@
 
 			bool flag17 = (float)new Point(num9 - num, num10 - num2).X * Main.windSpeedTarget > 0f;
 			spawnTileType = SpawnNPC_TryFindingProperGroundTileType(spawnTileType, num, num2);
+			spawnInfo.spawnTileX = num;
+			spawnInfo.spawnTileY = num2;
+			spawnInfo.spawnTileType = num3;
+			spawnInfo.player = Main.player[num7];
+			spawnInfo.sky = flag3;
+			spawnInfo.lihzahrd = flag4;
+			spawnInfo.playerSafe = flag5;
+			spawnInfo.invasion = flag6;
+			spawnInfo.water = flag7;
+			spawnInfo.granite = flag9;
+			spawnInfo.marble = flag10;
+			spawnInfo.spiderCave = flag11;
+			spawnInfo.playerInTown = flag12;
+			spawnInfo.desertCave = flag13;
+			spawnInfo.planteraDefeated = flag14;
+			spawnInfo.safeRangeX = flag15;
 			int newNPC = 200;
 			int cattailX;
 			int cattailY;
+			int? spawnChoice = NPCLoader.ChooseSpawn(spawnInfo);
+			if (!spawnChoice.HasValue)
+				return;
+
+			int spawn = spawnChoice.Value;
+			if (spawn != 0)
+				goto endVanillaSpawn;
+
 			if (Main.player[num7].ZoneTowerNebula) {
 				bool flag18 = true;
 				int num50 = 0;
@@ -63140,6 +_,10 @@
 				}
 			}
 
+			endVanillaSpawn: //this ugly code is just to minimize the diff file
+			if (spawn != 0)
+				newNPC = NPCLoader.SpawnNPC(spawn, num, num2);
+
 			if (Main.npc[newNPC].type == 1 && Main.player[num7].RollLuck(180) == 0)
 				Main.npc[newNPC].SetDefaults(-4);
 
@@ -63905,12 +_,14 @@
 			if (num2 < 0)
 				num2 = 0;
 
+			if (NPCLoader.StrikeNPC(this, ref num, num2, ref knockBack, hitDirection, ref crit)) {
-			num = Main.CalculateDamageNPCsTake((int)num, num2);
+				num = Main.CalculateDamageNPCsTake((int)num, num2);
-			if (crit)
+				if (crit)
-				num *= 2.0;
+					num *= 2.0;
 
-			if (takenDamageMultiplier > 1f)
+				if (takenDamageMultiplier > 1f)
-				num *= (double)takenDamageMultiplier;
+					num *= takenDamageMultiplier;
+			}
 
 			if ((takenDamageMultiplier > 1f || Damage != 9999) && lifeMax > 1) {
 				if (friendly) {
@@ -64185,6 +_,10 @@
 		}
 
 		public void HitEffect(int hitDirection = 0, double dmg = 10.0) {
+			NPCLoader.HitEffect(this, hitDirection, dmg);
+		}
+
+		public void VanillaHitEffect(int hitDirection = 0, double dmg = 10.0) {
 			if (!active)
 				return;
 
@@ -71202,6 +_,8 @@
 			return false;
 		}
 
+		public bool HasBuff(int type) => FindBuffIndex(type) != -1; //TODO: Make this an extension instead?
+
 		public int FindBuffIndex(int type) {
 			if (buffImmune[type])
 				return -1;
@@ -71228,6 +_,8 @@
 			int num = -1;
 			for (int i = 0; i < 5; i++) {
 				if (buffType[i] == type) {
+					if (BuffLoader.ReApply(type, this, time, i))
+						return;
 					if (buffTime[i] < time)
 						buffTime[i] = time;
 
@@ -71520,6 +_,7 @@
 			}
 
 			UpdateNPC_BuffFlagsReset();
+			NPCLoader.ResetEffects(this);
 			UpdateNPC_BuffSetFlags();
 			UpdateNPC_SoulDrainDebuff();
 			UpdateNPC_BuffClearExpiredBuffs();
@@ -72413,6 +_,7 @@
 					num = 5;
 			}
 
+			NPCLoader.UpdateLifeRegen(this, ref num);
 			if (lifeRegen <= -240 && num < 2)
 				num = 2;
 
@@ -72611,6 +_,7 @@
 
 					if (buffType[i] == 316)
 						markedByRainbowWhip = true;
+					BuffLoader.Update(buffType[i], this, ref i); // todo, move?
 				}
 			}
 		}
@@ -72812,7 +_,12 @@
 				if (acceptableNPCIDs[nPC.type] && nPC.active && !nPC.friendly && nPC.damage > 0) {
 					Rectangle npcRect = nPC.Hitbox;
 					GetMeleeCollisionData(hitbox, i, ref specialHitSetter, ref damageMultiplier, ref npcRect);
+					bool? modCanHit = NPCLoader.CanHitNPC(Main.npc[i], this);
+					if (modCanHit.HasValue && !modCanHit.Value)
+						continue;
+
+					// TODO: NPCLoader.CanHitNPC should return a plain bool. The only use of "override true" below is to force the skeleton merchant to be hit by skeletons
-					if (hitbox.Intersects(npcRect) && (type != 453 || !NPCID.Sets.Skeletons[nPC.type]) && nPC.type != 624)
+					if (hitbox.Intersects(npcRect) && ((modCanHit == true) || type != 453 || !NPCID.Sets.Skeletons[nPC.type])&& nPC.type != 624)
 						BeHurtByOtherNPC(i, nPC);
 				}
 			}
@@ -72824,14 +_,17 @@
 				num = 20;
 
 			int num2 = Main.DamageVar(thatNPC.damage);
-			int num3 = 6;
+			float num3 = 6;
 			int num4 = (!(thatNPC.Center.X > base.Center.X)) ? 1 : (-1);
-			double num5 = StrikeNPCNoInteraction(num2, num3, num4);
+			bool crit = false;
+			NPCLoader.ModifyHitNPC(thatNPC, this, ref num2, ref num3, ref crit);
+			double num5 = StrikeNPCNoInteraction(num2, num3, num4, crit, false, false);
 			if (Main.netMode != 0)
 				NetMessage.SendData(28, -1, -1, null, whoAmI, num2, num3, num4);
 
 			netUpdate = true;
 			immune[255] = num;
+			NPCLoader.OnHitNPC(thatNPC, this, (int)num5, num3, crit);
 			if (dryadWard) {
 				num2 = (int)num5 / 3;
 				num3 = 6;
@@ -73458,6 +_,10 @@
 			if (IsABestiaryIconDummy)
 				newColor = Color.White;
 
+			Color? modColor = NPCLoader.GetAlpha(this, newColor);
+			if (modColor.HasValue)
+				return modColor.Value;
+
 			float num = (float)(255 - alpha) / 255f;
 			int num2 = (int)((float)(int)newColor.R * num);
 			int num3 = (int)((float)(int)newColor.G * num);
@@ -74935,6 +_,7 @@
 				result = ((!HasSpecialEventText("Bunny", out specialEventText)) ? Lang.BunnyChat(this) : specialEventText);
 			}
 
+			NPCLoader.GetChat(this, ref result);
 			return result;
 		}
 
