--- src/Terraria/Terraria/PopupText.cs
+++ src/tModLoader/Terraria/PopupText.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
 using Terraria.GameContent;
+using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -25,6 +_,7 @@
 		public static int sonarText = -1;
 		public bool expert;
 		public bool master;
+		public int rarity;
 		public bool sonar;
 		public PopupTextContext context;
 		public int npcNetID;
@@ -250,10 +_,14 @@
 					popupText2.color = new Color(5, 200, 255);
 				else if (newItem.rare == 10)
 					popupText2.color = new Color(255, 40, 100);
-				else if (newItem.rare >= 11)
+				else if (newItem.rare == 11)
 					popupText2.color = new Color(180, 40, 255);
+				else if (newItem.rare >= ItemRarityID.Count)
+					popupText2.color = RarityLoader.GetRarity(newItem.rare).RarityColor;
 
+				popupText2.rarity = newItem.rare;
 				popupText2.expert = newItem.expert;
+				popupText2.master = newItem.master;
 				popupText2.name = newItem.AffixName();
 				popupText2.stack = stack;
 				popupText2.velocity.Y = -7f;
@@ -441,6 +_,9 @@
 				color = new Color((byte)Main.DiscoR, (byte)Main.DiscoG, (byte)Main.DiscoB, Main.mouseTextColor);
 			else if (master)
 				color = new Color(255, (byte)(Main.masterColor * 200f), 0, Main.mouseTextColor);
+
+			if (rarity > ItemRarityID.Purple)
+				color = RarityLoader.GetRarity(rarity).RarityColor;
 
 			bool flag = false;
 			Vector2 textHitbox = GetTextHitbox();
