--- src/Terraria/Terraria/PopupText.cs
+++ src/tModLoader/Terraria/PopupText.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
 using Terraria.GameContent;
+using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -25,6 +_,7 @@
 		public static int sonarText = -1;
 		public bool expert;
 		public bool master;
+		public int rarity;
 		public bool sonar;
 		public PopupTextContext context;
 		public int npcNetID;
@@ -184,7 +_,6 @@
 				Main.popupText[num2].position.X = newItem.position.X + (float)newItem.width * 0.5f - vector2.X * 0.5f;
 				Main.popupText[num2].position.Y = newItem.position.Y + (float)newItem.height * 0.25f - vector2.Y * 0.5f;
 				Main.popupText[num2].color = Color.White;
-				Main.popupText[num2].master = false;
 				if (newItem.rare == 1)
 					Main.popupText[num2].color = new Color(150, 150, 255);
 				else if (newItem.rare == 2)
@@ -195,8 +_,6 @@
 					Main.popupText[num2].color = new Color(255, 150, 150);
 				else if (newItem.rare == 5)
 					Main.popupText[num2].color = new Color(255, 150, 255);
-				else if (newItem.rare == -13)
-					Main.popupText[num2].master = true;
 				else if (newItem.rare == -11)
 					Main.popupText[num2].color = new Color(255, 175, 0);
 				else if (newItem.rare == -1)
@@ -211,10 +_,14 @@
 					Main.popupText[num2].color = new Color(5, 200, 255);
 				else if (newItem.rare == 10)
 					Main.popupText[num2].color = new Color(255, 40, 100);
-				else if (newItem.rare >= 11)
+				else if (newItem.rare == 11)
 					Main.popupText[num2].color = new Color(180, 40, 255);
+				else if (newItem.rare >= ItemRarityID.Count)
+					Main.popupText[num2].color = RarityLoader.GetRarity(newItem.rare).RarityColor;
 
+				Main.popupText[num2].rarity = newItem.rare;
 				Main.popupText[num2].expert = newItem.expert;
+				Main.popupText[num2].master = newItem.master;
 				Main.popupText[num2].name = newItem.AffixName();
 				Main.popupText[num2].stack = stack;
 				Main.popupText[num2].velocity.Y = -7f;
@@ -398,6 +_,9 @@
 				color = new Color((byte)Main.DiscoR, (byte)Main.DiscoG, (byte)Main.DiscoB, Main.mouseTextColor);
 			else if (master)
 				color = new Color(255, (byte)(Main.masterColor * 200f), 0, Main.mouseTextColor);
+
+			if (rarity > ItemRarityID.Purple)
+				color = RarityLoader.GetRarity(rarity).RarityColor;
 
 			bool flag = false;
 			string text = name;
