--- src/Terraria/Terraria/Program.cs
+++ src/tModLoader/Terraria/Program.cs
@@ -4,6 +_,7 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Text;
@@ -11,12 +_,15 @@
 using System.Windows.Forms;
 using Terraria.Initializers;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Core;
+using Terraria.ModLoader.Exceptions;
 using Terraria.Social;
 using Terraria.Utilities;
 
 namespace Terraria
 {
-	public static class Program
+	public static partial class Program
 	{
 		/* hide unused consts
 		public const bool IsServer = false;
@@ -55,21 +_,15 @@
 			LoadedEverything = true;
 		}
 
-		private static void ForceJITOnAssembly(Assembly assembly) {
-			Type[] types = assembly.GetTypes();
-			for (int i = 0; i < types.Length; i++) {
-#if WINDOWS || NETCORE
-				MethodInfo[] methods = types[i].GetMethods(BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
-#else
-				MethodInfo[] methods = types[i].GetMethods();
-#endif
-				foreach (MethodInfo methodInfo in methods) {
-					if (!methodInfo.IsAbstract && !methodInfo.ContainsGenericParameters && methodInfo.GetMethodBody() != null)
-#if WINDOWS || NETCORE
-						RuntimeHelpers.PrepareMethod(methodInfo.MethodHandle);
-#else
-						JitForcedMethodCache = methodInfo.MethodHandle.GetFunctionPointer();
-#endif
+		private static void ForceJITOnAssembly(IEnumerable<Type> types) {
+			var methodsToJIT = CollectMethodsToJIT(types);
+
+			if (Environment.ProcessorCount > 1) {
+				methodsToJIT.AsParallel().AsUnordered().ForAll(ForceJITOnMethod);
+			}
+			else {
+				foreach (var method in methodsToJIT) {
+					ForceJITOnMethod(method);
 				}
 
 				ThingsLoaded++;
@@ -85,10 +_,13 @@
 		}
 
 		private static void ForceLoadAssembly(Assembly assembly, bool initializeStaticMembers) {
-			ThingsToLoad = assembly.GetTypes().Length;
-			ForceJITOnAssembly(assembly);
-			if (initializeStaticMembers)
-				ForceStaticInitializers(assembly);
+			var types = assembly.GetTypes();
+			ThingsToLoad = types.Select(type => GetAllMethods(type).Count()).Sum();
+
+			ForceJITOnAssembly(types);
+			if (initializeStaticMembers) {
+				ForceStaticInitializers(types);
+			}
 		}
 
 		private static void ForceLoadAssembly(string name, bool initializeStaticMembers) {
@@ -124,8 +_,11 @@
 		}
 
 		private static void InitializeConsoleOutput() {
-			if (!Main.dedServ && Debugger.IsAttached)
+			if (!Main.dedServ && (Debugger.IsAttached || LaunchParameters.ContainsKey("-console")))
 				return;
+
+			if (!Main.dedServ)
+				Platform.Get<IWindowService>().HideConsole();
 
 			try {
 				Console.OutputEncoding = Encoding.UTF8;
@@ -138,32 +_,59 @@
 			}
 		}
 
+		// some versions of the .NET runtime will run the Main static initilizer as soon as LaunchGame is invoked
+		// causing Main.SavePath to be initialized before LaunchParameters is parsed.
+		// moving arg parsing to a separate function avoids this
 		public static void LaunchGame(string[] args, bool monoArgs = false) {
 			Thread.CurrentThread.Name = "Main Thread";
 			if (monoArgs)
 				args = Utils.ConvertMonoArgsToDotNet(args);
 
+			LaunchParameters = Utils.ParseArguements(args);
+			SavePath =
+				LaunchParameters.ContainsKey("-tmlsavedirectory") ? LaunchParameters["-tmlsavedirectory"] :
+				LaunchParameters.ContainsKey("-savedirectory") ? Path.Combine(LaunchParameters["-savedirectory"], "ModLoader") :
+				Platform.Get<IPathService>().GetStoragePath($"Terraria{Path.DirectorySeparatorChar}ModLoader{(!BuildInfo.IsRelease ? Path.DirectorySeparatorChar + "Beta" : "")}");
+
+			if (File.Exists("savehere.txt"))
+				SavePath = "ModLoader"; // Fallback for unresolveable antivirus/onedrive issues. Also makes the game portable I guess.
+
+			bool dedServ = LaunchParameters.ContainsKey("-server");
+			try {
+				Logging.Init(dedServ);
+			}
+			catch (Exception e) {
+				DisplayException(e);
+				return;
+			}
+
+			if (LaunchInitializer.TryParameter("-build") is string build)
+				// Inside the build from command line, the application exits
+				// Anything beyond this point will not be executed
+				ModCompile.BuildModCommandLine(build);
+
+			LaunchGame_(dedServ);
+		}
+
+		public static void LaunchGame_(bool dedServ) {
 			if (Platform.IsOSX) {
 				Main.OnEngineLoad += delegate {
 					Main.instance.IsMouseVisible = false;
 				};
 			}
 
-			LaunchParameters = Utils.ParseArguements(args);
-#if SERVER
-			Main.dedServ = true;
+			Main.dedServ = dedServ;
-#else
-			Main.dedServ = LaunchParameters.ContainsKey("-server");
-#endif
 			ThreadPool.SetMinThreads(8, 8);
 			LanguageManager.Instance.SetLanguage(GameCulture.DefaultCulture);
 			InitializeConsoleOutput();
-			SetupLogging();
+			//Program.SetupLogging();
 			Platform.Get<IWindowService>().SetQuickEditEnabled(enabled: false);
-			using (Main main = new Main()) {
-				try {
+			try {
+				ModLoader.ModLoader.EngineInit();
+				using (Main main = new Main()) {
 					Lang.InitializeLegacyLocalization();
 					SocialAPI.Initialize();
+					main.PostSocialInitialize();
 					LaunchInitializer.LoadParameters(main);
 					Main.OnEnginePreload += StartForceLoad;
 					if (!Main.dedServ)
@@ -171,9 +_,10 @@
 					else
 						main.DedServ();
 				}
-				catch (Exception e) {
-					DisplayException(e);
-				}
+			}
+			catch (Exception e) {
+				Logging.Terraria.Fatal("Main engine crash", e);
+				DisplayException(e);
 			}
 		}
 
@@ -208,9 +_,22 @@
 				}
 
 				if (Main.dedServ)
-					Console.WriteLine(Language.GetTextValue("Error.ServerCrash"), DateTime.Now, e);
+				Console.WriteLine(Language.GetTextValue("tModLoader.ServerCrash"), DateTime.Now, e);
 
-				MessageBox.Show(text, "Terraria: Error");
+				ModLoader.UI.Interface.MessageBoxShow(e, generateTip: true);
+
+				if (e is FolderCreationFailedException)
+					Process.Start(e.HelpLink);
+
+				if (e.Message.Contains("CSteamworks"))
+					Process.Start(@"https://github.com/tModLoader/tModLoader/wiki/Basic-tModLoader-Usage-FAQ#systemdllnotfoundexception-unable-to-load-dllcsteamworks");
+
+				if (e.StackTrace.Contains("LaunchLocalServer")) {
+					ModLoader.UI.Interface.MessageBoxShow(Language.GetTextValue("tModLoader.MissingServerExecutable"));
+
+					Process.Start(@"https://github.com/tModLoader/tModLoader/wiki/Basic-tModLoader-Usage-FAQ#host-and-play-the-system-cannot-find-the-file-specified");
+				}
+
 			}
 			catch {
 			}
