--- src/Terraria/Terraria/Program.cs
+++ src/tModLoader/Terraria/Program.cs
@@ -6,6 +_,7 @@
 using System.Diagnostics;
 #endif
 using System.IO;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Text;
@@ -15,12 +_,14 @@
 #endif
 using Terraria.Initializers;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
 using Terraria.Social;
 using Terraria.Utilities;
 
 namespace Terraria
 {
-	public static class Program
+	public static partial class Program
 	{
 		/* hide unused consts
 		public const bool IsServer = false;
@@ -89,10 +_,13 @@
 		}
 
 		private static void ForceLoadAssembly(Assembly assembly, bool initializeStaticMembers) {
-			ThingsToLoad = assembly.GetTypes().Length;
-			ForceJITOnAssembly(assembly);
-			if (initializeStaticMembers)
-				ForceStaticInitializers(assembly);
+			var types = assembly.GetTypes();
+			ThingsToLoad = types.Select(type => GetMethodsCrossPlatform(type).Count()).Sum();
+
+			ForceJITOnAssembly(types);
+			if (initializeStaticMembers) {
+				ForceStaticInitializers(types);
+			}
 		}
 
 		private static void ForceLoadAssembly(string name, bool initializeStaticMembers) {
@@ -144,27 +_,52 @@
 			}
 		}
 
+		// some versions of the .NET runtime will run the Main static initilizer as soon as LaunchGame is invoked
+		// causing Main.SavePath to be initialized before LaunchParameters is parsed.
+		// moving arg parsing to a separate function avoids this
 		public static void LaunchGame(string[] args, bool monoArgs = false) {
 			Thread.CurrentThread.Name = "Main Thread";
 			if (monoArgs)
 				args = Utils.ConvertMonoArgsToDotNet(args);
 
+			LaunchParameters = Utils.ParseArguements(args);
+			SavePath =
+				LaunchParameters.ContainsKey("-tmlsavedirectory") ? LaunchParameters["-tmlsavedirectory"] :
+				LaunchParameters.ContainsKey("-savedirectory") ? Path.Combine(LaunchParameters["-savedirectory"], "ModLoader") :
+				Platform.Get<IPathService>().GetStoragePath($"Terraria{Path.DirectorySeparatorChar}ModLoader{(ModLoader.ModLoader.IsBeta ? Path.DirectorySeparatorChar + "Beta" : "")}");
+
+			if (File.Exists("savehere.txt"))
+				SavePath = "ModLoader"; // Fallback for unresolveable antivirus/onedrive issues. Also makes the game portable I guess.
+
+			try {
+				Logging.Init();
+			}
+			catch (Exception e) {
+				DisplayException(e);
+				return;
+			}
+
+			LaunchGame_();
+		}
+
+		public static void LaunchGame_() {
 			if (Platform.IsOSX) {
 				Main.OnEngineLoad += delegate {
 					Main.instance.IsMouseVisible = false;
 				};
 			}
 
-			LaunchParameters = Utils.ParseArguements(args);
 			ThreadPool.SetMinThreads(8, 8);
 			LanguageManager.Instance.SetLanguage(GameCulture.DefaultCulture);
 			InitializeConsoleOutput();
-			SetupLogging();
+			//Program.SetupLogging();
 			Platform.Get<IWindowService>().SetQuickEditEnabled(enabled: false);
-			using (Main main = new Main()) {
-				try {
+			try {
+				ModLoader.ModLoader.EngineInit();
+				using (Main main = new Main()) {
 					Lang.InitializeLegacyLocalization();
 					SocialAPI.Initialize();
+					main.PostSocialInitialize();
 					LaunchInitializer.LoadParameters(main);
 					Main.OnEnginePreload += StartForceLoad;
 #if CLIENT
@@ -173,9 +_,10 @@
 					main.DedServ();
 #endif
 				}
-				catch (Exception e) {
-					DisplayException(e);
-				}
+			}
+			catch (Exception e) {
+				Logging.Terraria.Fatal("Main engine crash", e);
+				DisplayException(e);
 			}
 		}
 
@@ -209,10 +_,22 @@
 					streamWriter.WriteLine("");
 				}
 
-#if CLIENT && WINDOWS
-				MessageBox.Show(e.ToString(), Language.GetTextValue("Error.Error"));
+#if CLIENT
+				ModLoader.UI.Interface.MessageBoxShow(e, generateTip: true);
+
+				if (e is FolderCreationFailedException)
+					Process.Start(e.HelpLink);
+
+				if (e.Message.Contains("CSteamworks"))
+					Process.Start(@"https://github.com/tModLoader/tModLoader/wiki/Basic-tModLoader-Usage-FAQ#systemdllnotfoundexception-unable-to-load-dllcsteamworks");
+
+				if (e.StackTrace.Contains("LaunchLocalServer")) {
+					ModLoader.UI.Interface.MessageBoxShow(Language.GetTextValue("tModLoader.MissingServerExecutable"));
+
+					Process.Start(@"https://github.com/tModLoader/tModLoader/wiki/Basic-tModLoader-Usage-FAQ#host-and-play-the-system-cannot-find-the-file-specified");
+				}
 #else
-				Console.WriteLine(Language.GetTextValue("Error.ServerCrash"), DateTime.Now, e);
+				Console.WriteLine(Language.GetTextValue("tModLoader.ServerCrash"), DateTime.Now, e);
 #endif
 			}
 			catch {
