--- src/Terraria\Terraria\Program.cs
+++ src/tModLoader\Terraria\Program.cs
@@ -18,6 +_,7 @@
 #endif
 using Terraria.Initializers;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -205,8 +_,12 @@
 #endif
 			try
 			{
-				Console.OutputEncoding = Encoding.Unicode;
-				Console.InputEncoding = Encoding.Unicode;
+				Console.OutputEncoding = Encoding.UTF8;
+
+				if (Platform.IsWindows)
+					Console.InputEncoding = Encoding.Unicode;
+				else
+					Console.InputEncoding = Encoding.UTF8;
 			}
 			catch
 			{
@@ -227,6 +_,15 @@
 				};
 			}
 			Program.LaunchParameters = Utils.ParseArguements(args);
+			if (Program.LaunchParameters.ContainsKey("-savedirectory"))
+			{
+				Program.LaunchParameters["-savedirectory"] = Path.Combine(Program.LaunchParameters["-savedirectory"], "ModLoader");
+			}
+			if (LaunchParameters.ContainsKey("-build"))
+			{
+				ModCompile.BuildModCommandLine(LaunchParameters["-build"]);
+				return;
+			}
 			ThreadPool.SetMinThreads(8, 8);
 			LanguageManager.Instance.SetLanguage(GameCulture.English);
 			Program.SetupLogging();

