--- src/Terraria\Terraria\Projectile.cs
+++ src/tModLoader\Terraria\Projectile.cs
@@ -5,11 +5,18 @@
 using Terraria.GameContent.Achievements;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
 	public class Projectile : Entity
 	{
+		public ModProjectile modProjectile
+		{
+			get;
+			internal set;
+		}
+
 		public bool arrow;
 		public int numHits;
 		public bool bobber;
@@ -98,6 +105,7 @@
 
 		public void SetDefaults(int Type)
 		{
+			this.modProjectile = null;
 			this.counterweight = false;
 			this.arrow = false;
 			this.bobber = false;
@@ -6930,10 +6938,7 @@
 				this.tileCollide = false;
 				this.manualDirectionChange = true;
 			}
-			else
-			{
-				this.active = false;
-			}
+			ProjectileLoader.SetupProjectile(this);
 			this.width = (int)((float)this.width * this.scale);
 			this.height = (int)((float)this.height * this.scale);
 			this.maxPenetrate = this.penetrate;
@@ -11954,6 +11959,8 @@
 						return 0.31f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.03f;
 					case "Random":
 						return Main.rand.NextFloat();
+					case "bluemagic123":
+						return 0.55f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.1f;
 				}
 			}
 			return (float)((int)indexing) / 6f;
@@ -12009,6 +12016,11 @@
 		}
 
 		public void AI()
+		{
+			ProjectileLoader.ProjectileAI(this);
+		}
+
+		public void VanillaAI()
 		{
 			if (this.aiStyle == 1)
 			{

