--- src/Terraria\Terraria\Recipe.cs
+++ src/tModLoader\Terraria\Recipe.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -39,7 +_,7 @@
 		{
 			foreach (int current in this.acceptedGroups)
 			{
-				if (RecipeGroup.recipeGroups[current].ValidItems.Contains(type))
+				if (RecipeGroup.recipeGroups[current].ContainsItem(type))
 				{
 					theText = RecipeGroup.recipeGroups[current].GetText();
 					return true;
@@ -53,7 +_,7 @@
 		{
 			foreach (int current in this.acceptedGroups)
 			{
-				if (RecipeGroup.recipeGroups[current].ValidItems.Contains(invType) && RecipeGroup.recipeGroups[current].ValidItems.Contains(reqType))
+				if (RecipeGroup.recipeGroups[current].ContainsItem(invType) && RecipeGroup.recipeGroups[current].ContainsItem(reqType))
 				{
 					return true;
 				}
@@ -80,6 +_,11 @@
 					break;
 				}
 				int num = item.stack;
+				ModRecipe modRecipe = this as ModRecipe;
+				if (modRecipe != null)
+				{
+					num = modRecipe.ConsumeItem(item.type, item.stack);
+				}
 				if (this.alchemy && Main.player[Main.myPlayer].alchemyTable)
 				{
 					if (num > 1)
@@ -184,78 +_,26 @@
 			{
 				return false;
 			}
-			if (reqType <= 621)
-			{
-				if (reqType == 9)
-				{
-					goto IL_59;
-				}
-				switch (reqType)
-				{
-					case 619:
-					case 620:
-					case 621:
-						goto IL_59;
-				}
-			}
-			else
-			{
-				if (reqType == 911 || reqType == 1729)
-				{
-					goto IL_59;
-				}
-				switch (reqType)
-				{
-					case 2503:
-					case 2504:
-						goto IL_59;
-				}
-			}
-			return false;
-			IL_59:
-			if (invType <= 621)
-			{
-				if (invType == 9)
-				{
-					return true;
-				}
-				switch (invType)
-				{
-					case 619:
-					case 620:
-					case 621:
-						return true;
-				}
-			}
-			else
-			{
-				if (invType == 911 || invType == 1729)
-				{
-					return true;
-				}
-				switch (invType)
-				{
-					case 2503:
-					case 2504:
-						return true;
-				}
-			}
-			return false;
+			RecipeGroup rec = RecipeGroup.recipeGroups[RecipeGroupID.Wood];
+			return rec.ContainsItem(invType) && rec.ContainsItem(reqType);
 		}
 
 		public bool useIronBar(int invType, int reqType)
 		{
-			return this.anyIronBar && (reqType == 22 || reqType == 704) && (invType == 22 || invType == 704);
+			RecipeGroup rec = RecipeGroup.recipeGroups[RecipeGroupID.IronBar];
+			return this.anyIronBar && rec.ContainsItem(invType) && rec.ContainsItem(reqType);
 		}
 
 		public bool useSand(int invType, int reqType)
 		{
-			return (reqType == 169 || reqType == 408 || reqType == 1246 || reqType == 370 || reqType == 3272) && (this.anySand && (invType == 169 || invType == 408 || invType == 1246 || invType == 370 || invType == 3272));
+			RecipeGroup rec = RecipeGroup.recipeGroups[RecipeGroupID.Sand];
+			return this.anySand && rec.ContainsItem(invType) && rec.ContainsItem(reqType);
 		}
 
 		public bool useFragment(int invType, int reqType)
 		{
-			return (reqType == 3458 || reqType == 3456 || reqType == 3457 || reqType == 3459) && (this.anyFragment && (invType == 3458 || invType == 3456 || invType == 3457 || invType == 3459));
+			RecipeGroup rec = RecipeGroup.recipeGroups[RecipeGroupID.Fragment];
+			return this.anyFragment && rec.ContainsItem(invType) && rec.ContainsItem(reqType);
 		}
 
 		public bool usePressurePlate(int invType, int reqType)
@@ -264,67 +_,8 @@
 			{
 				return false;
 			}
-			if (reqType <= 543)
-			{
-				if (reqType == 529)
-				{
-					goto IL_54;
-				}
-				switch (reqType)
-				{
-					case 541:
-					case 542:
-					case 543:
-						goto IL_54;
-				}
-			}
-			else
-			{
-				switch (reqType)
-				{
-					case 852:
-					case 853:
-						goto IL_54;
-					default:
-						if (reqType == 1151)
-						{
-							goto IL_54;
-						}
-						break;
-				}
-			}
-			return false;
-			IL_54:
-			if (invType <= 543)
-			{
-				if (invType == 529)
-				{
-					return true;
-				}
-				switch (invType)
-				{
-					case 541:
-					case 542:
-					case 543:
-						return true;
-				}
-			}
-			else
-			{
-				switch (invType)
-				{
-					case 852:
-					case 853:
-						return true;
-					default:
-						if (invType == 1151)
-						{
-							return true;
-						}
-						break;
-				}
-			}
-			return false;
+			RecipeGroup rec = RecipeGroup.recipeGroups[RecipeGroupID.PressurePlate];
+			return rec.ContainsItem(invType) && rec.ContainsItem(reqType);
 		}
 
 		public static void FindRecipes()
@@ -474,7 +_,7 @@
 							flag2 = false;
 						}
 					}
-					if (flag2)
+					if (flag2 && RecipeHooks.RecipeAvailable(Main.recipe[num4]))
 					{
 						Main.availableRecipe[Main.numAvailableRecipes] = num4;
 						Main.numAvailableRecipes++;
@@ -563,6 +_,8 @@
 					2007
 				});
 			RecipeGroupID.Snails = RecipeGroup.RegisterGroup("Snails", rec);
+			RecipeGroupHelper.AddOldVanillaGroups();
+			RecipeGroupHelper.AddRecipeGroups();
 		}
 
 		public static void SetupRecipes()
@@ -10918,6 +_,9 @@
 			Recipe.AddRecipe();
 			Recipe.WallReturn();
 			Recipe.PlatformReturn();
+			RecipeHooks.AddRecipes();
+			RecipeHooks.PostAddRecipes();
+			Item.PopulateMaterialCache();
 			for (int num6 = 0; num6 < Recipe.numRecipes; num6++)
 			{
 				int num7 = 0;

