--- src/Terraria\Terraria\Recipe.cs
+++ src/tModLoader\Terraria\Recipe.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -79,6 +_,11 @@
 					break;
 				}
 				int num = item.stack;
+				ModRecipe modRecipe = this as ModRecipe;
+				if (modRecipe != null)
+				{
+					num = modRecipe.ConsumeItem(item.type, item.stack);
+				}
 				if (this.alchemy && Main.player[Main.myPlayer].alchemyTable)
 				{
 					if (num > 1)
@@ -464,7 +_,7 @@
 							flag2 = false;
 						}
 					}
-					if (flag2)
+					if (flag2 && RecipeHooks.RecipeAvailable(Main.recipe[num4]))
 					{
 						Main.availableRecipe[Main.numAvailableRecipes] = num4;
 						Main.numAvailableRecipes++;
@@ -497,6 +_,10 @@
 
 		public static void SetupRecipeGroups()
 		{
+			if (RecipeGroup.nextRecipeGroupIndex > 0)
+			{
+				return;
+			}
 			RecipeGroup rec = new RecipeGroup(() => Lang.misc[37] + " " + Main.npcName[74], new int[]
 				{
 					2015,
@@ -10563,6 +_,7 @@
 			Recipe.AddRecipe();
 			Recipe.WallReturn();
 			Recipe.PlatformReturn();
+			ModLoader.ModLoader.AddRecipes();
 			for (int num4 = 0; num4 < Recipe.numRecipes; num4++)
 			{
 				int num5 = 0;

