--- src/Terraria\Terraria\Recipe.cs
+++ src/tModLoader\Terraria\Recipe.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -79,6 +_,11 @@
 					break;
 				}
 				int num = item.stack;
+				ModRecipe modRecipe = this as ModRecipe;
+				if (modRecipe != null)
+				{
+					num = modRecipe.ConsumeItem(item.type, item.stack);
+				}
 				if (this.alchemy && Main.player[Main.myPlayer].alchemyTable)
 				{
 					if (num > 1)
@@ -464,7 +_,7 @@
 							flag2 = false;
 						}
 					}
-					if (flag2)
+					if (flag2 && RecipeHooks.RecipeAvailable(Main.recipe[num4]))
 					{
 						Main.availableRecipe[Main.numAvailableRecipes] = num4;
 						Main.numAvailableRecipes++;
@@ -553,6 +_,8 @@
 					2007
 				});
 			RecipeGroupID.Snails = RecipeGroup.RegisterGroup("Snails", rec);
+			RecipeGroupHelper.AddOldVanillaGroups();
+			RecipeGroupHelper.AddRecipeGroups();
 		}
 
 		public static void SetupRecipes()
@@ -10568,6 +_,7 @@
 			Recipe.AddRecipe();
 			Recipe.WallReturn();
 			Recipe.PlatformReturn();
+			RecipeHooks.AddRecipes();
 			for (int num4 = 0; num4 < Recipe.numRecipes; num4++)
 			{
 				int num5 = 0;

