--- src/TerrariaNetCore/Terraria/RecipeGroup.cs
+++ src/tModLoader/Terraria/RecipeGroup.cs
@@ -58,6 +_,16 @@
 		if (RegisteredId >= 0)
 			throw new Exception("Already registered");
 
+		// TML: Handle registering to an existing group by merging items and keeping the original group.
+		var existingRecipeGroup = recipeGroups.Values.FirstOrDefault(x => x.GetPlaceholderItemType() == this.GetPlaceholderItemType());
+		if (existingRecipeGroup != null) {
+			foreach (var item in this.Items) {
+				if(!existingRecipeGroup.ValidItems.Contains(item))
+					existingRecipeGroup.Add(item);
+			}
+			return existingRecipeGroup;
+		}
+
 		int key = (RegisteredId = nextRecipeGroupIndex++);
 		recipeGroups.Add(key, this);
 		return this;
