--- src/TerrariaNetCore/Terraria/RecipeGroup.cs
+++ src/tModLoader/Terraria/RecipeGroup.cs
@@ -16,6 +_,8 @@
 	public static Dictionary<int, RecipeGroup> recipeGroups = new Dictionary<int, RecipeGroup>();
 	public static int nextRecipeGroupIndex;
 
+	internal bool[] ValidItemsLookup; // Added by TML.
+
 	public int RegisteredId { get; private set; }
 
 	private static Func<string> WithDefaultCombineFormat(string key)
@@ -56,6 +_,14 @@
 		if (RegisteredId >= 0)
 			throw new Exception("Already registered");
 
+		// TML: Handle registering to an existing group by merging items and keeping the original group.
+		foreach (var kvp in recipeGroups) {
+			if (kvp.Value.GetType() == this.GetType() && ReferenceEquals(kvp.Value.GetText, this.GetText)) {
+				kvp.Value.ValidItems.UnionWith(this.ValidItems);
+				return kvp.Value;
+			}
+		}
+
 		int key = (RegisteredId = nextRecipeGroupIndex++);
 		recipeGroups.Add(key, this);
 		return this;
@@ -76,4 +_,7 @@
 	public int GetGroupFakeItemId() => RegisteredId + FakeItemIdOffset;
 	public bool Contains(int itemType) => ValidItems.Contains(itemType);
 	public int GetPlaceholderItemType() => Items[0];
+
+	// Added by TML.
+	public bool ContainsItem(int type) => ValidItemsLookup?[type] ?? ValidItems.Contains(type);
 }
