--- src/TerrariaNetCore/Terraria/RecipeGroup.cs
+++ src/tModLoader/Terraria/RecipeGroup.cs
@@ -57,6 +_,12 @@
 	{
 		if (RegisteredId >= 0)
 			throw new Exception("Already registered");
+		
+		// TML: Handle registering to an existing group by merging items and keeping the original group.
+		if (recipeGroups.TryGetValue(name, out int existingRecipeGroup)) {
+			recipeGroups[existingRecipeGroup].ValidItems.UnionWith(rec.ValidItems);
+			return existingRecipeGroup;
+		}
 
 		int key = (RegisteredId = nextRecipeGroupIndex++);
 		recipeGroups.Add(key, this);
