--- src/TerrariaNetCore/Terraria/Recipe.cs
+++ src/tModLoader/Terraria/Recipe.cs
@@ -1,15 +_,19 @@
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text;
 using Terraria.GameContent;
 using Terraria.GameContent.UI;
 using Terraria.ID;
 using Terraria.Map;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
+using Terraria.ModLoader.Utilities;
 using Terraria.UI;
 
 namespace Terraria;
 
-public class Recipe
+public sealed partial class Recipe
 {
 	public struct RequiredItemEntry
 	{
@@ -50,32 +_,44 @@
 		}
 	}
 
-	public static int maxRequirements = 15;
+	internal static int maxVanillaRequirements = 15;
 	public static int maxRecipes = 3600;
 	public static int numRecipes;
 	public static bool[] TileUsedInRecipes = new bool[TileID.Count];
 	public static List<int>[] TileCountsAs = new List<int>[TileID.Count];
-	private static Recipe currentRecipe = new Recipe();
+	private static Recipe currentRecipe;
 	public Item createItem = new Item();
-	public Item[] requiredItem = new Item[maxRequirements];
+	// required... changed from [maxRequirements] to List
+	public List<Item> requiredItem = new();
 	public int requiredTile = -1;
-	public int[] acceptedGroups = new int[maxRequirements];
-	public RequiredItemEntry[] requiredItemQuickLookup = new RequiredItemEntry[maxRequirements];
+	public List<int> acceptedGroups = new();
+	internal RequiredItemEntry[] requiredItemQuickLookup; // TML private->internal
 	public List<Item> customShimmerResults;
+
+	// Following fields made internal, automatically replaced with corresponding Conditions
-	public bool needHoney;
+	internal bool needHoney;
-	public bool needWater;
+	internal bool needWater;
-	public bool needLava;
+	internal bool needLava;
-	public bool needTorchGodsFavor;
+	internal bool needTorchGodsFavor;
-	public bool alchemy;
+	//internal bool alchemy; TML: Recipe.alchemy is unused due to being replaced by consume item hooks
-	public bool needSnowBiome;
+	internal bool needSnowBiome;
-	public bool needGraveyardBiome;
+	internal bool needGraveyardBiome;
-	public bool needMechdusa;
+	internal bool needMechdusa;
+	// Following fields made internal, use DisableDecraft() and DecraftDisabled
-	public bool notDecraftable;
+	internal bool notDecraftable;
+	// Following fields made internal, automatically replaced with corresponding Decrafting Conditions
-	public bool crimson;
+	internal bool crimson;
-	public bool corruption;
+	internal bool corruption;
+
 	private static Dictionary<int, int> _ownedItems = new Dictionary<int, int>();
 	internal static List<Chest> _recipeChests = new List<Chest>();
 
+	// Added by TML.
+	static Recipe()
+	{
+		ResetCurrentRecipe();
+	}
+
 	public bool DoesNotNeedTileOrLiquid {
 		get {
 			if (requiredTile < 0 && !needHoney && !needWater)
@@ -89,7 +_,7 @@
 	{
 		StringBuilder stringBuilder = new StringBuilder();
 		stringBuilder.AppendLine(createItem.ToString());
-		Item[] array = requiredItem;
+		var array = requiredItem;
 		foreach (Item item in array) {
 			if (item.IsAir)
 				break;
@@ -97,7 +_,7 @@
 			stringBuilder.AppendLine("  " + item.ToString());
 		}
 
-		int[] array2 = acceptedGroups;
+		var array2 = acceptedGroups;
 		foreach (int num in array2) {
 			if (num < 0)
 				break;
@@ -108,9 +_,9 @@
 		return stringBuilder.ToString();
 	}
 
-	public void RequireGroup(RecipeGroup group)
+	private void RequireGroup(RecipeGroup group) // TML: private
 	{
-		for (int i = 0; i < maxRequirements; i++) {
+		for (int i = 0; i < acceptedGroups.Count; i++) {
 			if (acceptedGroups[i] == -1) {
 				acceptedGroups[i] = group.RegisteredId;
 				break;
@@ -120,12 +_,8 @@
 
 	public bool ProcessGroupsForText(int type, out string theText)
 	{
-		for (int i = 0; i < maxRequirements; i++) {
-			int num = acceptedGroups[i];
-			if (num == -1)
-				break;
-
-			if (RecipeGroup.recipeGroups[num].ValidItems.Contains(type)) {
+		foreach (int num in acceptedGroups) {
+			if (RecipeGroup.recipeGroups[num].Contains(type)) {
 				theText = RecipeGroup.recipeGroups[num].GetText();
 				return true;
 			}
@@ -135,7 +_,7 @@
 		return false;
 	}
 
-	public Item AddCustomShimmerResult(int itemType, int itemStack = 1)
+	public Recipe AddCustomShimmerResult(int itemType, int itemStack = 1)
 	{
 		if (customShimmerResults == null)
 			customShimmerResults = new List<Item>();
@@ -144,9 +_,10 @@
 		item.SetDefaults(itemType);
 		item.stack = itemStack;
 		customShimmerResults.Add(item);
-		return item;
+		return this;
 	}
 
+	/*
 	public Recipe()
 	{
 		for (int i = 0; i < maxRequirements; i++) {
@@ -154,10 +_,18 @@
 			acceptedGroups[i] = -1;
 		}
 	}
+	*/
+
+	// Added by TML, replaces the removed public constructor.
+	internal Recipe(Mod mod = null)
+	{
+		Mod = mod;
+		RecipeIndex = -1;
+	}
 
 	public void GetIngredientsForOneCraft(Player player, List<RequiredItemEntry> ingredients)
 	{
-		for (int i = 0; i < maxRequirements; i++) {
+		for (int i = 0; i < requiredItem.Count; i++) {
 			RequiredItemEntry requiredItemEntry = requiredItemQuickLookup[i];
 			if (requiredItemEntry.itemIdOrRecipeGroup != 0) {
 				requiredItemEntry.stack -= GetIngredientCraftingDiscount(player, requiredItemEntry);
@@ -173,6 +_,7 @@
 
 	private int GetIngredientCraftingDiscount(Player player, RequiredItemEntry req)
 	{
+		/*
 		int num = 0;
 		if (alchemy && player.alchemyTable) {
 			for (int i = 0; i < req.stack; i++) {
@@ -182,6 +_,9 @@
 		}
 
 		return num;
+		*/
+
+		return 0; // TODO, reimplement crafting discounts in terms of RequiredItemEntry
 	}
 
 	public static void UpdateRecipeList()
@@ -204,7 +_,14 @@
 			if (recipe.createItem.type == 0)
 				break;
 
+			if (recipe.Disabled)
+				continue;
+
+			bool requirementsMet = false;
 			if ((recipeFilterHack == null || recipeFilterHack.Accepts(recipe)) && recipe.PlayerMeetsEnvironmentConditions(localPlayer) && CollectedEnoughItemsToCraft(recipe))
+				requirementsMet = true;
+
+			if (requirementsMet && RecipeLoader.RecipeAvailable(recipe))
 				AddToAvailableRecipes(i);
 		}
 
@@ -297,9 +_,35 @@
 		CollectItems(player.inventory, 58);
 		CollectItemsFromChests(player);
 		AddFakeCountsForItemGroups();
+
+		foreach ((var items, _) in PlayerLoader.GetModdedCraftingMaterials(Main.LocalPlayer)) {
+			CollectItems(items);
+		}
+
 		CraftingRequests.SubtractPendingRequests();
 	}
 
+	// Added by TML. Replaces multiple vanilla crafting code.
+	/*
+	private bool ConsumeForCraft(Item item, Item requiredItem, ref int stackRequired)
+	{
+		if (item.stack == 0 || item.IsNotTheSameAs(requiredItem) && !AcceptedByItemGroups(item.type, requiredItem.type))
+			return false;
+
+		var consumed = item.Clone();
+		consumed.stack = Math.Min(item.stack, stackRequired);
+		RecipeLoader.ConsumedItems.Add(consumed);
+
+		item.stack -= consumed.stack;
+		stackRequired -= consumed.stack;
+
+		if (item.stack == 0)
+			item.TurnToAir();
+
+		return true;
+	}
+	*/
+
 	private static void CollectItemsFromChests(Player player)
 	{
 		_recipeChests.Clear();
@@ -336,8 +_,17 @@
 
 	private static void CollectItems(Item[] currentInventory, int slotCap)
 	{
+		CollectItems(currentInventory.Take(slotCap));
+	}
+
+	// Added by TML
+	private static void CollectItems(IEnumerable<Item> items)
+	{
+		/*
 		for (int i = 0; i < slotCap; i++) {
 			Item item = currentInventory[i];
+		*/
+		foreach (var item in items) {
 			if (item.stack > 0) {
 				int num = item.stack;
 				if (_ownedItems.TryGetValue(item.type, out var value))
@@ -361,7 +_,10 @@
 			if (recipe.createItem.type == 0)
 				break;
 
+			if (recipe.Disabled)
+				continue;
+
-			for (int j = 0; j < maxRequirements; j++) {
+			for (int j = 0; j < recipe.requiredItem.Count; j++) {
 				RequiredItemEntry requiredItemEntry = recipe.requiredItemQuickLookup[j];
 				if (requiredItemEntry.itemIdOrRecipeGroup == 0)
 					break;
@@ -434,6 +_,8 @@
 		RecipeGroups.Fragment = new RecipeGroup(Lang.misc[51].Key, 3458, 3456, 3457, 3459).Register();
 		RecipeGroups.PressurePlate = new RecipeGroup(Lang.misc[38].Key, 542, 852, 543, 541, 1151, 529, 853, 4261).Register();
 		RecipeGroups.Jellyfish = new RecipeGroup("Misc.Jellyfish", 2436, 2437, 2438).Register();
+		RecipeGroup.AssignKeysToVanillaGroups(); // TML
+		RecipeGroupHelper.AddRecipeGroups();
 	}
 
 	public static void UpdateItemVariants()
@@ -441,8 +_,9 @@
 		for (int i = 0; i < maxRecipes; i++) {
 			Recipe obj = Main.recipe[i];
 			obj.createItem.Refresh();
+
-			Item[] array = obj.requiredItem;
+			var array = obj.requiredItem;
-			for (int j = 0; j < array.Length; j++) {
+			for (int j = 0; j < array.Count; j++) {
 				array[j].Refresh();
 			}
 		}
@@ -497,6 +_,7 @@
 
 	public static void SetupRecipes()
 	{
+		Recipe.SetupRecipeGroups(); // TML
 		Array.Clear(TileUsedInRecipes, 0, TileUsedInRecipes.Length);
 		SetupTileInheritance();
 		int num = 5;
@@ -15022,11 +_,19 @@
 		AddRecipe();
 		CreateReverseWallRecipes();
 		CreateReversePlatformRecipes();
+
+		RecipeLoader.AddRecipes();
+		RecipeLoader.PostAddRecipes();
+		RecipeLoader.OrderRecipes();
+
 		CreateRequiredItemQuickLookups();
 		UpdateWhichItemsAreMaterials();
 		UpdateWhichItemsAreCrafted();
 		UpdateInheritedTilesUsedInRecipes();
 		UpdateMaterialFieldForAllRecipes();
+
+		FixRecipeGroups();
+
 		foreach (RecipeGroup value in RecipeGroup.recipeGroups.Values) {
 			value.SortDecraftingEntries();
 		}
@@ -15036,19 +_,14 @@
 
 	private static void CreateRequiredItemQuickLookups()
 	{
-		for (int i = 0; i < numRecipes; i++) {
+		for (int i = 0; i < Main.recipe.Length; i++) {
 			Recipe recipe = Main.recipe[i];
+			recipe.requiredItemQuickLookup = new RequiredItemEntry[recipe.requiredItem.Count];
-			for (int j = 0; j < maxRequirements; j++) {
+			for (int j = 0; j < recipe.requiredItem.Count; j++) {
 				Item item = recipe.requiredItem[j];
-				if (item.IsAir)
-					break;
 
 				RequiredItemEntry requiredItemEntry = new RequiredItemEntry(item.type, item.stack);
-				int[] array = recipe.acceptedGroups;
-				foreach (int num in array) {
+				foreach (int num in recipe.acceptedGroups) {
-					if (num < 0)
-						break;
-
 					RecipeGroup recipeGroup = RecipeGroup.recipeGroups[num];
 					if (recipeGroup.ValidItems.Contains(item.type))
 						requiredItemEntry.itemIdOrRecipeGroup = recipeGroup.GetGroupFakeItemId();
@@ -15062,8 +_,8 @@
 	private static void UpdateMaterialFieldForAllRecipes()
 	{
 		for (int i = 0; i < numRecipes; i++) {
-			for (int j = 0; Main.recipe[i].requiredItem[j].type > 0; j++) {
-				Main.recipe[i].requiredItem[j].Refresh(onlyIfVariantChanged: false);
+			foreach (Item item in Main.recipe[i].requiredItem) {
+				item.Refresh(onlyIfVariantChanged: false);
 			}
 
 			Main.recipe[i].createItem.Refresh(onlyIfVariantChanged: false);
@@ -15075,7 +_,11 @@
 		bool[] array = new bool[RecipeGroup.nextRecipeGroupIndex];
 		for (int i = 0; i < numRecipes; i++) {
 			Recipe recipe = Main.recipe[i];
+
+			if (Main.recipe[i].Disabled)
+				continue;
+			
-			for (int j = 0; j < maxRequirements; j++) {
+			for (int j = 0; j < recipe.requiredItem.Count; j++) {
 				RequiredItemEntry requiredItemEntry = recipe.requiredItemQuickLookup[j];
 				if (requiredItemEntry.itemIdOrRecipeGroup == 0)
 					break;
@@ -15099,7 +_,10 @@
 
 	public static void UpdateWhichItemsAreCrafted()
 	{
+		List<int>[] recipeLists = new List<int>[ItemLoader.ItemCount];
 		for (int i = 0; i < numRecipes; i++) {
+
+			/*
 			if (!Main.recipe[i].notDecraftable)
 				ItemID.Sets.IsCrafted[Main.recipe[i].createItem.type] = i;
 
@@ -15108,7 +_,15 @@
 
 			if (Main.recipe[i].corruption)
 				ItemID.Sets.IsCraftedCorruption[Main.recipe[i].createItem.type] = i;
+			*/
+
+			Recipe recipe = Main.recipe[i];
+
+			if (!recipe.notDecraftable && !recipe.Disabled)
+				(recipeLists[recipe.createItem.type] ??= new()).Add(i);
 		}
+
+		ItemID.Sets.CraftingRecipeIndices = recipeLists.Select(l => (l ??= new()).ToArray()).ToArray();
 	}
 
 	private static void AddSolarFurniture()
@@ -16730,7 +_,7 @@
 	{
 		int num = numRecipes;
 		for (int i = 0; i < num; i++) {
-			if (Main.recipe[i].createItem.createTile >= 0 && TileID.Sets.Platforms[Main.recipe[i].createItem.createTile] && Main.recipe[i].requiredItem[1].type == 0) {
+			if (Main.recipe[i].createItem.createTile >= 0 && TileID.Sets.Platforms[Main.recipe[i].createItem.createTile] && Main.recipe[i].requiredItem.Count == 1) {
 				currentRecipe.createItem.SetDefaults(Main.recipe[i].requiredItem[0].type);
 				currentRecipe.createItem.stack = Main.recipe[i].requiredItem[0].stack;
 				currentRecipe.requiredItem[0].SetDefaults(Main.recipe[i].createItem.type);
@@ -16740,10 +_,12 @@
 				Recipe recipe = Main.recipe[numRecipes - 1];
 				for (int num2 = numRecipes - 2; num2 > i; num2--) {
 					Main.recipe[num2 + 1] = Main.recipe[num2];
+					Main.recipe[num2].RecipeIndex = num2 + 1; // Added by TML.
 				}
 
 				Main.recipe[i + 1] = recipe;
 				Main.recipe[i + 1].notDecraftable = true;
+				recipe.RecipeIndex = i + 1; // Added by TML.
 			}
 		}
 	}
@@ -16752,7 +_,7 @@
 	{
 		int num = numRecipes;
 		for (int i = 0; i < num; i++) {
-			if (Main.recipe[i].createItem.createWall > 0 && Main.recipe[i].requiredItem[1].type == 0 && Main.recipe[i].requiredItem[0].createWall == -1) {
+			if (Main.recipe[i].createItem.createWall > 0 && Main.recipe[i].requiredItem.Count == 1 && Main.recipe[i].requiredItem[0].createWall == -1) {
 				currentRecipe.createItem.SetDefaults(Main.recipe[i].requiredItem[0].type);
 				currentRecipe.createItem.stack = Main.recipe[i].requiredItem[0].stack;
 				currentRecipe.requiredItem[0].SetDefaults(Main.recipe[i].createItem.type);
@@ -16762,15 +_,17 @@
 				Recipe recipe = Main.recipe[numRecipes - 1];
 				for (int num2 = numRecipes - 2; num2 > i; num2--) {
 					Main.recipe[num2 + 1] = Main.recipe[num2];
+					Main.recipe[num2].RecipeIndex = num2 + 1; // Added by TML.
 				}
 
 				Main.recipe[i + 1] = recipe;
 				Main.recipe[i + 1].notDecraftable = true;
+				recipe.RecipeIndex = i + 1; // Added by TML.
 			}
 		}
 	}
 
-	public void SetIngredients(params int[] ingredients)
+	private void SetIngredients(params int[] ingredients) // TML: private
 	{
 		if (ingredients.Length == 1) {
 			ingredients = new int[2] {
@@ -16789,22 +_,83 @@
 		}
 	}
 
-	public void SetCraftingStation(int tileType)
+	private void SetCraftingStation(int tileType) // TML: private
 	{
 		requiredTile = tileType;
 	}
 
 	private static void AddRecipe()
 	{
+		/*
 		if (currentRecipe.requiredTile == 13)
 			currentRecipe.alchemy = true;
 
 		Main.recipe[numRecipes] = currentRecipe;
+		*/
 		if (currentRecipe.requiredTile >= 0)
 			TileUsedInRecipes[currentRecipe.requiredTile] = true;
 
+		/*
 		currentRecipe = new Recipe();
 		numRecipes++;
+		*/
+
+		//TML: The whole method is rewritten.
+		if (currentRecipe.requiredTile == TileID.Bottles)
+			currentRecipe.AddConsumeIngredientCallback(IngredientQuantityRules.Alchemy);
+
+		static void ReplaceCondition(ref bool field, Condition cond)
+		{
+			if (field)
+				currentRecipe.AddCondition(cond);
+
+			field = false;
+		}
+
+		static void ReplaceDecraftCondition(ref bool field, Condition cond)
+		{
+			if (field)
+				currentRecipe.AddDecraftCondition(cond);
+
+			field = false;
+		}
+
+		ReplaceCondition(ref currentRecipe.needGraveyardBiome, Condition.InGraveyard);
+		ReplaceCondition(ref currentRecipe.needSnowBiome, Condition.InSnow);
+		ReplaceCondition(ref currentRecipe.needWater, Condition.NearWater);
+		ReplaceCondition(ref currentRecipe.needLava, Condition.NearLava);
+		ReplaceCondition(ref currentRecipe.needHoney, Condition.NearHoney);
+		ReplaceCondition(ref currentRecipe.needMechdusa, Condition.ZenithWorld);
+
+		ReplaceDecraftCondition(ref currentRecipe.crimson, Condition.CrimsonWorld);
+		ReplaceDecraftCondition(ref currentRecipe.corruption, Condition.CorruptWorld);
+
+		currentRecipe.requiredItem.RemoveAll(item => item.IsAir);
+		currentRecipe.acceptedGroups.RemoveAll(group => group == -1);
+
+		Main.recipe[numRecipes] = currentRecipe;
+		currentRecipe.RecipeIndex = numRecipes;
+
+		if (RecipeLoader.FirstRecipeForItem[currentRecipe.createItem.type] == null) {
+			RecipeLoader.FirstRecipeForItem[currentRecipe.createItem.type] = currentRecipe;
+		}
+
+		ResetCurrentRecipe();
+		numRecipes++;
+	}
+
+	// Added by TML.
+	private static void ResetCurrentRecipe()
+	{
+		currentRecipe = new Recipe {
+			requiredItem = new List<Item>(),
+			requiredTile = -1,
+			acceptedGroups = Enumerable.Repeat(-1, maxVanillaRequirements).ToList()
+		};
+
+		for (int i = 0; i < maxVanillaRequirements; i++) {
+			currentRecipe.requiredItem.Add(new Item());
+		}
 	}
 
 	public static int GetRequiredTileStyle(int tileID)
