--- src/Terraria\Terraria\RecipeGroup.cs
+++ src/tModLoader\Terraria\RecipeGroup.cs
@@ -8,6 +_,7 @@
 		public Func<string> GetText;
 		public List<int> ValidItems;
 		public int IconicItemIndex;
+		internal bool[] ValidItemsLookup;
 		public static Dictionary<int, RecipeGroup> recipeGroups = new Dictionary<int, RecipeGroup>();
 		public static Dictionary<string, int> recipeGroupIDs = new Dictionary<string, int>();
 		public static int nextRecipeGroupIndex = 0;
@@ -25,5 +_,14 @@
 			RecipeGroup.recipeGroupIDs.Add(name, num);
 			return num;
 		}
+
+		public bool ContainsItem(int type)
+		{
+			if (ValidItemsLookup == null)
+			{
+				return ValidItems.Contains(type);
+			}
+			return ValidItemsLookup[type];
+		}
 	}
 }

