--- src/Terraria\Terraria\RemoteServer.cs
+++ src/tModLoader\Terraria\RemoteServer.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Net.Sockets;
 
 namespace Terraria
@@ -57,6 +_,7 @@
 			{
 				try
 				{
+					Logging.Terraria.Error("Client Read Thread", value);
 					using (StreamWriter streamWriter = new StreamWriter("client-crashlog.txt", true))
 					{
 						streamWriter.WriteLine(DateTime.Now);

