--- src/TerrariaNetCore/Terraria/RemoteClient.cs
+++ src/tModLoader/Terraria/RemoteClient.cs
@@ -2,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Terraria.DataStructures;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Net.Sockets;
 
 namespace Terraria;
@@ -12,7 +_,10 @@
 	public int Id;
 	public string Name = "Anonymous";
 	public bool IsActive;
-	public bool PendingTermination;
+
+	// Changed from a field into a private setter property.
+	// Vanilla code must be changed to use SetPendingTermination instead.
+	public bool PendingTermination { get; private set; }
 	public bool PendingTerminationApproved;
 	public bool IsAnnouncementCompleted;
 	public int State;
@@ -21,6 +_,10 @@
 	public string StatusText2;
 	public int StatusCount;
 	public int StatusMax;
+	/// <summary>
+	/// Indicates if specific world sections (200 x 150 tile chunks) of the world are loaded by this client. The server will sync world sections to the client as the player visits them by calling <see cref="CheckSection"/>.
+	/// <br/><br/> <b>Only accessible on Server!</b>
+	/// </summary>
 	public bool[,] TileSections = new bool[Main.maxTilesX / 200 + 1, Main.maxTilesY / 150 + 1];
 	public uint[,] TileSectionsCheckTime = new uint[Main.maxTilesX / 200 + 1, Main.maxTilesY / 150 + 1];
 	public bool CheckingSections;
@@ -39,6 +_,13 @@
 
 	public static event Action<int, Point> NetSectionActivated;
 
+	// Added by TML to log a reason for terminating clients
+	internal void SetPendingTermination(string reason)
+	{
+		PendingTermination = true;
+		ModNet.Log(Id, "Terminating: " + reason);
+	}
+
 	public bool IsConnected()
 	{
 		if (Socket != null)
@@ -94,6 +_,11 @@
 		SpamWater = 0f;
 	}
 
+	/// <summary>
+	/// Ensures that the specified client (<paramref name="playerIndex"/>) has the world section containing the provided <paramref name="position"/> loaded. <paramref name="fluff"/> determines how many surrounding sections will be synced at the same time.
+	/// <br/><br/> The server should call this method before teleporting the player or when the player is viewing locations detached from the player, such as with Kwad Racer Drone or golf.
+	/// <br/><br/> <b>Call on Server only!</b>
+	/// </summary>
 	public static void CheckSection(int playerIndex, Vector2 position, int fluff = 1)
 	{
 		RemoteClient remoteClient = Netplay.Clients[playerIndex];
@@ -232,7 +_,7 @@
 			}
 		}
 		catch {
-			PendingTermination = true;
+			SetPendingTermination("Exception in RemoteClient.TryRead");
 		}
 	}
 
@@ -240,7 +_,7 @@
 	{
 		if (!Netplay.Disconnect) {
 			if (length == 0) {
-				PendingTermination = true;
+				SetPendingTermination("ServerReadCallBack returned no data");
 			}
 			else {
 				try {
@@ -283,7 +_,7 @@
 				StatusText = Language.GetTextValue("Net.ClientPlaying", Socket.GetRemoteAddress(), Name);
 			}
 			catch (Exception) {
-				PendingTermination = true;
+				SetPendingTermination("Exception in Socket.GetRemoteAddress while updating StatusText (State == 10)");
 			}
 		}
 	}
