--- src/TerrariaNetCore/Terraria/RemoteServer.cs
+++ src/tModLoader/Terraria/RemoteServer.cs
@@ -1,5 +_,6 @@
 using System;
 using System.IO;
+using Terraria.ModLoader;
 using Terraria.Net.Sockets;
 
 namespace Terraria;
@@ -67,6 +_,7 @@
 		}
 		catch (Exception value) {
 			try {
+				Logging.Terraria.Error("Client Read Thread", value);
 				using StreamWriter streamWriter = new StreamWriter("client-crashlog.txt", append: true);
 				streamWriter.WriteLine(DateTime.Now);
 				streamWriter.WriteLine(value);
