--- src/TerrariaNetCore/Terraria/SceneMetrics.cs
+++ src/tModLoader/Terraria/SceneMetrics.cs
@@ -3,6 +_,7 @@
 using Terraria.GameContent;
 using Terraria.GameContent.Events;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.WorldBuilding;
 
 namespace Terraria;
@@ -70,14 +_,20 @@
 	public bool InTorchGodMinigame;
 	public static int NPCEventZoneRadius = 4000;
 	public bool CanPlayCreditsRoll;
+	/// <summary>
+	/// Indexed by BannerIDs, indicates which enemies are currently subject to the banner damage bonus. Set to true in conjunction with <see cref="hasBanner"/>.
+	/// </summary>
 	public bool[] NPCBannerBuff = new bool[BannerSystem.MaxBannerTypes];
+	/// <summary>
+	/// Indicates that the player is in range of a buff providing Banner tile and should receive the <see cref="BuffID.MonsterBanner"/> buff. Set to true in conjunction with <see cref="NPCBannerBuff"/>.
+	/// </summary>
 	public bool hasBanner;
 	public Vector2[] ClosestNPCPosition = new Vector2[NPCID.Count];
 	private static Player _dummyPlayer = new Player();
-	private readonly int[] _tileCounts = new int[TileID.Count];
+	internal int[] _tileCounts = new int[TileLoader.TileCount];
 	private readonly int[] _liquidCounts = new int[LiquidID.Count];
 
 	public uint LastScanTime { get; private set; }
 
 	public Vector2 Center { get; private set; }
 
@@ -113,67 +_,67 @@
 
 	public int ActiveMusicBox { get; set; }
 
-	public int SandTileCount { get; private set; }
+	public int SandTileCount { get; set; }
 
-	public int MushroomTileCount { get; private set; }
+	public int MushroomTileCount { get; set; }
 
-	public int SnowTileCount { get; private set; }
+	public int SnowTileCount { get; set; }
 
-	public int WaterCandleCount { get; private set; }
+	public int WaterCandleCount { get; set; }
 
-	public int PeaceCandleCount { get; private set; }
+	public int PeaceCandleCount { get; set; }
 
-	public int ShadowCandleCount { get; private set; }
+	public int ShadowCandleCount { get; set; }
 
-	public int PartyMonolithCount { get; private set; }
+	public int PartyMonolithCount { get; set; }
 
-	public int MeteorTileCount { get; private set; }
+	public int MeteorTileCount { get; set; }
 
-	public int BloodTileCount { get; private set; }
+	public int BloodTileCount { get; set; }
 
-	public int JungleTileCount { get; private set; }
+	public int JungleTileCount { get; set; }
 
-	public int DungeonTileCount { get; private set; }
+	public int DungeonTileCount { get; set; }
 
-	public bool HasSunflower { get; private set; }
+	public bool HasSunflower { get; set; }
 
-	public bool HasGardenGnome { get; private set; }
+	public bool HasGardenGnome { get; set; }
 
-	public bool HasClock { get; private set; }
+	public bool HasClock { get; set; }
 
-	public bool HasCampfire { get; private set; }
+	public bool HasCampfire { get; set; }
 
-	public bool HasStarInBottle { get; private set; }
+	public bool HasStarInBottle { get; set; }
 
-	public bool HasHeartLantern { get; private set; }
+	public bool HasHeartLantern { get; set; }
 
-	public int ActiveFountainColor { get; private set; }
+	public int ActiveFountainColor { get; set; }
 
-	public int ActiveMonolithType { get; private set; }
+	public int ActiveMonolithType { get; set; }
 
-	public bool BloodMoonMonolith { get; private set; }
+	public bool BloodMoonMonolith { get; set; }
 
-	public bool MoonLordMonolith { get; private set; }
+	public bool MoonLordMonolith { get; set; }
 
-	public bool EchoMonolith { get; private set; }
+	public bool EchoMonolith { get; set; }
 
-	public int ShimmerMonolithState { get; private set; }
+	public int ShimmerMonolithState { get; set; }
 
-	public bool CRTMonolith { get; private set; }
+	public bool CRTMonolith { get; set; }
 
-	public bool RetroMonolith { get; private set; }
+	public bool RetroMonolith { get; set; }
 
-	public bool NoirMonolith { get; private set; }
+	public bool NoirMonolith { get; set; }
 
-	public bool RadioThingMonolith { get; private set; }
+	public bool RadioThingMonolith { get; set; }
 
-	public bool HasCatBast { get; private set; }
+	public bool HasCatBast { get; set; }
 
-	public int GraveyardTileCount { get; private set; }
+	public int GraveyardTileCount { get; set; }
 
-	public int DesertSandTileCount { get; private set; }
+	public int DesertSandTileCount { get; set; }
 
-	public int OceanSandTileCount { get; private set; }
+	public int OceanSandTileCount { get; set; }
 
 	public bool EnoughTilesForShimmer => ShimmerTileCount >= ShimmerTileThreshold;
 
@@ -251,2 +_,5 @@
 			TileCenter = Center.ToTileCoordinates().ClampedInWorld();
+
+			SystemLoader.ResetNearbyTileEffects();
+
 			ScanTiles();
@@ -264,5 +_,7 @@
 			CanPlayCreditsRoll = ActiveMusicBox == 85;
+
+			SystemLoader.TileCountsAvailable(_tileCounts);
 		}
 	}
 
 	private void ScanTiles()
@@ -343,4 +_,6 @@
 
 				UpdateOreFinder(new Point(i, j), tile);
+
+				TileLoader.NearbyEffects(i, j, tile.type, false);
 			}
 		}
@@ -355,7 +_,7 @@
 				if (tile == null || !tile.active())
 					continue;
 
-				if (tile.type == 104)
+				if (TileID.Sets.Clock[tile.type])
 					HasClock = true;
 
 				switch (tile.type) {
@@ -442,6 +_,12 @@
 							RadioThingMonolith = true;
 						break;
 				}
+
+				// This does not use TileLoader.IsModMusicBox because it needs the *exact* frameY for the second dict lookup
+				if (MusicLoader.tileToMusic.ContainsKey(tile.type) && MusicLoader.tileToMusic[tile.type].ContainsKey(tile.frameY) && tile.frameX == 36)
+					ActiveMusicBox = MusicLoader.tileToMusic[tile.type][tile.frameY];
+
+				TileLoader.NearbyEffects(i, j, tile.type, true);
 			}
 		}
 	}
@@ -460,6 +_,7 @@
 
 		ShimmerTileCount = _liquidCounts[3];
 		HoneyBlockCount = _tileCounts[229];
+		/*
 		HolyTileCount = _tileCounts[109] + _tileCounts[492] + _tileCounts[110] + _tileCounts[113] + _tileCounts[117] + _tileCounts[116] + _tileCounts[164] + _tileCounts[403] + _tileCounts[402];
 		SnowTileCount = _tileCounts[147] + _tileCounts[148] + _tileCounts[161] + _tileCounts[162] + _tileCounts[164] + _tileCounts[163] + _tileCounts[200];
 		if (Main.remixWorld) {
@@ -474,12 +_,18 @@
 		}
 
 		MushroomTileCount = _tileCounts[70] + _tileCounts[71] + _tileCounts[72] + _tileCounts[528];
+		*/
 		MeteorTileCount = _tileCounts[37];
+		/*
 		DungeonTileCount = _tileCounts[41] + _tileCounts[43] + _tileCounts[44] + _tileCounts[481] + _tileCounts[482] + _tileCounts[483];
 		SandTileCount = _tileCounts[53] + _tileCounts[112] + _tileCounts[116] + _tileCounts[234] + _tileCounts[397] + _tileCounts[398] + _tileCounts[402] + _tileCounts[399] + _tileCounts[396] + _tileCounts[400] + _tileCounts[403] + _tileCounts[401];
+		*/
 		PartyMonolithCount = _tileCounts[455];
 		GraveyardTileCount = _tileCounts[85];
 		GraveyardTileCount -= _tileCounts[27] / 2;
+
+		TileLoader.RecountTiles(this);
+
 		if (_tileCounts[27] > 0)
 			HasSunflower = true;
 
@@ -623,7 +_,9 @@
 
 	public int GetTileCount(ushort tileId) => _tileCounts[tileId];
 
+	public int GetLiquidCount(short liquidType) => _liquidCounts[liquidType];
+
 	public void Reset()
 	{
 		LastScanTime = uint.MaxValue;
 		Array.Clear(_tileCounts, 0, _tileCounts.Length);
@@ -707,4 +_,7 @@
 		TownNPCCount = 0;
 		PerspectivePlayer = _dummyPlayer;
+
+		if (NPCBannerBuff.Length < ModLoader.NPCLoader.NPCCount)
+			Array.Resize(ref NPCBannerBuff, ModLoader.NPCLoader.NPCCount);
 	}
 
