--- src/Terraria/Terraria/SceneMetrics.cs
+++ src/tModLoader/Terraria/SceneMetrics.cs
@@ -27,7 +_,7 @@
 		public static int GraveyardTileThreshold = 24;
 		public bool[] NPCBannerBuff = new bool[289];
 		public bool hasBanner;
-		private readonly int[] _tileCounts = new int[623];
+		internal int[] _tileCounts = new int[623];
 		private readonly World _world;
 		private readonly List<Point> _oreFinderTileLocations = new List<Point>(512);
 		public int bestOre;
