--- src/Terraria/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -7,7 +_,8 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © 2020 Re-Logic</Copyright>
     <RootNamespace>Terraria</RootNamespace>
-	<OutputName>Terraria</OutputName>
+	  <OutputName>tModLoader</OutputName>
+	<LangVersion>latest</LangVersion>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Server'))">
     <OutputType>Exe</OutputType>
@@ -15,15 +_,23 @@
   </PropertyGroup>
   <PropertyGroup>
     <PdbFile>bin\$(OutputName)</PdbFile>
-    <ApplicationIcon>Icon.ico</ApplicationIcon>
+    <ApplicationIcon>tModLoader.ico</ApplicationIcon>
   </PropertyGroup>
   <ItemGroup>
     <Reference Include="FNA" Condition="$(DefineConstants.Contains('FNA'))" />
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.Reduced" />
     <Reference Include="Microsoft.Xna.Framework" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="Microsoft.Xna.Framework.Game" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="Microsoft.Xna.Framework.Graphics" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('XNA'))" />
+    <Reference Include="Mono.Cecil" />
+    <Reference Include="Mono.Cecil.Pdb" />
+    <Reference Include="Mono.Cecil.Mdb" />
+    <Reference Include="MP3Sharp" />
+    <Reference Include="NVorbis" />
+    <Reference Include="log4net" />
+    <Reference Include="MonoMod.RuntimeDetour" />
+    <Reference Include="MonoMod.Utils" />
     <Reference Include="Newtonsoft.Json" />
     <Reference Include="RailSDK.Net">
       <HintPath Condition="$(DefineConstants.Contains('WINDOWS'))">Libraries/Windows/RailSDK.Net.dll</HintPath>
@@ -89,6 +_,6 @@
     <Copy SourceFiles="@(LibFolder)" DestinationFolder="$(TargetDir)" />
     <Copy SourceFiles="@(SteamFiles)" DestinationFolder="$(TargetDir)" />
     <!--Move bin to steam-->
-    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(TerrariaSteamPath)" />
+    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(tModLoaderSteamPath)" />
   </Target>
 </Project>
