--- src/TerrariaNetCore/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -9,27 +_,29 @@
 		<Version>1.4.3.6</Version>
 		<Company>Re-Logic</Company>
 		<Copyright>Copyright © 2022 Re-Logic</Copyright>
-		<AssemblyName>Terraria</AssemblyName>
+		<AssemblyName>tModLoader</AssemblyName>
 		<RootNamespace>Terraria</RootNamespace>
-		<ApplicationIcon>Icon.ico</ApplicationIcon>
+		<ApplicationIcon>tModLoader.ico</ApplicationIcon>
-	</PropertyGroup>
-
-	<!-- Conditionals -->
-	<PropertyGroup>
-		<!-- Obsolete: Append 'Server' if it's a server build. -->
-		<!-- <AssemblyName Condition="$(Configuration.Contains('Server'))">$(AssemblyName)Server</AssemblyName> -->
-
-		<!-- Avoid overwriting Terraria(Server).exe (if it's not Debug - it's Release.) -->
-		<AssemblyName Condition="$(Configuration.Contains('Debug'))">$(AssemblyName)Debug</AssemblyName>
-		<AssemblyName Condition="!$(Configuration.Contains('Debug'))">$(AssemblyName)Release</AssemblyName>
-
-		<!-- Set OutputName to the modified AssemblyName. -->
 		<OutputName>$(AssemblyName)</OutputName>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		<UseAppHost>false</UseAppHost>
+		<GenerateRuntimeConfigDevFile>true</GenerateRuntimeConfigDevFile>
+	</PropertyGroup>
+
+	<!-- Git & version information -->
+	<PropertyGroup>
+		<tMLVersion Condition="'$(TmlVersion)' == ''">9999.0</tMLVersion>
+		<BuildPurpose Condition="'$(BuildPurpose)' == ''">dev</BuildPurpose>
+		<BranchName Condition="'$(BranchName)' == ''">unknown</BranchName>
+		<BranchName>$(BranchName.Replace("|","-"))</BranchName>
+		<CommitSHA Condition="'$(CommitSHA)' == ''">unknown</CommitSHA>
+		<BuildDate>$([System.DateTime]::UtcNow.ToBinary())</BuildDate>
+		<SourceRevisionId>$(tMLVersion)|$(BranchName)|$(BuildPurpose)|$(CommitSHA)|$(BuildDate)</SourceRevisionId>
 	</PropertyGroup>
 
 	<!-- Task properties -->
 	<PropertyGroup>
-		<_ActualOutputDirectory>$(TerrariaSteamPath)</_ActualOutputDirectory>
+		<_ActualOutputDirectory>$(tModLoaderSteamPath)</_ActualOutputDirectory>
 
 		<LibraryInputDirectory>$(MSBuildThisFileDirectory)\Libraries</LibraryInputDirectory>
 		<LibraryOutputDirectory>Libraries</LibraryOutputDirectory>
@@ -42,21 +_,34 @@
 		<!-- Projects -->
 		<ProjectReference Include="..\ReLogic\ReLogic.csproj" />
 		<ProjectReference Include="..\..\..\FNA\FNA.Core.csproj" />
+		<ProjectReference Include="..\..\..\tModPorter\tModPorter\tModPorter.csproj" />
 		<!-- NuGet Packages -->
+		<PackageReference Include="Basic.Reference.Assemblies.Net60" Version="1.2.4" />
 		<PackageReference Include="CsvHelper" Version="2.16.3" /> <!-- Aeons old, but that's alright. -->
+		<PackageReference Include="Hjson" Version="3.0.0" />
+		<PackageReference Include="log4net" Version="2.0.15" />
+		<PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.2.0" />
+		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.2.0" />
+		<PackageReference Include="Microsoft.Win32.Registry" Version="6.0.0-preview.5.21301.5 " />
+		<PackageReference Include="MonoMod.RuntimeDetour" Version="22.5.31.2" />
 		<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
 		<PackageReference Include="Steamworks.NET" Version="20.1.0" />
+		<PackageReference Include="System.CodeDom" Version="6.0.0" />
+		<PackageReference Include="System.Diagnostics.PerformanceCounter" Version="6.0.0" />
+		<PackageReference Include="System.Reflection.MetadataLoadContext" Version="6.0.0" />
 		<!-- DLLs -->
-		<Reference Include="Libraries\Common\Ionic.Zip.CF.dll" />
+		<Reference Include="Libraries\Common\Ionic.Zip.Reduced.dll" />
 		<Reference Include="Libraries\Common\MP3Sharp.dll" />
 		<Reference Include="Libraries\Common\NVorbis.dll" /> <!-- The NuGet package is outdated. -->
 		<Reference Include="Libraries\Common\RailSDK.Net.dll" />
 		<Reference Include="Libraries\Common\SteelSeriesEngineWrapper.dll" />
+		<Reference Include="Libraries\Common\TerrariaHooks.dll" />
 	</ItemGroup>
 
 	<!-- Resources -->
 	<ItemGroup>
 		<!-- Embedded resources -->
+		<EmbeddedResource Include="**\*.png" />
 		<EmbeddedResource Include="GameContent\Creative\Content\*" />
 		<EmbeddedResource Include="GameContent\Metadata\MaterialData\*" />
 		<EmbeddedResource Include="GameContent\WorldBuilding\*" />
@@ -65,10 +_,12 @@
 		<EmbeddedResource Include="Microsoft\**" />
 		<!-- Copy to output folder -->
 		<Content Include="Libraries\Native\**" CopyToOutputDirectory="PreserveNewest" />
+		<Content Include="release_extras\**" CopyToOutputDirectory="PreserveNewest" Link="%(RecursiveDir)%(Filename)%(Extension)" />
 
 		<!-- Excluded code files -->
 		<Compile Remove="Social\WeGame\AsyncTaskHelper.cs" />
 		<Compile Remove="Social\WeGame\CurrentThreadRunner.cs" />
+		<Compile Remove="Libraries\Common\TerrariaHooks.dll" />
 		<!-- Hide libraries in IDEs -->
 		<None Remove="Libraries\**" />
 	</ItemGroup>
@@ -101,6 +_,9 @@
 
 	<!-- Output cleanups -->
 	<Target Name="OutputCleanup" AfterTargets="Build" >
+		<Delete Files="$(TargetDir)\tModPorter.runtimeconfig.json" />
+		<Delete Files="$(TargetDir)\tModPorter.deps.json" />
+
 		<DeleteEmptyDirectories Path="$(OutputPath)" />
 	</Target>
 
