--- src/TerrariaNetCore/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -57,9 +_,13 @@
 		<PackageReference Include="Steamworks.NET" Version="20.1.0" />
 	</ItemGroup>
 	<PropertyGroup>
-		<_ActualOutputDirectory>$(TerrariaSteamPath)</_ActualOutputDirectory>
+			<_ActualOutputDirectory>$(tModLoaderSteamPath)</_ActualOutputDirectory>
 	</PropertyGroup>
 	<Target Name="CopyToSteamDir" AfterTargets="Build">
+		<!-- Exclude tModPorter's metadata. -->
+		<Delete Files="$(TargetDir)\tModPorter.runtimeconfig.json" />
+		<Delete Files="$(TargetDir)\tModPorter.deps.json" />
+		
 		<!-- copy files systematically to output folder -->
 		<ItemGroup>
 			<BinFiles Include="$(TargetDir)**" />
