--- src/Terraria/Terraria/TerrariaNetCore.csproj
+++ src/tModLoader/Terraria/TerrariaNetCore.csproj
@@ -7,23 +_,23 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © 2020 Re-Logic</Copyright>
     <RootNamespace>Terraria</RootNamespace>
-	  <OutputName>Terraria</OutputName>
+    <OutputName>tModLoader</OutputName>
     <EffectivePlatform>Linux</EffectivePlatform>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Server'))">
     <OutputType>Exe</OutputType>
     <OutputName>$(OutputName)Server</OutputName>
   </PropertyGroup>
   <PropertyGroup>
     <PdbFile>bin\$(OutputName)</PdbFile>
     <ApplicationIcon>Icon.ico</ApplicationIcon>
     <AssemblyName>$(OutputName)</AssemblyName>
     <TargetName>$(OutputName)</TargetName>
   </PropertyGroup>
   <ItemGroup>
     <FrameworkReference Include="Microsoft.WindowsDesktop.App" />
     <Reference Include="FNA" />
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.Reduced" />
     <Reference Include="Newtonsoft.Json" />
     <Reference Include="RailSDK.Net">
       <HintPath>Libraries/Native/$(EffectivePlatform)/RailSDK.Net.dll</HintPath>
@@ -31,8 +_,8 @@
     </Reference>
     <ProjectReference Include="../ReLogic/ReLogicNetCore.csproj" />
     <EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll">
-	    <Link>Libraries/ReLogic/ReLogic.dll</Link>
-      <LogicalName>Terraria.Libraries.ReLogic.ReLogic.dll</LogicalName>
+      <Link>Libraries/ReLogic/ReLogic.dll</Link>
+      <LogicalName>ReLogic.dll</LogicalName>
     </EmbeddedResource>
     <Reference Include="Steamworks.NET">
       <HintPath>Libraries/Native/$(EffectivePlatform)/Steamworks.NET.dll</HintPath>
@@ -44,25 +_,25 @@
     <EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
     <EmbeddedResource Include="GameContent/WorldBuilding/*" />
     <EmbeddedResource Include="IO/Data/*" />
-    <EmbeddedResource Include="Libraries/DotNetZip/**" />
+    <EmbeddedResource Include="Libraries/Common/**" />
     <EmbeddedResource Include="Localization/Content/**" />
     <EmbeddedResource Include="Microsoft/**" />
   </ItemGroup>
   <ItemGroup>
     <None Remove="Libraries/Linux/**" />
     <None Remove="Libraries/Mac/**" />
     <None Remove="Libraries/Mono/**" />
     <None Remove="Libraries/Windows/**" />
     <None Remove="Libraries/XNA/**" />
     <None Remove="Libraries/FNA/**" />
   </ItemGroup>
   <Target Name="CopyToSteamDir" AfterTargets="Build">
     <ItemGroup>
       <BinFiles Include="$(TargetDir)**" />
       <LibFiles Include="Libraries/Native/**" />
     </ItemGroup>
     <Message Text="Copying $(OutputName) to Steam Dir..." Importance="high" />
-    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(TerrariaSteamPath)" />
-    <Copy SourceFiles="@(LibFiles)" DestinationFolder="$(TerrariaSteamPath)/Libraries/Native/%(RecursiveDir)" />
+    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(tModLoaderSteamPath)" />
+    <Copy SourceFiles="@(LibFiles)" DestinationFolder="$(tModLoaderSteamPath)/Libraries/Native/%(RecursiveDir)" />
   </Target>
 </Project>
