--- src/Terraria/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -7,7 +_,16 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © 2020 Re-Logic</Copyright>
     <RootNamespace>Terraria</RootNamespace>
-    <AssemblyName>Terraria</AssemblyName>
+    <AssemblyName>tModLoader</AssemblyName>
+  </PropertyGroup>
+  <PropertyGroup>
+    <tMLVersion>0.12</tMLVersion>
+	<BuildPurpose Condition="'$(BuildPurpose)' == ''">dev</BuildPurpose>
+	<BranchName Condition="'$(BranchName)' == ''">unknown</BranchName>
+  <BranchName>$(BranchName.Replace("-",""))</BranchName>
+	<CommitSHA Condition="'$(CommitSHA)' == ''">unknown</CommitSHA>
+    <BuildDate>$([System.DateTime]::UtcNow.ToBinary())</BuildDate>
+    <SourceRevisionId>$(tMLVersion)-$(BranchName)-$(BuildPurpose)-$(CommitSHA)-$(BuildDate)</SourceRevisionId>
   </PropertyGroup>
   <PropertyGroup Condition="$(Configuration.Contains('Server'))">
     <OutputType>Exe</OutputType>
@@ -18,20 +_,28 @@
   </PropertyGroup>
   <PropertyGroup>
     <OutputName>$(AssemblyName)</OutputName>
-    <ApplicationIcon>Icon.ico</ApplicationIcon>
+    <ApplicationIcon>tModLoader.ico</ApplicationIcon>
     <TargetFramework>net5.0</TargetFramework>
   </PropertyGroup>
   <ItemGroup>
     <Reference Include="FNA" />
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.Reduced" />
-      <Reference Include="Newtonsoft.Json" />
+    <Reference Include="Mono.Cecil" />
+    <Reference Include="Mono.Cecil.Pdb" />
+    <Reference Include="Mono.Cecil.Mdb" />
+    <Reference Include="MP3Sharp" />
+    <Reference Include="NVorbis" />
+    <Reference Include="log4net" />
+    <Reference Include="MonoMod.RuntimeDetour" />
+    <Reference Include="MonoMod.Utils" />
+    <Reference Include="Newtonsoft.Json" />
     <Reference Include="RailSDK.Net" />
     <ProjectReference Include="../ReLogic/ReLogic.csproj">
       <Private>false</Private>
     </ProjectReference>
     <EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll">
       <Link>Libraries/ReLogic/ReLogic.dll</Link>
-      <LogicalName>Terraria.Libraries.ReLogic.ReLogic.dll</LogicalName>
+      <LogicalName>ReLogic.dll</LogicalName>
     </EmbeddedResource>
     <Reference Include="Steamworks.NET" />
     <Reference Include="SteelSeriesEngineWrapper" />
@@ -40,29 +_,70 @@
     </Reference>
   </ItemGroup>
   <ItemGroup>
+    <EmbeddedResource Include="**\*.png" />
     <EmbeddedResource Include="GameContent/Creative/Content/*" />
     <EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
     <EmbeddedResource Include="GameContent/WorldBuilding/*" />
     <EmbeddedResource Include="IO/Data/*" />
-    <EmbeddedResource Include="Libraries/Common/**" />
-    <EmbeddedResource Include="Libraries/FNA/**" />
+    <EmbeddedResource Include="Libraries/Common/**">
+      <LogicalName>%(Filename)%(Extension)</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="Libraries/FNA/**">
+      <LogicalName>%(Filename)%(Extension)</LogicalName>
+    </EmbeddedResource>
     <EmbeddedResource Include="Localization/Content/**" />
     <EmbeddedResource Include="Microsoft/**" />
   </ItemGroup>
   <ItemGroup>
     <Compile Remove="Social/WeGame/AsyncTaskHelper.cs" />
     <Compile Remove="Social/WeGame/CurrentThreadRunner.cs" />
-    <None Remove="Libraries/Mono/**" />
     <None Remove="Libraries/Windows/**" />
     <None Remove="Libraries/XNA/**" />
     <None Remove="Libraries/net40/**" />
     <None Update="Libraries/Native/**" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.7.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="3.7.0" />
+    <PackageReference Include="Microsoft.Win32.Registry" Version="5.0.0" />
+    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.3" />
+    <PackageReference Include="System.CodeDom" Version="5.0.0" />
+    <PackageReference Include="System.Diagnostics.PerformanceCounter" Version="5.0.1" />
+  </ItemGroup>
+  
   <Target Name="CopyToSteamDir" AfterTargets="Build">
     <ItemGroup>
       <BinFiles Include="$(TargetDir)**" />
     </ItemGroup>
     <Message Text="Copying $(AssemblyName) to Steam Dir..." Importance="high" />
-    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(TerrariaSteamPath)/%(RecursiveDir)" />
+    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(tModLoaderSteamPath)/%(RecursiveDir)" />
   </Target>
+
+  <ItemDefinitionGroup>
+    <ReferenceCopyLocalPaths>
+      <DestinationSubDirectory>NuGet\</DestinationSubDirectory>
+    </ReferenceCopyLocalPaths>
+  </ItemDefinitionGroup>
+  <Target Name="RedirectBuildFilesToLib" AfterTargets="ResolveLockFileCopyLocalFiles">
+    <ItemGroup>
+      <ReferenceCopyLocalPaths Condition=" '%(ReferenceCopyLocalPaths.AssetType)' == 'runtime' ">
+        <DestinationSubDirectory>Libraries\%(DestinationSubDirectory)</DestinationSubDirectory>
+      </ReferenceCopyLocalPaths>
+      <ReferenceCopyLocalPaths Condition=" '%(ReferenceCopyLocalPaths.AssetType)' == 'resources' ">
+        <DestinationSubDirectory>Libraries\Resources\%(DestinationSubDirectory)</DestinationSubDirectory>
+      </ReferenceCopyLocalPaths>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="RedirectPublishFilesToLib" AfterTargets="ComputeResolvedFilesToPublishList">
+    <ItemGroup>
+      <ResolvedFileToPublish Condition=" '%(ResolvedFileToPublish.AssetType)' == 'runtime' ">
+        <RelativePath>%(RelativePath)</RelativePath>
+      </ResolvedFileToPublish>
+      <ResolvedFileToPublish Condition=" '%(ResolvedFileToPublish.AssetType)' == 'resources' ">
+        <RelativePath>Libraries\Resources\%(RelativePath)</RelativePath>
+      </ResolvedFileToPublish>
+    </ItemGroup>
+  </Target>
+  
 </Project>
