--- src/Terraria/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -7,26 +_,27 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © 2021 Re-Logic</Copyright>
     <RootNamespace>Terraria</RootNamespace>
-    <AssemblyName>Terraria</AssemblyName>
+    <AssemblyName>tModLoader</AssemblyName>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <UseAppHost>false</UseAppHost>
     <GenerateRuntimeConfigDevFile>true</GenerateRuntimeConfigDevFile>
   </PropertyGroup>
-  <PropertyGroup Condition="$(Configuration.Contains('Server'))">
-    <AssemblyName>$(AssemblyName)Server</AssemblyName>
-  </PropertyGroup>
-  <PropertyGroup Condition="$(Configuration.Contains('Debug'))">
-    <AssemblyName>$(AssemblyName)Debug</AssemblyName>
-  </PropertyGroup>
-  <!-- Avoid overwriting Terraria(Server).exe (if it's not Debug it's release) -->
-  <PropertyGroup Condition="!$(Configuration.Contains('Debug'))">
-    <AssemblyName>$(AssemblyName)Release</AssemblyName>
+  <PropertyGroup>
+    <tMLVersion>0.12</tMLVersion>
+    <BuildPurpose Condition="'$(BuildPurpose)' == ''">dev</BuildPurpose>
+    <BranchName Condition="'$(BranchName)' == ''">unknown</BranchName>
+  <BranchName>$(BranchName.Replace("-",""))</BranchName>
+    <CommitSHA Condition="'$(CommitSHA)' == ''">unknown</CommitSHA>
+    <BuildDate>$([System.DateTime]::UtcNow.ToBinary())</BuildDate>
+    <SourceRevisionId>$(tMLVersion)-$(BranchName)-$(BuildPurpose)-$(CommitSHA)-$(BuildDate)</SourceRevisionId>
   </PropertyGroup>
   <PropertyGroup>
     <OutputName>$(AssemblyName)</OutputName>
-    <ApplicationIcon>Icon.ico</ApplicationIcon>
+    <ApplicationIcon>tModLoader.ico</ApplicationIcon>
   </PropertyGroup>
   <ItemGroup>
     <Reference Include="CsvHelper" />
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.Reduced" />
     <ProjectReference Include="../../../FNA/FNA.Core.csproj" />
     <Reference Include="MP3Sharp" />
     <Reference Include="Newtonsoft.Json" />
@@ -35,8 +_,11 @@
     <ProjectReference Include="../ReLogic/ReLogic.csproj" />
     <Reference Include="Steamworks.NET" />
     <Reference Include="SteelSeriesEngineWrapper" />
+    <Reference Include="log4net" />
+    <Reference Include="TerrariaHooks" />
   </ItemGroup>
   <ItemGroup>
+    <EmbeddedResource Include="**\*.png" />
     <EmbeddedResource Include="GameContent/Creative/Content/*" />
     <EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
     <EmbeddedResource Include="GameContent/WorldBuilding/*" />
@@ -47,16 +_,28 @@
   <ItemGroup>
     <Compile Remove="Social/WeGame/AsyncTaskHelper.cs" />
     <Compile Remove="Social/WeGame/CurrentThreadRunner.cs" />
-    <None Remove="Libraries/Mono/**" />
+    <Compile Remove="Libraries/Common/TerrariaHooks.dll" />
     <None Remove="Libraries/Windows/**" />
     <None Remove="Libraries/XNA/**" />
     <None Remove="Libraries/net40/**" />
     <Content Include="Libraries/Native/**" CopyToOutputDirectory="PreserveNewest" />
+    <Content Include="release_extras/**" CopyToOutputDirectory="PreserveNewest" Link="%(RecursiveDir)%(Filename)%(Extension)" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="MonoMod.RuntimeDetour" Version="21.11.1.1" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.0.0-6.final" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.0.0-6.final" />
+    <PackageReference Include="Microsoft.Win32.Registry" Version="6.0.0-preview.5.21301.5 " />
+    <PackageReference Include="System.CodeDom" Version="6.0.0" />
+    <PackageReference Include="System.Diagnostics.PerformanceCounter" Version="6.0.0" />
+    <PackageReference Include="Hjson" Version="3.0.0" />
   </ItemGroup>
   <PropertyGroup>
-    <_ActualOutputDirectory>$(TerrariaSteamPath)</_ActualOutputDirectory>
+    <_ActualOutputDirectory>$(tModLoaderSteamPath)</_ActualOutputDirectory>
   </PropertyGroup>
   <Target Name="CopyToSteamDir" AfterTargets="Build">
+    <!-- Deal with Abnormal Managed Libaries that require deployment assembly replacements for varying architectures-->
+    <Copy SourceFiles="Libraries\Common\UNIX.Steamworks.NET.dll" DestinationFolder="$(TargetDir)\PlatformVariantLibs" SkipUnchangedFiles="True" />
     <!-- copy files systematically to output folder -->
     <ItemGroup>
       <BinFiles Include="$(TargetDir)**" />
