--- src/TerrariaNetCore/Terraria/Terraria.csproj
+++ src/tModLoader/Terraria/Terraria.csproj
@@ -3,6 +_,7 @@
 	<Import Project="../Configuration.targets" />
 	<Import Project="../../WorkspaceInfo.targets" />
 	<Import Project="../../../tModBuildTasks/BuildTasks.targets" />
+	<Import Project="../../../tModCodeAssist/tModCodeAssist.targets" />
 
 	<!-- General -->
 	<PropertyGroup>
@@ -11,59 +_,99 @@
 		<Company>Re-Logic</Company>
 		<Copyright>Copyright © 2022 Re-Logic</Copyright>
 		<RootNamespace>Terraria</RootNamespace>
-		<AssemblyName>Terraria</AssemblyName>
+		<AssemblyName>tModLoader</AssemblyName>
-		<ApplicationIcon>Icon.ico</ApplicationIcon>
+		<ApplicationIcon>tModLoader.ico</ApplicationIcon>
-
-		<!-- Avoid overwriting Terraria.exe (if it's not Debug it's release) -->
-		<AssemblyName Condition="$(Configuration.Contains('Debug'))">$(AssemblyName)Debug</AssemblyName>
-		<AssemblyName Condition="!$(Configuration.Contains('Debug'))">$(AssemblyName)Release</AssemblyName>
 		<OutputName>$(AssemblyName)</OutputName>
 	</PropertyGroup>
 
 	<!-- Build Configuration -->
 	<PropertyGroup>
+		<UseAppHost>false</UseAppHost>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
+		
 		<LibrariesDirectory>Libraries</LibrariesDirectory>
 		<GenerateRuntimeConfigDevFile>true</GenerateRuntimeConfigDevFile>
 		<CopyDebugSymbolFilesFromPackages>true</CopyDebugSymbolFilesFromPackages>
 		<CopyDocumentationFilesFromPackages>true</CopyDocumentationFilesFromPackages>
+		<ValidateExecutableReferencesMatchSelfContained>false</ValidateExecutableReferencesMatchSelfContained>
+
+		<DotNetName Condition=" '$(OS)' == 'Windows_NT' ">dotnet.exe</DotNetName>
+		<DotNetName Condition=" '$(OS)' == 'Unix' ">dotnet</DotNetName>
+		<DotNetName Condition=" '$(DotNetName)' == '' ">dotnet</DotNetName>
+
+		<!-- Include annotations in compiled binary for mods to see. -->
+		<DefineConstants>$(DefineConstants);JETBRAINS_ANNOTATIONS</DefineConstants>
+	</PropertyGroup>
+
+	<!-- Versioning -->
+	<PropertyGroup>
+		<tMLVersion Condition="'$(TmlVersion)' == ''">9999.0</tMLVersion>
+		<StableVersion Condition="'$(stableVersion)' == ''">0.0</StableVersion>
+		<BuildPurpose Condition="'$(BuildPurpose)' == ''">dev</BuildPurpose>
+		<BranchName Condition="'$(BranchName)' == ''">unknown</BranchName>
+		<BranchName>$(BranchName.Replace("|","-"))</BranchName>
+		<CommitSHA Condition="'$(CommitSHA)' == ''">unknown</CommitSHA>
+		<BuildDate>$([System.DateTime]::UtcNow.ToBinary())</BuildDate>
+		<SourceRevisionId>$(tMLVersion)|$(StableVersion)|$(BranchName)|$(BuildPurpose)|$(CommitSHA)|$(BuildDate)</SourceRevisionId>
 	</PropertyGroup>
 
 	<!-- References -->
 	<ItemGroup>
 		<ProjectReference Include="../../../FNA/FNA.Core.csproj" />
 		<ProjectReference Include="../ReLogic/ReLogic.csproj" />
-		<EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll">
-			<Link>Libraries/ReLogic/ReLogic.dll</Link>
-			<LogicalName>Terraria.Libraries.ReLogic.ReLogic.dll</LogicalName>
-		</EmbeddedResource>
+		<ProjectReference Include="../../../tModPorter/tModPorter/tModPorter.csproj" />
+		<ProjectReference Include="../../../tModCodeAssist/tModCodeAssist/tModCodeAssist.csproj"
+			ReferenceOutputAssembly="false" />
+		<ProjectReference Include="../../../tModCodeAssist/tModCodeAssist.CodeFixes/tModCodeAssist.CodeFixes.csproj"
+			ReferenceOutputAssembly="false" />
 
 		<Reference Include="CsvHelper" />
-		<Reference Include="Ionic.Zip.CF" />
 		<Reference Include="MP3Sharp" />
 		<Reference Include="Newtonsoft.Json" />
 		<Reference Include="NVorbis" />
 		<Reference Include="RailSDK.Net" />
 		<Reference Include="SteelSeriesEngineWrapper" />
+		<Reference Include="log4net" />
+		<Reference Include="TerrariaHooks" />
 
+		<PackageReference Include="DotNetZip" Version="1.16.0" />
+		<PackageReference Include="JetBrains.Annotations" Version="2025.2.4" />
 		<PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.4" />
+		<PackageReference Include="Basic.Reference.Assemblies.Net80" Version="1.5.0" />
+		<PackageReference Include="MonoMod.RuntimeDetour" Version="25.3.2" />
+		<PackageReference Include="Microsoft.Win32.Registry" Version="6.0.0-preview.5.21301.5 " />
+		<PackageReference Include="Hjson" Version="3.0.0" />
+		<PackageReference Include="System.Reflection.MetadataLoadContext" Version="8.0.0" />
+		<PackageReference Include="Stubble.Core" Version="1.10.8" PrivateAssets="All" />
 	</ItemGroup>
 
 	<!-- Embedded Resources -->
 	<ItemGroup>
+		<EmbeddedResource Include="**\*.png" />
+		<EmbeddedResource Include="**\*.fxc" />
 		<EmbeddedResource Include="GameContent/Creative/Content/*" />
 		<EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
 		<EmbeddedResource Include="GameContent/WorldBuilding/*" />
 		<EmbeddedResource Include="IO/Data/*" />
 		<EmbeddedResource Include="Localization/Content/**" />
 		<EmbeddedResource Include="Microsoft/**" />
+		
+		<!-- Mod Template -->
+		<None Include="ModLoader/Templates/**" />
+		<Compile Remove="ModLoader/Templates/**" />
+		<EmbeddedResource Remove="ModLoader/Templates/**" />
+		<EmbeddedResource Include="ModLoader/Templates/**">
+			<LogicalName>$([System.String]::Copy('Terraria/%(Identity)').Replace('\', '/'))</LogicalName>
+		</EmbeddedResource>
 	</ItemGroup>
 
 	<!-- Files -->
 	<ItemGroup>
-		<None Remove="Libraries/Mono/**" />
+		<Compile Remove="Libraries/Common/TerrariaHooks.dll" />
 		<None Remove="Libraries/Windows/**" />
 		<None Remove="Libraries/XNA/**" />
 		<Content Include="Libraries/Native/**" CopyToOutputDirectory="PreserveNewest" />
+		<Content Include="release_extras/**" CopyToOutputDirectory="PreserveNewest" Link="%(RecursiveDir)%(Filename)%(Extension)" />
 	</ItemGroup>
 
 	<!-- Import tasks -->
@@ -104,9 +_,13 @@
 	<!-- Installs the build's output into the relevant game folder. -->
 	<Target Name="CopyToActualOutputDirectory" AfterTargets="Build">
 		<PropertyGroup>
-			<_ActualOutputDirectory>$(TerrariaSteamPath)</_ActualOutputDirectory>
+			<_ActualOutputDirectory>$(tModLoaderSteamPath)</_ActualOutputDirectory>
 		</PropertyGroup>
 
+		<!-- Exclude tModPorter's metadata. -->
+		<Delete Files="$(TargetDir)\tModPorter.runtimeconfig.json" />
+		<Delete Files="$(TargetDir)\tModPorter.deps.json" />
+		
 		<!-- Copy files systematically to output folder -->
 		<ItemGroup>
 			<BinFiles Include="$(TargetDir)**" />
@@ -114,6 +_,7 @@
 
 		<Message Text="Copying $(AssemblyName) to '$(_ActualOutputDirectory)'..." Importance="high" />
 		<Copy SourceFiles="@(BinFiles)" DestinationFolder="$(_ActualOutputDirectory)/%(RecursiveDir)" SkipUnchangedFiles="True" />
+		<RemoveDir Directories="$(_ActualOutputDirectory)\Content" Condition="!Exists('$(TargetDir)\Content')" />
 
 		<!-- Copy the Libraries folder, purging excessive files -->
 		<SynchronizeDirectories Source="$(TargetDir)Libraries" Destination="$(_ActualOutputDirectory)/Libraries" />
