--- src/Terraria\Terraria\TileObject.cs
+++ src/tModLoader\Terraria\TileObject.cs
@@ -160,7 +_,7 @@
 			return true;
 		}
 
-		public static bool CanPlace(int x, int y, int type, int style, int dir, out TileObject objectData, bool onlyCheck = false)
+		public static bool CanPlace(int x, int y, int type, int style, int dir, out TileObject objectData, bool onlyCheck = false, bool checkStay = false)
 		{
 			TileObjectData tileData = TileObjectData.GetTileData(type, style, 0);
 			objectData = TileObject.Empty;
@@ -326,7 +_,7 @@
 						for (int k = 0; k < tileData2.Height; k++)
 						{
 							Tile tileSafely = Framing.GetTileSafely(num8 + j, num9 + k);
-							bool flag2 = !tileData2.LiquidPlace(tileSafely);
+							bool flag2 = !tileData2.LiquidPlace(tileSafely, checkStay);
 							bool flag3 = false;
 							if (tileData2.AnchorWall)
 							{
@@ -341,7 +_,7 @@
 								}
 							}
 							bool flag4 = false;
-							if (tileSafely.active() && !Main.tileCut[(int)tileSafely.type])
+							if (tileSafely.active() && !Main.tileCut[(int)tileSafely.type] && !checkStay)
 							{
 								flag4 = true;
 							}

