--- src/TerrariaNetCore/Terraria/TileObject.cs
+++ src/tModLoader/Terraria/TileObject.cs
@@ -4,6 +_,7 @@
 using Terraria.Enums;
 using Terraria.GameContent;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.ObjectData;
 
 namespace Terraria;
@@ -14,6 +_,7 @@
 	public int yCoord;
 	public int type;
 	public int style;
+	/// <summary> Note: The index of the alternate within the <see cref="TileObjectData.Alternates"/>, not the alternate placement style offset (<see cref="TileObjectData.Style"/>). This counts from 1, a value of 0 means it is not an alternate. </summary>
 	public int alternate;
 	public int random;
 	public static TileObject Empty = default(TileObject);
@@ -139,7 +_,8 @@
 		return true;
 	}
 
+	//TML: Added 'checkStay' parameter.
-	public static bool CanPlace(int x, int y, int type, int style, int dir, out TileObject objectData, bool onlyCheck = false, int? forcedRandom = null)
+	public static bool CanPlace(int x, int y, int type, int style, int dir, out TileObject objectData, bool onlyCheck = false, int? forcedRandom = null, bool checkStay = false)
 	{
 		TileObjectData tileData = TileObjectData.GetTileData(type, style);
 		objectData = Empty;
@@ -229,7 +_,7 @@
 					num16--;
 
 				if (num16 < rectangle.Y) {
-					rectangle.Width += rectangle.Y - num16;
+					rectangle.Height += rectangle.Y - num16; // Fix issue (OOB exception) with partial right and left anchors.
 					num11 += rectangle.Y - num16;
 					rectangle.Y = num16;
 				}
@@ -247,12 +_,12 @@
 				if (rectangle.X + rectangle.Width == tileData2.Width)
 					rectangle.Width++;
 
-				int num19 = tileData2.AnchorLeft.checkStart;
+				int num19 = tileData2.AnchorRight.checkStart; 
 				if ((tileData2.AnchorRight.type & AnchorType.Tree) == AnchorType.Tree)
 					num19--;
 
 				if (num19 < rectangle.Y) {
-					rectangle.Width += rectangle.Y - num19;
+					rectangle.Height += rectangle.Y - num19; // Fix issue (OOB exception) with partial right and left anchors.
 					num11 += rectangle.Y - num19;
 					rectangle.Y = num19;
 				}
@@ -284,7 +_,7 @@
 			for (int i = 0; i < tileData2.Width; i++) {
 				for (int j = 0; j < tileData2.Height; j++) {
 					Tile tileSafely = Framing.GetTileSafely(num8 + i, num9 + j);
-					bool flag3 = !tileData2.LiquidPlace(tileSafely);
+					bool flag3 = !tileData2.LiquidPlace(tileSafely, checkStay);
 					bool flag4 = false;
 					if (tileData2.AnchorWall) {
 						num25 += 1f;
@@ -295,7 +_,7 @@
 					}
 
 					bool flag5 = false;
-					if (tileSafely.active() && (!Main.tileCut[tileSafely.type] || tileSafely.type == 484 || tileSafely.type == 654) && !TileID.Sets.BreakableWhenPlacing[tileSafely.type])
+					if (tileSafely.active() && (!Main.tileCut[tileSafely.type] || tileSafely.type == 484 || tileSafely.type == 654) && !TileID.Sets.BreakableWhenPlacing[tileSafely.type] && !checkStay)
 						flag5 = true;
 
 					if (flag5 || flag3 || flag4) {
@@ -331,6 +_,7 @@
 									flag6 = true;
 							}
 							else if (Main.tileSolid[tileSafely.type] && Main.tileSolidTop[tileSafely.type]) {
+								// This is buggy since it doesn't check TopSlope and HalfBrick like WorldGen.SolidTileAllowBottomSlope does, causing issues with placing ontop of hammerable tiles like MetalBar. This bug left in until vanilla fix since it might be complicated.
 								flag6 = true;
 							}
 						}
@@ -376,7 +_,7 @@
 					Tile tileSafely = Framing.GetTileSafely(num8 + num31, num9 + num29);
 					bool flag7 = false;
 					if (tileSafely.nactive()) {
-						if (Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0)) {
+						if ((anchorBottom.type & AnchorType.SolidTile) == AnchorType.SolidTile && Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0)) { // AnchorTop
 							flag7 = tileData2.isValidTileAnchor(tileSafely.type);
 							if (flag7)
 								num30 = 0;
@@ -457,7 +_,7 @@
 					Tile tileSafely = Framing.GetTileSafely(num8 + width, num9 + num32);
 					bool flag8 = false;
 					if (tileSafely.nactive()) {
-						if (Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0))
+						if ((anchorBottom.type & AnchorType.SolidTile) == AnchorType.SolidTile && Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0)) // AnchorRight
 							flag8 = tileData2.isValidTileAnchor(tileSafely.type);
 
 						if (!flag8 && (anchorBottom.type & AnchorType.SolidSide) == AnchorType.SolidSide && Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type]) {
@@ -525,7 +_,7 @@
 					Tile tileSafely = Framing.GetTileSafely(num8 + num33, num9 + num34);
 					bool flag9 = false;
 					if (tileSafely.nactive()) {
-						if (Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0))
+						if ((anchorBottom.type & AnchorType.SolidTile) == AnchorType.SolidTile && Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type] && !Main.tileNoAttach[tileSafely.type] && (tileData2.FlattenAnchors || tileSafely.blockType() == 0)) // AnchorLeft
 							flag9 = tileData2.isValidTileAnchor(tileSafely.type);
 
 						if (!flag9 && (anchorBottom.type & AnchorType.SolidSide) == AnchorType.SolidSide && Main.tileSolid[tileSafely.type] && !Main.tileSolidTop[tileSafely.type]) {
@@ -638,8 +_,13 @@
 				Point16 objectStart = TileObjectPreviewData.randomCache.ObjectStart;
 				int num38 = coordinates.X + objectStart.X;
 				int num39 = coordinates.Y + objectStart.Y;
+				int num40 = x - tileObjectData.Origin.X;
+				int num41 = y - tileObjectData.Origin.Y;
+				/* Fix random cache not working with alternates with different origins (mistakenly cycles new random each update), it should use the alternates TOD, not the tile styles TOD
 				int num40 = x - tileData.Origin.X;
 				int num41 = y - tileData.Origin.Y;
+				*/
+
 				if (num38 != num40 || num39 != num41)
 					flag10 = true;
 			}
@@ -818,7 +_,14 @@
 					}
 				}
 				else {
+					/*
 					sb.Draw(value, vector, rectangle, color, 0f, Vector2.Zero, 1f, spriteEffects, 0f);
+					*/
+					bool validPlacement = num11 == 1;
+					if (TileLoader.PreDrawPlacementPreview(num9, num10, op.Type, sb, ref rectangle, ref vector, ref color, validPlacement, ref spriteEffects))
+						sb.Draw(sourceRectangle: rectangle, texture: value, position: vector, color: color, rotation: 0f, origin: Vector2.Zero, scale: 1f, effects: spriteEffects, layerDepth: 0f);
+
+					TileLoader.PostDrawPlacementPreview(num9, num10, op.Type, sb, rectangle, vector, color, validPlacement, spriteEffects);
 				}
 
 				num8 += num12 + tileData.CoordinatePadding;
