--- src/TerrariaNetCore/Terraria/Utils.cs
+++ src/tModLoader/Terraria/Utils.cs
@@ -25,7 +_,7 @@
 
 namespace Terraria;
 
-public static class Utils
+public static partial class Utils
 {
 	public delegate bool TileActionAttempt(int x, int y);
 
@@ -294,7 +_,8 @@
 		ArrayList arrayList = new ArrayList();
 		string text = "";
 		for (int i = 0; i < brokenArgs.Length; i++) {
-			if (brokenArgs[i].StartsWith("-")) {
+			//TML: Steam uses +connect_lobby as an argument, so had to add == "+" so this wouldn't break that argument.
+			if (brokenArgs[i].StartsWith("-") || brokenArgs[i].StartsWith("+")) {
 				if (text != "") {
 					arrayList.Add(text);
 					text = "";
@@ -375,6 +_,14 @@
 							num2 -= 16;
 
 						int num3 = Math.Min(list3[l].Text.Length, num2 / 8);
+
+						// Loop added by TML. Explain what this is?
+						for (int i = 0; i < list3[l].Text.Length; i++) {
+							// TODO: hypen?, more efficient, binary search?, use ChatManager.GetStringSize to support other tags better.
+							if (font.MeasureString(list3[l].Text.Substring(0, i)).X * list3[l].Scale < num2)
+								num3 = i;
+						}
+
 						if (num3 < 0)
 							num3 = 0;
 
@@ -669,7 +_,13 @@
 		return array;
 	}
 
+	/// <summary>
+	/// Generates a random value between 0f (inclusive) and 1f (exclusive). <br/>It will not return 1f.
+	/// </summary>
+	/// <param name="r"></param>
+	/// <returns></returns>
 	public static float NextFloat(this UnifiedRandom r) => (float)r.NextDouble();
+
 	public static float NextFloatDirection(this UnifiedRandom r) => (float)r.NextDouble() * 2f - 1f;
 	public static float NextFloat(this UnifiedRandom random, FloatRange range) => random.NextFloat() * (range.Maximum - range.Minimum) + range.Minimum;
 	public static T NextFromList<T>(this UnifiedRandom random, params T[] objs) => objs[random.Next(objs.Length)];
@@ -1383,14 +_,23 @@
 			text.Substring(0, maxCharactersDisplayed);
 
 		DynamicSpriteFont value = FontAssets.DeathText.Value;
+
+		/*
 		for (int i = -1; i < 2; i++) {
 			for (int j = -1; j < 2; j++) {
-				spriteBatch.DrawString(value, text, pos + new Vector2(i, j), Color.Black, 0f, new Vector2(anchorx, anchory) * value.MeasureString(text), scale, SpriteEffects.None, 0f);
+				spriteBatch.DrawString(value, text, pos + new Vector2((float)i, (float)j), Color.Black, 0f, new Vector2(anchorx, anchory) * value.MeasureString(text), scale, SpriteEffects.None, 0f);
 			}
 		}
 
 		spriteBatch.DrawString(value, text, pos, color, 0f, new Vector2(anchorx, anchory) * value.MeasureString(text), scale, SpriteEffects.None, 0f);
 		return value.MeasureString(text) * scale;
+		*/
+
+		Vector2 vector = ChatManager.GetStringSize(value, text, Vector2.One);
+
+		ChatManager.DrawColorCodedStringWithShadow(spriteBatch, value, text, pos, color, 0f, new Vector2(anchorx, anchory) * vector, new Vector2(scale), -1f, 2f);
+
+		return vector * scale;
 	}
 
 	public static void DrawInvBG(SpriteBatch sb, Rectangle R, Color c = default(Color))
