--- src/TerrariaNetCore/Terraria/WaterfallID.cs
+++ src/tModLoader/Terraria/WaterfallID.cs
@@ -2,7 +_,31 @@
 
 public static class WaterfallID
 {
+	public const int Water = 0;
 	public const int Lava = 1;
+	public const int Rainbow = 2;
+	public const int Corrupt = 3;
+	public const int Jungle = 4;
+	public const int Hallow = 5;
+	public const int Snow = 6;
+	public const int Desert = 7;
+	public const int Underground = 8;
+	public const int Cavern = 9;
+	public const int Blood = 10;
+	public const int Rain = 11;
+	public const int RainBackground = 12;
+	public const int Crimson = 13;
 	public const int Honey = 14;
+	public const int Amethyst = 15;
+	public const int Topaz = 16;
+	public const int Sapphire = 17;
+	public const int Emerald = 18;
+	public const int Ruby = 19;
+	public const int Diamond = 20;
+	public const int Amber = 21;
+	public const int Snowfall = 22;
+	public const int UndergroundDesert = 23;
+	public const int Ocean = 24;
 	public const int Shimmer = 25;
+	public static readonly int Count = 26;
 }
