--- src/TerrariaNetCore/Terraria/WaterfallManager.cs
+++ src/tModLoader/Terraria/WaterfallManager.cs
@@ -1,10 +_,14 @@
 using System;
+using System.Linq.Expressions;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using Microsoft.Xna.Framework.Graphics.PackedVector;
 using ReLogic.Content;
 using Terraria.GameContent.Liquid;
 using Terraria.ID;
 using Terraria.IO;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Core;
 
 namespace Terraria;
 
@@ -21,12 +_,12 @@
 	private const int minWet = 160;
 	private const int maxWaterfallCountDefault = 1000;
 	private const int maxLength = 100;
-	private const int maxTypes = 26;
+	internal const int maxTypes = 26;
 	public int maxWaterfallCount = 1000;
-	private int qualityMax;
+	public int qualityMax;
 	private int currentMax;
-	private WaterfallData[] waterfalls = new WaterfallData[1000];
+	public WaterfallData[] waterfalls = new WaterfallData[1000];
-	private Asset<Texture2D>[] waterfallTexture = new Asset<Texture2D>[26];
+	internal Asset<Texture2D>[] waterfallTexture = new Asset<Texture2D>[26];
 	private int wFallFrCounter;
 	private int regularFrame;
 	private int wFallFrCounter2;
@@ -36,8 +_,8 @@
 	private int rainFrameBackground;
 	private int snowFrameCounter;
 	private int snowFrameForeground;
-	private int findWaterfallCount;
+	public int findWaterfallCount;
-	private int waterfallDist = 100;
+	public int waterfallDist = 100;
 
 	public void BindTo(Preferences preferences)
 	{
@@ -130,15 +_,28 @@
 
 						if ((tile3.liquid > 160 || tile4.liquid > 160) && ((tile3.liquid == 0 && !WorldGen.SolidTile(tile3) && tile3.slope() == 0) || (tile4.liquid == 0 && !WorldGen.SolidTile(tile4) && tile4.slope() == 0)) && currentMax < qualityMax) {
 							waterfalls[currentMax].type = 0;
+							int liquidType = 0;
 							if (tile2.lava() || tile4.lava() || tile3.lava())
 								waterfalls[currentMax].type = 1;
 							else if (tile2.honey() || tile4.honey() || tile3.honey())
 								waterfalls[currentMax].type = 14;
 							else if (tile2.shimmer() || tile4.shimmer() || tile3.shimmer())
 								waterfalls[currentMax].type = 25;
+							else if (tile2.liquidType() >= LiquidID.Count)
+								liquidType = tile2.liquidType();
+							else if (tile4.liquidType() >= LiquidID.Count)
+								liquidType = tile4.liquidType();
+							else if (tile3.liquidType() >= LiquidID.Count)
+								liquidType = tile3.liquidType();
 							else
 								waterfalls[currentMax].type = 0;
 
+							if (liquidType >= LiquidID.Count) {
+								int? waterfallStyle = LiquidLoader.DrawWaterfall(i, j, liquidType);
+								if (waterfallStyle != null)
+									waterfalls[currentMax].type = (int)waterfallStyle;
+							}
+
 							waterfalls[currentMax].x = i;
 							waterfalls[currentMax].y = j;
 							currentMax++;
@@ -146,32 +_,41 @@
 					}
 				}
 
+				WaterfallData? data = TileLoader.CreateWaterfall(i, j, tile.type);
+				if (data == null) {
-				if (tile.type == 196) {
+					if (tile.type == 196) {
-					Tile tile5 = Main.tile[i, j + 1];
+						Tile tile5 = Main.tile[i, j + 1];
-					if (tile5 == null) {
+						if (tile5 == null) {
-						tile5 = new Tile();
+							tile5 = new Tile();
-						Main.tile[i, j + 1] = tile5;
+							Main.tile[i, j + 1] = tile5;
+						}
+
+						if (!WorldGen.SolidTile(tile5) && tile5.slope() == 0 && currentMax < qualityMax) {
+							waterfalls[currentMax].type = 11;
+							waterfalls[currentMax].x = i;
+							waterfalls[currentMax].y = j + 1;
+							currentMax++;
+						}
 					}
 
+					if (tile.type == 460) {
+						Tile tile6 = Main.tile[i, j + 1];
+						if (tile6 == null) {
+							tile6 = new Tile();
+							Main.tile[i, j + 1] = tile6;
+						}
+
-					if (!WorldGen.SolidTile(tile5) && tile5.slope() == 0 && currentMax < qualityMax) {
+						if (!WorldGen.SolidTile(tile6) && tile6.slope() == 0 && currentMax < qualityMax) {
-						waterfalls[currentMax].type = 11;
+							waterfalls[currentMax].type = 22;
-						waterfalls[currentMax].x = i;
+							waterfalls[currentMax].x = i;
-						waterfalls[currentMax].y = j + 1;
+							waterfalls[currentMax].y = j + 1;
-						currentMax++;
+							currentMax++;
+						}
 					}
 				}
-
-				if (tile.type == 460) {
-					Tile tile6 = Main.tile[i, j + 1];
-					if (tile6 == null) {
-						tile6 = new Tile();
-						Main.tile[i, j + 1] = tile6;
-					}
-
-					if (!WorldGen.SolidTile(tile6) && tile6.slope() == 0 && currentMax < qualityMax) {
-						waterfalls[currentMax].type = 22;
-						waterfalls[currentMax].x = i;
-						waterfalls[currentMax].y = j + 1;
+				else if (((WaterfallData)data).type > -1) {
+					if (currentMax < qualityMax) {
+						waterfalls[currentMax] = (WaterfallData)data;
 						currentMax++;
 					}
 				}
@@ -216,9 +_,46 @@
 			if (++snowFrameForeground > 7)
 				snowFrameForeground = 0;
 		}
+		
+		for (int i = 0; i < Main.wFallFrame.Length; i++) {
+			if (i >= WaterfallID.Count) {
+				ModWaterfallStyle fall = LoaderManager.Get<WaterFallStylesLoader>().Get(i);
+				if (LoaderUtils.HasOverride(fall, (Expression<Func<ModWaterfallStyle, WaterFallStylesLoader.DelegateAnimateWaterfall>>)((ModWaterfallStyle t) => t.AnimateWaterfall))) {
+					LoaderManager.Get<WaterFallStylesLoader>().Get(i).AnimateWaterfall(ref Main.wFallFrame[i], ref Main.wFallFrameBack[i], ref Main.wFallFrameCounter[i]);
+					continue;
+				}
+				Main.wFallFrameCounter[i]++;
+				if (Main.wFallFrameCounter[i] > 2) {
+					Main.wFallFrameCounter[i] = 0;
+					Main.wFallFrame[i]++;
+					if (Main.wFallFrame[i] > 15) {
+						Main.wFallFrame[i] = 0;
+					}
+				}
+			}
+			else {
+				if (i == WaterfallID.Lava || i == WaterfallID.Honey || i == WaterfallID.Shimmer) {
+					Main.wFallFrame[i] = slowFrame;
+					Main.wFallFrameCounter[i] = wFallFrCounter2;
+				}
+				else if (i == WaterfallID.Rain) {
+					Main.wFallFrame[i] = rainFrameForeground;
+					Main.wFallFrameBack[i] = rainFrameBackground;
+					Main.wFallFrameCounter[i] = rainFrameCounter;
+				}
+				else if (i == WaterfallID.Snow) {
+					Main.wFallFrame[i] = snowFrameForeground;
+					Main.wFallFrameCounter[i] = snowFrameCounter;
+				}
+				else {
+					Main.wFallFrame[i] = regularFrame;
+					Main.wFallFrameCounter[i] = wFallFrCounter;
+				}
+			}
+		}
 	}
 
-	private void DrawWaterfall(int Style = 0, float Alpha = 1f)
+	internal void DrawWaterfall(int Style = 0, float Alpha = 1f)
 	{
 		Main.tileSolid[546] = false;
 		float num = 0f;
@@ -244,6 +_,8 @@
 			int num20 = 0;
 			int num21;
 			int num22;
+			if (num12 >= WaterfallID.Count && !LoaderManager.Get<WaterFallStylesLoader>().Get(num12).PreDraw(waterfalls[i], num13, num14, Main.spriteBatch))
+				continue;
 			if (num12 == 1 || num12 == 14 || num12 == 25) {
 				if (Main.drewLava || waterfalls[i].stopAtStep == 0)
 					continue;
@@ -359,6 +_,9 @@
 				}
 
 				num21 = 32 * regularFrame;
+				if (num12 >= WaterfallID.Count) {
+					num21 = 32 * Main.wFallFrame[num12];
+				}
 			}
 
 			int num29 = 0;
@@ -480,15 +_,23 @@
 					case 267:
 					case 268:
 						num12 = 15 + num34 - 262;
+						// Patch note: ^ Both are used below.
 						break;
 				}
+
+				if (num34 != -1)
+					TileLoader.ChangeWaterfallStyle(num34, ref num12);
 
 				Color color5 = Lighting.GetColor(num13, num14);
 				if (k > 50)
 					TrySparkling(num13, num14, num17, color5);
 
-				float alpha = GetAlpha(Alpha, num22, num12, num14, k, tile3);
+				float alpha = GetAlpha(Alpha, num22, num12, num14, k, num13, num14);
 				color5 = StylizeColor(alpha, num22, num12, num14, k, tile3, color5);
+				bool canPlaySound = true;
+				if (num12 >= WaterfallID.Count) {
+					canPlaySound = LoaderManager.Get<WaterFallStylesLoader>().Get(num12).PlayWaterfallSounds();
+				}
 				if (num12 == 1) {
 					float num35 = Math.Abs((float)(num13 * 16 + 8) - (Main.screenPosition.X + (float)(Main.screenWidth / 2)));
 					float num36 = Math.Abs((float)(num14 * 16 + 8) - (Main.screenPosition.Y + (float)(Main.screenHeight / 2)));
@@ -509,7 +_,7 @@
 						num10 = num14 * 16 + 8;
 					}
 				}
-				else if (num12 != 1 && num12 != 14 && num12 != 25 && num12 != 11 && num12 != 12 && num12 != 22) {
+				else if (num12 != 1 && num12 != 14 && num12 != 25 && num12 != 11 && num12 != 12 && num12 != 22 && canPlaySound) {
 					float num39 = Math.Abs((float)(num13 * 16 + 8) - (Main.screenPosition.X + (float)(Main.screenWidth / 2)));
 					float num40 = Math.Abs((float)(num14 * 16 + 8) - (Main.screenPosition.Y + (float)(Main.screenHeight / 2)));
 					if (num39 < (float)(Main.screenWidth * 2) && num40 < (float)(Main.screenHeight * 2)) {
@@ -689,6 +_,8 @@
 				if ((tile4.active() && (tile4.type == 189 || tile4.type == 196)) || (tile6.active() && (tile6.type == 189 || tile6.type == 196)) || (tile5.active() && (tile5.type == 189 || tile5.type == 196)))
 					num22 = (int)(40f * ((float)Main.maxTilesX / 4200f) * Main.gfxQuality);
 			}
+			if (num12 >= WaterfallID.Count)
+				LoaderManager.Get<WaterFallStylesLoader>().Get(num12).PostDraw(waterfalls[i], num13, num14, Main.spriteBatch);
 		}
 
 		Main.ambientWaterfallX = num4;
@@ -700,7 +_,7 @@
 		Main.tileSolid[546] = true;
 	}
 
-	private void DrawWaterfall(int waterfallType, int x, int y, float opacity, Vector2 position, Rectangle sourceRect, Color color, SpriteEffects effects)
+	public void DrawWaterfall(int waterfallType, int x, int y, float opacity, Vector2 position, Rectangle sourceRect, Color color, SpriteEffects effects)
 	{
 		Texture2D value = waterfallTexture[waterfallType].Value;
 		if (waterfallType == 25) {
@@ -716,7 +_,7 @@
 		}
 	}
 
-	private static Color StylizeColor(float alpha, int maxSteps, int waterfallType, int y, int s, Tile tileCache, Color aColor)
+	public static Color StylizeColor(float alpha, int maxSteps, int waterfallType, int y, int s, Tile tileCache, Color aColor)
 	{
 		float num = (float)(int)aColor.R * alpha;
 		float num2 = (float)(int)aColor.G * alpha;
@@ -746,15 +_,21 @@
 				num = 255f * alpha;
 				num2 = 255f * alpha;
 				num3 = 255f * alpha;
+				// Patch note: ^ All 3 are used below.
 				break;
 		}
+
+		if (waterfallType >= maxTypes)
+			LoaderManager.Get<WaterFallStylesLoader>().Get(waterfallType).ColorMultiplier(ref num, ref num2, ref num3, alpha);
 
 		aColor = new Color((int)num, (int)num2, (int)num3, (int)num4);
 		return aColor;
 	}
 
+	// Added x/y parameters and removed the tileCache parameters
-	private static float GetAlpha(float Alpha, int maxSteps, int waterfallType, int y, int s, Tile tileCache)
+	public static float GetAlpha(float Alpha, int maxSteps, int waterfallType, int y, int s, int tileX, int tileY)
 	{
+		Tile tileCache = Main.tile[tileX, tileY];
 		float num;
 		switch (waterfallType) {
 			case 1:
@@ -774,10 +_,16 @@
 		if (s > maxSteps - 10)
 			num *= (float)(maxSteps - s) / 10f;
 
+		if (waterfallType >= WaterfallID.Count) {
+			float? alpha = LoaderManager.Get<WaterFallStylesLoader>().Get(waterfallType).Alpha(tileX, tileY, Alpha, maxSteps, s, tileCache);
+			if (alpha != null) {
+				num = (float)alpha;
+			}
+		}
 		return num;
 	}
 
-	private static void TrySparkling(int x, int y, int direction, Color aColor2)
+	public static void TrySparkling(int x, int y, int direction, Color aColor2)
 	{
 		if (aColor2.R > 20 || aColor2.B > 20 || aColor2.G > 20) {
 			float num = (int)aColor2.R;
@@ -794,8 +_,11 @@
 		}
 	}
 
-	private static void AddLight(int waterfallType, int x, int y)
+	public static void AddLight(int waterfallType, int x, int y)
 	{
+		if (waterfallType >= maxTypes)
+			LoaderManager.Get<WaterFallStylesLoader>().Get(waterfallType).AddLight(x, y);
+		
 		switch (waterfallType) {
 			case 1: {
 				float r;
@@ -913,5 +_,7 @@
 
 		if (Main.liquidAlpha[13] > 0f)
 			DrawWaterfall(24, Main.liquidAlpha[13]);
+
+		LoaderManager.Get<WaterStylesLoader>().DrawWaterfall(this);
 	}
 }
