--- src/Terraria/Terraria/WaterfallManager.cs
+++ src/tModLoader/Terraria/WaterfallManager.cs
@@ -2,6 +_,7 @@
 using Microsoft.Xna.Framework.Graphics;
 using System;
 using System.IO;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -18,7 +_,7 @@
 		private const int minWet = 160;
 		private const int maxCount = 200;
 		private const int maxLength = 100;
-		private const int maxTypes = 23;
+		internal const int maxTypes = 23;
 		private int qualityMax;
 		private int currentMax;
 		private WaterfallData[] waterfalls;
@@ -201,7 +_,7 @@
 			}
 		}
 
-		private void DrawWaterfall(SpriteBatch spriteBatch, int Style = 0, float Alpha = 1f) {
+		internal void DrawWaterfall(SpriteBatch spriteBatch, int Style = 0, float Alpha = 1f) {
 			float num = 0f;
 			float num2 = 99999f;
 			float num3 = 99999f;
@@ -344,6 +_,7 @@
 					if (num29 >= 2)
 						continue;
 
+					//patch file: num12
 					switch (num12) {
 						case 1: {
 								float r;
@@ -409,6 +_,8 @@
 								break;
 							}
 					}
+					if (num12 >= WaterfallStyleLoader.vanillaWaterfallCount)
+						WaterfallStyleLoader.GetWaterfallStyle(num12).AddLight(num13, num14);
 
 					int num31 = 0;
 					int num32 = 0;
@@ -515,6 +_,9 @@
 							num12 = 15 + num36 - 262;
 							break;
 					}
+					if(num36 != -1) {
+						TileLoader.ChangeWaterfallStyle(num36, ref num12);
+					}
 
 					if (WorldGen.SolidTile(tile4) && !tile2.halfBrick())
 						num11 = 8;
@@ -570,6 +_,9 @@
 							break;
 					}
 
+					if (num12 >= WaterfallStyleLoader.vanillaWaterfallCount)
+						WaterfallStyleLoader.GetWaterfallStyle(num12).ColorMultiplier(ref num38, ref num39, ref num40, num37);
+
 					color5 = new Color((int)num38, (int)num39, (int)num40, (int)num41);
 					if (num12 == 1) {
 						float num42 = Math.Abs((float)(num13 * 16 + 8) - (Main.screenPosition.X + (float)(Main.screenWidth / 2)));
@@ -809,6 +_,8 @@
 
 			if (Main.liquidAlpha[10] > 0f)
 				DrawWaterfall(spriteBatch, 13, Main.liquidAlpha[10]);
+
+			WaterStyleLoader.DrawWaterfall(this, spriteBatch);
 		}
 	}
 }
