--- src/Terraria\Terraria\WindowsLaunch.cs
+++ src/tModLoader\Terraria\WindowsLaunch.cs
@@ -63,12 +63,12 @@
 					result = Assembly.Load(array);
 				}
 				return result;
-			};
+            };
 #if SERVER
 			WindowsLaunch._handleRoutine = new WindowsLaunch.HandlerRoutine(WindowsLaunch.ConsoleCtrlCheck);
 			WindowsLaunch.SetConsoleCtrlHandler(WindowsLaunch._handleRoutine, true);
 #endif
-			Program.LaunchGame(args);
+            Program.LaunchGame(args);
 		}
 	}
 }

