--- src/Terraria\Terraria\Wiring.cs
+++ src/tModLoader\Terraria\Wiring.cs
@@ -3,6 +3,7 @@
 using System.Collections.Generic;
 using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria
 {
@@ -12,15 +13,15 @@
 		private const int MaxMech = 1000;
 		public static bool running;
 		private static Dictionary<Point16, bool> _wireSkip;
-		private static DoubleStack<Point16> _wireList;
-		private static Dictionary<Point16, byte> _toProcess;
-		private static Vector2[] _teleport;
-		private static int[] _inPumpX;
-		private static int[] _inPumpY;
-		private static int _numInPump;
-		private static int[] _outPumpX;
-		private static int[] _outPumpY;
-		private static int _numOutPump;
+		public static DoubleStack<Point16> _wireList;
+		public static Dictionary<Point16, byte> _toProcess;
+		public static Vector2[] _teleport;
+		public static int[] _inPumpX;
+		public static int[] _inPumpY;
+		public static int _numInPump;
+		public static int[] _outPumpX;
+		public static int[] _outPumpY;
+		public static int _numOutPump;
 		private static int[] _mechX;
 		private static int[] _mechY;
 		private static int _numMechs;
@@ -204,7 +205,7 @@
 			}
 		}
 
-		private static bool CheckMech(int i, int j, int time)
+		public static bool CheckMech(int i, int j, int time)
 		{
 			for (int k = 0; k < Wiring._numMechs; k++)
 			{
@@ -278,7 +279,7 @@
 			}
 		}
 
-		private static void TripWire(int left, int top, int width, int height)
+		public static void TripWire(int left, int top, int width, int height)
 		{
 			if (Main.netMode == 1)
 			{
@@ -656,7 +657,7 @@
 						NetMessage.SendData(19, -1, -1, "", 4 + flag.ToInt(), (float)i, (float)j, 0f, 0, 0, 0);
 						return;
 					}
-					if (type == 11)
+					if (TileLoader.CloseDoorID(Main.tile[i, j]) >= 0)
 					{
 						if (WorldGen.CloseDoor(i, j, true))
 						{
@@ -664,7 +665,7 @@
 							return;
 						}
 					}
-					else if (type == 10)
+					else if (TileLoader.OpenDoorID(Main.tile[i, j]) >= 0)
 					{
 						int num26 = 1;
 						if (Main.rand.Next(2) == 0)
@@ -762,7 +763,7 @@
 						}
 						else
 						{
-							if (type == 4)
+							if (TileLoader.IsTorch(type))
 							{
 								if (tile.frameX < 66)
 								{
@@ -1108,7 +1109,7 @@
 								}
 								else
 								{
-									if (type == 139 || type == 35)
+									if (type == 139 || type == 35 || TileLoader.IsModMusicBox(tile))
 									{
 										WorldGen.SwitchMB(i, j);
 										return;
@@ -1454,6 +1455,7 @@
 						}
 					}
 				}
+				TileLoader.HitWire(i, j, type);
 			}
 		}
 

