--- src/TerrariaNetCore/Terraria/WorldSections.cs
+++ src/tModLoader/Terraria/WorldSections.cs
@@ -3,7 +_,7 @@
 
 namespace Terraria;
 
-public class WorldSections
+public partial class WorldSections
 {
 	private struct IterationState
 	{
@@ -81,6 +_,10 @@
 		}
 	}
 
+	/// <summary>
+	/// Checks if the tile at the specified tile coordinate has been loaded for this client in a multiplayer game session. In multiplayer, sections of tiles are sent when the player visits them, so much of the map will not be loaded for a client. until visited.
+	/// <para/> Modders may need to check this for client code that could potentially access unloaded tiles.
+	/// </summary>
 	public bool TileLoaded(int tileX, int tileY) => SectionLoaded(Netplay.GetSectionX(tileX), Netplay.GetSectionY(tileY));
 
 	public bool SectionLoaded(int x, int y)
