--- src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/TerrariaNetCore/Terraria/Graphics/Capture/CaptureManager.cs
@@ -6,7 +_,12 @@
 
 public class CaptureManager : IDisposable
 {
+#if NETCORE
+	// Client initialization has been moved to Main.ClientInitialize to prevent FNA deadlocks
+	public static CaptureManager Instance = Main.dedServ ? new CaptureManager() : null;
+#else
 	public static CaptureManager Instance = new CaptureManager();
+#endif
 	private CaptureInterface _interface;
 	private CaptureCamera _camera;
 
