--- src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/TerrariaNetCore/Terraria/Graphics/Capture/CaptureManager.cs
@@ -6,7 +_,8 @@
 
 public class CaptureManager : IDisposable
 {
-	public static CaptureManager Instance = new CaptureManager();
+	// Client initialization has been moved to Main.LoadContent to prevent FNA deadlocks
+	public static CaptureManager Instance = Main.dedServ ? new CaptureManager() : null;
 	private CaptureInterface _interface;
 	private CaptureCamera _camera;
 
