--- src/TerrariaNetCore/Terraria/Graphics/Effects/FilterManager.cs
+++ src/tModLoader/Terraria/Graphics/Effects/FilterManager.cs
@@ -183,4 +_,13 @@
 
 	public bool HasActiveFilter() => _activeFilters.Count != 0;
 	public bool CanCapture() => HasActiveFilter();
+
+	// Added by TML.
+	internal void DeactivateAll()
+	{
+		foreach (string key in _effects.Keys) {
+			if (this[key].IsActive())
+				this[key].Deactivate(Array.Empty<object>());
+		}
+	}
 }
