--- src/TerrariaNetCore/Terraria/Graphics/Effects/FilterManager.cs
+++ src/tModLoader/Terraria/Graphics/Effects/FilterManager.cs
@@ -3,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.IO;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Effects;
 
@@ -158,7 +_,8 @@
 			linkedListNode = next;
 		}
 
-		graphicsDevice.SetRenderTarget(finalTexture);
+		bool requiresScreenTarget = SystemLoader.RequiresScreenTarget();
+		graphicsDevice.SetRenderTarget(requiresScreenTarget ? t2 : finalTexture);
 		graphicsDevice.Clear(Color.Transparent);
 		if (Main.player[Main.myPlayer].gravDir == -1f)
 			Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend, SamplerState.LinearClamp, DepthStencilState.Default, RasterizerState.CullNone, null, Main.GameViewMatrix.EffectMatrix);
@@ -174,6 +_,19 @@
 		}
 
 		Main.spriteBatch.End();
+
+		if (requiresScreenTarget) {
+			Main.finalScreenTarget = t2;
+
+			graphicsDevice.SetRenderTarget(finalTexture);
+			graphicsDevice.Clear(Color.Transparent);
+
+			Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend);
+
+			Main.spriteBatch.Draw(t2, Vector2.Zero, value, Color.White);
+
+			Main.spriteBatch.End();
+		}
 		for (int i = 0; i < 8; i++) {
 			graphicsDevice.Textures[i] = null;
 		}
@@ -183,4 +_,13 @@
 
 	public bool HasActiveFilter() => _activeFilters.Count != 0;
 	public bool CanCapture() => HasActiveFilter();
+
+	// Added by TML.
+	internal void DeactivateAll()
+	{
+		foreach (string key in _effects.Keys) {
+			if (this[key].IsActive())
+				this[key].Deactivate(Array.Empty<object>());
+		}
+	}
 }
