--- src/TerrariaNetCore/Terraria/Graphics/Effects/OverlayManager.cs
+++ src/tModLoader/Terraria/Graphics/Effects/OverlayManager.cs
@@ -77,12 +_,17 @@
 		}
 	}
 
-	public void Draw(SpriteBatch spriteBatch, RenderLayers layer)
+	//TML: Added 'beginSpriteBatch' parameter.
+	public void Draw(SpriteBatch spriteBatch, RenderLayers layer, bool beginSpriteBatch = false)
 	{
 		if (_overlayCount == 0)
 			return;
 
+		/*
 		bool flag = false;
+		*/
+		bool flag = !beginSpriteBatch;
+
 		for (int i = 0; i < _activeOverlays.Length; i++) {
 			for (LinkedListNode<Overlay> linkedListNode = _activeOverlays[i].First; linkedListNode != null; linkedListNode = linkedListNode.Next) {
 				Overlay value = linkedListNode.Value;
@@ -97,7 +_,10 @@
 			}
 		}
 
+		/*
 		if (flag)
+		*/
+		if (flag && beginSpriteBatch)
 			spriteBatch.End();
 	}
 }
