--- src/TerrariaNetCore/Terraria/Graphics/Effects/SkyManager.cs
+++ src/tModLoader/Terraria/Graphics/Effects/SkyManager.cs
@@ -1,3 +_,4 @@
+using System;
 using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
@@ -21,11 +_,13 @@
 
 	public void Update(GameTime gameTime)
 	{
+		/*
 		int num = Main.dayRate;
 		if (num < 1)
 			num = 1;
+		*/
 
-		for (int i = 0; i < num; i++) {
+		for (int i = 0; i < Main.worldEventUpdates; i++) {
 			LinkedListNode<CustomSky> linkedListNode = _activeSkies.First;
 			while (linkedListNode != null) {
 				CustomSky value = linkedListNode.Value;
@@ -98,5 +_,14 @@
 		}
 
 		return MathHelper.Clamp(num, 0f, 1f);
+	}
+
+	// Added by TML.
+	internal void DeactivateAll()
+	{
+		foreach (string key in _effects.Keys) {
+			if (this[key].IsActive())
+				this[key].Deactivate(Array.Empty<object>());
+		}
 	}
 }
