--- src/Terraria\Terraria.GameContent.Biomes\CaveHouseBiome.cs
+++ src/tModLoader\Terraria.GameContent.Biomes\CaveHouseBiome.cs
@@ -146,7 +_,7 @@
 		}
 
 		private const int VERTICAL_EXIT_WIDTH = 3;
-		private static readonly bool[] _blacklistedTiles = TileID.Sets.Factory.CreateBoolSet(true, 225, 41, 43, 44, 226, 203, 112, 25, 151);
+		public static readonly bool[] _blacklistedTiles = TileID.Sets.Factory.CreateBoolSet(true, 225, 41, 43, 44, 226, 203, 112, 25, 151);
 		private int _sharpenerCount;
 		private int _extractinatorCount;
 
@@ -409,7 +_,7 @@
 				}
 
 				int num18 = item14.Width / 8 + 3;
-				WorldGen.SetupStatueList();
+				// Commented out because it makes no sense: WorldGen.SetupStatueList();
 				while (num18 > 0) {
 					int num19 = GenBase._random.Next(item14.Width - 3) + 1 + item14.X;
 					int num20 = item14.Y + item14.Height - 2;
