--- src/TerrariaNetCore/Terraria/Chat/ChatCommandProcessor.cs
+++ src/tModLoader/Terraria/Chat/ChatCommandProcessor.cs
@@ -3,6 +_,7 @@
 using ReLogic.Utilities;
 using Terraria.Chat.Commands;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria.Chat;
 
@@ -107,6 +_,11 @@
 
 	public void ProcessIncomingMessage(ChatMessage message, int clientId)
 	{
+		if (message.CommandId._name == "Say" && CommandLoader.HandleCommand(message.Text, new PlayerCommandCaller(Main.player[clientId]))) {
+			message.Consume();
+			return;
+		}
+
 		if (_commands.TryGetValue(message.CommandId, out var value)) {
 			value.ProcessIncomingMessage(message.Text, (byte)clientId);
 			message.Consume();
