--- src/TerrariaNetCore/Terraria/GameContent/UI/ResourceSets/PlayerStatsSnapshot.cs
+++ src/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerStatsSnapshot.cs
@@ -5,10 +_,38 @@
 	public int Life;
 	public int LifeMax;
 	public int LifeFruitCount;
+	/*
 	public float LifePerSegment;
+	*/
 	public int Mana;
 	public int ManaMax;
+	/*
 	public float ManaPerSegment;
+	*/
+
+	// Fields & properties below this point were added by TML.
+
+	private int numLifeHearts;
+	private int numManaStars;
+
+	public float LifePerSegment => LifeMax / (float)AmountOfLifeHearts;
+	public float ManaPerSegment => ManaMax / (float)AmountOfManaStars;
+
+	/// <summary>
+	/// How many life hearts should be drawn
+	/// </summary>
+	public int AmountOfLifeHearts {
+		get => numLifeHearts;
+		set => numLifeHearts = Utils.Clamp(value, int.MinValue, 20);
+	}
+
+	/// <summary>
+	/// How many mana stars should be drawn
+	/// </summary>
+	public int AmountOfManaStars {
+		get => numManaStars;
+		set => numManaStars = Utils.Clamp(value, int.MinValue, 20);
+	}
 
 	public PlayerStatsSnapshot(Player player)
 	{
@@ -18,7 +_,11 @@
 		ManaMax = player.statManaMax2;
 		float num = 20f;
 		int num2 = player.statLifeMax / 20;
+
+		/*
 		int num3 = (player.statLifeMax - 400) / 5;
+		*/
+		int num3 = player.ConsumedLifeFruit;
 		if (num3 < 0)
 			num3 = 0;
 
@@ -30,7 +_,12 @@
 		int num4 = player.statLifeMax2 - player.statLifeMax;
 		num += (float)(num4 / num2);
 		LifeFruitCount = num3;
+
+		/*
 		LifePerSegment = num;
 		ManaPerSegment = 20f;
+		*/
+		numLifeHearts = (int)(LifeMax / num);
+		numManaStars = (int)(ManaMax / 20d);
 	}
 }
