--- src/TerrariaNetCore/Terraria/GameContent/UI/ResourceSets/PlayerResourceSetsManager.cs
+++ src/tModLoader/Terraria/GameContent/UI/ResourceSets/PlayerResourceSetsManager.cs
@@ -5,7 +_,7 @@
 
 namespace Terraria.GameContent.UI.ResourceSets
 {
-	public class PlayerResourceSetsManager
+	public partial class PlayerResourceSetsManager
 	{
 		private Dictionary<string, IPlayerResourcesDisplaySet> _sets = new Dictionary<string, IPlayerResourcesDisplaySet>();
 		private IPlayerResourcesDisplaySet _activeSet;
@@ -23,7 +_,7 @@
 		}
 
 		private void Configuration_OnLoad(Preferences obj) {
-			_activeSetConfigKey = Main.Configuration.Get("PlayerResourcesSet", "New");
+			_activeSetConfigKeyOriginal = _activeSetConfigKey = Main.Configuration.Get("PlayerResourcesSet", "New");
 			if (_loadedContent)
 				SetActiveFromLoadedConfigKey();
 		}
@@ -35,8 +_,8 @@
 
 		public void LoadContent(AssetRequestMode mode) {
 			_sets["New"] = new FancyClassicPlayerResourcesDisplaySet("New", "New", "FancyClassic", mode);
-			_sets["Default"] = new ClassicPlayerResourcesDisplaySet("Default", "Default");
 			_sets["HorizontalBars"] = new HorizontalBarsPlayerReosurcesDisplaySet("HorizontalBars", "HorizontalBars", "HorizontalBars", mode);
+			_sets["Default"] = new ClassicPlayerResourcesDisplaySet("Default", "Default");
 			_loadedContent = true;
 			SetActiveFromLoadedConfigKey();
 		}
@@ -69,7 +_,7 @@
 
 		public void CycleResourceSet() {
 			IPlayerResourcesDisplaySet lastFrame = null;
-			_sets.Values.FirstOrDefault(delegate (IPlayerResourcesDisplaySet frame) {
+			_sets.Values.Reverse().FirstOrDefault(delegate (IPlayerResourcesDisplaySet frame) {
 				if (frame == _activeSet)
 					return true;
 
@@ -78,7 +_,7 @@
 			});
 
 			if (lastFrame == null)
-				lastFrame = _sets.Values.Last();
+				lastFrame = _sets.Values.First();
 
 			SetActiveFrame(lastFrame);
 		}
