--- src/Terraria\Terraria.GameContent.UI.Elements\UICharacterListItem.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UICharacterListItem.cs
@@ -1,8 +_,10 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using System;
+using System.Reflection;
 using Terraria.Graphics;
 using Terraria.IO;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.UI;
 
@@ -22,6 +_,7 @@
 		private Texture2D _buttonFavoriteInactiveTexture;
 		private Texture2D _buttonPlayTexture;
 		private Texture2D _buttonDeleteTexture;
+		private Texture2D _errorTexture;
 		private UIImageButton _deleteButton;
 
 		public bool IsFavorite
@@ -43,6 +_,8 @@
 			this._buttonFavoriteInactiveTexture = TextureManager.Load("Images/UI/ButtonFavoriteInactive");
 			this._buttonPlayTexture = TextureManager.Load("Images/UI/ButtonPlay");
 			this._buttonDeleteTexture = TextureManager.Load("Images/UI/ButtonDelete");
+			this._errorTexture = Texture2D.FromStream(Main.instance.GraphicsDevice,
+				Assembly.GetExecutingAssembly().GetManifestResourceStream("Terraria.ModLoader.UI.ButtonError.png"));
 			this.Height.Set(96f, 0f);
 			this.Width.Set(0f, 1f);
 			base.SetPadding(6f);
@@ -89,6 +_,17 @@
 			{
 				base.Append(uIImageButton4);
 			}
+			if (data.customDataFail != null)
+			{
+				UIImageButton uIImageButton5 = new UIImageButton(this._errorTexture);
+				uIImageButton5.VAlign = 1f;
+				uIImageButton5.HAlign = 1f;
+				uIImageButton5.Left.Set(-24f, 0f);
+				uIImageButton5.OnClick += new UIElement.MouseEvent(this.ErrorButtonClick);
+				uIImageButton5.OnMouseOver += new UIElement.MouseEvent(this.ErrorMouseOver);
+				uIImageButton5.OnMouseOut += new UIElement.MouseEvent(this.DeleteMouseOut);
+				base.Append(uIImageButton5);
+			}
 			this._buttonLabel = new UIText("", 1f, false);
 			this._buttonLabel.VAlign = 1f;
 			this._buttonLabel.Left.Set(80f, 0f);
@@ -97,7 +_,7 @@
 			this._deleteButtonLabel = new UIText("", 1f, false);
 			this._deleteButtonLabel.VAlign = 1f;
 			this._deleteButtonLabel.HAlign = 1f;
-			this._deleteButtonLabel.Left.Set(-30f, 0f);
+			this._deleteButtonLabel.Left.Set(data.customDataFail == null ? -30f : -54f, 0f);
 			this._deleteButtonLabel.Top.Set(-3f, 0f);
 			base.Append(this._deleteButtonLabel);
 			uIImageButton.SetSnapPoint("Play", snapPointIndex, null, null);
@@ -133,6 +_,11 @@
 		private void DeleteMouseOver(UIMouseEvent evt, UIElement listeningElement)
 		{
 			this._deleteButtonLabel.SetText("Delete");
+		}
+
+		private void ErrorMouseOver(UIMouseEvent evt, UIElement listeningElement)
+		{
+			this._deleteButtonLabel.SetText(this._data.customDataFail.modName + " Error");
 		}
 
 		private void DeleteMouseOut(UIMouseEvent evt, UIElement listeningElement)
@@ -210,6 +_,11 @@
 			{
 				uIList.UpdateOrder();
 			}
+		}
+
+		private void ErrorButtonClick(UIMouseEvent evt, UIElement listeningElement)
+		{
+			ErrorLogger.LogException(this._data.customDataFail.InnerException);
 		}
 
 		public override int CompareTo(object obj)
@@ -269,11 +_,11 @@
 			this.DrawPanel(spriteBatch, vector2, num2);
 			spriteBatch.Draw(Main.heartTexture, vector2 + new Vector2(5f, 2f), Color.White);
 			vector2.X += 10f + (float)Main.heartTexture.Width;
-			Utils.DrawBorderString(spriteBatch, this._data.Player.statLifeMax + " HP", vector2 + new Vector2(0f, 3f), Color.White, 1f, 0f, 0f, -1);
+			Utils.DrawBorderString(spriteBatch, this._data.Player.statLifeMax2 + " HP", vector2 + new Vector2(0f, 3f), Color.White, 1f, 0f, 0f, -1);
 			vector2.X += 65f;
 			spriteBatch.Draw(Main.manaTexture, vector2 + new Vector2(5f, 2f), Color.White);
 			vector2.X += 10f + (float)Main.manaTexture.Width;
-			Utils.DrawBorderString(spriteBatch, this._data.Player.statManaMax + " MP", vector2 + new Vector2(0f, 3f), Color.White, 1f, 0f, 0f, -1);
+			Utils.DrawBorderString(spriteBatch, this._data.Player.statManaMax2 + " MP", vector2 + new Vector2(0f, 3f), Color.White, 1f, 0f, 0f, -1);
 			vector.X += num2 + 5f;
 			Vector2 vector3 = vector;
 			float num3 = 110f;

