--- src/Terraria\Terraria.GameContent.UI.Elements\UIKeybindingSimpleListItem.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UIKeybindingSimpleListItem.cs
@@ -6,7 +_,7 @@
 
 namespace Terraria.GameContent.UI.Elements
 {
-	internal class UIKeybindingSimpleListItem : UIElement
+	public class UIKeybindingSimpleListItem : UIElement
 	{
 		private Color _color;
 		private Func<string> _GetTextFunction;

