--- src/Terraria\Terraria.GameContent.UI.Elements\UIKeybindingToggleListItem.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UIKeybindingToggleListItem.cs
@@ -7,7 +_,7 @@
 
 namespace Terraria.GameContent.UI.Elements
 {
-	internal class UIKeybindingToggleListItem : UIElement
+	public class UIKeybindingToggleListItem : UIElement
 	{
 		private Color _color;
 		private Func<string> _TextDisplayFunction;

