--- src/Terraria/Terraria.GameContent.UI.Elements/UIPanel.cs
+++ src/tModLoader/Terraria.GameContent.UI.Elements/UIPanel.cs
@@ -14,13 +_,16 @@
 		public Color BorderColor = Color.Black;
 		public Color BackgroundColor = new Color(63, 82, 151) * 0.7f;
 
-		public UIPanel() {
+		public override void OnActivate() {
+			// moved from constructor to avoid texture loading on JIT thread
 			if (_borderTexture == null)
 				_borderTexture = Main.Assets.Request<Texture2D>("Images/UI/PanelBorder");
 
 			if (_backgroundTexture == null)
 				_backgroundTexture = Main.Assets.Request<Texture2D>("Images/UI/PanelBackground");
+		}
 
+		public UIPanel() {
 			SetPadding(_cornerSize);
 		}
 
