--- src/Terraria\Terraria.GameContent.UI.Elements\UIScrollbar.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UIScrollbar.cs
@@ -114,5 +_,38 @@
 			base.MouseUp(evt);
 			this._isDragging = false;
 		}
+
+		public override void ScrollWheel(UIScrollWheelEvent evt) {
+			base.ScrollWheel(evt);
+			if (evt.ScrollWheelValue != 0) {
+				Main.UiScrollHandled = true;
+			}
+		}
+	}
+
+	public class FixedUIScrollbar : UIScrollbar
+	{
+		UserInterface userInterface;
+
+		public FixedUIScrollbar(UserInterface userInterface)
+		{
+			this.userInterface = userInterface;
+		}
+
+		protected override void DrawSelf(SpriteBatch spriteBatch)
+		{
+			UserInterface temp = UserInterface.ActiveInstance;
+			UserInterface.ActiveInstance = userInterface;
+			base.DrawSelf(spriteBatch);
+			UserInterface.ActiveInstance = temp;
+		}
+
+		public override void MouseDown(UIMouseEvent evt)
+		{
+			UserInterface temp = UserInterface.ActiveInstance;
+			UserInterface.ActiveInstance = userInterface;
+			base.MouseDown(evt);
+			UserInterface.ActiveInstance = temp;
+		}
 	}
 }

