--- src/Terraria\Terraria.GameContent.UI.Elements\UITextPanel.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UITextPanel.cs
@@ -1,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.UI;
+using Terraria.UI.Chat;
 
 namespace Terraria.GameContent.UI.Elements
 {
@@ -67,7 +_,8 @@
 		}
 
 		public virtual void SetText(T text, float textScale, bool large) {
-			Vector2 textSize = new Vector2((large ? Main.fontDeathText : Main.fontMouseText).MeasureString(text.ToString()).X, large ? 32f : 16f) * textScale;
+			Vector2 textSize = ChatManager.GetStringSize(dynamicSpriteFont, text.ToString(), new Vector2(textScale));
+			textSize.Y = (large ? 32f : 16f) * textScale;
 			_text = text;
 			_textScale = textScale;
 			_textSize = textSize;
