--- src/Terraria\Terraria.GameContent.UI.Elements\UIWorldListItem.cs
+++ src/tModLoader\Terraria.GameContent.UI.Elements\UIWorldListItem.cs
@@ -5,6 +_,8 @@
 using Terraria.Graphics;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Config;
 using Terraria.Social;
 using Terraria.UI;
 
@@ -25,6 +_,7 @@
 		private Texture2D _buttonPlayTexture;
 		private Texture2D _buttonSeedTexture;
 		private Texture2D _buttonDeleteTexture;
+		private Texture2D _configTexture;
 		private UIImageButton _deleteButton;
 
 		public bool IsFavorite
@@ -87,6 +_,17 @@
 				base.Append(uIImageButton4);
 				num += 24f;
 			}
+			//if (ConfigManager.Configs.Count > 0)
+			//{
+			//	UIImageButton configButton = new UIImageButton(this._configTexture);
+			//	configButton.VAlign = 1f;
+			//	configButton.Left.Set(num, 0f);
+			//	configButton.OnClick += new UIElement.MouseEvent(this.ConfigButtonClick);
+			//	configButton.OnMouseOver += new UIElement.MouseEvent(this.ConfigMouseOver);
+			//	configButton.OnMouseOut += new UIElement.MouseEvent(this.ButtonMouseOut);
+			//	Append(configButton);
+			//	num += 24f;
+			//}
 			UIImageButton uIImageButton5 = new UIImageButton(this._buttonDeleteTexture);
 			uIImageButton5.VAlign = 1f;
 			uIImageButton5.HAlign = 1f;
@@ -126,6 +_,7 @@
 			this._buttonPlayTexture = TextureManager.Load("Images/UI/ButtonPlay");
 			this._buttonSeedTexture = TextureManager.Load("Images/UI/ButtonSeed");
 			this._buttonDeleteTexture = TextureManager.Load("Images/UI/ButtonDelete");
+			this._configTexture = ModLoader.UI.UICommon.ButtonConfigTexture;
 		}
 
 		private void InitializeAppearance()
@@ -169,6 +_,11 @@
 		private void SeedMouseOver(UIMouseEvent evt, UIElement listeningElement)
 		{
 			this._buttonLabel.SetText(Language.GetTextValue("UI.CopySeed", this._data.SeedText));
+		}
+
+		private void ConfigMouseOver(UIMouseEvent evt, UIElement listeningElement)
+		{
+			this._buttonLabel.SetText("Edit World Config");
 		}
 
 		private void DeleteMouseOver(UIMouseEvent evt, UIElement listeningElement)
@@ -242,8 +_,27 @@
 			}
 			if (!Main.menuMultiplayer)
 			{
+				ConfigManager.LoadAll(); // Makes sure MP configs are cleared.
+				if (ConfigManager.AnyModNeedsReload())
+				{
+					ModLoader.ModLoader.OnSuccessfulLoad += PlayReload();
+					ModLoader.ModLoader.Reload();
+					return;
+				}
+				ConfigManager.OnChangedAll();
 				WorldGen.playWorld();
 			}
+		}
+
+		internal static Action PlayReload() {
+			// Main.ActivePlayerFileData gets cleared during reload
+			var path = Main.ActivePlayerFileData.Path;
+			var isCloudSave = Main.ActivePlayerFileData.IsCloudSave;
+			return () => {
+				// re-select the current player
+				Player.GetFileData(path, isCloudSave).SetAsActive();
+				WorldGen.playWorld();
+			};
 		}
 
 		private void FavoriteButtonClick(UIMouseEvent evt, UIElement listeningElement)
@@ -272,6 +_,10 @@
 		{
 			Platform.Current.Clipboard = this._data.SeedText;
 			this._buttonLabel.SetText(Language.GetTextValue("UI.SeedCopied"));
+		}
+
+		private void ConfigButtonClick(UIMouseEvent evt, UIElement listeningElement)
+		{
 		}
 
 		public override int CompareTo(object obj)

