--- src/Terraria/Terraria/UwUPnP/UPnP.cs
+++ src/tModLoader/Terraria/UwUPnP/UPnP.cs
@@ -1,4 +_,3 @@
-#if SERVER
 using System;
 using System.Collections.Generic;
 using System.Linq;
@@ -7,6 +_,7 @@
 using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
+using Terraria;
 
 
 namespace UwUPnP
@@ -29,7 +_,9 @@
 
 		static UPnP()
 		{
+			if (Main.dedServ) {
-			FindGateway();
+				FindGateway();
+			}
 		}
 
 		private static Gateway Gateway
@@ -90,4 +_,3 @@
 			.Where(a => a.AddressFamily == AddressFamily.InterNetwork || a.AddressFamily == AddressFamily.InterNetworkV6);
 	}
 }
-#endif
