--- src/Terraria\Terraria.GameContent.Tile_Entities\TEItemFrame.cs
+++ src/tModLoader\Terraria.GameContent.Tile_Entities\TEItemFrame.cs
@@ -82,7 +_,7 @@
 
 		public override void WriteExtraData(BinaryWriter writer, bool networkSend)
 		{
-			writer.Write((short)this.item.netID);
+			writer.Write((short)(this.item.netID >= ItemID.Count ? 0 : this.item.netID));
 			writer.Write(this.item.prefix);
 			writer.Write((short)this.item.stack);
 		}

