--- src/Terraria\Terraria.Graphics.Effects\EffectManager.cs
+++ src/tModLoader\Terraria.Graphics.Effects\EffectManager.cs
@@ -4,7 +_,7 @@
 
 namespace Terraria.Graphics.Effects
 {
-	internal abstract class EffectManager<T> where T : GameEffect
+	public abstract class EffectManager<T> where T : GameEffect
 	{
 		protected bool _isLoaded;
 		protected Dictionary<string, T> _effects = new Dictionary<string, T>();

