--- src/Terraria\Terraria.Graphics.Effects\FilterManager.cs
+++ src/tModLoader\Terraria.Graphics.Effects\FilterManager.cs
@@ -4,7 +4,7 @@
 
 namespace Terraria.Graphics.Effects
 {
-	internal class FilterManager : EffectManager<Filter>
+	public class FilterManager : EffectManager<Filter>
 	{
 		private const float OPACITY_RATE = 0.05f;
 		private LinkedList<Filter> _activeFilters = new LinkedList<Filter>();
@@ -75,5 +75,16 @@
 		{
 			return this._activeFilters.Count != 0;
 		}
+
+		internal void DeactivateAll()
+		{
+			foreach (string key in this._effects.Keys)
+			{
+				if (this[key].IsActive())
+				{
+					this[key].Deactivate(new object[0]);
+				}
+			}
+		}
 	}
 }

