--- src/Terraria\Terraria.Graphics.Effects\Filter.cs
+++ src/tModLoader\Terraria.Graphics.Effects\Filter.cs
@@ -32,14 +_,14 @@
 			return this._shader;
 		}
 
-		internal override void Activate(Vector2 position, params object[] args)
+		public override void Activate(Vector2 position, params object[] args)
 		{
 			this._shader.UseGlobalOpacity(this.Opacity);
 			this._shader.UseTargetPosition(position);
 			this.Active = true;
 		}
 
-		internal override void Deactivate(params object[] args)
+		public override void Deactivate(params object[] args)
 		{
 			this.Active = false;
 		}

