--- src/Terraria\Terraria.Graphics.Effects\FilterManager.cs
+++ src/tModLoader\Terraria.Graphics.Effects\FilterManager.cs
@@ -227,5 +_,16 @@
 		{
 			return this.HasActiveFilter() || this.OnPostDraw != null;
 		}
+
+		internal void DeactivateAll()
+		{
+			foreach (string key in this._effects.Keys)
+			{
+				if (this[key].IsActive())
+				{
+					this[key].Deactivate(new object[0]);
+				}
+			}
+		}
 	}
 }

