--- src/Terraria\Terraria.Graphics.Effects\Filter.cs
+++ src/tModLoader\Terraria.Graphics.Effects\Filter.cs
@@ -26,13 +_,13 @@
 
 		public ScreenShaderData GetShader() => _shader;
 
-		internal override void Activate(Vector2 position, params object[] args) {
+		public override void Activate(Vector2 position, params object[] args) {
 			_shader.UseGlobalOpacity(Opacity);
 			_shader.UseTargetPosition(position);
 			Active = true;
 		}
 
-		internal override void Deactivate(params object[] args) {
+		public override void Deactivate(params object[] args) {
 			Active = false;
 		}
 
