--- src/Terraria\Terraria.Graphics.Effects\GameEffect.cs
+++ src/tModLoader\Terraria.Graphics.Effects\GameEffect.cs
@@ -41,8 +_,8 @@
 
 		public abstract bool IsVisible();
 
-		internal abstract void Activate(Vector2 position, params object[] args);
+		public abstract void Activate(Vector2 position, params object[] args);
 
-		internal abstract void Deactivate(params object[] args);
+		public abstract void Deactivate(params object[] args);
 	}
 }

