--- src/Terraria\Terraria.Graphics.Effects\SimpleOverlay.cs
+++ src/tModLoader\Terraria.Graphics.Effects\SimpleOverlay.cs
@@ -35,12 +_,12 @@
 			_shader.Update(gameTime);
 		}
 
-		internal override void Activate(Vector2 position, params object[] args) {
+		public override void Activate(Vector2 position, params object[] args) {
 			TargetPosition = position;
 			Mode = OverlayMode.FadeIn;
 		}
 
-		internal override void Deactivate(params object[] args) {
+		public override void Deactivate(params object[] args) {
 			Mode = OverlayMode.FadeOut;
 		}
 
