--- src/Terraria\Terraria.Initializers\LaunchInitializer.cs
+++ src/tModLoader\Terraria.Initializers\LaunchInitializer.cs
@@ -9,11 +9,7 @@
 		public static void LoadParameters(Main game)
 		{
 			LaunchInitializer.LoadSharedParameters(game);
-#if CLIENT
 			LaunchInitializer.LoadClientParameters(game);
-#else
-			LaunchInitializer.LoadServerParameters(game);
-#endif
 		}
 
 		private static void LoadSharedParameters(Main game)

