--- src/tModLoader/Terraria.ModLoader.UI/UIDeveloperModeHelp.cs
+++ src/64bit/Terraria.ModLoader.UI/UIDeveloperModeHelp.cs
@@ -191,6 +_,7 @@
 			Interface.downloadProgress.OnDownloadsComplete += () => {
 				Main.menuMode = Interface.developerModeHelpID;
 			};
+			// Change download name from ModCompile to ModCompile64
 			DownloadFileFromUrl("ModCompile64", url, file, () => {
 				try {
 					Extract(file);
