--- src/decompiled\Properties\AssemblyInfo.cs
+++ src/merged\Properties\AssemblyInfo.cs
@@ -1,14 +_,25 @@
 using System;
+#if WINDOWS
 using System.Diagnostics;
+#endif
 using System.Reflection;
 using System.Runtime.CompilerServices;
+#if WINDOWS || LINUX
 using System.Runtime.InteropServices;
+#endif
+#if WINDOWS
 using System.Runtime.Versioning;
+#endif
 using System.Security;
 using System.Security.Permissions;
 
+#if WINDOWS || LINUX
 [assembly: AssemblyVersion("1.3.1.0")]
+#endif
+#if WINDOWS
 [assembly: Debuggable(DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints)]
+#endif
+#if WINDOWS || LINUX
 [assembly: AssemblyCompany("Re-Logic")]
 [assembly: AssemblyCopyright("Copyright © Re-Logic 2016")]
 [assembly: AssemblyDescription("")]
@@ -16,10 +_,20 @@
 [assembly: AssemblyProduct("Terraria")]
 [assembly: AssemblyTitle("Terraria")]
 [assembly: AssemblyTrademark("")]
+#endif
+#if WINDOWS
 [assembly: CompilationRelaxations(8)]
+#endif
+#if MAC
+[assembly: AssemblyVersion("0.0.0.0")]
+#endif
 [assembly: RuntimeCompatibility(WrapNonExceptionThrows = true)]
+#if WINDOWS || LINUX
 [assembly: ComVisible(false)]
 [assembly: Guid("f571b16a-2c9b-44ab-b115-7c762c9e4e7e")]
+#endif
+#if WINDOWS
 [assembly: TargetFramework(".NETFramework,Version=v4.0,Profile=Client", FrameworkDisplayName = ".NET Framework 4 Client Profile")]
+#endif
 [assembly: SecurityPermission(SecurityAction.RequestMinimum, SkipVerification = true)]
 [module: UnverifiableCode]

