--- src/merged\Terraria.Utilities\PlatformUtilities.cs
+++ src/Terraria\Terraria.Utilities\PlatformUtilities.cs
@@ -73,7 +_,7 @@
 		{
 #if WINDOWS
 			string clipboardText = "";
-			Thread thread = new Thread(delegate
+			Thread thread = new Thread((ThreadStart)delegate
 			{
 				clipboardText = Clipboard.GetText();
 			});
@@ -107,7 +_,7 @@
 		public static void SetClipboard(string text)
 		{
 #if WINDOWS
-			Thread thread = new Thread(delegate
+			Thread thread = new Thread((ThreadStart)delegate
 			{
 				if (text.Length > 0)
 				{

