--- src/merged\Terraria.Utilities\PlatformUtilties.cs
+++ src/Terraria\Terraria.Utilities\PlatformUtilties.cs
@@ -47,7 +47,7 @@
 		{
 #if WINDOWS
 			string clipboardText = "";
-			Thread thread = new Thread(delegate
+			Thread thread = new Thread((ThreadStart)delegate
 			{
 				clipboardText = Clipboard.GetText();
 			});
@@ -81,7 +81,7 @@
 		public static void SetClipboard(string text)
 		{
 #if WINDOWS
-			Thread thread = new Thread(delegate
+			Thread thread = new Thread((ThreadStart)delegate
 			{
 				if (text.Length > 0)
 				{

