--- src/Terraria/Terraria/GameContent/ItemDropRules/CommonDrop.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/CommonDrop.cs
@@ -4,11 +_,11 @@
 {
 	public class CommonDrop : IItemDropRule
 	{
-		protected int _itemId;
+		public int _itemId;
-		protected int _dropsOutOfY;
+		public int _dropsOutOfY;
-		protected int _amtDroppedMinimum;
+		public int _amtDroppedMinimum;
-		protected int _amtDroppedMaximum;
+		public int _amtDroppedMaximum;
-		protected int _dropsXoutOfY;
+		public int _dropsXoutOfY;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
