--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/CommonCode.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/CommonCode.cs
@@ -1,12 +_,28 @@
 using Microsoft.Xna.Framework;
 using Terraria.Utilities;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.ItemDropRules
 {
 	public static class CommonCode
 	{
+		public static void DropItem(DropAttemptInfo info, int item, int stack, bool scattered = false) {
+			if (info.npc == null) {
+				int x = (int)info.player.position.X + info.player.width / 2;
+				int y = (int)info.player.position.Y + info.player.height / 2;
+				if (scattered) {
+					x = (int)info.player.position.X + Main.rand.Next(info.player.width + 1);
+					y = (int)info.player.position.Y + Main.rand.Next(info.player.height + 1);
+				}
+
+				Item.NewItem(info.player.GetItemSource_OpenItem(info.item), x, y, 0, 0, item, stack, noBroadcast: false, -1);
+			}
+			else
+				DropItemFromNPC(info.npc, item, 1, false);
+		}
+
 		public static void DropItemFromNPC(NPC npc, int itemId, int stack, bool scattered = false) {
-			if (itemId > 0 && itemId < 5125) {
+			if (itemId > 0 && itemId < ItemLoader.ItemCount) {
 				int x = (int)npc.position.X + npc.width / 2;
 				int y = (int)npc.position.Y + npc.height / 2;
 				if (scattered) {
@@ -20,7 +_,7 @@
 		}
 
 		public static void DropItemLocalPerClientAndSetNPCMoneyTo0(NPC npc, int itemId, int stack, bool interactionRequired = true) {
-			if (itemId <= 0 || itemId >= 5125)
+			if (itemId <= 0 || itemId >= ItemLoader.ItemCount)
 				return;
 
 			if (Main.netMode == 2) {
@@ -41,7 +_,7 @@
 		}
 
 		public static void DropItemForEachInteractingPlayerOnThePlayer(NPC npc, int itemId, UnifiedRandom rng, int chanceNumerator, int chanceDenominator, int stack = 1, bool interactionRequired = true) {
-			if (itemId <= 0 || itemId >= 5125)
+			if (itemId <= 0 || itemId >= ItemLoader.ItemCount)
 				return;
 
 			if (Main.netMode == 2) {
