--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/CommonDropWithRerolls.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/CommonDropWithRerolls.cs
@@ -16,12 +_,12 @@
 	{
 		bool flag = false;
 		for (int i = 0; i < timesToRoll; i++) {
-			flag = flag || info.player.RollLuck(chanceDenominator) < chanceNumerator;
+			flag = flag || info.RollLuck(chanceDenominator) < chanceNumerator;
 		}
 
 		ItemDropAttemptResult result;
 		if (flag) {
-			CommonCode.DropItemFromNPC(info.npc, itemId, info.rng.Next(amountDroppedMinimum, amountDroppedMaximum + 1));
+			CommonCode.DropItem(info, itemId, info.rng.Next(amountDroppedMinimum, amountDroppedMaximum + 1));
 			result = default(ItemDropAttemptResult);
 			result.State = ItemDropAttemptResultState.Success;
 			return result;
