--- src/Terraria/Terraria/GameContent/ItemDropRules/CommonDropWithRerolls.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/CommonDropWithRerolls.cs
@@ -4,7 +_,7 @@
 {
 	public class CommonDropWithRerolls : CommonDrop
 	{
-		private int _timesToRoll;
+		public int _timesToRoll;
 
 		public CommonDropWithRerolls(int itemId, int dropsOutOfY, int amountDroppedMinimum, int amountDroppedMaximum, int rerolls)
 			: base(itemId, dropsOutOfY, amountDroppedMinimum, amountDroppedMaximum) {
