--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/CommonDropNotScalingWithLuck.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/CommonDropNotScalingWithLuck.cs
@@ -2,14 +_,15 @@
 {
 	public class CommonDropNotScalingWithLuck : CommonDrop
 	{
+		// chanceNumerator parameter added by TML.
-		public CommonDropNotScalingWithLuck(int itemId, int chanceDenominator, int amountDroppedMinimum, int amountDroppedMaximum)
+		public CommonDropNotScalingWithLuck(int itemId, int chanceDenominator, int amountDroppedMinimum, int amountDroppedMaximum, int chanceNumerator = 1)
-			: base(itemId, chanceDenominator, amountDroppedMinimum, amountDroppedMaximum) {
+			: base(itemId, chanceDenominator, amountDroppedMinimum, amountDroppedMaximum, chanceNumerator) {
 		}
 
 		public override ItemDropAttemptResult TryDroppingItem(DropAttemptInfo info) {
 			ItemDropAttemptResult result;
 			if (info.rng.Next(chanceDenominator) < chanceNumerator) {
-				CommonCode.DropItemFromNPC(info.npc, itemId, info.rng.Next(amountDroppedMinimum, amountDroppedMaximum + 1));
+				CommonCode.DropItem(info, itemId, info.rng.Next(amountDroppedMinimum, amountDroppedMaximum + 1));
 				result = default(ItemDropAttemptResult);
 				result.State = ItemDropAttemptResultState.Success;
 				return result;
