--- src/Terraria/Terraria/GameContent/ItemDropRules/Conditions.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/Conditions.cs
@@ -13,8 +_,8 @@
 
 		public class IsUsingSpecificAIValues : IItemDropRuleCondition, IProvideItemConditionDescription
 		{
-			private int _aiSlotToCheck;
+			public int _aiSlotToCheck;
-			private float _valueToMatch;
+			public float _valueToMatch;
 
 			public IsUsingSpecificAIValues(int aislot, float valueToMatch) {
 				_aiSlotToCheck = aislot;
@@ -119,7 +_,7 @@
 
 		public class FromCertainWaveAndAbove : IItemDropRuleCondition, IProvideItemConditionDescription
 		{
-			private int _neededWave;
+			public int _neededWave;
 
 			public FromCertainWaveAndAbove(int neededWave) {
 				_neededWave = neededWave;
@@ -438,7 +_,7 @@
 
 		public class NamedNPC : IItemDropRuleCondition, IProvideItemConditionDescription
 		{
-			private string _neededName;
+			public string _neededName;
 
 			public NamedNPC(string neededName) {
 				_neededName = neededName;
