--- src/Terraria/Terraria/GameContent/ItemDropRules/DropOneByOne.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropOneByOne.cs
@@ -18,8 +_,8 @@
 			public float GetPersonalDropRate() => (float)DropsXOutOfYTimes_TheX / (float)DropsXOutOfYTimes_TheY;
 		}
 
-		private int _itemId;
+		public int _itemId;
-		private Parameters _parameters;
+		public Parameters _parameters;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
