--- src/Terraria/Terraria/GameContent/ItemDropRules/DropBasedOnExpertMode.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropBasedOnExpertMode.cs
@@ -2,7 +_,7 @@
 
 namespace Terraria.GameContent.ItemDropRules
 {
-	public class DropBasedOnExpertMode : IItemDropRule, INestedItemDropRule
+	public partial class DropBasedOnExpertMode : IItemDropRule, INestedItemDropRule
 	{
 		private IItemDropRule _ruleForNormalMode;
 		private IItemDropRule _ruleForExpertMode;
