--- src/Terraria/Terraria/GameContent/ItemDropRules/DropBasedOnMasterMode.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropBasedOnMasterMode.cs
@@ -4,8 +_,8 @@
 {
 	public class DropBasedOnMasterMode : IItemDropRule, INestedItemDropRule
 	{
-		private IItemDropRule _ruleForDefault;
+		public IItemDropRule _ruleForDefault;
-		private IItemDropRule _ruleForMasterMode;
+		public IItemDropRule _ruleForMasterMode;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
