--- src/Terraria/Terraria/GameContent/ItemDropRules/DropLocalPerClientAndResetsNPCMoneyTo0.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropLocalPerClientAndResetsNPCMoneyTo0.cs
@@ -2,7 +_,7 @@
 {
 	public class DropLocalPerClientAndResetsNPCMoneyTo0 : CommonDrop
 	{
-		private IItemDropRuleCondition _condition;
+		public IItemDropRuleCondition _condition;
 
 		public DropLocalPerClientAndResetsNPCMoneyTo0(int itemId, int dropsOutOfY, int amountDroppedMinimum, int amountDroppedMaximum, IItemDropRuleCondition optionalCondition)
 			: base(itemId, dropsOutOfY, amountDroppedMinimum, amountDroppedMaximum) {
