--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/DropAttemptInfo.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropAttemptInfo.cs
@@ -5,9 +_,17 @@
 public struct DropAttemptInfo
 {
 	public NPC npc;
+	public int item; // Added by TML.
+	public Chest chest; // Added by TML.
 	public Player player;
 	public UnifiedRandom rng;
 	public bool IsInSimulation;
 	public bool IsExpertMode;
 	public bool IsMasterMode;
+	public readonly int RollLuck(int chanceDenominator) // Added by TML.
+	{
+		if (WorldGen.generatingWorld)
+			return WorldGen.genRand.Next(chanceDenominator);
+		return player.RollLuck(chanceDenominator);
+	}
 }
