--- src/Terraria/Terraria/GameContent/ItemDropRules/DropBasedOnExpertMode.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/DropBasedOnExpertMode.cs
@@ -4,8 +_,8 @@
 {
 	public class DropBasedOnExpertMode : IItemDropRule, INestedItemDropRule
 	{
-		private IItemDropRule _ruleForNormalMode;
+		public IItemDropRule _ruleForNormalMode;
-		private IItemDropRule _ruleForExpertMode;
+		public IItemDropRule _ruleForExpertMode;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
