--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/ItemDropResolver.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropResolver.cs
@@ -13,7 +_,14 @@
 
 	public void TryDropping(DropAttemptInfo info)
 	{
+		/*
 		List<IItemDropRule> rulesForNPCID = _database.GetRulesForNPCID(info.npc.netID);
+		*/
+
+		var rulesForNPCID = info.npc != null
+			? _database.GetRulesForNPCID(info.npc.netID)
+			: _database.GetRulesForItemID(info.item);
+
 		for (int i = 0; i < rulesForNPCID.Count; i++) {
 			ResolveRule(rulesForNPCID[i], info);
 		}
