--- src/Terraria/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs
@@ -1,6 +_,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.ItemDropRules
 {
@@ -55,7 +_,7 @@
 			return entry;
 		}
 
-		private void RegisterToNPCNetId(int npcNetId, IItemDropRule entry) {
+		public void RegisterToNPCNetId(int npcNetId, IItemDropRule entry) {
 			if (!_entriesByNpcNetId.ContainsKey(npcNetId))
 				_entriesByNpcNetId[npcNetId] = new List<IItemDropRule>();
 
@@ -114,6 +_,11 @@
 			RegisterMimic();
 			RegisterEclipse();
 			RegisterBloodMoonFishing();
+
+			foreach (KeyValuePair<int, NPC> pair in ContentSamples.NpcsByNetId) {
+				NPCLoader.ModifyNPCLoot(pair.Value, new NPCLoot(pair.Key, this));
+			}
+
 			TrimDuplicateRulesForNegativeIDs();
 		}
 
@@ -231,6 +_,16 @@
 			RegisterBoss_PumpkinMoon();
 			RegisterBoss_HallowBoss();
 			RegisterBoss_QueenSlime();
+
+			for (int i = NPCID.Count; i < NPCLoader.NPCCount; i++) {
+				//For modded NPCs, automatically register boss bags as a drop rule
+				NPC npc = ContentSamples.NpcsByNetId[i];
+				int bagType = -1;
+				NPCLoader.BossBag(npc, ref bagType);
+
+				if (bagType > -1)
+					RegisterToNPC(i, ItemDropRule.BossBag(bagType));
+			}
 		}
 
 		private void RegisterBoss_QueenSlime() {
