--- src/Terraria/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs
@@ -1,12 +_,13 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.ItemDropRules
 {
 	public class ItemDropDatabase
 	{
-		private List<IItemDropRule> _globalEntries = new List<IItemDropRule>();
+		internal List<IItemDropRule> _globalEntries = new List<IItemDropRule>();
 		private Dictionary<int, List<IItemDropRule>> _entriesByNpcNetId = new Dictionary<int, List<IItemDropRule>>();
 		private Dictionary<int, List<int>> _npcNetIdsByType = new Dictionary<int, List<int>>();
 		private int _masterModeDropRng = 4;
@@ -55,7 +_,7 @@
 			return entry;
 		}
 
-		private void RegisterToNPCNetId(int npcNetId, IItemDropRule entry) {
+		public void RegisterToNPCNetId(int npcNetId, IItemDropRule entry) {
 			if (!_entriesByNpcNetId.ContainsKey(npcNetId))
 				_entriesByNpcNetId[npcNetId] = new List<IItemDropRule>();
 
@@ -114,6 +_,11 @@
 			RegisterMimic();
 			RegisterEclipse();
 			RegisterBloodMoonFishing();
+
+			foreach (KeyValuePair<int, NPC> pair in ContentSamples.NpcsByNetId) {
+				NPCLoader.ModifyNPCLoot(pair.Value, new NPCLoot(pair.Key, this));
+			}
+
 			TrimDuplicateRulesForNegativeIDs();
 		}
 
@@ -593,6 +_,7 @@
 			RegisterToGlobal(new ItemDropWithConditionRule(520, 5, 1, 1, new Conditions.SoulOfLight()));
 			RegisterToGlobal(new ItemDropWithConditionRule(521, 5, 1, 1, new Conditions.SoulOfNight()));
 			RegisterToGlobal(ItemDropRule.ByCondition(new Conditions.PirateMap(), 1315, 100));
+			NPCLoader.ModifyGlobalLoot(new GlobalLoot(this));
 		}
 
 		private void RegisterFoodDrops() {
