--- src/Terraria/Terraria/GameContent/ItemDropRules/ItemDropWithConditionRule.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropWithConditionRule.cs
@@ -4,7 +_,7 @@
 {
 	public class ItemDropWithConditionRule : CommonDrop
 	{
-		private IItemDropRuleCondition _condition;
+		public IItemDropRuleCondition _condition;
 
 		public ItemDropWithConditionRule(int itemId, int dropsOutOfY, int amountDroppedMinimum, int amountDroppedMaximum, IItemDropRuleCondition condition, int dropsXOutOfY = 1)
 			: base(itemId, dropsOutOfY, amountDroppedMinimum, amountDroppedMaximum, dropsXOutOfY) {
