--- src/TerrariaNetCore/Terraria/GameContent/ItemDropRules/MechBossSpawnersDropRule.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/MechBossSpawnersDropRule.cs
@@ -25,21 +_,21 @@
 	{
 		ItemDropAttemptResult result;
 		if (!NPC.downedMechBoss1 && info.player.RollLuck(2500) == 0) {
-			CommonCode.DropItemFromNPC(info.npc, 556, 1);
+			CommonCode.DropItem(info, 556, 1);
 			result = default(ItemDropAttemptResult);
 			result.State = ItemDropAttemptResultState.Success;
 			return result;
 		}
 
 		if (!NPC.downedMechBoss2 && info.player.RollLuck(2500) == 0) {
-			CommonCode.DropItemFromNPC(info.npc, 544, 1);
+			CommonCode.DropItem(info, 544, 1);
 			result = default(ItemDropAttemptResult);
 			result.State = ItemDropAttemptResultState.Success;
 			return result;
 		}
 
 		if (!NPC.downedMechBoss3 && info.player.RollLuck(2500) == 0) {
-			CommonCode.DropItemFromNPC(info.npc, 557, 1);
+			CommonCode.DropItem(info, 557, 1);
 			result = default(ItemDropAttemptResult);
 			result.State = ItemDropAttemptResultState.Success;
 			return result;
