--- src/Terraria/Terraria/GameContent/ItemDropRules/OneFromOptionsNotScaledWithLuckDropRule.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/OneFromOptionsNotScaledWithLuckDropRule.cs
@@ -4,9 +_,9 @@
 {
 	public class OneFromOptionsNotScaledWithLuckDropRule : IItemDropRule
 	{
-		private int[] _dropIds;
+		public int[] _dropIds;
-		private int _outOfY;
+		public int _outOfY;
-		private int _xoutOfY;
+		public int _xoutOfY;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
