--- src/Terraria/Terraria/GameContent/ItemDropRules/OneFromOptionsNotScaledWithLuckDropRule.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/OneFromOptionsNotScaledWithLuckDropRule.cs
@@ -2,7 +_,7 @@
 
 namespace Terraria.GameContent.ItemDropRules
 {
-	public class OneFromOptionsNotScaledWithLuckDropRule : IItemDropRule
+	public partial class OneFromOptionsNotScaledWithLuckDropRule : IItemDropRule
 	{
 		private int[] _dropIds;
 		private int _outOfY;
