--- src/Terraria/Terraria/GameContent/ItemDropRules/OneFromRulesRule.cs
+++ src/tModLoader/Terraria/GameContent/ItemDropRules/OneFromRulesRule.cs
@@ -4,8 +_,8 @@
 {
 	public class OneFromRulesRule : IItemDropRule, INestedItemDropRule
 	{
-		private IItemDropRule[] _options;
+		public IItemDropRule[] _options;
-		private int _outOfY;
+		public int _outOfY;
 
 		public List<IItemDropRuleChainAttempt> ChainedRules {
 			get;
