--- src/decompiled/Terraria/GameContent/Drawing/TileDrawing.cs
+++ src/Terraria/Terraria/GameContent/Drawing/TileDrawing.cs
@@ -256,6 +_,9 @@
 		}
 
 		public void Update() {
+			if (Main.dedServ)
+				return;
+
 			double num = Math.Abs(Main.WindForVisuals);
 			num = Utils.GetLerpValue(0.08f, 1.2f, (float)num, clamped: true);
 			_treeWindCounter += 0.004166666666666667 + 0.004166666666666667 * num * 2.0;
