--- src/TerrariaNetCore/Terraria/Localization/GameCulture.cs
+++ src/tModLoader/Terraria/Localization/GameCulture.cs
@@ -25,6 +_,8 @@
 	public readonly CultureInfo CultureInfo;
 	public readonly int LegacyId;
 
+	public static IEnumerable<GameCulture> KnownCultures => _legacyCultures.Values;
+
 	public static GameCulture DefaultCulture { get; set; }
 
 	public bool IsActive => Language.ActiveCulture == this;
