--- src/TerrariaNetCore/Terraria/Localization/LanguageManager.cs
+++ src/tModLoader/Terraria/Localization/LanguageManager.cs
@@ -13,14 +_,15 @@
 using ReLogic.Content.Sources;
 using ReLogic.Graphics;
 using ReLogic.Utilities;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 
 namespace Terraria.Localization;
 
-public class LanguageManager
+public partial class LanguageManager
 {
 	public static LanguageManager Instance = new LanguageManager();
-	private readonly Dictionary<string, LocalizedText> _localizedTexts = new Dictionary<string, LocalizedText>();
+	internal readonly Dictionary<string, LocalizedText> _localizedTexts = new Dictionary<string, LocalizedText>();
 	private readonly Dictionary<string, List<string>> _categoryGroupedKeys = new Dictionary<string, List<string>>();
 	private readonly Dictionary<string, Dictionary<string, string>> _textVariations = new Dictionary<string, Dictionary<string, string>>();
 	private GameCulture _fallbackCulture = GameCulture.DefaultCulture;
@@ -85,6 +_,9 @@
 
 	public void SetLanguage(GameCulture culture)
 	{
+		if (ActiveCulture != null) // Log selected language, but skip logging initial DefaultCulture.
+			Logging.tML.Info($"Language: {culture.Name}");
+
 		if (ActiveCulture != culture) {
 			Thread.CurrentThread.CurrentCulture = culture.CultureInfo;
 			Thread.CurrentThread.CurrentUICulture = culture.CultureInfo;
@@ -92,18 +_,29 @@
 		}
 	}
 
+	/*
 	private void ReloadLanguage(GameCulture targetCulture)
+	*/
+	private void ReloadLanguage(GameCulture targetCulture, bool resetValuesToKeysFirst = true)
 	{
 		if (ActiveCulture != _fallbackCulture) {
+			if (resetValuesToKeysFirst)
-			SetAllTextValuesToKeys();
+				SetAllTextValuesToKeys();
+
 			if (targetCulture != _fallbackCulture)
 				LoadLanguage(_fallbackCulture);
 		}
 
 		LoadLanguage(targetCulture);
+		RecalculateBoundTextValues(); // Added by TML
+
 		if (this.OnLanguageChanged != null)
 			this.OnLanguageChanged(this);
+
+		SystemLoader.OnLocalizationsLoaded(); // Added by TML
 	}
+
+	internal void ReloadLanguage(bool resetValuesToKeysFirst = true) => ReloadLanguage(ActiveCulture, resetValuesToKeysFirst);
 
 	private void LoadLanguage(GameCulture culture)
 	{
@@ -136,8 +_,12 @@
 				break;
 			}
 		}
+
+		LocalizationLoader.LoadModTranslations(culture);
 	}
 
+	// Replaced by ProcessCopyCommandsInTexts implementation in LanguageManager.tML.cs
+	/*
 	private void ProcessCopyCommandsInTexts()
 	{
 		Regex regex = new Regex("{\\$(\\w+\\.\\w+)}", RegexOptions.Compiled);
@@ -155,12 +_,17 @@
 					break;
 
 				value.SetValue(text);
+				// patch context
 			}
 		}
 	}
+	*/ // ProcessCopyCommandsInTexts
 
 	public void UseSources(List<IContentSource> sourcesFromLowestToHighest)
 	{
+		if (_contentSources.SequenceEqual(sourcesFromLowestToHighest)) // Added by tML
+			return;
+
 		_contentSources = sourcesFromLowestToHighest;
 		ReloadLanguage(ActiveCulture);
 	}
@@ -170,8 +_,9 @@
 		string name = ActiveCulture.Name;
 		string assetNameStart = ("Localization" + Path.DirectorySeparatorChar + name).ToLower();
 		foreach (IContentSource contentSource in _contentSources) {
-			foreach (string item in contentSource.GetAllAssetsStartingWith(assetNameStart)) {
-				string extension = contentSource.GetExtension(item);
+			foreach (string item in contentSource.GetAllAssetsStartingWith(assetNameStart, ignoreCase: true)) {
+				// TODO: GetExtension doesn't work for ZipContentSource. Might work if we do the same workaround that FileSystemContentSource does for assetExtensions, but that seems hacky/wrong. We can wait for reworks of asset system to fix these inconsistencies. Zip resource packs are rare.
+				string extension = contentSource.GetExtension(item) ?? "";
 				if (!(extension == ".json") && !(extension == ".csv"))
 					continue;
 
@@ -186,6 +_,15 @@
 						LoadLanguageFromFileTextCsv(fileText);
 				}
 				catch (Exception innerException) {
+					Logging.Terraria.Error($"An error occurred loading the \"{item}\" file from one of the enabled resource packs. The language changes contained within will not take effect.");
+					string basePath = null;
+					if(contentSource is FileSystemContentSource fileSystemContentSource) {
+						basePath = (string)typeof(FileSystemContentSource).GetField("_basePath", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(fileSystemContentSource);
+						var resourcePack = Main.AssetSourceController.ActiveResourcePackList.EnabledPacks.FirstOrDefault(x => basePath.Contains(x.FileName));
+						if (resourcePack != null) {
+							Logging.Terraria.Error($"The resource pack that caused this error is \"{resourcePack.Name}\".");
+						}
+					}
 					IAssetRepository assetRepository = Main.instance.Services.Get<IAssetRepository>();
 					if (assetRepository != null && assetRepository.AssetLoadFailHandler != null) {
 						string assetName = item + extension;
@@ -353,6 +_,9 @@
 		if (!_categoryGroupedKeys.TryGetValue(categoryName, out var value))
 			return new LocalizedText(categoryName + ".RANDOM", categoryName + ".RANDOM");
 
+		//TML: Changed to a GetKeysInCategory call instead of direct dictionary access.
+		var list = GetKeysInCategory(categoryName);
+
 		return GetText(categoryName + "." + value[(random ?? Main.rand).Next(value.Count)]);
 	}
 
@@ -361,6 +_,9 @@
 		if (!_categoryGroupedKeys.TryGetValue(categoryName, out var value))
 			return new LocalizedText(categoryName + ".INDEXED", categoryName + ".INDEXED");
 
+		//TML: Changed to a GetKeysInCategory call instead of direct dictionary access.
+		var list = GetKeysInCategory(categoryName);
+
 		int index2 = index % value.Count;
 		return GetText(categoryName + "." + value[index2]);
 	}
@@ -372,7 +_,20 @@
 		if (_localizedTexts.TryGetValue(key, out var value))
 			return value;
 
+		/*
 		return _localizedTexts[key] = new LocalizedText(key, key);
+		*/
+		return new LocalizedText(key, key);
+	}
+
+	public LocalizedText GetOrRegister(string key, Func<string> makeDefaultValue = null)
+	{
+		if (!_localizedTexts.TryGetValue(key, out var text)) {
+			_moddedKeys.Add(key);
+			_localizedTexts[key] = text = new LocalizedText(key, makeDefaultValue?.Invoke() ?? key);
+		}
+
+		return text;
 	}
 
 	public string GetTextValue(string key)
