--- src/TerrariaNetCore/Terraria/Localization/VariableText.cs
+++ src/tModLoader/Terraria/Localization/VariableText.cs
@@ -1,42 +_,56 @@
 using System;
 using System.Collections.Generic;
-using System.Text;
 using System.Text.RegularExpressions;
 
 namespace Terraria.Localization;
 
-internal class VariableText
+internal partial class VariableText
 {
 	private struct Condition
 	{
 		public bool RequiredValue;
 		public string Name;
+
+		public override string ToString() => $"{{{(RequiredValue ? "?" : "?!")}{Name}}}"; // Added by TML. Reproduces the pattern this was parsed from. See Bind
 	}
 
 	private readonly string _original;
+	/*
 	private readonly string _format;
+	*/
 	private readonly Condition[] _conditions;
 	private readonly string[] _variables;
 	private readonly object[] _formatArgBuffer;
+	/*
 	private readonly StringBuilder _formatBuffer = new StringBuilder();
+	*/
+	private readonly CompositeText _compositeText; // TML added, replacement for _format and _formatBuffer
+
-	private static readonly Regex _substitutionRegex = new Regex("{(\\?!?)?([a-zA-Z][\\w\\.]*)}", RegexOptions.Compiled);
+	private static readonly Regex _substitutionRegex = new Regex("{(\\?!?)?([a-zA-Z][\\w\\./]*)}", RegexOptions.Compiled); // '/' added by TML, for mod namespaces (eg {ExampleMod/Something})
 
-	private VariableText(string original, string format, Condition[] conditions, string[] variables)
+	private VariableText(string original, string format, Condition[] conditions, string[] variables, int positionalArgCount)
 	{
 		_original = original;
+		/*
 		_format = format;
+		*/
+		_compositeText = new CompositeText(format);
 		_conditions = conditions;
 		_variables = variables;
+		PositionalArgCount = positionalArgCount;
-		_formatArgBuffer = new object[variables.Length];
+		_formatArgBuffer = new object[positionalArgCount + variables.Length];
 	}
 
 	public static bool TryCreate(string s, out VariableText text)
 	{
+		if (_namedPluralRegex.IsMatch(s)) { } else
 		if (!_substitutionRegex.IsMatch(s)) {
 			text = null;
 			return false;
 		}
 
+		int positionalArgCount = CompositeText.CountArgs(s); // Added by TML
+
 		List<string> variables = new List<string>();
 		List<Condition> conditions = new List<Condition>();
 		string format = _substitutionRegex.Replace(s, delegate (Match match) {
@@ -57,10 +_,12 @@
 				variables.Add(text2);
 			}
 
+			num += positionalArgCount; // Added by TML
 			return "{" + num + "}";
 		});
 
+		format = ConvertNamedPlurals(format, variables, positionalArgCount); // Added by TML, named plural support
-		text = new VariableText(s, format, conditions.ToArray(), variables.ToArray());
+		text = new VariableText(s, format, conditions.ToArray(), variables.ToArray(), positionalArgCount);
 		return true;
 	}
 
@@ -69,8 +_,19 @@
 		Condition[] conditions = _conditions;
 		for (int i = 0; i < conditions.Length; i++) {
 			Condition condition = conditions[i];
+			/*
 			if (((lookup(condition.Name) as bool?) ?? false) != condition.RequiredValue)
 				return false;
+			*/
+			// Support non-bool as conditions. Non-null values act like true.
+			object value = lookup(condition.Name);
+			if (value is bool boolValue) {
+				if (boolValue != condition.RequiredValue)
+					return false;
+			}
+			else if ((value == null) == condition.RequiredValue) {
+				return false;
+			}
 		}
 
 		for (int j = 0; j < _variables.Length; j++) {
@@ -78,7 +_,7 @@
 			if (obj == null)
 				return false;
 
-			_formatArgBuffer[j] = obj;
+			_formatArgBuffer[PositionalArgCount + j] = obj;
 		}
 
 		return true;
@@ -86,16 +_,23 @@
 
 	public bool ConditionsMet(Func<string, object> lookup) => CheckConditionsAndLoadArgs(lookup);
 
-	public bool TryFormat(Func<string, object> lookup, out string formatted)
+	public bool TryFormat(Func<string, object> lookup, out string formatted) => TryFormat(lookup, [], out formatted);
+	public bool TryFormat(Func<string, object> lookup, object[] args, out string formatted) // overload added to supply positional args as well as named substitutions
 	{
 		if (!CheckConditionsAndLoadArgs(lookup)) {
 			formatted = null;
 			return false;
 		}
 
+		/*
 		_formatBuffer.AppendFormat(_format, _formatArgBuffer);
 		formatted = _formatBuffer.ToString();
 		_formatBuffer.Clear();
+		*/
+
+		CompositeText.PadArgs(ref args, PositionalArgCount);
+		Array.Copy(args, _formatArgBuffer, PositionalArgCount);
+		formatted = _compositeText.Format(_formatArgBuffer);
 		return true;
 	}
 
