--- src/TerrariaNetCore/ReLogic/Content/IAssetRepository.cs
+++ src/tModLoader/ReLogic/Content/IAssetRepository.cs
@@ -1,5 +_,6 @@
 using System;
 using System.Collections.Generic;
+using System.IO;
 using ReLogic.Content.Sources;
 
 namespace ReLogic.Content;
@@ -16,7 +_,15 @@
 
 	void SetSources(IEnumerable<IContentSource> sources, AssetRequestMode mode = AssetRequestMode.ImmediateLoad);
 
+	//TML: Default request mode for mods changed to Async.
-	Asset<T> Request<T>(string assetName, AssetRequestMode mode = AssetRequestMode.ImmediateLoad) where T : class;
+	Asset<T> Request<T>(string assetName, AssetRequestMode mode = AssetRequestMode.AsyncLoad) where T : class;
+
+	// Added by TML.
+	Asset<T> CreateUntracked<T>(Stream stream, string name, AssetRequestMode mode = AssetRequestMode.ImmediateLoad) where T : class;
 
 	void TransferCompletedAssets();
+
+	// Exists to change the default parameter of 'mode' (from Immediate to Async) for modders, but not the base game.
+	internal Asset<T> Request<T>(string assetName) where T : class
+		=> Request<T>(assetName, AssetRequestMode.ImmediateLoad);
 }
