--- src/decompiled\Terraria.csproj
+++ src/merged\Terraria.csproj
@@ -24,6 +24,12 @@
     <DebugSymbols>true</DebugSymbols>
     <DebugType>pdbonly</DebugType>
     <Optimize>true</Optimize>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
+    <DefineConstants>CLIENT; WINDOWS</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
+    <DefineConstants>CLIENT; WINDOWS</DefineConstants>
   </PropertyGroup>
   <ItemGroup>
     <Reference Include="Microsoft.Xna.Framework.Game" />
@@ -332,8 +338,10 @@
     <Compile Include="Terraria\keyBoardInput.cs" />
     <Compile Include="Terraria\Lang.cs" />
     <Compile Include="Terraria\Lighting.cs" />
+    <Compile Include="Terraria\LinuxLaunch.cs" />
     <Compile Include="Terraria\Liquid.cs" />
     <Compile Include="Terraria\LiquidBuffer.cs" />
+    <Compile Include="Terraria\MacLaunch.cs" />
     <Compile Include="Terraria\Main.cs" />
     <Compile Include="Terraria\MessageBuffer.cs" />
     <Compile Include="Terraria\Minecart.cs" />

