--- src/decompiled\Terraria.csproj
+++ src/merged\Terraria.csproj
@@ -1,48 +_,79 @@
 <?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <ProjectGuid>{3996D5FA-6E59-4FE4-9F2B-40EEEF9645D5}</ProjectGuid>
-    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
+    <AssemblyName>Terraria</AssemblyName>
+    <TargetFramework>net40</TargetFramework>
+    <Version>1.3.5.1</Version>
+    <Company>Re-Logic</Company>
+    <Copyright>Copyright © Re-Logic 2017</Copyright>
+    <RootNamespace />
+    <Configurations>WindowsDebug;WindowsServerDebug;WindowsRelease;WindowsServerRelease;MacRelease;MacServerRelease;LinuxRelease;LinuxServerRelease</Configurations>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <LangVersion>latest</LangVersion>
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
     <OutputType>WinExe</OutputType>
-    <AssemblyName>Terraria</AssemblyName>
-    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
-    <TargetFrameworkProfile>Client</TargetFrameworkProfile>
-    <WarningLevel>4</WarningLevel>
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Platform)' == 'x86' ">
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <Optimize>true</Optimize>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(Configuration.Contains('Server'))">
+    <OutputType>Exe</OutputType>
+    <DefineConstants>$(DefineConstants);SERVER</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="!$(Configuration.Contains('Server'))">
+    <DefineConstants>$(DefineConstants);CLIENT</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(Configuration.Contains('Debug'))">
+    <Optimize>false</Optimize>
+    <DefineConstants>$(DefineConstants);DEBUG</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(Configuration.Contains('Windows'))">
     <PlatformTarget>x86</PlatformTarget>
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
-    <OutputPath>bin\Debug\</OutputPath>
-    <DebugSymbols>true</DebugSymbols>
-    <DebugType>full</DebugType>
-    <Optimize>false</Optimize>
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
-    <OutputPath>bin\Release\</OutputPath>
-    <DebugSymbols>true</DebugSymbols>
-    <DebugType>pdbonly</DebugType>
-    <Optimize>true</Optimize>
+    <AssemblySearchPaths>$(AssemblySearchPaths);{GAC}</AssemblySearchPaths>
+    <DefineConstants>$(DefineConstants);WINDOWS;XNA</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(Configuration.Contains('Mac'))">
+    <DefineConstants>$(DefineConstants);MAC;FNA</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition="$(Configuration.Contains('Linux'))">
+    <DefineConstants>$(DefineConstants);LINUX;FNA</DefineConstants>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Xna.Framework.Game" />
-    <Reference Include="Microsoft.Xna.Framework.Graphics" />
-    <Reference Include="Microsoft.Xna.Framework" />
+    <Reference Include="Microsoft.Xna.Framework.Game" Condition="$(DefineConstants.Contains('XNA'))" />
+    <Reference Include="Microsoft.Xna.Framework.Graphics" Condition="$(DefineConstants.Contains('XNA'))" />
+    <Reference Include="Microsoft.Xna.Framework" Condition="$(DefineConstants.Contains('XNA'))" />
+    <Reference Include="Microsoft.Xna.Framework.Xact" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="Newtonsoft.Json" />
     <Reference Include="System" />
-    <Reference Include="Steamworks.NET" />
-    <Reference Include="ReLogic" />
+    <Reference Include="Steamworks.NET">
+      <HintPath Condition="$(DefineConstants.Contains('WINDOWS'))">..\..\references\Steamworks.NET.Windows.dll</HintPath>
+      <HintPath Condition="$(DefineConstants.Contains('MAC'))">..\..\references\Steamworks.NET.OSX.dll</HintPath>
+      <HintPath Condition="$(DefineConstants.Contains('LINUX'))">..\..\references\Steamworks.NET.Linux.dll</HintPath>
+    </Reference>
+    <Reference Include="ReLogic">
+      <HintPath Condition="$(DefineConstants.Contains('WINDOWS'))">..\..\references\ReLogicWindows.dll</HintPath>
+      <HintPath Condition="$(DefineConstants.Contains('MAC'))">..\..\references\ReLogicMac.dll</HintPath>
+      <HintPath Condition="$(DefineConstants.Contains('LINUX'))">..\..\references\ReLogicLinux.dll</HintPath>
+    </Reference>
     <Reference Include="System.Core" />
-    <Reference Include="Microsoft.Xna.Framework.Xact" />
     <Reference Include="System.Drawing" />
-    <Reference Include="Ionic.Zip.CF" />
-    <Reference Include="System.Windows.Forms" />
+    <Reference Include="Ionic.Zip.CF">
+      <HintPath>Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll</HintPath>
+    </Reference>
+    <Reference Include="FNA" Condition="$(DefineConstants.Contains('FNA'))">
+      <HintPath>..\..\references\FNA.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Windows.Forms">
+      <HintPath Condition="!$(DefineConstants.Contains('WINDOWS'))">..\..\references\System.Windows.Forms.Mono.dll</HintPath>
+    </Reference>
     <Reference Include="WindowsBase" />
   </ItemGroup>
   <ItemGroup>
     <Compile Include="Extensions\EnumerationExtensions.cs" />
-    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="NATUPNPLib\IStaticPortMapping.cs" Condition="$(DefineConstants.Contains('SERVER'))" />
+    <Compile Include="NATUPNPLib\IStaticPortMappingCollection.cs" Condition="$(DefineConstants.Contains('SERVER'))" />
+    <Compile Include="NATUPNPLib\IUPnPNAT.cs" Condition="$(DefineConstants.Contains('SERVER'))" />
+    <Compile Include="NATUPNPLib\UPnPNAT.cs" Condition="$(DefineConstants.Contains('SERVER'))" />
     <Compile Include="Terraria.Achievements\Achievement.cs" />
     <Compile Include="Terraria.Achievements\AchievementCategory.cs" />
     <Compile Include="Terraria.Achievements\AchievementCondition.cs" />
@@ -388,7 +_,7 @@
     <Compile Include="Terraria.UI\UIState.cs" />
     <Compile Include="Terraria.UI\UserInterface.cs" />
     <Compile Include="Terraria.Utilities\CrashDump.cs" />
-    <Compile Include="Terraria.Utilities\FileOperationAPIWrapper.cs" />
+    <Compile Include="Terraria.Utilities\FileOperationAPIWrapper.cs" Condition="$(DefineConstants.Contains('WINDOWS'))" />
     <Compile Include="Terraria.Utilities\FileUtilities.cs" />
     <Compile Include="Terraria.Utilities\NPCUtils.cs" />
     <Compile Include="Terraria.Utilities\PlatformUtilities.cs" />
@@ -437,8 +_,10 @@
     <Compile Include="Terraria\ItemText.cs" />
     <Compile Include="Terraria\Lang.cs" />
     <Compile Include="Terraria\Lighting.cs" />
+    <Compile Include="Terraria\LinuxLaunch.cs" Condition="$(DefineConstants.Contains('LINUX'))" />
     <Compile Include="Terraria\Liquid.cs" />
     <Compile Include="Terraria\LiquidBuffer.cs" />
+    <Compile Include="Terraria\MacLaunch.cs" Condition="$(DefineConstants.Contains('MAC'))" />
     <Compile Include="Terraria\Main.cs" />
     <Compile Include="Terraria\MessageBuffer.cs" />
     <Compile Include="Terraria\Minecart.cs" />
@@ -468,7 +_,7 @@
     <Compile Include="Terraria\TimeLogger.cs" />
     <Compile Include="Terraria\Utils.cs" />
     <Compile Include="Terraria\WaterfallManager.cs" />
-    <Compile Include="Terraria\WindowsLaunch.cs" />
+    <Compile Include="Terraria\WindowsLaunch.cs" Condition="$(DefineConstants.Contains('WINDOWS'))" />
     <Compile Include="Terraria\Wiring.cs" />
     <Compile Include="Terraria\WorldGen.cs" />
     <Compile Include="Terraria\WorldSections.cs" />
@@ -478,8 +_,24 @@
     <EmbeddedResource Include="Microsoft.Xna.Framework.RuntimeProfile" />
     <EmbeddedResource Include="Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll" />
     <EmbeddedResource Include="Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll" />
-    <EmbeddedResource Include="Terraria.Libraries\ReLogic\ReLogic.dll" />
-    <EmbeddedResource Include="Terraria.Libraries\Steamworks.NET.Windows\Steamworks.NET.dll" />
+    <EmbeddedResource Include="..\..\references\ReLogicWindows.dll" Condition="$(DefineConstants.Contains('WINDOWS'))">
+      <LogicalName>ReLogic.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\ReLogicMac.dll" Condition="$(DefineConstants.Contains('MAC'))">
+      <LogicalName>ReLogic.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\ReLogicLinux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
+      <LogicalName>ReLogic.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\Steamworks.NET.Windows.dll" Condition="$(DefineConstants.Contains('WINDOWS'))">
+      <LogicalName>Steamworks.NET.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\Steamworks.NET.OSX.dll" Condition="$(DefineConstants.Contains('MAC'))">
+      <LogicalName>Steamworks.NET.dll</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="..\..\references\Steamworks.NET.Linux.dll" Condition="$(DefineConstants.Contains('LINUX'))">
+      <LogicalName>Steamworks.NET.dll</LogicalName>
+    </EmbeddedResource>
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Game.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.Items.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.de-DE.json" />
@@ -544,5 +_,4 @@
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Projectiles.json" />
     <EmbeddedResource Include="Terraria.Localization.Content.zh-Hans.Town.json" />
   </ItemGroup>
-  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
 </Project>

