--- src/Terraria/Terraria/GameContent/CoinLossRevengeSystem.cs
+++ src/tModLoader/Terraria/GameContent/CoinLossRevengeSystem.cs
@@ -349,9 +_,9 @@
 
 						revengeMarker.SpawnEnemy();
 						list2.Add(revengeMarker);
-#if SERVER
-						NetMessage.SendData(127, -1, -1, null, revengeMarker.UniqueID);
-#endif
+						if (Main.dedServ) {
+							NetMessage.SendData(127, -1, -1, null, revengeMarker.UniqueID);
+						}
 					}
 				}
 
@@ -365,7 +_,9 @@
 				IEnumerable<RevengeMarker> enumerable2 = _markers.Where((RevengeMarker x) => x.IsInvalid());
 				_markers.RemoveAll((RevengeMarker x) => x.IsInvalid());
 				_markers.RemoveAll((RevengeMarker x) => x.IsExpired(_gameTime));
-#if SERVER
+				if (!Main.dedServ)
+					return;
+
 				foreach (RevengeMarker item in enumerable) {
 					NetMessage.SendData(127, -1, -1, null, item.UniqueID);
 				}
@@ -373,7 +_,6 @@
 				foreach (RevengeMarker item2 in enumerable2) {
 					NetMessage.SendData(127, -1, -1, null, item2.UniqueID);
 				}
-#endif
 			}
 		}
 
