--- src/TerrariaNetCore/Terraria/GameContent/NPCInteractions.cs
+++ src/tModLoader/Terraria/GameContent/NPCInteractions.cs
@@ -4,13 +_,14 @@
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent;
 
-public class NPCInteractions
+public partial class NPCInteractions
 {
-	public static class Actions
+	public static partial class Actions
 	{
 		public class OpenSign : NPCInteraction
 		{
@@ -191,12 +_,25 @@
 
 			public override void Interact()
 			{
-				Main.NPCChatText_DoNurseHeal(Main.GetNurseHealCost());
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				int healCost = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
+				Main.NPCChatText_DoNurseHeal(healCost, health, removeDebuffs, reason, canHeal);
 			}
 
 			public override bool TryAddCoins(ref Color chatColor, out int coinValue)
 			{
-				coinValue = Main.GetNurseHealCost();
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				coinValue = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
 				Main.GetCoinValueText_Nurse(ref chatColor, ref coinValue);
 				return coinValue > 0;
 			}
@@ -204,7 +_,7 @@
 
 		public class CloseChat : NPCInteraction
 		{
-			public override bool Condition() => true;
+			public override bool Condition() => Main.LocalPlayer.TalkNPC.type < NPCID.Count; // TML: Changed condition
 			public override string GetText() => Lang.inter[52].Value;
 
 			public override void Interact()
@@ -217,6 +_,9 @@
 		{
 			public override bool Condition()
 			{
+				if (Main.LocalPlayer.TalkNPC.type >= NPCID.Count) // TML: Changed condition
+					return false;
+
 				if (NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC))
 					return false;
 
@@ -238,7 +_,7 @@
 		{
 			public override bool ShowExcalmation => true;
 
-			public override bool Condition() => NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC);
+			public override bool Condition() => NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC) && Main.LocalPlayer.TalkNPC.type < NPCID.Count; // TML: Changed condition
 			public override string GetText() => Language.GetTextValue("UI.NPCHousing");
 
 			public override void Interact()
@@ -265,7 +_,10 @@
 					text = "Slime" + slimeType + "Chatter";
 				}
 
+				if (talkNPC.ModNPC is ModNPC modNPC)
+					Main.npcChatText = Language.GetTextValue(modNPC.GetLocalizationKey("TownNPCMood") + ".NoHome");
+				else
-				Main.npcChatText = Language.GetTextValue(text + ".NoHome");
+					Main.npcChatText = Language.GetTextValue(text + ".NoHome");
 				Main.npcChatText += "\n\n";
 				if (talkNPC.type == 160) {
 					Main.npcChatText += Language.GetTextValueWith("HousingText.HousingRequirements_Truffle", new {
@@ -406,7 +_,7 @@
 		Register(new Actions.OpenShop(npcType, shopIndex, customTextKey));
 	}
 
-	private static void Register(NPCInteraction interaction)
+	public static void Register(NPCInteraction interaction) // TML: Made public
 	{
 		All.Add(interaction);
 	}
