--- src/TerrariaNetCore/Terraria/GameContent/NPCInteraction.cs
+++ src/tModLoader/Terraria/GameContent/NPCInteraction.cs
@@ -4,12 +_,25 @@
 
 public abstract class NPCInteraction
 {
+	/// <summary>
+	/// Set to true to display a small exclamation point next to the chat button.
+	/// <br/>Defaults to false.
+	/// </summary>
 	public virtual bool ShowExcalmation => false;
 
+	/// <summary>
+	/// The local player instance of who is speaking to the NPC.
+	/// </summary>
 	public Player LocalPlayer => Main.LocalPlayer;
 
+	/// <summary>
+	/// The instance of the NPC the player is speaking to. Sort for <c>Main.npc[LocalPlayer.talkNPC]</c>
+	/// </summary>
 	public NPC TalkNPC => Main.npc[LocalPlayer.talkNPC];
 
+	/// <summary>
+	/// The NPC type of the NPC the player is speaking to. Will be 0 if the player is not speaking to any NPC.
+	/// </summary>
 	public int TalkNPCType {
 		get {
 			if (LocalPlayer.talkNPC == -1)
@@ -19,12 +_,32 @@
 		}
 	}
 
+	/// <summary>
+	/// Determines when this NPCInteraction can/will be shown.
+	/// </summary>
+	/// <returns></returns>
+	/// <remarks>Be sure to set the condition to only show on the Town NPCs you want it to show on, otherwise every Town NPC will receive the button.
+	/// <br/>Example: <c>TalkNPCType == ModContent.NPCType&lt;YourModNPCHere&gt;()</c>
+	/// </remarks>
 	public abstract bool Condition();
 
+	/// <summary>
+	/// The text of the button.
+	/// </summary>
+	/// <returns></returns>
 	public abstract string GetText();
 
+	/// <summary>
+	/// The action that happens when the button is clicked such as opening a shop.
+	/// </summary>
 	public abstract void Interact();
 
+	/// <summary>
+	/// Adds a coin counter display to the right of the button if set to true.
+	/// </summary>
+	/// <param name="chatColor">Allows for the color of the button to be changed.</param>
+	/// <param name="coinValue">The value of coins to display. The value can be negative and it will display properly.</param>
+	/// <returns></returns>
 	public virtual bool TryAddCoins(ref Color chatColor, out int coinValue)
 	{
 		coinValue = 0;
