--- src/TerrariaNetCore/Terraria/GameContent/NPCInteractions.cs
+++ src/tModLoader/Terraria/GameContent/NPCInteractions.cs
@@ -4,13 +_,14 @@
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent;
 
-public class NPCInteractions
+public partial class NPCInteractions
 {
-	public static class Actions
+	public static partial class Actions
 	{
 		public class OpenSign : NPCInteraction
 		{
@@ -35,18 +_,38 @@
 
 		public class OpenShop : NPCInteraction
 		{
-			private int _shopIndex;
-			private int _npcType;
+			private int _shopIndex; // Vanilla shop index
+			private string _shopName; // String shop name
+			private string _shopFullName; // String full shop name, if set.
+			// private int _npcType; TML: Removed
 			private string _customTextKey;
 
+			/// <inheritdoc cref="NPCInteractions.Shop(string, string)"/>
+			public OpenShop(string shopName, string customTextKey = null)
+			{
+				// Split the shopName by the / and assign _shopName to be the last part of the string.
+				// Example: Terraria/Painter/Shop
+				//		_shopName = Shop
+				//		_shopFullName = Terraria/Painter/Shop
+				// Example: Shop
+				//		_shopName = Shop
+				//		_shopFullName = Shop
+				string[] splitName = shopName.Split('/');
+				_shopIndex = -1;
+				_customTextKey = customTextKey;
+				_shopName = splitName[^1];
+				_shopFullName = shopName;
+			}
+
-			public OpenShop(int npcType, int shopIndex, string customTextKey = null)
+			public OpenShop(int shopIndex, string customTextKey = null)
 			{
-				_npcType = npcType;
 				_shopIndex = shopIndex;
 				_customTextKey = customTextKey;
+				_shopName = null;
+				_shopFullName = null;
 			}
 
-			public override bool Condition() => base.TalkNPCType == _npcType;
+			public override bool Condition() => true; // TML: Changed condition
 
 			public override string GetText()
 			{
@@ -58,7 +_,18 @@
 
 			public override void Interact()
 			{
+				// If vanilla shop, open by ID.
+				if (_shopIndex != -1) {
+					Main.instance.OpenShop(NPCShopDatabase.GetShopNameFromVanillaIndex(_shopIndex));
+				}
+				// If the full name was entered, use that.
+				else if (_shopFullName != _shopName) {
-				Main.instance.OpenShop(_shopIndex);
+					Main.instance.OpenShop(_shopFullName);
+				}
+				// Else, get the full name.
+				else if (_shopName != null) {
+					Main.instance.OpenShop(NPCShopDatabase.GetShopName(LocalPlayer.TalkNPC.type, _shopName));
+				}
 			}
 		}
 
@@ -191,12 +_,25 @@
 
 			public override void Interact()
 			{
-				Main.NPCChatText_DoNurseHeal(Main.GetNurseHealCost());
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				int healCost = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
+				Main.NPCChatText_DoNurseHeal(healCost, health, removeDebuffs, reason, canHeal);
 			}
 
 			public override bool TryAddCoins(ref Color chatColor, out int coinValue)
 			{
-				coinValue = Main.GetNurseHealCost();
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				coinValue = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
 				Main.GetCoinValueText_Nurse(ref chatColor, ref coinValue);
 				return coinValue > 0;
 			}
@@ -265,13 +_,22 @@
 					text = "Slime" + slimeType + "Chatter";
 				}
 
+				if (talkNPC.ModNPC is ModNPC modNPC)
+					Main.npcChatText = Language.GetTextValue(modNPC.GetLocalizationKey("TownNPCMood") + ".NoHome");
+				else
-				Main.npcChatText = Language.GetTextValue(text + ".NoHome");
+					Main.npcChatText = Language.GetTextValue(text + ".NoHome");
 				Main.npcChatText += "\n\n";
 				if (talkNPC.type == 160) {
 					Main.npcChatText += Language.GetTextValueWith("HousingText.HousingRequirements_Truffle", new {
 						NPCName = talkNPC.FullName
 					});
 				}
+				else if (talkNPC.ModNPC is ModNPC modNPC2 && Language.Exists($"{modNPC2.GetLocalizationKey("HousingText")}.HousingRequirements")) {
+					// If Mods.ModName.NPCs.NPCName.HousingText.HousingRequirements exists, use that as the text.
+					Main.npcChatText += Language.GetTextValueWith($"{modNPC2.GetLocalizationKey("HousingText")}.HousingRequirements", new {
+						NPCName = talkNPC.FullName
+					});
+				}
 				else {
 					Main.npcChatText += Language.GetTextValue("HousingText.HousingRequirements");
 				}
@@ -352,10 +_,12 @@
 		}
 	}
 
-	public static List<NPCInteraction> All = new List<NPCInteraction>();
+	internal static List<NPCInteraction> All = new List<NPCInteraction>(); // TML: Made internal
+
 
 	public static void Initialize()
 	{
+		// TML: Buttons are no longer initialized here. They are set in Terraria.ModLoader.NPCInteractionDatabase.Populate()
 		Shop(17, 1);
 		Shop(19, 2);
 		Shop(20, 3);
@@ -403,7 +_,7 @@
 
 	private static void Shop(int npcType, int shopIndex, string customTextKey = null)
 	{
-		Register(new Actions.OpenShop(npcType, shopIndex, customTextKey));
+		// Register(new Actions.OpenShop(npcType, shopIndex, customTextKey)); TML: Removed/Obsolete
 	}
 
 	private static void Register(NPCInteraction interaction)
