--- src/TerrariaNetCore/Terraria/GameContent/NPCInteractions.cs
+++ src/tModLoader/Terraria/GameContent/NPCInteractions.cs
@@ -4,13 +_,14 @@
 using Terraria.GameContent.Achievements;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent;
 
-public class NPCInteractions
+public partial class NPCInteractions
 {
-	public static class Actions
+	public static partial class Actions
 	{
 		public class OpenSign : NPCInteraction
 		{
@@ -33,6 +_,7 @@
 			}
 		}
 
+		/* TML: Re-implemented in NPCInteractions.TML.cs
 		public class OpenShop : NPCInteraction
 		{
 			private int _shopIndex;
@@ -61,6 +_,7 @@
 				Main.instance.OpenShop(_shopIndex);
 			}
 		}
+		*/
 
 		public class StardewValleyBit : NPCInteraction
 		{
@@ -191,20 +_,38 @@
 
 			public override void Interact()
 			{
-				Main.NPCChatText_DoNurseHeal(Main.GetNurseHealCost());
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				int healCost = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
+				Main.NPCChatText_DoNurseHeal(healCost, health, removeDebuffs, reason, canHeal);
 			}
 
 			public override bool TryAddCoins(ref Color chatColor, out int coinValue)
 			{
-				coinValue = Main.GetNurseHealCost();
+				// TML locals.
+				int health = LocalPlayer.statLifeMax2 - LocalPlayer.statLife;
+				bool removeDebuffs = true;
+				string reason = "";
+				bool canHeal = true;
+
+				coinValue = Main.GetNurseHealCost(ref health, ref removeDebuffs, ref reason, ref canHeal);
 				Main.GetCoinValueText_Nurse(ref chatColor, ref coinValue);
 				return coinValue > 0;
 			}
 		}
 
-		public class CloseChat : NPCInteraction
+		public class CloseChat(int type) : NPCInteraction // TML: Added int type
 		{
-			public override bool Condition() => true;
+			public override bool Condition() // TML: Changed condition. Added sign and NPC type check.
+			{
+				if (Main.LocalPlayer.sign > -1)
+					return false;
+				return Main.LocalPlayer.TalkNPC.type == type; 
+			}
 			public override string GetText() => Lang.inter[52].Value;
 
 			public override void Interact()
@@ -213,11 +_,11 @@
 			}
 		}
 
-		public class ReportHappiness : NPCInteraction
+		public class ReportHappiness(int type) : NPCInteraction // TML: Added int type
 		{
-			public override bool Condition()
+			public override bool Condition() // TML: Changed condition. Added NPC type check.
 			{
-				if (NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC))
+				if (NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC) && Main.LocalPlayer.TalkNPC.type != type)
 					return false;
 
 				return base.LocalPlayer.currentShoppingSettings.HappinessReport != "";
@@ -234,11 +_,11 @@
 			}
 		}
 
-		public class RequestHome : NPCInteraction
+		public class RequestHome(int type) : NPCInteraction // TML: Added int type
 		{
 			public override bool ShowExcalmation => true;
 
-			public override bool Condition() => NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC);
+			public override bool Condition() => NPC.CanShowHomelessText(Main.LocalPlayer.talkNPC) && Main.LocalPlayer.TalkNPC.type == type; // TML: Changed condition
 			public override string GetText() => Language.GetTextValue("UI.NPCHousing");
 
 			public override void Interact()
@@ -265,7 +_,10 @@
 					text = "Slime" + slimeType + "Chatter";
 				}
 
+				if (talkNPC.ModNPC is ModNPC modNPC)
+					Main.npcChatText = Language.GetTextValue(modNPC.GetLocalizationKey("TownNPCMood") + ".NoHome");
+				else
-				Main.npcChatText = Language.GetTextValue(text + ".NoHome");
+					Main.npcChatText = Language.GetTextValue(text + ".NoHome");
 				Main.npcChatText += "\n\n";
 				if (talkNPC.type == 160) {
 					Main.npcChatText += Language.GetTextValueWith("HousingText.HousingRequirements_Truffle", new {
@@ -383,7 +_,9 @@
 		Shop(227, 25, "GameUI.PainterDecor");
 		Register(new Actions.TaxCollectorCollectTaxes());
 		Register(new Actions.NurseHeal());
+		// Register(new Actions.CloseChat());
-		Register(new Actions.CloseChat());
+		Register(new Actions.CloseSign());
+		RegisterVanillaCloseButtons(); // TML: Moved to different method.
 		Register(new Actions.OpenSign());
 		Register(new Actions.StardewValleyBit());
 		Register(new Actions.DryadPurification());
@@ -397,8 +_,10 @@
 		Register(new Actions.StylistHairWindow());
 		Register(new Actions.DyeTraderRarePlant());
 		Register(new Actions.TavernkeepAdvice());
-		Register(new Actions.ReportHappiness());
-		Register(new Actions.RequestHome());
+		// Register(new Actions.ReportHappiness());
+		RegisterVanillaHappinessButtons(); // TML: Moved to different method.
+		// Register(new Actions.RequestHome());
+		RegisterVanillaHousingButtons(); // TML: Moved to different method.
 	}
 
 	private static void Shop(int npcType, int shopIndex, string customTextKey = null)
@@ -406,7 +_,13 @@
 		Register(new Actions.OpenShop(npcType, shopIndex, customTextKey));
 	}
 
-	private static void Register(NPCInteraction interaction)
+	/// <summary>
+	/// Register a chat button to the NPC.
+	/// <br/> Example: <c>NPCInteractions.Register(new Actions.CloseChat(NPC.type))</c>
+	/// <br/> For registering shops, use the helper <see cref="NPCInteractions.Shop(int, string, string)"/>
+	/// </summary>
+	/// <param name="interaction">The type of chat button interaction.</param>
+	public static void Register(NPCInteraction interaction) // TML: Made public
 	{
 		All.Add(interaction);
 	}
