--- src/TerrariaNetCore/Terraria/GameContent/PlayerSleepingHelper.cs
+++ src/tModLoader/Terraria/GameContent/PlayerSleepingHelper.cs
@@ -1,6 +_,8 @@
 using System;
 using Microsoft.Xna.Framework;
+using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
@@ -180,8 +_,20 @@
 				break;
 		}
 
-		anchorPosition = new Point(num4, num + 1).ToWorldCoordinates(8f, 16f);
 		visualoffset = SetOffsetbyBed(tileSafely.frameY / 36);
+
+		TileRestingInfo info = new TileRestingInfo(null, new Point(num4, num), visualoffset, targetDirection);
+		TileLoader.ModifySleepingTargetInfo(x, y, tileSafely.type, ref info);
+		num4 = info.AnchorTilePosition.X;
+		num = info.AnchorTilePosition.Y;
+		int directionOffset = info.DirectionOffset;
+		targetDirection = info.TargetDirection;
+		visualoffset = info.VisualOffset;
+		Vector2 finalOffset = info.FinalOffset;
+
+		anchorPosition = new Point(num4, num + 1).ToWorldCoordinates(8f, 16f); // Reordered so that anchorPosition is modifiable
+		anchorPosition.X += targetDirection * directionOffset; // Added to match PlayerSittingHelper
+		anchorPosition += finalOffset; // Added to match PlayerSittingHelper
 		return true;
 	}
 
