--- src/Terraria/Terraria/GameContent/PlayerSittingHelper.cs
+++ src/tModLoader/Terraria/GameContent/PlayerSittingHelper.cs
@@ -1,5 +_,7 @@
 using Microsoft.Xna.Framework;
+using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent
 {
@@ -230,6 +_,15 @@
 						break;
 					}
 			}
+
+			TileRestingInfo info = new TileRestingInfo(new Point(num, num2), seatDownOffset, targetDirection, num3, zero);
+			TileLoader.ModifySittingTargetInfo(x, y, tileSafely.type, ref info);
+			num = info.anchorTilePosition.X;
+			num2 = info.anchorTilePosition.Y;
+			num3 = info.directionOffset;
+			targetDirection = info.targetDirection;
+			seatDownOffset = info.visualOffset;
+			zero = info.finalOffset;
 
 			playerSittingPosition = new Point(num, num2 + 1).ToWorldCoordinates(8f, 16f);
 			playerSittingPosition.X += targetDirection * num3;
