--- src/TerrariaNetCore/Terraria/GameContent/PlayerHeadDrawRenderTargetContent.cs
+++ src/tModLoader/Terraria/GameContent/PlayerHeadDrawRenderTargetContent.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.DataStructures;
 
@@ -19,6 +_,7 @@
 	internal override void DrawTheContent(SpriteBatch spriteBatch)
 	{
 		if (_player != null && !_player.ShouldNotDraw) {
+			/*
 			_drawData.Clear();
 			_dust.Clear();
 			_gore.Clear();
@@ -34,6 +_,9 @@
 			PlayerDrawHeadLayers.DrawPlayer_07_JustHair(ref drawinfo);
 			PlayerDrawHeadLayers.DrawPlayer_08_FaceAcc(ref drawinfo);
 			PlayerDrawHeadLayers.DrawPlayer_RenderAllLayers(ref drawinfo);
+			*/
+
+			Main.PlayerRenderer.DrawPlayerHead(Main.Camera, _player, new Vector2(width * 0.5f, height * 0.5f));
 		}
 	}
 }
