--- src/Terraria/Terraria/GameContent/PlayerSittingHelper.cs
+++ src/tModLoader/Terraria/GameContent/PlayerSittingHelper.cs
@@ -1,5 +_,7 @@
 using Microsoft.Xna.Framework;
+using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent
 {
@@ -231,7 +_,18 @@
 					}
 			}
 
-			playerSittingPosition = new Point(num, num2 + 1).ToWorldCoordinates(8f, 16f);
+			// At this point, vanilla tile num/num2 represent the bottom-most tile coordinate of the tile (no vanilla tile for sitting is higher than 2),
+			// Modded ones are arbitrary, modder responsibility to find the bottom-post tile
+			TileRestingInfo info = new TileRestingInfo(player, new Point(num, num2), seatDownOffset, targetDirection, num3, zero);
+			TileLoader.ModifySittingTargetInfo(x, y, tileSafely.type, ref info);
+			num = info.anchorTilePosition.X;
+			num2 = info.anchorTilePosition.Y;
+			num3 = info.directionOffset;
+			targetDirection = info.targetDirection;
+			seatDownOffset = info.visualOffset;
+			zero = info.finalOffset;
+
+			playerSittingPosition = new Point(num, num2).ToWorldCoordinates(8f, 16f); // Moved +1 hardcode into ModifySittingTargetInfo
 			playerSittingPosition.X += targetDirection * num3;
 			playerSittingPosition += zero;
 			return true;
