--- src/Terraria/Terraria/GameContent/Profiles.cs
+++ src/tModLoader/Terraria/GameContent/Profiles.cs
@@ -17,10 +_,10 @@
 			public LegacyNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
-#if CLIENT
-				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
-				_defaultParty = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Party", AssetRequestMode.DoNotLoad);
-#endif
+				if (!Main.dedServ) {
+					_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
+					_defaultParty = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Party", AssetRequestMode.DoNotLoad);
+				}
 			}
 
 			public int RollVariation() => 0;
@@ -50,11 +_,12 @@
 			public TransformableNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+
-#if CLIENT
+				if (!Main.dedServ) {
-				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
+					_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
-				_defaultTransformed = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Transformed", AssetRequestMode.DoNotLoad);
+					_defaultTransformed = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Transformed", AssetRequestMode.DoNotLoad);
-				_defaultCredits = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Credits", AssetRequestMode.DoNotLoad);
+					_defaultCredits = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Credits", AssetRequestMode.DoNotLoad);
-#endif
+				}
 			}
 
 			public int RollVariation() => 0;
@@ -92,9 +_,9 @@
 				string[] variants = _variants;
 				foreach (string str in variants) {
 					string text = _rootFilePath + "_" + str;
-#if CLIENT
-					_variantTextures[text] = Main.Assets.Request<Texture2D>(text, AssetRequestMode.DoNotLoad);
-#endif
+					if (!Main.dedServ) {
+						_variantTextures[text] = Main.Assets.Request<Texture2D>(text, AssetRequestMode.DoNotLoad);
+					}
 				}
 			}
 
