--- src/TerrariaNetCore/Terraria/GameContent/Profiles.cs
+++ src/tModLoader/Terraria/GameContent/Profiles.cs
@@ -5,7 +_,7 @@
 
 namespace Terraria.GameContent;
 
-public class Profiles
+public partial class Profiles
 {
 	public class StackedNPCProfile : ITownNPCProfile
 	{
@@ -67,7 +_,7 @@
 		}
 
 		public int RollVariation() => 0;
-		public string GetNameForVariant(NPC npc) => NPC.getNewNPCName(npc.type);
+		public string GetNameForVariant(NPC npc) => npc.getNewNPCName();
 
 		public Asset<Texture2D> GetTextureNPCShouldUse(NPC npc)
 		{
@@ -104,7 +_,7 @@
 		}
 
 		public int RollVariation() => 0;
-		public string GetNameForVariant(NPC npc) => NPC.getNewNPCName(npc.type);
+		public string GetNameForVariant(NPC npc) => npc.getNewNPCName();
 
 		public Asset<Texture2D> GetTextureNPCShouldUse(NPC npc)
 		{
