--- src/TerrariaNetCore/Terraria/GameContent/SecretSeedsTracker.cs
+++ src/tModLoader/Terraria/GameContent/SecretSeedsTracker.cs
@@ -1,9 +_,10 @@
 using System.Collections.Generic;
 using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
-public static class SecretSeedsTracker
+public static partial class SecretSeedsTracker
 {
 	private static List<string> _seedsForConfig = new List<string>();
 	private static List<WorldGen.SecretSeed> _seedsForInterface = new List<WorldGen.SecretSeed>();
@@ -25,10 +_,11 @@
 				if (SeedHasSecret(item, out var seed))
 					_seedsForInterface.Add(seed);
 			}
+			SeedLoader.AddModdedSecretSeeds(_seedsForInterface);
 
 			_seedsForInterface = _seedsForInterface.Distinct().ToList();
 			_seedsForConfig.Clear();
-			_seedsForConfig.AddRange(_seedsForInterface.Select((WorldGen.SecretSeed x) => x.TextThatWasUsedToUnlock));
+			_seedsForConfig.AddRange(_seedsForInterface.Where(seed => seed.ModSecretSeed == null).Select((WorldGen.SecretSeed x) => x.TextThatWasUsedToUnlock));
 		}
 
 		_seedsForConfig.Sort();
@@ -40,7 +_,8 @@
 	public static void AddSeedToTrack(string seedString)
 	{
 		if (SeedHasSecret(seedString, out var seed) && !_seedsForInterface.Contains(seed)) {
+			if (seed.ModSecretSeed == null)
-			_seedsForConfig.Add(seed.TextThatWasUsedToUnlock);
+				_seedsForConfig.Add(seed.TextThatWasUsedToUnlock);
 			_seedsForInterface.Add(seed);
 			Main.SaveSettings();
 		}
