--- src/TerrariaNetCore/Terraria/GameContent/ShimmerTransforms.cs
+++ src/tModLoader/Terraria/GameContent/ShimmerTransforms.cs
@@ -1,4 +_,5 @@
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
@@ -12,15 +_,34 @@
 
 	public static int GetDecraftingRecipeIndex(int type)
 	{
+		/*
+		vanilla:
 		int num = ItemID.Sets.IsCrafted[type];
 		if (num < 0)
+		return false;
+		*/
+
+		int[] recipeIndecies = ItemID.Sets.IsCrafted[type];
+		if (recipeIndecies[0] < 0)
 			return -1;
 
+		/*  automatically replaced with corresponding Decrafting Conditions
 		if (WorldGen.crimson && ItemID.Sets.IsCraftedCrimson[type] >= 0)
 			return ItemID.Sets.IsCraftedCrimson[type];
 
 		if (!WorldGen.crimson && ItemID.Sets.IsCraftedCorruption[type] >= 0)
 			return ItemID.Sets.IsCraftedCorruption[type];
+		*/
+
+		//Deliberately preserving the "last first" order of the previous implementation, adding manual priority support
+		int num = recipeIndecies[recipeIndecies.Length -1];
+
+		foreach (int posRec in recipeIndecies) {
+			if (RecipeLoader.DecraftAvailable(Main.recipe[posRec])) {
+				if (Main.recipe[posRec].shimmerPriority >= Main.recipe[num].shimmerPriority)
+					num = posRec;
+			}
+		}
 
 		return num;
 	}
@@ -35,6 +_,9 @@
 			return true;
 
 		if (!NPC.downedGolemBoss && RecipeSets.PostGolem[decraftingRecipeIndex])
+			return true;
+
+		if (!RecipeLoader.DecraftAvailable(Main.recipe[decraftingRecipeIndex]))
 			return true;
 
 		return false;
