--- src/Terraria/Terraria/GameContent/ShopHelper.cs
+++ src/tModLoader/Terraria/GameContent/ShopHelper.cs
@@ -3,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Personalities;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.Localization;
 
 namespace Terraria.GameContent
@@ -118,7 +_,7 @@
 					_currentPriceAdjustment *= 0.95f;
 				}
 
-				bool[] array = new bool[668];
+				bool[] array = new bool[NPCLoader.NPCCount];
 				foreach (NPC item in nearbyResidentNPCs) {
 					array[item.type] = true;
 				}
@@ -182,11 +_,24 @@
 					break;
 			}
 
-			return Language.GetTextValue("TownNPCMoodBiomes." + text);
+			string result = Language.GetTextValue("TownNPCMoodBiomes." + text);
+			if (biomeID >= PrimaryBiomeID.Count) {
+				string key = LoaderManager.Get<BiomeLoader>().Get(biomeID).DisplayName.Key;
+				result = Language.GetTextValue(key);
+			}
+
+			return result;
 		}
 
 		private void AddHappinessReportText(string textKeyInCategory, object substitutes = null) {
+			string str;
+			if (_currentNPCBeingTalkedTo.type < NPCID.Count)
-			string str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
+				str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
+			else {
+				var modNpc = _currentNPCBeingTalkedTo.ModNPC;
+				str = $"Mods.{modNpc.Mod.Name}.TownNPCMood.{modNpc.Name}";
+			}
+
 			if (_currentNPCBeingTalkedTo.type == 633 && _currentNPCBeingTalkedTo.altTexture == 2)
 				str += "Transformed";
 
