--- src/TerrariaNetCore/Terraria/GameContent/ShopHelper.cs
+++ src/tModLoader/Terraria/GameContent/ShopHelper.cs
@@ -4,6 +_,7 @@
 using Terraria.GameContent.Personalities;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
@@ -16,19 +_,26 @@
 	private float _currentPriceAdjustment;
 	private NPC _currentNPCBeingTalkedTo;
 	private Player _currentPlayerTalking;
-	private PersonalityDatabase _database;
-	private AShoppingBiome[] _dangerousBiomes = new AShoppingBiome[3] {
+	internal PersonalityDatabase _database; //TML: Turned from private to internal.
+
+	private IShoppingBiome[] _dangerousBiomes = new IShoppingBiome[3] { //TML: Changed from AShoppingBiome to IShoppingBiome
 		new CorruptionBiome(),
 		new CrimsonBiome(),
 		new DungeonBiome()
 	};
-	private const float likeValue = 0.94f;
+	internal const float likeValue = 0.94f;
-	private const float dislikeValue = 1.06f;
+	internal const float dislikeValue = 1.06f;
-	private const float loveValue = 0.88f;
+	internal const float loveValue = 0.88f;
-	private const float hateValue = 1.12f;
+	internal const float hateValue = 1.12f;
 
 	public ShopHelper()
 	{
+		// Nothing! The below method will be called in NPCLoader.
+	}
+
+	// Split from the constructor.
+	internal void ReinitializePersonalityDatabase()
+	{
 		_database = new PersonalityDatabase();
 		new PersonalityDatabasePopulator().Populate(_database);
 	}
@@ -125,7 +_,7 @@
 			_currentPriceAdjustment *= 0.95f;
 		}
 
-		bool[] array = new bool[NPCID.Count];
+		bool[] array = new bool[NPCLoader.NPCCount];
 		foreach (NPC item in nearbyResidentNPCs) {
 			array[item.type] = true;
 		}
@@ -136,10 +_,17 @@
 		helperInfo.NearbyNPCs = nearbyResidentNPCs;
 		helperInfo.nearbyNPCsByType = array;
 		HelperInfo info = helperInfo;
-		foreach (IShopPersonalityTrait shopModifier in _database.GetByNPCID(npc.type).ShopModifiers) {
+
+		if (!_database.TryGetProfileByNPCID(npc.type, out var personalityProfile)) {
+			goto SkipModifierEnumeration;
+		}
+
+		foreach (IShopPersonalityTrait shopModifier in personalityProfile.ShopModifiers) {
 			shopModifier.ModifyShopPrice(info, this);
 		}
 
+		SkipModifierEnumeration:
+
 		new AllPersonalitiesModifier().ModifyShopPrice(info, this);
 		if (_currentHappiness == "")
 			AddHappinessReportText("Content");
@@ -154,11 +_,17 @@
 		return priceAdjustment;
 	}
 
-	private static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue("TownNPCMoodBiomes." + biomeNameKey);
+	//TML: Turned public.
+	//TODO: Anything else?
+	public static string BiomeNameByKey(string biomeNameKey) => Language.GetTextValue(biomeNameKey.StartsWith("Mods.") ? biomeNameKey : "TownNPCMoodBiomes." + biomeNameKey);
 
 	private void AddHappinessReportText(string textKeyInCategory, object substitutes = null)
 	{
 		string text = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
+
+		if (_currentNPCBeingTalkedTo.ModNPC is ModNPC modNPC)
+			text = modNPC.GetLocalizationKey("TownNPCMood");
+
 		if (_currentNPCBeingTalkedTo.type == 633 && _currentNPCBeingTalkedTo.altTexture == 2)
 			text += "Transformed";
 
@@ -166,6 +_,129 @@
 		_currentHappiness = _currentHappiness + textValueWith + " ";
 	}
 
+	// Added by TML
+	private void AddHappinessReportTextWithKey(string textKey, string textKeyInCategory, object substitutes = null, int otherNPCType = 0)
+	{
+		if (_currentNPCBeingTalkedTo.type == 633 && _currentNPCBeingTalkedTo.altTexture == 2) {
+			// If the Zoologist is talking about a Mod NPC, use ".Transformed" so the key becomes Mods.ModName.NPCs.NPCName.TownNPCMood.Transformed.BestiaryGirl_AffectionLevelNPC
+			// The reason this check is only in this method and not the other is because the other method is not used for trying to access the otherNPCType's localization keys.
+			if (otherNPCType > NPCID.Count)
+				textKey += ".Transformed";
+			// Else vanilla, the key is TownNPCMood_BestiaryGirlTransformed.AffectionLevelNPC_NPCName
+			else
+				textKey += "Transformed";
+		}
+
+		string textValueWith = Language.GetTextValueWith(textKey + "." + textKeyInCategory, substitutes);
+		_currentHappiness = _currentHappiness + textValueWith + " ";
+	}
+
+	// Added by TML.
+	internal void ApplyNpcRelationshipEffect(int otherNPCType, AffectionLevel affectionLevel)
+	{
+		if (affectionLevel == 0 || !Enum.IsDefined(affectionLevel)) {
+			return;
+		}
+
+		/*
+			If talking to ModNPC
+				Want to use ModNPC's LoveNPC_OtherPerson if exists
+			If talking to Vanilla NPC or ModNPC specific dialogue didn't exist.
+				If talking about a ModNPC
+					Want to use ModNPC's OtherPerson_LoveNPC if exists
+				If talking about a vanilla NPC
+					Want to use TownNPCMood_TalkPerson.LoveNPC_OtherPerson
+				Else, use generic LoveNPC
+		*/
+
+		string otherNPCName = NPCID.Search.GetName(otherNPCType); // Get the name of the other Town NPC.
+		ModNPC talkModNPC = _currentNPCBeingTalkedTo.ModNPC; // ModNPC instance of the Town NPC the player is talking to, if it exists.
+	
+		// If the current Town NPC is a ModNPC
+		// Check to see if there is a localization entry for the other Town NPC defined in our current Town NPC.
+		// Example: Mods.ExampleMod.NPCs.ExamplePerson.TownNPCMood.LoveNPC_Guide
+		// Example: Mods.ExampleMod.NPCs.ExamplePerson.TownNPCMood.LoveNPC_MyMod/MyTownNPC
+		if (talkModNPC != null && Language.Exists($"{talkModNPC.GetLocalizationKey("TownNPCMood")}.{affectionLevel}NPC_{otherNPCName}")) {
+			// If it does exist, add the happiness report.
+			AddHappinessReportTextWithKey(talkModNPC.GetLocalizationKey("TownNPCMood"), $"{affectionLevel}NPC_{otherNPCName}", new {
+				NPCName = NPC.GetFullnameByID(otherNPCType)
+			});
+		}
+		// Vanilla NPC or ModNPC specific dialogue didn't exist.
+		else {
+			ModNPC otherModNPC = ModContent.GetModNPC(otherNPCType); // ModNPC instance of the other Town NPC, if it exists.
+			string talkNPCName = NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID); // Get the name of the Town NPC the player is talking to.
+
+			// Talking about a ModNPC. We want to use the OtherModNPC's entry for the Town NPC the player is talking to, if it exists.
+			// Example: Mods.ExampleMod.NPCs.ExamplePerson.TownNPCMood.Guide_LoveNPC
+			// Example: Mods.ExampleMod.NPCs.ExamplePerson.TownNPCMood.MyMod/MyTownNPC_LoveNPC
+			if (otherModNPC != null && Language.Exists($"{otherModNPC.GetLocalizationKey("TownNPCMood")}.{talkNPCName}_{affectionLevel}sNPC")) {
+				AddHappinessReportTextWithKey(otherModNPC.GetLocalizationKey("TownNPCMood"), $"{talkNPCName}_{affectionLevel}sNPC", new {
+					NPCName = NPC.GetFullnameByID(otherNPCType)
+				},
+				otherNPCType);
+			}
+			// Try for vanilla NPC talking about another vanilla NPC.
+			// Example: TownNPCMood_Guide.LoveNPC_Merchant
+			else if (Language.Exists($"TownNPCMood_{talkNPCName}.{affectionLevel}NPC_{otherNPCName}")) {
+				AddHappinessReportTextWithKey($"TownNPCMood_{talkNPCName}", $"{affectionLevel}NPC_{otherNPCName}", new {
+					NPCName = NPC.GetFullnameByID(otherNPCType)
+				});
+			}
+			// If nothing above was true, add the generic dialogue.
+			else {
+				AddHappinessReportText($"{affectionLevel}NPC", new {
+					NPCName = NPC.GetFullnameByID(otherNPCType)
+				});
+			}
+		}
+
+		_currentPriceAdjustment *= NPCHappiness.AffectionLevelToPriceMultiplier[affectionLevel];
+	}
+
+	// Added by TML.
+	internal void ApplyBiomeRelationshipEffect(string biomeNameKey, AffectionLevel affectionLevel)
+	{
+		if (affectionLevel == 0 || !Enum.IsDefined(affectionLevel)) {
+			return;
+		}
+
+		ModNPC talkModNPC = _currentNPCBeingTalkedTo.ModNPC; // ModNPC instance of the Town NPC the player is talking to, if it exists.
+		string talkNPCName = NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID); // Get the name of the Town NPC the player is talking to.
+
+		// Vanilla biomes are just a name. Example: "Forest"
+		// Modded biomes are localization path. Example: "Mods.ExampleMod.Biomes.ExampleSurfaceBiome.TownNPCDialogueName"
+		string biomeName = biomeNameKey;
+		if (biomeNameKey.StartsWith("Mods.")) {
+			string[] splitBiomeName = biomeName.Split('.');
+			biomeName = $"{splitBiomeName[1]}/{splitBiomeName[3]}"; // Construct the ModName/ModBiomeName. Example: "ExampleMod/ExampleSurfaceBiome"
+		}
+
+		// If the current Town NPC is a ModNPC
+		// Check to see if there is a localization entry for the biome defined in our current Town NPC.
+		if (talkModNPC != null && Language.Exists($"{talkModNPC.GetLocalizationKey("TownNPCMood")}.{affectionLevel}Biome_{biomeName}")) {
+			// If it does exist, add the happiness report.
+			AddHappinessReportTextWithKey(talkModNPC.GetLocalizationKey("TownNPCMood"), $"{affectionLevel}Biome_{biomeName}", new {
+				BiomeName = BiomeNameByKey(biomeNameKey)
+			});
+		}
+		// Try for vanilla NPC talking about the biome.
+		else if (Language.Exists($"TownNPCMood_{talkNPCName}.{affectionLevel}Biome_{biomeName}")) {
+			AddHappinessReportTextWithKey($"TownNPCMood_{talkNPCName}", $"{affectionLevel}Biome_{biomeName}", new {
+				BiomeName = BiomeNameByKey(biomeNameKey)
+			});
+		}
+		// Else, generic biome dialogue.
+		else {
+			AddHappinessReportText($"{affectionLevel}Biome", new {
+				BiomeName = BiomeNameByKey(biomeNameKey)
+			});
+		}
+
+		_currentPriceAdjustment *= NPCHappiness.AffectionLevelToPriceMultiplier[affectionLevel];
+	}
+
+	/*
 	public void LikeBiome(string nameKey)
 	{
 		AddHappinessReportText("LikeBiome", new {
@@ -255,6 +_,7 @@
 
 		_currentPriceAdjustment *= 1.12f;
 	}
+	*/
 
 	private List<NPC> GetNearbyResidentNPCs(NPC npc, out int npcsWithinHouse, out int npcsWithinVillage)
 	{
@@ -312,7 +_,8 @@
 	private bool IsPlayerInEvilBiomes(Player player)
 	{
 		for (int i = 0; i < _dangerousBiomes.Length; i++) {
-			AShoppingBiome aShoppingBiome = _dangerousBiomes[i];
+			IShoppingBiome aShoppingBiome = _dangerousBiomes[i];
+
 			if (aShoppingBiome.IsInBiome(player)) {
 				AddHappinessReportText("HateBiome", new {
 					BiomeName = BiomeNameByKey(aShoppingBiome.NameKey)
@@ -328,7 +_,10 @@
 	private bool IsNotReallyTownNPC(NPC npc)
 	{
 		int type = npc.type;
+		/*
 		if (type == 37 || type == 368 || type == 453)
+		*/
+		if (NPCID.Sets.NoTownNPCHappiness[type])
 			return true;
 
 		return false;
