--- src/Terraria/Terraria/GameContent/ShopHelper.cs
+++ src/tModLoader/Terraria/GameContent/ShopHelper.cs
@@ -3,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Personalities;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.Localization;
 
 namespace Terraria.GameContent
@@ -118,7 +_,7 @@
 					_currentPriceAdjustment *= 0.95f;
 				}
 
-				bool[] array = new bool[668];
+				bool[] array = new bool[NPCLoader.NPCCount];
 				foreach (NPC item in nearbyResidentNPCs) {
 					array[item.type] = true;
 				}
@@ -182,7 +_,13 @@
 					break;
 			}
 
-			return Language.GetTextValue("TownNPCMoodBiomes." + text);
+			string result = Language.GetTextValue("TownNPCMoodBiomes." + text);
+			if (biomeID >= BiomeLoader.VanillaPrimaryBiomeCount) {
+				string key = LoaderManager.Get<BiomeLoader>().Get(biomeID).DisplayName.Key;
+				result = Language.GetTextValue(key);
+			}
+
+			return result;
 		}
 
 		private void AddHappinessReportText(string textKeyInCategory, object substitutes = null) {
