--- src/Terraria/Terraria/GameContent/ShopHelper.cs
+++ src/tModLoader/Terraria/GameContent/ShopHelper.cs
@@ -3,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.Personalities;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.Localization;
 
 namespace Terraria.GameContent
@@ -118,7 +_,7 @@
 					_currentPriceAdjustment *= 0.95f;
 				}
 
-				bool[] array = new bool[667];
+				bool[] array = new bool[NPCLoader.NPCCount];
 				foreach (NPC item in nearbyResidentNPCs) {
 					array[item.type] = true;
 				}
@@ -186,7 +_,7 @@
 		}
 
 		private void AddHappinessReportText(string textKeyInCategory, object substitutes = null) {
-			string str = "TownNPCMood_" + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID);
+			string str = (_currentNPCBeingTalkedTo.type > 664 ? "Mods." + _currentNPCBeingTalkedTo.ModNPC.Mod.Name + "." : "") + "TownNPCMood_"  + NPCID.Search.GetName(_currentNPCBeingTalkedTo.netID); //Temporary until localization rework
 			if (_currentNPCBeingTalkedTo.type == 633 && _currentNPCBeingTalkedTo.altTexture == 2)
 				str += "Transformed";
 
