--- src/TerrariaNetCore/Terraria/GameContent/SmartCursorHelper.cs
+++ src/tModLoader/Terraria/GameContent/SmartCursorHelper.cs
@@ -6,6 +_,7 @@
 using Terraria.GameContent.UI;
 using Terraria.GameInput;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
@@ -158,7 +_,8 @@
 		bool result = false;
 		Tile tile = Main.tile[providedInfo.screenTargetX, providedInfo.screenTargetY];
 		if (tile.active()) {
 			switch (tile.type) {
+				/*
 				case 10:
 				case 11:
 				case 13:
@@ -211,6 +_,9 @@
 				case 699:
 				case 707:
 					result = true;
+				*/
+				default:
+					result = TileID.Sets.DisableSmartCursor[tile.type];
 					break;
 				case 314:
 					if (providedInfo.player.gravDir == 1f)
@@ -278,7 +_,7 @@
 			return;
 
 		int type = providedInfo.item.type;
-		if (type < 0 || type >= ItemID.Count || !ItemID.Sets.GrassSeeds[type])
+		if (type < 0 || !ItemID.Sets.GrassSeeds[type])
 			return;
 
 		_targets.Clear();
@@ -739,7 +_,7 @@
 	private static void Step_BlocksLines(SmartCursorUsageInfo providedInfo, ref int focusedX, ref int focusedY)
 	{
 		int type = providedInfo.item.type;
-		if (type < 0 || type >= ItemID.Count || !Player.SmartCursorSettings.SmartBlocksEnabled || providedInfo.item.createTile <= -1 || !AllowNormalBlockPlacementBehaviourForItemType(type) || !Main.tileSolid[providedInfo.item.createTile] || Main.tileSolidTop[providedInfo.item.createTile] || Main.tileFrameImportant[providedInfo.item.createTile] || focusedX != -1 || focusedY != -1)
+		if (type < 0 || !Player.SmartCursorSettings.SmartBlocksEnabled || providedInfo.item.createTile <= -1 || !AllowNormalBlockPlacementBehaviourForItemType(type) || !Main.tileSolid[providedInfo.item.createTile] || Main.tileSolidTop[providedInfo.item.createTile] || Main.tileFrameImportant[providedInfo.item.createTile] || focusedX != -1 || focusedY != -1)
 			return;
 
 		_targets.Clear();
@@ -799,7 +_,7 @@
 
 	private static void Step_Boulders(SmartCursorUsageInfo providedInfo, ref int focusedX, ref int focusedY)
 	{
-		if (providedInfo.item.createTile <= -1 || providedInfo.item.createTile >= TileID.Count || !TileID.Sets.Boulders[providedInfo.item.createTile] || focusedX != -1 || focusedY != -1)
+		if (providedInfo.item.createTile <= -1 || !TileID.Sets.Boulders[providedInfo.item.createTile] || focusedX != -1 || focusedY != -1)
 			return;
 
 		_targets.Clear();
@@ -942,7 +_,7 @@
 				for (int k = i; k <= i + 1; k++) {
 					for (int l = j - 1; l <= j; l++) {
 						Tile tile3 = Main.tile[k, l];
-						if (tile3.active() && (tile3.type < 0 || tile3.type >= TileID.Count || Main.tileSolid[tile3.type] || !WorldGen.CanCutTile(k, l, TileCuttingContext.TilePlacement)))
+						if (tile3.active() && (tile3.type < 0 || Main.tileSolid[tile3.type] || !WorldGen.CanCutTile(k, l, TileCuttingContext.TilePlacement)))
 							flag = false;
 					}
 				}
@@ -1758,13 +_,15 @@
 				if ((tile.active() && !Main.tileCut[tile.type] && !TileID.Sets.BreakableWhenPlacing[tile.type]) || (tile2.active() && !Main.tileCut[tile2.type] && !TileID.Sets.BreakableWhenPlacing[tile2.type]) || !tile3.active() || !WorldGen.SolidTile2(tile3))
 					continue;
 
 				bool flag = (tile4.active() && TileID.Sets.CommonSapling[tile4.type]) || (tile5.active() && TileID.Sets.CommonSapling[tile5.type]);
 				bool flag2 = flag || (tile6.active() && TileID.Sets.CommonSapling[tile6.type]) || (tile7.active() && TileID.Sets.CommonSapling[tile7.type]) || (tile8.active() && TileID.Sets.CommonSapling[tile8.type]) || (tile9.active() && TileID.Sets.CommonSapling[tile9.type]);
+				bool tileTypeValid = false;
+
 				switch (tile3.type) {
 					case 60:
 						if (!flag2 && WorldGen.EmptyTileCheck(i - 2, i + 2, j - num2 + 1, j, 20))
 							_targets.Add(new Point(i, j));
-						break;
+						continue;
 					case 2:
 					case 23:
 					case 109:
@@ -1782,19 +_,30 @@
 					case 112:
 					case 116:
 					case 234:
-						if (!flag && tile.liquid == 0 && WorldGen.EmptyTileCheck(i, i, j - num3, j, 20))
-							_targets.Add(new Point(i, j));
+						tileTypeValid = true;
 						break;
 				}
+
+				if (!tileTypeValid)
+					tileTypeValid = TileLoader.CanGrowModTree(tile3.type) || TileLoader.CanGrowModPalmTree(tile3.type);
+
+				if (tileTypeValid && tile4.liquid == 0 && tile.liquid == 0 && tile5.liquid == 0 && WorldGen.EmptyTileCheck(i - 2, i + 2, j - 20, j, 20))
+					_targets.Add(new Tuple<int, int>(i, j));
 			}
 		}
 
 		_toRemove.Clear();
 		for (int k = 0; k < _targets.Count; k++) {
 			bool flag3 = false;
+
 			for (int l = -1; l < 2; l += 2) {
 				Tile tile10 = Main.tile[_targets[k].X + l, _targets[k].Y + 1];
 				if (tile10.active()) {
+					if (TileLoader.CanGrowModTree(tile10.type) || TileLoader.CanGrowModPalmTree(tile10.type)) {
+						flag3 = true;
+						continue;
+					}
+
 					switch (tile10.type) {
 						case 2:
 						case 23:
@@ -2291,7 +_,7 @@
 			return;
 
 		int type = providedInfo.item.type;
-		if (type < 0 || type >= ItemID.Count)
+		if (type < 0)
 			return;
 
 		int reachableStartX = providedInfo.reachableStartX;
@@ -2377,12 +_,15 @@
 		int reachableEndY = providedInfo.reachableEndY;
 		_ = providedInfo.screenTargetX;
 		_ = providedInfo.screenTargetY;
 		int type = providedInfo.item.type;
-		if (type < 0 || type >= ItemID.Count || !ItemID.Sets.Torches[type] || fX != -1 || fY != -1)
+		if (type < 0 || !ItemID.Sets.Torches[type] || fX != -1 || fY != -1)
 			return;
 
 		_targets.Clear();
+		/*
 		bool flag = !ItemID.Sets.WaterTorches[type];
+		*/
+		bool flag = !ItemID.Sets.WaterTorches[providedInfo.player.BiomeTorchHoldStyle(type)];
 		for (int i = reachableStartX; i <= reachableEndX; i++) {
 			for (int j = reachableStartY; j <= reachableEndY; j++) {
 				Tile tile = Main.tile[i, j];
