--- src/TerrariaNetCore/Terraria/GameContent/TownRoomManager.cs
+++ src/tModLoader/Terraria/GameContent/TownRoomManager.cs
@@ -3,14 +_,15 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent;
 
 public class TownRoomManager
 {
 	public static object EntityCreationLock = new object();
-	private List<Tuple<int, Point>> _roomLocationPairs = new List<Tuple<int, Point>>();
+	internal List<Tuple<int, Point>> _roomLocationPairs = new List<Tuple<int, Point>>();
-	private bool[] _hasRoom = new bool[NPCID.Count];
+	internal bool[] _hasRoom = new bool[NPCLoader.NPCCount];
 
 	public void AddOccupantsToList(int x, int y, List<int> occupantsList)
 	{
@@ -82,8 +_,22 @@
 	public void Save(BinaryWriter writer)
 	{
 		lock (EntityCreationLock) {
+			/*
 			writer.Write(_roomLocationPairs.Count);
+			*/
+
+			int count = 0;
+			foreach (Tuple<int, Point> current in _roomLocationPairs) {
+				if (current.Item1 < NPCID.Count)
+					count++;
+			}
+
+			writer.Write(count);
+
 			foreach (Tuple<int, Point> roomLocationPair in _roomLocationPairs) {
+				if (roomLocationPair.Item1 >= NPCID.Count)
+					continue;
+
 				writer.Write(roomLocationPair.Item1);
 				writer.Write(roomLocationPair.Item2.X);
 				writer.Write(roomLocationPair.Item2.Y);
