--- src/TerrariaNetCore/Terraria/GameContent/TreePaintingSettings.cs
+++ src/tModLoader/Terraria/GameContent/TreePaintingSettings.cs
@@ -15,12 +_,12 @@
 
 	public void ApplyShader(int paintColor, Effect shader)
 	{
-		shader.Parameters["leafHueTestOffset"].SetValue(HueTestOffset);
+		shader.Parameters["leafHueTestOffset"]?.SetValue(HueTestOffset);
-		shader.Parameters["leafMinHue"].SetValue(SpecialGroupMinimalHueValue);
+		shader.Parameters["leafMinHue"]?.SetValue(SpecialGroupMinimalHueValue);
-		shader.Parameters["leafMaxHue"].SetValue(SpecialGroupMaximumHueValue);
+		shader.Parameters["leafMaxHue"]?.SetValue(SpecialGroupMaximumHueValue);
-		shader.Parameters["leafMinSat"].SetValue(SpecialGroupMinimumSaturationValue);
+		shader.Parameters["leafMinSat"]?.SetValue(SpecialGroupMinimumSaturationValue);
-		shader.Parameters["leafMaxSat"].SetValue(SpecialGroupMaximumSaturationValue);
+		shader.Parameters["leafMaxSat"]?.SetValue(SpecialGroupMaximumSaturationValue);
-		shader.Parameters["invertSpecialGroupResult"].SetValue(InvertSpecialGroupResult);
+		shader.Parameters["invertSpecialGroupResult"]?.SetValue(InvertSpecialGroupResult);
 		int index = Main.ConvertPaintIdToTileShaderIndex(paintColor, UseSpecialGroups, UseWallShaderHacks);
 		shader.CurrentTechnique.Passes[index].Apply();
 	}
