--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/tModLoader/ReLogic/ReLogic.csproj
@@ -8,7 +_,7 @@
     <RootNamespace>ReLogic</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
     <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
     <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
   </ItemGroup>
@@ -24,4 +_,9 @@
   <ItemGroup>
     <ProjectReference Include="../../../FNA/FNA.csproj" />
   </ItemGroup>
+  <ItemDefinitionGroup>
+    <ReferenceCopyLocalPaths>
+      <DestinationSubDirectory>Libraries\NuGet\</DestinationSubDirectory>
+    </ReferenceCopyLocalPaths>
+  </ItemDefinitionGroup>
 </Project>
