--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/tModLoader/ReLogic/ReLogic.csproj
@@ -9,7 +_,7 @@
     <TargetFramework>net5.0</TargetFramework>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
     <Reference Include="../Terraria/Libraries/FNA/FNA.dll" />
     <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
     <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
@@ -19,5 +_,8 @@
 	<Compile Remove="Localization/IME/WindowsIme.cs" />
 	<Compile Remove="OS/Windows/WindowsMessageHook.cs" />
     <EmbeddedResource Include="Microsoft/**" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Drawing.Common" Version="5.0.1" />
   </ItemGroup>
 </Project>
