--- src/Terraria/ReLogic/ReLogicNetCore.csproj
+++ src/tModLoader/ReLogic/ReLogicNetCore.csproj
@@ -11,9 +_,9 @@
   </PropertyGroup>
   <ItemGroup>
     <FrameworkReference Include="Microsoft.WindowsDesktop.App" />
-    <Reference Include="../Terraria/Libraries/DotNetZip/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
     <Reference Include="../Terraria/Libraries/Native/FNA.dll" />
-    <Reference Include="../Terraria/Libraries/JSON.NET/Newtonsoft.Json.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft/**" />
