--- src/TerrariaNetCore/ReLogic/ReLogic.csproj
+++ src/tModLoader/ReLogic/ReLogic.csproj
@@ -6,12 +_,18 @@
 		<Company>Re-Logic</Company>
 		<Copyright>Copyright © Re-Logic 2017</Copyright>
 		<RootNamespace>ReLogic</RootNamespace>
+		<GenerateDocumentationFile>true</GenerateDocumentationFile>
 	</PropertyGroup>
 	<ItemGroup>
-		<Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
 		<ProjectReference Include="../../../FNA/FNA.Core.csproj" />
 		<Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
 		<Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
+		<PackageReference Include="DotNetZip" Version="1.16.0" />
 	</ItemGroup>
 	<ItemGroup />
+	<ItemDefinitionGroup>
+		<ReferenceCopyLocalPaths>
+			<DestinationSubDirectory>Libraries\NuGet\</DestinationSubDirectory>
+		</ReferenceCopyLocalPaths>
+	</ItemDefinitionGroup>
 </Project>
