--- src/Terraria/ReLogic/ReLogicNetCore.csproj
+++ src/tModLoader/ReLogic/ReLogicNetCore.csproj
@@ -10,7 +_,7 @@
     <AssemblyName>ReLogic</AssemblyName>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
     <Reference Include="../Terraria/Libraries/FNA/FNA.dll" />
     <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
     <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
