--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/tModLoader/ReLogic/ReLogic.csproj
@@ -8,18 +_,18 @@
     <RootNamespace>ReLogic</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <FrameworkReference Include="Microsoft.WindowsDesktop.App" />
-    <Reference Include="../Terraria/Libraries/DotNetZip/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
     <Reference Include="../Terraria/Libraries/FNA/FNA.dll" Condition="$(DefineConstants.Contains('FNA'))" />
     <Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.dll" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.Game.dll" Condition="$(DefineConstants.Contains('XNA'))" />
     <Reference Include="../Terraria/Libraries/XNA/Microsoft.Xna.Framework.Graphics.dll" Condition="$(DefineConstants.Contains('XNA'))" />
-    <Reference Include="../Terraria/Libraries/JSON.NET/Newtonsoft.Json.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
-    <Reference Include="System" />
-    <Reference Include="System.Core" />
-    <Reference Include="System.Drawing" />
-    <Reference Include="System.Windows.Forms" />
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="Microsoft/**" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Drawing.Common" Version="4.7.0" />
   </ItemGroup>
 </Project>
