--- src/Terraria/ReLogic/ReLogic.csproj
+++ src/tModLoader/ReLogic/ReLogic.csproj
@@ -6,10 +_,10 @@
     <Company>Re-Logic</Company>
     <Copyright>Copyright © Re-Logic 2017</Copyright>
     <RootNamespace>ReLogic</RootNamespace>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.CF.dll" />
+    <Reference Include="../Terraria/Libraries/Common/Ionic.Zip.Reduced.dll" />
-    <Reference Include="../Terraria/Libraries/FNA/FNA.dll" />
     <Reference Include="../Terraria/Libraries/Common/Newtonsoft.Json.dll" />
     <Reference Include="../Terraria/Libraries/Common/SteelSeriesEngineWrapper.dll" />
   </ItemGroup>
@@ -22,4 +_,12 @@
   <ItemGroup>
       <PackageReference Include="SixLabors.ImageSharp" Version="1.0.3" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\FNA\FNA.csproj" />
+  </ItemGroup>
+  <ItemDefinitionGroup>
+    <ReferenceCopyLocalPaths>
+        <DestinationSubDirectory>Libraries\NuGet\</DestinationSubDirectory>
+    </ReferenceCopyLocalPaths>
+  </ItemDefinitionGroup>
 </Project>
