--- src/TerrariaNetCore/Terraria/Graphics/TileBatch.cs
+++ src/tModLoader/Terraria/Graphics/TileBatch.cs
@@ -571,7 +_,10 @@
 					}
 
+					// FNA spams a warning with SetDataOptions.None, I think Discard is correct here?
+					// - ChickenBones
+					// Note: The new vanilla code already uses SetDataOptions.Discard here.
 					_vertexBuffer.SetData(_vertices, 0, vbCount * 4, SetDataOptions.Discard);
 				}
 
 				int num3 = Math.Min(num, vbCount - _vertexBufferPosition);
 				_graphicsDevice.DrawIndexedPrimitives(PrimitiveType.TriangleList, _vertexBufferPosition * 4, 0, num3 * 4, 0, num3 * 2);
