--- src/TerrariaNetCore/Terraria/Graphics/TileBatch.cs
+++ src/tModLoader/Terraria/Graphics/TileBatch.cs
@@ -630,7 +_,14 @@
 		for (int i = 0; i <= _lastBatchDrawGroupIndex; i++) {
 			BatchDrawGroup batchDrawGroup = _batchDrawGroups[i];
 			int vertexCount = batchDrawGroup.VertexCount;
+
+			// FNA spams a warning with SetDataOptions.None, I think Discard is correct here?
+			// - ChickenBones
+			vertexBuffer.SetData(_batchDrawGroups[i].VertexArray, 0, vertexCount, SetDataOptions.Discard);
+			/*
 			vertexBuffer.SetData(_batchDrawGroups[i].VertexArray, 0, vertexCount);
+			*/
+
 			for (int j = 0; j < batchDrawGroup.BatchDrawCount; j++) {
 				BatchDrawInfo batchDrawInfo = batchDrawGroup.BatchDraws[j];
 				_graphicsDevice.Textures[0] = batchDrawInfo.Texture;
