--- src/decompiled\Terraria.Initializers\LaunchInitializer.cs
+++ src/merged\Terraria.Initializers\LaunchInitializer.cs
@@ -10,7 +_,11 @@
 		public static void LoadParameters(Main game)
 		{
 			LaunchInitializer.LoadSharedParameters(game);
+#if CLIENT
 			LaunchInitializer.LoadClientParameters(game);
+#else
+			LaunchInitializer.LoadServerParameters(game);
+#endif
 		}
 
 		private static void LoadSharedParameters(Main game)

