--- src/merged\Terraria.GameContent.UI.States\UIVirtualKeyboard.cs
+++ src/Terraria\Terraria.GameContent.UI.States\UIVirtualKeyboard.cs
@@ -698,7 +_,7 @@
 					this.CopyTextToChest();
 				}
 			}
-			byte b = (255 + Main.tileColor.R * 2) / 3;
+			byte b = (byte)((255 + Main.tileColor.R * 2) / 3);
 			Color value = new Color((int)b, (int)b, (int)b, 255);
 			this._textBox.TextColor = Color.Lerp(Color.White, value, 0.2f);
 			this._label.TextColor = Color.Lerp(Color.White, value, 0.2f);

