--- src/decompiled\Terraria.Utilities\CrashDump.cs
+++ src/merged\Terraria.Utilities\CrashDump.cs
@@ -67,7 +_,11 @@
 			DateTime dateTime = DateTime.Now.ToLocalTime();
 			return string.Format("{0}_{1}_{2}_{3}.dmp", new object[]
 			{
+#if CLIENT
 				"Terraria",
+#else
+				"TerrariaServer",
+#endif
 				Main.versionNumber,
 				dateTime.ToString("MM-dd-yy_HH-mm-ss-ffff", CultureInfo.InvariantCulture),
 				Thread.CurrentThread.ManagedThreadId

