--- src/merged\Terraria\Chest.cs
+++ src/Terraria\Terraria\Chest.cs
@@ -268,12 +_,14 @@
 			}
 			int num = (int)(Main.tile[X, Y].frameX / 36);
 			int num2 = num;
+			short num3;
+			int type;
 			switch (num2)
 			{
 			case 2:
 			{
-				short num3 = 36;
-				int type = 11;
+				num3 = 36;
+				type = 11;
 				AchievementsHelper.NotifyProgressionEvent(19);
 				goto IL_B7;
 			}
@@ -281,8 +_,8 @@
 				break;
 			case 4:
 			{
-				short num3 = 36;
-				int type = 11;
+				num3 = 36;
+				type = 11;
 				goto IL_B7;
 			}
 			default:
@@ -298,8 +_,8 @@
 					{
 						return false;
 					}
-					short num3 = 180;
-					int type = 11;
+					num3 = 180;
+					type = 11;
 					AchievementsHelper.NotifyProgressionEvent(20);
 					goto IL_B7;
 				}
@@ -310,8 +_,8 @@
 					case 38:
 					case 40:
 					{
-						short num3 = 36;
-						int type = 11;
+						num3 = 36;
+						type = 11;
 						goto IL_B7;
 					}
 					}
@@ -327,11 +_,9 @@
 				for (int j = Y; j <= Y + 1; j++)
 				{
 					Tile expr_E8 = Main.tile[i, j];
-					short num3;
 					expr_E8.frameX -= num3;
 					for (int k = 0; k < 4; k++)
 					{
-						int type;
 						Dust.NewDust(new Vector2((float)(i * 16), (float)(j * 16)), 16, 16, type, 0f, 0f, 0, default(Color), 1f);
 					}
 				}

