--- src/merged\Terraria\Dust.cs
+++ src/Terraria\Terraria\Dust.cs
@@ -1199,7 +_,7 @@
 						{
 							dust.rotation += 0.1f * dust.scale;
 							Color color = Lighting.GetColor((int)(dust.position.X / 16f), (int)(dust.position.Y / 16f));
-							byte b = (color.R + color.G + color.B) / 3;
+							byte b = (byte)((color.R + color.G + color.B) / 3);
 							float num76 = ((float)b / 270f + 1f) / 2f;
 							float num77 = ((float)b / 270f + 1f) / 2f;
 							float num78 = ((float)b / 270f + 1f) / 2f;

