--- src/merged\Terraria\HitTile.cs
+++ src/Terraria\Terraria\HitTile.cs
@@ -141,11 +_,10 @@
 			hitTileObject.animationDirection = (Main.rand.NextFloat() * 6.28318548f).ToRotationVector2() * 2f;
 			if (tileId != this.bufferLocation)
 			{
-				for (int i = 0; i <= 20; i++)
-				{
-					if (this.order[i] == tileId)
-					{
-						IL_10D:
+				for (int i = 0; i <= 20 + 1; i++)
+				{
+					if (i == 20 + 1 || this.order[i] == tileId)
+					{
 						while (i > 1)
 						{
 							int num = this.order[i - 1];
@@ -157,7 +_,6 @@
 						goto IL_11A;
 					}
 				}
-				goto IL_10D;
 			}
 			this.bufferLocation = this.order[20];
 			this.data[this.bufferLocation].Clear();
@@ -177,11 +_,10 @@
 				return;
 			}
 			this.data[tileId].Clear();
-			for (int i = 0; i < 20; i++)
-			{
-				if (this.order[i] == tileId)
-				{
-					IL_47:
+			for (int i = 0; i < 20 + 1; i++)
+			{
+				if (i == 20 || this.order[i] == tileId)
+				{
 					while (i < 20)
 					{
 						this.order[i] = this.order[i + 1];
@@ -191,7 +_,6 @@
 					return;
 				}
 			}
-			goto IL_47;
 		}
 
 		public void Prune()

