--- src/decompiled/Terraria/Item.cs
+++ src/Terraria/Terraria/Item.cs
@@ -45938,6 +_,9 @@
 		}
 
 		public static Rectangle GetDrawHitbox(int type, Player user) {
+#if SERVER
+			return Rectangle.Empty;
+#else
 			Main.instance.LoadItem(type);
 			if (ItemID.Sets.IsFood[type])
 				return TextureAssets.Item[type].Frame(1, 3, 0, 1);
@@ -45963,6 +_,7 @@
 				default:
 					return TextureAssets.Item[type].Frame();
 			}
+#endif
 		}
 
 		public static int NewItem(Vector2 pos, Vector2 randomBox, int Type, int Stack = 1, bool noBroadcast = false, int prefixGiven = 0, bool noGrabDelay = false, bool reverseLookup = false) => NewItem((int)pos.X, (int)pos.Y, (int)randomBox.X, (int)randomBox.Y, Type, Stack, noBroadcast, prefixGiven, noGrabDelay, reverseLookup);
