--- src/decompiled/Terraria/Main.cs
+++ src/Terraria/Terraria/Main.cs
@@ -75,7 +_,21 @@
 
 namespace Terraria;
 
+#if !WINDOWS
+class Control
+{
+	internal static Form FromHandle(IntPtr handle)
+	{
+		throw new NotImplementedException();
+	}
+}
+#endif
+
+#if SERVER
+public class Main : Terraria.Server.Game
+#else
 public class Main : Game
+#endif
 {
 	public delegate void OnPlayerSelected(PlayerFileData player);
 
@@ -4969,7 +_,7 @@
 		if (!dedServ) {
 			Map = new WorldMap(maxTilesX, maxTilesY);
 			Configuration.Load();
-			graphics = new GraphicsDeviceManager(this as Game);
+			graphics = new GraphicsDeviceManager(((object)this) as Game);
 			base.Content.RootDirectory = "Content";
 		}
 	}
@@ -39052,6 +_,9 @@
 
 		Assets.TransferCompletedAssets();
 		Exit();
+#if MAC && !SERVER
+		SDL2.SDL.SDL_DestroyWindow(base.Window.Handle);
+#endif
 	}
 
 	private void Main_Exiting(object sender, EventArgs e)
