--- src/decompiled/Terraria/Main.cs
+++ src/Terraria/Terraria/Main.cs
@@ -79,7 +_,21 @@
 
 namespace Terraria;
 
+#if !WINDOWS
+class Control
+{
+	internal static Form FromHandle(IntPtr handle)
+	{
+		throw new NotImplementedException();
+	}
+}
+#endif
+
+#if SERVER
+public class Main : Terraria.Server.Game
+#else
 public class Main : Game
+#endif
 {
 	public static class CurrentFrameFlags
 	{
@@ -3821,8 +_,10 @@
 			_windowMover.TryMovingToScreen(screenDeviceName);
 			ThrottleWhenInactive = configuration.Get("ThrottleWhenInactive", defaultValue: true);
 			UnpinFromCore0 = configuration.Get("UnpinFromCore0", defaultValue: false);
+#if !SERVER
 			if (UnpinFromCore0 && Platform.IsWindows)
 				WindowsPerformanceDiagnostics.UnpinFromCore0();
+#endif
 		}
 
 		if (!flag)
@@ -5368,7 +_,7 @@
 		if (!dedServ) {
 			Map = new WorldMap(maxTilesX, maxTilesY);
 			Configuration.Load();
-			graphics = new GraphicsDeviceManager(this as Game);
+			graphics = new GraphicsDeviceManager(((object)this) as Game);
 			base.Content.RootDirectory = "Content";
 		}
 	}
