--- src/decompiled/Terraria/Main.cs
+++ src/Terraria/Terraria/Main.cs
@@ -80,7 +_,21 @@
 
 namespace Terraria;
 
+#if !WINDOWS
+class Control
+{
+	internal static Form FromHandle(IntPtr handle)
+	{
+		throw new NotImplementedException();
+	}
+}
+#endif
+
+#if SERVER
+public class Main : Terraria.Server.Game
+#else
 public class Main : Game
+#endif
 {
 	public static class CurrentFrameFlags
 	{
@@ -5401,7 +_,10 @@
 		if (!dedServ) {
 			Map = new WorldMap(maxTilesX, maxTilesY);
 			Configuration.Load();
-			graphics = new GraphicsDeviceManager(this as Game);
+			graphics = new GraphicsDeviceManager(((object)this) as Game);
+#if FNA
+			Console.WriteLine("SDL Video Driver: " + SDL3.SDL.SDL_GetCurrentVideoDriver());
+#endif
 			base.Content.RootDirectory = "Content";
 		}
 	}
