--- src/decompiled/Terraria/Main.cs
+++ src/Terraria/Terraria/Main.cs
@@ -82,7 +_,11 @@
 
 namespace Terraria;
 
+#if SERVER
+public class Main : Terraria.Server.Game
+#else
 public class Main : Game
+#endif
 {
 	public static class CurrentFrameFlags
 	{
@@ -5582,6 +_,9 @@
 			Map = new WorldMap(maxTilesX, maxTilesY);
 			Configuration.Load();
 			graphics = new GraphicsDeviceManager(this as Game);
+#if FNA
+			Console.WriteLine("SDL Video Driver: " + SDL3.SDL.SDL_GetCurrentVideoDriver());
+#endif
 			base.Content.RootDirectory = "Content";
 		}
 	}
