--- src/merged\Terraria\MessageBuffer.cs
+++ src/Terraria\Terraria\MessageBuffer.cs
@@ -773,7 +_,7 @@
 				if (bitsByte9[0])
 				{
 					player5.pulley = true;
-					player5.pulleyDir = (bitsByte9[1] ? 2 : 1);
+					player5.pulleyDir = (bitsByte9[1] ? (byte)2 : (byte)1);
 				}
 				else
 				{
@@ -1034,11 +_,11 @@
 						bitsByte10 = this.reader.ReadByte();
 						bitsByte11 = this.reader.ReadByte();
 						tile.active(bitsByte10[0]);
-						tile.wall = (bitsByte10[2] ? 1 : 0);
+						tile.wall = (bitsByte10[2] ? (byte)1 : (byte)0);
 						bool flag8 = bitsByte10[3];
 						if (Main.netMode != 2)
 						{
-							tile.liquid = (flag8 ? 1 : 0);
+							tile.liquid = (flag8 ? (byte)1 : (byte)0);
 						}
 						tile.wire(bitsByte10[4]);
 						tile.halfBrick(bitsByte10[5]);

