--- src/merged\Terraria\MessageBuffer.cs
+++ src/Terraria\Terraria\MessageBuffer.cs
@@ -804,7 +_,7 @@
 				if (bitsByte10[0])
 				{
 					player5.pulley = true;
-					player5.pulleyDir = (bitsByte10[1] ? 2 : 1);
+					player5.pulleyDir = (bitsByte10[1] ? (byte)2 : (byte)1);
 				}
 				else
 				{
@@ -1119,11 +_,11 @@
 						bitsByte11 = this.reader.ReadByte();
 						bitsByte12 = this.reader.ReadByte();
 						tile.active(bitsByte11[0]);
-						tile.wall = (bitsByte11[2] ? 1 : 0);
+						tile.wall = (bitsByte11[2] ? (byte)1 : (byte)0);
 						bool flag9 = bitsByte11[3];
 						if (Main.netMode != 2)
 						{
-							tile.liquid = (flag9 ? 1 : 0);
+							tile.liquid = (flag9 ? (byte)1 : (byte)0);
 						}
 						tile.wire(bitsByte11[4]);
 						tile.halfBrick(bitsByte11[5]);
@@ -3482,7 +_,7 @@
 				PlayerDeathReason playerDeathReason2 = PlayerDeathReason.FromReader(this.reader);
 				int num216 = (int)this.reader.ReadInt16();
 				int num217 = (int)(this.reader.ReadByte() - 1);
-				bool pvp2 = this.reader.ReadByte()[0];
+				bool pvp2 = ((BitsByte)this.reader.ReadByte())[0];
 				Main.player[num215].KillMe(playerDeathReason2, (double)num216, num217, pvp2);
 				if (Main.netMode == 2)
 				{

