--- src/decompiled/Terraria/Minecart.cs
+++ src/Terraria/Terraria/Minecart.cs
@@ -55,7 +_,7 @@
 		private static int[][] _tileHeight;
 
 		public static void Initialize() {
-			if ((float)TextureAssets.MinecartMount.Width() != 50f)
+			if (!Main.dedServ && (float)TextureAssets.MinecartMount.Width() != 50f)
 				throw new Exception("Be sure to update Minecart.textureWidth to match the actual texture size of " + 50f + ".");
 
 			_rightSideConnection = new int[36];
