--- src/decompiled/Terraria/NetMessage.cs
+++ src/Terraria/Terraria/NetMessage.cs
@@ -2065,6 +_,12 @@
 
 		public static void CheckBytes(int bufferIndex = 256) {
 			lock (buffer[bufferIndex]) {
+				if (Main.dedServ && Netplay.Clients[bufferIndex].PendingTermination) {
+					Netplay.Clients[bufferIndex].PendingTerminationApproved = true;
+					buffer[bufferIndex].checkBytes = false;
+					return;
+				}
+
 				int num = 0;
 				int num2 = buffer[bufferIndex].totalData;
 				try {
@@ -2073,6 +_,13 @@
 						if (num2 >= num3) {
 							long position = buffer[bufferIndex].reader.BaseStream.Position;
 							buffer[bufferIndex].GetData(num + 2, num3 - 2, out int _);
+
+							if (Main.dedServ && Netplay.Clients[bufferIndex].PendingTermination) {
+								Netplay.Clients[bufferIndex].PendingTerminationApproved = true;
+								buffer[bufferIndex].checkBytes = false;
+								break;
+							}
+
 							buffer[bufferIndex].reader.BaseStream.Position = position + num3;
 							num2 -= num3;
 							num += num3;
@@ -2083,6 +_,9 @@
 					}
 				}
 				catch (Exception) {
+					if (Main.dedServ && num < NetMessage.buffer.Length - 100)
+						Console.WriteLine("Error on message " + NetMessage.buffer[num + 2]);
+
 					num2 = 0;
 					num = 0;
 				}
