--- src/merged\Terraria\NetMessage.cs
+++ src/Terraria\Terraria\NetMessage.cs
@@ -341,7 +_,7 @@
 					writer.Write((byte)number5);
 					break;
 				case 18:
-					writer.Write(Main.dayTime ? 1 : 0);
+					writer.Write((byte)(Main.dayTime ? 1 : 0));
 					writer.Write((int)Main.time);
 					writer.Write(Main.sunModY);
 					writer.Write(Main.moonModY);
@@ -350,7 +_,7 @@
 					writer.Write((byte)number);
 					writer.Write((short)number2);
 					writer.Write((short)number3);
-					writer.Write((number4 == 1f) ? 1 : 0);
+					writer.Write((byte)((number4 == 1f) ? 1 : 0));
 					break;
 				case 20:
 				{
@@ -1616,7 +_,7 @@
 								{
 									short[] arg_1FB_0 = array2;
 									short expr_1F3 = num2;
-									num2 = expr_1F3 + 1;
+									num2 = (short)(expr_1F3 + 1);
 									arg_1FB_0[(int)expr_1F3] = num9;
 								}
 							}
@@ -1627,7 +_,7 @@
 								{
 									short[] arg_23C_0 = array2;
 									short expr_234 = num2;
-									num2 = expr_234 + 1;
+									num2 = (short)(expr_234 + 1);
 									arg_23C_0[(int)expr_234] = num10;
 								}
 							}
@@ -1638,7 +_,7 @@
 								{
 									short[] arg_280_0 = array2;
 									short expr_278 = num2;
-									num2 = expr_278 + 1;
+									num2 = (short)(expr_278 + 1);
 									arg_280_0[(int)expr_278] = num11;
 								}
 							}
@@ -1649,7 +_,7 @@
 								{
 									short[] arg_2C0_0 = array3;
 									short expr_2B7 = num3;
-									num3 = expr_2B7 + 1;
+									num3 = (short)(expr_2B7 + 1);
 									arg_2C0_0[(int)expr_2B7] = (short)num12;
 								}
 							}
@@ -1660,7 +_,7 @@
 								{
 									short[] arg_300_0 = array3;
 									short expr_2F7 = num3;
-									num3 = expr_2F7 + 1;
+									num3 = (short)(expr_2F7 + 1);
 									arg_300_0[(int)expr_2F7] = (short)num13;
 								}
 							}
@@ -1966,7 +_,7 @@
 								}
 								else
 								{
-									tile.slope(b4 - 1);
+									tile.slope((byte)(b4 - 1));
 								}
 							}
 						}

