--- src/decompiled/Terraria/Netplay.cs
+++ src/Terraria/Terraria/Netplay.cs
@@ -1,8 +_,10 @@
-using NATUPNPLib;
 using System;
 using System.IO;
 using System.Net;
 using System.Net.Sockets;
+#if !WINDOWS
+using System.Runtime.CompilerServices;
+#endif
 using System.Threading;
 using Terraria.IO;
 using Terraria.Localization;
@@ -10,6 +_,12 @@
 using Terraria.Net.Sockets;
 using Terraria.Social;
 using Terraria.Utilities;
+#if !WINDOWS
+using Terraria.Social.Base;
+#endif
+#if SERVER
+using NATUPNPLib;
+#endif
 
 namespace Terraria
 {
@@ -32,8 +_,10 @@
 		public static bool spamCheck = false;
 		public static bool anyClients = false;
 		private static Thread ServerThread;
+#if SERVER
 		public static UPnPNAT upnpnat;
 		public static IStaticPortMappingCollection mappings;
+#endif
 		public static string portForwardIP;
 		public static int portForwardPort;
 		public static bool portForwardOpen;
@@ -43,28 +_,37 @@
 		private static void OpenPort() {
 			portForwardIP = GetLocalIPAddress();
 			portForwardPort = ListenPort;
-			if (upnpnat == null) {
-				upnpnat = (UPnPNAT)Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid("AE1E00AA-3FD5-403C-8A27-2BBDC30CD0E1")));
-				mappings = upnpnat.StaticPortMappingCollection;
-			}
-
-			if (mappings == null)
-				return;
-
-			foreach (IStaticPortMapping mapping in mappings) {
-				if (mapping.InternalPort == portForwardPort && mapping.InternalClient == portForwardIP && mapping.Protocol == "TCP")
-					portForwardOpen = true;
-			}
-
-			if (!portForwardOpen) {
-				mappings.Add(portForwardPort, "TCP", portForwardPort, portForwardIP, bEnabled: true, "Terraria Server");
-				portForwardOpen = true;
-			}
+#if SERVER
+			if (Netplay.upnpnat == null)
+			{
+				Netplay.upnpnat = (UPnPNAT)Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid("AE1E00AA-3FD5-403C-8A27-2BBDC30CD0E1")));
+				Netplay.mappings = Netplay.upnpnat.StaticPortMappingCollection;
+			}
+			if (Netplay.mappings != null)
+			{
+				foreach (IStaticPortMapping staticPortMapping in Netplay.mappings)
+				{
+					if (staticPortMapping.InternalPort == Netplay.portForwardPort && staticPortMapping.InternalClient == Netplay.portForwardIP && staticPortMapping.Protocol == "TCP")
+					{
+						Netplay.portForwardOpen = true;
+					}
+				}
+				if (!Netplay.portForwardOpen)
+				{
+					Netplay.mappings.Add(Netplay.portForwardPort, "TCP", Netplay.portForwardPort, Netplay.portForwardIP, true, "Terraria Server");
+					Netplay.portForwardOpen = true;
+				}
+			}
+#endif
 		}
 
 		public static void closePort() {
-			if (portForwardOpen)
-				mappings.Remove(portForwardPort, "TCP");
+#if SERVER
+			if (Netplay.portForwardOpen)
+			{
+				Netplay.mappings.Remove(Netplay.portForwardPort, "TCP");
+			}
+#endif
 		}
 
 		public static string GetLocalIPAddress() {
@@ -174,6 +_,11 @@
 					flag = false;
 				}
 				catch {
+#if MAC
+					Thread.Sleep(200);
+					Netplay.Connection.Socket.Close();
+					Netplay.Connection.Socket = new TcpSocket();
+#endif
 					if (!disconnect && Main.gameMenu)
 						flag = true;
 				}
@@ -594,13 +_,20 @@
 		}
 
 		public static void Initialize() {
+#if CLIENT
+			NetMessage.buffer[256] = new MessageBuffer();
+			NetMessage.buffer[256].whoAmI = 256;
+#else
-			for (int i = 0; i < 257; i++) {
+			for (int i = 0; i < 257; i++)
+			{
 				if (i < 256)
-					Clients[i] = new RemoteClient();
-
+				{
+					Netplay.Clients[i] = new RemoteClient();
+				}
 				NetMessage.buffer[i] = new MessageBuffer();
 				NetMessage.buffer[i].whoAmI = i;
 			}
+#endif
 		}
 
 		public static int GetSectionX(int x) => x / 200;
