--- src/merged\Terraria\NPC.cs
+++ src/Terraria\Terraria\NPC.cs
@@ -33766,7 +_,7 @@
 																				this.ai[0] = -300f;
 																				this.netUpdate = true;
 																			}
-																			base.Center + Vector2.UnitX * (float)this.direction * 200f;
+																			Vector2 _ = base.Center + Vector2.UnitX * (float)this.direction * 200f;
 																			Vector2 vector223 = base.Center + Vector2.UnitX * (float)this.direction * 50f - Vector2.UnitY * 6f;
 																			if (this.ai[0] == 54f && Main.netMode != 1)
 																			{
@@ -50721,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_2E2B;
+							break;
 						case 3:
-							IL_2E0C:
+						default:
 							if (num33 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_2E2B;
-							}
-							goto IL_2E2B;
-						}
-						goto IL_2E0C;
-						IL_2E2B:
+							}
+							break;
+						}
 						int num34;
 						if (this.frameCounter < 10.0)
 						{
@@ -50865,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_328A;
+							break;
 						case 3:
-							IL_326B:
+						default:
 							if (num37 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_328A;
-							}
-							goto IL_328A;
-						}
-						goto IL_326B;
-						IL_328A:
+							}
+							break;
+						}
 						int num38 = 0;
 						if (this.frameCounter < 16.0)
 						{
@@ -50923,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_3435;
+							break;
 						case 3:
-							IL_3416:
+						default:
 							if (num39 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_3435;
-							}
-							goto IL_3435;
-						}
-						goto IL_3416;
-						IL_3435:
+							}
+							break;
+						}
 						int num40;
 						if (this.frameCounter < 10.0)
 						{
@@ -50969,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_3522;
+							break;
 						case 3:
-							IL_3503:
+						default:
 							if (num41 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_3522;
-							}
-							goto IL_3522;
-						}
-						goto IL_3503;
-						IL_3522:
+							}
+							break;
+						}
 						int num42;
 						if (this.frameCounter >= 10.0)
 						{
@@ -51149,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_396C;
+							break;
 						case 3:
-							IL_394D:
+						default:
 							if (num55 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_396C;
-							}
-							goto IL_396C;
-						}
-						goto IL_394D;
-						IL_396C:
+							}
+							break;
+						}
 						bool flag2 = this.ai[0] == 3f;
 						int num56 = 0;
 						int num57 = 0;
@@ -51274,19 +_,16 @@
 						case 2:
 						case 4:
 						case 5:
-							goto IL_3CE0;
+							break;
 						case 3:
-							IL_3CC1:
+						default:
 							if (num60 != 0)
 							{
 								this.frame.Y = 0;
 								this.frameCounter = 0.0;
-								goto IL_3CE0;
-							}
-							goto IL_3CE0;
-						}
-						goto IL_3CC1;
-						IL_3CE0:
+							}
+							break;
+						}
 						bool flag3 = this.ai[0] == 16f;
 						int num61 = 0;
 						int num62 = -1;
@@ -62526,7 +_,7 @@
 						{
 							if ((n == num53 - num55 || n == num53 + num55 || num56 == num54 - num55 || num56 == num54 + num55) && !Main.tile[n, num56].active())
 							{
-								Main.tile[n, num56].type = (WorldGen.crimson ? 347 : 140);
+								Main.tile[n, num56].type = (ushort)(WorldGen.crimson ? 347 : 140);
 								Main.tile[n, num56].active(true);
 							}
 							Main.tile[n, num56].lava(false);
@@ -70205,7 +_,7 @@
 					while ((double)num105 < 10.0 + dmg / 10.0)
 					{
 						int num106 = this.width / 4;
-						((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
+						Vector2 _ = ((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
 						int num107 = Dust.NewDust(base.Center - Vector2.One * (float)num106, num106 * 2, num106 * 2, 229, 0f, 0f, 0, default(Color), 1f);
 						Dust dust31 = Main.dust[num107];
 						Vector2 vector3 = Vector2.Normalize(dust31.position - base.Center);
@@ -70228,7 +_,7 @@
 					for (int num108 = 0; num108 < 60; num108++)
 					{
 						int num109 = this.width / 4;
-						((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
+						Vector2 _ = ((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
 						int num110 = Dust.NewDust(base.Center - Vector2.One * (float)num109, num109 * 2, num109 * 2, 229, 0f, 0f, 0, default(Color), 1f);
 						Dust dust32 = Main.dust[num110];
 						Vector2 vector4 = Vector2.Normalize(dust32.position - base.Center);
@@ -71106,7 +_,7 @@
 					for (int num207 = 0; num207 < 60; num207++)
 					{
 						int num208 = 25;
-						((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
+						Vector2 _ = ((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
 						int num209 = Dust.NewDust(base.Center - Vector2.One * (float)num208, num208 * 2, num208 * 2, 212, 0f, 0f, 0, default(Color), 1f);
 						Dust dust47 = Main.dust[num209];
 						Vector2 vector7 = Vector2.Normalize(dust47.position - base.Center);
@@ -77107,7 +_,7 @@
 					{
 						if (this.boss)
 						{
-							this.oldPosition - this.position;
+							Vector2 _ = this.oldPosition - this.position;
 							if (this.netSpam <= 15)
 							{
 								this.netSpam += 5;

