--- src/decompiled/Terraria/NPC.cs
+++ src/Terraria/Terraria/NPC.cs
@@ -24710,7 +_,8 @@
 						float num1181 = 1f + Main.rand.NextFloat();
 						float fadeIn = 0.4f + Main.rand.NextFloat();
 						int num1182 = Utils.SelectRandom<int>(Main.rand, 31, 229);
-						if (flag76) {
+
+						if (!Main.dedServ && flag76) {
 							MoonlordDeathDrama.AddExplosion(vector146);
 							for (float num1183 = 0f; num1183 < num1178 * 2f; num1183++) {
 								Dust dust6 = Main.dust[Dust.NewDust(vector146, 0, 0, 229)];
