--- src/merged\Terraria\Player.cs
+++ src/Terraria\Terraria\Player.cs
@@ -12704,7 +_,7 @@
 						{
 							list10.Add(tuple9);
 						}
-						Main.tile[tuple9.Item1, tuple9.Item2];
+						Tile _ = Main.tile[tuple9.Item1, tuple9.Item2];
 						if (Collision.HitWallSubstep(tuple9.Item1, tuple9.Item2))
 						{
 							list10.Add(tuple9);
@@ -24981,8 +_,8 @@
 						flag2 = true;
 						int num22 = (int)(Main.tile[myX, myY].frameX / 54);
 						int num23 = (int)(Main.tile[myX, myY].frameY / 54);
-						short arg_BFD_0 = Main.tile[myX, myY].frameX % 54 / 18;
-						short arg_C17_0 = Main.tile[myX, myY].frameY % 54 / 18;
+						short arg_BFD_0 = (short)(Main.tile[myX, myY].frameX % 54 / 18);
+						short arg_C17_0 = (short)(Main.tile[myX, myY].frameY % 54 / 18);
 						int num24 = -1;
 						switch (num22)
 						{
@@ -36019,7 +_,7 @@
 													{
 														num258 = 0;
 													}
-													Main.tile[num242, num243].frameX = 22 * num258;
+													Main.tile[num242, num243].frameX = (short) (22 * num258);
 													if (Main.netMode == 1)
 													{
 														NetMessage.SendTileSquare(-1, Player.tileTargetX, Player.tileTargetY, 1, TileChangeType.None);
@@ -39297,7 +_,7 @@
 				FileUtilities.Copy(path, path + ".bak", isCloudSave, true);
 			}
 			RijndaelManaged rijndaelManaged = new RijndaelManaged();
-			using (Stream stream = isCloudSave ? new MemoryStream(2000) : new FileStream(path, FileMode.Create))
+			using (Stream stream = isCloudSave ? (Stream)new MemoryStream(2000) : (Stream)new FileStream(path, FileMode.Create))
 			{
 				using (CryptoStream cryptoStream = new CryptoStream(stream, rijndaelManaged.CreateEncryptor(Player.ENCRYPTION_KEY, Player.ENCRYPTION_KEY), CryptoStreamMode.Write))
 				{
@@ -40211,7 +_,6 @@
 			this._shaderObstructionInternalValue = 1f;
 			int[] array2 = new int[2];
 			this.hurtCooldowns = array2;
-			base..ctor();
 			this.width = 20;
 			this.height = 42;
 			this.name = string.Empty;

