--- src/decompiled/Terraria/Player.cs
+++ src/Terraria/Terraria/Player.cs
@@ -22458,10 +_,12 @@
 		if ((Main.netMode != 2 && whoAmI == Main.myPlayer) || whoAmI == Main.LocalPlayer.spectating)
 			ActiveSections.CheckSection(position);
 
+#if !SERVER
 		if (DebugOptions.ShowSections && whoAmI == Main.myPlayer) {
 			Point point = new Point(Netplay.GetSectionX((int)position.X >> 4), Netplay.GetSectionY((int)position.Y >> 4));
 			DebugLineDraw.World.AddRectangle(new Vector2(point.X * 200 * 16, point.Y * 150 * 16), new Vector2((point.X + 1) * 200 * 16, (point.Y + 1) * 150 * 16), Color.Yellow);
 		}
+#endif
 	}
 
 	private void Update_AdjustTileTargetForDisplayJars(int i)
