--- src/decompiled/Terraria/Program.cs
+++ src/Terraria/Terraria/Program.cs
@@ -18,10 +_,19 @@
 {
 	public static class Program
 	{
+#if XNA
 		public static bool IsXna = true;
 		public static bool IsFna = false;
+#else
+		public static bool IsXna = false;
+		public static bool IsFna = true;
+#endif
+#if MONO
+		public static bool IsMono = true;
+#else
 		public static bool IsMono = false;
+#endif
-		public const bool IsDebug = false;
+		// public const bool IsDebug = false;
 		public static Dictionary<string, string> LaunchParameters = new Dictionary<string, string>();
 		private static int ThingsToLoad;
 		private static int ThingsLoaded;
@@ -147,6 +_,9 @@
 			}
 
 			LaunchParameters = Utils.ParseArguements(args);
+#if SERVER
+			Main.dedServ = true;
+#endif
 			ThreadPool.SetMinThreads(8, 8);
 			LanguageManager.Instance.SetLanguage(GameCulture.DefaultCulture);
 			InitializeConsoleOutput();
