--- src/merged\Terraria\Projectile.cs
+++ src/Terraria\Terraria\Projectile.cs
@@ -4513,13 +_,12 @@
 					case 412:
 					case 413:
 					case 414:
-						IL_7A81:
+					default:
 						this.width = 10;
 						this.height = 10;
 						this.aiStyle = 10;
 						goto IL_C73D;
 					}
-					goto IL_7A81;
 				}
 				if (this.type == 415 || this.type == 416 || this.type == 417 || this.type == 418)
 				{
@@ -12706,100 +_,34 @@
 			string name;
 			if (Main.player[this.owner].active && (name = Main.player[this.owner].name) != null)
 			{
-				if (<PrivateImplementationDetails>{E90A5A2F-CD10-4A2C-9D2A-6B036D4E8877}.$$method0x6000dca-1 == null)
-				{
-					<PrivateImplementationDetails>{E90A5A2F-CD10-4A2C-9D2A-6B036D4E8877}.$$method0x6000dca-1 = new Dictionary<string, int>(14)
-					{
-						{
-							"Tsuki",
-							0
-						},
-						{
-							"Yoraiz0r",
-							1
-						},
-						{
-							"Ghostar",
-							2
-						},
-						{
-							"Devalaous",
-							3
-						},
-						{
-							"Leinfors",
-							4
-						},
-						{
-							"Aeroblop",
-							5
-						},
-						{
-							"Doylee",
-							6
-						},
-						{
-							"Darkhalis",
-							7
-						},
-						{
-							"Arkhalis",
-							8
-						},
-						{
-							"Nike Leon",
-							9
-						},
-						{
-							"Suweeka",
-							10
-						},
-						{
-							"W1K",
-							11
-						},
-						{
-							"Grox The Great",
-							12
-						},
-						{
-							"Random",
-							13
-						}
-					};
-				}
-				int num;
-				if (<PrivateImplementationDetails>{E90A5A2F-CD10-4A2C-9D2A-6B036D4E8877}.$$method0x6000dca-1.TryGetValue(name, out num))
-				{
-					switch (num)
-					{
-					case 0:
-					case 1:
-						return 0.85f;
-					case 2:
-						return 0.33f;
-					case 3:
-						return 0.66f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.1f;
-					case 4:
-						return 0.77f;
-					case 5:
-						return 0.25f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.1f;
-					case 6:
-						return 0f;
-					case 7:
-					case 8:
-						return 0.75f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.07f;
-					case 9:
-						return 0.075f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.07f;
-					case 10:
-						return 0.5f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.18f;
-					case 11:
-						return 0.75f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.05f;
-					case 12:
-						return 0.31f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.03f;
-					case 13:
-						return Main.rand.NextFloat();
-					}
+				switch (name)
+				{
+				case "Tsuki":
+				case "Yoraiz0r":
+					return 0.85f;
+				case "Ghostar":
+					return 0.33f;
+				case "Devalaous":
+					return 0.66f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.1f;
+				case "Leinfors":
+					return 0.77f;
+				case "Aeroblop":
+					return 0.25f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.1f;
+				case "Doylee":
+					return 0f;
+				case "Darkhalis":
+				case "Arkhalis":
+					return 0.75f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.07f;
+				case "Nike Leon":
+					return 0.075f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.07f;
+				case "Suweeka":
+					return 0.5f + (float)Math.Cos(Main.time / 180.0 * 6.2831854820251465) * 0.18f;
+				case "W1K":
+					return 0.75f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.05f;
+				case "Grox The Great":
+					return 0.31f + (float)Math.Cos(Main.time / 120.0 * 6.2831854820251465) * 0.03f;
+				case "Random":
+					return Main.rand.NextFloat();
 				}
 			}
 			return (float)((int)indexing) / 6f;
@@ -21384,7 +_,7 @@
 																int num677 = (int)(this.position.Y + (float)this.height) / 16;
 																if (WorldGen.SolidTile(num676, num677) || Main.tile[num676, num677].halfBrick() || Main.tile[num676, num677].slope() > 0)
 																{
-																	goto IL_1BA21;
+																	goto Block_1935;
 																}
 																if (TileID.Sets.Platforms[(int)Main.tile[num676, num677].type] && Main.tile[num676, num677].active() && !Main.tile[num676, num677].inActive())
 																{
@@ -36671,7 +_,7 @@
 							for (int n = 0; n < 5; n++)
 							{
 								int num31 = this.width / 4;
-								((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
+								Vector2 _ = ((float)Main.rand.NextDouble() * 6.28318548f).ToRotationVector2() * (float)Main.rand.Next(24, 41) / 8f;
 								int num32 = Dust.NewDust(base.Center - Vector2.One * (float)num31, num31 * 2, num31 * 2, 88, 0f, 0f, 0, default(Color), 1f);
 								Dust dust5 = Main.dust[num32];
 								Vector2 value3 = Vector2.Normalize(dust5.position - base.Center);
@@ -39220,7 +_,6 @@
 				{
 					Vector2 mouseWorld3 = Main.MouseWorld;
 					Vector2 value3 = player.DirectionTo(mouseWorld3);
-					value3 * 0f;
 				}
 				float num13 = this.rotation - 0.7853982f * (float)num10;
 				value2 = (num13 + ((num10 == -1) ? 3.14159274f : 0f)).ToRotationVector2() * (this.ai[0] / num) * scaleFactor;

