--- src/decompiled/Terraria/TimeLogger.cs
+++ src/Terraria/Terraria/TimeLogger.cs
@@ -617,6 +_,7 @@
 
 	private static void DrawExtras()
 	{
+#if !SERVER
 		if (Platform.IsWindows) {
 			WindowsPerformanceDiagnostics.Data data = WindowsPerformanceDiagnostics.GetData();
 			int num = Main.screenWidth - 180;
@@ -634,6 +_,7 @@
 				DrawString("reducing performance", new Vector2(num, num2 += 12), Color.Orange);
 			}
 		}
+#endif
 
 		if (!PlayerInput.UsingGamepad)
 			return;
