--- src/merged\Terraria\WaterfallManager.cs
+++ src/Terraria\Terraria\WaterfallManager.cs
@@ -289,12 +_,13 @@
 				int num18 = 0;
 				int num19 = 0;
 				int num20 = 0;
+				int num21;
 				int num22;
 				if (num12 == 1 || num12 == 14)
 				{
 					if (!Main.drewLava && this.waterfalls[i].stopAtStep != 0)
 					{
-						int num21 = 32 * this.slowFrame;
+						num21 = 32 * this.slowFrame;
 						goto IL_46D;
 					}
 				}
@@ -310,7 +_,7 @@
 						{
 							goto IL_18F5;
 						}
-						int num21 = 32 * this.regularFrame;
+						num21 = 32 * this.regularFrame;
 						goto IL_46D;
 					}
 					if (!Main.drewLava)
@@ -760,7 +_,6 @@
 								Main.dust[num55].velocity *= 0f;
 							}
 						}
-						int num21;
 						if (num15 == 0 && num36 != 0 && num16 == 1 && num17 != num18)
 						{
 							num36 = 0;

