--- src/decompiled/Terraria/WindowsLaunch.cs
+++ src/Terraria/Terraria/WindowsLaunch.cs
@@ -1,3 +_,4 @@
+#if WINDOWS
 using System;
 using System.IO;
 using System.Reflection;
@@ -64,6 +_,11 @@
 			return Assembly.Load(array);
 		};
 
+#if SERVER
+		WindowsLaunch._handleRoutine = new WindowsLaunch.HandlerRoutine(WindowsLaunch.ConsoleCtrlCheck);
+		WindowsLaunch.SetConsoleCtrlHandler(WindowsLaunch._handleRoutine, true);
+#endif
 		Program.LaunchGame(args);
 	}
 }
+#endif
