--- src/decompiled/WindowsLaunch.cs
+++ src/Terraria/WindowsLaunch.cs
@@ -1,3 +_,4 @@
+#if WINDOWS
 using System;
 using System.IO;
 using System.Reflection;
@@ -63,7 +_,12 @@
 				}
 			};
 
+#if SERVER
+			WindowsLaunch._handleRoutine = new WindowsLaunch.HandlerRoutine(WindowsLaunch.ConsoleCtrlCheck);
+			WindowsLaunch.SetConsoleCtrlHandler(WindowsLaunch._handleRoutine, true);
+#endif
 			Program.LaunchGame(args);
 		}
 	}
 }
+#endif
