--- src/merged\Terraria\Wiring.cs
+++ src/Terraria\Terraria\Wiring.cs
@@ -719,7 +_,7 @@
 					}
 					if (flag5 != flag2 || flag6 || flag7)
 					{
-						short arg_183_0 = tile.frameX % 18 / 18;
+						short arg_183_0 = (short)(tile.frameX % 18 / 18);
 						tile.frameX = (short)(18 * flag5.ToInt());
 						if (flag4)
 						{
@@ -868,11 +_,11 @@
 									{
 										Dictionary<Point16, byte> pixelBoxTriggers;
 										Point16 key;
-										(pixelBoxTriggers = Wiring._PixelBoxTriggers)[key = point2] = (pixelBoxTriggers[key] | ((j == 0 | j == 1) ? 2 : 1));
+										(pixelBoxTriggers = Wiring._PixelBoxTriggers)[key = point2] = (byte)(pixelBoxTriggers[key] | ((j == 0 | j == 1) ? 2 : 1));
 									}
 									else
 									{
-										Wiring._PixelBoxTriggers[point2] = ((j == 0 | j == 1) ? 2 : 1);
+										Wiring._PixelBoxTriggers[point2] = (byte)((j == 0 | j == 1) ? 2 : 1);
 									}
 								}
 								bool flag;

