--- src/decompiled/Terraria/WorldGen.cs
+++ src/Terraria/Terraria/WorldGen.cs
@@ -39511,8 +_,10 @@
 				if (!noItem && FixExploitManEaters.SpotProtected(i, j))
 					return;
 
+#if CLIENT
 				if (!gen && !Main.gameMenu)
 					KillTile_PlaySounds(i, j, fail, tile);
+#endif
 			}
 
 			if (tile.type == 128 || tile.type == 269) {
@@ -51120,6 +_,7 @@
 
 		public static bool UpdateMapTile(int i, int j, bool addToList = true) {
 			bool result = false;
+#if CLIENT
 			if (Main.mapEnabled && !noMapUpdate && !gen && Main.Map[i, j].Light > 0 && Main.Map.UpdateType(i, j) && addToList) {
 				result = true;
 				if (MapHelper.numUpdateTile < MapHelper.maxUpdateTile - 1) {
@@ -51131,6 +_,7 @@
 					Main.refreshMap = true;
 				}
 			}
+#endif
 
 			return result;
 		}
