--- src/tModLoader/Terraria.ModLoader.Engine/HiDefGraphicsIssues.cs
+++ src/64bit/Terraria.ModLoader.Engine/HiDefGraphicsIssues.cs
@@ -13,6 +_,7 @@
 using System.Threading;
 using System.Windows.Forms;
 using Terraria.ModLoader.IO;
+using Terraria.Utilities;
 
 namespace Terraria.ModLoader.Engine
 {
@@ -135,11 +_,11 @@
 
 			//var modsAffected = ModContent.HiDefMods.Count == 0 ? "No mods will be affected." : $"The following mods will be affected {string.Join(", ", ModContent.HiDefMods.Select(m => m.DisplayName))}";
 			string message = $"tML encountered a crash when testing some experimental graphics features. If this issue persists consistently, you may have to edit config.json and set the Support4K setting to false. \nPlease restart your game.\nReport Status: {reportStatus}";
-#if !MAC
-			MessageBox.Show(message, "Graphics Engine Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
-#else
-			UI.Interface.MessageBoxShow(message, "Graphics Engine Failure");
-#endif
+			if(!PlatformUtilities.IsOSX)
+				MessageBox.Show(message, "Graphics Engine Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
+			else
+				UI.Interface.MessageBoxShow(message, "Graphics Engine Failure");
+
 			Environment.Exit(1);
 		}
 
