--- src/tModLoader/Terraria.ModLoader.Engine/HiDefGraphicsIssues.cs
+++ src/64bit/Terraria.ModLoader.Engine/HiDefGraphicsIssues.cs
@@ -14,6 +_,7 @@
 using System.Windows.Forms;
 using Terraria.Localization;
 using Terraria.ModLoader.IO;
+using Terraria.Utilities;
 
 namespace Terraria.ModLoader.Engine
 {
@@ -50,7 +_,7 @@
 #endif
 #if DEBUG
 	        mainThreadId = Thread.CurrentThread.ManagedThreadId;
-	        new Hook(typeof(Texture2D).FindMethod("ValidateCreationParameters"), new hook_ValidateCreationParameters(HookValidateCreationParameters));
+	        //new Hook(typeof(Texture2D).FindMethod("ValidateCreationParameters"), new hook_ValidateCreationParameters(HookValidateCreationParameters));
 #endif
         }
 
@@ -136,11 +_,11 @@
 
 			//var modsAffected = ModContent.HiDefMods.Count == 0 ? "No mods will be affected." : $"The following mods will be affected {string.Join(", ", ModContent.HiDefMods.Select(m => m.DisplayName))}";
 			string message = Language.GetTextValue("tModLoader.GraphicsEngineFailureMessage", reportStatus);
-#if !MAC
-			MessageBox.Show(message, Language.GetTextValue("tModLoader.GraphicsEngineFailure"), MessageBoxButtons.OK, MessageBoxIcon.Error);
-#else
-			UI.Interface.MessageBoxShow(message, Language.GetTextValue("tModLoader.GraphicsEngineFailure"));
-#endif
+			if(!PlatformUtilities.IsOSX)
+				MessageBox.Show(message, "Graphics Engine Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
+			else
+				UI.Interface.MessageBoxShow(message, "Graphics Engine Failure");
+
 			Environment.Exit(1);
 		}
 
