--- src/TerrariaNetCore/Terraria/GameContent/Liquid/LiquidRenderer.cs
+++ src/tModLoader/Terraria/GameContent/Liquid/LiquidRenderer.cs
@@ -1,4 +_,5 @@
 using System;
+using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
@@ -10,7 +_,7 @@
 
 public class LiquidRenderer
 {
-	private struct LiquidCache
+	public struct LiquidCache
 	{
 		public float LiquidLevel;
 		public float VisibleLiquidLevel;
@@ -37,7 +_,7 @@
 		public byte VisibleType;
 	}
 
-	private struct LiquidDrawCache
+	public struct LiquidDrawCache
 	{
 		public Rectangle SourceRectangle;
 		public Vector2 LiquidOffset;
@@ -48,7 +_,7 @@
 		public bool HasWall;
 	}
 
-	private struct SpecialLiquidDrawCache
+	public struct SpecialLiquidDrawCache
 	{
 		public int X;
 		public int Y;
@@ -64,13 +_,13 @@
 	private const int ANIMATION_FRAME_COUNT = 16;
 	private const int CACHE_PADDING = 2;
 	private const int CACHE_PADDING_2 = 4;
-	private static readonly int[] WATERFALL_LENGTH = new int[4] {
+	public static readonly int[] WATERFALL_LENGTH = new int[4] {
 		10,
 		3,
 		2,
 		10
 	};
-	private static readonly float[] DEFAULT_OPACITY = new float[4] {
+	public static readonly float[] DEFAULT_OPACITY = new float[4] {
 		0.6f,
 		0.95f,
 		0.95f,
@@ -86,7 +_,7 @@
 	};
 	public const float MIN_LIQUID_SIZE = 0.25f;
 	public static LiquidRenderer Instance;
-	private readonly Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
+	public Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
 	private LiquidCache[] _cache = new LiquidCache[1];
 	private LiquidDrawCache[] _drawCache = new LiquidDrawCache[1];
 	private SpecialLiquidDrawCache[] _drawCacheForShimmer = new SpecialLiquidDrawCache[1];
@@ -96,7 +_,9 @@
 	private Color[] _waveMask = new Color[1];
 	private float _frameState;
 
+	/*
 	private static Tile[,] Tiles => Main.tile;
+	*/
 
 	public event Action<Color[], Rectangle> WaveFilters;
 
@@ -131,14 +_,14 @@
 		if (_waveMask.Length < drawArea.Width * drawArea.Height)
 			_waveMask = new Color[drawArea.Width * drawArea.Height];
 
-		Tile tile = null;
+		Tile tile = default;
 		fixed (LiquidCache* ptr = &_cache[1]) {
 			LiquidCache* ptr2 = ptr;
 			int num = rectangle.Height * 2 + 2;
 			ptr2 = ptr;
 			for (int i = rectangle.X; i < rectangle.X + rectangle.Width; i++) {
 				for (int j = rectangle.Y; j < rectangle.Y + rectangle.Height; j++) {
-					tile = Tiles[i, j];
+					tile = Main.tile[i, j];
 					if (tile == null)
 						tile = new Tile();
 
@@ -524,7 +_,17 @@
 			LiquidDrawCache* ptr2 = ptr;
 			for (int i = drawArea.X; i < drawArea.X + drawArea.Width; i++) {
 				for (int j = drawArea.Y; j < drawArea.Y + drawArea.Height; j++) {
-					if (ptr2->IsVisible) {
+					// TML(liquid-slopes): Also attempt to draw if the
+					// coordinate is on the edge of a tile.
+					bool isEdge = Main.instance.TilesRenderer.EdgeTiles.TryGetValue(new Point(i, j), out var edgeCache);
+
+					if (ptr2->IsVisible || isEdge) {
+						// IsEdge is unique because it means there isn't
+						// actually a liquid present in the tile, so a lot of
+						// data isn't present.
+						// The below data is mostly irrelevant for edge tiles,
+						// so it's fine for this data to be missing.
+
 						Rectangle sourceRectangle = ptr2->SourceRectangle;
 						if (ptr2->IsSurfaceLiquid)
 							sourceRectangle.Y = 1280;
@@ -532,8 +_,14 @@
 							sourceRectangle.Y += _animationFrame * 80;
 
 						Vector2 liquidOffset = ptr2->LiquidOffset;
-						float num = ptr2->Opacity * (isBackgroundDraw ? 1f : DEFAULT_OPACITY[ptr2->Type]);
-						int num2 = ptr2->Type;
+
+						// Opacity is important for edge tiles since we want to
+						// support clean transitions between biomes.
+						var type = isEdge ? Main.tile[i, j].LiquidType : ptr2->Type;
+						var opacity = isEdge ? 1f : ptr2->Opacity;
+
+						float num = opacity * (isBackgroundDraw ? 1f : DEFAULT_OPACITY[type]);
+						int num2 = type;
 						switch (num2) {
 							case 0:
 								num2 = waterStyle;
@@ -544,14 +_,25 @@
 								break;
 						}
 
+						// Colors are important for edge tiles, especially in
+						// case mods make edits to them for custom liquid
+						// rendering.
 						num = Math.Min(1f, num);
 						Lighting.GetCornerColors(i, j, out var vertices);
 						vertices.BottomLeftColor *= num;
 						vertices.BottomRightColor *= num;
 						vertices.TopLeftColor *= num;
 						vertices.TopRightColor *= num;
+
+						// Explicitly handle IsVisible vs IsEdge rendering here.
+						if (!isEdge) {
-						Main.DrawTileInWater(drawOffset, i, j);
+							Main.DrawTileInWater(drawOffset, i, j);
-						Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(i << 4, j << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+							Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(i << 4, j << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+						}
+						else {
+							var tile = Main.tile[i, j];
+							Main.instance.TilesRenderer.DrawPartialLiquid(isBackgroundDraw, tile, new Vector2(i << 4, j << 4) + drawOffset + edgeCache.LiquidOffset, edgeCache.LiquidFrame, edgeCache.LiquidType, vertices, edgeCache.IsWaterForFading);
+						}
 					}
 
 					ptr2++;
@@ -570,28 +_,50 @@
 			SpecialLiquidDrawCache* ptr2 = ptr;
 			int num = _drawCacheForShimmer.Length;
 			for (int i = 0; i < num; i++) {
-				if (!ptr2->IsVisible)
+				int num3 = ptr2->X + drawArea.X - 2;
+				int num4 = ptr2->Y + drawArea.Y - 2;
+
+				// TML(liquid-slopes): Also attempt to draw if the
+				// coordinate is on the edge of a tile.
+				bool isEdge = Main.instance.TilesRenderer.EdgeTiles.TryGetValue(new Point(num3, num4), out var edgeCache);
+
+				if (!ptr2->IsVisible && !isEdge)
 					break;
 
+				// IsEdge is unique because it means there isn't actually a
+				// liquid present in the tile, so a lot of data isn't present.
+				// The below data is mostly irrelevant for edge tiles, so it's
+				// fine for this data to be missing.
+
 				Rectangle sourceRectangle = ptr2->SourceRectangle;
 				if (ptr2->IsSurfaceLiquid)
 					sourceRectangle.Y = 1280;
 				else
 					sourceRectangle.Y += _animationFrame * 80;
 
+				// Opacity is important for edge tiles since we want to
+				// support clean transitions between biomes.
+				var opacity = isEdge ? 1f : ptr2->Opacity;
+
 				Vector2 liquidOffset = ptr2->LiquidOffset;
-				float val = ptr2->Opacity * (isBackgroundDraw ? 1f : 0.75f);
+				float val = opacity * (isBackgroundDraw ? 1f : 0.75f);
 				int num2 = 14;
 				val = Math.Min(1f, val);
-				int num3 = ptr2->X + drawArea.X - 2;
-				int num4 = ptr2->Y + drawArea.Y - 2;
 				Lighting.GetCornerColors(num3, num4, out var vertices);
 				SetShimmerVertexColors(ref vertices, val, num3, num4);
+
+				if (!isEdge) {
-				Main.DrawTileInWater(drawOffset, num3, num4);
+					Main.DrawTileInWater(drawOffset, num3, num4);
-				Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(num3 << 4, num4 << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+					Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(num3 << 4, num4 << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+				}
+				else {
+					var tile = Main.tile[num3, num4];
+					Main.instance.TilesRenderer.DrawPartialLiquid(isBackgroundDraw, tile, new Vector2(num3 << 4, num4 << 4) + drawOffset + edgeCache.LiquidOffset, edgeCache.LiquidFrame, edgeCache.LiquidType, vertices, fadeWater: false);
+				}
+
 				sourceRectangle = ptr2->SourceRectangle;
 				bool flag = sourceRectangle.X != 16 || sourceRectangle.Y % 80 != 48;
-				if (flag || (num3 + num4) % 2 == 0) {
+				if (flag || (num3 + num4) % 2 == 0 && !isEdge) {
 					sourceRectangle.X += 48;
 					sourceRectangle.Y += 80 * GetShimmerFrame(flag, num3, num4);
 					SetShimmerVertexColors_Sparkle(ref vertices, ptr2->Opacity, num3, num4, flag);
