--- src/TerrariaNetCore/Terraria/GameContent/Liquid/LiquidRenderer.cs
+++ src/tModLoader/Terraria/GameContent/Liquid/LiquidRenderer.cs
@@ -10,7 +_,7 @@
 
 public class LiquidRenderer
 {
-	private struct LiquidCache
+	public struct LiquidCache
 	{
 		public float LiquidLevel;
 		public float VisibleLiquidLevel;
@@ -37,7 +_,7 @@
 		public byte VisibleType;
 	}
 
-	private struct LiquidDrawCache
+	public struct LiquidDrawCache
 	{
 		public Rectangle SourceRectangle;
 		public Vector2 LiquidOffset;
@@ -48,7 +_,7 @@
 		public bool HasWall;
 	}
 
-	private struct SpecialLiquidDrawCache
+	public struct SpecialLiquidDrawCache
 	{
 		public int X;
 		public int Y;
@@ -61,16 +_,24 @@
 		public bool HasWall;
 	}
 
+	// TML(liquid-slopes): Introduce new cache for tiles bordering liquids.
+	public struct LiquidEdgeCache
+	{
+		public bool IsVisible;
+		public bool IsHalfBrick;
+		public SlopeType SlopeType;
+	}
+
 	private const int ANIMATION_FRAME_COUNT = 16;
 	private const int CACHE_PADDING = 2;
 	private const int CACHE_PADDING_2 = 4;
-	private static readonly int[] WATERFALL_LENGTH = new int[4] {
+	public static readonly int[] WATERFALL_LENGTH = new int[4] {
 		10,
 		3,
 		2,
 		10
 	};
-	private static readonly float[] DEFAULT_OPACITY = new float[4] {
+	public static readonly float[] DEFAULT_OPACITY = new float[4] {
 		0.6f,
 		0.95f,
 		0.95f,
@@ -86,17 +_,20 @@
 	};
 	public const float MIN_LIQUID_SIZE = 0.25f;
 	public static LiquidRenderer Instance;
-	private readonly Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
+	public Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
 	private LiquidCache[] _cache = new LiquidCache[1];
 	private LiquidDrawCache[] _drawCache = new LiquidDrawCache[1];
 	private SpecialLiquidDrawCache[] _drawCacheForShimmer = new SpecialLiquidDrawCache[1];
+	private LiquidEdgeCache[] _cacheForEdges = new LiquidEdgeCache[1];
 	private int _animationFrame;
 	private Rectangle _drawArea = new Rectangle(0, 0, 1, 1);
 	private readonly UnifiedRandom _random = new UnifiedRandom();
 	private Color[] _waveMask = new Color[1];
 	private float _frameState;
 
+	/*
 	private static Tile[,] Tiles => Main.tile;
+	*/
 
 	public event Action<Color[], Rectangle> WaveFilters;
 
@@ -128,17 +_,20 @@
 		if (_drawCacheForShimmer.Length < drawArea.Width * drawArea.Height + 1)
 			_drawCacheForShimmer = new SpecialLiquidDrawCache[drawArea.Width * drawArea.Height + 1];
 
+		if (_cacheForEdges.Length < drawArea.Width * drawArea.Height + 1)
+			_cacheForEdges = new LiquidEdgeCache[drawArea.Width * drawArea.Height + 1];
+
 		if (_waveMask.Length < drawArea.Width * drawArea.Height)
 			_waveMask = new Color[drawArea.Width * drawArea.Height];
 
-		Tile tile = null;
+		Tile tile = default;
 		fixed (LiquidCache* ptr = &_cache[1]) {
 			LiquidCache* ptr2 = ptr;
 			int num = rectangle.Height * 2 + 2;
 			ptr2 = ptr;
 			for (int i = rectangle.X; i < rectangle.X + rectangle.Width; i++) {
 				for (int j = rectangle.Y; j < rectangle.Y + rectangle.Height; j++) {
-					tile = Tiles[i, j];
+					tile = Main.tile[i, j];
 					if (tile == null)
 						tile = new Tile();
 
