--- src/TerrariaNetCore/Terraria/GameContent/Liquid/LiquidRenderer.cs
+++ src/tModLoader/Terraria/GameContent/Liquid/LiquidRenderer.cs
@@ -1,4 +_,5 @@
 using System;
+using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
@@ -10,7 +_,7 @@
 
 public class LiquidRenderer
 {
-	private struct LiquidCache
+	public struct LiquidCache
 	{
 		public float LiquidLevel;
 		public float VisibleLiquidLevel;
@@ -37,7 +_,7 @@
 		public byte VisibleType;
 	}
 
-	private struct LiquidDrawCache
+	public struct LiquidDrawCache
 	{
 		public Rectangle SourceRectangle;
 		public Vector2 LiquidOffset;
@@ -48,7 +_,7 @@
 		public bool HasWall;
 	}
 
-	private struct SpecialLiquidDrawCache
+	public struct SpecialLiquidDrawCache
 	{
 		public int X;
 		public int Y;
@@ -64,13 +_,13 @@
 	private const int ANIMATION_FRAME_COUNT = 16;
 	private const int CACHE_PADDING = 2;
 	private const int CACHE_PADDING_2 = 4;
-	private static readonly int[] WATERFALL_LENGTH = new int[4] {
+	public static readonly int[] WATERFALL_LENGTH = new int[4] {
 		10,
 		3,
 		2,
 		10
 	};
-	private static readonly float[] DEFAULT_OPACITY = new float[4] {
+	public static readonly float[] DEFAULT_OPACITY = new float[4] {
 		0.6f,
 		0.95f,
 		0.95f,
@@ -86,7 +_,7 @@
 	};
 	public const float MIN_LIQUID_SIZE = 0.25f;
 	public static LiquidRenderer Instance;
-	private readonly Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
+	public Asset<Texture2D>[] _liquidTextures = new Asset<Texture2D>[15];
 	private LiquidCache[] _cache = new LiquidCache[1];
 	private LiquidDrawCache[] _drawCache = new LiquidDrawCache[1];
 	private SpecialLiquidDrawCache[] _drawCacheForShimmer = new SpecialLiquidDrawCache[1];
@@ -96,7 +_,9 @@
 	private Color[] _waveMask = new Color[1];
 	private float _frameState;
 
+	/*
 	private static Tile[,] Tiles => Main.tile;
+	*/
 
 	public event Action<Color[], Rectangle> WaveFilters;
 
@@ -131,19 +_,21 @@
 		if (_waveMask.Length < drawArea.Width * drawArea.Height)
 			_waveMask = new Color[drawArea.Width * drawArea.Height];
 
-		Tile tile = null;
+		Tile tile = default;
 		fixed (LiquidCache* ptr = &_cache[1]) {
 			LiquidCache* ptr2 = ptr;
 			int num = rectangle.Height * 2 + 2;
 			ptr2 = ptr;
 			for (int i = rectangle.X; i < rectangle.X + rectangle.Width; i++) {
 				for (int j = rectangle.Y; j < rectangle.Y + rectangle.Height; j++) {
-					tile = Tiles[i, j];
+					tile = Main.tile[i, j];
 					if (tile == null)
 						tile = new Tile();
 
+					// TML(liquid-slopes): Half-brick rendering is now entirely
+					// handled as a slope/partial liquid, so disable it here.
 					ptr2->LiquidLevel = (float)(int)tile.liquid / 255f;
-					ptr2->IsHalfBrick = tile.halfBrick() && ptr2[-1].HasLiquid && !TileID.Sets.Platforms[tile.type];
+					ptr2->IsHalfBrick = LiquidEdgeRenderer.Active ? false : (tile.halfBrick() && ptr2[-1].HasLiquid && !TileID.Sets.Platforms[tile.type]);
 					ptr2->IsSolid = WorldGen.SolidOrSlopedTile(tile);
 					ptr2->HasLiquid = tile.liquid != 0;
 					ptr2->VisibleLiquidLevel = 0f;
@@ -516,7 +_,7 @@
 			this.WaveFilters(_waveMask, GetCachedDrawArea());
 	}
 
-	public unsafe void DrawNormalLiquids(SpriteBatch spriteBatch, Vector2 drawOffset, int waterStyle, float globalAlpha, bool isBackgroundDraw)
+	public unsafe void DrawNormalLiquids(SpriteBatch spriteBatch, Vector2 drawOffset, int waterStyle, float globalAlpha, bool isBackgroundDraw, bool drawSinglePassLiquids)
 	{
 		Rectangle drawArea = _drawArea;
 		Main.tileBatch.Begin();
@@ -539,9 +_,15 @@
 								num2 = waterStyle;
 								num *= globalAlpha;
 								break;
+							case LiquidID.Lava:
+								num2 = WaterStyleID.Lava;
+								break;
 							case 2:
 								num2 = 11;
 								break;
+							case LiquidID.Shimmer:
+								num2 = 14;
+								break;
 						}
 
 						num = Math.Min(1f, num);
@@ -550,6 +_,7 @@
 						vertices.BottomRightColor *= num;
 						vertices.TopLeftColor *= num;
 						vertices.TopRightColor *= num;
+
 						Main.DrawTileInWater(drawOffset, i, j);
 						Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(i << 4, j << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
 					}
@@ -558,7 +_,46 @@
 				}
 			}
 		}
+		if (LiquidEdgeRenderer.Active) {
+			foreach (var edgeCache in LiquidEdgeRenderer.Edges.Values) {
+				var tile = Main.tile[edgeCache.TileX, edgeCache.TileY];
 
+				float opacity = isBackgroundDraw ? 1f : DEFAULT_OPACITY[edgeCache.LiquidType];
+				int liquidType = edgeCache.LiquidType;
+				switch (liquidType) {
+					case LiquidID.Water:
+						liquidType = waterStyle;
+						opacity *= globalAlpha;
+						break;
+					case LiquidID.Lava:
+						liquidType = WaterStyleID.Lava;
+						break;
+					case LiquidID.Honey:
+						liquidType = WaterStyleID.Honey;
+						;
+						break;
+					case LiquidID.Shimmer:
+						liquidType = 14;
+						break;
+				}
+
+				opacity = Math.Min(1f, opacity);
+				Lighting.GetCornerColors(edgeCache.TileX, edgeCache.TileY, out var vertices);
+				vertices.BottomLeftColor *= opacity;
+				vertices.BottomRightColor *= opacity;
+				vertices.TopLeftColor *= opacity;
+				vertices.TopRightColor *= opacity;
+
+				if (liquidType == 14) {
+					if (!drawSinglePassLiquids)
+						break;
+
+					SetShimmerVertexColors(ref vertices, opacity, edgeCache.TileX, edgeCache.TileY);
+				}
+
+				Main.instance.TilesRenderer.DrawPartialLiquid(isBackgroundDraw, tile, edgeCache.LiquidPosition + drawOffset, edgeCache.LiquidFrame, liquidType, vertices);
+			}
+		}
 		Main.tileBatch.End();
 	}
 
@@ -589,6 +_,7 @@
 				SetShimmerVertexColors(ref vertices, val, num3, num4);
 				Main.DrawTileInWater(drawOffset, num3, num4);
 				Main.tileBatch.Draw(_liquidTextures[num2].Value, new Vector2(num3 << 4, num4 << 4) + drawOffset + liquidOffset, sourceRectangle, vertices, Vector2.Zero, 1f, SpriteEffects.None);
+
 				sourceRectangle = ptr2->SourceRectangle;
 				bool flag = sourceRectangle.X != 16 || sourceRectangle.Y % 80 != 48;
 				if (flag || (num3 + num4) % 2 == 0) {
