--- src/Terraria/Configuration.targets
+++ src/TerrariaNetCore/Configuration.targets
@@ -1,26 +_,15 @@
 <Project>
 	<PropertyGroup>
-		<TargetFramework>net40</TargetFramework>
-		<Configurations>WindowsDebug;WindowsRelease;WindowsServerDebug;WindowsServerRelease;LinuxRelease;LinuxServerRelease;MacRelease;MacServerRelease</Configurations>
-		<PlatformTarget>x86</PlatformTarget>
+		<TargetFramework>net8.0</TargetFramework>
+		<!-- Defines the Minimum patch version that is required to play. -->
+		<!-- <RuntimeFrameworkVersion>6.0.14</RuntimeFrameworkVersion> -->
+		<!-- Eliminates the Refs folder from output-->
+		<ProduceReferenceAssembly>false</ProduceReferenceAssembly> 
+		<Configurations>Debug;Release</Configurations>
+		<PlatformTarget>AnyCPU</PlatformTarget>
 		<AllowUnsafeBlocks>true</AllowUnsafeBlocks>
 		<Optimize>true</Optimize>
-		<LangVersion>latest</LangVersion>
-	</PropertyGroup>
-	<PropertyGroup Condition="$(Configuration.Contains('Windows'))">
-		<DefineConstants>$(DefineConstants);WINDOWS;XNA</DefineConstants>
+		<DefineConstants>$(DefineConstants);FNA;NETCORE</DefineConstants>
-	</PropertyGroup>
-	<PropertyGroup Condition="$(Configuration.Contains('Linux'))">
-		<DefineConstants>$(DefineConstants);LINUX;FNA;MONO</DefineConstants>
-	</PropertyGroup>
-	<PropertyGroup Condition="$(Configuration.Contains('Mac'))">
-		<DefineConstants>$(DefineConstants);MAC;FNA;MONO</DefineConstants>
-	</PropertyGroup>
-	<PropertyGroup Condition="$(Configuration.Contains('Server'))">
-		<DefineConstants>$(DefineConstants);SERVER</DefineConstants>
-	</PropertyGroup>
-	<PropertyGroup Condition="!$(Configuration.Contains('Server'))">
-		<DefineConstants>$(DefineConstants);CLIENT</DefineConstants>
 	</PropertyGroup>
 	<PropertyGroup Condition="$(Configuration.Contains('Debug'))">
 		<Optimize>false</Optimize>
