--- src/Terraria/Configuration.targets
+++ src/TerrariaNetCore/Configuration.targets
@@ -2,12 +_,16 @@
 
 	<!-- Build properties -->
 	<PropertyGroup>
-		<TargetFramework>net40</TargetFramework>
-		<Configurations>WindowsDebug;WindowsRelease;WindowsServerDebug;WindowsServerRelease;LinuxRelease;LinuxServerRelease;MacRelease;MacServerRelease</Configurations>
-		<PlatformTarget>x86</PlatformTarget>
+		<TargetFramework>net6.0</TargetFramework>
+		<!-- Defines the Minimum patch version that is required to play. -->
+		<RuntimeFrameworkVersion>6.0.6</RuntimeFrameworkVersion>
+		<!-- Eliminates the Refs folder from output.-->
+		<ProduceReferenceAssembly>false</ProduceReferenceAssembly> 
+		<Configurations>Debug;Release</Configurations>
+		<PlatformTarget>AnyCPU</PlatformTarget>
 		<AllowUnsafeBlocks>true</AllowUnsafeBlocks>
 		<Optimize>true</Optimize>
-		<LangVersion>latest</LangVersion>
+		<DefineConstants>$(DefineConstants);FNA;NETCORE</DefineConstants>
 	</PropertyGroup>
 
 	<!-- Debug properties -->
@@ -16,19 +_,9 @@
 		<DefineConstants>$(DefineConstants);DEBUG</DefineConstants>
 	</PropertyGroup>
 
-	<!-- Conditional constants -->
+	<!-- Task properties -->
 	<PropertyGroup>
-		<!-- Windows -->
-		<DefineConstants Condition="$(Configuration.Contains('Windows'))">$(DefineConstants);WINDOWS;XNA</DefineConstants>
-		<!-- Linux -->
-		<DefineConstants Condition="$(Configuration.Contains('Linux'))">$(DefineConstants);LINUX;FNA;MONO</DefineConstants>
-		<!-- Mac -->
-		<DefineConstants Condition="$(Configuration.Contains('Mac'))">$(DefineConstants);MAC;FNA;MONO</DefineConstants>
-
-		<!-- Server -->
-		<DefineConstants Condition="$(Configuration.Contains('Server'))">$(DefineConstants);SERVER</DefineConstants>
-		<!-- Client -->
-		<DefineConstants Condition="!$(Configuration.Contains('Server'))">$(DefineConstants);CLIENT</DefineConstants>
+		<DependencyOrganizerAssemblyFile>$(MSBuildThisFileDirectory)\..\..\BuildTools\DependencyOrganizer\Libraries\DependencyOrganizer.dll</DependencyOrganizerAssemblyFile>
 	</PropertyGroup>
 
 </Project>
