--- src/Terraria\Terraria.GameContent.Generation\TrackGenerator.cs
+++ src/tModLoader\Terraria.GameContent.Generation\TrackGenerator.cs
@@ -5,7 +_,7 @@
 
 namespace Terraria.GameContent.Generation
 {
-	internal class TrackGenerator
+	public class TrackGenerator
 	{
 		private struct TrackHistory
 		{
@@ -95,7 +_,7 @@
 			{
 				return false;
 			}
-			byte wall = Main.tile[x, y].wall;
+			ushort wall = Main.tile[x, y].wall;
 			for (int i = 0; i < TrackGenerator.INVALID_WALLS.Length; i++)
 			{
 				if (wall == TrackGenerator.INVALID_WALLS[i])

