--- src/TerrariaNetCore/Terraria/GameContent/Personalities/AShoppingBiome.cs
+++ src/tModLoader/Terraria/GameContent/Personalities/AShoppingBiome.cs
@@ -1,12 +_,30 @@
+using Terraria.ModLoader;
+
 namespace Terraria.GameContent.Personalities
 {
-	public abstract class AShoppingBiome
+	// Interface added by TML.
+	public interface IShoppingBiome
+	{
+		string NameKey { get; }
+
+		bool IsInBiome(Player player);
+	}
+
+	//TML: Added ILoadable
+	public abstract class AShoppingBiome : IShoppingBiome, ILoadable
 	{
 		public string NameKey {
 			get;
 			protected set;
 		}
 
+		// Added by TML to prevent unintended usages.
+		internal AShoppingBiome() { }
+
 		public abstract bool IsInBiome(Player player);
+
+		void ILoadable.Load(Mod mod) { }
+
+		void ILoadable.Unload() { }
 	}
 }
