--- src/TerrariaNetCore/Terraria/GameContent/Personalities/AShoppingBiome.cs
+++ src/tModLoader/Terraria/GameContent/Personalities/AShoppingBiome.cs
@@ -1,8 +_,28 @@
+using Terraria.ModLoader;
+
 namespace Terraria.GameContent.Personalities;
 
+// Interface added by TML.
+public interface IShoppingBiome
+{
+	string NameKey { get; }
+
+	bool IsInBiome(Player player);
+}
+
+/*
 public abstract class AShoppingBiome
+*/
+public abstract class AShoppingBiome : IShoppingBiome, ILoadable
 {
 	public string NameKey { get; protected set; }
 
+	// Added by TML to prevent unintended usages.
+	internal AShoppingBiome() { }
+
 	public abstract bool IsInBiome(Player player);
+
+	void ILoadable.Load(Mod mod) { }
+
+	void ILoadable.Unload() { }
 }
