--- src/decompiled\Terraria.Social\SocialAPI.cs
+++ src/merged\Terraria.Social\SocialAPI.cs
@@ -34,7 +_,14 @@
 			if (!mode.HasValue)
 			{
 				mode = new SocialMode?(SocialMode.None);
+#if CLIENT
 				mode = new SocialMode?(SocialMode.Steam);
+#else
+				if (Program.LaunchParameters.ContainsKey("-steam"))
+				{
+					mode = new SocialMode?(SocialMode.Steam);
+				}
+#endif
 			}
 			SocialAPI._mode = mode.Value;
 			SocialAPI._modules = new List<ISocialModule>();
@@ -60,7 +_,11 @@
 
 		private static T LoadModule<T>() where T : ISocialModule, new()
 		{
+#if MAC
+			T t = Activator.CreateInstance<T>();
+#else
 			T t = (default(T) == null) ? Activator.CreateInstance<T>() : default(T);
+#endif
 			SocialAPI._modules.Add(t);
 			return t;
 		}
@@ -78,7 +_,11 @@
 			SocialAPI.Achievements = SocialAPI.LoadModule<Terraria.Social.Steam.AchievementsSocialModule>();
 			SocialAPI.Cloud = SocialAPI.LoadModule<Terraria.Social.Steam.CloudSocialModule>();
 			SocialAPI.Overlay = SocialAPI.LoadModule<Terraria.Social.Steam.OverlaySocialModule>();
+#if CLIENT
 			SocialAPI.Network = SocialAPI.LoadModule<NetClientSocialModule>();
+#else
+			SocialAPI.Network = SocialAPI.LoadModule<NetServerSocialModule>();
+#endif
 		}
 	}
 }

