--- src/decompiled/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
@@ -6,9 +_,23 @@
 {
 	public class CaptureManager : IDisposable
 	{
+		// 'Terraria' patchstep: Client initialization has been moved to LoadContent to prevent FNA deadlocks
+		
+		/*
 		public static CaptureManager Instance = new CaptureManager();
+		*/
+		private static CaptureManager _instance;
+		
 		private CaptureInterface _interface;
 		private CaptureCamera _camera;
+		public static CaptureManager Instance {
+			get => Main.dedServ ? new CaptureManager() : _instance;
+			set => _instance = value;
+		}
+
+		/*
+		*/
+		public bool IsCapturing => !Main.dedServ && _camera.IsCapturing;
 
 		public bool IsCapturing {
 			get {
