--- src/Terraria/Terraria/Graphics/WindowStateController.cs
+++ src/TerrariaNetCore/Terraria/Graphics/WindowStateController.cs
@@ -6,6 +_,11 @@
 {
 	public class WindowStateController
 	{
+#if NETCORE
+		public bool CanMoveWindowAcrossScreens => false;
+		public string ScreenDeviceName => "";
+		public void TryMovingToScreen(string screenDeviceName) {}
+#else
 		public bool CanMoveWindowAcrossScreens => Platform.IsWindows;
 
 		public string ScreenDeviceName {
@@ -54,5 +_,6 @@
 
 			return false;
 		}
+#endif
 	}
 }
