--- src/decompiled/Terraria/Utilities/CrashDump.cs
+++ src/Terraria/Terraria/Utilities/CrashDump.cs
@@ -55,7 +_,8 @@
 
 		private static string CreateDumpName() {
 			DateTime dateTime = DateTime.Now.ToLocalTime();
+			string Name = Main.dedServ ? "TerrariaServer" : "Terraria";
-			return string.Format("{0}_{1}_{2}_{3}.dmp", "Terraria", Main.versionNumber, dateTime.ToString("MM-dd-yy_HH-mm-ss-ffff", CultureInfo.InvariantCulture), Thread.CurrentThread.ManagedThreadId);
+			return string.Format("{0}_{1}_{2}_{3}.dmp", Name, Main.versionNumber, dateTime.ToString("MM-dd-yy_HH-mm-ss-ffff", CultureInfo.InvariantCulture), Thread.CurrentThread.ManagedThreadId);
 		}
 
 		private static bool Write(Options options, ExceptionInfo exceptionInfo, string outputDirectory) {
