--- src/decompiled/Terraria/Utilities/FileOperationAPIWrapper.cs
+++ src/Terraria/Terraria/Utilities/FileOperationAPIWrapper.cs
@@ -1,3 +_,4 @@
+#if WINDOWS
 using System;
 using System.Runtime.InteropServices;
 
@@ -78,3 +_,4 @@
 		private static bool DeleteCompletelySilent(string path) => DeleteFile(path, FileOperationFlags.FOF_SILENT | FileOperationFlags.FOF_NOCONFIRMATION | FileOperationFlags.FOF_NOERRORUI);
 	}
 }
+#endif
