--- src/decompiled/Terraria/Utilities/PlatformUtilities.cs
+++ src/Terraria/Terraria/Utilities/PlatformUtilities.cs
@@ -1,11 +_,41 @@
 using System;
+using System.IO;
+#if FNA
+using SDL3;
+#endif
 
 namespace Terraria.Utilities;
 
 public static class PlatformUtilities
 {
+#if FNA
+	public unsafe static void SavePng(string path, int width, int height, byte[] data)
+	{
+		if (width * height * 4 > data.Length)
+			throw new IndexOutOfRangeException();
+
+		fixed (byte* ptr = data) {
+			IntPtr intPtr = SDL.SDL_CreateSurfaceFrom(width, height, SDL.SDL_PixelFormat.SDL_PIXELFORMAT_ABGR8888, (IntPtr)ptr, width * 4);
+			if (intPtr == IntPtr.Zero)
+				throw new IOException("Failed to create SDL surface: " + SDL.SDL_GetError());
+
+			try {
+				IntPtr intPtr2 = SDL.SDL_IOFromFile(path, "wb");
+				if (intPtr2 == IntPtr.Zero)
+					throw new IOException("Failed to open file for writing: " + SDL.SDL_GetError());
+
+				if (!SDL.SDL_SavePNG_IO(intPtr, intPtr2, true))
+					throw new IOException("Failed to save PNG: " + SDL.SDL_GetError());
+			}
+			finally {
+				SDL.SDL_DestroySurface(intPtr);
+			}
+		}
+	}
+#else
 	public static void SavePng(string path, int width, int height, byte[] data)
 	{
 		throw new NotSupportedException("Use Bitmap to save png images on windows");
 	}
+#endif
 }
