--- src/Terraria\Terraria.DataStructures\PlayerDeathReason.cs
+++ src/tModLoader\Terraria.DataStructures\PlayerDeathReason.cs
@@ -6,14 +_,14 @@
 {
 	public class PlayerDeathReason
 	{
-		private int SourcePlayerIndex = -1;
-		private int SourceNPCIndex = -1;
-		private int SourceProjectileIndex = -1;
-		private int SourceOtherIndex = -1;
-		private int SourceProjectileType;
-		private int SourceItemType;
-		private int SourceItemPrefix;
-		private string SourceCustomReason;
+		public int SourcePlayerIndex = -1;
+		public int SourceNPCIndex = -1;
+		public int SourceProjectileIndex = -1;
+		public int SourceOtherIndex = -1;
+		public int SourceProjectileType;
+		public int SourceItemType;
+		public int SourceItemPrefix;
+		public string SourceCustomReason;
 
 		public static PlayerDeathReason LegacyEmpty()
 		{

