--- src/Terraria/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs
+++ src/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.Bestiary
 {
@@ -268,16 +_,17 @@
 			public const int ItemSpawn = 61;
 		}
 
-		private BestiaryDatabase _currentDatabase;
+		private static BestiaryDatabase _currentDatabase;
 
-		private BestiaryEntry FindEntryByNPCID(int npcNetId) => _currentDatabase.FindEntryByNPCID(npcNetId);
+		public static BestiaryEntry FindEntryByNPCID(int npcNetId) => _currentDatabase.FindEntryByNPCID(npcNetId);
+
 		private BestiaryEntry Register(BestiaryEntry entry) => _currentDatabase.Register(entry);
 		private IBestiaryEntryFilter Register(IBestiaryEntryFilter filter) => _currentDatabase.Register(filter);
 
 		public void Populate(BestiaryDatabase database) {
 			_currentDatabase = database;
-			AddEmptyEntries_CrittersAndEnemies_Automated();
-			AddTownNPCs_Manual();
+			AddEmptyEntries_Automated();
+			// AddTownNPCs_Manual();
 			AddNPCBiomeRelationships_Automated();
 			AddNPCBiomeRelationships_Manual();
 			AddNPCBiomeRelationships_AddDecorations_Automated();
@@ -542,53 +_,58 @@
 			}
 		}
 
-		private void AddTownNPCs_Manual() {
-			Register(BestiaryEntry.TownNPC(22));
-			Register(BestiaryEntry.TownNPC(17));
-			Register(BestiaryEntry.TownNPC(18));
-			Register(BestiaryEntry.TownNPC(19));
-			Register(BestiaryEntry.TownNPC(20));
-			Register(BestiaryEntry.TownNPC(37));
-			Register(BestiaryEntry.TownNPC(54));
-			Register(BestiaryEntry.TownNPC(38));
-			Register(BestiaryEntry.TownNPC(107));
-			Register(BestiaryEntry.TownNPC(108));
-			Register(BestiaryEntry.TownNPC(124));
-			Register(BestiaryEntry.TownNPC(142));
-			Register(BestiaryEntry.TownNPC(160));
-			Register(BestiaryEntry.TownNPC(178));
-			Register(BestiaryEntry.TownNPC(207));
-			Register(BestiaryEntry.TownNPC(208));
-			Register(BestiaryEntry.TownNPC(209));
-			Register(BestiaryEntry.TownNPC(227));
-			Register(BestiaryEntry.TownNPC(228));
-			Register(BestiaryEntry.TownNPC(229));
-			Register(BestiaryEntry.TownNPC(353));
-			Register(BestiaryEntry.TownNPC(369));
-			Register(BestiaryEntry.TownNPC(441));
-			Register(BestiaryEntry.TownNPC(550));
-			Register(BestiaryEntry.TownNPC(588));
-			Register(BestiaryEntry.TownNPC(368));
-			Register(BestiaryEntry.TownNPC(453));
-			Register(BestiaryEntry.TownNPC(633));
-			Register(BestiaryEntry.TownNPC(663));
-			Register(BestiaryEntry.TownNPC(638));
-			Register(BestiaryEntry.TownNPC(637));
-			Register(BestiaryEntry.TownNPC(656));
-		}
+		//private void AddTownNPCs_Manual() {
+		//	Register(BestiaryEntry.TownNPC(22));
+		//	Register(BestiaryEntry.TownNPC(17));
+		//	Register(BestiaryEntry.TownNPC(18));
+		//	Register(BestiaryEntry.TownNPC(19));
+		//	Register(BestiaryEntry.TownNPC(20));
+		//	Register(BestiaryEntry.TownNPC(37));
+		//	Register(BestiaryEntry.TownNPC(54));
+		//	Register(BestiaryEntry.TownNPC(38));
+		//	Register(BestiaryEntry.TownNPC(107));
+		//	Register(BestiaryEntry.TownNPC(108));
+		//	Register(BestiaryEntry.TownNPC(124));
+		//	Register(BestiaryEntry.TownNPC(142));
+		//	Register(BestiaryEntry.TownNPC(160));
+		//	Register(BestiaryEntry.TownNPC(178));
+		//	Register(BestiaryEntry.TownNPC(207));
+		//	Register(BestiaryEntry.TownNPC(208));
+		//	Register(BestiaryEntry.TownNPC(209));
+		//	Register(BestiaryEntry.TownNPC(227));
+		//	Register(BestiaryEntry.TownNPC(228));
+		//	Register(BestiaryEntry.TownNPC(229));
+		//	Register(BestiaryEntry.TownNPC(353));
+		//	Register(BestiaryEntry.TownNPC(369));
+		//	Register(BestiaryEntry.TownNPC(441));
+		//	Register(BestiaryEntry.TownNPC(550));
+		//	Register(BestiaryEntry.TownNPC(588));
+		//	Register(BestiaryEntry.TownNPC(368));
+		//	Register(BestiaryEntry.TownNPC(453));
+		//	Register(BestiaryEntry.TownNPC(633));
+		//	Register(BestiaryEntry.TownNPC(656));
+		//	Register(BestiaryEntry.TownNPC(638));
+		//	Register(BestiaryEntry.TownNPC(637));
+		//	Register(BestiaryEntry.TownNPC(656));
+		//}
 
 		private void AddMultiEntryNPCS_Manual() {
 			Register(BestiaryEntry.Enemy(85)).Icon = new UnlockableNPCEntryIcon(85, 0f, 0f, 0f, 3f);
 		}
 
-		private void AddEmptyEntries_CrittersAndEnemies_Automated() {
+		private void AddEmptyEntries_Automated() {
+			BestiaryEntry bestiaryEntry;
 			HashSet<int> exclusions = GetExclusions();
 			foreach (KeyValuePair<int, NPC> item in ContentSamples.NpcsByNetId) {
-				if (!exclusions.Contains(item.Key) && !item.Value.isLikeATownNPC) {
+				if (!exclusions.Contains(item.Key)) {
 					if (item.Value.CountsAsACritter)
-						Register(BestiaryEntry.Critter(item.Key));
+						bestiaryEntry = Register(BestiaryEntry.Critter(item.Key));
+					else if (item.Value.isLikeATownNPC)
+						bestiaryEntry = Register(BestiaryEntry.TownNPC(item.Key));
 					else
-						Register(BestiaryEntry.Enemy(item.Key));
+						bestiaryEntry = Register(BestiaryEntry.Enemy(item.Key));
+
+					NPCLoader.SetBestiary(item.Value, _currentDatabase, bestiaryEntry);
 				}
 			}
 		}
