--- src/TerrariaNetCore/Terraria/GameContent/Bestiary/BestiaryDatabase.cs
+++ src/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabase.cs
@@ -1,10 +_,11 @@
 using System.Collections.Generic;
 using Terraria.GameContent.ItemDropRules;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.Bestiary;
 
-public class BestiaryDatabase
+public partial class BestiaryDatabase
 {
 	public delegate void BestiaryEntriesPass(BestiaryEntry entry);
 
@@ -28,6 +_,14 @@
 				_byNpcId[nPCNetIdBestiaryInfoElement.NetId] = entry;
 		}
 
+		Mod mod = ID.ContentSamples.NpcsByNetId[((NPCNetIdBestiaryInfoElement)entry.Info[0]).NetId].ModNPC?.Mod;
+		if (mod == null)
+			_vanillaEntries.Add(entry);
+		else if (_byMod.ContainsKey(mod))
+			_byMod[mod].Add(entry);
+		else
+			_byMod.Add(mod, new List<BestiaryEntry> { entry });
+
 		return entry;
 	}
 
@@ -54,7 +_,7 @@
 
 	public void Merge(ItemDropDatabase dropsDatabase)
 	{
-		for (int i = -65; i < NPCID.Count; i++) {
+		for (int i = -65; i < NPCLoader.NPCCount; i++) {
 			ExtractDropsForNPC(dropsDatabase, i);
 		}
 	}
