--- src/TerrariaNetCore/Terraria/GameContent/Bestiary/NPCKillsTracker.cs
+++ src/tModLoader/Terraria/GameContent/Bestiary/NPCKillsTracker.cs
@@ -1,5 +_,6 @@
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Terraria.GameContent.NetModules;
 using Terraria.ID;
 using Terraria.Net;
@@ -52,8 +_,12 @@
 	public void Save(BinaryWriter writer)
 	{
 		lock (_killCountsByNpcId) {
-			writer.Write(_killCountsByNpcId.Count);
-			foreach (KeyValuePair<string, int> item in _killCountsByNpcId) {
+			var vanillaOnly = _killCountsByNpcId
+				.Where(pair => ContentSamples.NpcNetIdsByPersistentIds.TryGetValue(pair.Key, out int netId) && netId < NPCID.Count);
+
+			writer.Write(vanillaOnly.Count());
+
+			foreach (KeyValuePair<string, int> item in vanillaOnly) {
 				writer.Write(item.Key);
 				writer.Write(item.Value);
 			}
