--- src/TerrariaNetCore/Terraria/GameContent/Bestiary/NPCWasNearPlayerTracker.cs
+++ src/tModLoader/Terraria/GameContent/Bestiary/NPCWasNearPlayerTracker.cs
@@ -1,5 +_,6 @@
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Terraria.GameContent.NetModules;
 using Terraria.ID;
@@ -55,8 +_,12 @@
 	public void Save(BinaryWriter writer)
 	{
 		lock (_entryCreationLock) {
-			writer.Write(_wasNearPlayer.Count);
-			foreach (string item in _wasNearPlayer) {
+			var vanillaOnly = _wasNearPlayer
+				.Where(persistentId => ContentSamples.NpcNetIdsByPersistentIds.TryGetValue(persistentId, out int netId) && netId < NPCID.Count);
+
+			writer.Write(vanillaOnly.Count());
+
+			foreach (string item in vanillaOnly) {
 				writer.Write(item);
 			}
 		}
