--- src/TerrariaNetCore/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs
+++ src/tModLoader/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs
@@ -3,6 +_,7 @@
 using ReLogic.Content;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.Bestiary;
 
@@ -60,7 +_,7 @@
 				positionOffsetCache.Y = value.PortraitPositionYOverride.Value;
 
 			if (value.CustomTexturePath != null)
-				asset = Main.Assets.Request<Texture2D>(value.CustomTexturePath);
+				asset = ModContent.Request<Texture2D>(value.CustomTexturePath);
 
 			if (asset != null && asset.IsLoaded)
 				_customTexture = asset;
@@ -171,7 +_,7 @@
 			return;
 		}
 
-		if (_npcCache.townNPC && TownNPCProfiles.Instance.GetProfile(_npcCache.type, out var profile))
+		if (_npcCache.townNPC && TownNPCProfiles.Instance.GetProfile(_npcCache, out var profile))
 			TextureAssets.Npc[_npcCache.type] = profile.GetTextureNPCShouldUse(_npcCache);
 
 		Main.instance.DrawNPCDirect(spriteBatch, _npcCache, _npcCache.behindTiles, Vector2.Zero);
