--- src/TerrariaNetCore/Terraria/Audio/LegacyAudioSystem.cs
+++ src/Unified/Terraria/Audio/LegacyAudioSystem.cs
@@ -102,13 +_,14 @@
 
 	public LegacyAudioSystem()
 	{
-		Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
+		var contentDir = Main.instance.Content.RootDirectory;
+		Engine = new AudioEngine(Path.Combine(contentDir, "TerrariaMusic.xgs"));
 #if FNA
 		Console.WriteLine("SDL Audio Driver: " + SDL3.SDL.SDL_GetCurrentAudioDriver());
 #endif
-		SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
+		SoundBank = new SoundBank(Engine, Path.Combine(contentDir, "Sound Bank.xsb"));
 		Engine.Update();
-		WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512); //TODO, investigate history of windows looping errors with streaming wavebank from disk in FNA in Windows.
+		WaveBank = new WaveBank(Engine, Path.Combine(contentDir, "Wave Bank.xwb"), 0, 512); //TODO, investigate history of windows looping errors with streaming wavebank from disk in FNA in Windows.
 		Engine.Update();
 		AudioTracks = new IAudioTrack[Main.maxMusic];
 		TrackNamesByIndex = new Dictionary<int, string>();
