--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/TerrariaNetCore/Terraria/Audio/LegacyAudioSystem.cs
@@ -101,11 +_,7 @@
 			Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
 			SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
 			Engine.Update();
-#if XNA
-			WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
-#else
-			WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb");
-#endif
+			WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb"); //, 0, 512); // Removing those 2 parameters fixes music not looping on FNA.
 			Engine.Update();
 			AudioTracks = new IAudioTrack[91];
 			TrackNamesByIndex = new Dictionary<int, string>();
