--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/TerrariaNetCore/Terraria/Audio/LegacyAudioSystem.cs
@@ -103,7 +_,7 @@
 		Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
 		SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
 		Engine.Update();
-		WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
+		WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512); //TODO, investigate history of windows looping errors with streaming wavebank from disk in FNA in Windows.
 		Engine.Update();
 		AudioTracks = new IAudioTrack[Main.maxMusic];
 		TrackNamesByIndex = new Dictionary<int, string>();
