--- src/decompiled/Terraria/Localization/LanguageManager.cs
+++ src/Terraria/Terraria/Localization/LanguageManager.cs
@@ -60,7 +_,7 @@
 
 		private string[] GetLanguageFilesForCulture(GameCulture culture) {
 			Assembly.GetExecutingAssembly();
-			return Array.FindAll(typeof(Program).Assembly.GetManifestResourceNames(), (string element) => element.StartsWith("Terraria.Localization.Content." + culture.CultureInfo.Name) && element.EndsWith(".json"));
+			return Array.FindAll(typeof(Program).Assembly.GetManifestResourceNames(), (string element) => element.StartsWith("Terraria.Localization.Content." + culture.CultureInfo.Name.Replace('-', '_')) && element.EndsWith(".json"));
 		}
 
 		public void SetLanguage(GameCulture culture) {
@@ -92,7 +_,7 @@
 				try {
 					string text2 = Utils.ReadEmbeddedResource(text);
 					if (text2 == null || text2.Length < 2)
-						throw new FileFormatException();
+						throw new FormatException();
 
 					LoadLanguageFromFileText(text2);
 				}
