--- src/Terraria/Terraria/NATUPNPLib/IStaticPortMapping.cs
+++ src/TerrariaNetCore/Terraria/NATUPNPLib/IStaticPortMapping.cs
@@ -1,9 +_,16 @@
-#if SERVER
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace NATUPNPLib;
 
+#if NETCORE
+public interface IStaticPortMapping
+{
+	public abstract int InternalPort { get; }
+	public abstract string Protocol { get; }
+	public abstract string InternalClient { get; }
+}
+#else
 [ComImport]
 [CompilerGenerated]
 [Guid("6F10711F-729B-41E5-93B8-F21D0F818DF1")]
