--- src/Terraria/Terraria/NATUPNPLib/IUPnPNAT.cs
+++ src/TerrariaNetCore/Terraria/NATUPNPLib/IUPnPNAT.cs
@@ -1,9 +_,14 @@
-#if SERVER
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace NATUPNPLib;
 
+#if NETCORE
+public interface IUPnPNAT
+{
+	public abstract IStaticPortMappingCollection StaticPortMappingCollection { get; }
+}
+#else
 [ComImport]
 [CompilerGenerated]
 [Guid("B171C812-CC76-485A-94D8-B6B3A2794E99")]
