--- src/Terraria\Terraria.GameContent.Events\DD2Event.cs
+++ src/tModLoader\Terraria.GameContent.Events\DD2Event.cs
@@ -32,6 +_,7 @@
 		private static int _crystalsDropping_toDrop = 0;
 		private static int _crystalsDropping_alreadyDropped = 0;
 		private static int _timeLeftUntilSpawningBegins = 0;
+		public static string TimeLeftBetweenWavesTimer = string.Format("{0}", TimeLeftBetweenWaves / 60);
 
 		public static bool ReadyToFindBartender
 		{

