--- src/Terraria\Terraria.GameContent.Events\Sandstorm.cs
+++ src/tModLoader\Terraria.GameContent.Events\Sandstorm.cs
@@ -183,8 +_,9 @@
 								vector += Utils.RandomVector2(Main.rand, -10f, 10f) + dust.velocity * -1.1f;
 								num16 = (int)vector.X / 16;
 								num17 = (int)vector.Y / 16;
-								if (WorldGen.InWorld(num16, num17, 10) && Main.tile[num16, num17] != null)
+								if (WorldGen.InWorld(num16, num17, 10) && Main.tile[num16, num17] != null) {
-									byte arg_5F6_0 = Main.tile[num16, num17].wall;
+									//byte arg_5F6_0 = Main.tile[num16, num17].wall;
+								}
 							}
 						}
 
