--- src/Terraria/Terraria.GameContent.Skies/MartianSky.cs
+++ src/tModLoader/Terraria.GameContent.Skies/MartianSky.cs
@@ -250,7 +_,7 @@
 			}
 		}
 
-		internal override void Activate(Vector2 position, params object[] args) {
+		public override void Activate(Vector2 position, params object[] args) {
 			_activeUfos = 0;
 			GenerateUfos();
 			Array.Sort(_ufos, (Ufo ufo1, Ufo ufo2) => ufo2.Depth.CompareTo(ufo1.Depth));
@@ -258,7 +_,7 @@
 			_leaving = false;
 		}
 
-		internal override void Deactivate(params object[] args) {
+		public override void Deactivate(params object[] args) {
 			_leaving = true;
 		}
 
