--- src/Terraria\Terraria.GameContent.Skies\MartianSky.cs
+++ src/tModLoader\Terraria.GameContent.Skies\MartianSky.cs
@@ -300,7 +_,7 @@
 			}
 		}
 
-		internal override void Activate(Vector2 position, params object[] args)
+		public override void Activate(Vector2 position, params object[] args)
 		{
 			this._activeUfos = 0;
 			this.GenerateUfos();
@@ -309,7 +_,7 @@
 			this._leaving = false;
 		}
 
-		internal override void Deactivate(params object[] args)
+		public override void Deactivate(params object[] args)
 		{
 			this._leaving = true;
 		}

