--- src/Terraria\Terraria.GameContent.Skies\NebulaSky.cs
+++ src/tModLoader\Terraria.GameContent.Skies\NebulaSky.cs
@@ -95,7 +_,7 @@
 
 		public override float GetCloudAlpha() => (1f - _fadeOpacity) * 0.3f + 0.7f;
 
-		internal override void Activate(Vector2 position, params object[] args) {
+		public override void Activate(Vector2 position, params object[] args) {
 			_fadeOpacity = 0.002f;
 			_isActive = true;
 			_pillars = new LightPillar[40];
@@ -110,7 +_,7 @@
 
 		private int SortMethod(LightPillar pillar1, LightPillar pillar2) => pillar2.Depth.CompareTo(pillar1.Depth);
 
-		internal override void Deactivate(params object[] args) {
+		public override void Deactivate(params object[] args) {
 			_isActive = false;
 		}
 
