--- src/Terraria\Terraria.GameContent.Skies\NebulaSky.cs
+++ src/tModLoader\Terraria.GameContent.Skies\NebulaSky.cs
@@ -109,7 +_,7 @@
 			return (1f - this._fadeOpacity) * 0.3f + 0.7f;
 		}
 
-		internal override void Activate(Vector2 position, params object[] args)
+		public override void Activate(Vector2 position, params object[] args)
 		{
 			this._fadeOpacity = 0.002f;
 			this._isActive = true;
@@ -128,7 +_,7 @@
 			return pillar2.Depth.CompareTo(pillar1.Depth);
 		}
 
-		internal override void Deactivate(params object[] args)
+		public override void Deactivate(params object[] args)
 		{
 			this._isActive = false;
 		}

