--- src/Terraria\Terraria.GameContent.Skies\PartySky.cs
+++ src/tModLoader\Terraria.GameContent.Skies\PartySky.cs
@@ -243,7 +_,7 @@
 			}
 		}
 
-		internal override void Activate(Vector2 position, params object[] args)
+		public override void Activate(Vector2 position, params object[] args)
 		{
 			if (this._active)
 			{
@@ -256,7 +_,7 @@
 			this._leaving = false;
 		}
 
-		internal override void Deactivate(params object[] args)
+		public override void Deactivate(params object[] args)
 		{
 			this._leaving = true;
 		}

