--- src/Terraria\Terraria.GameContent.Skies\SandstormSky.cs
+++ src/tModLoader\Terraria.GameContent.Skies\SandstormSky.cs
@@ -54,13 +_,13 @@
 			}
 		}
 
-		internal override void Activate(Vector2 position, params object[] args)
+		public override void Activate(Vector2 position, params object[] args)
 		{
 			this._isActive = true;
 			this._isLeaving = false;
 		}
 
-		internal override void Deactivate(params object[] args)
+		public override void Deactivate(params object[] args)
 		{
 			this._isLeaving = true;
 		}

