--- src/Terraria\Terraria.Graphics.Shaders\ArmorShaderDataSet.cs
+++ src/tModLoader\Terraria.Graphics.Shaders\ArmorShaderDataSet.cs
@@ -6,12 +_,16 @@
 {
 	public class ArmorShaderDataSet
 	{
-		protected List<ArmorShaderData> _shaderData = new List<ArmorShaderData>();
-		protected Dictionary<int, int> _shaderLookupDictionary = new Dictionary<int, int>();
-		protected int _shaderDataCount;
+		internal List<ArmorShaderData> _shaderData = new List<ArmorShaderData>();
+		internal Dictionary<int, int> _shaderLookupDictionary = new Dictionary<int, int>();
+		internal int _shaderDataCount;
 
 		public T BindShader<T>(int itemId, T shaderData) where T : ArmorShaderData
 		{
+			if (this._shaderDataCount == short.MaxValue)
+			{
+				throw new Exception("Too many shaders bound.");
+			}
 			this._shaderLookupDictionary[itemId] = ++this._shaderDataCount;
 			this._shaderData.Add(shaderData);
 			return shaderData;

