--- src/Terraria\Terraria.Graphics.Shaders\ArmorShaderDataSet.cs
+++ src/tModLoader\Terraria.Graphics.Shaders\ArmorShaderDataSet.cs
@@ -15,6 +_,16 @@
 			this._shaderLookupDictionary[itemId] = ++this._shaderDataCount;
 			this._shaderData.Add(shaderData);
 			return shaderData;
+		}
+
+		public void Remove(int itemId)
+		{
+			int shaderId = GetShaderIdFromItemId(itemId);
+			if (shaderId > 0)
+			{
+				_shaderLookupDictionary.Remove(itemId);
+				_shaderData[shaderId - 1] = null;
+			}
 		}
 
 		public void Apply(int shaderId, Entity entity, DrawData? drawData = null)

