--- src/Terraria\Terraria.Graphics.Shaders\HairShaderDataSet.cs
+++ src/tModLoader\Terraria.Graphics.Shaders\HairShaderDataSet.cs
@@ -7,15 +_,17 @@
 {
 	public class HairShaderDataSet
 	{
-		protected List<HairShaderData> _shaderData = new List<HairShaderData>();
+		internal List<HairShaderData> _shaderData = new List<HairShaderData>();
-		protected Dictionary<int, short> _shaderLookupDictionary = new Dictionary<int, short>();
+		internal Dictionary<int, short> _shaderLookupDictionary = new Dictionary<int, short>();
+		internal Dictionary<short, int> _reverseShaderLookupDictionary = new Dictionary<short, int>();
-		protected byte _shaderDataCount;
+		internal byte _shaderDataCount;
 
 		public T BindShader<T>(int itemId, T shaderData) where T : HairShaderData {
 			if (_shaderDataCount == 255)
 				throw new Exception("Too many shaders bound.");
 
 			_shaderLookupDictionary[itemId] = _shaderDataCount += 1;
+			_reverseShaderLookupDictionary[_shaderLookupDictionary[itemId]] = itemId;
 			_shaderData.Add(shaderData);
 			return shaderData;
 		}
