--- src/TerrariaNetCore/Terraria/Chest.cs
+++ src/Unified/Terraria/Chest.cs
@@ -319,7 +_,9 @@
 				Tile tileSafely2 = Framing.GetTileSafely(i, j);
 				if (tileSafely2.type == type2) {
 					tileSafely2.frameX -= num;
+					/*
 					Main.tile[i, j] = tileSafely2;
+					*/
 					for (int k = 0; k < 4; k++) {
 						Dust.NewDust(new Vector2(i * 16, j * 16), 16, 16, type);
 					}
@@ -384,7 +_,9 @@
 				Tile tileSafely2 = Framing.GetTileSafely(i, j);
 				if (tileSafely2.type == type) {
 					tileSafely2.frameX += num;
+					/*
 					Main.tile[i, j] = tileSafely2;
+					*/
 				}
 			}
 		}
