--- src/TerrariaNetCore/Terraria/Player.cs
+++ src/Unified/Terraria/Player.cs
@@ -18821,7 +_,7 @@
 
 	public static Tile GetFloorTile(int x, int y)
 	{
-		Tile result = null;
+		Tile result = default;
 		if (Main.tile[x - 1, y] == null)
 			Main.tile[x - 1, y] = new Tile();
 
@@ -18841,7 +_,11 @@
 		return result;
 	}
 
+	/*
 	public static int GetFloorTileType(int x, int y) => ((int?)GetFloorTile(x, y)?.type) ?? (-1);
+	*/
+
+	public static int GetFloorTileType(int x, int y) => GetFloorTile(x, y).type;
 
 	private void MakeFloorDust(bool Falling, int type, int frameX, int frameY, int paintColor)
 	{
@@ -19429,6 +_,8 @@
 		if (whoAmI != Main.myPlayer) {
 			int num2 = (int)(position.X + (float)(width / 2)) / 16;
 			int num3 = (int)(position.Y + (float)(height / 2)) / 16;
+
+			/*
 			if (!WorldGen.InWorld(num2, num3, 4))
 				flag = true;
 			else if (Main.tile[num2, num3] == null)
@@ -19440,6 +_,10 @@
 			else if (Main.tile[num2, num3 - 3] == null)
 				flag = true;
 			else if (Main.tile[num2, num3 + 3] == null)
+				flag = true;
+			*/
+
+			if (Main.netMode == NetmodeID.MultiplayerClient && !Main.sectionManager.TilesLoaded(num2 - 3, num3 - 3, num2 + 3, num3 + 3))
 				flag = true;
 
 			if (flag) {
