--- src/TerrariaNetCore/Terraria/Terraria.csproj
+++ src/Unified/Terraria/Terraria.csproj
@@ -15,8 +_,7 @@
 		<ApplicationIcon>Icon.ico</ApplicationIcon>
 
 		<!-- Avoid overwriting Terraria.exe (if it's not Debug it's release) -->
-		<AssemblyName Condition="$(Configuration.Contains('Debug'))">$(AssemblyName)Debug</AssemblyName>
-		<AssemblyName Condition="!$(Configuration.Contains('Debug'))">$(AssemblyName)Release</AssemblyName>
+		<AssemblyName>$(AssemblyName)Unified</AssemblyName>
 		<OutputName>$(AssemblyName)</OutputName>
 	</PropertyGroup>
 
@@ -30,7 +_,7 @@
 
 	<!-- References -->
 	<ItemGroup>
-		<ProjectReference Include="../../../FNA/FNA.Core.csproj" />
+		<ProjectReference Include="../../../FNA2/src/FNA/FNA.csproj" />
 		<ProjectReference Include="../ReLogic/ReLogic.csproj" />
 		<EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll">
 			<Link>Libraries/ReLogic/ReLogic.dll</Link>
@@ -46,6 +_,15 @@
 		<Reference Include="SteelSeriesEngineWrapper" />
 
 		<PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.4" />
+
+		<Using Alias="Vector2" Include="System.Numerics.Vector2"/>
+		<Using Alias="Vector3" Include="System.Numerics.Vector3"/>
+		<Using Alias="Vector4" Include="System.Numerics.Vector4"/>
+		<Using Alias="Quaternion" Include="System.Numerics.Quaternion"/>
+		<Using Alias="Point" Include="System.Drawing.Point"/>
+		<Using Alias="Rectangle" Include="System.Drawing.Rectangle"/>
+		<Using Alias="Matrix" Include="System.Numerics.Matrix4x4"/>
+		<Using Static="true" Include="Microsoft.Xna.Framework.NumericsExtensions"/>
 	</ItemGroup>
 
 	<!-- Embedded Resources -->
