--- src/TerrariaNetCore/Terraria/Terraria.csproj
+++ src/Unified/Terraria/Terraria.csproj
@@ -14,9 +_,7 @@
 		<AssemblyName>Terraria</AssemblyName>
 		<ApplicationIcon>Icon.ico</ApplicationIcon>
 
-		<!-- Avoid overwriting Terraria.exe (if it's not Debug it's release) -->
-		<AssemblyName Condition="$(Configuration.Contains('Debug'))">$(AssemblyName)Debug</AssemblyName>
-		<AssemblyName Condition="!$(Configuration.Contains('Debug'))">$(AssemblyName)Release</AssemblyName>
+		<AssemblyName>$(AssemblyName)</AssemblyName>
 		<OutputName>$(AssemblyName)</OutputName>
 	</PropertyGroup>
 
@@ -30,7 +_,7 @@
 
 	<!-- References -->
 	<ItemGroup>
-		<ProjectReference Include="../../../FNA/FNA.Core.csproj" />
+		<ProjectReference Include="../../../FNA2/src/FNA/FNA.csproj" />
 		<ProjectReference Include="../ReLogic/ReLogic.csproj" />
 		<EmbeddedResource Include="../ReLogic/bin/$(Configuration)/$(TargetFramework)/ReLogic.dll">
 			<Link>Libraries/ReLogic/ReLogic.dll</Link>
@@ -45,11 +_,30 @@
 		<Reference Include="RailSDK.Net" />
 		<Reference Include="SteelSeriesEngineWrapper" />
 
+		<PackageReference Include="Microsoft.Extensions.Hosting" Version="10.0.2" />
+		<PackageReference Include="Microsoft.Extensions.Logging" Version="10.0.2" />
 		<PackageReference Include="Steamworks.NET.AnyCPU" Version="2025.162.4" />
+
+		<PackageReference Include="Krafs.Publicizer" Version="2.3.0">
+			<PrivateAssets>all</PrivateAssets>
+			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+		</PackageReference>
+		<Publicize Include="Microsoft.Extensions.Logging.Console" IncludeCompilerGeneratedMembers="false" IncludeVirtualMembers="false" />
+
+		<Using Alias="Vector2" Include="System.Numerics.Vector2" />
+		<Using Alias="Vector3" Include="System.Numerics.Vector3" />
+		<Using Alias="Vector4" Include="System.Numerics.Vector4" />
+		<Using Alias="Quaternion" Include="System.Numerics.Quaternion" />
+		<Using Alias="Point" Include="System.Drawing.Point" />
+		<Using Alias="Rectangle" Include="System.Drawing.Rectangle" />
+		<Using Alias="Matrix" Include="System.Numerics.Matrix4x4" />
+		<Using Static="true" Include="Microsoft.Xna.Framework.NumericsExtensions" />
 	</ItemGroup>
 
 	<!-- Embedded Resources -->
 	<ItemGroup>
+		<EmbeddedResource Include="**\*.fxc" />
+		<EmbeddedResource Include="**\*.png" />
 		<EmbeddedResource Include="GameContent/Creative/Content/*" />
 		<EmbeddedResource Include="GameContent/Metadata/MaterialData/*" />
 		<EmbeddedResource Include="GameContent/WorldBuilding/*" />
@@ -64,6 +_,10 @@
 		<None Remove="Libraries/Windows/**" />
 		<None Remove="Libraries/XNA/**" />
 		<Content Include="Libraries/Native/**" CopyToOutputDirectory="PreserveNewest" />
+
+		<None Update="steam_appid.txt">
+			<CopyToOutputDirectory>Always</CopyToOutputDirectory>
+		</None>
 	</ItemGroup>
 
 	<!-- Import tasks -->
@@ -99,24 +_,6 @@
 		</PropertyGroup>
 
 		<WriteLinesToFile File="$(ProjectRuntimeConfigDevFilePath)" Lines="$(DevRuntimeConfig)" Overwrite="true" Encoding="UTF-8" />
-	</Target>
-
-	<!-- Installs the build's output into the relevant game folder. -->
-	<Target Name="CopyToActualOutputDirectory" AfterTargets="Build">
-		<PropertyGroup>
-			<_ActualOutputDirectory>$(TerrariaSteamPath)</_ActualOutputDirectory>
-		</PropertyGroup>
-
-		<!-- Copy files systematically to output folder -->
-		<ItemGroup>
-			<BinFiles Include="$(TargetDir)**" />
-		</ItemGroup>
-
-		<Message Text="Copying $(AssemblyName) to '$(_ActualOutputDirectory)'..." Importance="high" />
-		<Copy SourceFiles="@(BinFiles)" DestinationFolder="$(_ActualOutputDirectory)/%(RecursiveDir)" SkipUnchangedFiles="True" />
-
-		<!-- Copy the Libraries folder, purging excessive files -->
-		<SynchronizeDirectories Source="$(TargetDir)Libraries" Destination="$(_ActualOutputDirectory)/Libraries" />
 	</Target>
 
 </Project>
