--- src/TerrariaNetCore/Terraria/WorldGen.cs
+++ src/Unified/Terraria/WorldGen.cs
@@ -768,6 +_,7 @@
 					continue;
 				}
 
+				/*
 				Tile tile = new Tile(Main.tile[num6, num7]);
 				Tile tile2 = new Tile(Main.tile[num8, num9]);
 				Main.tile[num6, num7].active(tile2.active());
@@ -780,6 +_,29 @@
 				Main.tile[num8, num9].color(tile.color());
 				Main.tile[num8, num9].fullbrightBlock(tile.fullbrightBlock());
 				Main.tile[num8, num9].invisibleBlock(tile.invisibleBlock());
+				*/
+				Tile tileA = Main.tile[num6, num7];
+				Tile tileB = Main.tile[num8, num9];
+				var aActive = tileA.active();
+				var aType = tileA.type;
+				var aColor = tileA.color();
+				var aFullbrightBlock = tileA.fullbrightBlock();
+				var aInvisibleBlock = tileA.invisibleBlock();
+				var bActive = tileB.active();
+				var bType = tileB.type;
+				var bColor = tileB.color();
+				var bFullbrightBlock = tileB.fullbrightBlock();
+				var bInvisibleBlock = tileB.invisibleBlock();
+				tileA.active(bActive);
+				tileA.type = bType;
+				tileA.color(bColor);
+				tileA.fullbrightBlock(bFullbrightBlock);
+				tileA.invisibleBlock(bInvisibleBlock);
+				tileB.active(aActive);
+				tileB.type = aType;
+				tileB.color(aColor);
+				tileB.fullbrightBlock(aFullbrightBlock);
+				tileB.invisibleBlock(aInvisibleBlock);
 			}
 
 			num3 = Main.maxTilesX / 2 / Variations.errorWorldAdjustment(1.0);
@@ -841,6 +_,7 @@
 
 				for (int num17 = 0; num17 < num10; num17++) {
 					for (int num18 = 0; num18 < num11; num18++) {
+						/*
 						Tile tile3 = new Tile(Main.tile[num12 + num17, num13 + num18]);
 						Tile tile4 = new Tile(Main.tile[num14 + num17, num15 + num18]);
 						Main.tile[num12 + num17, num13 + num18].active(tile4.active());
@@ -853,6 +_,29 @@
 						Main.tile[num14 + num17, num15 + num18].color(tile3.color());
 						Main.tile[num14 + num17, num15 + num18].fullbrightBlock(tile3.fullbrightBlock());
 						Main.tile[num14 + num17, num15 + num18].invisibleBlock(tile3.invisibleBlock());
+						*/
+						Tile tileA = Main.tile[num12 + num17, num13 + num18];
+						Tile tileB = Main.tile[num14 + num17, num15 + num18];
+						var aActive = tileA.active();
+						var aType = tileA.type;
+						var aColor = tileA.color();
+						var aFullbrightBlock = tileA.fullbrightBlock();
+						var aInvisibleBlock = tileA.invisibleBlock();
+						var bActive = tileB.active();
+						var bType = tileB.type;
+						var bColor = tileB.color();
+						var bFullbrightBlock = tileB.fullbrightBlock();
+						var bInvisibleBlock = tileB.invisibleBlock();
+						tileA.active(bActive);
+						tileA.type = bType;
+						tileA.color(bColor);
+						tileA.fullbrightBlock(bFullbrightBlock);
+						tileA.invisibleBlock(bInvisibleBlock);
+						tileB.active(aActive);
+						tileB.type = aType;
+						tileB.color(aColor);
+						tileB.fullbrightBlock(aFullbrightBlock);
+						tileB.invisibleBlock(aInvisibleBlock);
 					}
 				}
 			}
@@ -883,7 +_,15 @@
 				int num22 = genRand.Next(5, 21);
 				int num23 = num20;
 				int num24 = num21;
+				/*
 				Tile tile5 = new Tile(Main.tile[num20, num21]);
+				*/
+				Tile tile5 = Main.tile[num20, num21];
+				var aActive = tile5.active();
+				var aType = tile5.type;
+				var aColor = tile5.color();
+				var aFullbrightBlock = tile5.fullbrightBlock();
+				var aInvisibleBlock = tile5.invisibleBlock();
 				for (int num25 = 0; num25 < num22; num25++) {
 					num23 += (int)vector2D.X;
 					num24 += (int)vector2D.Y;
@@ -892,11 +_,18 @@
 
 					Main.tile[num20, num21].slope(0);
 					Main.tile[num20, num21].halfBrick(halfBrick: false);
+					/*
 					Main.tile[num23, num24].active(tile5.active());
 					Main.tile[num23, num24].type = tile5.type;
 					Main.tile[num23, num24].color(tile5.color());
 					Main.tile[num23, num24].fullbrightBlock(tile5.fullbrightBlock());
 					Main.tile[num23, num24].invisibleBlock(tile5.invisibleBlock());
+					*/
+					Main.tile[num23, num24].active(aActive);
+					Main.tile[num23, num24].type = aType;
+					Main.tile[num23, num24].color(aColor);
+					Main.tile[num23, num24].fullbrightBlock(aFullbrightBlock);
+					Main.tile[num23, num24].invisibleBlock(aInvisibleBlock);
 				}
 			}
 		}
@@ -3811,7 +_,7 @@
 			return false;
 
 		Rectangle hitbox = Main.LocalPlayer.Hitbox;
-		hitbox.Inflate(Main.MaxWorldViewSize.X / 2, Main.MaxWorldViewSize.Y / 2);
+		hitbox.Inflate(Main.MaxGameplayViewSize.X / 2, Main.MaxGameplayViewSize.Y / 2);
 		NPC nPC = Main.npc[n];
 		Rectangle value = new Rectangle(nPC.homeTileX * 16, nPC.homeTileY * 16, 16, 16);
 		return hitbox.Intersects(value);
@@ -5805,6 +_,7 @@
 		Main.treeStyle[3] = 0;
 		Liquid.numLiquid = 0;
 		LiquidBuffer.numLiquidBuffer = 0;
+		/*
 		if (Main.netMode == 1 || lastMaxTilesX > Main.maxTilesX || lastMaxTilesY > Main.maxTilesY) {
 			for (int j = 0; j < lastMaxTilesX; j++) {
 				float num = (float)j / (float)lastMaxTilesX;
@@ -5814,12 +_,14 @@
 				}
 			}
 		}
+		*/
 
 		lastMaxTilesX = Main.maxTilesX;
 		lastMaxTilesY = Main.maxTilesY;
 		if (Main.netMode != 2)
 			Main.sectionManager = new WorldSections(Main.maxTilesX / 200, Main.maxTilesY / 150);
 
+		/*
 		if (Main.netMode != 1) {
 			for (int l = 0; l < Main.maxTilesX; l++) {
 				float num2 = (float)l / (float)Main.maxTilesX;
@@ -5832,6 +_,8 @@
 				}
 			}
 		}
+		*/
+		Main.tile.ClearEverything();
 
 		for (int n = 0; n < Main.countsAsHostForGameplay.Length; n++) {
 			Main.countsAsHostForGameplay[n] = false;
@@ -24188,7 +_,7 @@
 	{
 		int treeHeight = 1;
 		int passStyle = -1;
-		Tile topTile = null;
+		Tile topTile = default;
 		for (int num = -1; num > -100; num--) {
 			Tile tile = Main.tile[x, y + num];
 			if (!tile.active() || !TileID.Sets.GetsCheckedForLeaves[tile.type])
@@ -52039,7 +_,7 @@
 			return false;
 
 		Tile tile = Main.tile[i, j];
-		Tile tile2 = null;
+		Tile tile2 = default;
 		if (tile == null)
 			return false;
 
@@ -60120,7 +_,7 @@
 				continue;
 
 			Rectangle tileRectangle = Utils.CenteredRectangle(player.Center.ToTileCoordinates(), Point.Zero);
-			tileRectangle.Inflate(Main.MaxWorldViewSize.X / 32, Main.MaxWorldViewSize.Y / 32);
+			tileRectangle.Inflate(Main.MaxGameplayViewSize.X / 32, Main.MaxGameplayViewSize.Y / 32);
 			tileRectangle = WorldUtils.ClampToWorld(tileRectangle, 10);
 			for (int j = 0; j < 300; j++) {
 				Point point = Main.rand.NextFromRectangle(tileRectangle);
@@ -60368,7 +_,7 @@
 	private static void SpawnStormLightningNearPlayer(List<Rectangle> safespots, Player player)
 	{
 		Rectangle tileRectangle = Utils.CenteredRectangle(player.Center.ToTileCoordinates(), Point.Zero);
-		tileRectangle.Inflate(Main.MaxWorldViewSize.X / 32, Main.MaxWorldViewSize.Y / 32);
+		tileRectangle.Inflate(Main.MaxGameplayViewSize.X / 32, Main.MaxGameplayViewSize.Y / 32);
 		tileRectangle = WorldUtils.ClampToWorld(tileRectangle, 10);
 		if (tileRectangle.Width == 0 || tileRectangle.Height == 0)
 			return;
@@ -68169,8 +_,12 @@
 								break;
 						}
 
+						/*
 						int num2 = tile2?.frameY ?? (-1);
 						int num3 = tile?.frameY ?? (-1);
+						*/
+						int num2 = tile2.frameY;
+						int num3 = tile.frameY;
 						bool flag = false;
 						switch (num) {
 							default:
@@ -68201,8 +_,12 @@
 						if (up == 165 && flag && flag3)
 							up = num;
 
+						/*
 						int num4 = tile2?.frameY ?? (-1);
 						int num5 = tile?.frameY ?? (-1);
+						*/
+						int num4 = tile2.frameY;
+						int num5 = tile.frameY;
 						bool flag4 = false;
 						switch (num) {
 							default:
