--- src/TerrariaNetCore/Terraria/UI/Chat/ChatMessageContainer.cs
+++ src/tModLoader/Terraria/UI/Chat/ChatMessageContainer.cs
@@ -55,7 +_,7 @@
 			_prepared = true;
 			int num = _widthLimitInPixels;
 			if (num == -1)
-				num = Main.screenWidth - 320;
+				num = (int)(Main.screenWidth / Main.UIScale - 320); // TML fix: scale NewText by UIScale.
 
 			List<List<TextSnippet>> list = Utils.WordwrapStringSmart(OriginalText, _color, FontAssets.MouseText.Value, num, 10);
 			_parsedText.Clear();
