--- src/TerrariaNetCore/Terraria/UI/Chat/TextSnippet.cs
+++ src/tModLoader/Terraria/UI/Chat/TextSnippet.cs
@@ -12,6 +_,7 @@
 		public float Scale = 1f;
 		public bool CheckForHover;
 		public bool DeleteWhole;
+		public bool NoBlinking;
 
 		public TextSnippet(string text = "") {
 			Text = text;
@@ -34,7 +_,7 @@
 		public virtual void OnClick() {
 		}
 
-		public virtual Color GetVisibleColor() => ChatManager.WaveColor(Color);
+		public virtual Color GetVisibleColor() => NoBlinking ? Color : ChatManager.WaveColor(Color);
 
 		public virtual bool UniqueDraw(bool justCheckingString, out Vector2 size, SpriteBatch spriteBatch, Vector2 position = default(Vector2), Color color = default(Color), float scale = 1f) {
 			size = Vector2.Zero;
