--- src/TerrariaNetCore/Terraria/Map/MapHelper.cs
+++ src/tModLoader/Terraria/Map/MapHelper.cs
@@ -6,7 +_,9 @@
 using Microsoft.Xna.Framework;
 using Terraria.ID;
 using Terraria.IO;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Testing;
 using Terraria.Utilities;
 
@@ -131,7 +_,8 @@
 	private static ushort dirtPosition;
 	private static ushort rockPosition;
 	private static ushort hellPosition;
+	internal static ushort modPosition; // Added by TML.
-	private static Color[] colorLookup;
+	internal static Color[] colorLookup;
 	private static ushort[] snowTypes;
 	private static ushort wallRangeStart;
 	private static ushort wallRangeEnd;
@@ -1707,6 +_,7 @@
 
 		hellPosition = num22;
 		colorLookup[num22] = color11;
+		modPosition = (ushort)(num22 + 1); // Added by TML.
 		snowTypes = new ushort[6];
 		snowTypes[0] = tileLookup[147];
 		snowTypes[1] = tileLookup[161];
@@ -1813,8 +_,14 @@
 			newLight = Light;
+			// Extra patch context.
 			baseType = ((backgroundOverride == 0) ? GetBackgroundType(i, j, ref newLight) : backgroundOverride);
 		}
 
+		/*
 		return MapTile.Create((ushort)(baseType + baseOption), (byte)newLight, (byte)newColor);
+		*/
+		ushort mapType = (ushort)(baseType + baseOption);
+		MapLoader.ModMapOption(ref mapType, i, j);
+		return MapTile.Create(mapType, (byte)newLight, (byte)newColor);
 	}
 
 	private static void GetTileType(int i, int j, Tile tileCache, ref int newColor, ref int newLight, ref int baseType, ref int baseOption)
@@ -1874,8 +_,8 @@
 		if (tileCache.liquid > 32) {
 			int num = tileCache.liquidType();
 			baseType = liquidPosition + num;
 		}
-		else if (!tileCache.invisibleWall() && tileCache.wall > 0 && tileCache.wall < WallID.Count) {
+		else if (!tileCache.invisibleWall() && tileCache.wall > 0 && tileCache.wall < WallLoader.WallCount) {
 			int wall = tileCache.wall;
 			baseType = wallLookup[wall];
 			newColor = tileCache.wallColor();
@@ -2716,7 +_,10 @@
 					byte b6 = 0;
 					int num7;
 					ushort num8;
+					/*
 					if (mapTile.Light <= 18) {
+					*/
+					if (mapTile.Light <= 18 || mapTile.Type >= modPosition) {
 						flag2 = false;
 						flag = false;
 						num7 = 0;
@@ -2869,6 +_,8 @@
 
 			deflateStream.Dispose();
 			FileUtilities.WriteAllBytes(text, memoryStream.ToArray(), isCloudSave);
+
+			MapIO.WriteModFile(text, isCloudSave);
 		}
 
 		noStatusText = false;
@@ -3105,6 +_,7 @@
 
 	public static void LoadMapVersion1(BinaryReader fileIO, int release)
 	{
+		Main.MapFileMetadata = FileMetadata.FromCurrentSettings(FileType.Map); // TML fix. Fix map saving if loading old map file.
 		string text = fileIO.ReadString();
 		int num = fileIO.ReadInt32();
 		int num2 = fileIO.ReadInt32();
