--- src/Terraria/Terraria/Map/MapHelper.cs
+++ src/tModLoader/Terraria/Map/MapHelper.cs
@@ -5,6 +_,8 @@
 using System.IO;
 using System.Threading;
 using Terraria.IO;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -110,7 +_,8 @@
 		private static ushort dirtPosition;
 		private static ushort rockPosition;
 		private static ushort hellPosition;
+		internal static ushort modPosition;
-		private static Color[] colorLookup;
+		internal static Color[] colorLookup;
 		private static ushort[] snowTypes;
 		private static ushort wallRangeStart;
 		private static ushort wallRangeEnd;
@@ -1437,6 +_,7 @@
 
 			hellPosition = num12;
 			colorLookup[num12] = color9;
+			modPosition = (ushort)(num12 + 1);
 			snowTypes = new ushort[6];
 			snowTypes[0] = tileLookup[147];
 			snowTypes[1] = tileLookup[161];
@@ -1626,7 +_,9 @@
 				}
 			}
 
-			return MapTile.Create((ushort)(num3 + baseOption), (byte)num2, (byte)num);
+			ushort mapType = (ushort)(num3 + baseOption);
+			MapLoader.ModMapOption(ref mapType, i, j);
+			return MapTile.Create(mapType, (byte)num2, (byte)num);
 		}
 
 		public static void GetTileBaseOption(int y, Tile tileCache, ref int baseOption) {
@@ -2259,7 +_,7 @@
 								byte b5 = 0;
 								int num7;
 								ushort num8;
-								if (mapTile.Light <= 18) {
+								if (mapTile.Light <= 18 || mapTile.Type >= modPosition) {
 									flag2 = false;
 									flag = false;
 									num7 = 0;
@@ -2403,7 +_,9 @@
 
 						deflateStream.Dispose();
 						FileUtilities.WriteAllBytes(text, memoryStream.ToArray(), isCloudSave);
+						//patch file: text
 					}
+					MapIO.WriteModFile(text, isCloudSave);
 				}
 			}
 
