--- src/TerrariaNetCore/Terraria/Map/MapIconOverlay.cs
+++ src/tModLoader/Terraria/Map/MapIconOverlay.cs
@@ -7,7 +_,17 @@
 {
 	private readonly List<IMapLayer> _layers = new List<IMapLayer>();
 
+	// Added by TML.
+	private IReadOnlyList<IMapLayer> _readOnlyLayers;
+
+	// Added by TML.
+	public MapIconOverlay()
+	{
+		_readOnlyLayers = _layers.AsReadOnly();
+	}
+
+	//TML: Made internal.
-	public MapIconOverlay AddLayer(IMapLayer layer)
+	internal MapIconOverlay AddLayer(IMapLayer layer)
 	{
 		_layers.Add(layer);
 		return this;
@@ -16,7 +_,18 @@
 	public void Draw(Vector2 mapPosition, Vector2 mapOffset, Rectangle? clippingRect, float mapScale, float drawScale, ref string text)
 	{
 		MapOverlayDrawContext context = new MapOverlayDrawContext(mapPosition, mapOffset, clippingRect, mapScale, drawScale);
+
+		// Reset layer visibility.
+		foreach (var layer in _layers) {
+			layer.Visible = true;
+		}
+
+		ModLoader.SystemLoader.PreDrawMapIconOverlay(_readOnlyLayers, context);
+
 		foreach (IMapLayer layer in _layers) {
+			if (!layer.Visible)
+				continue;
+
 			layer.Draw(ref context, ref text);
 		}
 	}
