--- src/TerrariaNetCore/Terraria/Map/TeleportPylonsMapLayer.cs
+++ src/tModLoader/Terraria/Map/TeleportPylonsMapLayer.cs
@@ -7,12 +_,15 @@
 using Terraria.GameContent;
 using Terraria.GameContent.Tile_Entities;
 using Terraria.GameInput;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.Map;
 
 public class TeleportPylonsMapLayer : IMapLayer
 {
+	// Added by TML.
+	public bool Visible { get; set; } = true;
 	public const int BorderSize = 10;
 
 	public static void DrawLine(SpriteBatch spriteBatch, Vector2 start, Vector2 end, Color col, float width)
@@ -26,17 +_,27 @@
 		}
 	}
 
+
 	public void Draw(ref MapOverlayDrawContext context, ref string text)
 	{
+		/*
 		List<TeleportPylonInfo> pylons = Main.PylonSystem.Pylons;
+		*/
 		float num = 1f;
+		/*
 		float scaleIfSelected = num * 2f;
+		*/
 		float scaleIfOffscreen = num * 0.5f;
+		IReadOnlyList<TeleportPylonInfo> pylons = Main.PylonSystem.Pylons;
+
 		Texture2D value = TextureAssets.Extra[182].Value;
 		Texture2D value2 = TextureAssets.Extra[299].Value;
+
+		/*
 		Color color = Color.White;
 		if (!TeleportPylonsSystem.IsPlayerNearAPylon(Main.LocalPlayer))
 			color = Color.Gray * 0.5f;
+		*/
 
 		bool flag = false;
 		int num2 = -1;
@@ -64,6 +_,28 @@
 				continue;
 
 			bool onScreen = true;
+
+			num = 1f;
+			float scaleIfSelected = num * 2f;
+			Color color = Color.White;
+
+
+			bool isNearPylon = TeleportPylonsSystem.IsPlayerNearAPylon(Main.LocalPlayer);
+			/*
+			if (!TeleportPylonsSystem.IsPlayerNearAPylon(Main.LocalPlayer))
+			*/
+			if (!isNearPylon) {
+				color = Color.Gray * 0.5f;
+			}
+
+			if (!PylonLoader.PreDrawMapIcon(ref context, ref text, ref info2, ref isNearPylon, ref color, ref num, ref scaleIfSelected))
+				continue;
+
+			if (info2.ModPylon is ModPylon pylon) {
+				pylon.DrawMapIcon(ref context, ref text, info2, isNearPylon, color, num, scaleIfSelected);
+				continue;
+			}
+
 			if (((!Main.mapFullscreen) ? context.Draw(value, info2.PositionInTiles.ToVector2() + new Vector2(1.5f, 2f), color, new SpriteFrame(11, 1, (byte)info2.TypeOfPylon, 0) {
 				PaddingY = 0
 			}, num, scaleIfSelected, Alignment.Center) : context.DrawClamped(value, value2, info2.PositionInTiles.ToVector2() + new Vector2(1.5f, 2f), color, new SpriteFrame(11, 1, (byte)info2.TypeOfPylon, 0) {
