--- src/TerrariaNetCore/Terraria/Map/TeleportPylonsMapLayer.cs
+++ src/tModLoader/Terraria/Map/TeleportPylonsMapLayer.cs
@@ -6,14 +_,17 @@
 using Terraria.GameContent;
 using Terraria.GameContent.Tile_Entities;
 using Terraria.GameInput;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.Map
 {
 	public class TeleportPylonsMapLayer : IMapLayer
 	{
+		public bool Visible { get; set; } = true;
+
 		public void Draw(ref MapOverlayDrawContext context, ref string text) {
-			List<TeleportPylonInfo> pylons = Main.PylonSystem.Pylons;
+			List<TeleportPylonInfo> pylons = Main.PylonSystem.VanillaPylons;
 			float num = 1f;
 			float scaleIfSelected = num * 2f;
 			Texture2D value = TextureAssets.Extra[182].Value;
@@ -24,6 +_,9 @@
 
 			for (int i = 0; i < pylons.Count; i++) {
 				TeleportPylonInfo info = pylons[i];
+				if (!PylonLoader.PreDrawMapIcon(ref context, ref text, ref info, ref num2, ref color, ref num, ref scaleIfSelected)) {
+					continue;
+				}
 				if (context.Draw(value, info.PositionInTiles.ToVector2() + new Vector2(1.5f, 2f), color, new SpriteFrame(9, 1, (byte)info.TypeOfPylon, 0) {
 					PaddingY = 0
 				}, num, scaleIfSelected, Alignment.Center).IsMouseOver) {
