--- src/TerrariaNetCore/Terraria/Map/WorldMap.cs
+++ src/tModLoader/Terraria/Map/WorldMap.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.IO;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -86,6 +_,7 @@
 		}
 
 		if (!FileUtilities.Exists(text, isCloudSave)) {
+			// Patch note: text ^, used below.
 			Main.MapFileMetadata = FileMetadata.FromCurrentSettings(FileType.Map);
 			return;
 		}
@@ -101,6 +_,9 @@
 					MapHelper.LoadMapVersion2(binaryReader, num);
 
 				ClearEdges();
+
+				MapIO.ReadModFile(text, isCloudSave);
+
 				Main.clearMap = true;
 				Main.loadMap = true;
 				Main.loadMapLock = true;
