--- src/Terraria/Terraria/Map/WorldMap.cs
+++ src/tModLoader/Terraria/Map/WorldMap.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.IO;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -81,6 +_,7 @@
 
 			if (!FileUtilities.Exists(text, isCloudSave)) {
 				Main.MapFileMetadata = FileMetadata.FromCurrentSettings(FileType.Map);
+				//patch file: text
 				return;
 			}
 
@@ -95,6 +_,7 @@
 								MapHelper.LoadMapVersion2(binaryReader, num);
 
 							ClearEdges();
+							MapIO.ReadModFile(text, isCloudSave);
 							Main.clearMap = true;
 							Main.loadMap = true;
 							Main.loadMapLock = true;
