--- src/TerrariaNetCore/Terraria/Map/WorldMap.cs
+++ src/tModLoader/Terraria/Map/WorldMap.cs
@@ -1,7 +_,8 @@
 using System;
 using System.IO;
 using Terraria.IO;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Testing;
 using Terraria.Utilities;
 
@@ -152,11 +_,12 @@
 			return;
 
 		if (!TryGetMapPath(Main.ActivePlayerFileData, Main.ActiveWorldFileData, out var mapPath)) {
+			// Patch note: mapPath ^, used below.
 			Main.MapFileMetadata = FileMetadata.FromCurrentSettings(FileType.Map);
 			return;
 		}
 
-		using MemoryStream input = new MemoryStream(FileUtilities.ReadAllBytes(mapPath, isCloudSave));
+		using MemoryStream input = FileUtilities.ReadAllBytes(mapPath, isCloudSave).ToMemoryStream();
 		using BinaryReader binaryReader = new BinaryReader(input);
 		try {
 			int num = binaryReader.ReadInt32();
@@ -170,12 +_,17 @@
 				else
 					MapHelper.LoadMapVersion2(binaryReader, num);
 
+				MapIO.ReadModFile(text, isCloudSave);
+
 				ClearEdges();
 				Main.clearMap = true;
 				Main.loadMap = true;
 				Main.loadMapLock = true;
 				Main.refreshMap = false;
 			}
+			else {
+				throw new Exception($"Map release version too high ({num}), the map file \'{text}\' is either corrupted or from a future version of Terraria."); // TML fix. Fix map saving if loading corrupted map file with higher release version number.
+			}
 		}
 		catch (Exception value) {
 			using (StreamWriter streamWriter = new StreamWriter("client-crashlog.txt", append: true)) {
@@ -188,6 +_,7 @@
 				File.Copy(mapPath, mapPath + ".bad", overwrite: true);
 
 			Clear();
+			Main.MapFileMetadata = FileMetadata.FromCurrentSettings(FileType.Map); // TML fix. Fix map saving if loading corrupted map file.
 		}
 	}
 
