--- src/decompiled/GameContent/NetModules/NetBestiaryModule.cs
+++ src/Terraria/GameContent/NetModules/NetBestiaryModule.cs
@@ -1,5 +_,7 @@
 using System.IO;
+#if CLIENT
 using Terraria.ID;
+#endif
 using Terraria.Net;
 
 namespace Terraria.GameContent.NetModules
@@ -36,6 +_,9 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
+#if SERVER
+			return false;
+#else
 			switch (reader.ReadByte()) {
 				case 0: {
 						short key3 = reader.ReadInt16();
@@ -59,6 +_,7 @@
 			}
 
 			return true;
+#endif
 		}
 	}
 }
