--- src/Terraria/Terraria/GameContent/Biomes/DeadMansChestBiome.cs
+++ src/tModLoader/Terraria/GameContent/Biomes/DeadMansChestBiome.cs
@@ -2,6 +_,7 @@
 using Newtonsoft.Json;
 using System.Collections.Generic;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.Utilities;
 using Terraria.WorldBuilding;
 
@@ -200,7 +_,7 @@
 		}
 
 		private void PlaceBoulderTrapSpot(Point position, int yPush) {
-			int[] array = new int[624];
+			int[] array = new int[TileLoader.TileCount];
 			for (int i = position.X; i < position.X + 2; i++) {
 				for (int j = position.Y - 4; j <= position.Y; j++) {
 					Tile tile = Main.tile[i, j];
@@ -399,7 +_,7 @@
 
 			item[1] = new Item();
 			item[1].SetDefaults(5007);
-			Main.chest[num3].item = item;
+			Main.chest[num3].item.GetItemArray() = item;
 		}
 
 		private void ActuallyPlaceDartTrap(Point position, int directionX, int x, int y, int xPush, Tile t) {
