--- src/Terraria\Terraria.GameContent\ChildSafety.cs
+++ src/tModLoader\Terraria.GameContent\ChildSafety.cs
@@ -7,7 +_,7 @@
 	{
 		private static SetFactory factoryDust = new SetFactory(275);
 		private static SetFactory factoryGore = new SetFactory(1087);
-		private static readonly bool[] SafeGore = factoryGore.CreateBoolSet(new int[] {
+		public static bool[] SafeGore = factoryGore.CreateBoolSet(new int[] {
 			11,
 			12,
 			13,
@@ -210,7 +_,7 @@
 			925,
 			926
 		});
-		private static readonly bool[] SafeDust = factoryDust.CreateBoolSet(true, new int[] {
+		public static bool[] SafeDust = factoryDust.CreateBoolSet(true, new int[] {
 			5,
 			227,
 			273
