--- src/decompiled/ReLogic/OS/Platform.cs
+++ src/Terraria/ReLogic/OS/Platform.cs
@@ -1,12 +_,25 @@
 using System;
+using System.Runtime.InteropServices;
+#if WINDOWS
 using ReLogic.OS.Windows;
+#elif LINUX
+using ReLogic.OS.Linux;
+#elif MAC
+using ReLogic.OS.OSX;
+#endif
 using ReLogic.Utilities;
 
 namespace ReLogic.OS;
 
 public abstract class Platform : IDisposable
 {
+#if WINDOWS
 	public static readonly Platform Current = new WindowsPlatform();
+#elif LINUX
+	public static readonly Platform Current = new LinuxPlatform();
+#elif MAC
+	public static readonly Platform Current = new OsxPlatform();
+#endif
 	public readonly PlatformType Type;
 	private TypeInstanceCollection<object> _services = new TypeInstanceCollection<object>();
 	private bool _disposedValue;
