--- src/decompiled/Terraria.Graphics/EmpressBladeDrawer.cs
+++ src/Terraria/Terraria.Graphics/EmpressBladeDrawer.cs
@@ -1,5 +_,7 @@
 using Microsoft.Xna.Framework;
+#if XNA
 using Terraria.Graphics.Shaders;
+#endif
 
 namespace Terraria.Graphics
 {
@@ -12,6 +_,7 @@
 		public Color ColorEnd;
 
 		public void Draw(Projectile proj) {
+#if XNA
 			_ = proj.ai[1];
 			MiscShaderData miscShaderData = GameShaders.Misc["EmpressBlade"];
 			int num = 1;
@@ -23,6 +_,7 @@
 			_vertexStrip.PrepareStrip(proj.oldPos, proj.oldRot, StripColors, StripWidth, -Main.screenPosition + proj.Size / 2f, proj.oldPos.Length, includeBacksides: true);
 			_vertexStrip.DrawTrail();
 			Main.pixelShader.CurrentTechnique.Passes[0].Apply();
+#endif
 		}
 
 		private Color StripColors(float progressOnStrip) {
