--- src/decompiled/Terraria.Graphics/MagicMissileDrawer.cs
+++ src/Terraria/Terraria.Graphics/MagicMissileDrawer.cs
@@ -1,6 +_,8 @@
 using Microsoft.Xna.Framework;
 using System.Runtime.InteropServices;
+#if XNA
 using Terraria.Graphics.Shaders;
+#endif
 
 namespace Terraria.Graphics
 {
@@ -10,6 +_,7 @@
 		private static VertexStrip _vertexStrip = new VertexStrip();
 
 		public void Draw(Projectile proj) {
+#if XNA
 			MiscShaderData miscShaderData = GameShaders.Misc["MagicMissile"];
 			miscShaderData.UseSaturation(-2.8f);
 			miscShaderData.UseOpacity(2f);
@@ -17,6 +_,7 @@
 			_vertexStrip.PrepareStripWithProceduralPadding(proj.oldPos, proj.oldRot, StripColors, StripWidth, -Main.screenPosition + proj.Size / 2f);
 			_vertexStrip.DrawTrail();
 			Main.pixelShader.CurrentTechnique.Passes[0].Apply();
+#endif
 		}
 
 		private Color StripColors(float progressOnStrip) {
