--- src/decompiled/Audio/ActiveSound.cs
+++ src/Terraria/Audio/ActiveSound.cs
@@ -1,3 +_,6 @@
+#if FNA
+using System.Threading;
+#endif
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
 
@@ -22,11 +_,17 @@
 		public bool IsPlaying => Sound.State == SoundState.Playing;
 
 		public ActiveSound(SoundStyle style, Vector2 position) {
+#if FNA
+			Monitor.Enter(SoundEngine.AudioLock);
+#endif
 			Position = position;
 			Volume = 1f;
 			IsGlobal = false;
 			Style = style;
 			Play();
+#if FNA
+			Monitor.Exit(SoundEngine.AudioLock);
+#endif
 		}
 
 		public ActiveSound(SoundStyle style) {
