--- src/Terraria\Properties\AssemblyInfo.cs
+++ src/tModLoader\Properties\AssemblyInfo.cs
@@ -8,7 +_,9 @@
 using System.Security.Permissions;
 
 [assembly: AssemblyVersion("1.3.5.1")]
+#if !DEBUG
 [assembly: Debuggable(DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints)]
+#endif
 [assembly: AssemblyCompany("Re-Logic")]
 [assembly: AssemblyCopyright("Copyright © Re-Logic 2017")]
 [assembly: AssemblyDescription("")]
@@ -23,3 +_,4 @@
 //[assembly: TargetFramework(".NETFramework,Version=v4.0,Profile=Client", FrameworkDisplayName = ".NET Framework 4 Client Profile")]
 [assembly: SecurityPermission(SecurityAction.RequestMinimum, SkipVerification = true)]
 [module: UnverifiableCode]
+[assembly: InternalsVisibleTo("tModLoaderTests")]

