--- src/Terraria/Terraria/GameInput/PlayerInputProfile.cs
+++ src/tModLoader/Terraria/GameInput/PlayerInputProfile.cs
@@ -3,10 +_,12 @@
 using Newtonsoft.Json.Linq;
 using System;
 using System.Collections.Generic;
+using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput
 {
-	public class PlayerInputProfile
+	public partial class PlayerInputProfile
 	{
 		public Dictionary<InputMode, KeyConfiguration> InputModes = new Dictionary<InputMode, KeyConfiguration> {
 			{ InputMode.Keyboard, new KeyConfiguration() },
@@ -185,9 +_,12 @@
 		}
 
 		public void CopyGameplaySettingsFrom(PlayerInputProfile profile, InputMode mode) {
-			string[] keysToCopy = new string[19] {
+			string[] keysToCopy = new string[] {
 				"MouseLeft",
 				"MouseRight",
+				"MouseMiddle",
+				"MouseXButton1",
+				"MouseXButton2",
 				"Up",
 				"Down",
 				"Left",
