--- src/Terraria/Terraria/GameInput/PlayerInput.cs
+++ src/tModLoader/Terraria/GameInput/PlayerInput.cs
@@ -8,13 +_,14 @@
 using Terraria.GameContent.UI.Chat;
 using Terraria.GameContent.UI.States;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
 namespace Terraria.GameInput
 {
-	public class PlayerInput
+	public partial class PlayerInput
 	{
 		public class MiscSettingsTEMP
 		{
@@ -26,6 +_,9 @@
 		public static List<string> KnownTriggers = new List<string> {
 			"MouseLeft",
 			"MouseRight",
+			"MouseMiddle",
+			"MouseXButton1",
+			"MouseXButton2",
 			"Up",
 			"Down",
 			"Left",
@@ -264,7 +_,7 @@
 			if (_fastUseItemInventorySlot >= 0 && _fastUseItemInventorySlot != -2) {
 				Player localPlayer = Main.LocalPlayer;
 				if (localPlayer.inventory[_fastUseItemInventorySlot].IsAir)
-					Utils.Swap(ref Main.mouseItem, ref localPlayer.inventory[_fastUseItemInventorySlot]);
+					Utils.Swap(ref Main.mouseItem, ref ((Item[])localPlayer.inventory)[_fastUseItemInventorySlot]);
 			}
 
 			_fastUseItemInventorySlot = -1;
@@ -277,7 +_,7 @@
 			if (Main.mouseItem.stack <= 0) {
 				int uIPointForBuildingMode = _UIPointForBuildingMode;
 				if (uIPointForBuildingMode < 50 && uIPointForBuildingMode >= 0 && Main.player[Main.myPlayer].inventory[uIPointForBuildingMode].stack > 0)
-					Utils.Swap(ref Main.mouseItem, ref Main.player[Main.myPlayer].inventory[uIPointForBuildingMode]);
+					Utils.Swap(ref Main.mouseItem, ref ((Item[])Main.player[Main.myPlayer].inventory)[uIPointForBuildingMode]);
 			}
 		}
 
@@ -288,7 +_,7 @@
 			if (Main.mouseItem.stack > 0 && Main.player[Main.myPlayer].itemAnimation == 0) {
 				int uIPointForBuildingMode = _UIPointForBuildingMode;
 				if (uIPointForBuildingMode < 50 && uIPointForBuildingMode >= 0 && Main.player[Main.myPlayer].inventory[uIPointForBuildingMode].stack <= 0)
-					Utils.Swap(ref Main.mouseItem, ref Main.player[Main.myPlayer].inventory[uIPointForBuildingMode]);
+					Utils.Swap(ref Main.mouseItem, ref ((Item[])Main.player[Main.myPlayer].inventory)[uIPointForBuildingMode]);
 			}
 
 			_UIPointForBuildingMode = -1;
@@ -316,9 +_,21 @@
 			}
 		}
 
+		// To allow mod keybinds to load, need to reinitialize after load.
+		internal static bool reinitialize;
+		private static void ReInitialize() {
+			Profiles.Clear();
+			OriginalProfiles.Clear();
+			Initialize_Inner(); //Mirsario: To not repeat code in patches, reuse the one that was previously in Initialize().
+			Load(); // Loads the JSON into the Profiles above
+			reinitialize = false;
+		}
 		public static void Initialize() {
 			Main.InputProfiles.OnProcessText += PrettyPrintProfiles;
 			Player.Hooks.OnEnterWorld += Hook_OnEnterWorld;
+			Initialize_Inner();
+		}
+		private static void Initialize_Inner() {
 			PlayerInputProfile playerInputProfile = new PlayerInputProfile("Redigit's Pick");
 			playerInputProfile.Initialize(PresetProfiles.Redigit);
 			Profiles.Add(playerInputProfile.Name, playerInputProfile);
@@ -366,7 +_,9 @@
 		}
 
 		public static void Load() {
-			Main.InputProfiles.Load();
+			if (!Main.InputProfiles.Load())
+				return; // fixes strange bug caused by deleting input profiles.json in game and reloading mods.
+
 			Dictionary<string, PlayerInputProfile> dictionary = new Dictionary<string, PlayerInputProfile>();
 			string currentValue = null;
 			Main.InputProfiles.Get("Selected Profile", ref currentValue);
@@ -496,6 +_,9 @@
 		}
 
 		public static void UpdateInput() {
+			if (reinitialize)
+				ReInitialize();
+
 			Triggers.Reset();
 			ScrollWheelValueOld = ScrollWheelValue;
 			ScrollWheelValue = 0;
@@ -526,6 +_,9 @@
 			UpdateMainMouse();
 			Main.mouseLeft = Triggers.Current.MouseLeft;
 			Main.mouseRight = Triggers.Current.MouseRight;
+			Main.mouseMiddle = Triggers.Current.MouseMiddle;
+			Main.mouseXButton1 = Triggers.Current.MouseXButton1;
+			Main.mouseXButton2 = Triggers.Current.MouseXButton2;
 			CacheZoomableValues();
 			if (num != 0 && PlayerInput.OnActionableInput != null)
 				PlayerInput.OnActionableInput();
@@ -716,6 +_,8 @@
 				flag = true;
 			}
 
+			// TODO: We might just need a bool for while reloading.
+			if (player.HeldItem.type >= ItemID.Sets.GamepadWholeScreenUseRange.Length) return false;
 			bool flag4 = ItemID.Sets.GamepadWholeScreenUseRange[player.inventory[player.selectedItem].type] || player.scope;
 			int num3 = player.inventory[player.selectedItem].tileBoost + ItemID.Sets.GamepadExtraRange[player.inventory[player.selectedItem].type];
 			if (player.yoyoString && ItemID.Sets.Yoyo[player.inventory[player.selectedItem].type])
@@ -996,6 +_,9 @@
 			List<string> list = new List<string> {
 				"MouseLeft",
 				"MouseRight",
+				"MouseMiddle",
+				"MouseXButton1",
+				"MouseXButton2",
 				"Inventory"
 			};
 
@@ -1177,7 +_,7 @@
 					if (DpadSnapCooldown[i] == 0) {
 						int num = 6;
 						if (ItemSlot.IsABuildingItem(player.inventory[player.selectedItem]))
-							num = player.inventory[player.selectedItem].useTime;
+							num = CombinedHooks.TotalUseTime(player.inventory[player.selectedItem].useTime, player, player.inventory[player.selectedItem]);
 
 						DpadSnapCooldown[i] = num;
 						zero += vector;
@@ -1419,6 +_,9 @@
 						case InputMode.Keyboard:
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Down"].Add("S");
 							c.KeyStatus["Left"].Add("A");
@@ -1456,6 +_,9 @@
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseLeft"].Add("Space");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Up"].Add("Up");
 							c.KeyStatus["Down"].Add("S");
@@ -1527,6 +_,9 @@
 						case InputMode.Keyboard:
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Down"].Add("S");
 							c.KeyStatus["Left"].Add("A");
@@ -1564,6 +_,9 @@
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseLeft"].Add("Space");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Up"].Add("Up");
 							c.KeyStatus["Down"].Add("S");
@@ -1634,6 +_,9 @@
 						case InputMode.Keyboard:
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Down"].Add("S");
 							c.KeyStatus["Left"].Add("A");
@@ -1671,6 +_,9 @@
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseLeft"].Add("Space");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Up"].Add("Up");
 							c.KeyStatus["Down"].Add("S");
@@ -1741,6 +_,9 @@
 						case InputMode.Keyboard:
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Down"].Add("S");
 							c.KeyStatus["Left"].Add("A");
@@ -1778,6 +_,9 @@
 							c.KeyStatus["MouseLeft"].Add("Mouse1");
 							c.KeyStatus["MouseLeft"].Add("Space");
 							c.KeyStatus["MouseRight"].Add("Mouse2");
+							c.KeyStatus["MouseMiddle"].Add("Mouse3");
+							c.KeyStatus["MouseXButton1"].Add("Mouse4");
+							c.KeyStatus["MouseXButton2"].Add("Mouse5");
 							c.KeyStatus["Up"].Add("W");
 							c.KeyStatus["Up"].Add("Up");
 							c.KeyStatus["Down"].Add("S");
