--- src/Terraria/Terraria/GameInput/PlayerInputProfile.cs
+++ src/tModLoader/Terraria/GameInput/PlayerInputProfile.cs
@@ -3,6 +_,8 @@
 using Newtonsoft.Json.Linq;
 using System;
 using System.Collections.Generic;
+using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput
 {
@@ -185,9 +_,12 @@
 		}
 
 		public void CopyGameplaySettingsFrom(PlayerInputProfile profile, InputMode mode) {
-			string[] keysToCopy = new string[19] {
+			string[] keysToCopy = new string[] {
 				"MouseLeft",
 				"MouseRight",
+				"MouseMiddle",
+				"MouseXButton1",
+				"MouseXButton2",
 				"Up",
 				"Down",
 				"Left",
@@ -240,6 +_,21 @@
 			};
 
 			CopyKeysFrom(profile, mode, keysToCopy);
+		}
+
+		public void CopyModHotkeySettingsFrom(PlayerInputProfile profile, InputMode mode) {
+			foreach (var modHotkey in HotkeyLoader.Hotkeys) {
+				InputModes[mode].KeyStatus[modHotkey.uniqueName].Clear();
+				if (!string.IsNullOrEmpty(modHotkey.defaultKey))
+					InputModes[mode].KeyStatus[modHotkey.uniqueName].Add(modHotkey.defaultKey);
+			}
+		}
+
+		public void CopyIndividualModHotkeySettingsFrom(PlayerInputProfile profile, InputMode mode, string uniqueName) {
+			var modHotKey = HotkeyLoader.modHotkeys[uniqueName];
+			InputModes[mode].KeyStatus[uniqueName].Clear();
+			if (!string.IsNullOrEmpty(modHotKey.defaultKey))
+				InputModes[mode].KeyStatus[uniqueName].Add(modHotKey.defaultKey);
 		}
 
 		public void CopyGamepadSettingsFrom(PlayerInputProfile profile, InputMode mode) {
