--- src/TerrariaNetCore/Terraria/GameInput/PlayerInputProfile.cs
+++ src/tModLoader/Terraria/GameInput/PlayerInputProfile.cs
@@ -1,12 +_,14 @@
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using Microsoft.Xna.Framework.Input;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput;
 
-public class PlayerInputProfile
+public partial class PlayerInputProfile
 {
 	public Dictionary<InputMode, KeyConfiguration> InputModes = new Dictionary<InputMode, KeyConfiguration> {
 		{ InputMode.Keyboard, new KeyConfiguration() },
@@ -206,9 +_,13 @@
 
 	public void CopyGameplaySettingsFrom(PlayerInputProfile profile, InputMode mode)
 	{
-		string[] keysToCopy = new string[23] {
+		string[] keysToCopy = new string[] {
 			"MouseLeft",
 			"MouseRight",
+			// Extra mouse buttons support
+			TriggerNames.MouseMiddle,
+			TriggerNames.MouseXButton1,
+			TriggerNames.MouseXButton2,
 			"Up",
 			"Down",
 			"Left",
