--- src/Terraria/Terraria/GameInput/TriggersSet.cs
+++ src/tModLoader/Terraria/GameInput/TriggersSet.cs
@@ -1,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using System.Collections.Generic;
 using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput
 {
@@ -29,6 +_,33 @@
 			}
 		}
 
+		public bool MouseMiddle {
+			get {
+				return KeyStatus["MouseMiddle"];
+			}
+			set {
+				KeyStatus["MouseMiddle"] = value;
+			}
+		}
+
+		public bool MouseXButton1 {
+			get {
+				return KeyStatus["MouseXButton1"];
+			}
+			set {
+				KeyStatus["MouseXButton1"] = value;
+			}
+		}
+
+		public bool MouseXButton2 {
+			get {
+				return KeyStatus["MouseXButton2"];
+			}
+			set {
+				KeyStatus["MouseXButton2"] = value;
+			}
+		}
+
 		public bool Up {
 			get {
 				return KeyStatus["Up"];
@@ -505,6 +_,10 @@
 			foreach (string knownTrigger in PlayerInput.KnownTriggers) {
 				KeyStatus.Add(knownTrigger, value: false);
 			}
+
+			foreach (var keybind in KeybindLoader.Keybinds) {
+				KeyStatus.Add(keybind.uniqueName, false);
+			}
 		}
 
 		public Vector2 GetNavigatorDirections() {
@@ -567,6 +_,8 @@
 
 			if (HotbarScrollCD > 0 && (!(HotbarScrollCD == 1 && flag) || PlayerInput.CurrentProfile.HotbarRadialHoldTimeRequired <= 0))
 				HotbarScrollCD--;
+
+			PlayerLoader.ProcessTriggers(p, this);
 		}
 
 		public void CopyIntoDuringChat(Player p) {
