--- src/TerrariaNetCore/Terraria/Net/Sockets/SocialSocket.cs
+++ src/tModLoader/Terraria/Net/Sockets/SocialSocket.cs
@@ -1,5 +_,6 @@
 using System.Threading;
 using System.Threading.Tasks;
+using Terraria.ModLoader;
 using Terraria.Social;
 
 namespace Terraria.Net.Sockets;
@@ -22,6 +_,8 @@
 	void ISocket.Close()
 	{
 		if (_remoteAddress != null) {
+			ModNet.Log(_remoteAddress, "Closing SocialSocket");
+
 			SocialAPI.Network.Close(_remoteAddress);
 			_remoteAddress = null;
 		}
@@ -37,6 +_,15 @@
 
 	void ISocket.AsyncSend(byte[] data, int offset, int size, SocketSendCallback callback, object state)
 	{
+		// Replace first chance exception with a cleaner log message, because Send fails on a null _remoteAddress
+		if (_remoteAddress == null) {
+			ModNet.Warn("SocialSocket, AsyncSend after connection closed.");
+			return;
+		}
+
+		if (ModNet.DetailedLogging)
+			ModNet.Debug(_remoteAddress, $"send {size}");
+
 		SocialAPI.Network.Send(_remoteAddress, data, size);
 #if NETCORE
 		Task.Run(() => callback.Invoke(state));
