--- src/Terraria\Terraria.Initializers\LaunchInitializer.cs
+++ src/tModLoader\Terraria.Initializers\LaunchInitializer.cs
@@ -1,6 +_,9 @@
 using System;
 using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
 using Terraria.Localization;
+using Terraria.ModLoader.Core;
 using Terraria.Social;
 
 namespace Terraria.Initializers
@@ -30,6 +_,12 @@
 				"-port"
 			})) != null && int.TryParse(s, out listenPort))
 				Netplay.ListenPort = listenPort;
+
+			string modPack;
+			if ((modPack = TryParameter(new string[] {
+				"-modpack"
+			})) != null)
+				ModOrganizer.commandLineModPack = modPack;
 		}
 
 		private static void LoadClientParameters(Main game) {
@@ -53,6 +_,30 @@
 				"-host"
 			}))
 				game.AutoHost();
+
+			if (HasParameter("-skipselect")) {
+				string skipSelectArgs = TryParameter(new string[] { "-skipselect" });
+				string playerName = null;
+				string worldName = null;
+				if (skipSelectArgs != null) {
+					Regex cmdRegEx = new Regex(@"(?<name>.*?):(?<val>.*)");
+					Match m = cmdRegEx.Match(skipSelectArgs);
+					if (m.Success) {
+						playerName = m.Groups["name"].Value;
+						worldName = m.Groups["val"].Value;
+					}
+				}
+
+				Main.showSplash = false;
+				ModLoader.ModLoader.OnSuccessfulLoad += () => {
+					WorldGen.clearWorld();
+					Main.LoadPlayers();
+					(Main.PlayerList.FirstOrDefault(x => x.Name == playerName) ?? Main.PlayerList[0]).SetAsActive();
+					Main.LoadWorlds();
+					(Main.WorldList.FirstOrDefault(x => x.Name == worldName) ?? Main.WorldList[0]).SetAsActive();
+					WorldGen.playWorld();
+				};
+			}
 		}
 
 		private static void LoadServerParameters(Main game) {
@@ -126,6 +_,18 @@
 			})) != null)
 				LanguageManager.Instance.SetLanguage(text);
 
+			// We handle the build parameter this far in
+			// so vanilla can take care of loading the appropriate localization keys beforehand
+			if ((text = TryParameter(new string[] {
+				"-build"
+			})) != null)
+				// Inside the build from command line, the application exits
+				// Anything beyond this point will not be executed
+				ModCompile.BuildModCommandLine(text);
+
+			if ((text = TryParameter("-install")) != null)
+				FileAssociationSupport.HandleFileAssociation(text);
+
 			string worldName;
 			if ((worldName = TryParameter(new string[] {
 				"-worldname"
@@ -137,6 +_,15 @@
 				"-motd"
 			})) != null)
 				game.NewMOTD(newMOTD);
+
+			string modPath;
+			if ((modPath = TryParameter(new string[] {
+				"-modpath"
+			})) != null)
+				ModOrganizer.modPath = modPath;
+
+			if (HasParameter("-showserverconsole"))
+				Main.showServerConsole = true;
 
 			string banFilePath;
 			if ((banFilePath = TryParameter(new string[] {
