--- src/Terraria\Terraria.Initializers\LaunchInitializer.cs
+++ src/tModLoader\Terraria.Initializers\LaunchInitializer.cs
@@ -37,6 +_,14 @@
 			{
 				Netplay.ListenPort = listenPort;
 			}
+			string modPack;
+			if ((modPack = LaunchInitializer.TryParameter(new string[]
+				{
+					"-modpack"
+				})) != null)
+			{
+				ModLoader.ModLoader.commandLineModPack = modPack;
+			}
 		}
 
 		private static void LoadClientParameters(Main game)
@@ -66,6 +_,19 @@
 				}))
 			{
 				game.AutoHost();
+			}
+			if (LaunchInitializer.HasParameter("-skipselect"))
+			{
+				Main.showSplash = false;
+				ModLoader.ModLoader.PostLoad += () =>
+				{
+					WorldGen.clearWorld();
+					Main.LoadPlayers();
+					Main.PlayerList[0].SetAsActive();
+					Main.LoadWorlds();
+					Main.WorldList[0].SetAsActive();
+					WorldGen.playWorld();
+				};
 			}
 		}
 
@@ -172,6 +_,14 @@
 			{
 				game.NewMOTD(newMOTD);
 			}
+			string modPath;
+			if ((modPath = LaunchInitializer.TryParameter(new string[]
+				{
+					"-modpath"
+				})) != null)
+			{
+				ModLoader.ModLoader.modPath = modPath;
+			}
 			string banFilePath;
 			if ((banFilePath = LaunchInitializer.TryParameter(new string[]
 				{

