--- src/Terraria/Terraria.Initializers/UILinksInitializer.cs
+++ src/tModLoader/Terraria.Initializers/UILinksInitializer.cs
@@ -2,6 +_,9 @@
 using System;
 using Terraria.GameContent.UI.States;
 using Terraria.GameInput;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
+using Terraria.ModLoader.UI.ModBrowser;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -490,7 +_,7 @@
 				return PlayerInput.BuildCommand(Lang.misc[56].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + PlayerInput.BuildCommand(Lang.misc[64].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["HotbarMinus"], PlayerInput.ProfileGamepadUI.KeyStatus["HotbarPlus"]) + PlayerInput.BuildCommand(Lang.misc[70].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]) + PlayerInput.BuildCommand(Lang.misc[69].Value, true, PlayerInput.ProfileGamepadUI.KeyStatus["SmartSelect"]);
 			};
 
-			for (int num8 = 600; num8 <= 650; num8++) {
+			for (int num8 = 600; num8 <= 698; num8++) {
 				UILinkPoint value11 = new UILinkPoint(num8, enabled: true, num8 + 10, num8 - 10, num8 - 1, num8 + 1);
 				cp13.LinkMap.Add(num8, value11);
 			}
@@ -500,7 +_,7 @@
 				if (num50 == 0)
 					num50 = 100;
 
-				for (int num51 = 0; num51 < 50; num51++) {
+				for (int num51 = 0; num51 < 98; num51++) {
 					cp13.LinkMap[600 + num51].Up = ((num51 % num50 == 0) ? (-1) : (600 + num51 - 1));
 					if (cp13.LinkMap[600 + num51].Up == -1) {
 						if (num51 >= num50 * 2)
@@ -1370,6 +_,13 @@
 					break;
 				case 6:
 					UIVirtualKeyboard.Cancel();
+					break;
+				case 100:
+					Main.PlaySound(11, -1, -1, 1, 1f, 0f);
+					Main.menuMode = UILinkPointNavigator.Shortcuts.BackButtonGoto;
+					break;
+				case 101:
+					Interface.modBrowser.BackClick(null, null);
 					break;
 			}
 		}
