--- src/Terraria/Terraria.Social/SocialAPI.cs
+++ src/tModLoader/Terraria.Social/SocialAPI.cs
@@ -1,4 +_,5 @@
 using System.Collections.Generic;
+using Terraria.ModLoader.Engine;
 using Terraria.Social.Base;
 using Terraria.Social.Steam;
 
@@ -20,7 +_,11 @@
 			if (!mode.HasValue) {
 				mode = SocialMode.None;
 #if CLIENT
+#if DEBUG
+				if (!Program.LaunchParameters.ContainsKey("-nosteam"))
+#endif
+				if (!GoGVerifier.IsGoG)
-				mode = SocialMode.Steam;
+					mode = SocialMode.Steam;
 #else
 				if (Program.LaunchParameters.ContainsKey("-steam"))
 					mode = SocialMode.Steam;
