--- src/Terraria\Terraria.IO\PlayerFileData.cs
+++ src/tModLoader\Terraria.IO\PlayerFileData.cs
@@ -3,6 +_,8 @@
 using System.Diagnostics;
 using System.IO;
 using System.Text.RegularExpressions;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -14,6 +_,7 @@
 		private TimeSpan _playTime = TimeSpan.Zero;
 		private Stopwatch _timer = new Stopwatch();
 		private bool _isTimerActive;
+        public CustomModDataException customDataFail = null;
 
 		public Player Player
 		{
@@ -96,6 +_,7 @@
 				this._isCloudSave = true;
 				this._path = playerPathFromName;
 				Main.CloudFavoritesData.SaveFavorite(this);
+				PlayerIO.MoveToCloud(base.Path, playerPathFromName);
 			}
 		}
 
@@ -112,6 +_,8 @@
 				char directorySeparatorChar = System.IO.Path.DirectorySeparatorChar;
 				string matchPattern = Regex.Escape(Main.CloudPlayerPath) + "/" + Regex.Escape(fileName) + "/.+\\.map";
 				List<string> files = SocialAPI.Cloud.GetFiles(matchPattern);
+				matchPattern = Regex.Escape(Main.CloudPlayerPath) + "/" + Regex.Escape(fileName) + "/.+\\.tmap";
+				files.AddRange(SocialAPI.Cloud.GetFiles(matchPattern));
 				for (int i = 0; i < files.Count; i++)
 				{
 					string localPath = string.Concat(new object[]
@@ -128,6 +_,7 @@
 				this._isCloudSave = false;
 				this._path = playerPathFromName;
 				Main.LocalFavoriteData.SaveFavorite(this);
+				PlayerIO.MoveToLocal(base.Path, playerPathFromName);
 			}
 		}
 

