--- src/Terraria\Terraria.IO\PlayerFileData.cs
+++ src/tModLoader\Terraria.IO\PlayerFileData.cs
@@ -3,6 +_,8 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -14,6 +_,7 @@
 		private TimeSpan _playTime = TimeSpan.Zero;
 		private Stopwatch _timer = new Stopwatch();
 		private bool _isTimerActive;
+        public CustomModDataException customDataFail = null;
 
 		public Player Player
 		{
@@ -96,6 +_,7 @@
 				this._isCloudSave = true;
 				this._path = playerPathFromName;
 				Main.CloudFavoritesData.SaveFavorite(this);
+				PlayerIO.MoveToCloud(base.Path, playerPathFromName);
 			}
 		}
 
@@ -111,7 +_,8 @@
 				string fileName = base.GetFileName(false);
 				string mapPath = System.IO.Path.Combine(Main.CloudPlayerPath, fileName);
 				IEnumerable<string> enumerable = from path in SocialAPI.Cloud.GetFiles()
-				                                 where path.StartsWith(mapPath, StringComparison.CurrentCultureIgnoreCase) && path.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase)
+				                                 where path.StartsWith(mapPath, StringComparison.CurrentCultureIgnoreCase)
+													&& (path.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase) || path.EndsWith(".tmap", StringComparison.CurrentCultureIgnoreCase))
 				                                 select path;
 				foreach (string current in enumerable)
 				{
@@ -122,6 +_,7 @@
 				this._isCloudSave = false;
 				this._path = playerPathFromName;
 				Main.LocalFavoriteData.SaveFavorite(this);
+				PlayerIO.MoveToLocal(base.Path, playerPathFromName);
 			}
 		}
 

