--- src/Terraria\Terraria.IO\PlayerFileData.cs
+++ src/tModLoader\Terraria.IO\PlayerFileData.cs
@@ -3,6 +_,8 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -14,6 +_,7 @@
 		private TimeSpan _playTime = TimeSpan.Zero;
 		private Stopwatch _timer = new Stopwatch();
 		private bool _isTimerActive;
+        public CustomModDataException customDataFail = null;
 
 		public Player Player
 		{
@@ -68,6 +_,7 @@
 			string playerPathFromName = Main.GetPlayerPathFromName(this.Name, true);
 			if (FileUtilities.MoveToCloud(base.Path, playerPathFromName))
 			{
+				PlayerIO.MoveToCloud(base.Path, playerPathFromName);
 				string fileName = base.GetFileName(false);
 				string path = string.Concat(new object[]
 					{
@@ -108,10 +_,12 @@
 			string playerPathFromName = Main.GetPlayerPathFromName(this.Name, false);
 			if (FileUtilities.MoveToLocal(base.Path, playerPathFromName))
 			{
+				PlayerIO.MoveToLocal(base.Path, playerPathFromName);
 				string fileName = base.GetFileName(false);
 				string mapPath = System.IO.Path.Combine(Main.CloudPlayerPath, fileName);
 				IEnumerable<string> enumerable = from path in SocialAPI.Cloud.GetFiles()
-				                                 where path.StartsWith(mapPath, StringComparison.CurrentCultureIgnoreCase) && path.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase)
+				                                 where path.StartsWith(mapPath, StringComparison.CurrentCultureIgnoreCase)
+													&& (path.EndsWith(".map", StringComparison.CurrentCultureIgnoreCase) || path.EndsWith(".tmap", StringComparison.CurrentCultureIgnoreCase))
 				                                 select path;
 				foreach (string current in enumerable)
 				{

