--- src/Terraria\Terraria.IO\WorldFileData.cs
+++ src/tModLoader\Terraria.IO\WorldFileData.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.Localization;
+using Terraria.ModLoader.IO;
 using Terraria.Utilities;
 
 namespace Terraria.IO
@@ -167,6 +_,7 @@
 				this._isCloudSave = true;
 				this._path = worldPathFromName;
 				Main.CloudFavoritesData.SaveFavorite(this);
+				WorldIO.MoveToCloud(base.Path, worldPathFromName);
 			}
 		}
 
@@ -183,6 +_,7 @@
 				this._isCloudSave = false;
 				this._path = worldPathFromName;
 				Main.LocalFavoriteData.SaveFavorite(this);
+				WorldIO.MoveToLocal(base.Path, worldPathFromName);
 			}
 		}
 	}

