--- src/Terraria\Terraria.IO\WorldFileData.cs
+++ src/tModLoader\Terraria.IO\WorldFileData.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.Localization;
+using Terraria.ModLoader.IO;
 using Terraria.Utilities;
 
 namespace Terraria.IO
@@ -133,6 +_,7 @@
 
 			string worldPathFromName = Main.GetWorldPathFromName(Name, true);
 			if (FileUtilities.MoveToCloud(Path, worldPathFromName)) {
+				WorldIO.MoveToCloud(Path, worldPathFromName);
 				Main.LocalFavoriteData.ClearEntry(this);
 				_isCloudSave = true;
 				_path = worldPathFromName;
@@ -146,6 +_,7 @@
 
 			string worldPathFromName = Main.GetWorldPathFromName(Name, false);
 			if (FileUtilities.MoveToLocal(Path, worldPathFromName)) {
+				WorldIO.MoveToLocal(Path, worldPathFromName);
 				Main.CloudFavoritesData.ClearEntry(this);
 				_isCloudSave = false;
 				_path = worldPathFromName;
