--- src/Terraria\Terraria.IO\WorldFileData.cs
+++ src/tModLoader\Terraria.IO\WorldFileData.cs
@@ -1,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.Localization;
+using Terraria.ModLoader.IO;
 using Terraria.Utilities;
 
 namespace Terraria.IO
@@ -163,6 +_,7 @@
 			string worldPathFromName = Main.GetWorldPathFromName(this.Name, true);
 			if (FileUtilities.MoveToCloud(base.Path, worldPathFromName))
 			{
+				WorldIO.MoveToCloud(base.Path, worldPathFromName);
 				Main.LocalFavoriteData.ClearEntry(this);
 				this._isCloudSave = true;
 				this._path = worldPathFromName;
@@ -179,6 +_,7 @@
 			string worldPathFromName = Main.GetWorldPathFromName(this.Name, false);
 			if (FileUtilities.MoveToLocal(base.Path, worldPathFromName))
 			{
+				WorldIO.MoveToLocal(base.Path, worldPathFromName);
 				Main.CloudFavoritesData.ClearEntry(this);
 				this._isCloudSave = false;
 				this._path = worldPathFromName;

