--- src/Terraria\Terraria.GameContent.UI.States\UICharacterSelect.cs
+++ src/tModLoader\Terraria.GameContent.UI.States\UICharacterSelect.cs
@@ -3,6 +_,7 @@
 using Terraria.GameContent.UI.Elements;
 using Terraria.IO;
 using Terraria.UI;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.UI.States
 {
@@ -63,12 +_,7 @@
 		{
 			Main.PlaySound(10, -1, -1, 1);
 			Player player = new Player();
-			player.inventory[0].SetDefaults("Copper Shortsword");
-			player.inventory[0].Prefix(-1);
-			player.inventory[1].SetDefaults("Copper Pickaxe");
-			player.inventory[1].Prefix(-1);
-			player.inventory[2].SetDefaults("Copper Axe");
-			player.inventory[2].Prefix(-1);
+			PlayerHooks.SetStartInventory(player);
 			Main.PendingPlayer = player;
 			Main.menuMode = 2;
 		}

