--- src/Terraria\Terraria.GameContent.UI.States\UICharacterSelect.cs
+++ src/tModLoader\Terraria.GameContent.UI.States\UICharacterSelect.cs
@@ -9,6 +_,7 @@
 using Terraria.Localization;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
+using Terraria.ModLoader;
 
 namespace Terraria.GameContent.UI.States
 {
@@ -80,12 +_,7 @@
 		{
 			Main.PlaySound(10, -1, -1, 1, 1f, 0f);
 			Player player = new Player();
-			player.inventory[0].SetDefaults(3507, false);
-			player.inventory[0].Prefix(-1);
-			player.inventory[1].SetDefaults(3509, false);
-			player.inventory[1].Prefix(-1);
-			player.inventory[2].SetDefaults(3506, false);
-			player.inventory[2].Prefix(-1);
+			PlayerHooks.SetupStartInventory(player);
 			Main.PendingPlayer = player;
 			Main.menuMode = 2;
 		}

