--- src/Terraria/ReLogic/OS/Windows/Clipboard.cs
+++ src/TerrariaNetCore/ReLogic/OS/Windows/Clipboard.cs
@@ -1,12 +_,28 @@
 using System;
 using System.Threading;
+#if WINDOWS
 using System.Windows.Forms;
+#endif
 using ReLogic.OS.Base;
+#if FNA
+using SDL3;
+#endif
 
 namespace ReLogic.OS.Windows;
 
 internal class Clipboard : ReLogic.OS.Base.Clipboard
 {
+#if NETCORE
+	protected override string GetClipboard()
+	{
+		return SDL.SDL_GetClipboardText();
+	}
+
+	protected override void SetClipboard(string text)
+	{
+		SDL.SDL_SetClipboardText(text);
+	}
+#else
 	protected override string GetClipboard() => TryToGetClipboardText();
 
 	protected override void SetClipboard(string text)
@@ -77,4 +_,5 @@
 			return ApartmentState.MTA;
 		}
 	}
+#endif
 }
