--- src/decompiled/ReLogic/OS/Windows/WindowService.cs
+++ src/Terraria/ReLogic/OS/Windows/WindowService.cs
@@ -9,7 +_,7 @@
 	{
 		public float GetScaling() {
 			try {
-				IntPtr hdc = System.Drawing.Graphics.FromHwnd(IntPtr.Zero).GetHdc();
+				IntPtr hdc = NativeMethods.GetDC(IntPtr.Zero);
 				int deviceCaps = NativeMethods.GetDeviceCaps(hdc, NativeMethods.DeviceCap.VertRes);
 				return (float)NativeMethods.GetDeviceCaps(hdc, NativeMethods.DeviceCap.DesktopVertRes) / (float)deviceCaps;
 			}
@@ -27,10 +_,14 @@
 		}
 
 		public void SetUnicodeTitle(GameWindow window, string title) {
+#if WINDOWS
 			NativeMethods.WndProcCallback d = NativeMethods.DefWindowProc;
 			int dwNewLong = NativeMethods.SetWindowLong(window.Handle, -4, (int)Marshal.GetFunctionPointerForDelegate((Delegate)d));
 			window.Title = title;
 			NativeMethods.SetWindowLong(window.Handle, -4, dwNewLong);
+#else
+			SDL2.SDL.SDL_SetWindowTitle(window.Handle, title);
+#endif
 		}
 
 		public void StartFlashingIcon(GameWindow window) {
