--- src/Terraria/ReLogic/OS/Windows/WindowsPlatform.cs
+++ src/TerrariaNetCore/ReLogic/OS/Windows/WindowsPlatform.cs
@@ -19,10 +_,13 @@
 
 	public override void InitializeClientServices(IntPtr windowHandle)
 	{
+		windowHandle = Platform.SdlHandleToWindowsHandle(windowHandle);
+
 		if (_wndProcHook == null)
 			_wndProcHook = new WindowsMessageHook(windowHandle);
 
+		// NetCore change from WindowsIme to WinImm32Ime
-		RegisterService((IImeService)new WindowsIme(_wndProcHook, windowHandle));
+		RegisterService((IImeService)new WinImm32Ime(_wndProcHook, windowHandle));
 		RegisterService((IMouseNotifier)new WindowsMouseNotifier(_wndProcHook));
 	}
 
