--- src/decompiled/Terraria/UI/UserInterface.cs
+++ src/Terraria/Terraria/UI/UserInterface.cs
@@ -56,10 +_,12 @@
 		}
 
 		private void ResetState() {
+			if (!Main.dedServ) {
 			GetMousePosition();
 			_wasMouseDown = Main.mouseLeft;
 			if (_lastElementHover != null)
 				_lastElementHover.MouseOut(new UIMouseEvent(_lastElementHover, MousePosition));
+			}
 
 			_lastElementHover = null;
 			_lastElementDown = null;
