--- src/merged\Terraria.Map\MapHelper.cs
+++ src/Terraria\Terraria.Map\MapHelper.cs
@@ -1225,9 +_,9 @@
 				oldColor.B = (byte)((float)(255 - oldColor.B) * 0.5f);
 				return;
 			}
-			oldColor.R = 255 - oldColor.R;
-			oldColor.G = 255 - oldColor.G;
-			oldColor.B = 255 - oldColor.B;
+			oldColor.R = (byte)(255 - oldColor.R);
+			oldColor.G = (byte)(255 - oldColor.G);
+			oldColor.B = (byte)(255 - oldColor.B);
 		}
 
 		public static Color GetMapTileXnaColor(ref MapTile tile)
@@ -2102,12 +_,12 @@
 								binaryWriter.Write(Main.worldID);
 								binaryWriter.Write(Main.maxTilesY);
 								binaryWriter.Write(Main.maxTilesX);
-								binaryWriter.Write(470);
-								binaryWriter.Write(231);
-								binaryWriter.Write(3);
-								binaryWriter.Write(256);
-								binaryWriter.Write(256);
-								binaryWriter.Write(256);
+								binaryWriter.Write((short)470);
+								binaryWriter.Write((short)231);
+								binaryWriter.Write((short)3);
+								binaryWriter.Write((short)256);
+								binaryWriter.Write((short)256);
+								binaryWriter.Write((short)256);
 								byte b = 1;
 								byte b2 = 0;
 								int i;
@@ -2874,13 +_,13 @@
 						case 3:
 						case 4:
 						case 5:
-							num27 += num21 + (ushort)(b5 - 3);
+							num27 += (ushort)(num21 + (ushort)(b5 - 3));
 							break;
 						case 6:
 							if ((double)l < Main.worldSurface)
 							{
 								ushort num28 = (ushort)((double)num7 * ((double)l / Main.worldSurface));
-								num27 += num22 + num28;
+								num27 += (ushort)(num22 + num28);
 							}
 							else
 							{

