--- src/merged\Terraria.Map\MapTile.cs
+++ src/Terraria\Terraria.Map\MapTile.cs
@@ -31,11 +_,11 @@
 		{
 			get
 			{
-				return this._extraData & 127;
+				return (byte)(this._extraData & 127);
 			}
 			set
 			{
-				this._extraData = ((this._extraData & 128) | (value & 127));
+				this._extraData = (byte)((this._extraData & 128) | (value & 127));
 			}
 		}
 
@@ -65,12 +_,12 @@
 
 		public MapTile WithLight(byte light)
 		{
-			return new MapTile(this.Type, light, this._extraData | 128);
+			return new MapTile(this.Type, light, (byte)(this._extraData | 128));
 		}
 
 		public static MapTile Create(ushort type, byte light, byte color)
 		{
-			return new MapTile(type, light, color | 128);
+			return new MapTile(type, light, (byte)(color | 128));
 		}
 	}
 }

