--- src/decompiled/Graphics/FinalFractalHelper.cs
+++ src/Terraria/Graphics/FinalFractalHelper.cs
@@ -2,7 +_,9 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Runtime.InteropServices;
+#if XNA
 using Terraria.Graphics.Shaders;
+#endif
 
 namespace Terraria.Graphics
 {
@@ -91,6 +_,7 @@
 		}
 
 		public void Draw(Projectile proj) {
+#if XNA
 			FinalFractalProfile finalFractalProfile = GetFinalFractalProfile((int)proj.ai[1]);
 			MiscShaderData miscShaderData = GameShaders.Misc["FinalFractal"];
 			int num = 4;
@@ -104,6 +_,7 @@
 			_vertexStrip.PrepareStrip(proj.oldPos, proj.oldRot, finalFractalProfile.colorMethod, finalFractalProfile.widthMethod, -Main.screenPosition + proj.Size / 2f, proj.oldPos.Length, includeBacksides: true);
 			_vertexStrip.DrawTrail();
 			Main.pixelShader.CurrentTechnique.Passes[0].Apply();
+#endif
 		}
 
 		public static FinalFractalProfile GetFinalFractalProfile(int usedSwordId) {
