--- src/decompiled/Terraria/Graphics/FlameLashDrawer.cs
+++ src/Terraria/Terraria/Graphics/FlameLashDrawer.cs
@@ -1,5 +_,7 @@
 using Microsoft.Xna.Framework;
+#if XNA
 using Terraria.Graphics.Shaders;
+#endif
 
 namespace Terraria.Graphics
 {
@@ -9,6 +_,7 @@
 		private float transitToDark;
 
 		public void Draw(Projectile proj) {
+#if XNA
 			transitToDark = Utils.GetLerpValue(0f, 6f, proj.localAI[0], clamped: true);
 			MiscShaderData miscShaderData = GameShaders.Misc["FlameLash"];
 			miscShaderData.UseSaturation(-2f);
@@ -17,6 +_,7 @@
 			_vertexStrip.PrepareStripWithProceduralPadding(proj.oldPos, proj.oldRot, StripColors, StripWidth, -Main.screenPosition + proj.Size / 2f);
 			_vertexStrip.DrawTrail();
 			Main.pixelShader.CurrentTechnique.Passes[0].Apply();
+#endif
 		}
 
 		private Color StripColors(float progressOnStrip) {
