--- src/decompiled/Terraria/Graphics/WindowStateController.cs
+++ src/Terraria/Terraria/Graphics/WindowStateController.cs
@@ -1,15 +_,26 @@
+#if WINDOWS
 using System.Drawing;
 using System.Windows.Forms;
+#endif
 
 namespace Terraria.Graphics
 {
 	public class WindowStateController
 	{
+#if WINDOWS
 		public bool CanMoveWindowAcrossScreens => true;
+#else
+		public bool CanMoveWindowAcrossScreens => false;
+#endif
 
+#if WINDOWS
 		public string ScreenDeviceName => Main.instance.Window.ScreenDeviceName;
+#else
+		public string ScreenDeviceName => "";
+#endif
 
 		public void TryMovingToScreen(string screenDeviceName) {
+#if WINDOWS
 			if (CanMoveWindowAcrossScreens && TryGetBounds(screenDeviceName, out Rectangle bounds) && IsVisibleOnAnyScreen(bounds)) {
 				Form form = (Form)Control.FromHandle(Main.instance.Window.Handle);
 				if (WouldViewFitInScreen(form.Bounds, bounds))
@@ -45,6 +_,7 @@
 			}
 
 			return false;
+#endif
 		}
 	}
 }
