--- src/decompiled\Terraria.Graphics.Capture\CaptureInterface.cs
+++ src/merged\Terraria.Graphics.Capture\CaptureInterface.cs
@@ -1128,7 +_,17 @@
 					if (i == num2++ && flag2 && flag)
 					{
 						string fileName = Path.Combine(Main.SavePath, "Captures");
+#if LINUX
+						Process.Start(new ProcessStartInfo(fileName)
+						{
+							FileName = "open-folder",
+							Arguments = fileName,
+							UseShellExecute = true,
+							CreateNoWindow = true
+						});
+#else
 						Process.Start(fileName);
+#endif
 					}
 					if (i == num2++ && flag2)
 					{

