--- src/decompiled\Terraria.Graphics.Capture\CaptureManager.cs
+++ src/merged\Terraria.Graphics.Capture\CaptureManager.cs
@@ -11,7 +_,11 @@
 
 		private CaptureCamera _camera;
 
+#if CLIENT
 		public bool IsCapturing => _camera.IsCapturing;
+#else
+		public bool IsCapturing => false;
+#endif
 
 		public bool Active {
 			get {
@@ -34,7 +_,9 @@
 
 		public CaptureManager() {
 			_interface = new CaptureInterface();
+#if CLIENT
 			_camera = new CaptureCamera(Main.instance.GraphicsDevice);
+#endif
 		}
 
 		public void Scrolling() {
