--- src/decompiled\Terraria.Graphics.Capture\CaptureManager.cs
+++ src/merged\Terraria.Graphics.Capture\CaptureManager.cs
@@ -16,7 +_,11 @@
 		{
 			get
 			{
+#if CLIENT
 				return this._camera.IsCapturing;
+#else
+				return false;
+#endif
 			}
 		}
 
@@ -50,7 +_,9 @@
 		public CaptureManager()
 		{
 			this._interface = new CaptureInterface();
+#if CLIENT
 			this._camera = new CaptureCamera(Main.instance.GraphicsDevice);
+#endif
 		}
 
 		public void Scrolling()

