--- src/decompiled\Terraria.Graphics.Capture\CaptureInterface.cs
+++ src/merged\Terraria.Graphics.Capture\CaptureInterface.cs
@@ -1129,7 +_,17 @@
 					}
 					if (i == num2++ && (flag2 & flag))
 					{
+#if LINUX
+						Process.Start(new ProcessStartInfo(Path.Combine(Main.SavePath, "Captures"))
+						{
+							FileName = "open-folder",
+							Arguments = Path.Combine(Main.SavePath, "Captures"),
+							UseShellExecute = true,
+							CreateNoWindow = true
+						});
+#else
 						Process.Start(Path.Combine(Main.SavePath, "Captures"));
+#endif
 					}
 					if (i == num2++ && flag2)
 					{

