--- src/decompiled\Terraria.Graphics.Capture\CaptureCamera.cs
+++ src/merged\Terraria.Graphics.Capture\CaptureCamera.cs
@@ -5,9 +_,14 @@
 using System.Drawing;
 using System.Drawing.Imaging;
 using System.IO;
+#if WINDOWS
 using System.Runtime.InteropServices;
+#endif
 using System.Threading;
 using Terraria.Localization;
+#if !WINDOWS
+using Terraria.Utilities;
+#endif
 
 namespace Terraria.Graphics.Capture
 {
@@ -206,10 +_,17 @@
 					{
 						for (int j = 0; j < area.Width; j++)
 						{
-							ptr2[2] = *ptr4;
+#if WINDOWS
+							ptr2[2] = ptr4[0];
 							ptr2[1] = ptr4[1];
-							*ptr2 = ptr4[2];
+							ptr2[0] = ptr4[2];
 							ptr2[3] = ptr4[3];
+#else
+							ptr2[0] = ptr4[0];
+							ptr2[1] = ptr4[1];
+							ptr2[2] = ptr4[2];
+							ptr2[3] = ptr4[3];
+#endif
 							ptr4 += 4;
 							ptr2 += 4;
 						}
@@ -237,6 +_,7 @@
 					"Captures",
 					Path.DirectorySeparatorChar
 				}));
+#if WINDOWS
 				using (Bitmap bitmap = new Bitmap(width, height))
 				{
 					System.Drawing.Rectangle rect = new System.Drawing.Rectangle(0, 0, width, height);
@@ -247,6 +_,12 @@
 					bitmap.Save(filename, imageFormat);
 					bitmap.Dispose();
 				}
+#else
+				using (FileStream fileStream = File.Create(filename))
+				{
+					PlatformUtilities.SavePng(fileStream, width, height, width, height, this._outputData);
+ 				}
+#endif
 				result = true;
 			}
 			catch (Exception value)
@@ -269,9 +_,6 @@
 			});
 			string filename2 = Path.Combine(text, filename);
 			Directory.CreateDirectory(text);
-			using (Bitmap bitmap = new Bitmap(width, height))
-			{
-				System.Drawing.Rectangle rect = new System.Drawing.Rectangle(0, 0, width, height);
 				int elementCount = texture.Width * texture.Height * 4;
 				texture.GetData<byte>(this._outputData, 0, elementCount);
 				int num = 0;
@@ -280,22 +_,39 @@
 				{
 					for (int j = 0; j < width; j++)
 					{
+#if WINDOWS
 						byte b = this._outputData[num + 2];
 						this._outputData[num2 + 2] = this._outputData[num];
 						this._outputData[num2] = b;
 						this._outputData[num2 + 1] = this._outputData[num + 1];
 						this._outputData[num2 + 3] = this._outputData[num + 3];
+#else
+						this._outputData[num2] = this._outputData[num];
+						this._outputData[num2 + 1] = this._outputData[num + 1];
+						this._outputData[num2 + 2] = this._outputData[num + 2];
+						this._outputData[num2 + 3] = this._outputData[num + 3];
+#endif
 						num += 4;
 						num2 += 4;
 					}
 					num += texture.Width - width << 2;
 				}
+#if WINDOWS
+			using (Bitmap bitmap = new Bitmap(width, height))
+			{
+				System.Drawing.Rectangle rect = new System.Drawing.Rectangle(0, 0, width, height);
 				BitmapData bitmapData = bitmap.LockBits(rect, ImageLockMode.WriteOnly, PixelFormat.Format32bppPArgb);
 				IntPtr scan = bitmapData.Scan0;
 				Marshal.Copy(this._outputData, 0, scan, width * height * 4);
 				bitmap.UnlockBits(bitmapData);
 				bitmap.Save(filename2, imageFormat);
 			}
+#else
+			using (FileStream fileStream = File.Create(filename2))
+			{
+				PlatformUtilities.SavePng(fileStream, width, height, width, height, this._outputData);
+ 			}
+#endif
 		}
 
 		private void FinishCapture()
@@ -338,6 +_,7 @@
 
 		public void Dispose()
 		{
+#if CLIENT
 			Monitor.Enter(this._captureLock);
 			if (this._isDisposed)
 			{
@@ -352,6 +_,7 @@
 			CaptureCamera.CameraExists = false;
 			this._isDisposed = true;
 			Monitor.Exit(this._captureLock);
+#endif
 		}
 	}
 }

