--- src/Terraria\Terraria.Map\MapHelper.cs
+++ src/tModLoader\Terraria.Map\MapHelper.cs
@@ -4,6 +_,8 @@
 using System.Diagnostics;
 using System.IO;
 using Terraria.IO;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -106,7 +_,8 @@
 		private static ushort dirtPosition;
 		private static ushort rockPosition;
 		private static ushort hellPosition;
-		private static Color[] colorLookup;
+		internal static ushort modPosition;
+		internal static Color[] colorLookup;
 		private static ushort[] snowTypes;
 		private static ushort wallRangeStart;
 		private static ushort wallRangeEnd;
@@ -1129,6 +_,7 @@
 			}
 			MapHelper.hellPosition = num11;
 			MapHelper.colorLookup[(int)num11] = color7;
+			MapHelper.modPosition = (ushort)(num11 + 1);
 			MapHelper.snowTypes = new ushort[6];
 			MapHelper.snowTypes[0] = MapHelper.tileLookup[147];
 			MapHelper.snowTypes[1] = MapHelper.tileLookup[161];
@@ -2006,6 +_,7 @@
 				}
 			}
 			int num16 = num3 + num4;
+			MapLoader.ModMapOption(ref num16, i, j);
 			return MapTile.Create((ushort)num16, (byte)num2, (byte)num);
 		}
 
@@ -2060,7 +_,7 @@
 							{
 								int num = 0;
 								byte[] array = new byte[16384];
-								binaryWriter.Write(193);
+								binaryWriter.Write(194);
 								Main.MapFileMetadata.IncrementAndWrite(binaryWriter);
 								binaryWriter.Write(Main.worldName);
 								binaryWriter.Write(Main.worldID);
@@ -2162,7 +_,7 @@
 										int num5;
 										ushort num6;
 										int num7;
-										if (mapTile.Light <= 18)
+										if (mapTile.Light <= 18 || mapTile.Type >= MapHelper.modPosition)
 										{
 											flag4 = false;
 											flag3 = false;
@@ -2344,9 +_,11 @@
 								}
 								deflateStream.Dispose();
 								FileUtilities.WriteAllBytes(text, memoryStream.ToArray(), isCloudSave);
+                                //patch file: text
 							}
 						}
 					}
+					MapIO.WriteModFile(text, isCloudSave);
 				}
 				catch (Exception value)
 				{

