--- src/Terraria\Terraria.Map\WorldMap.cs
+++ src/tModLoader\Terraria.Map\WorldMap.cs
@@ -1,6 +1,7 @@
 using System;
 using System.IO;
 using Terraria.IO;
+using Terraria.ModLoader.IO;
 using Terraria.Social;
 using Terraria.Utilities;
 
@@ -115,6 +116,7 @@
 							{
 								MapHelper.LoadMapVersion2(binaryReader, num);
 							}
+							MapIO.ReadModFile(text2, isCloudSave);
 							Main.clearMap = true;
 							Main.loadMap = true;
 							Main.loadMapLock = true;

