--- src/Terraria/ReLogic/Content/Readers/IAssetReader.cs
+++ src/tModLoader/ReLogic/Content/Readers/IAssetReader.cs
@@ -1,3 +_,4 @@
+using System;
 using System.IO;
 
 namespace ReLogic.Content.Readers
@@ -5,5 +_,10 @@
 	public interface IAssetReader
 	{
 		T FromStream<T>(Stream stream) where T : class;
+
+		//Added by tML:
+
+		/// <summary> Returns an array of types that should be associated with this reader's supported formats. It's okay to return null here, if the reader's possibilities are too ambiguous. </summary>
+		Type[] GetAssociatedTypes();
 	}
 }
