--- src/TerrariaNetCore/Terraria/GameContent/Achievements/AchievementsHelper.cs
+++ src/tModLoader/Terraria/GameContent/Achievements/AchievementsHelper.cs
@@ -1,3 +_,5 @@
+using System.Linq;
+
 namespace Terraria.GameContent.Achievements;
 
 public class AchievementsHelper
@@ -18,6 +_,9 @@
 	private static bool mayhem2down;
 	private static bool mayhem3down;
 
+	/// <summary>
+	/// Tracks if the local player is actively mining to determine if <see cref="TileDestroyedCondition"/> should be satisfied or not. Modders can set to true before any custom action that should count as player controlled mining and set to false after. This prevents world updates that happen to kill tiles from satisfying <see cref="TileDestroyedCondition"/>, to avoid rewarding players for mining tiles they didn't actually mine.
+	/// </summary>
 	public static bool CurrentlyMining {
 		get {
 			return _isMining;
@@ -30,6 +_,7 @@
 	public static event ItemPickupEvent OnItemPickup;
 	public static event ItemCraftEvent OnItemCraft;
 	public static event TileDestroyedEvent OnTileDestroyed;
+	/// <summary> Called when an NPC is killed and the local player damaged/"interacted" with the NPC (<see cref="NPC.playerInteraction"/>). Unlike <see cref="ModLoader.ModNPC.OnKill"/>, this will be called on clients in multiplayer (if they damaged the NPC), making it suitable for Achievement logic. </summary>
 	public static event NPCKilledEvent OnNPCKilled;
 	public static event ProgressionEventEvent OnProgressionEvent;
 
@@ -114,10 +_,10 @@
 			}
 		}
 
-		if (player.statManaMax > 20)
+		if (player.ConsumedManaCrystals > 0)
 			Main.Achievements.GetCondition("STAR_POWER", "Use").Complete();
 
-		if (player.statLifeMax == 500 && player.statManaMax == 200)
+		if (player.ConsumedLifeCrystals == Player.LifeCrystalMax && player.ConsumedLifeFruit == Player.LifeFruitMax && player.ConsumedManaCrystals == Player.ManaCrystalMax)
 			Main.Achievements.GetCondition("TOPPED_OFF", "Use").Complete();
 
 		if (player.miscEquips[4].type > 0)
@@ -191,7 +_,7 @@
 		if (context == 17)
 			Main.Achievements.GetCondition("THE_CAVALRY", "Equip").Complete();
 
-		if ((context == 10 || context == 11) && item.wingSlot > 0)
+		if (new int[4] { 10, 11, -10, 11 }.Contains(context) && item.wingSlot > 0)
 			Main.Achievements.GetCondition("HEAD_IN_THE_CLOUDS", "Equip").Complete();
 
 		if (context == 8 && player.armor[0].stack > 0 && player.armor[1].stack > 0 && player.armor[2].stack > 0)
@@ -224,12 +_,12 @@
 		switch (eventID) {
 			case 1:
 				Main.Achievements.GetCondition("STAR_POWER", "Use").Complete();
-				if (player.statLifeMax == 500 && player.statManaMax == 200)
+				if (player.ConsumedLifeCrystals == Player.LifeCrystalMax && player.ConsumedLifeFruit == Player.LifeFruitMax && player.ConsumedManaCrystals == Player.ManaCrystalMax)
 					Main.Achievements.GetCondition("TOPPED_OFF", "Use").Complete();
 				break;
 			case 2:
 				Main.Achievements.GetCondition("GET_A_LIFE", "Use").Complete();
-				if (player.statLifeMax == 500 && player.statManaMax == 200)
+				if (player.ConsumedLifeCrystals == Player.LifeCrystalMax && player.ConsumedLifeFruit == Player.LifeFruitMax && player.ConsumedManaCrystals == Player.ManaCrystalMax)
 					Main.Achievements.GetCondition("TOPPED_OFF", "Use").Complete();
 				break;
 			case 3:
