--- src/merged\Terraria\HitTile.cs
+++ src/Terraria\Terraria\HitTile.cs
@@ -148,7 +148,15 @@
 						goto IL_EE;
 					}
 				}
-				goto IL_E1;
+				for (int k = 20; k > 1; k--)
+				{
+					int num = this.order[k - 1];
+					this.order[k - 1] = this.order[k];
+					this.order[k] = num;
+				}
+				this.order[1] = tileId;
+				goto IL_EE;
+				//goto IL_E1;
 			}
 			this.bufferLocation = this.order[20];
 			this.data[this.bufferLocation].Clear();
@@ -182,7 +190,8 @@
 					return;
 				}
 			}
-			goto IL_47;
+			this.order[20] = tileId;
+			//goto IL_47;
 		}
 
 		public void Prune()

