--- src/Terraria/Terraria.GameContent.Drawing/TileDrawing.cs
+++ src/tModLoader/Terraria.GameContent.Drawing/TileDrawing.cs
@@ -9,6 +_,7 @@
 using Terraria.GameContent.Tile_Entities;
 using Terraria.Graphics.Capture;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.ObjectData;
 using Terraria.UI;
 using Terraria.Utilities;
@@ -354,6 +_,8 @@
 						short frameY = tile.frameY;
 						if (!TextureAssets.Tile[type].IsLoaded)
 							Main.instance.LoadTiles(type);
+						if (!TileLoader.PreDraw(j, i, type, Main.spriteBatch))
+							goto PostDraw;
 
 						switch (type) {
 							case 541:
@@ -481,6 +_,8 @@
 						}
 
 						DrawSingleTile(value, solidLayer, waterStyleOverride, unscaledPosition, vector, j, i);
+						PostDraw:
+						TileLoader.PostDraw(j, i, type, Main.spriteBatch);
 					}
 				}
 			}
@@ -4355,6 +_,7 @@
 							glowColor = _martianGlow;
 						}
 
+						//*extra patch context*
 						if (num51 == 29) {
 							glowTexture = TextureAssets.GlowMask[122].Value;
 							glowSourceRect = new Rectangle(tileFrameX, tileFrameY % 38, tileWidth, tileHeight);
@@ -4364,6 +_,8 @@
 						break;
 					}
 			}
+			TileLoader.SetDrawPositions(x, y, ref tileWidth, ref addFrX, ref tileHeight); //TODO: Is this correct?
+			TileLoader.SetSpriteEffects(x, y, tileCache.type, ref tileSpriteEffect);
 		}
 
 		private bool IsWindBlocked(int x, int y) {
@@ -5120,8 +_,11 @@
 			int type = 706;
 			if (Main.waterStyle == 12)
 				type = 1147;
-			else if (Main.waterStyle > 1)
+			else if (Main.waterStyle > 1) {
 				type = 706 + Main.waterStyle - 1;
+				if (Main.waterStyle >= WaterStyleLoader.vanillaWaterCount)
+					type = WaterStyleLoader.GetWaterStyle(Main.waterStyle).GetDropletGore();
+			}
 
 			if (typeCache == 374)
 				type = 716;
