--- src/Terraria/Terraria.GameContent.Drawing/TileDrawing.cs
+++ src/tModLoader/Terraria.GameContent.Drawing/TileDrawing.cs
@@ -9,6 +_,7 @@
 using Terraria.GameContent.Tile_Entities;
 using Terraria.Graphics.Capture;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.ObjectData;
 using Terraria.UI;
 using Terraria.Utilities;
@@ -553,7 +_,7 @@
 			if (TileID.Sets.HasOutlines[drawData.typeCache])
 				GetTileOutlineInfo(tileX, tileY, drawData.typeCache, ref drawData.tileLight, ref highlightTexture, ref highlightColor);
 
-			if (_localPlayer.dangerSense && IsTileDangerous(_localPlayer, drawData.tileCache, drawData.typeCache)) {
+			if (_localPlayer.dangerSense && IsTileDangerous(_localPlayer, drawData.tileCache, drawData.typeCache, tileX, tileY)) {
 				if (drawData.tileLight.R < byte.MaxValue)
 					drawData.tileLight.R = byte.MaxValue;
 
@@ -2794,8 +_,11 @@
 			}
 		}
 
-		private static bool IsTileDangerous(Player localPlayer, Tile tileCache, ushort typeCache) {
+		private static bool IsTileDangerous(Player localPlayer, Tile tileCache, ushort typeCache, int x, int y) {
 			bool flag = false || typeCache == 135 || typeCache == 137 || typeCache == 138 || typeCache == 484 || typeCache == 141 || typeCache == 210 || typeCache == 442 || typeCache == 443 || typeCache == 444 || typeCache == 411 || typeCache == 485 || typeCache == 85;
+
+			flag |= TileLoader.Dangersense(x,y,typeCache,localPlayer);
+
 			if (tileCache.slope() == 0 && !tileCache.inActive()) {
 				flag = (flag || typeCache == 32 || typeCache == 69 || typeCache == 48 || typeCache == 232 || typeCache == 352 || typeCache == 483 || typeCache == 482 || typeCache == 481 || typeCache == 51 || typeCache == 229);
 				if (!localPlayer.fireWalk)
@@ -5744,7 +_,7 @@
 		}
 
 		private void DrawAnimatedTile_AdjustForVisionChangers(int i, int j, Tile tileCache, ushort typeCache, short tileFrameX, short tileFrameY, ref Color tileLight, bool canDoDust) {
-			if (_localPlayer.dangerSense && IsTileDangerous(_localPlayer, tileCache, typeCache)) {
+			if (_localPlayer.dangerSense && IsTileDangerous(_localPlayer, tileCache, typeCache, i, j)) {
 				if (tileLight.R < byte.MaxValue)
 					tileLight.R = byte.MaxValue;
 
