--- src/Terraria\Terraria.Graphics.Capture\CaptureCamera.cs
+++ src/tModLoader\Terraria.Graphics.Capture\CaptureCamera.cs
@@ -320,7 +320,6 @@
 
 		public void Dispose()
 		{
-#if CLIENT
 			Monitor.Enter(this._captureLock);
 			if (this._isDisposed)
 			{
@@ -335,7 +334,6 @@
 			CaptureCamera.CameraExists = false;
 			this._isDisposed = true;
 			Monitor.Exit(this._captureLock);
-#endif
 		}
 	}
 }

