--- src/TerrariaNetCore/Terraria/DataStructures/TileDrawInfo.cs
+++ src/Unified/Terraria/DataStructures/TileDrawInfo.cs
@@ -1,9 +_,10 @@
+using System;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 
 namespace Terraria.DataStructures;
 
-public class TileDrawInfo
+public ref struct TileDrawInfo
 {
 	public Tile tileCache;
 	public ushort typeCache;
@@ -21,7 +_,7 @@
 	public Texture2D glowTexture;
 	public Rectangle glowSourceRect;
 	public Color glowColor;
-	public Vector3[] colorSlices = new Vector3[9];
+	public Span<Vector3> colorSlices;
 	public Color finalColor;
 	public Color colorTint;
 }
