--- src/decompiled/Terraria/GameContent/NetModules/NetAmbienceModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetAmbienceModule.cs
@@ -1,7 +_,9 @@
 using System.IO;
 using Terraria.GameContent.Ambience;
+#if CLIENT
 using Terraria.GameContent.Skies;
 using Terraria.Graphics.Effects;
+#endif
 using Terraria.Net;
 
 namespace Terraria.GameContent.NetModules
@@ -18,6 +_,9 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
+#if SERVER
+			return false;
+#else
 			byte playerId = reader.ReadByte();
 			int seed = reader.ReadInt32();
 			SkyEntityType type = (SkyEntityType)reader.ReadByte();
@@ -26,6 +_,7 @@
 			});
 
 			return true;
+#endif
 		}
 	}
 }
