--- src/decompiled/Terraria/GameContent/NetModules/NetAmbienceModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetAmbienceModule.cs
@@ -18,6 +_,9 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
+			if(Main.dedServ)
+			return false;
+
 			byte playerId = reader.ReadByte();
 			int seed = reader.ReadInt32();
 			SkyEntityType type = (SkyEntityType)reader.ReadByte();
