--- src/decompiled/Terraria/GameContent/NetModules/NetBestiaryModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetBestiaryModule.cs
@@ -36,6 +_,9 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
+			if(Main.dedServ)
+			return false;
+
 			switch (reader.ReadByte()) {
 				case 0: {
 						short key3 = reader.ReadInt16();
@@ -59,6 +_,7 @@
 			}
 
 			return true;
+
 		}
 	}
 }
