--- src/decompiled/Terraria/GameContent/NetModules/NetCreativeUnlocksModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetCreativeUnlocksModule.cs
@@ -14,6 +_,9 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
+			if(Main.dedServ)
+			return false;
+
 			short key = reader.ReadInt16();
 			string persistentId = ContentSamples.ItemPersistentIdsByNetIds[key];
 			ushort sacrificeCount = reader.ReadUInt16();
