--- src/decompiled/Terraria/GameContent/NetModules/NetPingModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetPingModule.cs
@@ -14,7 +_,14 @@
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
 			Vector2 position = reader.ReadVector2();
+
+			if (Main.dedServ) { // #if SERVER
+				NetManager.Instance.Broadcast(Serialize(position), userId);
+				return true;
+			}
+
 			Main.Pings.Add(position);
+
 			return true;
 		}
 	}
