--- src/decompiled/Terraria/GameContent/NetModules/NetTextModule.cs
+++ src/Terraria/Terraria/GameContent/NetModules/NetTextModule.cs
@@ -39,6 +_,11 @@
 			return true;
 		}
 
-		public override bool Deserialize(BinaryReader reader, int senderPlayerId) => DeserializeAsClient(reader, senderPlayerId);
+		public override bool Deserialize(BinaryReader reader, int senderPlayerId) {
+			if (Main.dedServ) // #if SERVER
+				return DeserializeAsServer(reader, senderPlayerId);
+
+			return DeserializeAsClient(reader, senderPlayerId);
+		}
 	}
 }
