--- src/TerrariaNetCore/Terraria/Social/Base/AWorkshopTagsCollection.cs
+++ src/tModLoader/Terraria/Social/Base/AWorkshopTagsCollection.cs
@@ -2,7 +_,7 @@
 
 namespace Terraria.Social.Base;
 
-public abstract class AWorkshopTagsCollection
+public abstract partial class AWorkshopTagsCollection
 {
 	public readonly List<WorkshopTagOption> WorldTags = new List<WorkshopTagOption>();
 	public readonly List<WorkshopTagOption> ResourcePackTags = new List<WorkshopTagOption>();
