--- src/Terraria/Terraria/Social/Base/AWorkshopTagsCollection.cs
+++ src/tModLoader/Terraria/Social/Base/AWorkshopTagsCollection.cs
@@ -2,17 +_,12 @@
 
 namespace Terraria.Social.Base
 {
-	public abstract class AWorkshopTagsCollection
+	public abstract partial class AWorkshopTagsCollection
 	{
-		public readonly List<WorkshopTagOption> WorldTags = new List<WorkshopTagOption>();
-		public readonly List<WorkshopTagOption> ResourcePackTags = new List<WorkshopTagOption>();
+		public readonly List<WorkshopTagOption> ModTags = new List<WorkshopTagOption>();
 
-		protected void AddWorldTag(string tagNameKey, string tagInternalName) {
+		protected void AddModTag(string tagNameKey, string tagInternalName) {
-			WorldTags.Add(new WorkshopTagOption(tagNameKey, tagInternalName));
+			ModTags.Add(new WorkshopTagOption(tagNameKey, tagInternalName));
-		}
-
-		protected void AddResourcePackTag(string tagNameKey, string tagInternalName) {
-			ResourcePackTags.Add(new WorkshopTagOption(tagNameKey, tagInternalName));
 		}
 	}
 }
