--- src/TerrariaNetCore/Terraria/Social/Base/WorkshopItemPublishSettings.cs
+++ src/tModLoader/Terraria/Social/Base/WorkshopItemPublishSettings.cs
@@ -8,5 +_,8 @@
 	public WorkshopItemPublicSettingId Publicity;
 	public string PreviewImagePath;
 
+	// Added by TML.
+	public string ChangeNotes;
+
 	public string[] GetUsedTagsInternalNames() => UsedTags.Select((WorkshopTagOption x) => x.InternalNameForAPIs).ToArray();
 }
