--- src/TerrariaNetCore/Terraria/Social/Base/WorkshopIssueReporter.cs
+++ src/tModLoader/Terraria/Social/Base/WorkshopIssueReporter.cs
@@ -39,13 +_,24 @@
 	public void ReportInstantUploadProblem(string textKey)
 	{
 		string textValue = Language.GetTextValue(textKey);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValue);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValue);
+
 		AddReport(textValue);
 		OpenReportsScreen();
 	}
 
 	public void ReportInstantUploadProblem(string textKey, object subs)
 	{
 		string textValueWith = Language.GetTextValueWith(textKey, subs);
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValueWith);
+
 		AddReport(textValueWith);
 		OpenReportsScreen();
 	}
@@ -53,6 +_,12 @@
 	public void ReportDelayedUploadProblem(string textKey)
 	{
 		string textValue = Language.GetTextValue(textKey);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValue);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValue);
+
 		AddReport(textValue);
 		NotifyReportsScreen();
 	}
@@ -60,6 +_,12 @@
 	public void ReportDelayedUploadProblem(string textKey, object subs)
 	{
 		string textValueWith = Language.GetTextValueWith(textKey, subs);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			return;
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
@@ -72,6 +_,12 @@
 		};
 
 		string textValueWith = Language.GetTextValueWith(textKey, obj);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValueWith);
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
@@ -83,6 +_,12 @@
 		};
 
 		string textValueWith = Language.GetTextValueWith(textKey, obj);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			return;
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
