--- src/TerrariaNetCore/Terraria/Social/Base/WorkshopIssueReporter.cs
+++ src/tModLoader/Terraria/Social/Base/WorkshopIssueReporter.cs
@@ -39,12 +_,23 @@
 	public void ReportInstantUploadProblem(string textKey)
 	{
 		string textValue = Language.GetTextValue(textKey);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValue);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValue);
+
 		AddReport(textValue);
 		OpenReportsScreen();
 	}
 
 	public void ReportInstantUploadProblemFromValue(string text)
 	{
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(text);
+		if (Main.dedServ)
+			throw new NotImplementedException(text);
+
 		AddReport(text);
 		OpenReportsScreen();
 	}
@@ -52,6 +_,12 @@
 	public void ReportDelayedUploadProblem(string textKey)
 	{
 		string textValue = Language.GetTextValue(textKey);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValue);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValue);
+
 		AddReport(textValue);
 		NotifyReportsScreen();
 	}
@@ -63,6 +_,12 @@
 		};
 
 		string textValueWith = Language.GetTextValueWith(textKey, obj);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			throw new NotImplementedException(textValueWith);
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
@@ -75,6 +_,12 @@
 		};
 
 		string textValueWith = Language.GetTextValueWith(textKey, obj);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			return;
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
@@ -86,6 +_,12 @@
 		};
 
 		string textValueWith = Language.GetTextValueWith(textKey, obj);
+
+		// Logging and dedServ check added by TML.
+		Utils.LogAndConsoleErrorMessage(textValueWith);
+		if (Main.dedServ)
+			return;
+
 		AddReport(textValueWith);
 		NotifyReportsScreen();
 	}
