--- src/Terraria\Terraria.UI\IngameFancyUI.cs
+++ src/tModLoader\Terraria.UI\IngameFancyUI.cs
@@ -5,6 +_,7 @@
 using Terraria.GameContent.UI.States;
 using Terraria.Localization;
 using Terraria.UI.Gamepad;
+using Terraria.ModLoader;
 
 namespace Terraria.UI
 {
@@ -83,6 +_,14 @@
 						{
 							Main.defaultChestName = Lang.dresserType[(int)(Main.tile[player.chestX, player.chestY].frameX / 54)];
 						}
+						if (TileLoader.IsChest(Main.tile[player.chestX, player.chestY].type))
+						{
+							Main.defaultChestName = TileLoader.ModChestName(Main.tile[player.chestX, player.chestY].type);
+						}
+						if (TileLoader.IsDresser(Main.tile[player.chestX, player.chestY].type))
+						{
+							Main.defaultChestName = TileLoader.ModDresserName(Main.tile[player.chestX, player.chestY].type);
+						}
 						if (Main.npcChatText == "")
 						{
 							Main.npcChatText = Main.defaultChestName;

