--- src/Terraria/Terraria.UI/ItemSorting.cs
+++ src/tModLoader/Terraria.UI/ItemSorting.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.UI
 {
@@ -809,7 +_,7 @@
 			list.Add(ItemSortingLayers.LastTilesCommon);
 			list.Add(ItemSortingLayers.LastNotTrash);
 			list.Add(ItemSortingLayers.LastTrash);
-			for (int i = -48; i < 5043; i++) {
+			for (int i = -48; i < ItemLoader.ItemCount; i++) {
 				Item item = new Item();
 				item.netDefaults(i);
 				list2.Add(item);
