--- src/Terraria\Terraria.UI\UIElement.cs
+++ src/tModLoader\Terraria.UI\UIElement.cs
@@ -49,6 +_,22 @@
 		public event UIElement.MouseEvent OnMouseOver;
 		public event UIElement.MouseEvent OnMouseOut;
 		public event UIElement.MouseEvent OnDoubleClick;
+		public event UIElement.MouseEvent OnRightMouseDown;
+		public event UIElement.MouseEvent OnRightMouseUp;
+		public event UIElement.MouseEvent OnRightClick;
+		public event UIElement.MouseEvent OnRightDoubleClick;
+		public event UIElement.MouseEvent OnMiddleMouseDown;
+		public event UIElement.MouseEvent OnMiddleMouseUp;
+		public event UIElement.MouseEvent OnMiddleClick;
+		public event UIElement.MouseEvent OnMiddleDoubleClick;
+		public event UIElement.MouseEvent OnXButton1MouseDown;
+		public event UIElement.MouseEvent OnXButton1MouseUp;
+		public event UIElement.MouseEvent OnXButton1Click;
+		public event UIElement.MouseEvent OnXButton1DoubleClick;
+		public event UIElement.MouseEvent OnXButton2MouseDown;
+		public event UIElement.MouseEvent OnXButton2MouseUp;
+		public event UIElement.MouseEvent OnXButton2Click;
+		public event UIElement.MouseEvent OnXButton2DoubleClick;
 		public event UIElement.ScrollWheelEvent OnScrollWheel;
 
 		public bool IsMouseHovering
@@ -135,6 +_,11 @@
 				current.Parent = null;
 			}
 			this.Elements.Clear();
+		}
+
+		public bool HasChild(UIElement child)
+		{
+			return Elements.Contains(child);
 		}
 
 		public virtual void Draw(SpriteBatch spriteBatch)
@@ -160,12 +_,14 @@
 			{
 				spriteBatch.End();
 				Rectangle clippingRectangle = this.GetClippingRectangle(spriteBatch);
-				spriteBatch.GraphicsDevice.ScissorRectangle = clippingRectangle;
+				Rectangle adjustedClippingRectangle = Rectangle.Intersect(clippingRectangle, spriteBatch.GraphicsDevice.ScissorRectangle);
+				spriteBatch.GraphicsDevice.ScissorRectangle = adjustedClippingRectangle;
 				spriteBatch.Begin(SpriteSortMode.Deferred, BlendState.AlphaBlend, anisotropicClamp, DepthStencilState.None, UIElement._overflowHiddenRasterizerState, null, Main.UIScaleMatrix);
 			}
 			this.DrawChildren(spriteBatch);
 			if (overflowHidden)
 			{
+				rasterizerState = spriteBatch.GraphicsDevice.RasterizerState;
 				spriteBatch.End();
 				spriteBatch.GraphicsDevice.ScissorRectangle = scissorRectangle;
 				spriteBatch.Begin(SpriteSortMode.Deferred, BlendState.AlphaBlend, anisotropicClamp, DepthStencilState.None, rasterizerState, null, Main.UIScaleMatrix);
@@ -404,6 +_,198 @@
 			}
 		}
 
+		public virtual void RightMouseDown(UIMouseEvent evt)
+		{
+			if (this.OnRightMouseDown != null)
+			{
+				this.OnRightMouseDown(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.RightMouseDown(evt);
+			}
+		}
+
+		public virtual void RightMouseUp(UIMouseEvent evt)
+		{
+			if (this.OnRightMouseUp != null)
+			{
+				this.OnRightMouseUp(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.RightMouseUp(evt);
+			}
+		}
+
+		public virtual void RightClick(UIMouseEvent evt)
+		{
+			if (this.OnRightClick != null)
+			{
+				this.OnRightClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.RightClick(evt);
+			}
+		}
+
+		public virtual void RightDoubleClick(UIMouseEvent evt)
+		{
+			if (this.OnRightDoubleClick != null)
+			{
+				this.OnRightDoubleClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.RightDoubleClick(evt);
+			}
+		}
+
+		public virtual void MiddleMouseDown(UIMouseEvent evt)
+		{
+			if (this.OnMiddleMouseDown != null)
+			{
+				this.OnMiddleMouseDown(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.MiddleMouseDown(evt);
+			}
+		}
+
+		public virtual void MiddleMouseUp(UIMouseEvent evt)
+		{
+			if (this.OnMiddleMouseUp != null)
+			{
+				this.OnMiddleMouseUp(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.MiddleMouseUp(evt);
+			}
+		}
+
+		public virtual void MiddleClick(UIMouseEvent evt)
+		{
+			if (this.OnMiddleClick != null)
+			{
+				this.OnMiddleClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.MiddleClick(evt);
+			}
+		}
+
+		public virtual void MiddleDoubleClick(UIMouseEvent evt)
+		{
+			if (this.OnMiddleDoubleClick != null)
+			{
+				this.OnMiddleDoubleClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.MiddleDoubleClick(evt);
+			}
+		}
+
+		public virtual void XButton1MouseDown(UIMouseEvent evt)
+		{
+			if (this.OnXButton1MouseDown != null)
+			{
+				this.OnXButton1MouseDown(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton1MouseDown(evt);
+			}
+		}
+
+		public virtual void XButton1MouseUp(UIMouseEvent evt)
+		{
+			if (this.OnXButton1MouseUp != null)
+			{
+				this.OnXButton1MouseUp(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton1MouseUp(evt);
+			}
+		}
+
+		public virtual void XButton1Click(UIMouseEvent evt)
+		{
+			if (this.OnXButton1Click != null)
+			{
+				this.OnXButton1Click(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton1Click(evt);
+			}
+		}
+
+		public virtual void XButton1DoubleClick(UIMouseEvent evt)
+		{
+			if (this.OnXButton1DoubleClick != null)
+			{
+				this.OnXButton1DoubleClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton1DoubleClick(evt);
+			}
+		}
+
+		public virtual void XButton2MouseDown(UIMouseEvent evt)
+		{
+			if (this.OnXButton2MouseDown != null)
+			{
+				this.OnXButton2MouseDown(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton2MouseDown(evt);
+			}
+		}
+
+		public virtual void XButton2MouseUp(UIMouseEvent evt)
+		{
+			if (this.OnXButton2MouseUp != null)
+			{
+				this.OnXButton2MouseUp(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton2MouseUp(evt);
+			}
+		}
+
+		public virtual void XButton2Click(UIMouseEvent evt)
+		{
+			if (this.OnXButton2Click != null)
+			{
+				this.OnXButton2Click(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton2Click(evt);
+			}
+		}
+
+		public virtual void XButton2DoubleClick(UIMouseEvent evt)
+		{
+			if (this.OnXButton2DoubleClick != null)
+			{
+				this.OnXButton2DoubleClick(evt, this);
+			}
+			if (this.Parent != null)
+			{
+				this.Parent.XButton2DoubleClick(evt);
+			}
+		}
+
 		public virtual void ScrollWheel(UIScrollWheelEvent evt)
 		{
 			if (this.OnScrollWheel != null)

