--- src/merged\Terraria.GameContent.Tile_Entities\TELogicSensor.cs
+++ src/Terraria\Terraria.GameContent.Tile_Entities\TELogicSensor.cs
@@ -143,7 +_,7 @@
 			{
 				return;
 			}
-			Main.tile[(int)this.Position.X, (int)this.Position.Y].frameX = (onState ? 18 : 0);
+			Main.tile[(int)this.Position.X, (int)this.Position.Y].frameX = (onState ? (short)18 : (short)0);
 			this.On = onState;
 			if (Main.netMode == 2)
 			{
@@ -283,7 +_,7 @@
 
 		public static void GetFrame(int x, int y, TELogicSensor.LogicCheckType type, bool on)
 		{
-			Main.tile[x, y].frameX = (on ? 18 : 0);
+			Main.tile[x, y].frameX = (on ? (short)18 : (short)0);
 			switch (type)
 			{
 			case TELogicSensor.LogicCheckType.Day:

