--- src/Terraria/Terraria/NATUPNPLib/IStaticPortMappingCollection.cs
+++ src/tModLoader/Terraria/NATUPNPLib/IStaticPortMappingCollection.cs
@@ -1,4 +_,3 @@
-#if SERVER
 using System.Collections;
 using System.Reflection;
 using System.Runtime.CompilerServices;
@@ -39,4 +_,3 @@
 		public abstract IStaticPortMapping Add(int lExternalPort, string bstrProtocol, int lInternalPort, string bstrInternalClient, bool bEnabled, string bstrDescription);
 	}
 }
-#endif
