--- src/TerrariaNetCore/Terraria/Graphics/Shaders/HairShaderData.cs
+++ src/tModLoader/Terraria/Graphics/Shaders/HairShaderData.cs
@@ -1,3 +_,4 @@
+using System;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
@@ -29,57 +_,62 @@
 
 	public bool ShaderDisabled => _shaderDisabled;
 
+	[Obsolete("Removed in 1.4.5. Use Asset<Effect> version instead. Asset version works with AsyncLoad")]
+	public HairShaderData(Ref<Effect> shader, string passName)
+		: base(shader, passName)
+	{
+	}
 	public HairShaderData(Asset<Effect> shader, string passName)
 		: base(shader, passName)
 	{
 	}
 
 	private void CheckCachedParameters()
 	{
 		if (_effect == null || _effect != base.Shader) {
 			_effect = base.Shader;
 			uColor = base.Shader.GetParameter<Vector3>("uColor");
 			uSaturation = base.Shader.GetParameter<float>("uSaturation");
 			uSecondaryColor = base.Shader.GetParameter<Vector3>("uSecondaryColor");
 			uTime = base.Shader.GetParameter<float>("uTime");
 			uOpacity = base.Shader.GetParameter<float>("uOpacity");
 			uDirection = base.Shader.GetParameter<float>("uDirection");
 			uSourceRect = base.Shader.GetParameter<Vector4>("uSourceRect");
 			uDrawPosition = base.Shader.GetParameter<Vector2>("uDrawPosition");
 			uTargetPosition = base.Shader.GetParameter<Vector2>("uTargetPosition");
 			uImageSize0 = base.Shader.GetParameter<Vector2>("uImageSize0");
 			uImageSize1 = base.Shader.GetParameter<Vector2>("uImageSize1");
 		}
 	}
 
 	public virtual void Apply(Player player, DrawData? drawData = null)
 	{
 		if (!_shaderDisabled) {
 			CheckCachedParameters();
-			uColor.SetValue(_uColor);
+			uColor?.SetValue(_uColor);
-			uSaturation.SetValue(_uSaturation);
+			uSaturation?.SetValue(_uSaturation);
-			uSecondaryColor.SetValue(_uSecondaryColor);
+			uSecondaryColor?.SetValue(_uSecondaryColor);
-			uTime.SetValue(Main.GlobalTimeWrappedHourly);
+			uTime?.SetValue(Main.GlobalTimeWrappedHourly);
-			uOpacity.SetValue(_uOpacity);
+			uOpacity?.SetValue(_uOpacity);
-			uTargetPosition.SetValue(_uTargetPosition);
+			uTargetPosition?.SetValue(_uTargetPosition);
 			if (drawData.HasValue) {
 				DrawData value = drawData.Value;
 				Vector4 value2 = new Vector4(value.sourceRect.Value.X, value.sourceRect.Value.Y, value.sourceRect.Value.Width, value.sourceRect.Value.Height);
-				uSourceRect.SetValue(value2);
+				uSourceRect?.SetValue(value2);
-				uDrawPosition.SetValue(value.position);
+				uDrawPosition?.SetValue(value.position);
-				uImageSize0.SetValue(new Vector2(value.texture.Width, value.texture.Height));
+				uImageSize0?.SetValue(new Vector2(value.texture.Width, value.texture.Height));
 			}
 			else {
-				uSourceRect.SetValue(new Vector4(0f, 0f, 4f, 4f));
+				uSourceRect?.SetValue(new Vector4(0f, 0f, 4f, 4f));
 			}
 
 			if (_uImage != null) {
 				Main.graphics.GraphicsDevice.Textures[1] = _uImage.Value;
-				uImageSize1.SetValue(new Vector2(_uImage.Width(), _uImage.Height()));
+				uImageSize1?.SetValue(new Vector2(_uImage.Width(), _uImage.Height()));
 			}
 
 			if (player != null)
-				uDirection.SetValue(player.direction);
+				uDirection?.SetValue(player.direction);
 
 			Apply();
 		}
@@ -99,6 +_,14 @@
 	{
 		if (!Main.dedServ)
 			_uImage = Main.Assets.Request<Texture2D>(path);
+
+		return this;
+	}
+
+	public HairShaderData UseImage(Asset<Texture2D> asset)
+	{
+		if (!Main.dedServ)
+			_uImage = asset;
 
 		return this;
 	}
