--- src/TerrariaNetCore/Terraria/Graphics/Shaders/HairShaderData.cs
+++ src/tModLoader/Terraria/Graphics/Shaders/HairShaderData.cs
@@ -1,3 +_,4 @@
+using System;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
@@ -17,10 +_,15 @@
 
 	public bool ShaderDisabled => _shaderDisabled;
 
+	[Obsolete("Removed in 1.4.5. Use Asset<Effect> version instead. Asset version works with AsyncLoad")]
 	public HairShaderData(Ref<Effect> shader, string passName)
 		: base(shader, passName)
 	{
 	}
+	public HairShaderData(Asset<Effect> shader, string passName)
+		: base(shader, passName)
+	{
+	}
 
 	public virtual void Apply(Player player, DrawData? drawData = null)
 	{
@@ -67,6 +_,12 @@
 	public HairShaderData UseImage(string path)
 	{
 		_uImage = Main.Assets.Request<Texture2D>(path);
+		return this;
+	}
+
+	public HairShaderData UseImage(Asset<Texture2D> asset)
+	{
+		_uImage = asset;
 		return this;
 	}
 
