--- src/TerrariaNetCore/Terraria/Graphics/Shaders/MiscShaderData.cs
+++ src/tModLoader/Terraria/Graphics/Shaders/MiscShaderData.cs
@@ -193,5 +_,35 @@
 		_uImage2Tex = texture;
 		_uImage2 = null;
 		return this;
 	}
+
+	public MiscShaderData UseImage0(Asset<Texture2D> asset)
+	{
+		if (Main.dedServ)
+			return this;
+
+		_uImage0Tex = null;
+		_uImage0 = asset;
+		return this;
+	}
+
+	public MiscShaderData UseImage1(Asset<Texture2D> asset)
+	{
+		if (Main.dedServ)
+			return this;
+
+		_uImage1Tex = null;
+		_uImage1 = asset;
+		return this;
+	}
+
+	public MiscShaderData UseImage2(Asset<Texture2D> asset)
+	{
+		if (Main.dedServ)
+			return this;
+
+		_uImage2Tex = null;
+		_uImage2 = asset;
+ 		return this;
+	}
 
