--- src/TerrariaNetCore/Terraria/Graphics/Shaders/MiscShaderData.cs
+++ src/tModLoader/Terraria/Graphics/Shaders/MiscShaderData.cs
@@ -105,6 +_,24 @@
 		return this;
 	}
 
+	public MiscShaderData UseImage0(Asset<Texture2D> asset)
+	{
+		_uImage0 = asset;
+		return this;
+	}
+
+	public MiscShaderData UseImage1(Asset<Texture2D> asset)
+	{
+		_uImage1 = asset;
+		return this;
+	}
+
+	public MiscShaderData UseImage2(Asset<Texture2D> asset)
+	{
+		_uImage2 = asset;
+		return this;
+	}
+
 	private static bool IsPowerOfTwo(int n) => (int)Math.Ceiling(Math.Log(n) / Math.Log(2.0)) == (int)Math.Floor(Math.Log(n) / Math.Log(2.0));
 
 	public MiscShaderData UseOpacity(float alpha)
