--- src/decompiled\Terraria.UI.Chat\ChatManager.cs
+++ src/merged\Terraria.UI.Chat\ChatManager.cs
@@ -52,7 +_,11 @@
 
 		public static void Register<T>(params string[] names) where T : ITagHandler, new()
 		{
+#if MAC
+			T t = Activator.CreateInstance<T>();
+#else
 			T t = (default(T) == null) ? Activator.CreateInstance<T>() : default(T);
+#endif
 			for (int i = 0; i < names.Length; i++)
 			{
 				ChatManager._handlers[names[i].ToLower()] = t;

