--- src/decompiled/Terraria/Social/Steam/AchievementsSocialModule.cs
+++ src/Terraria/Terraria/Social/Steam/AchievementsSocialModule.cs
@@ -16,7 +_,7 @@
 
 		public override void Initialize() {
 			_userStatsReceived = Callback<UserStatsReceived_t>.Create(OnUserStatsReceived);
-			SteamUserStats.RequestCurrentStats();
+			var t = SteamUserStats.RequestCurrentStats();
 			while (!_areStatsReceived) {
 				CoreSocialModule.Pulse();
 				Thread.Sleep(10);
