--- src/decompiled/Terraria/Social/Steam/CoreSocialModule.cs
+++ src/Terraria/Terraria/Social/Steam/CoreSocialModule.cs
@@ -2,7 +_,9 @@
 using Steamworks;
 using System;
 using System.Threading;
+#if WINDOWS
 using System.Windows.Forms;
+#endif
 using Terraria.Localization;
 
 namespace Terraria.Social.Steam
@@ -20,13 +_,17 @@
 
 		public void Initialize() {
 			_instance = this;
+#if CLIENT
 			if (SteamAPI.RestartAppIfNecessary(new AppId_t(105600u))) {
 				Environment.Exit(1);
 				return;
 			}
+#endif
 
 			if (!SteamAPI.Init()) {
+#if WINDOWS
 				MessageBox.Show(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
+#endif
 				Environment.Exit(1);
 			}
 
@@ -39,8 +_,13 @@
 			thread2.Start();
 			Main.OnTickForThirdPartySoftwareOnly += PulseSteamTick;
 			Main.OnTickForThirdPartySoftwareOnly += PulseSteamCallback;
+#if CLIENT
 			if (Platform.IsOSX)
 				_onOverlayActivated = Callback<GameOverlayActivated_t>.Create(OnOverlayActivated);
+#if MAC
+			this._onOverlayActivated = Callback<GameOverlayActivated_t>.Create(new Callback<GameOverlayActivated_t>.DispatchDelegate(this.OnOverlayActivated));
+#endif
+#endif
 		}
 
 		public void PulseSteamTick() {
@@ -86,7 +_,12 @@
 		}
 
 		public void Shutdown() {
+#if WINDOWS
 			Application.ApplicationExit += delegate {
+#else
+			IsSteamValid = false;
+			AppDomain.CurrentDomain.ProcessExit += delegate {
+#endif
 				IsSteamValid = false;
 			};
 		}
