--- src/decompiled/Terraria/Social/Steam/CoreSocialModule.cs
+++ src/Terraria/Terraria/Social/Steam/CoreSocialModule.cs
@@ -24,7 +_,7 @@
 
 		public void Initialize() {
 			_instance = this;
-			if (SteamAPI.RestartAppIfNecessary(new AppId_t(105600u))) {
+			if (!Main.dedServ && SteamAPI.RestartAppIfNecessary(new AppId_t(105600u))) {
 				Environment.Exit(1);
 				return;
 			}
@@ -43,7 +_,8 @@
 			thread2.Start();
 			Main.OnTickForThirdPartySoftwareOnly += PulseSteamTick;
 			Main.OnTickForThirdPartySoftwareOnly += PulseSteamCallback;
-			if (Platform.IsOSX)
+
+			if (!Main.dedServ && Platform.IsOSX)
 				_onOverlayActivated = Callback<GameOverlayActivated_t>.Create(OnOverlayActivated);
 		}
 
@@ -100,7 +_,12 @@
 		}
 
 		public void Shutdown() {
+#if WINDOWS
 			Application.ApplicationExit += delegate {
+#else
+			IsSteamValid = false;
+			AppDomain.CurrentDomain.ProcessExit += delegate {
+#endif
 				IsSteamValid = false;
 			};
 		}
