--- src/decompiled/Terraria/GameContent/Profiles.cs
+++ src/Terraria/Terraria/GameContent/Profiles.cs
@@ -17,6 +_,10 @@
 			public LegacyNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+
+				if (Main.dedServ) // #if SERVER
+					return;
+
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultParty = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Party", AssetRequestMode.DoNotLoad);
 			}
@@ -48,6 +_,10 @@
 			public TransformableNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+
+				if (Main.dedServ) // #if SERVER
+					return;
+
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultTransformed = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Transformed", AssetRequestMode.DoNotLoad);
 				_defaultCredits = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Credits", AssetRequestMode.DoNotLoad);
@@ -85,6 +_,10 @@
 				_npcBaseName = npcBaseName;
 				_variantHeadIDs = variantHeadIds;
 				_variants = variantTextureNames;
+
+				if (Main.dedServ) // #if SERVER
+					return;
+				
 				string[] variants = _variants;
 				foreach (string str in variants) {
 					string text = _rootFilePath + "_" + str;
