--- src/decompiled/Terraria/GameContent/Profiles.cs
+++ src/Terraria/Terraria/GameContent/Profiles.cs
@@ -17,8 +_,10 @@
 			public LegacyNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+#if CLIENT
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultParty = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Party", AssetRequestMode.DoNotLoad);
+#endif
 			}
 
 			public int RollVariation() => 0;
@@ -48,9 +_,11 @@
 			public TransformableNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+#if CLIENT
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultTransformed = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Transformed", AssetRequestMode.DoNotLoad);
 				_defaultCredits = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Credits", AssetRequestMode.DoNotLoad);
+#endif
 			}
 
 			public int RollVariation() => 0;
@@ -88,7 +_,9 @@
 				string[] variants = _variants;
 				foreach (string str in variants) {
 					string text = _rootFilePath + "_" + str;
+#if CLIENT
 					_variantTextures[text] = Main.Assets.Request<Texture2D>(text, AssetRequestMode.DoNotLoad);
+#endif
 				}
 			}
 
