--- src/decompiled/Terraria/GameContent/Profiles.cs
+++ src/Terraria/Terraria/GameContent/Profiles.cs
@@ -17,6 +_,9 @@
 			public LegacyNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+				if (Main.dedServ)
+					return;
+
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultParty = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Party", AssetRequestMode.DoNotLoad);
 			}
@@ -48,6 +_,9 @@
 			public TransformableNPCProfile(string npcFileTitleFilePath, int defaultHeadIndex) {
 				_rootFilePath = npcFileTitleFilePath;
 				_defaultVariationHeadIndex = defaultHeadIndex;
+				if (Main.dedServ)
+					return;
+
 				_defaultNoAlt = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default", AssetRequestMode.DoNotLoad);
 				_defaultTransformed = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Transformed", AssetRequestMode.DoNotLoad);
 				_defaultCredits = Main.Assets.Request<Texture2D>(npcFileTitleFilePath + "_Default_Credits", AssetRequestMode.DoNotLoad);
@@ -88,6 +_,7 @@
 				string[] variants = _variants;
 				foreach (string str in variants) {
 					string text = _rootFilePath + "_" + str;
+					if(!Main.dedServ)
 					_variantTextures[text] = Main.Assets.Request<Texture2D>(text, AssetRequestMode.DoNotLoad);
 				}
 			}
