--- src/merged\Terraria.GameContent.Biomes\CaveHouseBiome.cs
+++ src/Terraria\Terraria.GameContent.Biomes\CaveHouseBiome.cs
@@ -796,7 +_,7 @@
 				{
 					CaveHouseBiome.BuildData.Default.Wall
 				}),
-				((double)room.Y > Main.worldSurface) ? new Actions.ClearWall(true) : new Actions.PlaceWall(2, true)
+				((double)room.Y > Main.worldSurface) ? (GenAction)new Actions.ClearWall(true) : (GenAction)new Actions.PlaceWall(2, true)
 			}));
 			WorldUtils.Gen(new Point(room.X, room.Y), new Shapes.Rectangle(room.Width, room.Height), Actions.Chain(new GenAction[]
 			{
@@ -849,7 +_,7 @@
 			{
 				new Modifiers.Dither(0.85000002384185791),
 				new Modifiers.Blotches(2, 0.8),
-				((double)room.Y > Main.worldSurface) ? new Actions.ClearWall(true) : new Actions.PlaceWall(40, true)
+				((double)room.Y > Main.worldSurface) ? (GenAction)new Actions.ClearWall(true) : (GenAction)new Actions.PlaceWall(40, true)
 			}));
 		}
 

