--- src/merged\Terraria.GameContent.Biomes\DesertBiome.cs
+++ src/Terraria\Terraria.GameContent.Biomes\DesertBiome.cs
@@ -257,7 +_,7 @@
 					tile.liquid = 0;
 					Tile testTile = GenBase._tiles[num19, num20 + 1];
 					Tile testTile2 = GenBase._tiles[num19, num20 + 2];
-					tile.type = ((WorldGen.SolidTile(testTile) && WorldGen.SolidTile(testTile2)) ? 53 : 397);
+					tile.type = ((WorldGen.SolidTile(testTile) && WorldGen.SolidTile(testTile2)) ? (ushort)53 : (ushort)397);
 					if (m > num10 + 5)
 					{
 						tile.wall = 187;

