--- src/merged\Terraria.GameContent.Biomes\MiningExplosivesBiome.cs
+++ src/Terraria\Terraria.GameContent.Biomes\MiningExplosivesBiome.cs
@@ -15,13 +_,13 @@
 			}
 			ushort type = Utils.SelectRandom<ushort>(GenBase._random, new ushort[]
 			{
-				(WorldGen.goldBar == 19) ? 8 : 169,
-				(WorldGen.silverBar == 21) ? 9 : 168,
-				(WorldGen.ironBar == 22) ? 6 : 167,
-				(WorldGen.copperBar == 20) ? 7 : 166
+				(WorldGen.goldBar == 19) ? (ushort)8 : (ushort)169,
+				(WorldGen.silverBar == 21) ? (ushort)9 : (ushort)168,
+				(WorldGen.ironBar == 22) ? (ushort)6 : (ushort)167,
+				(WorldGen.copperBar == 20) ? (ushort)7 : (ushort)166
 			});
 			double num = GenBase._random.NextDouble() * 2.0 - 1.0;
-			if (!WorldUtils.Find(origin, Searches.Chain((num > 0.0) ? new Searches.Right(40) : new Searches.Left(40), new GenCondition[]
+			if (!WorldUtils.Find(origin, Searches.Chain((num > 0.0) ? (GenSearch)new Searches.Right(40) : (GenSearch)new Searches.Left(40), new GenCondition[]
 			{
 				new Conditions.IsSolid()
 			}), out origin))

