--- src/Terraria/Terraria/nativefiledialog.cs
+++ src/TerrariaNetCore/Terraria/nativefiledialog.cs
@@ -58,8 +_,16 @@
 		return result;
 	}
 
+#if NETCORE
+	private unsafe static void free(IntPtr ptr)
+	{
+		// Cross-platform C deallocation.
+		NativeMemory.Free((void*)ptr);
+	}
+#else
 	[DllImport("msvcrt", CallingConvention = CallingConvention.Cdecl)]
 	private static extern void free(IntPtr ptr);
+#endif
 
 	[DllImport("nfd", CallingConvention = CallingConvention.Cdecl, EntryPoint = "NFD_OpenDialog")]
 	private unsafe static extern nfdresult_t INTERNAL_NFD_OpenDialog(byte* filterList, byte* defaultPath, out IntPtr outPath);
