--- src/Terraria/Terraria/Program.cs
+++ src/TerrariaNetCore/Terraria/Program.cs
@@ -165,7 +_,19 @@
 		SavePath = (LaunchParameters.ContainsKey("-savedirectory") ? LaunchParameters["-savedirectory"] : Platform.Get<IPathService>().GetStoragePath("Terraria"));
 #if SERVER
 		Main.dedServ = true;
+#else
+		// Unify server and client dll via launch param
+		Main.dedServ = LaunchParameters.ContainsKey("-server");
 #endif
+
+		// Remove bad environment variables that might be set by other app installs
+		if (Main.dedServ) {
+			Environment.SetEnvironmentVariable("FNA_PLATFORM_BACKEND", "NONE");
+		}
+		else {
+			Environment.SetEnvironmentVariable("FNA_PLATFORM_BACKEND", "SDL3");
+		}
+
 		ThreadPool.SetMinThreads(8, 8);
 		InitializeConsoleOutput();
 		SetupLogging();
@@ -200,8 +_,8 @@
 			Main.OnEnginePreload += StartForceLoad;
 			if (Main.dedServ)
 				main.DedServ();
-
-			main.Run();
+			else
+				main.Run();
 		}
 		catch (Exception e) {
 			DisplayException(e);
@@ -226,8 +_,15 @@
 			}
 		}
 
+		// TODO: Do we always want to default to SDLGPU? Should OpenGL be
+		// preferred on some systems?
+#if NETCORE
+		if (!flag)
+			Environment.SetEnvironmentVariable("FNA3D_FORCE_DRIVER", "SDLGPU");
+#else
 		if (!flag)
 			Environment.SetEnvironmentVariable("FNA3D_FORCE_DRIVER", Platform.IsWindows ? "D3D12" : (Platform.IsOSX ? "Metal" : "Vulkan"));
+#endif
 	}
 
 	private static void DisplayException(Exception e)
