--- src/Terraria/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
@@ -38,12 +_,18 @@
 		private UICharacterNameButton _previewImagePathPlate;
 		private Texture2D _previewImageTransientTexture;
 		private UIImage _previewImageUIElement;
-		private string _previewImagePath;
+		protected string _previewImagePath;
 		private Asset<Texture2D> _defaultPreviewImageTexture;
 		private UIElement _steamDisclaimerButton;
 		private UIText _disclaimerText;
 		private UIGamepadHelper _helper;
 
+		// Added by TML.
+		public UIState PreviousUIState {
+			get => _previousUIState;
+			set => _previousUIState = value;
+		}
+
 		public AWorkshopPublishInfoState(UIState stateToGoBackTo, TPublishedObjectType dataObject) {
 			_previousUIState = stateToGoBackTo;
 			_dataObject = dataObject;
@@ -232,7 +_,11 @@
 		}
 
 		private void steamDisclaimerText_OnClick(UIMouseEvent evt, UIElement listeningElement) {
-			ProcessStart("https://steamcommunity.com/sharedfiles/workshoplegalagreement");
+			var ps = new ProcessStartInfo("https://steamcommunity.com/sharedfiles/workshoplegalagreement") {
+				UseShellExecute = true,
+				Verb = "open"
+			};
+			Process.Start(ps);
 		}
 
 		public static void ProcessStart(string path) {
