--- src/TerrariaNetCore/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
@@ -10,6 +_,7 @@
 using Terraria.GameContent.UI.Elements;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader.Core;
 using Terraria.Social;
 using Terraria.Social.Base;
 using Terraria.UI;
@@ -34,20 +_,37 @@
 	private UIElement _publishButton;
 	private WorkshopItemPublicSettingId _optionPublicity = WorkshopItemPublicSettingId.Public;
 	private GroupOptionButton<WorkshopItemPublicSettingId>[] _publicityOptions;
-	private List<GroupOptionButton<WorkshopTagOption>> _tagOptions;
+	protected List<GroupOptionButton<WorkshopTagOption>> _tagOptions;
-	private UICharacterNameButton _previewImagePathPlate;
+	protected UICharacterNameButton _previewImagePathPlate;
-	private Texture2D _previewImageTransientTexture;
+	protected Texture2D _previewImageTransientTexture;
-	private UIImage _previewImageUIElement;
+	protected UIImage _previewImageUIElement;
-	private string _previewImagePath;
+	protected string _previewImagePath;
 	private Asset<Texture2D> _defaultPreviewImageTexture;
 	private UIElement _steamDisclaimerButton;
-	private UIText _disclaimerText;
+	protected UIText _disclaimerText;
 	private UIGamepadHelper _helper;
 
+	// Added by TML [[
+	protected UIText _tMLDisclaimerText;
+	protected UIElement _tMLDisclaimerButton;
+
+	private bool _isMod;
+	private bool _isUpdate;
+
+	public UIState PreviousUIState {
+		get => _previousUIState;
+		set => _previousUIState = value;
+	}
+	// ]]
+
 	public AWorkshopPublishInfoState(UIState stateToGoBackTo, TPublishedObjectType dataObject)
 	{
 		_previousUIState = stateToGoBackTo;
 		_dataObject = dataObject;
+
+		if (dataObject is TmodFile) {
+			_isMod = true;
+		}
 	}
 
 	public override void OnInitialize()
@@ -67,7 +_,6 @@
 		uIPanel.Height.Set(-num3, 1f);
 		uIPanel.BackgroundColor = new Color(33, 43, 79) * 0.8f;
 		AddBackButton(num, uIElement);
-		AddPublishButton(num, uIElement);
 		int num5 = 6 + num4;
 		UIList uiList = AddUIList(uIPanel, num5);
 		FillUIList(uiList);
@@ -76,17 +_,28 @@
 		uIElement.Append(uIPanel);
 		Append(uIElement);
 		SetDefaultOptions();
+		AddPublishButton(num, uIElement);
 	}
 
 	private void SetDefaultOptions()
 	{
+		// Moved from below.
+		SetTagsFromFoundEntry();
+
+		// Seemingly moved to SetTagsFromFoundEntry
+		/*
 		_optionPublicity = WorkshopItemPublicSettingId.Public;
+		*/
+
 		GroupOptionButton<WorkshopItemPublicSettingId>[] publicityOptions = _publicityOptions;
 		for (int i = 0; i < publicityOptions.Length; i++) {
 			publicityOptions[i].SetCurrentOption(_optionPublicity);
 		}
 
+		/*
 		SetTagsFromFoundEntry();
+		*/
+
 		UpdateImagePreview();
 	}
 
@@ -100,6 +_,13 @@
 		uIElement.SetPadding(0f);
 		uiList.Add(uIElement);
 		uiList.Add(CreateSteamDisclaimer("disclaimer"));
+
+		// Added by tML
+		if (_isMod) {
+			uiList.Add((this as WorkshopPublishInfoStateForMods).CreateTmlDisclaimer("disclaimer"));
+			(this as WorkshopPublishInfoStateForMods).AddNonModOwnerPublishWarning(uiList);
+		}
+
 		uiList.Add(CreatePreviewImageSelectionPanel("image"));
 		uiList.Add(CreatePublicSettingsRow(0f, 44f, "public"));
 		uiList.Add(CreateTagOptionsPanel(0f, 44, "tags"));
@@ -124,7 +_,10 @@
 		uIElement.SetPadding(0f);
 		obj.Append(uIElement);
 		float num = 86f;
+		if (!_isMod)
-		_defaultPreviewImageTexture = Main.Assets.Request<Texture2D>("Images/UI/Workshop/DefaultPreviewImage");
+			_defaultPreviewImageTexture = Main.Assets.Request<Texture2D>("Images/UI/Workshop/DefaultPreviewImage");
+		else
+			_defaultPreviewImageTexture = Main.Assets.Request<Texture2D>("Images/UI/DefaultResourcePackIcon");
 		UIImage uIImage = new UIImage(_defaultPreviewImageTexture) {
 			Width = new StyleDimension(-4f, 1f),
 			Height = new StyleDimension(-4f, 1f),
@@ -140,7 +_,8 @@
 		};
 
 		uIElement.Append(uIImage);
+		if (!_isMod)
-		uIElement.Append(element);
+			uIElement.Append(element);
 		_previewImageUIElement = uIImage;
 		UICharacterNameButton uICharacterNameButton = new UICharacterNameButton(Language.GetText("Workshop.PreviewImagePathTitle"), Language.GetText("Workshop.PreviewImagePathEmpty"), Language.GetText("Workshop.PreviewImagePathDescription")) {
 			Width = StyleDimension.FromPixelsAndPercent(0f - num, 1f),
@@ -161,6 +_,8 @@
 		if (!TryFindingTags(out var info))
 			return;
 
+		_isUpdate = true;
+
 		if (info.tags != null) {
 			foreach (GroupOptionButton<WorkshopTagOption> tagOption in _tagOptions) {
 				bool flag = info.tags.Contains(tagOption.OptionValue.InternalNameForAPIs);
@@ -169,10 +_,14 @@
 			}
 		}
 
+		/*
 		GroupOptionButton<WorkshopItemPublicSettingId>[] publicityOptions = _publicityOptions;
 		for (int i = 0; i < publicityOptions.Length; i++) {
 			publicityOptions[i].SetCurrentOption(info.publicity);
 		}
+		*/
+
+		_optionPublicity = info.publicity;
 	}
 
 	protected abstract bool TryFindingTags(out FoundWorkshopEntryInfo info);
@@ -328,41 +_,46 @@
 	private UIElement CreatePublicSettingsRow(float accumulatedHeight, float height, string tagGroup)
 	{
 		CreateStylizedCategoryPanel(height, "Workshop.CategoryTitlePublicity", out var entirePanel, out var innerPanel);
-		WorkshopItemPublicSettingId[] array = new WorkshopItemPublicSettingId[3] {
+		WorkshopItemPublicSettingId[] array = new WorkshopItemPublicSettingId[4] {
 			WorkshopItemPublicSettingId.Public,
 			WorkshopItemPublicSettingId.FriendsOnly,
-			WorkshopItemPublicSettingId.Private
+			WorkshopItemPublicSettingId.Private,
+			WorkshopItemPublicSettingId.Unlisted
 		};
 
-		LocalizedText[] array2 = new LocalizedText[3] {
+		LocalizedText[] array2 = new LocalizedText[4] {
 			Language.GetText("Workshop.SettingsPublicityPublic"),
 			Language.GetText("Workshop.SettingsPublicityFriendsOnly"),
-			Language.GetText("Workshop.SettingsPublicityPrivate")
+			Language.GetText("Workshop.SettingsPublicityPrivate"),
+			Language.GetText("tModLoader.SettingsPublicityUnlisted")
 		};
 
-		LocalizedText[] array3 = new LocalizedText[3] {
+		LocalizedText[] array3 = new LocalizedText[4] {
 			Language.GetText("Workshop.SettingsPublicityPublicDescription"),
 			Language.GetText("Workshop.SettingsPublicityFriendsOnlyDescription"),
-			Language.GetText("Workshop.SettingsPublicityPrivateDescription")
+			Language.GetText("Workshop.SettingsPublicityPrivateDescription"),
+			Language.GetText("tModLoader.SettingsPublicityUnlistedDescription")
 		};
 
-		Color[] array4 = new Color[3] {
+		Color[] array4 = new Color[4] {
+			Color.White,
 			Color.White,
 			Color.White,
 			Color.White
 		};
 
-		string[] array5 = new string[3] {
+		string[] array5 = new string[4] {
 			"Images/UI/Workshop/PublicityPublic",
 			"Images/UI/Workshop/PublicityFriendsOnly",
-			"Images/UI/Workshop/PublicityPrivate"
+			"Images/UI/Workshop/PublicityPrivate",
+			"Images/UI/WorldCreation/IconRandomName" //tmod added
 		};
 
 		float num = 0.98f;
 		GroupOptionButton<WorkshopItemPublicSettingId>[] array6 = new GroupOptionButton<WorkshopItemPublicSettingId>[array.Length];
 		for (int i = 0; i < array6.Length; i++) {
-			GroupOptionButton<WorkshopItemPublicSettingId> groupOptionButton = new GroupOptionButton<WorkshopItemPublicSettingId>(array[i], array2[i], array3[i], array4[i], array5[i], 1f, 1f, 16f);
+			GroupOptionButton<WorkshopItemPublicSettingId> groupOptionButton = new GroupOptionButton<WorkshopItemPublicSettingId>(array[i], array2[i], array3[i], array4[i], array5[i], 0.8f, 1f, 16f);
-			groupOptionButton.Width = StyleDimension.FromPixelsAndPercent(-4 * (array6.Length - 1), 1f / (float)array6.Length * num);
+			groupOptionButton.Width = StyleDimension.FromPixelsAndPercent(-2 * (array6.Length - 1), 1f / (float)array6.Length * num);
 			groupOptionButton.HAlign = (float)i / (float)(array6.Length - 1);
 			groupOptionButton.Left = StyleDimension.FromPercent((1f - num) * (1f - groupOptionButton.HAlign * 2f));
 			groupOptionButton.Top.Set(accumulatedHeight, 0f);
@@ -382,22 +_,47 @@
 	{
 		List<WorkshopTagOption> tagsToShow = GetTagsToShow();
 		int num = 3;
-		int num2 = (int)Math.Ceiling((float)tagsToShow.Count / (float)num);
+		int num2 = (int)Math.Ceiling((float)tagsToShow.Count / (float)num) + 1;
 		int num3 = heightPerRow * num2;
 		CreateStylizedCategoryPanel(num3, "Workshop.CategoryTitleTags", out var entirePanel, out var innerPanel);
 		float num4 = 0.98f;
+		int additionalIOffset = 0;
 		List<GroupOptionButton<WorkshopTagOption>> list = new List<GroupOptionButton<WorkshopTagOption>>();
 		for (int i = 0; i < tagsToShow.Count; i++) {
 			WorkshopTagOption workshopTagOption = tagsToShow[i];
 			GroupOptionButton<WorkshopTagOption> groupOptionButton = new GroupOptionButton<WorkshopTagOption>(workshopTagOption, Language.GetText(workshopTagOption.NameKey), Language.GetText(workshopTagOption.NameKey + "Description"), Color.White, null, 1f, 0.5f, 16f);
 			groupOptionButton.ShowHighlightWhenSelected = false;
 			groupOptionButton.SetCurrentOption(null);
-			int num5 = i / num;
-			int num6 = i - num5 * num;
+			int num5 = (i + additionalIOffset) / num;
+			if (workshopTagOption.InternalNameForAPIs == "English") {
+				num5++;
+				if ((i + additionalIOffset) % num != 0) {
+					additionalIOffset += 3;
+					num5++;
+				}
+				additionalIOffset += 3 - ((i + additionalIOffset) % num);
+				AddHorizontalSeparator(entirePanel, 44 + num5 * heightPerRow, 4);
+				UIText uIText = new UIText(Language.GetText("tModLoader.TagsCategoryLanguage")) {
+					HAlign = 0f,
+					VAlign = 0f,
+					Width = StyleDimension.FromPixelsAndPercent(-40f, 1f),
+					Height = StyleDimension.FromPixelsAndPercent(44, 0f),
+					Top = StyleDimension.FromPixelsAndPercent(5f + num5 * heightPerRow, 0f)
+				};
+				uIText.PaddingLeft = 20f;
+				uIText.PaddingRight = 20f;
+				uIText.PaddingTop = 6f;
+				uIText.IsWrapped = false;
+				entirePanel.Append(uIText);
+			}
+			num5 = (i + additionalIOffset) / num;
+			int num6 = (i + additionalIOffset) % num;
 			groupOptionButton.Width = StyleDimension.FromPixelsAndPercent(-4 * (num - 1), 1f / (float)num * num4);
 			groupOptionButton.HAlign = (float)num6 / (float)(num - 1);
 			groupOptionButton.Left = StyleDimension.FromPercent((1f - num4) * (1f - groupOptionButton.HAlign * 2f));
 			groupOptionButton.Top.Set(num5 * heightPerRow, 0f);
+			// If we want to enforce TagsTranslation meaning later:
+			// if(workshopTagOption.NameKey != "tModLoader.TagsTranslation")
 			groupOptionButton.OnLeftMouseDown += ClickTagOption;
 			groupOptionButton.OnMouseOver += ShowOptionDescription;
 			groupOptionButton.OnMouseOut += ClearOptionDescription;
@@ -593,7 +_,8 @@
 
 	private void AddPublishButton(int backButtonYLift, UIElement outerContainer)
 	{
-		UITextPanel<LocalizedText> uITextPanel = new UITextPanel<LocalizedText>(Language.GetText("Workshop.Publish"), 0.7f, large: true);
+		UITextPanel<LocalizedText> uITextPanel = new UITextPanel<LocalizedText>(Language.GetText(_isUpdate ? "tModLoader.PublishUpdateButton" : "tModLoader.PublishNewButton"), 0.7f, large: true);
+		uITextPanel.TextColor = _isUpdate ? Color.White : Color.Orange;
 		uITextPanel.Width.Set(-10f, 0.5f);
 		uITextPanel.Height.Set(50f, 0f);
 		uITextPanel.VAlign = 1f;
@@ -643,7 +_,9 @@
 			UsedTags = (from x in _tagOptions
 						where x.IsSelected
 						select x.OptionValue).ToArray(),
-			PreviewImagePath = _previewImagePath
+			PreviewImagePath = _previewImagePath,
+			// Introduced by TML:
+			ChangeNotes = (this as WorkshopPublishInfoStateForMods).changeNotes,
 		};
 	}
 
@@ -653,6 +_,7 @@
 			new ExtensionFilter("Image files", "png", "jpg", "jpeg")
 		};
 
+		// TODO: Support default folder, set default folder to mod source folder for mods.
 		string text = FileBrowser.OpenFilePanel("Open icon", extensions);
 		if (text != null) {
 			_previewImagePath = text;
@@ -680,14 +_,15 @@
 		return path;
 	}
 
-	private void UpdateImagePreview()
+	protected virtual void UpdateImagePreview()
 	{
 		Texture2D texture2D = null;
 		string contents = PrettifyPath(_previewImagePath);
 		_previewImagePathPlate.SetContents(contents);
 		if (_previewImagePath != null) {
 			try {
+				//TML: Prefixed 'using' as a bugfix.
-				FileStream stream = File.OpenRead(_previewImagePath);
+				using FileStream stream = File.OpenRead(_previewImagePath);
 				texture2D = Texture2D.FromStream(Main.graphics.GraphicsDevice, stream);
 			}
 			catch (Exception exception) {
