--- src/Terraria/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/AWorkshopPublishInfoState.cs
@@ -38,11 +_,17 @@
 		private UICharacterNameButton _previewImagePathPlate;
 		private Texture2D _previewImageTransientTexture;
 		private UIImage _previewImageUIElement;
-		private string _previewImagePath;
+		protected string _previewImagePath;
 		private Asset<Texture2D> _defaultPreviewImageTexture;
 		private UIElement _steamDisclaimerButton;
 		private UIText _disclaimerText;
 		private UIGamepadHelper _helper;
+
+		// Added by TML.
+		public UIState PreviousUIState {
+			get => _previousUIState;
+			set => _previousUIState = value;
+		}
 
 		public AWorkshopPublishInfoState(UIState stateToGoBackTo, TPublishedObjectType dataObject) {
 			_previousUIState = stateToGoBackTo;
