--- src/Terraria/Terraria/GameContent/UI/States/UIAchievementsMenu.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIAchievementsMenu.cs
@@ -66,7 +_,7 @@
 			uITextPanel2.OnClick += GoBackClick;
 			uIElement.Append(uITextPanel2);
 			_backpanel = uITextPanel2;
-			List<Achievement> list = Main.Achievements.CreateAchievementsList();
+			List<IAchievement> list = Main.Achievements.CreateAchievementsList();
 			for (int i = 0; i < list.Count; i++) {
 				UIAchievementListItem item = new UIAchievementListItem(list[i], flag);
 				_achievementsList.Add(item);
@@ -139,7 +_,7 @@
 			SetupGamepadPoints(spriteBatch);
 		}
 
-		public void GotoAchievement(Achievement achievement) {
+		public void GotoAchievement(IAchievement achievement) {
 			_achievementsList.Goto(delegate (UIElement element) {
 				UIAchievementListItem uIAchievementListItem = element as UIAchievementListItem;
 				return uIAchievementListItem != null && uIAchievementListItem.GetAchievement() == achievement;
