--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIAchievementsMenu.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIAchievementsMenu.cs
@@ -1,13 +_,16 @@
+using System;
 using System.Collections.Generic;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
+using ReLogic.OS.Windows;
 using Terraria.Achievements;
 using Terraria.Audio;
 using Terraria.GameContent.UI.Elements;
 using Terraria.GameInput;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader.UI;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -20,9 +_,22 @@
 	private List<UIToggleImage> _categoryButtons = new List<UIToggleImage>();
 	private UIElement _backpanel;
 	private UIElement _outerContainer;
+	private UIPanel blockingPanel;
+
+	private UIInputTextField uISearchBar;
+	private UIPanel uISearchBackground;
+
+	private bool _filterModded = false;
+
+	Asset<Texture2D> blockingBackground;
+	Asset<Texture2D> blockingBorder;
+
+	private UIPanel achievementResetAreYouSure;
 
 	public void InitializePage()
 	{
+		blockingBackground = ModLoader.ModLoader.ManifestAssets.Request<Texture2D>($"Terraria.ModLoader.UI.InnerPanelNotRounded");
+		blockingBorder = ModLoader.ModLoader.ManifestAssets.Request<Texture2D>($"Terraria.ModLoader.UI.PanelBorderNotRounded");
 		RemoveAllChildren();
 		_categoryButtons.Clear();
 		_achievementElements.Clear();
@@ -50,22 +_,36 @@
 		_achievementsList.Top.Set(50f, 0f);
 		_achievementsList.ListPadding = 5f;
 		uIPanel.Append(_achievementsList);
+
 		UITextPanel<LocalizedText> uITextPanel = new UITextPanel<LocalizedText>(Language.GetText("UI.Achievements"), 1f, large: true);
 		uITextPanel.HAlign = 0.5f;
 		uITextPanel.Top.Set(-33f, 0f);
 		uITextPanel.SetPadding(13f);
 		uITextPanel.BackgroundColor = new Color(73, 94, 171);
 		uIElement.Append(uITextPanel);
+
 		UITextPanel<LocalizedText> uITextPanel2 = new UITextPanel<LocalizedText>(Language.GetText("UI.Back"), 0.7f, large: true);
-		uITextPanel2.Width.Set(-10f, 0.5f);
+		uITextPanel2.Width.Set(-10f, 0.8f);
 		uITextPanel2.Height.Set(50f, 0f);
 		uITextPanel2.VAlign = 1f;
-		uITextPanel2.HAlign = 0.5f;
+		uITextPanel2.HAlign = 0f;
 		uITextPanel2.Top.Set(-45f, 0f);
 		uITextPanel2.OnMouseOver += FadedMouseOver;
 		uITextPanel2.OnMouseOut += FadedMouseOut;
 		uITextPanel2.OnLeftClick += GoBackClick;
 		uIElement.Append(uITextPanel2);
+
+		UITextPanel<LocalizedText> resetAchievementsButton = new UITextPanel<LocalizedText>(Language.GetOrRegister("tModLoader.AchievementsReset", () => "Reset"), 0.7f, large: true);
+		resetAchievementsButton.Width.Set(-10f, 1 - uITextPanel2.Width.Percent);
+		resetAchievementsButton.Height.Set(50f, 0f);
+		resetAchievementsButton.VAlign = 1f;
+		resetAchievementsButton.HAlign = 1f;
+		resetAchievementsButton.Top.Set(-45f, 0f);
+		resetAchievementsButton.OnMouseOver += FadedMouseOver;
+		resetAchievementsButton.OnMouseOut += FadedMouseOut;
+		resetAchievementsButton.OnLeftClick += ResetAchievementsConfirm;
+		uIElement.Append(resetAchievementsButton);
+
 		_backpanel = uITextPanel2;
 		List<Achievement> list = Main.Achievements.CreateAchievementsList();
 		for (int i = 0; i < list.Count; i++) {
@@ -74,6 +_,17 @@
 			_achievementElements.Add(item);
 		}
 
+		uISearchBar = new UIInputTextField(Language.GetOrRegister("tModLoader.AchievementSearch", () => "Search..").Value);
+		uISearchBar.HAlign = 1;
+		uISearchBar.Height.Set(40, 0);
+		uISearchBar.Width.Set(0f, 0.8f);
+		uISearchBar.MaxWidth.Set(275f, 0f);
+		uISearchBar.MinWidth.Set(200f, 0f);
+		uISearchBar.Top.Set(15, 0);
+		uISearchBar.OnRightClick += (a, b) => uISearchBar.Text = "";
+		uISearchBar.OnTextChange += FilterSearch;
+		uIPanel.Append(uISearchBar);
+
 		UIScrollbar uIScrollbar = new UIScrollbar();
 		uIScrollbar.SetView(100f, 1000f);
 		uIScrollbar.Height.Set(-50f, 1f);
@@ -94,7 +_,6 @@
 			_categoryButtons.Add(uIToggleImage);
 			uIElement2.Append(uIToggleImage);
 		}
-
 		uIPanel.Append(uIElement2);
 	}
 
@@ -152,6 +_,102 @@
 		IngameFancyUI.Close();
 	}
 
+	private void ResetAchievements(UIMouseEvent evt, UIElement listeningElement)
+	{
+		CloseAchievementConfirm(evt, listeningElement);
+		Main.Achievements.ClearAll();
+		Main.menuMode = 0;
+		IngameFancyUI.Close();
+	}
+
+	private void ResetAchievementsConfirm(UIMouseEvent evt, UIElement listeningElement)
+	{
+		blockingPanel = new UIPanel(blockingBackground, blockingBorder);
+		blockingPanel.Width.Set(0, 1f);
+		blockingPanel.Height.Set(0, 1f);
+		blockingPanel.HAlign = 0.5f;
+		blockingPanel.VAlign = 0.5f;
+		blockingPanel.BackgroundColor = new Color(0, 0, 0, 125);
+		blockingPanel.BorderColor = blockingPanel.BackgroundColor;
+		blockingPanel.OnLeftClick += BlockInput;
+		blockingPanel.OnMouseOver += BlockInput;
+		Append(blockingPanel);
+
+		achievementResetAreYouSure = new UIPanel();
+		achievementResetAreYouSure.Width.Set(400f, 0f);
+		achievementResetAreYouSure.Height.Set(300, 0f);
+		achievementResetAreYouSure.VAlign = 0.5f;
+		achievementResetAreYouSure.HAlign = 0.5f;
+		Append(achievementResetAreYouSure);
+
+		UITextPanel<LocalizedText> areYouSureText = new UITextPanel<LocalizedText>(Language.GetOrRegister("tModLoader.AchievementResetConfirm", () => "Are you sure? Reg"), 0.6f, large: true);
+		areYouSureText.HAlign = 0.5f;
+		areYouSureText.SetPadding(13f);
+		areYouSureText.Top.Set(-33, 0f);
+
+		areYouSureText.BackgroundColor = new Color(73, 94, 171);
+		achievementResetAreYouSure.Append(areYouSureText);
+
+		string text = FontAssets.ItemStack.Value.CreateWrappedText(Language.GetOrRegister("tModLoader.AchievementResetConfirmTooltip", () => "Resetting your achievements means you'll lose them permanently, are you sure you want to continue?").Value, 310, Language.ActiveCulture.CultureInfo);
+		UITextPanel<string> areYouSureDesciption = new UITextPanel<string>(text, 1f, large: false);
+		areYouSureDesciption.HAlign = 0.5f;
+		areYouSureDesciption.Top.Set(20, 0f);
+		areYouSureDesciption.SetPadding(13f);
+		areYouSureDesciption.Width.Set(-10, 1);
+		areYouSureDesciption.Height.Set(-50, 0.9f);
+		achievementResetAreYouSure.Append(areYouSureDesciption);
+
+		// Confirm Button
+		UITextPanel<LocalizedText> yesButton = new UITextPanel<LocalizedText>(Language.GetOrRegister("tModLoader.AchievementsReset", () => "Reset"), 0.7f, large: true);
+		yesButton.Width.Set(0, 0.5f);
+		yesButton.Height.Set(40f, 0f);
+		yesButton.VAlign = 1;
+		yesButton.HAlign = 1;
+		yesButton.OnMouseOver += FadedMouseOver;
+		yesButton.OnMouseOut += FadedMouseOut;
+		yesButton.OnLeftClick += ResetAchievements;
+		achievementResetAreYouSure.Append(yesButton);
+
+		// Cancel Button
+		UITextPanel<LocalizedText> noButton = new UITextPanel<LocalizedText>(Language.GetText("UI.Cancel"), 0.7f, large: true);
+		noButton.Width.Set(-10, 0.5f);
+		noButton.Height.Set(40f, 0f);
+		noButton.VAlign = 1;
+		noButton.OnMouseOver += FadedMouseOver;
+		noButton.OnMouseOut += FadedMouseOut;
+		noButton.OnLeftClick += CloseAchievementConfirm;
+		achievementResetAreYouSure.Append(noButton);
+	}
+
+	private void CloseAchievementConfirm(UIMouseEvent evt, UIElement listeningElement)
+	{
+		RemoveChild(blockingPanel);
+		RemoveChild(achievementResetAreYouSure);
+		blockingPanel = null;
+		achievementResetAreYouSure = null;
+	}
+
+	private void BlockInput(UIMouseEvent evt, UIElement listeningElement)
+	{
+	}
+
+	private void FilterSearch(object sender, EventArgs e)
+	{
+		_achievementsList.Clear();
+		string searchText = uISearchBar.Text?.ToLowerInvariant() ?? string.Empty; // Get the search text, ensuring it's lowercase
+
+		foreach (UIAchievementListItem achievementElement in _achievementElements) {
+			string friendlyName = achievementElement.GetAchievement().FriendlyName.Value.ToLowerInvariant(); // Convert to lowercase for case-insensitive comparison
+			string modName = achievementElement.GetAchievement().Modded ? achievementElement.GetAchievement().ModAchievement.Mod.DisplayName.ToLowerInvariant()  : string.Empty; // Convert to lowercase for case-insensitive comparison
+
+			if (friendlyName.Contains(searchText) || modName.Contains(searchText)) {
+				_achievementsList.Add(achievementElement);
+			}
+		}
+
+		Recalculate();
+	}
+
 	private void FadedMouseOver(UIMouseEvent evt, UIElement listeningElement)
 	{
 		SoundEngine.PlaySound(12);
@@ -176,6 +_,13 @@
 
 		Recalculate();
 	}
+
+	private void ToggleFilterModded(UIMouseEvent evt, UIElement listeningElement)
+	{
+		_filterModded = !_filterModded;
+		FilterList(evt, listeningElement);
+	}
+
 
 	public override void OnActivate()
 	{
