--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UICharacterSelect.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UICharacterSelect.cs
@@ -1,6 +_,8 @@
 using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
+using System.Threading.Tasks;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.Audio;
@@ -9,14 +_,17 @@
 using Terraria.ID;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
+using Terraria.Utilities;
 
 namespace Terraria.GameContent.UI.States;
 
-public class UICharacterSelect : UIState
+public partial class UICharacterSelect : UIState
 {
-	private UIList _playerList;
+	internal UIList _playerList;
 	private UITextPanel<LocalizedText> _backPanel;
 	private UITextPanel<LocalizedText> _newPanel;
 	private UIPanel _containerPanel;
@@ -76,6 +_,7 @@
 		uITextPanel2.SetSnapPoint("New", 0);
 		_newPanel = uITextPanel3;
 		Append(uIElement);
+		InitializeMigrationPanel();
 	}
 
 	public override void Recalculate()
@@ -127,6 +_,7 @@
 	{
 		Main.LoadPlayers();
 		Main.ActivePlayerFileData = new PlayerFileData();
+		ActivateMigrationPanel();
 		UpdatePlayersList();
 		if (PlayerInput.UsingGamepadUI)
 			UILinkPointNavigator.ChangePoint(3000 + ((_playerList.Count == 0) ? 1 : 2));
@@ -149,6 +_,12 @@
 		int num = 0;
 		foreach (PlayerFileData item in list) {
 			_playerList.Add(new UICharacterListItem(item, num++));
+		}
+
+		_playerList.Add(_migrationPanel);
+
+		if (list.Count == 0) {
+			AddAutomaticPlayerMigrationButtons();
 		}
 	}
 
