--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UICharacterCreation.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UICharacterCreation.cs
@@ -13,6 +_,7 @@
 using Terraria.Initializers;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -1054,6 +_,15 @@
 				break;
 		}
 
+		if (player.hair >= HairID.Count) {
+		    player.Male = HairLoader.GetHair(player.hair).RandomizedCharacterCreationGender switch {
+				Gender.Male => true,
+				Gender.Female => false,
+				Gender.Unspecified => Main.rand.NextBool(),
+				_ => Main.rand.NextBool()
+			};
+		}
+
 		Click_CharClothStyle(null, null);
 		UpdateSelectedGender();
 		UpdateColorPickers();
@@ -1117,8 +_,14 @@
 		int num = 0;
 		byte difficulty = _player.difficulty;
 		if (difficulty == 3) {
+			PlayerLoader.ModifyMaxStats(_player);
+
+			/*
 			_player.statLife = (_player.statLifeMax = 100);
 			_player.statMana = (_player.statManaMax = 20);
+			*/
+			_player.statLife = _player.statLifeMax;
+			_player.statMana = _player.statManaMax;
 			_player.inventory[num].SetDefaults(6);
 			_player.inventory[num++].Prefix(-1);
 			_player.inventory[num].SetDefaults(1);
@@ -1156,6 +_,10 @@
 
 		_player.savedPerPlayerFieldsThatArentInThePlayerClass = new Player.SavedPlayerDataWithAnnoyingRules();
 		CreativePowerManager.Instance.ResetDataForNewPlayer(_player);
+
+		var vanillaItems = _player.inventory.Where(item => !item.IsAir).Select(x => x.Clone());
+		var startingItems = PlayerLoader.GetStartingItems(_player, vanillaItems);
+		PlayerLoader.SetStartInventory(_player, startingItems);
 	}
 
 	private bool GetHexColor(string hexString, out Vector3 hsl)
