--- src/Terraria/Terraria/GameContent/UI/States/UICharacterCreation.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UICharacterCreation.cs
@@ -13,6 +_,7 @@
 using Terraria.Initializers;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -1053,6 +_,10 @@
 
 			_player.savedPerPlayerFieldsThatArentInThePlayerClass = new Player.SavedPlayerDataWithAnnoyingRules();
 			CreativePowerManager.Instance.ResetDataForNewPlayer(_player);
+
+			var inventoryAdditions = PlayerHooks.AddStartingItems(_player);
+			PlayerHooks.ModifyStartingInventory(_player, inventoryAdditions);
+			PlayerHooks.SetStartInventory(_player, inventoryAdditions.Values.SelectMany(i => i).ToList());
 		}
 
 		private bool GetHexColor(string hexString, out Vector3 hsl) {
