--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIResourcePackSelectionMenu.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIResourcePackSelectionMenu.cs
@@ -33,6 +_,12 @@
 	private const string _snapPointName_openFolder = "OpenFolder";
 	private UIGamepadHelper _helper;
 
+	// Added by TML.
+	public UIState PreviousUIState {
+		get => _uiStateToGoBackTo;
+		set => _uiStateToGoBackTo = value;
+	}
+
 	public UIResourcePackSelectionMenu(UIState uiStateToGoBackTo, AssetSourceController sourceController, ResourcePackList currentResourcePackList)
 	{
 		_sourceController = sourceController;
