--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIVirtualKeyboard.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIVirtualKeyboard.cs
@@ -359,7 +_,7 @@
 		Append(uIText);
 		_label = uIText;
 		Append(uIElement);
-		int textMaxLength = (_edittingSign ? 1200 : 20);
+		int textMaxLength = (_edittingSign ? 1200 : _edittingChest ? Chest.MaxNameLength : 20);
 		_textBox.SetTextMaxLength(textMaxLength);
 		Text = startingText;
 		if (Text.Length == 0)
