--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIVirtualKeyboard.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIVirtualKeyboard.cs
@@ -358,7 +_,7 @@
 		Append(uIText);
 		_label = uIText;
 		Append(uIElement);
-		_textBox.SetTextMaxLength(maxLength);
+		_textBox.SetTextMaxLength(_edittingChest ? Chest.MaxNameLength : maxLength);
 		Text = startingText;
 		if (Text.Length == 0)
 			SetKeyState(KeyState.Shift);
