--- src/Terraria/Terraria/GameContent/UI/States/UIWorldCreation.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIWorldCreation.cs
@@ -632,8 +_,10 @@
 					Main.maxTilesY = 1800;
 					break;
 				case WorldSizeId.Large:
+					//Main.maxTilesX = 8400;
+					//Main.maxTilesY = 2400;
-					Main.maxTilesX = 8400;
+					Main.maxTilesX = 1750;
-					Main.maxTilesY = 2400;
+					Main.maxTilesY = 900;
 					break;
 			}
 
