--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIWorldCreationAdvanced.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIWorldCreationAdvanced.cs
@@ -7,6 +_,7 @@
 using Terraria.GameContent.UI.Elements;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 using Terraria.WorldBuilding;
@@ -36,6 +_,12 @@
 	private bool _allowScrolling;
 	private UIGamepadHelper _helper;
 
+	// Added by TML.
+	public UIState PreviousUIState {
+		get => throw new NotSupportedException();
+		set => throw new NotSupportedException();
+	}
+
 	public UIWorldCreationAdvanced(UIWorldCreation state, bool allowScrolling = false)
 	{
 		_creationState = state;
@@ -70,7 +_,9 @@
 			VAlign = 0f
 		};
 
+		/*
 		if (!_allowScrolling)
+		*/
 			uIElement.MaxHeight = StyleDimension.FromPixels(400f);
 
 		uIElement.SetPadding(0f);
@@ -132,8 +_,11 @@
 		UIScrollbar uIScrollbar = new UIScrollbar {
 			Height = StyleDimension.FromPixelsAndPercent(-138 - num * 2, 1f),
 			Top = StyleDimension.FromPixels(44 + num),
-			HAlign = 1f
+			HAlign = 1f,
+			AutoHide = true
 		};
+		uIScrollbar.OnScroll += () => _creationState.RecalculateSpecialSeedRing = true;
+
 
 		uIList.SetScrollbar(uIScrollbar);
 		infoContainer.Append(uIList);
@@ -196,7 +_,7 @@
 		GroupOptionButton<AWorldGenerationOption>[] array = (_seedButtons = PrepareSeedButtons());
 		_seedButtonRegion = new UIElement {
 			Width = StyleDimension.FromPercent(1f),
-			Height = StyleDimension.FromPixels((float)Math.Ceiling((double)array.Length / (double)num) * 70f - 10f)
+			Height = StyleDimension.FromPixels((float)Math.Ceiling((double)(array.Length+1) / (double)num) * 70f)
 		};
 
 		listArea.Add(_seedButtonRegion);
@@ -218,7 +_,7 @@
 	private GroupOptionButton<AWorldGenerationOption>[] PrepareSeedButtons()
 	{
 		List<GroupOptionButton<AWorldGenerationOption>> list = new List<GroupOptionButton<AWorldGenerationOption>>();
-		foreach (AWorldGenerationOption option in WorldGenerationOptions.Options) {
+		foreach (AWorldGenerationOption option in SeedLoader.AddModSeedOptions(WorldGenerationOptions.Options)) {
 			option.Load();
 			list.Add(CreateButton(new WorldSpecialSeedOption {
 				Seed = option,
@@ -235,6 +_,9 @@
 	{
 		AWorldGenerationOption optionValue = ((GroupOptionButton<AWorldGenerationOption>)listeningElement).OptionValue;
 		_creationState.ToggleSeedOption(optionValue);
+		if (optionValue is ModSpecialSeedUIOption modOption) {
+			modOption.OnClick();
+		}
 		UpdateContents();
 	}
 
