--- src/TerrariaNetCore/Terraria/GameContent/UI/States/UIWorldCreation.cs
+++ src/tModLoader/Terraria/GameContent/UI/States/UIWorldCreation.cs
@@ -14,6 +_,7 @@
 using Terraria.ID;
 using Terraria.IO;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.Testing;
 using Terraria.UI;
@@ -118,6 +_,12 @@
 		ResetSpecialSeedRing();
 	}
 
+	// Added by TML.
+	public UIState PreviousUIState {
+		get => _goBackTarget;
+		set => _goBackTarget = value;
+	}
+
 	public void SetGoBackTarget(UIState state)
 	{
 		_goBackTarget = state;
@@ -484,7 +_,7 @@
 	private void ClickAdvancedSeedMenu(UIMouseEvent evt, UIElement listeningElement)
 	{
 		ResetSpecialSeedRing();
-		UIWorldCreationAdvanced uIWorldCreationAdvanced = new UIWorldCreationAdvanced(this);
+		UIWorldCreationAdvanced uIWorldCreationAdvanced = new UIWorldCreationAdvanced(this, true);
 		SetGoBackTarget(uIWorldCreationAdvanced);
 		Main.MenuUI.SetState(uIWorldCreationAdvanced);
 	}
@@ -793,6 +_,9 @@
 				WorldGenerationOptions.SelectOption(optionFromSeedText);
 				SoundEngine.PlaySound(24);
 			}
+			else {
+				_isSpecialSeedText = SpecialSeedLoader.CanEnableSeedsFromText(_optionSeed);
+			}
 
 			if (WorldGen.SecretSeed.CheckInputForSecretSeed(_optionSeed, out var secretSeed)) {
 				if (!secretSeed.Enabled) {
@@ -941,6 +_,8 @@
 			string seed = string.Join("|", _secretSeedTextsEntered) + "|" + Main.ActiveWorldFileData.SeedText;
 			Main.ActiveWorldFileData.SetSeed(seed);
 		}
+
+		SpecialSeedLoader.SetWorldGenMenu();
 
 		WorldGenerator.Controller controller = new WorldGenerator.Controller {
 			Paused = (DebugOptions.enableDebugCommands && Main.keyState.PressingControl())
