--- src/TerrariaNetCore/Terraria/Graphics/Renderers/FakeFishParticle.cs
+++ src/tModLoader/Terraria/Graphics/Renderers/FakeFishParticle.cs
@@ -176,7 +_,10 @@
 						break;
 				}
 
+				/*
 				SoundEngine.PlaySound(SoundID.FishSplash, (int)vector.X, (int)vector.Y, 5f);
+				*/
+				SoundEngine.PlaySound(SoundID.FishSplash.WithPitchOffset(5), (int)vector.X, (int)vector.Y);
 			}
 		}
 		else if (honeyWet) {
@@ -189,7 +_,7 @@
 				Main.dust[num4].noGravity = true;
 			}
 
-			SoundEngine.PlaySound(SoundID.FishSplash, (int)vector.X, (int)vector.Y, 1f);
+			SoundEngine.PlaySound(SoundID.FishSplash.WithPitchOffset(1f), (int)vector.X, (int)vector.Y);
 		}
 		else if (lavaWet) {
 			for (int k = 0; k < 10; k++) {
@@ -201,7 +_,10 @@
 				Main.dust[num5].noGravity = true;
 			}
 
+			/*
 			SoundEngine.PlaySound(SoundID.FishSplash, (int)vector.X, (int)vector.Y, 1f);
+			*/
+			SoundEngine.PlaySound(SoundID.FishSplash.WithPitchOffset(1f), (int)vector.X, (int)vector.Y);
 		}
 		else {
 			for (int l = 0; l < 10; l++) {
@@ -213,7 +_,7 @@
 				Main.dust[num6].noGravity = true;
 			}
 
-			SoundEngine.PlaySound(SoundID.FishSplash, (int)vector.X, (int)vector.Y, 1f);
+			SoundEngine.PlaySound(SoundID.FishSplash.WithPitchOffset(1f), (int)vector.X, (int)vector.Y);
 		}
 	}
 
