--- src/Terraria/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
+++ src/tModLoader/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
@@ -2,9 +_,12 @@
 using Microsoft.Xna.Framework.Graphics;
 using System;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
 using Terraria.DataStructures;
 using Terraria.GameContent;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Renderers
 {
@@ -30,6 +_,7 @@
 		}
 
 		public void DrawPlayerHead(Camera camera, Player drawPlayer, Vector2 position, float alpha = 1f, float scale = 1f, Color borderColor = default(Color)) {
+			/*
 			if (!drawPlayer.ShouldNotDraw) {
 				_drawData.Clear();
 				_dust.Clear();
@@ -48,6 +_,9 @@
 				CreateOutlines(alpha, scale, borderColor);
 				PlayerDrawHeadLayers.DrawPlayer_RenderAllLayers(ref drawinfo);
 			}
+			*/
+
+			DrawPlayerInternal(camera, drawPlayer, position + Main.screenPosition, 0f, Vector2.Zero, 0f, scale, false, true);
 		}
 
 		private void CreateOutlines(float alpha, float scale, Color borderColor) {
@@ -110,87 +_,77 @@
 		}
 
 		public void DrawPlayer(Camera camera, Player drawPlayer, Vector2 position, float rotation, Vector2 rotationOrigin, float shadow = 0f, float scale = 1f) {
+			DrawPlayerInternal(camera, drawPlayer, position, rotation, rotationOrigin, shadow, scale);
+		}
+
+		// Compare this with the vanilla layer drawing call order to make sure it's accurate when updating
+		internal static IList<PlayerDrawLayer> GetVanillaLayers() => new List<PlayerDrawLayer> {
+			PlayerDrawLayer.JimsCloak,
+			PlayerDrawLayer.MountBack,
+			PlayerDrawLayer.Carpet,
+			PlayerDrawLayer.PortableStool,
+			PlayerDrawLayer.ElectrifiedDebuffBack,
+			PlayerDrawLayer.ForbiddenSetRing,
+			PlayerDrawLayer.SafemanSun,
+			PlayerDrawLayer.WebbedDebuffBack,
+			PlayerDrawLayer.LeinforsHairShampoo,
+			PlayerDrawLayer.Backpacks,
+			PlayerDrawLayer.Wings,
+			PlayerDrawLayer.BackAcc,
+			PlayerDrawLayer.HairBack,
+			PlayerDrawLayer.HeadBack,
+			PlayerDrawLayer.BalloonAcc,
+			PlayerDrawLayer.HeldItemBehindBackArm, // drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindBackArm
+			PlayerDrawLayer.Skin,
+			PlayerDrawLayer.Leggings, // !(drawinfo.drawPlayer.wearsRobe && drawinfo.drawPlayer.body != 166)
+			PlayerDrawLayer.Shoes,
+			PlayerDrawLayer.Robe, // drawinfo.drawPlayer.wearsRobe && drawinfo.drawPlayer.body != 166
+			PlayerDrawLayer.SkinLongCoat,
+			PlayerDrawLayer.ArmorLongCoat,
+			PlayerDrawLayer.Torso,
+			PlayerDrawLayer.OffhandAcc,
+			PlayerDrawLayer.WaistAcc,
+			PlayerDrawLayer.NeckAcc,
+			PlayerDrawLayer.Head,
+			PlayerDrawLayer.FaceAcc,
+			PlayerDrawLayer.FrontAccFrontNeck, // drawinfo.drawFrontAccInNeckAccLayer
+			PlayerDrawLayer.MountFront,
+			PlayerDrawLayer.Pulley,
+			PlayerDrawLayer.FrontAccBack,
+			PlayerDrawLayer.Shield,
+			PlayerDrawLayer.SolarShield,
+			PlayerDrawLayer.HeldItemBehindFrontArm, // drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindFrontArm
+			PlayerDrawLayer.ArmOverItem,
+			PlayerDrawLayer.HandOnAcc,
+			PlayerDrawLayer.BladedGlove,
+			PlayerDrawLayer.FrontAccFront, // !drawinfo.drawFrontAccInNeckAccLayer
+			PlayerDrawLayer.HeldItemOverFrontArm, // drawinfo.weaponDrawOrder == WeaponDrawOrder.OverFrontArm
+			PlayerDrawLayer.ProjectileOverArm,
+			PlayerDrawLayer.FrozenOrWebbedDebuff,
+			PlayerDrawLayer.ElectrifiedDebuffFront,
+			PlayerDrawLayer.IceBarrier,
+			PlayerDrawLayer.CaptureTheGem,
+			PlayerDrawLayer.BeetleBuff
+		};
+
+		private void DrawPlayerInternal(Camera camera, Player drawPlayer, Vector2 position, float rotation, Vector2 rotationOrigin, float shadow = 0f, float scale = 1f, bool useLighting = true, bool headOnly = false) {
 			if (drawPlayer.ShouldNotDraw)
 				return;
 
-			PlayerDrawSet drawinfo = default(PlayerDrawSet);
+			PlayerDrawSet drawinfo = default;
 			_drawData.Clear();
 			_dust.Clear();
 			_gore.Clear();
-			drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position, shadow, rotation, rotationOrigin);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_2_JimsCloak(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_02_MountBehindPlayer(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_03_Carpet(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_03_PortableStool(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_04_ElectrifiedDebuffBack(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_05_ForbiddenSetRing(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_05_2_SafemanSun(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_06_WebbedDebuffBack(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_07_LeinforsHairShampoo(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_08_Backpacks(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_09_Wings(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_10_BackAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_BackHair(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_3_BackHead(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_11_Balloons(ref drawinfo);
-			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindBackArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
-
-			PlayerDrawLayers.DrawPlayer_12_Skin(ref drawinfo);
-			if (drawinfo.drawPlayer.wearsRobe && drawinfo.drawPlayer.body != 166) {
-				PlayerDrawLayers.DrawPlayer_14_Shoes(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_13_Leggings(ref drawinfo);
-			}
-			else {
-				PlayerDrawLayers.DrawPlayer_13_Leggings(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_14_Shoes(ref drawinfo);
-			}
-
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_15_SkinLongCoat(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_16_ArmorLongCoat(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_17_Torso(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_18_OffhandAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_19_WaistAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_20_NeckAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_21_Head(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_22_FaceAcc(ref drawinfo);
-			if (drawinfo.drawFrontAccInNeckAccLayer) {
-				PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_32_FrontAcc_FrontPart(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			}
-
-			PlayerDrawLayers.DrawPlayer_23_MountFront(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_24_Pulley(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_32_FrontAcc_BackPart(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_25_Shield(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_MountPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_26_SolarShield(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_MountMinus(ref drawinfo);
-			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindFrontArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
-
-			PlayerDrawLayers.DrawPlayer_28_ArmOverItem(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_29_OnhandAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_30_BladedGlove(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-			if (!drawinfo.drawFrontAccInNeckAccLayer)
-				PlayerDrawLayers.DrawPlayer_32_FrontAcc_FrontPart(ref drawinfo);
-
-			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.OverFrontArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
-
-			PlayerDrawLayers.DrawPlayer_31_ProjectileOverArm(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_33_FrozenOrWebbedDebuff(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_34_ElectrifiedDebuffFront(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_35_IceBarrier(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_36_CTG(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_37_BeetleBuff(ref drawinfo);
+			drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position, shadow, rotation, rotationOrigin, useLighting);
+
+			PlayerHooks.ModifyDrawInfo(ref drawinfo);
+
+			foreach (var layer in PlayerDrawLayerHooks.GetDrawLayers(drawinfo)) {
+				if (!headOnly || layer.IsHeadLayer) {
+					layer.PositionAndDraw(ref drawinfo);
+				}
+			}
+
 			PlayerDrawLayers.DrawPlayer_MakeIntoFirstFractalAfterImage(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_TransformDrawData(ref drawinfo);
 			if (scale != 1f)
