--- src/Terraria/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
+++ src/tModLoader/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
@@ -5,6 +_,7 @@
 using Terraria.DataStructures;
 using Terraria.GameContent;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Renderers
 {
@@ -30,6 +_,7 @@
 		}
 
 		public void DrawPlayerHead(Camera camera, Player drawPlayer, Vector2 position, float alpha = 1f, float scale = 1f, Color borderColor = default(Color)) {
+			/*
 			if (!drawPlayer.ShouldNotDraw) {
 				_drawData.Clear();
 				_dust.Clear();
@@ -48,6 +_,9 @@
 				CreateOutlines(alpha, scale, borderColor);
 				PlayerDrawHeadLayers.DrawPlayer_RenderAllLayers(ref drawinfo);
 			}
+			*/
+
+			DrawPlayerInternal(camera, drawPlayer, position + Main.screenPosition, 0f, Vector2.Zero, alpha: alpha, scale: scale, headOnly: true);
 		}
 
 		private void CreateOutlines(float alpha, float scale, Color borderColor) {
@@ -110,6 +_,10 @@
 		}
 
 		public void DrawPlayer(Camera camera, Player drawPlayer, Vector2 position, float rotation, Vector2 rotationOrigin, float shadow = 0f, float scale = 1f) {
+			DrawPlayerInternal(camera, drawPlayer, position, rotation, rotationOrigin, shadow, scale);
+		}
+
+		private void DrawPlayerInternal(Camera camera, Player drawPlayer, Vector2 position, float rotation, Vector2 rotationOrigin, float shadow = 0f, float alpha = 1f, float scale = 1f, bool headOnly = false) {
 			if (drawPlayer.ShouldNotDraw)
 				return;
 
@@ -117,7 +_,15 @@
 			_drawData.Clear();
 			_dust.Clear();
 			_gore.Clear();
+
+			if (headOnly) {
+				drawinfo.HeadOnlySetup(drawPlayer, _drawData, _dust, _gore, position.X, position.Y, alpha, scale);
+			}
+			else {
-			drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position, shadow, rotation, rotationOrigin);
+				drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position, shadow, rotation, rotationOrigin);
+			}
+
+			/*
 			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_01_2_JimsCloak(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
@@ -198,6 +_,15 @@
 			PlayerDrawLayers.DrawPlayer_36_CTG(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_37_BeetleBuff(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_38_EyebrellaCloud(ref drawinfo);
+			*/
+			PlayerLoader.ModifyDrawInfo(ref drawinfo);
+
+			foreach (var layer in PlayerDrawLayerLoader.GetDrawLayers(drawinfo)) {
+				if (!headOnly || layer.IsHeadLayer) {
+					layer.DrawWithTransformationAndChildren(ref drawinfo);
+				}
+			}
+
 			PlayerDrawLayers.DrawPlayer_MakeIntoFirstFractalAfterImage(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_TransformDrawData(ref drawinfo);
 			if (scale != 1f)
