--- src/Terraria/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
+++ src/tModLoader/Terraria/Graphics/Renderers/LegacyPlayerRenderer.cs
@@ -2,9 +_,12 @@
 using Microsoft.Xna.Framework.Graphics;
 using System;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
 using Terraria.DataStructures;
 using Terraria.GameContent;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.Graphics.Renderers
 {
@@ -35,15 +_,25 @@
 			_gore.Clear();
 			PlayerDrawHeadSet drawinfo = default(PlayerDrawHeadSet);
 			drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position.X, position.Y, alpha, scale);
-			PlayerDrawHeadLayers.DrawPlayer_00_BackHelmet(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_01_FaceSkin(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_02_DrawArmorWithFullHair(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_03_HelmetHair(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_04_JungleRose(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_05_TallHats(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_06_NormalHats(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_07_JustHair(ref drawinfo);
-			PlayerDrawHeadLayers.DrawPlayer_08_FaceAcc(ref drawinfo);
+
+			var layers = new List<PlayerHeadLayer> {
+				PlayerHeadLayer.BackHelmet,
+				PlayerHeadLayer.FaceSkin,
+				PlayerHeadLayer.DrawArmorWithFullHair,
+				PlayerHeadLayer.HelmetHair,
+				PlayerHeadLayer.JungleRose,
+				PlayerHeadLayer.TallHats,
+				PlayerHeadLayer.NormalHats,
+				PlayerHeadLayer.JustHair,
+				PlayerHeadLayer.FaceAcc
+			};
+
+			PlayerHooks.ModifyDrawHeadLayers(drawPlayer, layers);
+
+			foreach (var layer in layers) {
+				layer.Draw(ref drawinfo);
+			}
+
 			CreateOutlines(alpha, scale, borderColor);
 			PlayerDrawHeadLayers.DrawPlayer_RenderAllLayers(ref drawinfo);
 		}
@@ -116,78 +_,100 @@
 			_dust.Clear();
 			_gore.Clear();
 			drawinfo.BoringSetup(drawPlayer, _drawData, _dust, _gore, position, shadow, rotation, rotationOrigin);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_BackHair(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_2_JimsCloak(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_01_3_BackHead(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_02_MountBehindPlayer(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_03_Carpet(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_03_PortableStool(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_04_ElectrifiedDebuffBack(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_05_ForbiddenSetRing(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_05_2_SafemanSun(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_06_WebbedDebuffBack(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_07_LeinforsHairShampoo(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_08_Backpacks(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_09_BackAc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_10_Wings(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_11_Balloons(ref drawinfo);
+
+			var vanillaLayers = new List<PlayerLayer> {
+				PlayerLayer.TorsoIndent, //{
+				PlayerLayer.HairBack,
+				PlayerLayer.JimsCloak,
+				PlayerLayer.HeadBack,
+				PlayerLayer.TorsoUnindent, //}
+				PlayerLayer.MountBack,
+				PlayerLayer.Carpet,
+				PlayerLayer.PortableStool,
+				PlayerLayer.TorsoIndent, //{
+				PlayerLayer.ElectrifiedDebuffBack,
+				PlayerLayer.ForbiddenSetRing,
+				PlayerLayer.SafemanSun,
+				PlayerLayer.WebbedDebuffBack,
+				PlayerLayer.LeinforsHairShampoo,
+				PlayerLayer.TorsoUnindent, //}
+				PlayerLayer.Backpacks,
+				PlayerLayer.BackAcc,
+				PlayerLayer.Wings,
+				PlayerLayer.BalloonAcc,
+			};
+
 			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindBackArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.HeldItem);
 
-			PlayerDrawLayers.DrawPlayer_12_Skin(ref drawinfo);
+			vanillaLayers.Add(PlayerLayer.Skin);
 			if (drawinfo.drawPlayer.wearsRobe && drawinfo.drawPlayer.body != 166) {
-				PlayerDrawLayers.DrawPlayer_14_Shoes(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_13_Leggings(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.Shoes);
+				vanillaLayers.Add(PlayerLayer.Leggings);
 			}
 			else {
-				PlayerDrawLayers.DrawPlayer_13_Leggings(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_14_Shoes(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.Leggings);
+				vanillaLayers.Add(PlayerLayer.Shoes);
 			}
 
-			PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_15_SkinLongCoat(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_16_ArmorLongCoat(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_17_Torso(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_18_OffhandAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_19_WaistAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_20_NeckAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_21_Head(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_22_FaceAcc(ref drawinfo);
+			vanillaLayers.Add(PlayerLayer.TorsoIndent); //{
+			vanillaLayers.Add(PlayerLayer.SkinLongCoat);
+			vanillaLayers.Add(PlayerLayer.ArmorLongCoat);
+			vanillaLayers.Add(PlayerLayer.Torso);
+			vanillaLayers.Add(PlayerLayer.OffhandAcc);
+			vanillaLayers.Add(PlayerLayer.WaistAcc);
+			vanillaLayers.Add(PlayerLayer.NeckAcc);
+			vanillaLayers.Add(PlayerLayer.Head);
+			vanillaLayers.Add(PlayerLayer.FaceAcc);
 			if (drawinfo.drawFrontAccInNeckAccLayer) {
-				PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_32_FrontAcc(ref drawinfo);
-				PlayerDrawLayers.DrawPlayer_extra_TorsoPlus(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.TorsoUnindent); //}
+				vanillaLayers.Add(PlayerLayer.FrontAcc);
+				vanillaLayers.Add(PlayerLayer.TorsoIndent); //{
 			}
 
-			PlayerDrawLayers.DrawPlayer_23_MountFront(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_24_Pulley(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_25_Shield(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_MountPlus(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_26_SolarShield(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_MountMinus(ref drawinfo);
+			vanillaLayers.Add(PlayerLayer.MountFront);
+			vanillaLayers.Add(PlayerLayer.Pulley);
+			vanillaLayers.Add(PlayerLayer.Shield);
+			vanillaLayers.Add(PlayerLayer.MountIndent); //{
+			vanillaLayers.Add(PlayerLayer.SolarShield);
+			vanillaLayers.Add(PlayerLayer.MountUnindent); //}
 			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.BehindFrontArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.HeldItem);
 
-			PlayerDrawLayers.DrawPlayer_28_ArmOverItem(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_29_OnhandAcc(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_30_BladedGlove(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_extra_TorsoMinus(ref drawinfo);
+			vanillaLayers.Add(PlayerLayer.ArmOverItem);
+			vanillaLayers.Add(PlayerLayer.HandOnAcc);
+			vanillaLayers.Add(PlayerLayer.BladedGlove);
+			vanillaLayers.Add(PlayerLayer.TorsoUnindent); //}
 			if (!drawinfo.drawFrontAccInNeckAccLayer)
-				PlayerDrawLayers.DrawPlayer_32_FrontAcc(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.FrontAcc);
 
 			if (drawinfo.weaponDrawOrder == WeaponDrawOrder.OverFrontArm)
-				PlayerDrawLayers.DrawPlayer_27_HeldItem(ref drawinfo);
-
-			PlayerDrawLayers.DrawPlayer_31_ProjectileOverArm(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_33_FrozenOrWebbedDebuff(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_34_ElectrifiedDebuffFront(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_35_IceBarrier(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_36_CTG(ref drawinfo);
-			PlayerDrawLayers.DrawPlayer_37_BeetleBuff(ref drawinfo);
+				vanillaLayers.Add(PlayerLayer.HeldItem);
+
+			vanillaLayers.Add(PlayerLayer.ProjectileOverArm);
+			vanillaLayers.Add(PlayerLayer.FrozenOrWebbedDebuff);
+			vanillaLayers.Add(PlayerLayer.ElectrifiedDebuffFront);
+			vanillaLayers.Add(PlayerLayer.IceBarrier);
+			vanillaLayers.Add(PlayerLayer.CaptureTheGem);
+			vanillaLayers.Add(PlayerLayer.BeetleBuff);
+
+			for (int i = 0; i < vanillaLayers.Count; i++) {
+				vanillaLayers[i].depth = i;
+			}
+
+			var readOnlyVanillaLayers = vanillaLayers.ToList().AsReadOnly();
+			var layers = new Dictionary<string, List<PlayerLayer>> {
+				{ "Terraria", vanillaLayers }
+			};
+
+			PlayerLayerHooks.ModifyDrawLayers(drawPlayer, layers, readOnlyVanillaLayers);
+			PlayerHooks.AddDrawLayers(drawPlayer, layers);
+			PlayerHooks.ModifyDrawLayers(drawPlayer, new ReadOnlyDictionary<string, List<PlayerLayer>>(layers));
+
+			foreach (var layer in layers.SelectMany(pair => pair.Value).OrderBy(l => l.depth)) {
+				layer.Draw(ref drawinfo);
+			}
+
 			PlayerDrawLayers.DrawPlayer_MakeIntoFirstFractalAfterImage(ref drawinfo);
 			PlayerDrawLayers.DrawPlayer_TransformDrawData(ref drawinfo);
 			if (scale != 1f)
