--- src/TerrariaNetCore/Terraria/Audio/ActiveSound.cs
+++ src/tModLoader/Terraria/Audio/ActiveSound.cs
@@ -1,48 +_,76 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
+using System;
+
+#nullable enable
 
 namespace Terraria.Audio;
 
+//TML: Expanded, moved and renamed version of vanilla's LoopedPlayCondition.
+/// <summary>
+/// <see cref="ActiveSound"/>'s update callback.
+/// <br/> Returning false here will force the sound to end abruptly.
+/// <br/> Tip: Use <see cref="ProjectileAudioTracker.IsActiveAndInGame"/> to tie sounds to projectiles.
+/// </summary>
+/// <param name="soundInstance"> The sound object instance. </param>
+/// <returns> Whether the sound effect should continue to play. </returns>
+public delegate bool SoundUpdateCallback(ActiveSound soundInstance);
+
 public class ActiveSound
 {
+	/*
 	public delegate bool LoopedPlayCondition();
 
 	public readonly bool IsGlobal;
-	public Vector2 Position;
+	*/
+
+	public Vector2? Position; //TML: Made nullable
 	public float Volume;
 	public float Pitch;
+	public SoundUpdateCallback? Callback;
+	/*
 	public LoopedPlayCondition Condition;
+	*/
 
-	public SoundEffectInstance Sound { get; private set; }
+	public SoundEffectInstance? Sound { get; private set; }
 
 	public SoundStyle Style { get; private set; }
 
 	public bool IsPlaying {
 		get {
 			if (Sound != null)
-				return Sound.State == SoundState.Playing;
+				return Sound?.IsDisposed == false && Sound.State == SoundState.Playing; //TML: Made error-safe
 
 			return false;
 		}
 	}
 
+	public bool IsPlayingOrPaused => Sound?.IsDisposed == false && Sound.State != SoundState.Stopped;
+
+	/*
 	private void UseOverrides(SoundPlayOverrides overrides)
 	{
 		if (overrides.Volume.HasValue)
 			Volume = overrides.Volume.Value;
 	}
+	*/
 
-	public ActiveSound(SoundStyle style, Vector2 position, SoundPlayOverrides overrides)
+	//TML: Position parameter made nullable, SoundUpdateCallback added with changed type. SoundPlayOverrides removed in favor of SoundStyle.WithVolume
+	public ActiveSound(SoundStyle style, Vector2? position = null, SoundUpdateCallback? updateCallback = null)
 	{
 		Position = position;
 		Volume = 1f;
 		Pitch = style.PitchVariance;
-		IsGlobal = false;
-		Style = style;
-		UseOverrides(overrides);
+		//UseOverrides(overrides);
+		//IsGlobal = false;
+		Style = style.WithSelectedVariant(style.SelectedVariant); // Select a variant if for some reason not set at this point.
+		Callback = updateCallback;
+
 		Play();
 	}
 
+	//TML: Merged with the above constructor.
+	/*
 	public ActiveSound(SoundStyle style)
 	{
 		Position = Vector2.Zero;
@@ -63,18 +_,26 @@
 		UseOverrides(overrides);
 		PlayLooped(condition);
 	}
+	*/
 
 	private void Play()
 	{
+		if (!Program.IsMainThread) {
+			RunOnMainThreadAndWait(Play);
+			return;
+		}
+
-		SoundEffectInstance soundEffectInstance2 = (Sound = Style.GetRandomSound().CreateInstance());
+		SoundEffectInstance soundEffectInstance2 = (Sound = Style.GetSoundEffect().CreateInstance());
 		soundEffectInstance2.Pitch += Style.GetRandomPitch();
 		Pitch = soundEffectInstance2.Pitch;
 		soundEffectInstance2.Volume = DetermineIntendedVolume();
+		soundEffectInstance2.IsLooped = Style.IsLooped;
 		soundEffectInstance2.Play();
 		SoundInstanceGarbageCollector.Track(soundEffectInstance2);
 		Update();
 	}
 
+	/*
 	private void PlayLooped(LoopedPlayCondition condition)
 	{
 		SoundEffectInstance soundEffectInstance2 = (Sound = Style.GetRandomSound().CreateInstance());
@@ -86,21 +_,37 @@
 		SoundInstanceGarbageCollector.Track(soundEffectInstance2);
 		Update();
 	}
+	*/
 
 	public void Stop()
 	{
+		if (!Program.IsMainThread) {
+			RunOnMainThreadAndWait(Stop);
+			return;
+		}
+
 		if (Sound != null)
 			Sound.Stop();
 	}
 
 	public void Pause()
 	{
+		if (!Program.IsMainThread) {
+			RunOnMainThreadAndWait(Pause);
+			return;
+		}
+
 		if (Sound != null && Sound.State == SoundState.Playing)
 			Sound.Pause();
 	}
 
 	public void Resume()
 	{
+		if (!Program.IsMainThread) {
+			RunOnMainThreadAndWait(Resume);
+			return;
+		}
+
 		if (Sound != null && Sound.State == SoundState.Paused)
 			Sound.Resume();
 	}
@@ -108,7 +_,15 @@
 	public void Update()
 	{
 		if (Sound != null) {
-			if (Condition != null && !Condition()) {
+			if (!Program.IsMainThread) {
+				RunOnMainThreadAndWait(Update);
+				return;
+			}
+
+			if (Sound.IsDisposed)
+				return;
+
+			if (Callback?.Invoke(this) == false) {
 				Sound.Stop(immediate: true);
 				return;
 			}
@@ -122,8 +_,11 @@
 	private float DetermineIntendedVolume()
 	{
 		float num = 1f;
+		/*
 		if (!IsGlobal) {
+		*/
+		if (Position.HasValue) {
-			Vector2 vector = Position - Main.Camera.Center;
+			Vector2 vector = Position.Value - Main.Camera.Center;
 			Sound.Pan = MathHelper.Clamp(vector.X / ((float)Main.MaxWorldViewSize.X * 0.5f), -1f, 1f);
 			num = MathHelper.Clamp(1f - vector.Length() / LegacySoundPlayer.SoundAttenuationDistance, 0f, 1f);
 		}
@@ -135,6 +_,12 @@
 				break;
 			case SoundType.Ambient:
 				num *= Main.ambientVolume;
+
+				// Added by TML to mimic the behavior of the LegacySoundPlayer code.
+				if (FocusHelper.QuietAmbientSounds) {
+					num = 0f;
+				}
+
 				break;
 			case SoundType.Music:
 				num *= Main.musicVolume;
@@ -143,4 +_,9 @@
 
 		return MathHelper.Clamp(num, 0f, 1f);
 	}
+
+	// Added by TML.
+	// As the name states, runs code on the main thread and waits for it to finish.
+	private static void RunOnMainThreadAndWait(Action action)
+		=> Main.RunOnMainThread(action).GetAwaiter().GetResult();
 }
