--- src/Terraria/Terraria/Audio/LegacySoundPlayer.cs
+++ src/tModLoader/Terraria/Audio/LegacySoundPlayer.cs
@@ -5,6 +_,7 @@
 using System;
 using System.IO;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.Audio
 {
@@ -38,12 +_,12 @@
 		private SoundEffectInstance _soundInstanceGrab;
 		private Asset<SoundEffect> _soundPixie;
 		private SoundEffectInstance _soundInstancePixie;
-		private Asset<SoundEffect>[] _soundItem = new Asset<SoundEffect>[SoundID.ItemSoundCount];
+		internal Asset<SoundEffect>[] _soundItem = new Asset<SoundEffect>[SoundID.ItemSoundCount];
-		private SoundEffectInstance[] _soundInstanceItem = new SoundEffectInstance[SoundID.ItemSoundCount];
+		internal SoundEffectInstance[] _soundInstanceItem = new SoundEffectInstance[SoundID.ItemSoundCount];
-		private Asset<SoundEffect>[] _soundNpcHit = new Asset<SoundEffect>[58];
+		internal Asset<SoundEffect>[] _soundNpcHit = new Asset<SoundEffect>[SoundID.NPCHitCount];
-		private SoundEffectInstance[] _soundInstanceNpcHit = new SoundEffectInstance[58];
+		internal SoundEffectInstance[] _soundInstanceNpcHit = new SoundEffectInstance[SoundID.NPCHitCount];
-		private Asset<SoundEffect>[] _soundNpcKilled = new Asset<SoundEffect>[SoundID.NPCDeathCount];
+		internal Asset<SoundEffect>[] _soundNpcKilled = new Asset<SoundEffect>[SoundID.NPCDeathCount];
-		private SoundEffectInstance[] _soundInstanceNpcKilled = new SoundEffectInstance[SoundID.NPCDeathCount];
+		internal SoundEffectInstance[] _soundInstanceNpcKilled = new SoundEffectInstance[SoundID.NPCDeathCount];
 		private SoundEffectInstance _soundInstanceMoonlordCry;
 		private Asset<SoundEffect> _soundDoorOpen;
 		private SoundEffectInstance _soundInstanceDoorOpen;
@@ -319,6 +_,8 @@
 
 					SoundEffectInstance soundEffectInstance = null;
 					switch (type) {
+						case int _ when SoundLoader.PlayModSound(type, num, num2, num3, ref soundEffectInstance):
+							break;
 						case 0: {
 								int num15 = Main.rand.Next(3);
 								if (_soundInstanceDig[num15] != null)
