--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/tModLoader/Terraria/Audio/LegacyAudioSystem.cs
@@ -1,8 +_,9 @@
 using Microsoft.Xna.Framework.Audio;
 using ReLogic.Content.Sources;
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using Terraria.ModLoader.Engine;
 
 namespace Terraria.Audio
 {
@@ -96,9 +_,11 @@
 		}
 
 		public LegacyAudioSystem() {
-			Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
-			SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
+			var contentManager = (TMLContentManager)Main.instance.Content;
+
+			engine = new AudioEngine(contentManager.GetPath("TerrariaMusic.xgs"));
+			SoundBank = new SoundBank(Engine, contentManager.GetPath("Sound Bank.xsb"));
 			Engine.Update();
-			WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
+			WaveBank = new WaveBank(Engine, contentManager.GetPath("Wave Bank.xwb"), 0, 512);
 			Engine.Update();
 			AudioTracks = new IAudioTrack[90];
@@ -112,5 +_,7 @@
 				yield return null;
 			}
 		}
+
+		internal Cue GetCueInternal(string cueName) => soundBank.GetCue(cueName);
 
 		public void LoadCue(int cueIndex, string cueName) {
