--- src/Terraria/Terraria/Audio/LegacySoundPlayer.cs
+++ src/tModLoader/Terraria/Audio/LegacySoundPlayer.cs
@@ -41,8 +_,8 @@
 		public SoundEffectInstance SoundInstancePixie;
 		public Asset<SoundEffect>[] SoundItem = new Asset<SoundEffect>[SoundID.ItemSoundCount];
 		public SoundEffectInstance[] SoundInstanceItem = new SoundEffectInstance[SoundID.ItemSoundCount];
-		public Asset<SoundEffect>[] SoundNpcHit = new Asset<SoundEffect>[58];
+		public Asset<SoundEffect>[] SoundNpcHit = new Asset<SoundEffect>[SoundID.NPCHitCount];
-		public SoundEffectInstance[] SoundInstanceNpcHit = new SoundEffectInstance[58];
+		public SoundEffectInstance[] SoundInstanceNpcHit = new SoundEffectInstance[SoundID.NPCHitCount];
 		public Asset<SoundEffect>[] SoundNpcKilled = new Asset<SoundEffect>[SoundID.NPCDeathCount];
 		public SoundEffectInstance[] SoundInstanceNpcKilled = new SoundEffectInstance[SoundID.NPCDeathCount];
 		public SoundEffectInstance SoundInstanceMoonlordCry;
