--- src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
+++ src/tModLoader/Terraria/Audio/LegacyAudioSystem.cs
@@ -4,6 +_,7 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.IO;
+using Terraria.ModLoader.Engine;
 
 namespace Terraria.Audio
 {
@@ -98,10 +_,12 @@
 		}
 
 		public LegacyAudioSystem() {
-			Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
-			SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
+			var contentManager = (TMLContentManager)Main.instance.Content;
+
+			Engine = new AudioEngine(contentManager.GetPath("TerrariaMusic.xgs"));
+			SoundBank = new SoundBank(Engine, contentManager.GetPath("Sound Bank.xsb"));
 			Engine.Update();
-			WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
+			WaveBank = new WaveBank(Engine, contentManager.GetPath("Wave Bank.xwb"), 0, 512);
 			Engine.Update();
 			AudioTracks = new IAudioTrack[90];
 			TrackNamesByIndex = new Dictionary<int, string>();
@@ -114,6 +_,8 @@
 				yield return null;
 			}
 		}
+
+		internal Cue GetCueInternal(string cueName) => SoundBank.GetCue(cueName);
 
 		public void LoadCue(int cueIndex, string cueName) {
 			CueAudioTrack cueAudioTrack = new CueAudioTrack(SoundBank, cueName);
