--- src/Terraria/Terraria/Audio/SoundEngine.cs
+++ src/tModLoader/Terraria/Audio/SoundEngine.cs
@@ -9,8 +_,8 @@
 	public static class SoundEngine
 	{
 		public static readonly object AudioLock = new object();
-		private static LegacySoundPlayer _legacyPlayer;
+		internal static LegacySoundPlayer _legacyPlayer;
-		private static SoundPlayer _player;
+		internal static SoundPlayer _player;
 		private static bool _areSoundsPaused;
 
 		public static bool IsAudioSupported {
