--- src/TerrariaNetCore/Terraria/Audio/SoundStyle.cs
+++ src/tModLoader/Terraria/Audio/SoundStyle.cs
@@ -1,3 +_,6 @@
+// Completely reimplemented in SoundStyle.TML.cs
+
+/*
 using Microsoft.Xna.Framework.Audio;
 using Terraria.Utilities;
 
@@ -35,3 +_,4 @@
 	public float GetRandomPitch() => _random.NextFloat() * PitchVariance - PitchVariance * 0.5f;
 	public abstract SoundEffect GetRandomSound();
 }
+*/
