--- src/Terraria/Terraria/Audio/SoundEngine.cs
+++ src/tModLoader/Terraria/Audio/SoundEngine.cs
@@ -20,6 +_,8 @@
 
 		public static void Initialize() {
 			IsAudioSupported = TestAudioSupport();
+			if (!IsAudioSupported)
+				Utils.ShowFancyErrorMessage(Localization.Language.GetTextValue("tModLoader.AudioNotSupported"), ModLoader.UI.Interface.loadModsID);
 		}
 
 		public static void Load(IServiceProvider services) {
@@ -324,7 +_,7 @@
 				}
 			}
 			catch (NoAudioHardwareException) {
-				Console.WriteLine("No audio hardware found. Disabling all audio.");
+				ModLoader.Logging.tML.Warn("No audio hardware found. Disabling all audio.");
 				return false;
 			}
 			catch {
