--- src/Terraria/ReLogic/Peripherals/RGB/Logitech/LogitechDeviceGroup.cs
+++ src/TerrariaNetCore/ReLogic/Peripherals/RGB/Logitech/LogitechDeviceGroup.cs
@@ -1,5 +_,6 @@
 using System;
 using System.Collections.Generic;
+using System.Runtime.InteropServices;
 
 namespace ReLogic.Peripherals.RGB.Logitech;
 
@@ -8,6 +_,9 @@
 	private readonly List<RgbDevice> _devices = new List<RgbDevice>();
 	private bool _isInitialized;
 	private readonly VendorColorProfile _colorProfiles;
+#if NETCORE
+	private bool _initializationFailed;
+#endif
 
 	public LogitechDeviceGroup(VendorColorProfile colorProfiles)
 	{
@@ -18,6 +_,21 @@
 	{
 		if (_isInitialized)
 			return;
+
+#if NETCORE
+		if (_initializationFailed)
+			return;
+
+		// TML: Added additional parameters to make the dll actually load. See changes in NativeMethods.
+		/*
+		if (!NativeLibrary.TryLoad(NativeMethods.DLL_NAME, out _)) {
+		*/
+		if (!NativeLibrary.TryLoad(NativeMethods.DLL_NAME, System.Reflection.Assembly.GetEntryAssembly(), DllImportSearchPath.SafeDirectories, out _)) {
+			_initializationFailed = true;
+
+			return;
+		}
+#endif
 
 		try {
 			if (!NativeMethods.LogiLedInit())
