--- src/Terraria\Terraria.World.Generation\WorldGenerator.cs
+++ src/tModLoader\Terraria.World.Generation\WorldGenerator.cs
@@ -8,9 +_,9 @@
 {
 	public class WorldGenerator
 	{
-		private List<GenPass> _passes = new List<GenPass>();
-		private float _totalLoadWeight;
-		private int _seed;
+		internal List<GenPass> _passes = new List<GenPass>();
+		internal float _totalLoadWeight;
+		internal int _seed;
 
 		public WorldGenerator(int seed)
 		{

