--- src/Terraria/Terraria.World.Generation/WorldGenerator.cs
+++ src/tModLoader/Terraria.World.Generation/WorldGenerator.cs
@@ -1,15 +_,19 @@
+using System;
 using System.Collections.Generic;
 using System.Diagnostics;
 using Terraria.GameContent.UI.States;
+using Terraria.Localization;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
 using Terraria.Utilities;
 
 namespace Terraria.World.Generation
 {
 	public class WorldGenerator
 	{
-		private List<GenPass> _passes = new List<GenPass>();
+		internal List<GenPass> _passes = new List<GenPass>();
-		private float _totalLoadWeight;
+		internal float _totalLoadWeight;
-		private int _seed;
+		internal int _seed;
 
 		public WorldGenerator(int seed) {
 			_seed = seed;
@@ -38,7 +_,23 @@
 				Main.rand = new UnifiedRandom(_seed);
 				stopwatch.Start();
 				progress.Start(pass2.Weight);
+				try {
-				pass2.Apply(progress);
+					pass2.Apply(progress);
+				}
+				catch(Exception e) {
+					string message = string.Join (
+						"\n",
+						Language.GetTextValue("tModLoader.WorldGenError"),
+						pass2.Name,
+						e
+					);
+					Logging.tML.Error(message);
+					Interface.errorMessage.Show(message, 0);
+
+					//We need to shutdown the thread without it saving.
+					//TODO: Allow returning a bool to signify if it should save or not.
+					throw e;
+				}
 				progress.End();
 				stopwatch.Reset();
 			}
