--- src/Terraria\Terraria.World.Generation\WorldGenerator.cs
+++ src/tModLoader\Terraria.World.Generation\WorldGenerator.cs
@@ -8,9 +_,9 @@
 {
 	public class WorldGenerator
 	{
-		private List<GenPass> _passes = new List<GenPass>();
-		private float _totalLoadWeight;
-		private int _seed;
+		internal List<GenPass> _passes = new List<GenPass>();
+		internal float _totalLoadWeight;
+		internal int _seed;
 
 		public WorldGenerator(int seed)
 		{
@@ -44,7 +_,22 @@
 				Main.rand = new UnifiedRandom(this._seed);
 				stopwatch.Start();
 				progress.Start(current2.Weight);
-				current2.Apply(progress);
+				try
+				{
+					current2.Apply(progress);
+				}
+				catch(Exception e)
+				{
+					string message = string.Join
+					(
+						"\n",
+						Localization.Language.GetTextValue("tModLoader.WorldGenError"),
+						current2.Name,
+						e.Message
+					);
+					Main.MenuUI.SetState(new ModLoader.UI.UIWorldLoadCrash(message));
+					throw e;
+				}
 				progress.End();
 				stopwatch.Reset();
 			}

