--- src/Terraria\Terraria.World.Generation\WorldGenerator.cs
+++ src/tModLoader\Terraria.World.Generation\WorldGenerator.cs
@@ -7,8 +_,8 @@
 {
 	public class WorldGenerator
 	{
-		private List<GenPass> _passes = new List<GenPass>();
-		private float _totalLoadWeight;
+		internal List<GenPass> _passes = new List<GenPass>();
+		internal float _totalLoadWeight;
 
 		public void Append(GenPass pass)
 		{

