--- src/decompiled/ReLogic/Content/Readers/PngReader.cs
+++ src/Terraria/ReLogic/Content/Readers/PngReader.cs
@@ -1,5 +_,9 @@
-using Microsoft.Xna.Framework;
+using Color = Microsoft.Xna.Framework.Color;
 using Microsoft.Xna.Framework.Graphics;
+#if NETCORE
+using SixLabors.ImageSharp;
+using SixLabors.ImageSharp.PixelFormats;
+#endif
 using System;
 using System.IO;
 using System.Threading;
@@ -48,5 +_,7 @@
 		public void Dispose() {
 			Dispose(disposing: true);
 		}
+
+		public Type[] GetAssociatedTypes() => new[] { typeof(Texture2D) };
 	}
 }
