--- src/decompiled/Terraria.Audio/ActiveSound.cs
+++ src/Terraria/Terraria.Audio/ActiveSound.cs
@@ -1,3 +_,4 @@
+using System.Threading;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Audio;
 
@@ -38,12 +_,18 @@
 		}
 
 		private void Play() {
+#if !WINDOWS
+			Monitor.Enter(SoundEngine.AudioLock);
+#endif
 			SoundEffectInstance soundEffectInstance = Style.GetRandomSound().CreateInstance();
 			soundEffectInstance.Pitch += Style.GetRandomPitch();
 			soundEffectInstance.Play();
 			SoundInstanceGarbageCollector.Track(soundEffectInstance);
 			Sound = soundEffectInstance;
 			Update();
+#if !WINDOWS
+			Monitor.Exit(SoundEngine.AudioLock);
+#endif
 		}
 
 		public void Stop() {
