--- src/decompiled/Terraria.Audio/SoundEngine.cs
+++ src/Terraria/Terraria.Audio/SoundEngine.cs
@@ -8,6 +_,7 @@
 {
 	public static class SoundEngine
 	{
+		public static readonly object AudioLock = new object();
 		private static LegacySoundPlayer _legacyPlayer;
 		private static SoundPlayer _player;
 		private static bool _areSoundsPaused;
@@ -30,6 +_,9 @@
 
 		public static void Update() {
 			if (IsAudioSupported) {
+#if FNA
+				Main.engine?.Update();
+#endif
 				SoundInstanceGarbageCollector.Update();
 				bool flag = (!Main.hasFocus || Main.gamePaused) && Main.netMode == 0;
 				if (!_areSoundsPaused && flag)
@@ -59,7 +_,10 @@
 			if (!IsAudioSupported)
 				return null;
 
+#if FNA
+			lock(AudioLock)
+#endif
-			return _legacyPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+				return _legacyPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
 		}
 
 		public static ActiveSound GetActiveSound(SlotId id) {
