--- src/Terraria/Terraria/Initializers/ChatInitializer.cs
+++ src/tModLoader/Terraria/Initializers/ChatInitializer.cs
@@ -6,6 +_,7 @@
 
 namespace Terraria.Initializers
 {
+	// @todo: if need be, we can add register tag handlers here
 	public static class ChatInitializer
 	{
 		public static void Load() {
@@ -24,10 +_,12 @@
 				"name"
 			});
 
+			/* AchievementTagHandler removed since achievements also removed 
 			ChatManager.Register<AchievementTagHandler>(new string[2] {
 				"a",
 				"achievement"
 			});
+			*/
 
 			ChatManager.Register<GlyphTagHandler>(new string[2] {
 				"g",
