--- src/TerrariaNetCore/Terraria/Initializers/ChatInitializer.cs
+++ src/tModLoader/Terraria/Initializers/ChatInitializer.cs
@@ -2,10 +_,12 @@
 using Terraria.GameContent.UI;
 using Terraria.GameContent.UI.Chat;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 
 namespace Terraria.Initializers;
 
+//TODO: If need be, we can register tag handlers here
 public static class ChatInitializer
 {
 	public static void Load()
@@ -56,6 +_,13 @@
 			string name = EmoteID.Search.GetName(i);
 			string key = "EmojiCommand." + name;
 			ChatManager.Commands.AddAlias(Language.GetText(key), NetworkText.FromFormattable("{0} {1}", Language.GetText("ChatCommand.Emoji_1"), Language.GetText("EmojiName." + name)));
+		}
+
+		foreach (var modEmoteBubble in EmoteBubbleLoader.emoteBubbles)
+		{
+			// Vanilla uses 2 keys, one with the name and the other with /${namekey}. Since the name is only used for the command, we can avoid that for simplicity
+			var command = new LocalizedText(modEmoteBubble.Command.Key, $"/{modEmoteBubble.Command.Value.ToLower()}");
+			ChatManager.Commands.AddAlias(command, NetworkText.FromFormattable("{0} {1}", Language.GetText("ChatCommand.Emoji_1"), modEmoteBubble.Command));
 		}
 	}
 }
