--- src/TerrariaNetCore/Terraria/Initializers/ChatInitializer.cs
+++ src/tModLoader/Terraria/Initializers/ChatInitializer.cs
@@ -1,11 +_,15 @@
+using Microsoft.Xna.Framework.Input;
+using System.Text.RegularExpressions;
 using Terraria.Chat.Commands;
 using Terraria.GameContent.UI;
 using Terraria.GameContent.UI.Chat;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 
 namespace Terraria.Initializers;
 
+//TODO: If need be, we can register tag handlers here
 public static class ChatInitializer
 {
 	public static void Load()
@@ -50,6 +_,12 @@
 				string name = EmoteID.Search.GetName(i);
 				string key = "EmojiCommand." + name;
 				ChatManager.Commands.AddAlias(Language.GetText(key), NetworkText.FromFormattable("{0} {1}", Language.GetText("ChatCommand.Emoji_1"), Language.GetText("EmojiName." + name)));
+			}
+
+			foreach (var modEmoteBubble in EmoteBubbleLoader.emoteBubbles)
+			{
+				var command = new LocalizedText(modEmoteBubble.EmoteName.Key, $"/{Lang.GetEmojiName(modEmoteBubble.Type)}");
+				ChatManager.Commands.AddAlias(command, NetworkText.FromFormattable("{0} {1}", Language.GetText("ChatCommand.Emoji_1"), Lang.GetEmojiName(modEmoteBubble.Type)));
 			}
 		}
 	}
