--- src/TerrariaNetCore/Terraria/Initializers/ChatInitializer.cs
+++ src/tModLoader/Terraria/Initializers/ChatInitializer.cs
@@ -1,9 +_,11 @@
 using Terraria.Chat.Commands;
 using Terraria.GameContent.UI.Chat;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 
 namespace Terraria.Initializers;
 
+//TODO: If need be, we can register tag handlers here
 public static class ChatInitializer
 {
 	public static void Load()
