--- src/TerrariaNetCore/Terraria/Initializers/LinkButtonsInitializer.cs
+++ src/tModLoader/Terraria/Initializers/LinkButtonsInitializer.cs
@@ -3,6 +_,7 @@
 using Microsoft.Xna.Framework.Graphics;
 using ReLogic.Content;
 using Terraria.DataStructures;
+using Terraria.Social;
 
 namespace Terraria.Initializers;
 
@@ -18,13 +_,28 @@
 		titleLinks.Add(MakeSimpleButton("TitleLinks.Forums", "https://forums.terraria.org/index.php", 4));
 		titleLinks.Add(MakeSimpleButton("TitleLinks.Merch", "https://terraria.org/store", 5));
 		titleLinks.Add(MakeSimpleButton("TitleLinks.Wiki", "https://terraria.wiki.gg/", 6));
+
+		// Add TML links.
+		titleLinks = Main.tModLoaderTitleLinks;
+		titleLinks.Add(MakeSimpleButton("TitleLinks.Discord", "https://tmodloader.net/discord", 0));
+		titleLinks.Add(MakeSimpleButton("TitleLinks.Twitter", "https://twitter.com/tModLoader", 3));
+		titleLinks.Add(MakeSimpleButton("TitleLinks.Wiki", "https://github.com/tModLoader/tModLoader/wiki", 6));
+
+		bool showPatreon = SocialAPI.Mode != SocialMode.Steam;
+		if (showPatreon)
+			titleLinks.Add(MakeSimpleButton("TitleLinks.Patreon", "https://www.patreon.com/tmodloader", 7));
 	}
 
 	private static TitleLinkButton MakeSimpleButton(string textKey, string linkUrl, int horizontalFrameIndex)
 	{
+		/*
 		Asset<Texture2D> asset = Main.Assets.Request<Texture2D>("Images/UI/TitleLinkButtons");
 		Rectangle value = asset.Frame(7, 2, horizontalFrameIndex);
 		Rectangle value2 = asset.Frame(7, 2, horizontalFrameIndex, 1);
+		*/
+		Asset<Texture2D> asset = Terraria.ModLoader.UI.UICommon.tModLoaderTitleLinkButtonsTexture;
+		Rectangle value = asset.Frame(8, 2, horizontalFrameIndex);
+		Rectangle value2 = asset.Frame(8, 2, horizontalFrameIndex, 1);
 		value.Width--;
 		value.Height--;
 		value2.Width--;
