--- src/Terraria/Terraria/Initializers/UILinksInitializer.cs
+++ src/tModLoader/Terraria/Initializers/UILinksInitializer.cs
@@ -6,6 +_,9 @@
 using Terraria.GameContent.UI.States;
 using Terraria.GameInput;
 using Terraria.Social;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
+using Terraria.ModLoader.UI.ModBrowser;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -632,7 +_,7 @@
 				return PlayerInput.BuildCommand(Lang.misc[56].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + PlayerInput.BuildCommand(Lang.misc[64].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["HotbarMinus"], PlayerInput.ProfileGamepadUI.KeyStatus["HotbarPlus"]) + PlayerInput.BuildCommand(Lang.misc[70].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]) + PlayerInput.BuildCommand(Lang.misc[69].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["SmartSelect"]) + (flag6 ? PlayerInput.BuildCommand("Evict", false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseRight"]) : "");
 			};
 
-			for (int num12 = 600; num12 <= 650; num12++) {
+			for (int num12 = 600; num12 <= 698; num12++) {
 				UILinkPoint value13 = new UILinkPoint(num12, enabled: true, num12 + 10, num12 - 10, num12 - 1, num12 + 1);
 				cp14.LinkMap.Add(num12, value13);
 			}
@@ -642,7 +_,7 @@
 				if (num56 == 0)
 					num56 = 100;
 
-				for (int num57 = 0; num57 < 50; num57++) {
+				for (int num57 = 0; num57 < 98; num57++) {
 					cp14.LinkMap[600 + num57].Up = ((num57 % num56 == 0) ? (-1) : (600 + num57 - 1));
 					if (cp14.LinkMap[600 + num57].Up == -1) {
 						if (num57 >= num56 * 2)
@@ -1584,6 +_,13 @@
 					break;
 				case 7:
 					(Main.MenuUI.CurrentState as IHaveBackButtonCommand)?.HandleBackButtonUsage();
+					break;
+				case 100:
+					SoundEngine.PlaySound(11, -1, -1, 1, 1f, 0f);
+					Main.menuMode = UILinkPointNavigator.Shortcuts.BackButtonGoto;
+					break;
+				case 101:
+					Interface.modBrowser.BackClick(null, null);
 					break;
 			}
 		}
