--- src/TerrariaNetCore/Terraria/Initializers/UILinksInitializer.cs
+++ src/tModLoader/Terraria/Initializers/UILinksInitializer.cs
@@ -9,7 +_,10 @@
 using Terraria.GameContent.UI.States;
 using Terraria.GameInput;
 using Terraria.Localization;
 using Terraria.Social;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
+using Terraria.ModLoader.UI.ModBrowser;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -772,7 +_,8 @@
 			return text3;
 		};
 
+		// Increased (600 -> 698) for modded TownNPC support. #461
-		for (int num14 = 600; num14 <= 650; num14++) {
+		for (int num14 = 600; num14 <= 698; num14++) {
 			UILinkPoint value13 = new UILinkPoint(num14, enabled: true, num14 + 10, num14 - 10, num14 - 1, num14 + 1);
 			cp15.LinkMap.Add(num14, value13);
 		}
@@ -782,7 +_,8 @@
 			if (num58 == 0)
 				num58 = 100;
 
+			// Increased (50 -> 98) for modded TownNPC support. #461
-			for (int num59 = 0; num59 < 50; num59++) {
+			for (int num59 = 0; num59 < 98; num59++) {
 				cp15.LinkMap[600 + num59].Up = ((num59 % num58 == 0) ? (-1) : (600 + num59 - 1));
 				if (cp15.LinkMap[600 + num59].Up == -1) {
 					if (num59 >= num58 * 2)
@@ -796,6 +_,11 @@
 				cp15.LinkMap[600 + num59].Down = (((num59 + 1) % num58 == 0 || num59 == UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - 1) ? 308 : (600 + num59 + 1));
 				cp15.LinkMap[600 + num59].Left = ((num59 < UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - num58) ? (600 + num59 + num58) : (-3));
 				cp15.LinkMap[600 + num59].Right = ((num59 < num58) ? (-4) : (600 + num59 - num58));
+				// Avoid navigating to crafting menu when a lot of TownNPC. #4790
+				if (cp15.LinkMap[600 + num59].Down >= 698)
+					cp15.LinkMap[600 + num59].Down = 308;
+				if (cp15.LinkMap[600 + num59].Left >= 698)
+					cp15.LinkMap[600 + num59].Left = -3;
 			}
 		};
 
@@ -1555,5 +_,16 @@
 		cp4.OnSpecialInteracts += () => (Main.InGameUI.CurrentState is UIBestiaryTest) ? (PlayerInput.BuildCommand(Lang.misc[82].Value, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + PlayerInput.BuildCommand(Language.GetText("UI.SwitchPage").Value, PlayerInput.ProfileGamepadUI.KeyStatus["HotbarMinus"], PlayerInput.ProfileGamepadUI.KeyStatus["HotbarPlus"]) + PlayerInput.BuildCommand(Lang.misc[53].Value, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]) + FancyUISpecialInstructions()) : (PlayerInput.BuildCommand(Lang.misc[53].Value, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]) + PlayerInput.BuildCommand(Lang.misc[82].Value, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + FancyUISpecialInstructions());
 		cp4.UpdateEvent += delegate {
+			bool flag11 = PlayerInput.Triggers.JustPressed.Inventory;
+			if (Main.inputTextEscape) {
+				Main.inputTextEscape = false;
+				flag11 = true;
+			}
+
+			if (CanExecuteInputCommand() && flag11)
+				FancyExit();
+
+			/* Above: Copied over Main.inputTextEscape bypass from MainMenu (1000) to FancyUI (1004)
 			if (CanExecuteInputCommand() && PlayerInput.Triggers.JustPressed.Inventory)
 				FancyExit();
+			*/
 		};
@@ -1808,6 +_,12 @@
 				if (Main.MenuUI.CurrentState is IHaveBackButtonCommand haveBackButtonCommand)
 					haveBackButtonCommand.HandleBackButtonUsage();
 				break;
+			// Added by TML: [[
+			case 100:
+				SoundEngine.PlaySound(11, -1, -1, 1, 1f, 0f);
+				Main.menuMode = UILinkPointNavigator.Shortcuts.BackButtonGoto;
+				break;
+			// ]]
 		}
 	}
 
