--- src/TerrariaNetCore/Terraria/Initializers/UILinksInitializer.cs
+++ src/tModLoader/Terraria/Initializers/UILinksInitializer.cs
@@ -10,6 +_,9 @@
 using Terraria.GameInput;
 using Terraria.Localization;
 using Terraria.Social;
+using Terraria.ModLoader;
+using Terraria.ModLoader.UI;
+using Terraria.ModLoader.UI.ModBrowser;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
@@ -772,7 +_,8 @@
 			return text3;
 		};
 
+		// Increased (600 -> 698) for modded TownNPC support. #461
-		for (int num14 = 600; num14 <= 650; num14++) {
+		for (int num14 = 600; num14 <= 698; num14++) {
 			UILinkPoint value13 = new UILinkPoint(num14, enabled: true, num14 + 10, num14 - 10, num14 - 1, num14 + 1);
 			cp15.LinkMap.Add(num14, value13);
 		}
@@ -782,9 +_,10 @@
 			if (num58 == 0)
 				num58 = 100;
 
+			// Increased (50 -> 98) for modded TownNPC support. #461
-			for (int num59 = 0; num59 < 50; num59++) {
+			for (int num59 = 0; num59 < 98; num59++) {
-				cp15.LinkMap[600 + num59].Up = ((num59 % num58 == 0) ? (-1) : (600 + num59 - 1));
+				cp14.LinkMap[600 + num59].Up = ((num59 % num58 == 0) ? (-1) : (600 + num59 - 1));
-				if (cp15.LinkMap[600 + num59].Up == -1) {
+				if (cp14.LinkMap[600 + num59].Up == -1) {
 					if (num59 >= num58 * 2)
 						cp15.LinkMap[600 + num59].Up = 307;
 					else if (num59 >= num58)
@@ -793,9 +_,14 @@
 						cp15.LinkMap[600 + num59].Up = 305;
 				}
 
-				cp15.LinkMap[600 + num59].Down = (((num59 + 1) % num58 == 0 || num59 == UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - 1) ? 308 : (600 + num59 + 1));
+				cp14.LinkMap[600 + num59].Down = (((num59 + 1) % num58 == 0 || num59 == UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - 1) ? 308 : (600 + num59 + 1));
-				cp15.LinkMap[600 + num59].Left = ((num59 < UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - num58) ? (600 + num59 + num58) : (-3));
+				cp14.LinkMap[600 + num59].Left = ((num59 < UILinkPointNavigator.Shortcuts.NPCS_IconsTotal - num58) ? (600 + num59 + num58) : (-3));
-				cp15.LinkMap[600 + num59].Right = ((num59 < num58) ? (-4) : (600 + num59 - num58));
+				cp14.LinkMap[600 + num59].Right = ((num59 < num58) ? (-4) : (600 + num59 - num58));
+				// Avoid navigating to crafting menu when a lot of TownNPC. #4790
+				if (cp14.LinkMap[600 + num59].Down >= 698)
+					cp14.LinkMap[600 + num59].Down = 308;
+				if (cp14.LinkMap[600 + num59].Left >= 698)
+					cp14.LinkMap[600 + num59].Left = -3;
 			}
 		};
 
@@ -986,20 +_,23 @@
 				}
 			}
 
-			cp12.LinkMap[1501].Left = -3;
-			if (num54 > 0)
-				cp12.LinkMap[1500 + num54 - 1].Right = -4;
-
-			cp12.LinkMap[1500].Down = ((num54 >= 2) ? 1502 : (-2));
-			cp12.LinkMap[1500].Left = ((num54 >= 1) ? 1501 : (-3));
-			cp12.LinkMap[1500].Up = 11001;
-			cp12.LinkMap[11001].Left = (Main.InPipCrafting ? 1501 : 12000);
-			cp12.LinkMap[11001].Down = ((!Main.InPipCrafting) ? (-1) : (Main.InGuideCraftMenu ? 1500 : 11003));
-			cp12.LinkMap[11001].Right = ((!Main.bannerUI.AnyAvailableBanners || Main.InGuideCraftMenu) ? (-1) : 11002);
-			cp12.LinkMap[11001].Up = (flag3 ? (-1) : 1502);
-			cp12.LinkMap[11002].Down = ((!Main.InPipCrafting) ? (-1) : 11003);
-			cp12.LinkMap[11002].Up = (flag3 ? (-1) : ((num54 >= 5) ? 1503 : 1502));
-			cp12.LinkMap[11003].Down = (flag3 ? 1502 : (-1));
+			bool flag4 = Main.mouseItem.stack <= 0;
+			/*
+			if (flag4 || (Main.mouseItem.type == createItem.type && Main.mouseItem.stack < Main.mouseItem.maxStack))
+			*/
+			if (flag4 || (Main.mouseItem.type == createItem.type && Main.mouseItem.stack < Main.mouseItem.maxStack && ItemLoader.CanStack(Main.mouseItem, createItem)))
+				text2 = ((!flag4) ? (text2 + PlayerInput.BuildCommand(Lang.misc[72].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"])) : (text2 + PlayerInput.BuildCommand(Lang.misc[72].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"], PlayerInput.ProfileGamepadUI.KeyStatus["MouseRight"])));
+
+			/*
+			if (!flag4 && Main.mouseItem.type == createItem.type && Main.mouseItem.stack < Main.mouseItem.maxStack)
+			*/
+			if (!flag4 && Main.mouseItem.type == createItem.type && Main.mouseItem.stack < Main.mouseItem.maxStack && ItemLoader.CanStack(Main.mouseItem, createItem))
+				text2 += PlayerInput.BuildCommand(Lang.misc[93].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseRight"]);
+
+			if (flag3)
+				text2 += PlayerInput.BuildCommand(Lang.misc[71].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]);
+
+			return text2 + HandleItem2();
 		};
 
 		cp12.LinkMap[1501].OnSpecialInteracts += () => ItemSlot.GetCraftSlotGamepadInstructions();
@@ -1554,10 +_,23 @@
 			cp4.LinkMap.Add(num28, new UILinkPoint(num28, enabled: true, -3, -4, -1, -2));
 		}
 
-		cp4.OnSpecialInteracts += () => (Main.InGameUI.CurrentState is UIBestiaryTest) ? (PlayerInput.BuildCommand(Lang.misc[82].Value, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + PlayerInput.BuildCommand(Language.GetText("UI.SwitchPage").Value, PlayerInput.ProfileGamepadUI.KeyStatus["HotbarMinus"], PlayerInput.ProfileGamepadUI.KeyStatus["HotbarPlus"]) + PlayerInput.BuildCommand(Lang.misc[53].Value, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]) + FancyUISpecialInstructions()) : (PlayerInput.BuildCommand(Lang.misc[53].Value, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]) + PlayerInput.BuildCommand(Lang.misc[82].Value, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + FancyUISpecialInstructions());
-		cp4.UpdateEvent += delegate {
+		cp3.OnSpecialInteracts += () => PlayerInput.BuildCommand(Lang.misc[53].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]) + PlayerInput.BuildCommand(Lang.misc[82].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Inventory"]) + FancyUISpecialInstructions();
+		cp3.UpdateEvent += delegate {
+			bool flag11 = PlayerInput.Triggers.JustPressed.Inventory;
+			if (Main.inputTextEscape) {
+				Main.inputTextEscape = false;
+				flag11 = true;
+			}
+
+			if (CanExecuteInputCommand() && flag11)
+				FancyExit();
+
+			/* Above: Copied over Main.inputTextEscape bypass from MainMenu (1000) to FancyUI (1004) because tModLoader has some FancyUI/UIState menus that have text input.
 			if (CanExecuteInputCommand() && PlayerInput.Triggers.JustPressed.Inventory)
 				FancyExit();
+			*/
+
+			UILinkPointNavigator.Shortcuts.BackButtonInUse = false;
 		};
 
 		cp4.EnterEvent += delegate {
@@ -1810,6 +_,12 @@
 				if (Main.MenuUI.CurrentState is IHaveBackButtonCommand haveBackButtonCommand)
 					haveBackButtonCommand.HandleBackButtonUsage();
 				break;
+			// Added by TML: [[
+			case 100:
+				SoundEngine.PlaySound(11, -1, -1, 1, 1f, 0f);
+				Main.menuMode = UILinkPointNavigator.Shortcuts.BackButtonGoto;
+				break;
+			// ]]
 		}
 	}
 
