--- src/Terraria/Terraria/GameContent/NetModules/NetAmbienceModule.cs
+++ src/tModLoader/Terraria/GameContent/NetModules/NetAmbienceModule.cs
@@ -1,9 +_,7 @@
 using System.IO;
 using Terraria.GameContent.Ambience;
-#if CLIENT
 using Terraria.GameContent.Skies;
 using Terraria.Graphics.Effects;
-#endif
 using Terraria.Net;
 
 namespace Terraria.GameContent.NetModules
@@ -20,9 +_,10 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int userId) {
-#if SERVER
-			return false;
-#else
+			if (Main.dedServ) {
+				return false;
+			}
+
 			byte playerId = reader.ReadByte();
 			int seed = reader.ReadInt32();
 			SkyEntityType type = (SkyEntityType)reader.ReadByte();
@@ -31,7 +_,6 @@
 			});
 
 			return true;
-#endif
 		}
 	}
 }
