--- src/Terraria/Terraria/GameContent/NetModules/NetTextModule.cs
+++ src/tModLoader/Terraria/GameContent/NetModules/NetTextModule.cs
@@ -40,11 +_,12 @@
 		}
 
 		public override bool Deserialize(BinaryReader reader, int senderPlayerId) {
-#if CLIENT
-			return DeserializeAsClient(reader, senderPlayerId);
-#else
-			return DeserializeAsServer(reader, senderPlayerId);
-#endif
+			if (!Main.dedServ) {
+				return DeserializeAsClient(reader, senderPlayerId);
+			}
+			else {
+				return DeserializeAsServer(reader, senderPlayerId);
+			}
 		}
 	}
 }
