--- src/TerrariaNetCore/Terraria/GameContent/NetModules/NetTeleportPylonModule.cs
+++ src/tModLoader/Terraria/GameContent/NetModules/NetTeleportPylonModule.cs
@@ -14,20 +_,24 @@
 		}
 
 		public static NetPacket SerializePylonWasAddedOrRemoved(TeleportPylonInfo info, SubPacketType packetType) {
-			NetPacket result = NetModule.CreatePacket<NetTeleportPylonModule>(6);
+			NetPacket result = NetModule.CreatePacket<NetTeleportPylonModule>(32);
 			result.Writer.Write((byte)packetType);
 			result.Writer.Write(info.PositionInTiles.X);
 			result.Writer.Write(info.PositionInTiles.Y);
 			result.Writer.Write((byte)info.TypeOfPylon);
+			result.Writer.Write(info.ModName);
+			result.Writer.Write(info.ModPylonName);
 			return result;
 		}
 
 		public static NetPacket SerializeUseRequest(TeleportPylonInfo info) {
-			NetPacket result = NetModule.CreatePacket<NetTeleportPylonModule>(6);
+			NetPacket result = NetModule.CreatePacket<NetTeleportPylonModule>(32);
 			result.Writer.Write((byte)2);
 			result.Writer.Write(info.PositionInTiles.X);
 			result.Writer.Write(info.PositionInTiles.Y);
 			result.Writer.Write((byte)info.TypeOfPylon);
+			result.Writer.Write(info.ModName);
+			result.Writer.Write(info.ModPylonName);
 			return result;
 		}
 
@@ -37,9 +_,11 @@
 						if (Main.dedServ) // #if SERVER
 							return false;
 
-						TeleportPylonInfo info3 = default(TeleportPylonInfo);
+						TeleportPylonInfo info3 = new TeleportPylonInfo();
 						info3.PositionInTiles = new Point16(reader.ReadInt16(), reader.ReadInt16());
 						info3.TypeOfPylon = (TeleportPylonType)reader.ReadByte();
+						info3.ModName = reader.ReadString();
+						info3.ModPylonName = reader.ReadString();
 						Main.PylonSystem.AddForClient(info3);
 						break;
 					}
@@ -47,17 +_,21 @@
 						if (Main.dedServ) // #if SERVER
 							return false;
 
-						TeleportPylonInfo info2 = default(TeleportPylonInfo);
+						TeleportPylonInfo info2 = new TeleportPylonInfo();
 						info2.PositionInTiles = new Point16(reader.ReadInt16(), reader.ReadInt16());
 						info2.TypeOfPylon = (TeleportPylonType)reader.ReadByte();
+						info2.ModName = reader.ReadString();
+						info2.ModPylonName = reader.ReadString();
 						Main.PylonSystem.RemoveForClient(info2);
 						break;
 
 				}
 				case 2: {
-						TeleportPylonInfo info = default(TeleportPylonInfo);
+						TeleportPylonInfo info = new TeleportPylonInfo();
 						info.PositionInTiles = new Point16(reader.ReadInt16(), reader.ReadInt16());
 						info.TypeOfPylon = (TeleportPylonType)reader.ReadByte();
+						info.ModName = reader.ReadString();
+						info.ModPylonName = reader.ReadString();
 						Main.PylonSystem.HandleTeleportRequest(info, userId);
 						break;
 					}
