--- src/Terraria\Terraria.Graphics\TileBatch.cs
+++ src/tModLoader\Terraria.Graphics\TileBatch.cs
@@ -498,6 +_,16 @@
 			this.InternalDraw(texture, destination, null, colors, 0f, Vector2.Zero, SpriteEffects.None, 0f, true);
 		}
 
+		// I made these 2 up.
+		public void Draw(Texture2D texture, Vector4 destination, Rectangle? sourceRectangle, VertexColors colors)
+		{
+			this.InternalDraw(texture, destination, sourceRectangle, colors, 0f, Vector2.Zero, SpriteEffects.None, 0f, true);
+		}
+		internal void Draw(Texture2D texture, Vector4 destinationRectangle, Rectangle? sourceRectangle, VertexColors colors, Vector2 origin, SpriteEffects effect, float rotation)
+		{
+			this.InternalDraw(texture, destinationRectangle, sourceRectangle, colors, rotation, origin, effect, 0f, true);
+		}
+
 		public void Draw(Texture2D texture, Vector2 position, Rectangle? sourceRectangle, VertexColors colors, Vector2 origin, float scale, SpriteEffects effects)
 		{
 			float z;

