--- src/TerrariaNetCore/Terraria/GameContent/Items/ItemVariant.cs
+++ src/tModLoader/Terraria/GameContent/Items/ItemVariant.cs
@@ -2,10 +_,20 @@
 
 namespace Terraria.GameContent.Items;
 
+/// <summary>
+/// Describes a variant of an <see cref="Item"/>.
+/// </summary>
 public class ItemVariant
 {
+	/// <summary>
+	/// The localized description of this <see cref="ItemVariant"/>.
+	/// </summary>
 	public readonly NetworkText Description;
 
+	/// <summary>
+	/// Creates a new <see cref="ItemVariant"/> with the provided localized description.
+	/// </summary>
+	/// <param name="description">The localized description to use.</param>
 	public ItemVariant(NetworkText description)
 	{
 		Description = description;
