--- src/TerrariaNetCore/Terraria/GameContent/Items/TagEffectState.cs
+++ src/tModLoader/Terraria/GameContent/Items/TagEffectState.cs
@@ -79,20 +_,26 @@
 		Array.Clear(ProcTimeLeftOnNPC, 0, ProcTimeLeftOnNPC.Length);
 	}
 
+	/*
 	public void ModifyHit(Projectile optionalProjectile, NPC npcHit, ref TagDamageChanges changes)
+	*/
+	public void ModifyHit(Projectile optionalProjectile, NPC npcHit, ref NPC.HitModifiers modifiers)
 	{
 		if (IsNPCTagged(npcHit.whoAmI) && _effect.CanRunHitEffects(_owner, optionalProjectile, npcHit)) {
-			_effect.ModifyTaggedHit(_owner, optionalProjectile, npcHit, ref changes);
+			_effect.ModifyTaggedHit(_owner, optionalProjectile, npcHit, ref modifiers);
 			if (CanProcOnNPC(npcHit.whoAmI))
-				_effect.ModifyProcHit(_owner, optionalProjectile, npcHit, ref changes);
+				_effect.ModifyProcHit(_owner, optionalProjectile, npcHit, ref modifiers);
 		}
 	}
 
+	/*
 	public void OnHit(Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	*/
+	public void OnHit(Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit)
 	{
 		if (IsNPCTagged(npcHit.whoAmI) && _effect.CanRunHitEffects(_owner, optionalProjectile, npcHit)) {
-			_effect.OnTaggedHit(_owner, optionalProjectile, npcHit, calcDamage);
+			_effect.OnTaggedHit(_owner, optionalProjectile, npcHit, hit);
-			if (CanProcOnNPC(npcHit.whoAmI) && _effect.OnProcHit(_owner, optionalProjectile, npcHit, calcDamage))
+ 			if (CanProcOnNPC(npcHit.whoAmI) && _effect.OnProcHit(_owner, optionalProjectile, npcHit, hit))
 				ClearProcOnNPC(npcHit.whoAmI);
 		}
 	}
