--- src/TerrariaNetCore/Terraria/GameContent/Items/TagEffectState.cs
+++ src/tModLoader/Terraria/GameContent/Items/TagEffectState.cs
@@ -242,7 +_,10 @@
 		}
 	}
 
+	/*
 	public void ModifyHit(Projectile optionalProjectile, NPC npcHit, ref int damageDealt, ref bool crit)
+	*/
+	public void ModifyHit(Projectile optionalProjectile, NPC npcHit, ref NPC.HitModifiers damageDealt, ref bool crit)
 	{
 		if (_effect != null && IsNPCTagged(npcHit.whoAmI) && _effect.CanRunHitEffects(_owner, optionalProjectile, npcHit)) {
 			_effect.ModifyTaggedHit(_owner, optionalProjectile, npcHit, ref damageDealt, ref crit);
@@ -251,7 +_,7 @@
 		}
 	}
 
-	public void OnHit(Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	public void OnHit(Projectile optionalProjectile, NPC npcHit, float calcDamage)
 	{
 		if (_effect != null && IsNPCTagged(npcHit.whoAmI) && _effect.CanRunHitEffects(_owner, optionalProjectile, npcHit)) {
 			_effect.OnTaggedHit(_owner, optionalProjectile, npcHit, calcDamage);
