--- src/TerrariaNetCore/Terraria/GameContent/Items/WhipTagEffect_ElectricEel.cs
+++ src/tModLoader/Terraria/GameContent/Items/WhipTagEffect_ElectricEel.cs
@@ -7,14 +_,14 @@
 
 public class WhipTagEffect_ElectricEel : WhipTagEffect
 {
-	public override void ModifyProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, ref TagDamageChanges changes)
+	public override void ModifyProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, ref NPC.HitModifiers modifiers)
 	{
-		CreateLightningBlast(owner, npcHit, ref changes);
+		CreateLightningBlast(owner, npcHit, ref modifiers);
 	}
 
-	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, int calcDamage) => true;
+	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit) => true;
 
-	public static void CreateLightningBlast(Player owner, NPC npcHit, ref TagDamageChanges changes)
+	public static void CreateLightningBlast(Player owner, NPC npcHit, ref NPC.HitModifiers modifiers)
 	{
 		int num = 0;
 		int num2 = 5;
