--- src/TerrariaNetCore/Terraria/GameContent/Items/WhipTagEffect_Starcrash.cs
+++ src/tModLoader/Terraria/GameContent/Items/WhipTagEffect_Starcrash.cs
@@ -5,16 +_,22 @@
 
 public class WhipTagEffect_Starcrash : WhipTagEffect
 {
+	/*
 	public override void OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	*/
+	public override void OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit)
 	{
-		SpawnMeteorWhipMeteorOn(optionalProjectile, npcHit, calcDamage);
+		SpawnMeteorWhipMeteorOn(optionalProjectile, npcHit, hit);
 	}
 
+	/*
 	private void SpawnMeteorWhipMeteorOn(Projectile projectile, NPC targetNPC, int calcDamage)
+	*/
+	private void SpawnMeteorWhipMeteorOn(Projectile projectile, NPC targetNPC, NPC.HitInfo hit)
 	{
 		int num = 200;
 		int num2 = 600;
-		int damage = (int)((float)calcDamage * 1.33f);
+		int damage = (int)(hit.SourceDamage * 1.33f);
 		Vector2 vector = new Vector2(-num + Main.rand.Next(num * 2), -num2);
 		Vector2 vector2 = targetNPC.Center + vector;
 		Vector2 vector3 = vector.SafeNormalize(Vector2.Zero) * -12f;
