--- src/TerrariaNetCore/Terraria/GameContent/Items/WhipTagEffect_ViolentDisplayOfFlower.cs
+++ src/tModLoader/Terraria/GameContent/Items/WhipTagEffect_ViolentDisplayOfFlower.cs
@@ -6,6 +_,9 @@
 public class WhipTagEffect_ViolentDisplayOfFlower : WhipTagEffect
 {
+	/*
 	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	*/
+	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit)
 	{
 		if (optionalProjectile != null)
 			SpawnFlowerExplosionOn(optionalProjectile, npcHit, calcDamage);
