--- src/TerrariaNetCore/Terraria/GameContent/Items/WhipTagEffect_Kaleidoscope.cs
+++ src/tModLoader/Terraria/GameContent/Items/WhipTagEffect_Kaleidoscope.cs
@@ -4,7 +_,10 @@
 
 public class WhipTagEffect_Kaleidoscope : WhipTagEffect
 {
+	/*
 	public override void OnTaggedHit(Player owner, Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	*/
+	public override void OnTaggedHit(Player owner, Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit)
 	{
 		ParticleOrchestrator.RequestParticleSpawn(clientOnly: false, ParticleOrchestraType.RainbowRodHit, new ParticleOrchestraSettings {
 			PositionInWorld = optionalProjectile.Center
