--- src/TerrariaNetCore/Terraria/GameContent/Items/WhipTagEffect_Possession.cs
+++ src/tModLoader/Terraria/GameContent/Items/WhipTagEffect_Possession.cs
@@ -3,7 +_,10 @@
 public class WhipTagEffect_Possession : WhipTagEffect
 {
+	/*
 	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, int calcDamage)
+	*/
+	public override bool OnProcHit(Player owner, Projectile optionalProjectile, NPC npcHit, NPC.HitInfo hit)
 	{
 		Projectile.SpawnMoonLordWhipProc(optionalProjectile, npcHit, 20, 0);
 		return true;
 	}
