--- src/TerrariaNetCore/Terraria/GameContent/Events/CultistRitual.cs
+++ src/tModLoader/Terraria/GameContent/Events/CultistRitual.cs
@@ -10,19 +_,20 @@
 	public const int respawnDelay = 43200;
 	private const int timePerCultist = 3600;
 	private const int recheckStart = 600;
+	//TML: Changed from int to double.
-	public static int delay;
+	public static double delay;
-	public static int recheck;
+	public static double recheck;
 
 	public static void UpdateTime()
 	{
 		if (Main.netMode == 1)
 			return;
 
-		delay -= Main.dayRate;
+		delay -= Main.desiredWorldEventsUpdateRate;
 		if (delay < 0)
 			delay = 0;
 
-		recheck -= Main.dayRate;
+		recheck -= Main.desiredWorldEventsUpdateRate;
 		if (recheck < 0)
 			recheck = 0;
 
