--- src/TerrariaNetCore/Terraria/GameContent/Events/MysticLogFairiesEvent.cs
+++ src/tModLoader/Terraria/GameContent/Events/MysticLogFairiesEvent.cs
@@ -9,7 +_,7 @@
 public class MysticLogFairiesEvent
 {
 	private bool _canSpawnFairies;
-	private int _delayUntilNextAttempt;
+	private double _delayUntilNextAttempt; //TML: Changed from int to double.
 	private const int DELAY_BETWEEN_ATTEMPTS = 60;
 	private List<Point> _stumpCoords = new List<Point>();
 
@@ -38,7 +_,7 @@
 	public void UpdateTime()
 	{
 		if (Main.netMode != 1 && _canSpawnFairies && IsAGoodTime()) {
-			_delayUntilNextAttempt = Math.Max(0, _delayUntilNextAttempt - Main.dayRate);
+			_delayUntilNextAttempt = Math.Max(0, _delayUntilNextAttempt - Main.desiredWorldEventsUpdateRate);
 			if (_delayUntilNextAttempt == 0) {
 				_delayUntilNextAttempt = 60;
 				TrySpawningFairies();
@@ -69,7 +_,8 @@
 
 		int oneOverSpawnChance = GetOneOverSpawnChance();
 		bool flag = false;
+
-		for (int i = 0; i < Main.dayRate; i++) {
+		for (int i = 0; i < Main.worldEventUpdates; i++) {
 			if (Main.rand.Next(oneOverSpawnChance) == 0) {
 				flag = true;
 				break;
