--- src/Terraria/Terraria/GameContent/Events/Sandstorm.cs
+++ src/tModLoader/Terraria/GameContent/Events/Sandstorm.cs
@@ -10,15 +_,17 @@
 		private const int SANDSTORM_DURATION_MINIMUM = 28800;
 		private const int SANDSTORM_DURATION_MAXIMUM = 86400;
 		public static bool Happening;
-		public static int TimeLeft;
+		public static double TimeLeft;
 		public static float Severity;
 		public static float IntendedSeverity;
 		private static bool _effectsUp;
+		private static double timePass;
 
 		private static bool HasSufficientWind() => Math.Abs(Main.windSpeedCurrent) >= 0.6f;
 
 		public static void WorldClear() {
 			Happening = false;
+			timePass = 0.0;
 		}
 
 		public static void UpdateTime() {
@@ -41,9 +_,14 @@
 					int num = 21600;
 					num = ((!Main.hardMode) ? (num * 3) : (num * 2));
 					if (HasSufficientWind()) {
+						timePass += Main.desiredWorldEventsUpdateRate;
+						if (timePass >= 1) {
+							double timePassReference = timePass;
-						for (int i = 0; i < Main.dayRate; i++) {
+							for (int i = 0; i < timePassReference; i++) {
-							if (Main.rand.Next(num) == 0)
+								if (Main.rand.Next(num) == 0)
-								StartSandstorm();
+									StartSandstorm();
+								timePass -= 1.0;
+							}
 						}
 					}
 				}
