--- src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureManager.cs
@@ -5,7 +_,8 @@
 {
 	public class CaptureManager
 	{
-		private static CaptureManager _instance = new CaptureManager();
+		// client initialization has been moved to LoadContent to prevent FNA deadlocks
+		private static CaptureManager _instance;
 		private CaptureInterface _interface;
 		private CaptureCamera _camera;
 
