--- src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureManager.cs
@@ -3,17 +_,18 @@
 
 namespace Terraria.Graphics.Capture
 {
-	public class CaptureManager
-	{
-		public static CaptureManager Instance = new CaptureManager();
+	public class CaptureManager {
+		//client initialization has been moved to LoadContent to prevent FNA deadlocks
+		public static CaptureManager Instance 
+		{
+			get { return Main.dedServ ? new CaptureManager() : _instance; }
+			set { _instance = value;  }
+		}
+		private static CaptureManager _instance;
 		private CaptureInterface _interface;
 		private CaptureCamera _camera;
 
-#if CLIENT
-		public bool IsCapturing => _camera.IsCapturing;
-#else
-		public bool IsCapturing => false;
-#endif
+		public bool IsCapturing => Main.dedServ ? false : _camera.IsCapturing;
 
 		public bool Active {
 			get {
@@ -36,9 +_,9 @@
 
 		public CaptureManager() {
 			_interface = new CaptureInterface();
-#if CLIENT
-			_camera = new CaptureCamera(Main.instance.GraphicsDevice);
-#endif
+			if (!Main.dedServ) {
+				_camera = new CaptureCamera(Main.instance.GraphicsDevice);
+			}
 		}
 
 		public void Scrolling() {
