--- src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureManager.cs
@@ -5,7 +_,11 @@
 {
 	public class CaptureManager
 	{
+#if SERVER
 		private static CaptureManager _instance;
+#else //client initialization has been moved to LoadContent to prevent FNA deadlocks
+		public static CaptureManager Instance;
+#endif
 		private CaptureInterface _interface;
 		private CaptureCamera _camera;
 
