--- src/TerrariaNetCore/Terraria/Graphics/Capture/CaptureBiome.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureBiome.cs
@@ -1,3 +_,4 @@
+using Terraria.ModLoader;
 namespace Terraria.Graphics.Capture;
 
 public class CaptureBiome
@@ -135,6 +_,13 @@
 				return captureBiome;
 		}
 
+		var sceneEffect = Main.LocalPlayer.CurrentSceneEffect;
+		if (sceneEffect.waterStyle.value >= Main.maxLiquidTypes) {
+			// This makes the capture match whatever the end result is of all modded effects, but that isn't quite right.
+			//TODO: Future improvement should be using the SceneEffect layer associated to background.
+			return new CaptureBiome(sceneEffect.surfaceBackground.value, sceneEffect.waterStyle.value, sceneEffect.tileColorStyle);
+		}
+
 		return null;
 	}
 
@@ -168,6 +_,13 @@
 			case 9:
 				return Styles.Mushroom;
 			default:
+				var sceneEffect = Main.LocalPlayer.CurrentSceneEffect;
+				if (sceneEffect.surfaceBackground.value >= Main.BG_STYLES_COUNT) {
+					// This makes the capture match whatever the end result is of all modded effects, but that isn't quite right.
+					//TODO: Future improvement should be using the SceneEffect layer associated to background.
+					return new CaptureBiome(sceneEffect.surfaceBackground.value, sceneEffect.waterStyle.value, sceneEffect.tileColorStyle);
+				}
+
 				return null;
 		}
 	}
