--- src/Terraria/Terraria/Graphics/Capture/CaptureCamera.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureCamera.cs
@@ -244,13 +_,15 @@
 			using (FileStream fileStream = File.Create(filename)) {
 #if !NETCORE
 				PlatformUtilities.SavePng(fileStream, width, height, width, height, _outputData);
-#elif CLIENT
-				//var tex = new Texture2D(Main.graphics.GraphicsDevice, width, height);
-				//tex.SetData(_outputData);
-				//tex.SaveAsPng(fileStream, width, height);
-				var image = Image.LoadPixelData<Rgba32>(_outputData, width, height);
-				image.SaveAsPng(fileStream);
+				return;
 #endif
+				if (!Main.dedServ) {
+					//var tex = new Texture2D(Main.graphics.GraphicsDevice, width, height);
+					//tex.SetData(_outputData);
+					//tex.SaveAsPng(fileStream, width, height);
+					var image = Image.LoadPixelData<Rgba32>(_outputData, width, height);
+					image.SaveAsPng(fileStream);
+				}
 			}
 		}
 #endif
@@ -334,7 +_,9 @@
 		}
 
 		public void Dispose() {
-#if CLIENT
+			if (Main.dedServ) {
+				return;
+			}
 			Monitor.Enter(_captureLock);
 			if (!_isDisposed) {
 				_frameBuffer.Dispose();
@@ -349,7 +_,6 @@
 				_isDisposed = true;
 				Monitor.Exit(_captureLock);
 			}
-#endif
 		}
 	}
 }
