--- src/Terraria/Terraria/Graphics/Capture/CaptureBiome.cs
+++ src/tModLoader/Terraria/Graphics/Capture/CaptureBiome.cs
@@ -1,3 +_,4 @@
+using Terraria.ModLoader;
 namespace Terraria.Graphics.Capture
 {
 	public class CaptureBiome
@@ -164,7 +_,11 @@
 				case 9:
 					return Styles.Mushroom;
 				default:
-					return null;
+					var atmos = Main.LocalPlayer.currentAVFX;
+					if (atmos.surfaceBG >= Loaders.SurfaceBgs.vanillaCount || atmos.waterStyle >= Loaders.Waters.vanillaCount) {
+						return new CaptureBiome(atmos.surfaceBG, atmos.waterStyle, atmos.tileColorStyle);
+					}
+					return null; 	
 			}
 		}
 
