--- src/decompiled/Terraria/Audio/LegacyAudioSystem.cs
+++ src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
@@ -68,6 +_,7 @@
 			if (!contentSource.HasAsset(assetPath))
 				continue;
 
+#if !SERVER
 			string extension = contentSource.GetExtension(assetPath);
 			try {
 				IAudioTrack audioTrack2 = null;
@@ -93,6 +_,7 @@
 				Main.IssueReporter.AddReport(textToShow);
 				Main.IssueReporterIndicator.AttemptLettingPlayerKnow();
 			}
+#endif
 		}
 
 		return audioTrack;
@@ -101,6 +_,9 @@
 	public LegacyAudioSystem()
 	{
 		Engine = new AudioEngine("Content\\TerrariaMusic.xgs");
+#if FNA
+		Console.WriteLine("SDL Audio Driver: " + SDL3.SDL.SDL_GetCurrentAudioDriver());
+#endif
 		SoundBank = new SoundBank(Engine, "Content\\Sound Bank.xsb");
 		Engine.Update();
 		WaveBank = new WaveBank(Engine, "Content\\Wave Bank.xwb", 0, 512);
