--- src/decompiled/Terraria/Audio/LegacyAudioSystem.cs
+++ src/Terraria/Terraria/Audio/LegacyAudioSystem.cs
@@ -58,6 +_,11 @@
 		private IAudioTrack FindReplacementTrack(List<IContentSource> sources, string assetPath) {
 			IAudioTrack audioTrack = null;
 			for (int i = 0; i < sources.Count; i++) {
+				if (Main.dedServ) {
+					sources[i].HasAsset(assetPath);
+					continue;
+				}
+				
 				IContentSource contentSource = sources[i];
 				if (!contentSource.HasAsset(assetPath))
 					continue;
