--- src/decompiled/Terraria/Audio/SoundEngine.cs
+++ src/Terraria/Terraria/Audio/SoundEngine.cs
@@ -8,6 +_,7 @@
 {
 	public static class SoundEngine
 	{
+		public static readonly object AudioLock = new object();
 		public static LegacySoundPlayer LegacySoundPlayer;
 		public static SoundPlayer SoundPlayer;
 		public static bool AreSoundsPaused;
@@ -69,13 +_,23 @@
 		}
 
 		public static SoundEffectInstance PlaySound(int type, int x = -1, int y = -1, int Style = 1, float volumeScale = 1f, float pitchOffset = 0f) {
+			if (Main.dedServ) // #if SERVER
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
+#if FNA
+			lock(AudioLock)
+#endif
-			return LegacySoundPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+				return LegacySoundPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+
 		}
 
 		public static ActiveSound GetActiveSound(SlotId id) {
+			if (Main.dedServ) // #if SERVER
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
@@ -83,6 +_,9 @@
 		}
 
 		public static SlotId PlayTrackedSound(SoundStyle style, Vector2 position) {
+			if (Main.dedServ) // #if SERVER
+				return SlotId.Invalid;
+
 			if (!IsAudioSupported)
 				return SlotId.Invalid;
 
@@ -90,6 +_,9 @@
 		}
 
 		public static SlotId PlayTrackedSound(SoundStyle style) {
+			if (Main.dedServ) // #if SERVER
+				return SlotId.Invalid;
+
 			if (!IsAudioSupported)
 				return SlotId.Invalid;
 
@@ -102,6 +_,9 @@
 		}
 
 		public static SoundEffect GetTrackableSoundByStyleId(int id) {
+			if (Main.dedServ) // #if SERVER
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
@@ -114,6 +_,9 @@
 		}
 
 		public static ActiveSound FindActiveSound(SoundStyle style) {
+			if (Main.dedServ) // #if SERVER
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
