--- src/decompiled/Terraria/Audio/SoundEngine.cs
+++ src/Terraria/Terraria/Audio/SoundEngine.cs
@@ -8,6 +_,7 @@
 {
 	public static class SoundEngine
 	{
+		public static readonly object AudioLock = new object();
 		public static LegacySoundPlayer LegacySoundPlayer;
 		public static SoundPlayer SoundPlayer;
 		public static bool AreSoundsPaused;
@@ -69,31 +_,52 @@
 		}
 
 		public static SoundEffectInstance PlaySound(int type, int x = -1, int y = -1, int Style = 1, float volumeScale = 1f, float pitchOffset = 0f) {
+			if (Main.dedServ)
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
+#if FNA
+			lock(AudioLock)
+#endif
-			return LegacySoundPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+				return LegacySoundPlayer.PlaySound(type, x, y, Style, volumeScale, pitchOffset);
+
 		}
 
 		public static ActiveSound GetActiveSound(SlotId id) {
+			if (Main.dedServ)
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
 			return SoundPlayer.GetActiveSound(id);
+
 		}
 
-		public static SlotId PlayTrackedSound(SoundStyle style, Vector2 position) {
+		public static SlotId PlayTrackedSound(SoundStyle style, Vector2 position)
+		{
+			if(Main.dedServ)
+				return SlotId.Invalid;
+
 			if (!IsAudioSupported)
 				return SlotId.Invalid;
 
 			return SoundPlayer.Play(style, position);
+
 		}
 
-		public static SlotId PlayTrackedSound(SoundStyle style) {
+		public static SlotId PlayTrackedSound(SoundStyle style)
+		{
+			if(Main.dedServ)
+				return SlotId.Invalid;
+
 			if (!IsAudioSupported)
 				return SlotId.Invalid;
 
 			return SoundPlayer.Play(style);
+
 		}
 
 		public static void StopTrackedSounds() {
@@ -102,6 +_,9 @@
 		}
 
 		public static SoundEffect GetTrackableSoundByStyleId(int id) {
+			if(Main.dedServ)
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
@@ -114,6 +_,9 @@
 		}
 
 		public static ActiveSound FindActiveSound(SoundStyle style) {
+			if(Main.dedServ)
+				return null;
+
 			if (!IsAudioSupported)
 				return null;
 
