--- src/decompiled/Terraria/Audio/SoundPlayer.cs
+++ src/Terraria/Terraria/Audio/SoundPlayer.cs
@@ -1,6 +_,9 @@
 using Microsoft.Xna.Framework;
 using ReLogic.Utilities;
 using System.Collections.Generic;
+#if FNA
+using System.Threading;
+#endif
 
 namespace Terraria.Audio
 {
@@ -59,6 +_,9 @@
 		}
 
 		public void Update() {
+#if FNA
+			Monitor.Enter(SoundEngine.AudioLock);
+#endif
 			foreach (SlotVector<ActiveSound>.ItemPair item in (IEnumerable<SlotVector<ActiveSound>.ItemPair>)_trackedSounds) {
 				try {
 					item.Value.Update();
@@ -69,6 +_,9 @@
 					_trackedSounds.Remove(item.Id);
 				}
 			}
+#if FNA
+			Monitor.Exit(SoundEngine.AudioLock);
+#endif
 		}
 
 		public ActiveSound FindActiveSound(SoundStyle style) {
