--- src/TerrariaNetCore/Terraria/WorldBuilding/Passes.cs
+++ src/tModLoader/Terraria/WorldBuilding/Passes.cs
@@ -13,6 +_,7 @@
 
 		protected override void ApplyPass(GenerationProgress progress, GameConfiguration configuration)
 		{
+			/*
 			for (int i = 0; i < GenBase._worldWidth; i++) {
 				for (int j = 0; j < GenBase._worldHeight; j++) {
 					if (GenBase._tiles[i, j] == null)
@@ -21,6 +_,8 @@
 						GenBase._tiles[i, j].ClearEverything();
 				}
 			}
+			*/
+			GenBase._tiles.ClearEverything();
 		}
 	}
 
