--- src/Terraria/WorldBuilding/WorldGenerator.cs
+++ src/tModLoader/WorldBuilding/WorldGenerator.cs
@@ -6,9 +_,9 @@
 {
 	public class WorldGenerator
 	{
-		private readonly List<GenPass> _passes = new List<GenPass>();
+		internal readonly List<GenPass> _passes = new List<GenPass>();
-		private float _totalLoadWeight;
+		internal float _totalLoadWeight;
-		private readonly int _seed;
+		internal readonly int _seed;
 		private readonly WorldGenConfiguration _configuration;
 		public static GenerationProgress CurrentGenerationProgress;
 
