--- src/TerrariaNetCore/Terraria/GameContent/Tile_Entities/TEDeadCellsDisplayJar.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TEDeadCellsDisplayJar.cs
@@ -1,9 +_,12 @@
 using System.IO;
+using Microsoft.Xna.Framework;
 using Terraria.DataStructures;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 
 namespace Terraria.GameContent.Tile_Entities;
 
-public class TEDeadCellsDisplayJar : TileEntityType<TEDeadCellsDisplayJar>, IFixLoadedData
+public partial class TEDeadCellsDisplayJar : TileEntityType<TEDeadCellsDisplayJar>, IFixLoadedData
 {
 	public Item item;
 
@@ -35,8 +_,8 @@
 
 	public override void WriteExtraData(BinaryWriter writer, bool networkSend)
 	{
-		writer.Write((short)item.type);
-		writer.Write(item.prefix);
+		ItemIO.WriteShortVanillaID(item, writer);
+		ItemIO.WriteByteVanillaPrefix(item, writer);
 		writer.Write((short)item.stack);
 	}
 
@@ -53,30 +_,38 @@
 	public void DropItem()
 	{
 		if (Main.netMode != 1)
-			Item.NewItem(new EntitySource_TileBreak(Position.X, Position.Y), Position.X * 16, Position.Y * 16, 32, 32, item.type, 1, noBroadcast: false, item.prefix);
+			Item.NewItem(new EntitySource_TileBreak(Position.X, Position.Y), Position.X * 16, Position.Y * 16, 32, 32, item, noBroadcast: false);
 
 		item = new Item();
 	}
 
+	// TML: replaced type and prefix with item
-	public static void TryPlacing(int x, int y, int type, int prefix, int stack)
+	public static void TryPlacing(int x, int y, Item item, int stack)
 	{
 		WorldGen.RangeFrame(x, y, x + 1, y + 2);
 		if (!TileEntity.TryGetAt<TEDeadCellsDisplayJar>(x, y, out var result)) {
+			/*
 			int num = Item.NewItem(new EntitySource_TileBreak(x, y), x * 16, y * 16, 32, 32, 1);
 			Main.item[num].SetDefaults(type);
 			Main.item[num].Prefix(prefix);
 			Main.item[num].stack = stack;
 			NetMessage.SendData(21, -1, -1, null, num);
+			*/
+			Item.NewItem(new EntitySource_TileBreak(x, y), new Rectangle(x * 16, y * 16, 32, 32), item);
 			return;
 		}
 
 		if (result.item.stack > 0)
 			result.DropItem();
 
+		/*
 		result.item = new Item();
 		result.item.SetDefaults(type);
 		result.item.Prefix(prefix);
 		result.item.stack = stack;
+		*/
+		result.item = ItemLoader.TransferWithLimit(item, stack);
+
 		NetMessage.SendData(86, -1, -1, null, result.ID, x, y);
 	}
 
@@ -126,16 +_,20 @@
 				NetMessage.SendData(17, -1, -1, null, 0, Player.tileTargetX, y, 1f);
 		}
 
-		if (Main.netMode == 1)
+		if (Main.netMode == 1) {
 			NetMessage.SendData(149, -1, -1, null, x, y, player.selectedItem, player.whoAmI, 1);
+			ItemLoader.TransferWithLimit(player.inventory[player.selectedItem], 1);
+		}
 		else
-			TryPlacing(x, y, player.inventory[player.selectedItem].type, player.inventory[player.selectedItem].prefix, 1);
+			TryPlacing(x, y, player.inventory[player.selectedItem], 1);
 
+		/*
 		player.inventory[player.selectedItem].stack--;
 		if (player.inventory[player.selectedItem].stack <= 0) {
 			player.inventory[player.selectedItem].SetDefaults(0);
 			Main.mouseItem.SetDefaults(0);
 		}
+		*/
 
 		if (player.selectedItem == 58)
 			Main.mouseItem = player.inventory[player.selectedItem].Clone();
