--- src/TerrariaNetCore/Terraria/GameContent/Tile_Entities/TEDisplayDoll.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TEDisplayDoll.cs
@@ -6,10 +_,12 @@
 using Terraria.Audio;
 using Terraria.DataStructures;
 using Terraria.GameInput;
 using Terraria.Graphics.Renderers;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.UI;
 
 namespace Terraria.GameContent.Tile_Entities;
 
-public class TEDisplayDoll : TileEntityType<TEDisplayDoll>, IFixLoadedData
+public sealed partial class TEDisplayDoll : TileEntityType<TEDisplayDoll>, IFixLoadedData
 {
@@ -168,8 +_,8 @@
 		Item[] equip = _equip;
 		foreach (Item item in equip) {
 			if (!item.IsAir) {
-				writer.Write((short)item.type);
-				writer.Write(item.prefix);
+				ItemIO.WriteShortVanillaID(item, writer);
+				ItemIO.WriteByteVanillaPrefix(item, writer);
 				writer.Write((short)item.stack);
 			}
 		}
@@ -177,8 +_,8 @@
 		equip = _dyes;
 		foreach (Item item2 in equip) {
 			if (!item2.IsAir) {
-				writer.Write((short)item2.type);
-				writer.Write(item2.prefix);
+				ItemIO.WriteShortVanillaID(item2, writer);
+				ItemIO.WriteByteVanillaPrefix(item2, writer);
 				writer.Write((short)item2.stack);
 			}
 		}
@@ -278,11 +_,14 @@
 			return;
 
 		TileEntityType<TEDisplayDoll>.Kill(num, num2);
+		if (!TileLoader.Drop(callX, callY, 470))
+			goto skipDrops;
 		if (Main.tile[callX, callY].frameX / 72 != 1)
 			Item.NewItem(new EntitySource_TileBreak(num, num2), num * 16, num2 * 16, 32, 48, 498);
 		else
 			Item.NewItem(new EntitySource_TileBreak(num, num2), num * 16, num2 * 16, 32, 48, 1989);
 
+		skipDrops:
 		WorldGen.destroyObject = true;
 		for (int k = num; k < num + 2; k++) {
 			for (int l = num2; l < num2 + 3; l++) {
@@ -671,18 +_,28 @@
 	public void WriteItem(int itemIndex, BinaryWriter writer, Item[] collection)
 	{
 		Item item = collection[itemIndex];
+		
+		ItemIO.Send(item, writer, writeStack: true);
+		return;
+		
 		writer.Write((ushort)item.type);
 		writer.Write((ushort)item.stack);
 		writer.Write(item.prefix);
 	}
 
 	public void ReadItem(int itemIndex, BinaryReader reader, Item[] collection)
 	{
+		/*
 		int num = reader.ReadUInt16();
 		int stack = reader.ReadUInt16();
 		int prefixWeWant = reader.ReadByte();
+		*/
 		if (itemIndex < collection.Length) {
 			Item obj = collection[itemIndex];
+			
+			ItemIO.Receive(item, reader, readStack: true);
+			return;
+			
 			obj.SetDefaults(num);
 			obj.stack = stack;
 			obj.Prefix(prefixWeWant);
