--- src/TerrariaNetCore/Terraria/GameContent/Tile_Entities/TEDisplayDoll.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TEDisplayDoll.cs
@@ -4,11 +_,13 @@
 using Terraria.Audio;
 using Terraria.DataStructures;
 using Terraria.GameInput;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.UI;
 
 namespace Terraria.GameContent.Tile_Entities;
 
-public class TEDisplayDoll : TileEntity, IFixLoadedData
+public sealed partial class TEDisplayDoll : TileEntity, IFixLoadedData
 {
 	private static byte _myEntityID;
 	private const int MyTileID = 470;
@@ -118,8 +_,8 @@
 		for (int i = 0; i < 8; i++) {
 			Item item = _items[i];
 			if (!item.IsAir) {
-				writer.Write((short)item.netID);
-				writer.Write(item.prefix);
+				ItemIO.WriteShortVanillaID(item, writer);
+				ItemIO.WriteByteVanillaPrefix(item, writer);
 				writer.Write((short)item.stack);
 			}
 		}
@@ -127,8 +_,8 @@
 		for (int j = 0; j < 8; j++) {
 			Item item2 = _dyes[j];
 			if (!item2.IsAir) {
-				writer.Write((short)item2.netID);
-				writer.Write(item2.prefix);
+				ItemIO.WriteShortVanillaID(item2, writer);
+				ItemIO.WriteByteVanillaPrefix(item2, writer);
 				writer.Write((short)item2.stack);
 			}
 		}
@@ -187,11 +_,14 @@
 			return;
 
 		Kill(num, num2);
+		if (!TileLoader.Drop(callX, callY, 470))
+			goto skipDrops;
 		if (Main.tile[callX, callY].frameX / 72 != 1)
 			Item.NewItem(new EntitySource_TileBreak(num, num2), num * 16, num2 * 16, 32, 48, 498);
 		else
 			Item.NewItem(new EntitySource_TileBreak(num, num2), num * 16, num2 * 16, 32, 48, 1989);
 
+		skipDrops:
 		WorldGen.destroyObject = true;
 		for (int k = num; k < num + 2; k++) {
 			for (int l = num2; l < num2 + 3; l++) {
@@ -468,6 +_,9 @@
 		if (dye)
 			item = _dyes[itemIndex];
 
+		ItemIO.Send(item, writer, writeStack: true);
+		return;
+
 		writer.Write((ushort)item.netID);
 		writer.Write((ushort)item.stack);
 		writer.Write(item.prefix);
@@ -475,12 +_,19 @@
 
 	public void ReadItem(int itemIndex, BinaryReader reader, bool dye)
 	{
+		Item item = (dye ? _dyes : _items)[itemIndex];
+
+		ItemIO.Receive(item, reader, readStack: true);
+		return;
+
 		int defaults = reader.ReadUInt16();
 		int stack = reader.ReadUInt16();
 		int prefixWeWant = reader.ReadByte();
+		/*
 		Item item = _items[itemIndex];
 		if (dye)
 			item = _dyes[itemIndex];
+		*/
 
 		item.SetDefaults(defaults);
 		item.stack = stack;
