--- src/Terraria/Terraria/GameContent/Tile_Entities/TEHatRack.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TEHatRack.cs
@@ -4,11 +_,13 @@
 using Terraria.Audio;
 using Terraria.DataStructures;
 using Terraria.GameInput;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.UI;
 
 namespace Terraria.GameContent.Tile_Entities
 {
-	public class TEHatRack : TileEntity
+	public sealed partial class TEHatRack : TileEntity
 	{
 		private static byte _myEntityID;
 		private const int MyTileID = 475;
