--- src/TerrariaNetCore/Terraria/GameContent/Tile_Entities/TETrainingDummy.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TETrainingDummy.cs
@@ -5,7 +_,7 @@
 
 namespace Terraria.GameContent.Tile_Entities;
 
-public class TETrainingDummy : TileEntityType<TETrainingDummy>
+public sealed class TETrainingDummy : TileEntityType<TETrainingDummy>
 {
 	private static List<Rectangle> playerBoxes = new List<Rectangle>();
 	private static bool playerBoxFilled;
