--- src/Terraria/Terraria/GameContent/Tile_Entities/TEWeaponsRack.cs
+++ src/tModLoader/Terraria/GameContent/Tile_Entities/TEWeaponsRack.cs
@@ -1,10 +_,12 @@
 using System.IO;
 using Terraria.DataStructures;
 using Terraria.ID;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 
 namespace Terraria.GameContent.Tile_Entities
 {
-	public class TEWeaponsRack : TileEntity
+	public sealed partial class TEWeaponsRack : TileEntity
 	{
 		private static byte _myEntityID;
 		public Item item;
@@ -73,16 +_,11 @@
 		}
 
 		public override void WriteExtraData(BinaryWriter writer, bool networkSend) {
-			writer.Write((short)item.netID);
-			writer.Write(item.prefix);
-			writer.Write((short)item.stack);
+			ItemIO.Write(item, writer, !networkSend || ModNet.AllowVanillaClients, writeStack: true);
 		}
 
 		public override void ReadExtraData(BinaryReader reader, bool networkSend) {
-			item = new Item();
-			item.netDefaults(reader.ReadInt16());
-			item.Prefix(reader.ReadByte());
-			item.stack = reader.ReadInt16();
+			item = ItemIO.Read(reader, !networkSend || ModNet.AllowVanillaClients, readStack: true);
 		}
 
 		public override string ToString() => Position.X + "x  " + Position.Y + "y item: " + item;
