--- src/decompiled\Terraria.GameContent.UI.Elements\UIHeader.cs
+++ src/merged\Terraria.GameContent.UI.Elements\UIHeader.cs
@@ -21,9 +_,11 @@
 				if (this._text != value)
 				{
 					this._text = value;
+#if CLIENT
 					Vector2 vector = Main.fontDeathText.MeasureString(this.Text);
 					this.Width.Pixels = vector.X;
 					this.Height.Pixels = vector.Y;
+#endif
 					this.Width.Precent = 0f;
 					this.Height.Precent = 0f;
 					this.Recalculate();

