--- src/Terraria/Terraria.GameContent.Liquid/LiquidRenderer.cs
+++ src/tModLoader/Terraria.GameContent.Liquid/LiquidRenderer.cs
@@ -74,9 +_,13 @@
 			0
 		};
 		public const float MIN_LIQUID_SIZE = 0.25f;
+#if SERVER
 		public static LiquidRenderer Instance = new LiquidRenderer();
+#else //client initialization has been moved to LoadContent to prevent FNA deadlocks
+		public static LiquidRenderer Instance;
+#endif
-		private Tile[,] _tiles = Main.tile;
+		//private Tile[,] _tiles = Main.tile;
-		private Texture2D[] _liquidTextures = new Texture2D[12];
+		public Texture2D[] _liquidTextures = new Texture2D[12];
 		private LiquidCache[] _cache = new LiquidCache[1];
 		private LiquidDrawCache[] _drawCache = new LiquidDrawCache[1];
 		private int _animationFrame;
@@ -112,7 +_,7 @@
 				ptr2 = ptr;
 				for (int i = rectangle.X; i < rectangle.X + rectangle.Width; i++) {
 					for (int j = rectangle.Y; j < rectangle.Y + rectangle.Height; j++) {
-						tile = _tiles[i, j];
+						tile = Main.tile[i, j];
 						if (tile == null)
 							tile = new Tile();
 
