--- src/merged\Terraria.GameInput\PlayerInput.cs
+++ src/Terraria\Terraria.GameInput\PlayerInput.cs
@@ -2008,7 +_,7 @@
 			Vector2 value4 = value - vector;
 			Vector2 value5 = value2 - vector;
 			Vector2 value6 = value3 - vector;
-			arg_AF_0 - vector;
+			Vector2 _ = arg_AF_0 - vector;
 			float scaleFactor = 1f / Main.GameViewMatrix.Zoom.X;
 			float num = 1f;
 			Vector2 arg_118_0 = vector - Main.screenPosition + value4 * scaleFactor;

