--- src/merged\Terraria.GameInput\PlayerInput.cs
+++ src/Terraria\Terraria.GameInput\PlayerInput.cs
@@ -2008,7 +_,7 @@
 			Vector2 value5 = value - vector;
 			Vector2 value6 = value2 - vector;
 			Vector2 value7 = value3 - vector;
-			value4 - vector;
+			Vector2 _ = value4 - vector;
 			float scaleFactor = 1f / Main.GameViewMatrix.Zoom.X;
 			float num = 1f;
 			Vector2 vector2 = vector - Main.screenPosition + value5 * scaleFactor;

