--- src/TerrariaNetCore/ReLogic/OS/Windows/NativeMethods.cs
+++ src/tModLoader/ReLogic/OS/Windows/NativeMethods.cs
@@ -185,5 +_,15 @@
 		}
 
 		private static bool DeleteCompletelySilent(string path) => DeleteFile(path, FileOperationFlags.FOF_SILENT | FileOperationFlags.FOF_NOCONFIRMATION | FileOperationFlags.FOF_NOERRORUI);
+
+		[DllImport("kernel32.dll")]
+		static extern IntPtr GetConsoleWindow();
+		[DllImport("user32.dll")]
+		static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);
+
+		public static void HideConsole() {
+			IntPtr hWnd = GetConsoleWindow();
+			ShowWindow(hWnd, 0);
+		}
 	}
 }
