--- src/TerrariaNetCore/ReLogic/OS/Windows/WindowService.cs
+++ src/tModLoader/ReLogic/OS/Windows/WindowService.cs
@@ -1,5 +_,6 @@
 using System;
 using System.Drawing;
+using System.IO;
 using System.Runtime.InteropServices;
 using Microsoft.Xna.Framework;
 
@@ -54,5 +_,41 @@
 	{
 		NativeMethods.FlashInfo flashInfo = NativeMethods.FlashInfo.CreateStop(window.Handle);
 		NativeMethods.FlashWindowEx(ref flashInfo);
+	}
+
+	// Added by TML.
+	public void HideConsole()
+	{
+		NativeMethods.HideConsole();
+	}
+
+	public void ShowConsole()
+	{
+		NativeMethods.ShowConsole();
+	}
+
+	public void ReleaseConsole()
+	{
+		NativeMethods.FreeConsole();
+	}
+
+	public void CreateAndRedirectConsole()
+	{
+		// Allocate a new console window
+		if (NativeMethods.AllocConsole()) {
+			// Redirect Console.Out to the new console's output stream
+			IntPtr stdOutHandle = NativeMethods.GetStdHandle(0xFFFFFFF5);
+			if (stdOutHandle != IntPtr.Zero) {
+				// Reopen standard output stream to the new console
+				TextWriter writer = new StreamWriter(Console.OpenStandardOutput()) { AutoFlush = true };
+				Console.SetOut(writer);
+			}
+			else {
+				Console.WriteLine("Failed to get standard output handle.");
+			}
+		}
+		else {
+			Console.WriteLine("Failed to allocate console.");
+		}
 	}
 }
