--- src/Terraria/Terraria/Social/Steam/CoreSocialModule.cs
+++ src/tModLoader/Terraria/Social/Steam/CoreSocialModule.cs
@@ -1,6 +_,7 @@
 using ReLogic.OS;
 using Steamworks;
 using System;
+using System.IO;
 using System.Threading;
 using System.Windows.Forms;
 using Terraria.Localization;
@@ -10,7 +_,7 @@
 	public class CoreSocialModule : ISocialModule
 	{
 		private static CoreSocialModule _instance;
-		public const int SteamAppId = 105600;
+		public const uint SteamAppId = ModLoader.Engine.Steam.TMLAppID; // No references to this?
 		private bool IsSteamValid;
 		private object _steamTickLock = new object();
 		private object _steamCallbackLock = new object();
@@ -24,13 +_,15 @@
 
 		public void Initialize() {
 			_instance = this;
-			if (!Main.dedServ && SteamAPI.RestartAppIfNecessary(new AppId_t(105600u))) {
+			if (System.Diagnostics.Debugger.IsAttached && !File.Exists("steam_appid.txt"))
+				ModLoader.Engine.Steam.EnsureSteamAppIdTMLFile();
+			if (SteamAPI.RestartAppIfNecessary(ModLoader.Engine.Steam.TMLAppID_t) && !System.Diagnostics.Debugger.IsAttached) {
 				Environment.Exit(1);
 				return;
 			}
 
 			if (!SteamAPI.Init()) {
-				MessageBox.Show(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
+				ModLoader.UI.Interface.MessageBoxShow(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
 				Environment.Exit(1);
 			}
 
