--- src/TerrariaNetCore/Terraria/Social/Steam/CoreSocialModule.cs
+++ src/tModLoader/Terraria/Social/Steam/CoreSocialModule.cs
@@ -1,16 +_,22 @@
 using System;
+using System.IO;
 using System.Threading;
 using System.Windows.Forms;
 using ReLogic.OS;
 using Steamworks;
 using Terraria.Localization;
+using Terraria.ModLoader.Engine;
 
 namespace Terraria.Social.Steam;
 
-public class CoreSocialModule : ISocialModule
+public partial class CoreSocialModule : ISocialModule
 {
 	private static CoreSocialModule _instance;
+
+	/*
 	public const int SteamAppId = 105600;
+	*/
+
 	private bool IsSteamValid;
 	private object _steamTickLock = new object();
 	private object _steamCallbackLock = new object();
@@ -26,15 +_,18 @@
 	public void Initialize()
 	{
 		_instance = this;
+		/*
 		if (!Main.dedServ && SteamAPI.RestartAppIfNecessary(new AppId_t(105600u))) {
 			Environment.Exit(1);
 			return;
 		}
+		*/
+		ModLoader.Engine.Steam.SetAppId(ModLoader.Engine.Steam.TMLAppID_t);
 
 		if (!SteamAPI.Init()) {
-			MessageBox.Show(Language.GetTextValue("Error.LaunchFromSteam"), Language.GetTextValue("Error.Error"));
-			Environment.Exit(1);
+			ErrorReporting.FatalExit(Language.GetTextValue("Error.LaunchFromSteam"));
 		}
+		PortFilesToCurrentStructure();
 
 		IsSteamValid = true;
 		Thread thread = new Thread(SteamCallbackLoop);
