--- src/TerrariaNetCore/Terraria/Social/Steam/NetServerSocialModule.cs
+++ src/tModLoader/Terraria/Social/Steam/NetServerSocialModule.cs
@@ -118,8 +_,12 @@
 
 	private void OnLobbyCreated(LobbyCreated_t result, bool failure)
 	{
-		if (!failure)
+		if (!failure) {
 			SteamFriends.SetRichPresence("status", Language.GetTextValue("Social.StatusInGame"));
+
+			// Extra logging added by TML.
+			Utils.LogAndConsoleInfoMessage("Currently Hosting LobbyID: " + GetLobbyId());
+		}
 	}
 
 	private bool OnPacketRead(byte[] data, int length, CSteamID userId)
@@ -190,6 +_,10 @@
 
 			_connectionStateMap[steamIDRemote] = ConnectionState.Authenticating;
 			_connectionAcceptedCallback(new SocialSocket(new SteamAddress(steamIDRemote)));
+		}
+		else {
+			// Extra logging added by TML.
+			Utils.LogAndConsoleInfoMessage("User " + SteamFriends.GetFriendPersonaName(steamIDRemote) + "connection rejected. Not a friend of lobby Owner.\nSet allow Friends of Friends to accept non-Friends of Host.");
 		}
 	}
 }
