--- src/decompiled/Terraria/Graphics/Capture/CaptureManager.cs
+++ src/Terraria/Terraria/Graphics/Capture/CaptureManager.cs
@@ -6,11 +_,25 @@
 {
 	public class CaptureManager : IDisposable
 	{
+		// 'Terraria' patchstep: Client initialization has been moved to LoadContent to prevent FNA deadlocks
+		
+		/*
 		public static CaptureManager Instance = new CaptureManager();
+		*/
+		private static CaptureManager _instance;
+		
 		private CaptureInterface _interface;
 		private CaptureCamera _camera;
 
+		public static CaptureManager Instance {
+			get => Main.dedServ ? new CaptureManager() : _instance;
+			set => _instance = value;
+		}
+
+		/*
 		public bool IsCapturing => _camera.IsCapturing;
+		*/
+		public bool IsCapturing => !Main.dedServ && _camera.IsCapturing;
 
 		public bool Active {
 			get {
@@ -33,6 +_,10 @@
 
 		public CaptureManager() {
 			_interface = new CaptureInterface();
+
+			if (Main.dedServ) // #if SERVER
+				return;
+
 			_camera = new CaptureCamera(Main.instance.GraphicsDevice);
 		}
 
