--- src/Terraria\Terraria.GameInput\PlayerInputProfile.cs
+++ src/tModLoader\Terraria.GameInput\PlayerInputProfile.cs
@@ -3,6 +_,8 @@
 using Newtonsoft.Json.Linq;
 using System;
 using System.Collections.Generic;
+using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput
 {
@@ -158,7 +_,7 @@
 		{
 			Dictionary<string, object> dictionary = new Dictionary<string, object>();
 			Dictionary<string, object> dictionary2 = new Dictionary<string, object>();
-			dictionary.Add("Last Launched Version", 193);
+			dictionary.Add("Last Launched Version", 194);
 			dictionary2.Add("Edittable", this.AllowEditting);
 			dictionary2.Add("Gamepad - HotbarRadialHoldTime", this.HotbarRadialHoldTimeRequired);
 			dictionary2.Add("Gamepad - LeftThumbstickDeadzoneX", this.LeftThumbstickDeadzoneX);
@@ -238,6 +_,9 @@
 			{
 				"MouseLeft",
 				"MouseRight",
+				"MouseMiddle",
+				"MouseXButton1",
+				"MouseXButton2",
 				"Up",
 				"Down",
 				"Left",
@@ -292,6 +_,25 @@
 			this.CopyKeysFrom(profile, mode, keysToCopy);
 		}
 
+		public void CopyModHotkeySettingsFrom(PlayerInputProfile profile, InputMode mode)
+		{
+			foreach (var modHotkey in HotKeyLoader.HotKeys) {
+				this.InputModes[mode].KeyStatus[modHotkey.uniqueName].Clear();
+				if (!string.IsNullOrEmpty(modHotkey.defaultKey)) {
+					this.InputModes[mode].KeyStatus[modHotkey.uniqueName].Add(modHotkey.defaultKey);
+				}
+			}
+		}
+
+		public void CopyIndividualModHotkeySettingsFrom(PlayerInputProfile profile, InputMode mode, string uniqueName)
+		{
+			var modHotKey = HotKeyLoader.modHotKeys[uniqueName];
+			this.InputModes[mode].KeyStatus[uniqueName].Clear();
+			if (!string.IsNullOrEmpty(modHotKey.defaultKey)) {
+				this.InputModes[mode].KeyStatus[uniqueName].Add(modHotKey.defaultKey);
+			}
+		}
+
 		public void CopyGamepadSettingsFrom(PlayerInputProfile profile, InputMode mode)
 		{
 			string[] keysToCopy = new string[]

