--- src/Terraria\Terraria.GameInput\PlayerInput.cs
+++ src/tModLoader\Terraria.GameInput\PlayerInput.cs
@@ -8,6 +_,7 @@
 using Terraria.GameContent.UI.States;
 using Terraria.ID;
 using Terraria.IO;
+using Terraria.ModLoader;
 using Terraria.Social;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
@@ -307,6 +_,42 @@
 			}
 		}
 
+		// To allow mod hotkeys to load, need to reinitialize after load.
+		public static void ReInitialize()
+		{
+			PlayerInput.Profiles.Clear();
+			PlayerInput.OriginalProfiles.Clear();
+			PlayerInputProfile playerInputProfile = new PlayerInputProfile("Redigit's Pick");
+			playerInputProfile.Initialize(PresetProfiles.Redigit);
+			PlayerInput.Profiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Yoraiz0r's Pick");
+			playerInputProfile.Initialize(PresetProfiles.Yoraiz0r);
+			PlayerInput.Profiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Console (Playstation)");
+			playerInputProfile.Initialize(PresetProfiles.ConsolePS);
+			PlayerInput.Profiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Console (Xbox)");
+			playerInputProfile.Initialize(PresetProfiles.ConsoleXBox);
+			PlayerInput.Profiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Custom");
+			playerInputProfile.Initialize(PresetProfiles.Redigit);
+			PlayerInput.Profiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Redigit's Pick");
+			playerInputProfile.Initialize(PresetProfiles.Redigit);
+			PlayerInput.OriginalProfiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Yoraiz0r's Pick");
+			playerInputProfile.Initialize(PresetProfiles.Yoraiz0r);
+			PlayerInput.OriginalProfiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Console (Playstation)");
+			playerInputProfile.Initialize(PresetProfiles.ConsolePS);
+			PlayerInput.OriginalProfiles.Add(playerInputProfile.Name, playerInputProfile);
+			playerInputProfile = new PlayerInputProfile("Console (Xbox)");
+			playerInputProfile.Initialize(PresetProfiles.ConsoleXBox);
+			PlayerInput.OriginalProfiles.Add(playerInputProfile.Name, playerInputProfile);
+			PlayerInput.Triggers.Initialize(); // Populates triggers with known hotkeys
+			PlayerInput.Load(); // Loads the JSON into the Profiles above
+		}
+
 		public static void Initialize()
 		{
 			Main.InputProfiles.OnProcessText += new Preferences.TextProcessAction(PlayerInput.PrettyPrintProfiles);
@@ -487,7 +_,14 @@
 			PlayerInput.VerifyBuildingMode();
 			PlayerInput.MouseInput();
 			PlayerInput.KeyboardInput();
-			PlayerInput.GamePadInput();
+			// ugly hack: While reloading,  Main.player[Main.myPlayer]'s Selected item might be a mod item, so
+			// bool flag4 = ItemID.Sets.GamepadWholeScreenUseRange[player.inventory[player.selectedItem].type] || player.scope;
+			// Will thrown an index out of range.
+			try
+			{
+				PlayerInput.GamePadInput();
+			}
+			catch { }
 			PlayerInput.Triggers.Update();
 			PlayerInput.PostInput();
 			PlayerInput.ScrollWheelDelta = PlayerInput.ScrollWheelValue - PlayerInput.ScrollWheelValueOld;
@@ -682,7 +_,8 @@
 				keyConfiguration.Processkey(PlayerInput.Triggers.Current, Buttons.RightTrigger.ToString());
 				flag = true;
 			}
-			bool flag4 = ItemID.Sets.GamepadWholeScreenUseRange[player.inventory[player.selectedItem].type] || player.scope;
+			// We might just need a bool for while reloading. 
+			bool flag4 = player.inventory[player.selectedItem].type < ItemID.Sets.GamepadWholeScreenUseRange.Length && ItemID.Sets.GamepadWholeScreenUseRange[player.inventory[player.selectedItem].type] || player.scope;
 			int num3 = player.inventory[player.selectedItem].tileBoost + ItemID.Sets.GamepadExtraRange[player.inventory[player.selectedItem].type];
 			if (player.yoyoString && ItemID.Sets.Yoyo[player.inventory[player.selectedItem].type])
 			{
@@ -1158,7 +_,7 @@
 						int num = 6;
 						if (ItemSlot.IsABuildingItem(player.inventory[player.selectedItem]))
 						{
-							num = player.inventory[player.selectedItem].useTime;
+							num = (int)(player.inventory[player.selectedItem].useTime / PlayerHooks.TotalUseTimeMultiplier(player, player.inventory[player.selectedItem]));
 						}
 						PlayerInput.DpadSnapCooldown[i] = num;
 						value += value2;
@@ -1462,6 +_,10 @@
 							c.KeyStatus["Hotbar8"].Add("D8");
 							c.KeyStatus["Hotbar9"].Add("D9");
 							c.KeyStatus["Hotbar10"].Add("D0");
+							foreach (var item in ModLoader.ModLoader.modHotKeys)
+							{
+								c.KeyStatus[item.Value.displayName].Add(item.Value.defaultKey);
+							}
 							return;
 						case InputMode.KeyboardUI:
 							c.KeyStatus["MouseLeft"].Add("Mouse1");

