--- src/Terraria\Terraria.GameInput\TriggersSet.cs
+++ src/tModLoader\Terraria.GameInput\TriggersSet.cs
@@ -2,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using Terraria.ModLoader;
 
 namespace Terraria.GameInput
 {
@@ -36,6 +_,42 @@
 			}
 		}
 
+		public bool MouseMiddle
+		{
+			get
+			{
+				return this.KeyStatus["MouseMiddle"];
+			}
+			set
+			{
+				this.KeyStatus["MouseMiddle"] = value;
+			}
+		}
+
+		public bool MouseXButton1
+		{
+			get
+			{
+				return this.KeyStatus["MouseXButton1"];
+			}
+			set
+			{
+				this.KeyStatus["MouseXButton1"] = value;
+			}
+		}
+
+		public bool MouseXButton2
+		{
+			get
+			{
+				return this.KeyStatus["MouseXButton2"];
+			}
+			set
+			{
+				this.KeyStatus["MouseXButton2"] = value;
+			}
+		}
+
 		public bool Up
 		{
 			get
@@ -660,6 +_,10 @@
 			foreach (string current in PlayerInput.KnownTriggers)
 			{
 				this.KeyStatus.Add(current, false);
+			}
+			foreach (var hotkey in HotKeyLoader.HotKeys)
+			{
+				this.KeyStatus.Add(hotkey.uniqueName, false);
 			}
 		}
 
@@ -737,6 +_,8 @@
 			{
 				this.HotbarScrollCD--;
 			}
+			ModLoader.ModHooks.HotKeyPressed();
+			ModLoader.PlayerHooks.ProcessTriggers(p, this);
 		}
 
 		public void CopyIntoDuringChat(Player p)

