--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_Buff.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_Buff.cs
@@ -1,14 +_,29 @@
+#nullable enable
+
+using Terraria.ID;
+using Terraria.ModLoader;
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Intended for mods to use when spawning projectiles periodically from buffs. <br/><br/>
+///
+/// Note that this can be used for both NPCs and Players
+/// </summary>
-public class EntitySource_Buff : IEntitySource
+public class EntitySource_Buff : EntitySource_Parent
 {
-	public readonly Entity Entity;
-	public readonly int BuffId;
-	public readonly int BuffIndex;
-
-	public EntitySource_Buff(Entity entity, int buffId, int buffIndex)
+	/// <summary>
+	/// The type of the buff (<see cref="BuffID"/> or <see cref="ModContent.BuffType"/>)
+	/// </summary>
+	public int BuffId { get; }
+
+	/// <summary>
+	/// The index of the buff in the entity's <c>buffType</c> and <c>buffTime</c> arrays.
+	/// </summary>
+	public int BuffIndex { get; }
+
+	public EntitySource_Buff(Entity entity, int buffId, int buffIndex, string? context = null) : base(entity, context)
 	{
-		Entity = entity;
 		BuffId = buffId;
 		BuffIndex = buffIndex;
 	}
