--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_ItemOpen.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_ItemOpen.cs
@@ -1,13 +_,28 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when opening bags or gifts. Passed to <see cref="Player.QuickSpawnItem(IEntitySource, Item, int)"/>
+/// </summary>
 public class EntitySource_ItemOpen : IEntitySource
 {
-	public readonly Entity Entity;
-	public readonly int ItemType;
-
-	public EntitySource_ItemOpen(Entity entity, int itemType)
+	/// <summary>
+	/// The player opening the item
+	/// </summary>
+	public Player Player { get; }
+
+	/// <summary>
+	/// The type of item being opened
+	/// </summary>
+	public int ItemType { get; }
+
+	public string? Context { get; }
+
+	public EntitySource_ItemOpen(Player player, int itemType, string? context = null)
 	{
-		Entity = entity;
+		Player = player;
 		ItemType = itemType;
+		Context = context;
 	}
 }
