--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_OnHit_ByItemSourceID.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_OnHit_ByItemSourceID.cs
@@ -1,6 +_,8 @@
 namespace Terraria.DataStructures
 {
+	//TML: Removed, see EntitySource_ByEntitySourceId.cs
+	/*
-	public class EntitySource_OnHit_ByItemSourceID : AEntitySource_OnHit
+	public class EntitySource_OnHit_ByItemSourceID : EntitySource_OnHit
 	{
 		public readonly int SourceId;
 
@@ -9,4 +_,5 @@
 			SourceId = itemSourceId;
 		}
 	}
+	*/
 }
