--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_Gift.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_Gift.cs
@@ -1,11 +_,11 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when NPCs or pets/minions give gifts or rewards to a player.
+/// </summary>
-public class EntitySource_Gift : IEntitySource
+public class EntitySource_Gift : EntitySource_Parent
 {
-	public readonly Entity Entity;
-
-	public EntitySource_Gift(Entity entity)
-	{
-		Entity = entity;
-	}
+	public EntitySource_Gift(Entity entity, string? context = null) : base(entity, context) { }
 }
