--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_Loot.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_Loot.cs
@@ -1,11 +_,12 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when an NPC or Projectile is killed and drops loot. <br/>
+/// Vanilla projectile examples are geodes and life crystal boulders
+/// </summary>
-public class EntitySource_Loot : IEntitySource
+public class EntitySource_Loot : EntitySource_Parent
 {
-	public readonly Entity Entity;
-
-	public EntitySource_Loot(Entity entity)
-	{
-		Entity = entity;
-	}
+	public EntitySource_Loot(Entity entity, string? context = null) : base(entity, context) { }
 }
