--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_ShakeTree.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_ShakeTree.cs
@@ -1,9 +_,11 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when items/bomb projectiles/NPCs fall out of trees being cut down
+/// </summary>
 public class EntitySource_ShakeTree : AEntitySource_Tile
 {
-	public EntitySource_ShakeTree(int tileCoordsX, int tileCoordsY)
-		: base(tileCoordsX, tileCoordsY)
-	{
-	}
+	public EntitySource_ShakeTree(int tileCoordsX, int tileCoordsY, string? context = null) : base(tileCoordsX, tileCoordsY, context) { }
 }
