--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_DropAsItem.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_DropAsItem.cs
@@ -1,11 +_,11 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when projectiles convert themselves to items in the world as a result of hitting a block.
+/// </summary>
-public class EntitySource_DropAsItem : IEntitySource
+public class EntitySource_DropAsItem : EntitySource_Parent
 {
-	public readonly IEntitySourceTarget Entity;
-
-	public EntitySource_DropAsItem(IEntitySourceTarget entity)
-	{
-		Entity = entity;
-	}
+	public EntitySource_DropAsItem(IEntitySourceTarget entity, string? context = null) : base(entity, context) { }
 }
