--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_ItemUse.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_ItemUse.cs
@@ -1,13 +_,21 @@
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// When a player uses an item. <br/>
+/// Can also be used for accessories if, for some reason the accessory has weapon stats <br/>
+/// When used to spawn projectiles, a snapshot of the Player and Item stats will be stored on the projectile. See <see cref="Projectile.HandlePlayerStatModifiers"/> for implementation
+/// </summary>
-public class EntitySource_ItemUse : IEntitySource
+public class EntitySource_ItemUse : EntitySource_Parent
 {
-	public readonly Entity Entity;
-	public readonly Item Item;
+	/// <summary>
+	/// The item being used
+	/// </summary>
+	public Item Item { get; }
 
-	public EntitySource_ItemUse(Entity entity, Item item)
+	public EntitySource_ItemUse(Player player, Item item, string? context = null) : base(player, context)
 	{
-		Entity = entity;
 		Item = item;
 	}
 }
