--- src/TerrariaNetCore/Terraria/DataStructures/EntitySource_Sync.cs
+++ src/tModLoader/Terraria/DataStructures/EntitySource_Sync.cs
@@ -1,5 +_,19 @@
+#nullable enable
+
+using Terraria.ID;
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Used when an <see cref="Item"/> is spawned on the server as a result of <see cref="MessageID.SyncItemCannotBeTakenByEnemies"/> <br/>
+/// Only used by vanilla for lucky coin. Note that the item is spawned client-side with <see cref="IEntitySource.Context"/> = <see cref="ItemSourceID.LuckyCoin"/>
+/// </summary>
 public class EntitySource_Sync : IEntitySource
 {
+	public string? Context { get; }
+
+	public EntitySource_Sync(string? context = null)
+	{
+		Context = context;
+	}
 }
