--- src/TerrariaNetCore/Terraria/DataStructures/GameModeData.cs
+++ src/tModLoader/Terraria/DataStructures/GameModeData.cs
@@ -1,6 +_,7 @@
 namespace Terraria.DataStructures;
 
-public class GameModeData
+//TML: Turned into a record struct, setters made public or made init-only.
+public record struct GameModeData
 {
 	public static readonly GameModeData NormalMode = new GameModeData {
 		Id = 0,
@@ -47,25 +_,25 @@
 		EnemyMoneyDropMultiplier = 1f
 	};
 
-	public int Id { get; private set; }
+	public int Id { get; init; }
 
-	public bool IsExpertMode { get; private set; }
+	public bool IsExpertMode { get; set; }
 
-	public bool IsMasterMode { get; private set; }
+	public bool IsMasterMode { get; set; }
 
-	public bool IsJourneyMode { get; private set; }
+	public bool IsJourneyMode { get; set; }
 
-	public float EnemyMaxLifeMultiplier { get; private set; }
+	public float EnemyMaxLifeMultiplier { get; set; }
 
-	public float EnemyDamageMultiplier { get; private set; }
+	public float EnemyDamageMultiplier { get; set; }
 
-	public float DebuffTimeMultiplier { get; private set; }
+	public float DebuffTimeMultiplier { get; set; }
 
-	public float KnockbackToEnemiesMultiplier { get; private set; }
+	public float KnockbackToEnemiesMultiplier { get; set; }
 
-	public float TownNPCDamageMultiplier { get; private set; }
+	public float TownNPCDamageMultiplier { get; set; }
 
-	public float EnemyDefenseMultiplier { get; private set; }
+	public float EnemyDefenseMultiplier { get; set; }
 
-	public float EnemyMoneyDropMultiplier { get; private set; }
+	public float EnemyMoneyDropMultiplier { get; set; }
 }
