--- src/TerrariaNetCore/Terraria/DataStructures/IEntitySource.cs
+++ src/tModLoader/Terraria/DataStructures/IEntitySource.cs
@@ -1,5 +_,39 @@
+using System;
+using Terraria.ID;
+
+#nullable enable
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// This object encapsulates context information about the source of a particular spawning event of an Item/Projectile/NPC/etc. Aids in facilitating many modding situations and used in various OnSpawn hooks.<br/>
+/// The <see href="https://github.com/tModLoader/tModLoader/wiki/IEntitySource">IEntitySource Guide</see> teaches how and why to use this.
+/// </summary>
 public interface IEntitySource
 {
+	// All members here were added by TML.
+
+	/// <summary>
+	/// Additional context identifier, particularly useful for set bonuses or accessory affects. See <see cref="ItemSourceID"/> and <see cref="ProjectileSourceID"/> for vanilla uses
+	/// </summary>
+	string? Context { get; }
+
+
+	// Used to add contexts to vanilla code without rewriting every line. Be sure to pop responsibly.
+	private static IEntitySource? _goreFallback;
+	internal static IEntitySource? GetGoreFallback() => _goreFallback;
+
+	internal struct FallbackSourceRef : IDisposable
+	{
+		private IEntitySource? prev;
+		public FallbackSourceRef(IEntitySource source)
+		{
+			prev = _goreFallback;
+			_goreFallback = source;
+		}
+
+		public void Dispose() => _goreFallback = prev;
+	}
+
+	internal static FallbackSourceRef PushFallback(IEntitySource source) => new (source);
 }
