--- src/Terraria/Terraria/DataStructures/IProjectileSource.cs
+++ src/tModLoader/Terraria/DataStructures/IProjectileSource.cs
@@ -1,6 +_,12 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from IProjectileSource
+
+	/// <summary>
+	/// A marker interface for types that describe the context of an entity's creation.
+	/// </summary>
-	public interface IProjectileSource
+	public interface IEntitySource
 	{
+
 	}
 }
