--- src/Terraria/Terraria/DataStructures/PlayerDrawSet.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDrawSet.cs
@@ -5,10 +_,11 @@
 using Terraria.GameContent.Golf;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.DataStructures
 {
-	public struct PlayerDrawSet
+	public partial struct PlayerDrawSet
 	{
 		public List<DrawData> DrawDataCache;
 		public List<int> DustCache;
@@ -129,6 +_,8 @@
 		public Vector2 Center => new Vector2(Position.X + (float)(drawPlayer.width / 2), Position.Y + (float)(drawPlayer.height / 2));
 
 		public void BoringSetup(Player player, List<DrawData> drawData, List<int> dust, List<int> gore, Vector2 drawPosition, float shadowOpacity, float rotation, Vector2 rotationOrigin) {
+			headOnlyRender = false;
+
 			DrawDataCache = drawData;
 			DustCache = dust;
 			GoreCache = gore;
@@ -226,6 +_,8 @@
 				type = Main.projectile[drawPlayer.heldProj].type;
 				if (type == 460 || type == 535 || type == 600)
 					heldProjOverHand = true;
+
+				ProjectileLoader.DrawHeldProjInFrontOfHeldItemAndArms(Main.projectile[drawPlayer.heldProj], ref heldProjOverHand);
 			}
 
 			drawPlayer.GetHairSettings(out fullHair, out hatHair, out hideHair, out backHairDraw, out drawsBackHairWithoutHeadgear);
