--- src/TerrariaNetCore/Terraria/DataStructures/PlayerDrawLayers.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDrawLayers.cs
@@ -9,9 +_,10 @@
 using Terraria.Graphics;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.DataStructures;
 
-public static class PlayerDrawLayers
+public static partial class PlayerDrawLayers
 {
@@ -288,7 +_,7 @@
 
 	public static void DrawPlayer_01_3_BackHead(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.head >= 0 && drawinfo.drawPlayer.head < ArmorIDs.Head.Count) {
+		if (drawinfo.drawPlayer.head >= 0) {
 			int num = ArmorIDs.Head.Sets.FrontToBackID[drawinfo.drawPlayer.head];
 			if (num >= 0) {
 				Vector2 pos = drawinfo.helmetOffset;
@@ -358,6 +_,10 @@
 
 	public static void DrawPlayer_07_LeinforsHairShampoo(ref PlayerDrawSet drawinfo)
 	{
+		if (drawinfo.headOnlyRender) {
+			return;
+		}
+
 		if (drawinfo.drawPlayer.leinforsHair && (drawinfo.fullHair || drawinfo.hatHair || drawinfo.drawsBackHairWithoutHeadgear || drawinfo.drawPlayer.head == -1 || drawinfo.drawPlayer.head == 0) && drawinfo.drawPlayer.hair != 12 && drawinfo.shadow == 0f && Main.rgbToHsl(drawinfo.colorHead).Z > 0.2f) {
 			if (Main.rand.Next(20) == 0 && !drawinfo.hatHair) {
 				Rectangle r = Utils.CenteredRectangle(drawinfo.Position + drawinfo.drawPlayer.Size / 2f + new Vector2(0f, drawinfo.drawPlayer.gravDir * -20f), new Vector2(20f, 14f));
@@ -434,7 +_,7 @@
 			drawinfo.DrawDataCache.Add(item);
 		}
 
-		if (drawinfo.drawPlayer.backpack > 0 && drawinfo.drawPlayer.backpack < ArmorIDs.Back.Count && (!drawinfo.drawPlayer.mount.Active || (drawinfo.drawPlayer.mount.Type >= 0 && MountID.Sets.DoesNotOverrideBackpackDraw[drawinfo.drawPlayer.mount.Type]))) {
+		if (drawinfo.drawPlayer.backpack > 0 && (!drawinfo.drawPlayer.mount.Active || (drawinfo.drawPlayer.mount.Type >= 0 && MountID.Sets.DoesNotOverrideBackpackDraw[drawinfo.drawPlayer.mount.Type]))) {
 			Vector2 vector2 = new Vector2(0f, 8f);
 			Vector2 vec4 = drawinfo.Position - Main.screenPosition + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.width / 2, drawinfo.drawPlayer.height - drawinfo.drawPlayer.bodyFrame.Height / 2) + new Vector2(0f, -4f) + vector2;
 			vec4 = vec4.Floor();
@@ -514,7 +_,7 @@
 
 	public static void DrawPlayer_08_1_Tails(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.tail > 0 && drawinfo.drawPlayer.tail < ArmorIDs.Back.Count && !drawinfo.drawPlayer.mount.Active) {
+		if (drawinfo.drawPlayer.tail > 0 && !drawinfo.drawPlayer.mount.Active) {
 			Vector2 zero = Vector2.Zero;
 			if (drawinfo.isSitting)
 				zero.Y += -2f;
@@ -533,10 +_,10 @@
 
 	public static void DrawPlayer_10_BackAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.back <= 0 || drawinfo.drawPlayer.back >= ArmorIDs.Back.Count)
+		if (drawinfo.drawPlayer.back <= 0)
 			return;
 
-		if (drawinfo.drawPlayer.front >= 1 && drawinfo.drawPlayer.front <= 4) {
+		if (drawinfo.drawPlayer.front >= 1) {
 			int num = drawinfo.drawPlayer.bodyFrame.Y / 56;
 			if (num < 1 || num > 5) {
 				drawinfo.armorAdjust = 10;
@@ -1010,7 +_,7 @@
 
 	public static void DrawPlayer_12_1_BalloonFronts(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.balloonFront <= 0 || drawinfo.drawPlayer.balloonFront >= ArmorIDs.Balloon.Count)
+		if (drawinfo.drawPlayer.balloonFront <= 0)
 			return;
 
 		DrawData item;
@@ -1040,7 +_,7 @@
 
 	public static void DrawPlayer_11_Balloons(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.balloon <= 0 || drawinfo.drawPlayer.balloon >= ArmorIDs.Balloon.Count)
+		if (drawinfo.drawPlayer.balloon <= 0)
 			return;
 
 		DrawData item;
@@ -1087,7 +_,7 @@
 			drawinfo.Position.Y -= drawinfo.torsoOffset;
 		}
 
-		if (!drawinfo.hidesBottomSkin && !IsBottomOverridden(ref drawinfo)) {
+		if (!drawinfo.hidesBottomSkin && !drawinfo.isBottomOverriden) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 10].Value, drawinfo.colorLegs, 0, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 				return;
@@ -1098,7 +_,8 @@
 		}
 	}
 
+	//TML: Made internal.
-	public static bool IsBottomOverridden(ref PlayerDrawSet drawinfo)
+	internal static bool IsBottomOverridden(ref PlayerDrawSet drawinfo)
 	{
 		if (ShouldOverrideLegs_CheckPants(ref drawinfo))
 			return true;
@@ -1109,8 +_,12 @@
 		return false;
 	}
 
+	//TML: Made private
-	public static bool ShouldOverrideLegs_CheckPants(ref PlayerDrawSet drawinfo)
+	private static bool ShouldOverrideLegs_CheckPants(ref PlayerDrawSet drawinfo)
 	{
+		return drawinfo.drawPlayer.legs > 0 && ArmorIDs.Legs.Sets.OverridesLegs[drawinfo.drawPlayer.legs];
+
+		/*
 		switch (drawinfo.drawPlayer.legs) {
 			case 55:
 			case 63:
@@ -1127,15 +_,22 @@
 			default:
 				return false;
 		}
+		*/
 	}
 
+	//TML: Made private
-	public static bool ShouldOverrideLegs_CheckShoes(ref PlayerDrawSet drawinfo)
+	private static bool ShouldOverrideLegs_CheckShoes(ref PlayerDrawSet drawinfo)
 	{
-		sbyte shoe = drawinfo.drawPlayer.shoe;
+		int shoe = drawinfo.drawPlayer.shoe;
+
+		return shoe > 0 && ArmorIDs.Shoe.Sets.OverridesLegs[shoe];
+
+		/*
 		if (shoe == 15)
 			return true;
 
 		return false;
+		*/
 	}
 
 	public static void DrawPlayer_12_Skin_Composite(ref PlayerDrawSet drawinfo)
@@ -1168,7 +_,7 @@
 			});
 		}
 
-		if (!drawinfo.hidesBottomSkin && !drawinfo.drawPlayer.invis && !IsBottomOverridden(ref drawinfo)) {
+		if (!drawinfo.hidesBottomSkin && !drawinfo.drawPlayer.invis && !drawinfo.isBottomOverriden) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 10].Value, drawinfo.colorLegs, drawinfo.skinDyePacked, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 			}
@@ -1201,8 +_,8 @@
 		float rotation = bodyRotation + drawinfo.compositeBackArmRotation;
 		bool flag = !drawinfo.drawPlayer.invis;
 		bool flag2 = !drawinfo.drawPlayer.invis;
-		bool flag3 = drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count;
+		bool flag3 = drawinfo.drawPlayer.body > 0;
-		bool flag4 = drawinfo.drawPlayer.coat > 0 && drawinfo.drawPlayer.coat < ArmorIDs.Body.Count;
+ 		bool flag4 = drawinfo.drawPlayer.coat > 0;
 		bool flag5 = !drawinfo.hidesTopSkin;
 		bool flag6 = false;
 		if (flag3) {
@@ -1291,7 +_,7 @@
 			}, drawinfo.drawPlayer.coat);
 		}
 
-		if (drawinfo.drawPlayer.handoff > 0 && drawinfo.drawPlayer.handoff < ArmorIDs.HandOff.Count) {
+		if (drawinfo.drawPlayer.handoff > 0) {
 			Texture2D value5 = TextureAssets.AccHandsOffComposite[drawinfo.drawPlayer.handoff].Value;
 			DrawCompositeArmorPiece(ref drawinfo, CompositePlayerDrawContext.BackArmAccessory, new DrawData(value5, vector3, drawinfo.compBackArmFrame, drawinfo.colorArmorBody, rotation, bodyVect, 1f, drawinfo.playerEffect) {
 				shader = drawinfo.cHandOff
@@ -1331,5 +_,5 @@
 			return;
 
 		if (drawinfo.isSitting && drawinfo.drawPlayer.legs != 140) {
-			if (drawinfo.drawPlayer.legs > 0 && drawinfo.drawPlayer.legs < ArmorIDs.Legs.Count && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
+			if (drawinfo.drawPlayer.legs > 0 && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
 				if (drawinfo.drawPlayer.invis)
@@ -1392,7 +_,7 @@
 				drawinfo.DrawDataCache.Add(item);
 			}
 		}
-		else if (drawinfo.drawPlayer.legs > 0 && drawinfo.drawPlayer.legs < ArmorIDs.Legs.Count && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
+		else if (drawinfo.drawPlayer.legs > 0 && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
 			if (drawinfo.drawPlayer.invis)
 				return;
 
@@ -1574,5 +_,5 @@
 	public static void DrawPlayer_14_Shoes(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.shoe > 0 && drawinfo.drawPlayer.shoe < ArmorIDs.Shoe.Count && !ShouldOverrideLegs_CheckPants(ref drawinfo)) {
+		if (drawinfo.drawPlayer.shoe > 0 && !ShouldOverrideLegs_CheckPants(ref drawinfo)) {
 			Vector2 shoeDrawOffset = drawinfo.drawPlayer.GetShoeDrawOffset();
 			int num = drawinfo.cShoe;
@@ -1592,7 +_,7 @@
 
 	public static void DrawPlayer_15_SkinLongCoat(ref PlayerDrawSet drawinfo)
 	{
-		if ((drawinfo.skinVar == 3 || drawinfo.skinVar == 8 || drawinfo.skinVar == 7) && (drawinfo.drawPlayer.body <= 0 || drawinfo.drawPlayer.body >= ArmorIDs.Body.Count) && !drawinfo.drawPlayer.invis) {
+		if ((drawinfo.skinVar == 3 || drawinfo.skinVar == 8 || drawinfo.skinVar == 7) && (drawinfo.drawPlayer.body <= 0) && !drawinfo.drawPlayer.invis) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 14].Value, drawinfo.colorShirt, 0, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 				return;
@@ -1738,7 +_,7 @@
 		if (drawinfo.usesCompositeTorso) {
 			DrawPlayer_17_TorsoComposite(ref drawinfo);
 		}
-		else if (drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count) {
+		else if (drawinfo.drawPlayer.body > 0) {
 			Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
 			int num = drawinfo.armorAdjust;
 			bodyFrame.X += num;
@@ -1799,9 +_,9 @@
 		Vector2 compositeOffset_BackArm = GetCompositeOffset_BackArm(ref drawinfo);
 		_ = vector3 + compositeOffset_BackArm;
 		bodyVect += compositeOffset_BackArm;
 		bool flag = false;
-		if (drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count) {
+		if (drawinfo.drawPlayer.body > 0) {
 			flag = true;
 			if (!drawinfo.drawPlayer.invis || IsArmorDrawnWhenInvisible(drawinfo.drawPlayer.body)) {
 				Texture2D value = TextureAssets.ArmorBodyComposite[drawinfo.drawPlayer.body].Value;
 				DrawCompositeArmorPiece(ref drawinfo, CompositePlayerDrawContext.Torso, new DrawData(value, vector, drawinfo.compTorsoFrame, drawinfo.colorArmorBody, bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect) {
@@ -1840,7 +_,7 @@
 
 	public static void DrawPlayer_18_OffhandAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (!drawinfo.usesCompositeBackHandAcc && drawinfo.drawPlayer.handoff > 0 && drawinfo.drawPlayer.handoff < ArmorIDs.HandOff.Count) {
+		if (!drawinfo.usesCompositeBackHandAcc && drawinfo.drawPlayer.handoff > 0) {
 			DrawData item = new DrawData(TextureAssets.AccHandsOff[drawinfo.drawPlayer.handoff].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cHandOff;
 			drawinfo.DrawDataCache.Add(item);
@@ -1860,7 +_,7 @@
 
 	public static void DrawPlayer_19_WaistAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.waist > 0 && drawinfo.drawPlayer.waist < ArmorIDs.Waist.Count) {
+		if (drawinfo.drawPlayer.waist > 0) {
 			Rectangle value = drawinfo.drawPlayer.legFrame;
 			if (ArmorIDs.Waist.Sets.UsesTorsoFraming[drawinfo.drawPlayer.waist])
 				value = drawinfo.drawPlayer.bodyFrame;
@@ -1873,7 +_,7 @@
 
 	public static void DrawPlayer_20_NeckAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.neck > 0 && drawinfo.drawPlayer.neck < ArmorIDs.Neck.Count) {
+		if (drawinfo.drawPlayer.neck > 0) {
 			DrawData item = new DrawData(TextureAssets.AccNeck[drawinfo.drawPlayer.neck].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cNeck;
 			drawinfo.DrawDataCache.Add(item);
@@ -1884,8 +_,8 @@
 	{
 		Vector2 pos = Vector2.Zero;
 		drawinfo.drawPlayer.ApplyHeadOffsetFromMount(ref pos);
 		Vector2 helmetOffset = drawinfo.helmetOffset + pos;
 		DrawPlayer_21_Head_TheFace(ref drawinfo);
-		bool flag = drawinfo.drawPlayer.head == 14 || drawinfo.drawPlayer.head == 56 || drawinfo.drawPlayer.head == 114 || drawinfo.drawPlayer.head == 158 || drawinfo.drawPlayer.head == 69 || drawinfo.drawPlayer.head == 180;
+		bool flag = drawinfo.drawPlayer.head >= 0 && ArmorIDs.Head.Sets.IsTallHat[drawinfo.drawPlayer.head];
 		bool flag2 = drawinfo.drawPlayer.head == 28;
 		bool flag3 = drawinfo.drawPlayer.head == 39 || drawinfo.drawPlayer.head == 38;
@@ -2018,7 +_,7 @@
 				drawinfo.DrawDataCache.Add(item);
 			}
 		}
-		else if (flag4 && drawinfo.drawPlayer.head > 0 && drawinfo.drawPlayer.head < ArmorIDs.Head.Count && !flag2) {
+		else if (flag4 && drawinfo.drawPlayer.head > 0 && !flag2) {
 			if (!(drawinfo.drawPlayer.invis && flag3)) {
 				if (drawinfo.drawPlayer.head == 13) {
 					int hatStacks2 = GetHatStacks(ref drawinfo, 205);
@@ -2218,5 +_,10 @@
 				drawinfo.DrawDataCache.Add(item);
 			}
 		}
+	}
+
+	public static void DrawPlayer_21_2_FinchNest(ref PlayerDrawSet drawinfo)
+	{
+		DrawData item;
 
 		if (!drawinfo.drawPlayer.babyBird)
@@ -2313,1 +_,1 @@
-		else if (!flag && drawinfo.drawPlayer.faceHead > 0 && drawinfo.drawPlayer.faceHead < ArmorIDs.Face.Count) {
+		else if (!flag && drawinfo.drawPlayer.faceHead > 0) {
@@ -2320,7 +_,7 @@
 
 			float num = 0f;
 			if (drawinfo.drawPlayer.face == 5) {
-				sbyte faceHead = drawinfo.drawPlayer.faceHead;
+				int faceHead = drawinfo.drawPlayer.faceHead;
 				if ((uint)(faceHead - 10) <= 3u)
 					num = 2 * drawinfo.drawPlayer.direction;
 			}
@@ -2512,4 +_,4 @@
 		Vector2 vector = DrawPlayer_GetMountOffsetForFaceAcc(ref drawinfo, ref isVelociraptor, ref hasRockGolemHead);
-		if (drawinfo.drawPlayer.face > 0 && drawinfo.drawPlayer.face < ArmorIDs.Face.Count && !ArmorIDs.Face.Sets.DrawInFaceUnderHairLayer[drawinfo.drawPlayer.face] && (!isVelociraptor || ArmorIDs.Face.Sets.CanDrawOnVelociraptorMount[drawinfo.drawPlayer.face]) && (drawinfo.drawPlayer.face != 23 || !flag)) {
+		if (drawinfo.drawPlayer.face > 0 && !ArmorIDs.Face.Sets.DrawInFaceUnderHairLayer[drawinfo.drawPlayer.face] && (!isVelociraptor || ArmorIDs.Face.Sets.CanDrawOnVelociraptorMount[drawinfo.drawPlayer.face]) && (drawinfo.drawPlayer.face != 23 || !flag)) {
 			Vector2 faceDrawOffset = drawinfo.drawPlayer.GetFaceDrawOffset(drawinfo.drawPlayer.face);
 			if (isVelociraptor && hasRockGolemHead)
@@ -2533,3 +_,3 @@
-		if (drawinfo.drawPlayer.faceFlower > 0 && drawinfo.drawPlayer.faceFlower < ArmorIDs.Face.Count && (!isVelociraptor || ArmorIDs.Face.Sets.CanDrawOnVelociraptorMount[drawinfo.drawPlayer.faceFlower])) {
+		if (drawinfo.drawPlayer.faceFlower > 0 && (!isVelociraptor || ArmorIDs.Face.Sets.CanDrawOnVelociraptorMount[drawinfo.drawPlayer.faceFlower])) {
 			Vector2 faceDrawOffset3 = drawinfo.drawPlayer.GetFaceDrawOffset(drawinfo.drawPlayer.faceFlower);
 			if (isVelociraptor && hasRockGolemHead)
@@ -2719,7 +_,7 @@
 
 	public static void DrawPlayer_25_Shield(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.shield <= 0 || drawinfo.drawPlayer.shield >= ArmorIDs.Shield.Count)
+		if (drawinfo.drawPlayer.shield <= 0 || drawinfo.drawPlayer.shield >= TextureAssets.AccShield.Length)
 			return;
 
 		Vector2 zero = Vector2.Zero;
@@ -2835,12 +_,12 @@
 		if (!drawinfo.drawPlayer.CanVisuallyHoldItem(heldItem))
 			flag2 = false;
 
 		drawinfo.itemColor = Lighting.GetColor((int)((double)drawinfo.Position.X + (double)drawinfo.drawPlayer.width * 0.5) / 16, (int)(((double)drawinfo.Position.Y + (double)drawinfo.drawPlayer.height * 0.5) / 16.0));
 		if (num == 678)
 			drawinfo.itemColor = Color.White;
 
 		DrawPlayer_27_HeldItem_ApplyStealthToColor(ref drawinfo, heldItem, flag, flag2, ref drawinfo.itemColor);
-		if (drawinfo.shadow != 0f || drawinfo.drawPlayer.frozen || !(flag || flag2) || num <= 0 || drawinfo.drawPlayer.dead || heldItem.noUseGraphic || (drawinfo.drawPlayer.wet && heldItem.noWet) || (drawinfo.drawPlayer.happyFunTorchTime && drawinfo.drawPlayer.inventory[drawinfo.drawPlayer.selectedItem].createTile == 4 && drawinfo.drawPlayer.itemAnimation == 0))
+		if (drawinfo.shadow != 0f || drawinfo.drawPlayer.frozen || !(flag || flag2) || num <= 0 || drawinfo.drawPlayer.dead || heldItem.noUseGraphic || (drawinfo.drawPlayer.wet && heldItem.noWet && !ItemID.Sets.WaterTorches[num]/*Allow biome torches underwater.*/) || (drawinfo.drawPlayer.happyFunTorchTime && drawinfo.drawPlayer.inventory[drawinfo.drawPlayer.selectedItem].createTile == 4 && drawinfo.drawPlayer.itemAnimation == 0))
 			return;
 
 		_ = drawinfo.drawPlayer.name;
@@ -3036,6 +_,8 @@
 					num7 -= (float)rectangle.Width;
 				}
 
+				ItemLoader.HoldoutOrigin(drawinfo.drawPlayer, ref origin5); // DrawPlayer_27_HeldItem
+
 				item = new DrawData(value, new Vector2((int)(drawinfo.ItemLocation.X - Main.screenPosition.X + origin5.X + num7), (int)(drawinfo.ItemLocation.Y - Main.screenPosition.Y + num8)), rectangle, heldItem.GetAlpha(drawinfo.itemColor), num6, origin5, adjustedItemScale, drawinfo.itemEffect);
 				drawinfo.DrawDataCache.Add(item);
 				if (num == 3870) {
@@ -3215,7 +_,7 @@
 		else if (drawinfo.usesCompositeTorso) {
 			DrawPlayer_28_ArmOverItemComposite(ref drawinfo);
 		}
-		else if (drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count) {
+		else if (drawinfo.drawPlayer.body > 0) {
 			Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
 			int num = drawinfo.armorAdjust;
 			bodyFrame.X += num;
@@ -3294,8 +_,8 @@
 			vector += new Vector2(((drawinfo.playerEffect & SpriteEffects.FlipHorizontally) == 0) ? 1 : (-1), ((drawinfo.playerEffect & SpriteEffects.FlipVertically) == 0) ? 1 : (-1));
 
 		_ = drawinfo.drawPlayer.invis;
-		bool num = drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count;
+		bool num = drawinfo.drawPlayer.body > 0;
-		bool flag = drawinfo.drawPlayer.coat > 0 && drawinfo.drawPlayer.coat < ArmorIDs.Body.Count;
+ 		bool flag = drawinfo.drawPlayer.coat > 0;
 		int num2 = (drawinfo.compShoulderOverFrontArm ? 1 : 0);
 		int num3 = ((!drawinfo.compShoulderOverFrontArm) ? 1 : 0);
 		int num4 = ((!drawinfo.compShoulderOverFrontArm) ? 1 : 0);
@@ -3410,7 +_,7 @@
 			}
 		}
 
-		if (drawinfo.drawPlayer.handon > 0 && drawinfo.drawPlayer.handon < ArmorIDs.HandOn.Count) {
+		if (drawinfo.drawPlayer.handon > 0) {
 			Texture2D value6 = TextureAssets.AccHandsOnComposite[drawinfo.drawPlayer.handon].Value;
 			DrawCompositeArmorPiece(ref drawinfo, CompositePlayerDrawContext.FrontArmAccessory, new DrawData(value6, vector, drawinfo.compFrontArmFrame, drawinfo.colorArmorBody, rotation, bodyVect, 1f, drawinfo.playerEffect) {
 				shader = drawinfo.cHandOn
@@ -3420,7 +_,7 @@
 
 	public static void DrawPlayer_29_OnhandAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (!drawinfo.usesCompositeFrontHandAcc && drawinfo.drawPlayer.handon > 0 && drawinfo.drawPlayer.handon < ArmorIDs.HandOn.Count) {
+		if (!drawinfo.usesCompositeFrontHandAcc && drawinfo.drawPlayer.handon > 0) {
 			DrawData item = new DrawData(TextureAssets.AccHandsOn[drawinfo.drawPlayer.handon].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cHandOn;
 			drawinfo.DrawDataCache.Add(item);
@@ -3455,7 +_,7 @@
 
 	public static void DrawPlayer_32_FrontAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.front > 0 && drawinfo.drawPlayer.front < ArmorIDs.Front.Count && !drawinfo.drawPlayer.mount.Active) {
+		if (drawinfo.drawPlayer.front > 0 && !drawinfo.drawPlayer.mount.Active) {
 			Vector2 zero = Vector2.Zero;
 			DrawData item = new DrawData(TextureAssets.AccFront[drawinfo.drawPlayer.front].Value, zero + new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cFront;
@@ -3466,6 +_,6 @@
 	public static void DrawPlayer_32_FrontAcc_FrontPart(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.front <= 0 || drawinfo.drawPlayer.front >= ArmorIDs.Front.Count || (ArmorIDs.Front.Sets.DontDrawIfWearingAScarfOrCape[drawinfo.drawPlayer.front] && ((drawinfo.drawPlayer.neck > 0 && ArmorIDs.Neck.Sets.IsAScarf[drawinfo.drawPlayer.neck]) || (drawinfo.drawPlayer.back > 0 && ArmorIDs.Back.Sets.IsACape[drawinfo.drawPlayer.back]))))
+		if (drawinfo.drawPlayer.front <= 0 || (ArmorIDs.Front.Sets.DontDrawIfWearingAScarfOrCape[drawinfo.drawPlayer.front] && ((drawinfo.drawPlayer.neck > 0 && ArmorIDs.Neck.Sets.IsAScarf[drawinfo.drawPlayer.neck]) || (drawinfo.drawPlayer.back > 0 && ArmorIDs.Back.Sets.IsACape[drawinfo.drawPlayer.back]))))
 			return;
 
 		Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
@@ -3505,6 +_,6 @@
 	public static void DrawPlayer_32_FrontAcc_BackPart(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.front <= 0 || drawinfo.drawPlayer.front >= ArmorIDs.Front.Count || (ArmorIDs.Front.Sets.DontDrawIfWearingAScarfOrCape[drawinfo.drawPlayer.front] && ((drawinfo.drawPlayer.neck > 0 && ArmorIDs.Neck.Sets.IsAScarf[drawinfo.drawPlayer.neck]) || (drawinfo.drawPlayer.back > 0 && ArmorIDs.Back.Sets.IsACape[drawinfo.drawPlayer.back]))))
+		if (drawinfo.drawPlayer.front <= 0 || (ArmorIDs.Front.Sets.DontDrawIfWearingAScarfOrCape[drawinfo.drawPlayer.front] && ((drawinfo.drawPlayer.neck > 0 && ArmorIDs.Neck.Sets.IsAScarf[drawinfo.drawPlayer.neck]) || (drawinfo.drawPlayer.back > 0 && ArmorIDs.Back.Sets.IsACape[drawinfo.drawPlayer.back]))))
 			return;
 
 		Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
