--- src/TerrariaNetCore/Terraria/DataStructures/PlayerDeathReason.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDeathReason.cs
@@ -14,6 +_,14 @@
 		private int _sourceItemPrefix;
 		private string _sourceCustomReason;
 
+		public ref int SourcePlayerIndex => ref _sourcePlayerIndex;
+		public ref int SourceNPCIndex => ref _sourceNPCIndex;
+		public ref int SourceProjectileIndex => ref _sourceProjectileIndex;
+		public ref int SourceOtherIndex => ref _sourceOtherIndex;
+		public ref int SourceItemType => ref _sourceItemType;
+		public ref int SourceItemPrefix => ref _sourceItemPrefix;
+		public ref string SourceCustomReason => ref _sourceCustomReason;
+
 		public int? SourceProjectileType {
 			get {
 				if (_sourceProjectileIndex == -1)
@@ -21,6 +_,7 @@
 
 				return _sourceProjectileType;
 			}
+			set => _sourceProjectileType = value??-1;
 		}
 
 		public static PlayerDeathReason LegacyEmpty() {
