--- src/Terraria/Terraria/DataStructures/PlayerDrawSet.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDrawSet.cs
@@ -5,10 +_,11 @@
 using Terraria.GameContent.Golf;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.DataStructures
 {
-	public struct PlayerDrawSet
+	public partial struct PlayerDrawSet
 	{
 		public List<DrawData> DrawDataCache;
 		public List<int> DustCache;
@@ -129,6 +_,8 @@
 		public Vector2 Center => new Vector2(Position.X + (float)(drawPlayer.width / 2), Position.Y + (float)(drawPlayer.height / 2));
 
 		public void BoringSetup(Player player, List<DrawData> drawData, List<int> dust, List<int> gore, Vector2 drawPosition, float shadowOpacity, float rotation, Vector2 rotationOrigin) {
+			headOnlyRender = false;
+
 			DrawDataCache = drawData;
 			DustCache = dust;
 			GoreCache = gore;
@@ -226,6 +_,8 @@
 				type = Main.projectile[drawPlayer.heldProj].type;
 				if (type == 460 || type == 535 || type == 600)
 					heldProjOverHand = true;
+
+				ProjectileLoader.DrawHeldProjInFrontOfHeldItemAndArms(Main.projectile[drawPlayer.heldProj], ref heldProjOverHand);
 			}
 
 			drawPlayer.GetHairSettings(out fullHair, out hatHair, out hideHair, out backHairDraw, out drawsBackHairWithoutHeadgear);
@@ -1049,8 +_,10 @@
 				DustCache.Add(dust21.dustIndex);
 			}
 
+			bool fullBright = false;
+			PlayerLoader.DrawEffects(this, ref num12, ref num13, ref num14, ref num15, ref fullBright);
 			if (num12 != 1f || num13 != 1f || num14 != 1f || num15 != 1f) {
-				if (drawPlayer.onFire || drawPlayer.onFire2 || drawPlayer.onFrostBurn || drawPlayer.onFire3 || drawPlayer.onFrostBurn2) {
+				if (drawPlayer.onFire || drawPlayer.onFire2 || drawPlayer.onFrostBurn || drawPlayer.onFire3 || drawPlayer.onFrostBurn2 || fullBright) {
 					colorEyeWhites = drawPlayer.GetImmuneAlpha(Color.White, shadow);
 					colorEyes = drawPlayer.GetImmuneAlpha(drawPlayer.eyeColor, shadow);
 					colorHair = drawPlayer.GetImmuneAlpha(drawPlayer.GetHairColor(useLighting: false), shadow);
@@ -1327,9 +_,9 @@
 		private void CreateCompositeData() {
 			frontShoulderOffset = Vector2.Zero;
 			backShoulderOffset = Vector2.Zero;
-			usesCompositeTorso = (drawPlayer.body > 0 && drawPlayer.body < 244 && ArmorIDs.Body.Sets.UsesNewFramingCode[drawPlayer.body]);
-			usesCompositeFrontHandAcc = (drawPlayer.handon > 0 && drawPlayer.handon < 23 && ArmorIDs.HandOn.Sets.UsesNewFramingCode[drawPlayer.handon]);
-			usesCompositeBackHandAcc = (drawPlayer.handoff > 0 && drawPlayer.handoff < 15 && ArmorIDs.HandOff.Sets.UsesNewFramingCode[drawPlayer.handoff]);
+			usesCompositeTorso = (drawPlayer.body > 0 && ArmorIDs.Body.Sets.UsesNewFramingCode[drawPlayer.body]);
+			usesCompositeFrontHandAcc = (drawPlayer.handon > 0 && ArmorIDs.HandOn.Sets.UsesNewFramingCode[drawPlayer.handon]);
+			usesCompositeBackHandAcc = (drawPlayer.handoff > 0 && ArmorIDs.HandOff.Sets.UsesNewFramingCode[drawPlayer.handoff]);
 			if (drawPlayer.body < 1)
 				usesCompositeTorso = true;
 
