--- src/TerrariaNetCore/Terraria/DataStructures/PlayerDeathReason.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDeathReason.cs
@@ -1,5 +_,7 @@
+using System;
 using System.IO;
 using Terraria.Localization;
+using Terraria.ModLoader.IO;
 
 namespace Terraria.DataStructures;
 
@@ -7,13 +_,27 @@
 {
 	private int _sourcePlayerIndex = -1;
 	private int _sourceNPCIndex = -1;
+	/*
 	private int _sourceProjectileLocalIndex = -1;
+	*/
 	private int _sourceOtherIndex = -1;
 	private int _sourceProjectileType;
+
+	private byte _sourceNPCGeneration; // Added
+	private Item _sourceItem; // Added
+	private ProjectileKey _sourceProjectileKey; // Added
+
+	// Changed to reference _sourceItem
+	/*
 	private int _sourceItemType;
 	private int _sourceItemPrefix;
+	*/
+	private int _sourceItemType => _sourceItem?.type ?? 0;
+	private int _sourceItemPrefix => _sourceItem?.prefix ?? 0;
 	private string _sourceCustomReason;
+	private NetworkText _customReason;
 
+	/*
 	public int? SourceOtherIndex {
 		get {
 			if (_sourceOtherIndex >= 0)
@@ -33,17 +_,45 @@
 	}
 
 	public string CustomReason => _sourceCustomReason;
+	*/
 
+	// Ref fields added by TML.
+	public ref int SourcePlayerIndex => ref _sourcePlayerIndex;
+	public ref byte SourceNPCGeneration => ref _sourceNPCGeneration;
+	public ref int SourceNPCIndex => ref _sourceNPCIndex;
+	public ref ProjectileKey SourceProjectileKey => ref _sourceProjectileKey;
+	public ref int SourceOtherIndex => ref _sourceOtherIndex;
+	public ref int SourceProjectileType => ref _sourceProjectileType;
+	public ref Item SourceItem => ref _sourceItem;
+	[Obsolete("CustomReason should be used instead")]
+	public ref string SourceCustomReason => ref _sourceCustomReason;
+	public ref NetworkText CustomReason => ref _customReason;
+
+	/// <summary>
+	/// If this DamageSource has been sent over the network. <see cref="TryGetCausingEntity"/> may fail if so.
+	/// </summary>
+	public bool NetSynced { get; private set; }
+
+	/// <summary>
+	/// Not recommended for use when <see cref="NetSynced"/> is true as NPCs and Projectiles may have despawned/died by the time a strike/death packet arrives
+	/// </summary>
 	public bool TryGetCausingEntity(out Entity entity)
 	{
 		entity = null;
+		/*
 		if (Main.npc.IndexInRange(_sourceNPCIndex)) {
+		*/
+		if (Main.npc.IndexInRange(_sourceNPCIndex) && Main.npc[_sourceNPCIndex].generation == _sourceNPCGeneration) {
 			entity = Main.npc[_sourceNPCIndex];
 			return true;
 		}
 
+		/*
 		if (Main.projectile.IndexInRange(_sourceProjectileLocalIndex)) {
 			entity = Main.projectile[_sourceProjectileLocalIndex];
+		*/
+		if (_sourceProjectileKey.TryGetActive(out Projectile projectile)) {
+			entity = projectile;
 			return true;
 		}
 
@@ -65,10 +_,16 @@
 	public static PlayerDeathReason ByNPC(int index)
 	{
 		return new PlayerDeathReason {
-			_sourceNPCIndex = index
+			_sourceNPCIndex = index,
+			_sourceNPCGeneration = Main.npc[index].generation
 		};
 	}
 
+	/// <summary>
+	/// A custom death message.
+	/// <para/> This overload does not support localization, all players will see the same text regardless of their language settings. Use <see cref="ByCustomReason(NetworkText)"/> instead for localization support.
+	/// </summary>
+	[Obsolete("Use the NetworkText version instead")]
 	public static PlayerDeathReason ByCustomReason(string reasonInEnglish)
 	{
 		return new PlayerDeathReason {
@@ -76,18 +_,37 @@
 		};
 	}
 
+	/// <summary>
+	/// A localized custom death message. Each client will see the death message in their own language, unlike with <see cref="ByCustomReason(string)"/>.
+	/// <para/> Text substitutions can be provided by using <see cref="LocalizedText.ToNetworkText(object[])"/> or <see cref="NetworkText.FromKey(string, object[])"/>.
+	/// </summary>
+	public static PlayerDeathReason ByCustomReason(NetworkText reason)
+	{
+		return new PlayerDeathReason {
+			_customReason = reason
+		};
+	}
+
+	/*
 	public static PlayerDeathReason ByPlayer(int index)
+	*/
+	public static PlayerDeathReason ByPlayerItem(int index, Item item)
 	{
 		return new PlayerDeathReason {
 			_sourcePlayerIndex = index,
+			/*
 			_sourceItemType = Main.player[index].inventory[Main.player[index].selectedItem].type,
 			_sourceItemPrefix = Main.player[index].inventory[Main.player[index].selectedItem].prefix
+			*/
+			_sourceItem = item
 		};
 	}
 
-	public static PlayerDeathReason ByOther(int type)
+	// playerIndex added
+	public static PlayerDeathReason ByOther(int type, int playerIndex = -1)
 	{
 		return new PlayerDeathReason {
+			_sourcePlayerIndex = playerIndex,
 			_sourceOtherIndex = type
 		};
 	}
@@ -96,23 +_,32 @@
 	{
 		PlayerDeathReason playerDeathReason = new PlayerDeathReason {
 			_sourcePlayerIndex = playerIndex,
+			/*
 			_sourceProjectileLocalIndex = projectileIndex,
+			*/
-			_sourceProjectileType = Main.projectile[projectileIndex].type
+			_sourceProjectileType = Main.projectile[projectileIndex].type,
+			_sourceProjectileKey = Main.projectile[projectileIndex].key
 		};
 
+		/*
 		if (playerIndex >= 0 && playerIndex <= 255) {
 			playerDeathReason._sourceItemType = Main.player[playerIndex].inventory[Main.player[playerIndex].selectedItem].type;
 			playerDeathReason._sourceItemPrefix = Main.player[playerIndex].inventory[Main.player[playerIndex].selectedItem].prefix;
 		}
+		*/
 
 		return playerDeathReason;
 	}
 
 	public NetworkText GetDeathText(string deadPlayerName)
 	{
+		if (_customReason != null)
+			return _customReason;
+
 		if (_sourceCustomReason != null)
 			return NetworkText.FromLiteral(_sourceCustomReason);
 
+		var _sourceProjectileLocalIndex = -1; // TML
 		return Lang.CreateDeathMessage(deadPlayerName, _sourcePlayerIndex, _sourceNPCIndex, _sourceProjectileLocalIndex, _sourceOtherIndex, _sourceProjectileType, _sourceItemType);
 	}
 
@@ -121,21 +_,33 @@
 		BitsByte bitsByte = (byte)0;
 		bitsByte[0] = _sourcePlayerIndex != -1;
 		bitsByte[1] = _sourceNPCIndex != -1;
+		/*
 		bitsByte[2] = _sourceProjectileLocalIndex != -1;
+		*/
+		bitsByte[2] = _sourceProjectileKey != default;
 		bitsByte[3] = _sourceOtherIndex != -1;
 		bitsByte[4] = _sourceProjectileType != 0;
 		bitsByte[5] = _sourceItemType != 0;
+		/*
 		bitsByte[6] = _sourceItemPrefix != 0;
+		*/
+		bitsByte[6] = _customReason != null;
 		bitsByte[7] = _sourceCustomReason != null;
 		writer.Write(bitsByte);
 		if (bitsByte[0])
 			writer.Write((short)_sourcePlayerIndex);
 
-		if (bitsByte[1])
+		if (bitsByte[1]) {
 			writer.Write((short)_sourceNPCIndex);
+			writer.Write(_sourceNPCGeneration);
+		}
 
-		if (bitsByte[2])
+		if (bitsByte[2]) {
+			/*
 			writer.Write((short)_sourceProjectileLocalIndex);
+			*/
+			writer.Write((int)_sourceProjectileKey);
+		}
 
 		if (bitsByte[3])
 			writer.Write((byte)_sourceOtherIndex);
@@ -144,10 +_,16 @@
 			writer.Write((short)_sourceProjectileType);
 
 		if (bitsByte[5])
+			ItemIO.Send(_sourceItem, writer);
+		/*
+		if (bitsByte[5])
 			writer.Write((short)_sourceItemType);
 
 		if (bitsByte[6])
 			writer.Write((byte)_sourceItemPrefix);
+		*/
+		if (bitsByte[6])
+			_customReason.Serialize(writer);
 
 		if (bitsByte[7])
 			writer.Write(_sourceCustomReason);
@@ -160,11 +_,17 @@
 		if (bitsByte[0])
 			playerDeathReason._sourcePlayerIndex = reader.ReadInt16();
 
-		if (bitsByte[1])
+		if (bitsByte[1]) {
 			playerDeathReason._sourceNPCIndex = reader.ReadInt16();
+			playerDeathReason._sourceNPCGeneration = reader.ReadByte();
+		}
 
-		if (bitsByte[2])
+		if (bitsByte[2]) {
+			/*
 			playerDeathReason._sourceProjectileLocalIndex = reader.ReadInt16();
+			*/
+			playerDeathReason._sourceProjectileKey = (ProjectileKey)reader.ReadInt32();
+		}
 
 		if (bitsByte[3])
 			playerDeathReason._sourceOtherIndex = reader.ReadByte();
@@ -172,15 +_,23 @@
 		if (bitsByte[4])
 			playerDeathReason._sourceProjectileType = reader.ReadInt16();
 
+		/*
 		if (bitsByte[5])
 			playerDeathReason._sourceItemType = reader.ReadInt16();
 
 		if (bitsByte[6])
 			playerDeathReason._sourceItemPrefix = reader.ReadByte();
+		*/
+		if (bitsByte[5])
+			playerDeathReason._sourceItem = ItemIO.Receive(reader);
+
+		if (bitsByte[6])
+			playerDeathReason._customReason = NetworkText.Deserialize(reader);
 
 		if (bitsByte[7])
 			playerDeathReason._sourceCustomReason = reader.ReadString();
 
+		playerDeathReason.NetSynced = true;
 		return playerDeathReason;
 	}
 }
