--- src/TerrariaNetCore/Terraria/DataStructures/PlayerDrawLayers.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDrawLayers.cs
@@ -9,11 +_,12 @@
 using Terraria.Graphics;
 using Terraria.Graphics.Shaders;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.DataStructures;
 
-public static class PlayerDrawLayers
+public static partial class PlayerDrawLayers
 {
 	public static void DrawPlayer_extra_TorsoPlus(ref PlayerDrawSet drawinfo)
 	{
@@ -288,7 +_,7 @@
 
 	public static void DrawPlayer_01_3_BackHead(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.head >= 0 && drawinfo.drawPlayer.head < ArmorIDs.Head.Count) {
+		if (drawinfo.drawPlayer.head >= 0) {
 			int num = ArmorIDs.Head.Sets.FrontToBackID[drawinfo.drawPlayer.head];
 			if (num >= 0) {
 				Vector2 pos = drawinfo.helmetOffset;
@@ -358,6 +_,10 @@
 
 	public static void DrawPlayer_07_LeinforsHairShampoo(ref PlayerDrawSet drawinfo)
 	{
+		if (drawinfo.headOnlyRender) {
+			return;
+		}
+
 		if (drawinfo.drawPlayer.leinforsHair && (drawinfo.fullHair || drawinfo.hatHair || drawinfo.drawsBackHairWithoutHeadgear || drawinfo.drawPlayer.head == -1 || drawinfo.drawPlayer.head == 0) && drawinfo.drawPlayer.hair != 12 && drawinfo.shadow == 0f && Main.rgbToHsl(drawinfo.colorHead).Z > 0.2f) {
 			if (Main.rand.Next(20) == 0 && !drawinfo.hatHair) {
 				Rectangle r = Utils.CenteredRectangle(drawinfo.Position + drawinfo.drawPlayer.Size / 2f + new Vector2(0f, drawinfo.drawPlayer.gravDir * -20f), new Vector2(20f, 14f));
@@ -514,7 +_,7 @@
 
 	public static void DrawPlayer_08_1_Tails(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.tail > 0 && drawinfo.drawPlayer.tail < ArmorIDs.Back.Count && !drawinfo.drawPlayer.mount.Active) {
+		if (drawinfo.drawPlayer.tail > 0 && !drawinfo.drawPlayer.mount.Active) {
 			Vector2 zero = Vector2.Zero;
 			if (drawinfo.isSitting)
 				zero.Y += -2f;
@@ -533,10 +_,10 @@
 
 	public static void DrawPlayer_10_BackAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.back <= 0 || drawinfo.drawPlayer.back >= ArmorIDs.Back.Count)
+		if (drawinfo.drawPlayer.back <= 0)
 			return;
 
-		if (drawinfo.drawPlayer.front >= 1 && drawinfo.drawPlayer.front <= 4) {
+		if (drawinfo.drawPlayer.front >= 1) {
 			int num = drawinfo.drawPlayer.bodyFrame.Y / 56;
 			if (num < 1 || num > 5) {
 				drawinfo.armorAdjust = 10;
@@ -1010,7 +_,7 @@
 
 	public static void DrawPlayer_12_1_BalloonFronts(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.balloonFront <= 0 || drawinfo.drawPlayer.balloonFront >= ArmorIDs.Balloon.Count)
+		if (drawinfo.drawPlayer.balloonFront <= 0)
 			return;
 
 		DrawData item;
@@ -1040,7 +_,7 @@
 
 	public static void DrawPlayer_11_Balloons(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.balloon <= 0 || drawinfo.drawPlayer.balloon >= ArmorIDs.Balloon.Count)
+		if (drawinfo.drawPlayer.balloon <= 0)
 			return;
 
 		DrawData item;
@@ -1087,7 +_,7 @@
 			drawinfo.Position.Y -= drawinfo.torsoOffset;
 		}
 
-		if (!drawinfo.hidesBottomSkin && !IsBottomOverridden(ref drawinfo)) {
+		if (!drawinfo.hidesBottomSkin && !drawinfo.isBottomOverriden) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 10].Value, drawinfo.colorLegs, 0, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 				return;
@@ -1098,7 +_,8 @@
 		}
 	}
 
+	//TML: Made internal.
-	public static bool IsBottomOverridden(ref PlayerDrawSet drawinfo)
+	internal static bool IsBottomOverridden(ref PlayerDrawSet drawinfo)
 	{
 		if (ShouldOverrideLegs_CheckPants(ref drawinfo))
 			return true;
@@ -1109,8 +_,12 @@
 		return false;
 	}
 
+	//TML: Made private
-	public static bool ShouldOverrideLegs_CheckPants(ref PlayerDrawSet drawinfo)
+	private static bool ShouldOverrideLegs_CheckPants(ref PlayerDrawSet drawinfo)
 	{
+		return drawinfo.drawPlayer.legs > 0 && ArmorIDs.Legs.Sets.OverridesLegs[drawinfo.drawPlayer.legs];
+
+		/*
 		switch (drawinfo.drawPlayer.legs) {
 			case 55:
 			case 63:
@@ -1127,15 +_,22 @@
 			default:
 				return false;
 		}
+		*/
 	}
 
+	//TML: Made private
-	public static bool ShouldOverrideLegs_CheckShoes(ref PlayerDrawSet drawinfo)
+	private static bool ShouldOverrideLegs_CheckShoes(ref PlayerDrawSet drawinfo)
 	{
-		sbyte shoe = drawinfo.drawPlayer.shoe;
+		int shoe = drawinfo.drawPlayer.shoe;
+
+		return shoe > 0 && ArmorIDs.Shoe.Sets.OverridesLegs[shoe];
+
+		/*
 		if (shoe == 15)
 			return true;
 
 		return false;
+		*/
 	}
 
 	public static void DrawPlayer_12_Skin_Composite(ref PlayerDrawSet drawinfo)
@@ -1168,7 +_,7 @@
 			});
 		}
 
-		if (!drawinfo.hidesBottomSkin && !drawinfo.drawPlayer.invis && !IsBottomOverridden(ref drawinfo)) {
+		if (!drawinfo.hidesBottomSkin && !drawinfo.drawPlayer.invis && !drawinfo.isBottomOverriden) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 10].Value, drawinfo.colorLegs, drawinfo.skinDyePacked, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 			}
@@ -1392,7 +_,7 @@
 				drawinfo.DrawDataCache.Add(item);
 			}
 		}
-		else if (drawinfo.drawPlayer.legs > 0 && drawinfo.drawPlayer.legs < ArmorIDs.Legs.Count && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
+		else if (drawinfo.drawPlayer.legs > 0 && (!ShouldOverrideLegs_CheckShoes(ref drawinfo) || drawinfo.drawPlayer.wearsRobe)) {
 			if (drawinfo.drawPlayer.invis)
 				return;
 
@@ -1592,7 +_,7 @@
 
 	public static void DrawPlayer_15_SkinLongCoat(ref PlayerDrawSet drawinfo)
 	{
-		if ((drawinfo.skinVar == 3 || drawinfo.skinVar == 8 || drawinfo.skinVar == 7) && (drawinfo.drawPlayer.body <= 0 || drawinfo.drawPlayer.body >= ArmorIDs.Body.Count) && !drawinfo.drawPlayer.invis) {
+		if ((drawinfo.skinVar == 3 || drawinfo.skinVar == 8 || drawinfo.skinVar == 7) && (drawinfo.drawPlayer.body <= 0) && !drawinfo.drawPlayer.invis) {
 			if (drawinfo.isSitting) {
 				DrawSittingLegs(ref drawinfo, TextureAssets.Players[drawinfo.skinVar, 14].Value, drawinfo.colorShirt, 0, drawinfo.drawPlayer.legs, default(Vector2), skin: true);
 				return;
@@ -1738,7 +_,7 @@
 		if (drawinfo.usesCompositeTorso) {
 			DrawPlayer_17_TorsoComposite(ref drawinfo);
 		}
-		else if (drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count) {
+		else if (drawinfo.drawPlayer.body > 0) {
 			Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
 			int num = drawinfo.armorAdjust;
 			bodyFrame.X += num;
@@ -1840,7 +_,7 @@
 
 	public static void DrawPlayer_18_OffhandAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (!drawinfo.usesCompositeBackHandAcc && drawinfo.drawPlayer.handoff > 0 && drawinfo.drawPlayer.handoff < ArmorIDs.HandOff.Count) {
+		if (!drawinfo.usesCompositeBackHandAcc && drawinfo.drawPlayer.handoff > 0) {
 			DrawData item = new DrawData(TextureAssets.AccHandsOff[drawinfo.drawPlayer.handoff].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cHandOff;
 			drawinfo.DrawDataCache.Add(item);
@@ -1860,7 +_,7 @@
 
 	public static void DrawPlayer_19_WaistAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.waist > 0 && drawinfo.drawPlayer.waist < ArmorIDs.Waist.Count) {
+		if (drawinfo.drawPlayer.waist > 0) {
 			Rectangle value = drawinfo.drawPlayer.legFrame;
 			if (ArmorIDs.Waist.Sets.UsesTorsoFraming[drawinfo.drawPlayer.waist])
 				value = drawinfo.drawPlayer.bodyFrame;
@@ -1873,7 +_,7 @@
 
 	public static void DrawPlayer_20_NeckAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.neck > 0 && drawinfo.drawPlayer.neck < ArmorIDs.Neck.Count) {
+		if (drawinfo.drawPlayer.neck > 0) {
 			DrawData item = new DrawData(TextureAssets.AccNeck[drawinfo.drawPlayer.neck].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cNeck;
 			drawinfo.DrawDataCache.Add(item);
@@ -1886,7 +_,7 @@
 		drawinfo.drawPlayer.ApplyHeadOffsetFromMount(ref pos);
 		Vector2 helmetOffset = drawinfo.helmetOffset + pos;
 		DrawPlayer_21_Head_TheFace(ref drawinfo);
-		bool flag = drawinfo.drawPlayer.head == 14 || drawinfo.drawPlayer.head == 56 || drawinfo.drawPlayer.head == 114 || drawinfo.drawPlayer.head == 158 || drawinfo.drawPlayer.head == 69 || drawinfo.drawPlayer.head == 180;
+		bool flag = drawinfo.drawPlayer.head >= 0 && ArmorIDs.Head.Sets.IsTallHat[drawinfo.drawPlayer.head];
 		bool flag2 = drawinfo.drawPlayer.head == 28;
 		bool flag3 = drawinfo.drawPlayer.head == 39 || drawinfo.drawPlayer.head == 38;
 		bool flag4 = true;
@@ -2218,6 +_,11 @@
 				drawinfo.DrawDataCache.Add(item);
 			}
 		}
+	}
+
+	public static void DrawPlayer_21_2_FinchNest(ref PlayerDrawSet drawinfo)
+	{
+		DrawData item;
 
 		if (!drawinfo.drawPlayer.babyBird)
 			return;
@@ -2320,7 +_,7 @@
 
 			float num = 0f;
 			if (drawinfo.drawPlayer.face == 5) {
-				sbyte faceHead = drawinfo.drawPlayer.faceHead;
+				int faceHead = drawinfo.drawPlayer.faceHead;
 				if ((uint)(faceHead - 10) <= 3u)
 					num = 2 * drawinfo.drawPlayer.direction;
 			}
@@ -2719,7 +_,7 @@
 
 	public static void DrawPlayer_25_Shield(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.shield <= 0 || drawinfo.drawPlayer.shield >= ArmorIDs.Shield.Count)
+		if (drawinfo.drawPlayer.shield <= 0 || drawinfo.drawPlayer.shield >= TextureAssets.AccShield.Length)
 			return;
 
 		Vector2 zero = Vector2.Zero;
@@ -3031,11 +_,14 @@
 						origin5 = Vector2.Zero;
 					}
 				}
+				// Extra patch context.
 				else if (drawinfo.drawPlayer.direction == -1) {
 					origin5 = new Vector2(rectangle.Width, rectangle.Height);
 					num7 -= (float)rectangle.Width;
 				}
 
+				ItemLoader.HoldoutOrigin(drawinfo.drawPlayer, ref origin5);
+
 				item = new DrawData(value, new Vector2((int)(drawinfo.ItemLocation.X - Main.screenPosition.X + origin5.X + num7), (int)(drawinfo.ItemLocation.Y - Main.screenPosition.Y + num8)), rectangle, heldItem.GetAlpha(drawinfo.itemColor), num6, origin5, adjustedItemScale, drawinfo.itemEffect);
 				drawinfo.DrawDataCache.Add(item);
 				if (num == 3870) {
@@ -3215,7 +_,7 @@
 		else if (drawinfo.usesCompositeTorso) {
 			DrawPlayer_28_ArmOverItemComposite(ref drawinfo);
 		}
-		else if (drawinfo.drawPlayer.body > 0 && drawinfo.drawPlayer.body < ArmorIDs.Body.Count) {
+		else if (drawinfo.drawPlayer.body > 0) {
 			Rectangle bodyFrame = drawinfo.drawPlayer.bodyFrame;
 			int num = drawinfo.armorAdjust;
 			bodyFrame.X += num;
@@ -3420,7 +_,7 @@
 
 	public static void DrawPlayer_29_OnhandAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (!drawinfo.usesCompositeFrontHandAcc && drawinfo.drawPlayer.handon > 0 && drawinfo.drawPlayer.handon < ArmorIDs.HandOn.Count) {
+		if (!drawinfo.usesCompositeFrontHandAcc && drawinfo.drawPlayer.handon > 0) {
 			DrawData item = new DrawData(TextureAssets.AccHandsOn[drawinfo.drawPlayer.handon].Value, new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cHandOn;
 			drawinfo.DrawDataCache.Add(item);
@@ -3455,7 +_,7 @@
 
 	public static void DrawPlayer_32_FrontAcc(ref PlayerDrawSet drawinfo)
 	{
-		if (drawinfo.drawPlayer.front > 0 && drawinfo.drawPlayer.front < ArmorIDs.Front.Count && !drawinfo.drawPlayer.mount.Active) {
+		if (drawinfo.drawPlayer.front > 0 && !drawinfo.drawPlayer.mount.Active) {
 			Vector2 zero = Vector2.Zero;
 			DrawData item = new DrawData(TextureAssets.AccFront[drawinfo.drawPlayer.front].Value, zero + new Vector2((int)(drawinfo.Position.X - Main.screenPosition.X - (float)(drawinfo.drawPlayer.bodyFrame.Width / 2) + (float)(drawinfo.drawPlayer.width / 2)), (int)(drawinfo.Position.Y - Main.screenPosition.Y + (float)drawinfo.drawPlayer.height - (float)drawinfo.drawPlayer.bodyFrame.Height + 4f)) + drawinfo.drawPlayer.bodyPosition + new Vector2(drawinfo.drawPlayer.bodyFrame.Width / 2, drawinfo.drawPlayer.bodyFrame.Height / 2), drawinfo.drawPlayer.bodyFrame, drawinfo.colorArmorBody, drawinfo.drawPlayer.bodyRotation, drawinfo.bodyVect, 1f, drawinfo.playerEffect);
 			item.shader = drawinfo.cFront;
