--- src/Terraria/Terraria/DataStructures/PlayerDrawSet.cs
+++ src/tModLoader/Terraria/DataStructures/PlayerDrawSet.cs
@@ -10,6 +_,14 @@
 {
 	public struct PlayerDrawSet
 	{
+		//A sad trick to avoid big patches for disabling lighting below.
+		private class Lighting
+		{
+			public static bool use;
+			public static Color GetColor(int x, int y) => use ? Terraria.Lighting.GetColor(x, y) : Color.White;
+			public static Color GetColor(int x, int y, Color oldColor) => use ? Terraria.Lighting.GetColor(x, y, oldColor) : oldColor;
+		}
+
 		public List<DrawData> DrawDataCache;
 		public List<int> DustCache;
 		public List<int> GoreCache;
@@ -127,7 +_,9 @@
 
 		public Vector2 Center => new Vector2(Position.X + (float)(drawPlayer.width / 2), Position.Y + (float)(drawPlayer.height / 2));
 
-		public void BoringSetup(Player player, List<DrawData> drawData, List<int> dust, List<int> gore, Vector2 drawPosition, float shadowOpacity, float rotation, Vector2 rotationOrigin) {
+		public void BoringSetup(Player player, List<DrawData> drawData, List<int> dust, List<int> gore, Vector2 drawPosition, float shadowOpacity, float rotation, Vector2 rotationOrigin, bool useLighting = true) {
+			Lighting.use = useLighting;
+
 			DrawDataCache = drawData;
 			DustCache = dust;
 			GoreCache = gore;
@@ -252,7 +_,7 @@
 
 			playerEffect = SpriteEffects.None;
 			itemEffect = SpriteEffects.FlipHorizontally;
-			colorHair = drawPlayer.GetImmuneAlpha(drawPlayer.GetHairColor(), shadow);
+			colorHair = drawPlayer.GetImmuneAlpha(drawPlayer.GetHairColor(useLighting), shadow);
 			colorEyeWhites = drawPlayer.GetImmuneAlpha(Lighting.GetColor((int)((double)Position.X + (double)drawPlayer.width * 0.5) / 16, (int)(((double)Position.Y + (double)drawPlayer.height * 0.25) / 16.0), Color.White), shadow);
 			colorEyes = drawPlayer.GetImmuneAlpha(Lighting.GetColor((int)((double)Position.X + (double)drawPlayer.width * 0.5) / 16, (int)(((double)Position.Y + (double)drawPlayer.height * 0.25) / 16.0), drawPlayer.eyeColor), shadow);
 			colorHead = drawPlayer.GetImmuneAlpha(Lighting.GetColor((int)((double)Position.X + (double)drawPlayer.width * 0.5) / 16, (int)(((double)Position.Y + (double)drawPlayer.height * 0.25) / 16.0), drawPlayer.skinColor), shadow);
@@ -1046,7 +_,7 @@
 				if (drawPlayer.onFire || drawPlayer.onFire2 || drawPlayer.onFrostBurn || drawPlayer.onFire3 || drawPlayer.onFrostBurn2) {
 					colorEyeWhites = drawPlayer.GetImmuneAlpha(Color.White, shadow);
 					colorEyes = drawPlayer.GetImmuneAlpha(drawPlayer.eyeColor, shadow);
-					colorHair = drawPlayer.GetImmuneAlpha(drawPlayer.GetHairColor(useLighting: false), shadow);
+					colorHair = drawPlayer.GetImmuneAlpha(drawPlayer.GetHairColor(useLighting), shadow);
 					colorHead = drawPlayer.GetImmuneAlpha(drawPlayer.skinColor, shadow);
 					colorBodySkin = drawPlayer.GetImmuneAlpha(drawPlayer.skinColor, shadow);
 					colorShirt = drawPlayer.GetImmuneAlpha(drawPlayer.shirtColor, shadow);
