--- src/Terraria/Terraria/DataStructures/ProjectileSource_NPC.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_NPC.cs
@@ -1,10 +_,11 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from ProjectileSource*
-	public class ProjectileSource_NPC : IProjectileSource
+	public class EntitySource_NPC : IEntitySource
 	{
 		public readonly NPC NPC;
 
-		public ProjectileSource_NPC(NPC npc) {
+		public EntitySource_NPC(NPC npc) {
 			NPC = npc;
 		}
 	}
