--- src/Terraria/Terraria/DataStructures/ProjectileSource_BySourceId.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_BySourceId.cs
@@ -1,10 +_,11 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from ProjectileSource*
-	public class ProjectileSource_BySourceId : IProjectileSource
+	public class EntitySource_BySourceId : IEntitySource
 	{
 		public readonly int SourceId;
 
-		public ProjectileSource_BySourceId(int projectileSourceId) {
+		public EntitySource_BySourceId(int projectileSourceId) {
 			SourceId = projectileSourceId;
 		}
 	}
