--- src/Terraria/Terraria/DataStructures/ProjectileSource_Mount.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_Mount.cs
@@ -1,11 +_,12 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from ProjectileSource*
-	public class ProjectileSource_Mount : IProjectileSource
+	public class EntitySource_Mount : IEntitySource
 	{
 		public readonly Player Player;
 		public readonly int MountId;
 
-		public ProjectileSource_Mount(Player player, int mountId) {
+		public EntitySource_Mount(Player player, int mountId) {
 			Player = player;
 			MountId = mountId;
 		}
