--- src/Terraria/Terraria/DataStructures/ProjectileSource_Item.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_Item.cs
@@ -1,11 +_,12 @@
 namespace Terraria.DataStructures
 {
-	public class ProjectileSource_Item : IProjectileSource
+	// Renamed from ProjectileSource_Item
+	public class EntitySource_ItemUse : IEntitySource
 	{
 		public readonly Player Player;
 		public readonly Item Item;
 
-		public ProjectileSource_Item(Player player, Item item) {
+		public EntitySource_ItemUse(Player player, Item item) {
 			Player = player;
 			Item = item;
 		}
