--- src/Terraria/Terraria/DataStructures/ProjectileSource_ShakeTree.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_ShakeTree.cs
@@ -1,8 +_,9 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from ProjectileSource*
-	public class ProjectileSource_ShakeTree : AProjectileSource_Tile
+	public class EntitySource_ShakeTree : AEntitySource_Tile
 	{
-		public ProjectileSource_ShakeTree(int tileCoordsX, int tileCoordsY)
+		public EntitySource_ShakeTree(int tileCoordsX, int tileCoordsY)
 			: base(tileCoordsX, tileCoordsY) {
 		}
 	}
