--- src/Terraria/Terraria/DataStructures/ProjectileSource_Buff.cs
+++ src/tModLoader/Terraria/DataStructures/ProjectileSource_Buff.cs
@@ -1,12 +_,13 @@
 namespace Terraria.DataStructures
 {
+	// Renamed from ProjectileSource*
-	public class ProjectileSource_Buff : IProjectileSource
+	public class EntitySource_Buff : IEntitySource
 	{
 		public readonly Player Player;
 		public readonly int BuffId;
 		public readonly int BuffIndex;
 
-		public ProjectileSource_Buff(Player player, int buffId, int buffIndex) {
+		public EntitySource_Buff(Player player, int buffId, int buffIndex) {
 			Player = player;
 			BuffId = buffId;
 			BuffIndex = buffIndex;
