--- src/TerrariaNetCore/Terraria/DataStructures/RecipeItemCreationContext.cs
+++ src/tModLoader/Terraria/DataStructures/RecipeItemCreationContext.cs
@@ -1,11 +_,33 @@
+using System.Collections.Generic;
+
 namespace Terraria.DataStructures;
 
+/// <summary>
+/// Provides the context of an item being crafted from a recipe. Includes the recipe, consumed items, and destination, facilitating features using data from consumed items to affect the final crafted item.
+/// </summary>
 public class RecipeItemCreationContext : ItemCreationContext
 {
 	public readonly Recipe Recipe;
 
+	// Added by TML.
+	/// <summary>
+	/// An item stack that the created item will be combined with (via OnStack). For normal crafting, this is Main.mouseItem
+	/// </summary>
+	public Item DestinationStack;
+
+	// Added by TML.
+	/// <summary>
+	/// Cloned list of Items consumed when crafting.
+	/// </summary>
+	public readonly List<Item> ConsumedItems;
+
+	/*
 	public RecipeItemCreationContext(Recipe recipe)
+	*/
+	public RecipeItemCreationContext(Recipe recipe, List<Item> consumedItems, Item destinationStack)
 	{
 		Recipe = recipe;
+		ConsumedItems = consumedItems;
+		DestinationStack = destinationStack;
 	}
 }
