--- src/tModLoader/Terraria/Netplay.cs
+++ src/64bit/Terraria/Netplay.cs
@@ -214,11 +_,12 @@
 					flag = false;
 				}
 				catch {
-#if MAC
-					Thread.Sleep(200);
-					Netplay.Connection.Socket.Close();
-					Netplay.Connection.Socket = new TcpSocket();
-#endif
+					if (PlatformUtilities.IsOSX) {
+						Thread.Sleep(200);
+						Netplay.Connection.Socket.Close();
+						Netplay.Connection.Socket = new TcpSocket();
+					}
+
 					if (!disconnect && Main.gameMenu)
 						flag = true;
 				}
