--- src/merged\Terraria.IO\WorldFile.cs
+++ src/Terraria\Terraria.IO\WorldFile.cs
@@ -1148,7 +_,7 @@
 			writer.Write(NPC.savedTaxCollector);
 			writer.Write(Main.invasionSizeStart);
 			writer.Write(WorldFile.tempCultistDelay);
-			writer.Write(580);
+			writer.Write((short)580);
 			for (int j = 0; j < 580; j++)
 			{
 				writer.Write(NPC.killCount[j]);
@@ -1412,7 +_,7 @@
 				}
 			}
 			writer.Write(num);
-			writer.Write(40);
+			writer.Write((short)40);
 			for (int i = 0; i < 1000; i++)
 			{
 				Chest chest = Main.chest[i];
@@ -1426,7 +_,7 @@
 						Item item = chest.item[l];
 						if (item == null)
 						{
-							writer.Write(0);
+							writer.Write((short)0);
 						}
 						else
 						{
@@ -2057,7 +_,7 @@
 							}
 							else
 							{
-								tile.slope(b4 - 1);
+								tile.slope((byte)(b4 - 1));
 							}
 						}
 					}
@@ -2692,7 +_,7 @@
 			}
 			try
 			{
-				using (Stream stream = cloudSave ? new MemoryStream(SocialAPI.Cloud.Read(file)) : new FileStream(file, FileMode.Open))
+				using (Stream stream = cloudSave ? (Stream)new MemoryStream(SocialAPI.Cloud.Read(file)) : (Stream)new FileStream(file, FileMode.Open))
 				{
 					using (BinaryReader binaryReader = new BinaryReader(stream))
 					{
@@ -2862,7 +_,7 @@
 					buffer = new byte[num];
 					SocialAPI.Cloud.Read(file, buffer, num);
 				}
-				using (Stream stream = flag ? new MemoryStream(buffer) : new FileStream(file, FileMode.Open))
+				using (Stream stream = flag ? (Stream)new MemoryStream(buffer) : (Stream)new FileStream(file, FileMode.Open))
 				{
 					using (BinaryReader binaryReader = new BinaryReader(stream))
 					{

