--- src/TerrariaNetCore/ReLogic/OS/OSX/WindowService.cs
+++ src/tModLoader/ReLogic/OS/OSX/WindowService.cs
@@ -22,4 +_,24 @@
 	public void StopFlashingIcon(GameWindow window)
 	{
 	}
+
+	// Added by TML.
+	public void HideConsole()
+	{
+	}
+
+	// Added by TML.
+	public void ShowConsole()
+	{
+	}
+
+	// Added by TML.
+	public void ReleaseConsole()
+	{
+	}
+
+	// Added by TML.
+	public void CreateAndRedirectConsole()
+	{
+	}
 }
