--- src/TerrariaNetCore/ReLogic/OS/OSX/WindowService.cs
+++ src/tModLoader/ReLogic/OS/OSX/WindowService.cs
@@ -4,4 +_,23 @@
 
 internal class WindowService : SDLWindowService
 {
+	// Added by TML.
+	public void HideConsole()
+	{
+	}
+
+	// Added by TML.
+	public void ShowConsole()
+	{
+	}
+
+	// Added by TML.
+	public void ReleaseConsole()
+	{
+	}
+
+	// Added by TML.
+	public void CreateAndRedirectConsole()
+	{
+	}
 }
