--- src/TerrariaNetCore/Terraria/UI/AchievementAdvisor.cs
+++ src/tModLoader/Terraria/UI/AchievementAdvisor.cs
@@ -1,3 +_,4 @@
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.Xna.Framework;
@@ -146,7 +_,10 @@
 
 		Rectangle frame = card.frame;
 		frame.X += 528;
-		spriteBatch.Draw(_achievementsTexture.Value, position, frame, color, 0f, Vector2.Zero, scale, SpriteEffects.None, 0f);
+		if (card.achievement.ModAchievement != null) {
+			frame.X = 66;
+		}
+		spriteBatch.Draw(card.achievement.ModAchievement?.Texture.Value ?? _achievementsTexture.Value, position, frame, color, 0f, Vector2.Zero, scale, SpriteEffects.None, 0f);
 		spriteBatch.Draw(_achievementsBorderTexture.Value, position + vector, null, color, 0f, Vector2.Zero, scale, SpriteEffects.None, 0f);
 		if (hovered)
 			spriteBatch.Draw(value, position + vector2, null, Main.OurFavoriteColor, 0f, Vector2.Zero, scale, SpriteEffects.None, 0f);
@@ -207,5 +_,6 @@
 		_cards.Add(new AchievementAdvisorCard(Main.Achievements.GetAchievement("STAR_DESTROYER"), num++));
 		_cards.Add(new AchievementAdvisorCard(Main.Achievements.GetAchievement("CHAMPION_OF_TERRARIA"), num++));
 		_cards.OrderBy((AchievementAdvisorCard x) => x.order);
+		// TODO: Consider supporting modded achievements in the AchievementAdvisor subset/ordering
 	}
 }
