--- src/TerrariaNetCore/Terraria/UI/ChestUI.cs
+++ src/tModLoader/Terraria/UI/ChestUI.cs
@@ -10,6 +_,8 @@
 using Terraria.GameContent;
 using Terraria.GameContent.UI;
 using Terraria.GameInput;
+using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 using Terraria.UI.Gamepad;
 
@@ -97,6 +_,8 @@
 					text = Lang.chestType2[tile.frameX / 36].Value;
 				else if (tile.type == 88)
 					text = Lang.dresserType[tile.frameX / 54].Value;
+				else if (TileID.Sets.BasicChest[Main.tile[player.chestX, player.chestY].type] || TileID.Sets.BasicDresser[Main.tile[player.chestX, player.chestY].type])
+					text = TileLoader.DefaultContainerName(tile.type, tile.TileFrameX, tile.TileFrameY);
 			}
 		}
 		else if (player.chest == -2) {
@@ -404,12 +_,17 @@
 			if (chest.item[i].stack >= chest.item[i].maxStack || !Item.CanStack(player.inventory[playerInventorySlot], chest.item[i]))
 				continue;
 
+			ItemLoader.StackItems(chest.item[i], player.inventory[playerInventorySlot], out _);
+
+			/* #OnStackHook
 			int num = player.inventory[playerInventorySlot].stack;
 			if (player.inventory[playerInventorySlot].stack + chest.item[i].stack > chest.item[i].maxStack)
 				num = chest.item[i].maxStack - chest.item[i].stack;
 
 			player.inventory[playerInventorySlot].stack -= num;
 			chest.item[i].stack += num;
+			*/
+
 			SoundEngine.PlaySound(7);
 			if (player.inventory[playerInventorySlot].stack <= 0) {
 				player.inventory[playerInventorySlot].SetDefaults(0);
@@ -433,12 +_,17 @@
 	{
 		for (int i = 0; i < chest.maxItems; i++) {
 			if (chest.item[i].stack < chest.item[i].maxStack && Item.CanStack(player.inventory[p], chest.item[i])) {
+				ItemLoader.StackItems(chest.item[i], player.inventory[p], out _);
+
+				/* #OnStackHook
 				int num = player.inventory[p].stack;
 				if (player.inventory[p].stack + chest.item[i].stack > chest.item[i].maxStack)
 					num = chest.item[i].maxStack - chest.item[i].stack;
 
 				player.inventory[p].stack -= num;
 				chest.item[i].stack += num;
+				*/
+
 				SoundEngine.PlaySound(7);
 				if (player.inventory[p].stack <= 0) {
 					player.inventory[p].SetDefaults(0);
@@ -610,10 +_,15 @@
 						num4 = num5;
 
 					SoundEngine.PlaySound(7);
+
+					ItemLoader.TryStackItems(item[num3], array[item2.Key], out num4);
+
+					/* #OnStackHook
 					item[num3].stack += num4;
 					array[item2.Key].stack -= num4;
 					if (array[item2.Key].stack == 0)
 						array[item2.Key].SetDefaults(0);
+					*/
 
 					array2[num3] = true;
 				}
@@ -651,8 +_,12 @@
 					if (num7 > num8)
 						num7 = num8;
 
+					ItemLoader.TryStackItems(item[num6], array[item5.Key], out num7);
+
+					/* #OnStackHook
 					item[num6].stack += num7;
 					array[item5.Key].stack -= num7;
+					*/
 					if (array[item5.Key].stack == 0)
 						array[item5.Key] = new Item();
 
@@ -760,12 +_,16 @@
 				if (!Item.CanStack(inventory[num2], item[i]) || ItemSlot.PickItemMovementAction(inventory, context, num2, item[i]) == -1)
 					continue;
 
+				ItemLoader.StackItems(inventory[num2], item[i], out _); // #StackablePrefixWeapons, check moved here
+
+				/* #OnStackHook
 				int num3 = item[i].stack;
 				if (inventory[num2].maxStack - inventory[num2].stack < num3)
 					num3 = inventory[num2].maxStack - inventory[num2].stack;
 
 				inventory[num2].stack += num3;
 				item[i].stack -= num3;
+				*/
 				flag = true;
 				if (inventory[num2].stack == inventory[num2].maxStack) {
 					if (Main.netMode == 1 && Main.player[Main.myPlayer].chest > -1)
@@ -1015,8 +_,12 @@
 					break;
 				}
 
+				ItemLoader.StackItems(item2[i], item, out _);
+
+				/* #OnStackHook
 				item.stack -= num;
 				item2[i].stack += num;
+				*/
 				SoundEngine.PlaySound(7);
 				if (item.stack <= 0) {
 					item.SetDefaults(0);
