--- src/Terraria/Terraria/UI/ChestUI.cs
+++ src/tModLoader/Terraria/UI/ChestUI.cs
@@ -6,6 +_,8 @@
 using Terraria.GameContent;
 using Terraria.GameInput;
 using Terraria.Localization;
+using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 using Terraria.UI.Gamepad;
 
@@ -105,6 +_,8 @@
 						text = Lang.chestType2[tile.frameX / 36].Value;
 					else if (tile.type == 88)
 						text = Lang.dresserType[tile.frameX / 54].Value;
+					else if (TileID.Sets.BasicChest[Main.tile[player.chestX, player.chestY].type] || TileID.Sets.BasicDresser[Main.tile[player.chestX, player.chestY].type])
+						text = TileLoader.ContainerName(Main.tile[player.chestX, player.chestY].type);
 				}
 			}
 			else if (player.chest == -2) {
@@ -1078,10 +_,10 @@
 		}
 
 		public static bool IsBlockedFromTransferIntoChest(Item item, Item[] container) {
-			if (item.type == 3213 && item.favorited && container == Main.LocalPlayer.bank.item)
+			if (item.type == 3213 && item.favorited && container == Main.LocalPlayer.bank.item.GetItemArray())
 				return true;
 
-			if (item.type == 4131 && item.favorited && container == Main.LocalPlayer.bank4.item)
+			if (item.type == 4131 && item.favorited && container == Main.LocalPlayer.bank4.item.GetItemArray())
 				return true;
 
 			return false;
