--- src/TerrariaNetCore/Terraria/UI/FancyErrorPrinter.cs
+++ src/tModLoader/Terraria/UI/FancyErrorPrinter.cs
@@ -4,6 +_,7 @@
 using System.Text;
 using ReLogic.Content;
 using Terraria.Utilities;
+using Terraria.ModLoader.Engine;
 
 namespace Terraria.UI;
 
@@ -37,8 +_,7 @@
 			AppendSuggestions(stringBuilder, suggestions);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(stringBuilder.ToString());
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -71,8 +_,7 @@
 			AppendSuggestions(stringBuilder, list);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(stringBuilder.ToString());
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -105,8 +_,7 @@
 			AppendSuggestions(stringBuilder, list);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(exception);
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -124,9 +_,11 @@
 		}
 	}
 
+	/* Use ErrorReporting.MessageBoxShow instead
 	private static void ShowTheBox(string preparedMessage)
 	{
 		if (!Main.dedServ)
 			MessageBox.Show(preparedMessage, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
 	}
+	*/
 }
