--- src/TerrariaNetCore/Terraria/UI/FancyErrorPrinter.cs
+++ src/tModLoader/Terraria/UI/FancyErrorPrinter.cs
@@ -3,7 +_,11 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Text;
+using System.Threading;
 using System.Windows.Forms;
+using Terraria.ModLoader.Engine;
+using Terraria.ModLoader;
+using Terraria.Localization;
 
 namespace Terraria.UI
 {
@@ -36,8 +_,7 @@
 				AppendSuggestions(stringBuilder, suggestions);
 				stringBuilder.AppendLine();
 				IncludeOriginalMessage(stringBuilder, exception);
-				ShowTheBox(stringBuilder.ToString());
-				Console.WriteLine(stringBuilder.ToString());
+				ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 			}
 		}
 
@@ -69,8 +_,7 @@
 				AppendSuggestions(stringBuilder, list);
 				stringBuilder.AppendLine();
 				IncludeOriginalMessage(stringBuilder, exception);
-				ShowTheBox(stringBuilder.ToString());
-				Console.WriteLine(stringBuilder.ToString());
+				ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 			}
 		}
 
@@ -102,8 +_,7 @@
 				AppendSuggestions(stringBuilder, list);
 				stringBuilder.AppendLine();
 				IncludeOriginalMessage(stringBuilder, exception);
-				ShowTheBox(stringBuilder.ToString());
-				Console.WriteLine(exception);
+				ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 			}
 		}
 
@@ -117,13 +_,6 @@
 				string str = suggestions[i];
 				text.AppendLine(i + 1 + ". " + str);
 			}
-		}
-
-		private static void ShowTheBox(string preparedMessage) {
-			if (Main.dedServ) // if #SERVER
-				return;
-			
-			MessageBox.Show(preparedMessage, "Error", MessageBoxButtons.OK, MessageBoxIcon.Hand);
 		}
 	}
 }
