--- src/TerrariaNetCore/Terraria/UI/FancyErrorPrinter.cs
+++ src/tModLoader/Terraria/UI/FancyErrorPrinter.cs
@@ -5,6 +_,8 @@
 using System.Windows.Forms;
 using ReLogic.Content;
 using ReLogic.OS;
+using Terraria.ModLoader.Engine;
+using Terraria.ModLoader.Exceptions;
 
 namespace Terraria.UI;
 
@@ -38,8 +_,7 @@
 			AppendSuggestions(stringBuilder, suggestions);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(stringBuilder.ToString());
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -55,6 +_,14 @@
 		if (exception is DirectoryNotFoundException)
 			flag = true;
 
+		if (exception is CustomModDataException ex) {
+			StringBuilder stringBuilder = new StringBuilder();
+			stringBuilder.AppendLine($"Failed to save world because this mod is failing: {ex.modName}");
+			stringBuilder.AppendLine();
+			IncludeOriginalMessage(stringBuilder, ex);
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
+		}
+
 		if (flag) {
 			StringBuilder stringBuilder = new StringBuilder();
 			stringBuilder.AppendLine("Failed to create the file: \"" + filePath.Replace("/", "\\") + "\"!");
@@ -72,8 +_,7 @@
 			AppendSuggestions(stringBuilder, list);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(stringBuilder.ToString());
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -106,8 +_,7 @@
 			AppendSuggestions(stringBuilder, list);
 			stringBuilder.AppendLine();
 			IncludeOriginalMessage(stringBuilder, exception);
-			ShowTheBox(stringBuilder.ToString());
-			Console.WriteLine(exception);
+			ErrorReporting.MessageBoxShow(stringBuilder.ToString());
 		}
 	}
 
@@ -125,9 +_,11 @@
 		}
 	}
 
+	/* Use ErrorReporting.MessageBoxShow instead
 	private static void ShowTheBox(string preparedMessage)
 	{
 		if (Platform.IsWindows && !Main.dedServ)
 			MessageBox.Show(preparedMessage, "Error", MessageBoxButtons.OK, MessageBoxIcon.Hand);
 	}
+	*/
 }
