--- src/TerrariaNetCore/Terraria/UI/GameInterfaceLayer.cs
+++ src/tModLoader/Terraria/UI/GameInterfaceLayer.cs
@@ -10,14 +_,21 @@
 	public readonly string Name;
 	public InterfaceScaleType ScaleType;
 
+	// Added by TML.
+	public bool Active;
+
 	public GameInterfaceLayer(string name, InterfaceScaleType scaleType)
 	{
 		Name = name;
 		ScaleType = scaleType;
+		Active = true;
 	}
 
 	public bool Draw()
 	{
+		if (!Active)
+			return true;
+
 		Matrix transformMatrix;
 		if (ScaleType == InterfaceScaleType.Game) {
 			PlayerInput.SetZoom_World();
