--- src/TerrariaNetCore/Terraria/UI/IHaveBackButtonCommand.cs
+++ src/tModLoader/Terraria/UI/IHaveBackButtonCommand.cs
@@ -1,6 +_,26 @@
+using Terraria.Audio;
+using Terraria.ID;
+
 namespace Terraria.UI;
 
 public interface IHaveBackButtonCommand
 {
-	void HandleBackButtonUsage();
+	// Added by TML.
+	UIState PreviousUIState { get; set; }
+
+	//TML: Added default implementation.
+	void HandleBackButtonUsage() => GoBackTo(PreviousUIState);
+
+	public static void GoBackTo(UIState state)
+	{
+		if (state == null) {
+			Main.menuMode = 0;
+		}
+		else {
+			Main.menuMode = MenuID.FancyUI;
+			Main.MenuUI.SetState(state);
+		}
+
+		SoundEngine.PlaySound(SoundID.MenuClose);
+	}
 }
