--- src/Terraria/Terraria/UI/IHaveBackButtonCommand.cs
+++ src/tModLoader/Terraria/UI/IHaveBackButtonCommand.cs
@@ -1,7 +_,20 @@
+using Terraria.Audio;
+
 namespace Terraria.UI
 {
 	public interface IHaveBackButtonCommand
 	{
-		void HandleBackButtonUsage();
+		UIState PreviousUIState { get; set; }
+
+		void HandleBackButtonUsage() {
+			if (PreviousUIState == null) {
+				Main.menuMode = 0;
+				return;
+			}
+
+			Main.menuMode = 888;
+			Main.MenuUI.SetState(PreviousUIState);
+			SoundEngine.PlaySound(11);
+		}
 	}
 }
