--- src/Terraria/Terraria/UI/IngameFancyUI.cs
+++ src/tModLoader/Terraria/UI/IngameFancyUI.cs
@@ -4,7 +_,9 @@
 using Terraria.Audio;
 using Terraria.GameContent.UI.States;
 using Terraria.GameInput;
+using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Gamepad;
 
 namespace Terraria.UI
@@ -85,6 +_,9 @@
 							Main.defaultChestName = Lang.chestType2[tile.frameX / 36].Value;
 						else if (tile.type == 88)
 							Main.defaultChestName = Lang.dresserType[tile.frameX / 54].Value;
+
+						if (tile.type >= TileID.Count && (TileID.Sets.BasicChest[tile.type] || TileID.Sets.BasicDresser[tile.type]))
+							Main.defaultChestName = TileLoader.ContainerName(Main.tile[player.chestX, player.chestY].type);
 
 						if (Main.npcChatText == "")
 							Main.npcChatText = Main.defaultChestName;
