--- src/Terraria/Terraria/UI/ItemSlot.cs
+++ src/tModLoader/Terraria/UI/ItemSlot.cs
@@ -1,6 +_,7 @@
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using System;
+using System.Linq;
 using Terraria.Audio;
 using Terraria.DataStructures;
 using Terraria.GameContent;
@@ -10,6 +_,7 @@
 using Terraria.GameInput;
 using Terraria.ID;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.UI.Chat;
 using Terraria.UI.Gamepad;
 
@@ -61,7 +_,7 @@
 
 		public static bool DrawGoldBGForCraftingMaterial;
 		public static bool ShiftForcedOn;
-		private static Item[] singleSlotArray;
+		internal static Item[] singleSlotArray;
 		private static bool[] canFavoriteAt;
 		private static bool[] canShareAt;
 		private static float[] inventoryGlowHue;
@@ -352,6 +_,10 @@
 				return true;
 
 			Item item = inv[slot];
+			// TODO: Make this more generalized.
+			if (ShiftInUse && PlayerHooks.ShiftClickSlot(Main.player[Main.myPlayer], inv, context, slot)) {
+				return true;
+			}
 			if (Main.cursorOverride == 2) {
 				if (ChatManager.AddChatText(FontAssets.MouseText.Value, ItemTagHandler.GenerateTag(item), Vector2.One))
 					SoundEngine.PlaySound(12);
@@ -651,18 +_,24 @@
 					break;
 				case 3:
 					HandleShopSlot(inv, slot, rightClickIsValid: false, leftClickIsValid: true);
+						
 					break;
 				case 4: {
+						if (!PlayerHooks.CanSellItem(player, Main.npc[player.talkNPC], inv, Main.mouseItem))
+							break;
+							
 						Chest chest = Main.instance.shop[Main.npcShop];
 						if (player.SellItem(Main.mouseItem)) {
-							chest.AddItemToShop(Main.mouseItem);
+							int soldItemIndex = chest.AddItemToShop(Main.mouseItem);
 							Main.mouseItem.SetDefaults();
 							SoundEngine.PlaySound(18);
+							PlayerHooks.PostSellItem(player, Main.npc[player.talkNPC], chest.item, chest.item[soldItemIndex]);
 						}
 						else if (Main.mouseItem.value == 0) {
-							chest.AddItemToShop(Main.mouseItem);
+							int soldItemIndex = chest.AddItemToShop(Main.mouseItem);
 							Main.mouseItem.SetDefaults();
 							SoundEngine.PlaySound(7);
+							PlayerHooks.PostSellItem(player, Main.npc[player.talkNPC], chest.item, chest.item[soldItemIndex]);
 						}
 
 						Recipe.FindRecipes();
@@ -714,18 +_,20 @@
 
 			if (Main.npcShop > 0 && !inv[slot].favorited) {
 				Chest chest = Main.instance.shop[Main.npcShop];
-				if (inv[slot].type < 71 || inv[slot].type > 74) {
+				if (inv[slot].type < 71 || inv[slot].type > 74 && PlayerHooks.CanSellItem(player, Main.npc[player.talkNPC], chest.item, inv[slot])) {					
 					if (player.SellItem(inv[slot])) {
-						chest.AddItemToShop(inv[slot]);
+						int soldItemIndex = chest.AddItemToShop(inv[slot]);
 						inv[slot].SetDefaults();
 						SoundEngine.PlaySound(18);
 						Recipe.FindRecipes();
+						PlayerHooks.PostSellItem(player, Main.npc[player.talkNPC], chest.item, chest.item[soldItemIndex]);
 					}
 					else if (inv[slot].value == 0) {
-						chest.AddItemToShop(inv[slot]);
+						int soldItemIndex = chest.AddItemToShop(inv[slot]);
 						inv[slot].SetDefaults();
 						SoundEngine.PlaySound(7);
 						Recipe.FindRecipes();
+						PlayerHooks.PostSellItem(player, Main.npc[player.talkNPC], chest.item, chest.item[soldItemIndex]);
 					}
 				}
 			}
@@ -1000,10 +_,12 @@
 			switch (context) {
 				case 0:
 					result = true;
-					if (Main.mouseRight && ((inv[slot].type >= 3318 && inv[slot].type <= 3332) || inv[slot].type == 3860 || inv[slot].type == 3862 || inv[slot].type == 3861 || inv[slot].type == 4782 || inv[slot].type == 4957)) {
+					if (Main.mouseRight && ((inv[slot].type >= 3318 && inv[slot].type <= 3332) || inv[slot].type == 3860 || inv[slot].type == 3862 || inv[slot].type == 3861 || inv[slot].type == 4782 || inv[slot].type == 4957 || ItemLoader.IsModBossBag(inv[slot]))) {
 						if (Main.mouseRightRelease) {
 							player.OpenBossBag(inv[slot].type);
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1016,7 +_,9 @@
 					else if (Main.mouseRight && inv[slot].type > 0 && inv[slot].type < 5088 && ItemID.Sets.IsFishingCrate[inv[slot].type]) {
 						if (Main.mouseRightRelease) {
 							player.OpenFishingCrate(inv[slot].type);
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1029,7 +_,9 @@
 					else if (Main.mouseRight && inv[slot].type == 3093) {
 						if (Main.mouseRightRelease) {
 							player.OpenHerbBag();
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1093,7 +_,9 @@
 					}
 					else if (Main.mouseRight && inv[slot].type == 1774) {
 						if (Main.mouseRightRelease) {
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1106,7 +_,9 @@
 					}
 					else if (Main.mouseRight && inv[slot].type == 3085) {
 						if (Main.mouseRightRelease && player.ConsumeItem(327)) {
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1132,7 +_,9 @@
 					}
 					else if (Main.mouseRight && inv[slot].type == 1869) {
 						if (Main.mouseRightRelease) {
+							if (ItemLoader.ConsumeItem(inv[slot], player))
-							inv[slot].stack--;
+								inv[slot].stack--;
+
 							if (inv[slot].stack == 0)
 								inv[slot].SetDefaults();
 
@@ -1162,6 +_,9 @@
 
 						Recipe.FindRecipes();
 					}
+					else if (EquipLoader.CanRightClick(inv[slot])) {
+						EquipLoader.RightClick(inv[slot], player);
+					}
 					else {
 						result = false;
 					}
@@ -1175,22 +_,22 @@
 						if (Main.npcShop > 0)
 							return true;
 
-						if ((inv[slot].type <= 0 || inv[slot].stack <= 0) && (inv[slot - 10].type <= 0 || inv[slot - 10].stack <= 0))
+						if ((inv[slot].type <= 0 || inv[slot].stack <= 0) && (inv[slot - inv.Length/2].type <= 0 || inv[slot - inv.Length/2].stack <= 0))
 							break;
 
-						Item item = inv[slot - 10];
+						Item item = inv[slot - inv.Length / 2];
 						bool flag2 = context != 11 || item.FitsAccessoryVanitySlot || item.IsAir;
 						if (flag2 && context == 11 && inv[slot].wingSlot > 0) {
-							for (int i = 3; i < 10; i++) {
+							for (int i = 0; i < inv.Length / 2; i++) {
-								if (inv[i].wingSlot > 0 && i != slot - 10)
+								if (inv[i].wingSlot > 0 && i != slot - inv.Length / 2)
 									flag2 = false;
 							}
 						}
 
-						if (!flag2)
+						if (!flag2 || !ItemLoader.CanEquipAccessory(inv[slot], slot - inv.Length / 2))
 							break;
 
-						Utils.Swap(ref inv[slot], ref inv[slot - 10]);
+						Utils.Swap(ref inv[slot], ref inv[slot - inv.Length / 2]);
 						SoundEngine.PlaySound(7);
 						Recipe.FindRecipes();
 						if (inv[slot].stack > 0) {
@@ -1264,6 +_,8 @@
 			int num = Main.superFastStack + 1;
 			Player localPlayer = Main.LocalPlayer;
 			for (int i = 0; i < num; i++) {
+				if (!PlayerHooks.CanBuyItem(localPlayer, Main.npc[localPlayer.talkNPC], inv, inv[slot]))
+					continue;
 				if (Main.mouseItem.stack >= Main.mouseItem.maxStack && Main.mouseItem.type != 0)
 					continue;
 
@@ -1276,13 +_,12 @@
 						SoundEngine.PlaySound(18);
 					else
 						SoundEngine.PlaySound(7);
+					PlayerHooks.PostBuyItem(localPlayer, Main.npc[localPlayer.talkNPC], inv, Main.mouseItem);
 				}
 
 				if (Main.mouseItem.type == 0) {
-					Main.mouseItem.netDefaults(inv[slot].netID);
-					if (inv[slot].prefix != 0)
-						Main.mouseItem.Prefix(inv[slot].prefix);
-
+					Main.mouseItem = inv[slot].Clone();
+					Main.mouseItem.buyOnce = false;
 					Main.mouseItem.stack = 0;
 				}
 
@@ -1293,6 +_,8 @@
 				RefreshStackSplitCooldown();
 				if (inv[slot].buyOnce && --inv[slot].stack <= 0)
 					inv[slot].SetDefaults();
+
+				PlayerHooks.PostBuyItem(localPlayer, Main.npc[localPlayer.talkNPC], inv, Main.mouseItem);
 			}
 		}
 
@@ -1628,10 +_,15 @@
 				num11 *= inventoryScale;
 				Vector2 position2 = position + vector / 2f - rectangle2.Size() * num11 / 2f;
 				Vector2 origin = rectangle2.Size() * (scale3 / 2f - 0.5f);
+				if (!ItemLoader.PreDrawInInventory(item, spriteBatch, position2, rectangle2, item.GetAlpha(currentColor), item.GetColor(color), origin, num11 * scale3))
+					goto skip;
+				
 				spriteBatch.Draw(value7, position2, rectangle2, item.GetAlpha(currentColor), 0f, origin, num11 * scale3, SpriteEffects.None, 0f);
 				if (item.color != Color.Transparent)
 					spriteBatch.Draw(value7, position2, rectangle2, item.GetColor(color), 0f, origin, num11 * scale3, SpriteEffects.None, 0f);
 
+				skip:
+				ItemLoader.PostDrawInInventory(item, spriteBatch, position2, rectangle2, item.GetAlpha(currentColor), item.GetColor(color), origin, num11 * scale3);
 				if (ItemID.Sets.TrapSigned[item.type])
 					spriteBatch.Draw(TextureAssets.Wire.Value, position + new Vector2(40f, 40f) * inventoryScale, new Rectangle(4, 58, 8, 8), color, 0f, new Vector2(4f), 1f, SpriteEffects.None, 0f);
 
@@ -1702,7 +_,7 @@
 					spriteBatch.Draw(TextureAssets.Cd.Value, position3, null, color3, 0f, default(Vector2), num11, SpriteEffects.None, 0f);
 				}
 
-				if ((context == 10 || context == 18) && item.expertOnly && !Main.expertMode) {
+				if ((context == 10 || context == 18) && ((item.expertOnly && !Main.expertMode) || (item.masterOnly && !Main.masterMode))) {
 					Vector2 position4 = position + value.Size() * inventoryScale / 2f - TextureAssets.Cd.Value.Size() * inventoryScale / 2f;
 					Color white = Color.White;
 					spriteBatch.Draw(TextureAssets.Cd.Value, position4, null, white, 0f, default(Vector2), num11, SpriteEffects.None, 0f);
@@ -1901,15 +_,15 @@
 					return false;
 
 				if (itemCollection[slot].wingSlot > 0 && item.wingSlot > 0)
-					return false;
+					return !ItemLoader.CanEquipAccessory(item, slot);
 			}
-
-			for (int i = 0; i < itemCollection.Length; i++) {
-				if (slot < 10 && i < 10) {
+			int sizeColl = itemCollection.Length;
+			for (int i = 0; i < sizeColl; i++) {
+				if (slot < sizeColl / 2  && i < sizeColl / 2) {
 					if (item.wingSlot > 0 && itemCollection[i].wingSlot > 0)
 						return true;
 
-					if (slot >= 10 && i >= 10 && item.wingSlot > 0 && itemCollection[i].wingSlot > 0)
+					if (slot >= sizeColl / 2 && i >= sizeColl / 2 && item.wingSlot > 0 && itemCollection[i].wingSlot > 0)
 						return true;
 				}
 
@@ -1917,7 +_,7 @@
 					return true;
 			}
 
-			return false;
+			return !ItemLoader.CanEquipAccessory(item, slot);
 		}
 
 		private static Item DyeSwap(Item item, out bool success) {
@@ -1934,9 +_,15 @@
 				}
 			}
 
-			if (dyeSlotCount >= 10)
+			if (dyeSlotCount >= 10) 
+			{
+				item2 = EquipLoader.VanillaDyeSwapMirror(item, out success);
+				if (success)
+					return item2;
+
 				dyeSlotCount = 0;
-
+			}
+			
 			if (dyeSlotCount < 0)
 				dyeSlotCount = 9;
 
@@ -1993,8 +_,23 @@
 						accSlotToSwapTo = j - 3;
 				}
 
+				for (int k = 0; k < num2; k++) {
+					int index = 3 + (accSlotToSwapTo + num2) % num2;
+					if (ItemLoader.CanEquipAccessory(item, index)) {
+						accSlotToSwapTo = index - 3;
+						break;
+					}
+				}
+
-				if (accSlotToSwapTo > num2)
+				if (accSlotToSwapTo > num2) {
+					result = EquipLoader.VanillaArmorSwapMirror(item, out success);
+					if (success) {
+						return result;
+					}
+
 					return item;
+				}
+					
 
 				if (accSlotToSwapTo < 0)
 					accSlotToSwapTo = num2 - 3;
@@ -2008,6 +_,9 @@
 						num3 = k;
 				}
 
+				if (!ItemLoader.CanEquipAccessory(item, num3))
+					return item;
+
 				result = player.armor[num3].Clone();
 				player.armor[num3] = item.Clone();
 				accSlotToSwapTo = 0;
@@ -2105,7 +_,7 @@
 		}
 
 		public static Color GetItemLight(ref Color currentColor, ref float scale, int type, bool outInTheWorld = false) {
-			if (type < 0 || type > 5088)
+			if (type < 0)
 				return currentColor;
 
 			if (type == 662 || type == 663) {
@@ -2345,7 +_,7 @@
 						s += PlayerInput.BuildCommand(Lang.misc[54].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]);
 					}
 
-					if (context == 8 && slot >= 3) {
+					if (context == 8 && slot >= player.totalArmorSlots) {
 						bool flag = player.hideVisibleAccessory[slot];
 						s += PlayerInput.BuildCommand(Lang.misc[flag ? 77 : 78].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]);
 						if (PlayerInput.Triggers.JustPressed.Grapple) {
@@ -2356,7 +_,7 @@
 						}
 					}
 
-					if ((context == 16 || context == 17 || context == 18 || context == 19 || context == 20) && slot < 2) {
+					if ((context == 16 || context == 17 || context == 18 || context == 19 || context == 20) && slot < player.totalArmorSlots - 1) {
 						bool flag2 = player.hideMisc[slot];
 						s += PlayerInput.BuildCommand(Lang.misc[flag2 ? 77 : 78].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]);
 						if (PlayerInput.Triggers.JustPressed.Grapple) {
@@ -2371,9 +_,9 @@
 					if (Main.mouseItem.type > 0 && Equippable(ref Main.mouseItem, context))
 						s += PlayerInput.BuildCommand(Lang.misc[65].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["MouseLeft"]);
 
-					if (context == 8 && slot >= 3) {
+					if (context == 8 && slot >= player.totalArmorSlots) {
 						int num = slot;
-						if (num % 10 == 8 && !player.CanDemonHeartAccessoryBeShown())
+						if (num % player.totalEquipSlots == player.demonHeartSlot && !player.CanAccessoryBeShown(player.demonHeartSlot))
 							num++;
 
 						bool flag3 = player.hideVisibleAccessory[num];
@@ -2386,7 +_,7 @@
 						}
 					}
 
-					if ((context == 16 || context == 17 || context == 18 || context == 19 || context == 20) && slot < 2) {
+					if ((context == 16 || context == 17 || context == 18 || context == 19 || context == 20) && slot < player.totalArmorSlots - 1) {
 						bool flag4 = player.hideMisc[slot];
 						s += PlayerInput.BuildCommand(Lang.misc[flag4 ? 77 : 78].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]);
 						if (PlayerInput.Triggers.JustPressed.Grapple) {
@@ -2424,10 +_,10 @@
 								num2 = slot;
 
 							if (inv == player.miscDyes)
-								num2 = 10 + slot;
+								num2 = player.totalEquipSlots + slot;
 
 							if (num2 != -1) {
-								if (num2 < 10) {
+								if (num2 < player.totalEquipSlots) {
 									bool flag7 = player.hideVisibleAccessory[slot];
 									s += PlayerInput.BuildCommand(Lang.misc[flag7 ? 77 : 78].Value, false, PlayerInput.ProfileGamepadUI.KeyStatus["Grapple"]);
 									if (PlayerInput.Triggers.JustPressed.Grapple) {
