--- src/TerrariaNetCore/Terraria/UI/ItemSorting.cs
+++ src/tModLoader/Terraria/UI/ItemSorting.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Terraria.ID;
+using Terraria.ModLoader;
 
 namespace Terraria.UI;
 
@@ -985,11 +_,17 @@
 
 	private struct DamageTypeSortingLayerEntry
 	{
+		/*
 		public float Multiplier;
+		*/
+		public StatModifier Multiplier;
 		public ItemSortingLayer Layer;
 		public int Index;
 
+		/*
 		public DamageTypeSortingLayerEntry(float multiplier, ItemSortingLayer layer, int index)
+		*/
+		public DamageTypeSortingLayerEntry(StatModifier multiplier, ItemSortingLayer layer, int index)
 		{
 			Multiplier = multiplier;
 			Layer = layer;
@@ -1121,7 +_,7 @@
 		list.Add(ItemSortingLayers.LastTilesCommon);
 		list.Add(ItemSortingLayers.LastNotTrash);
 		list.Add(ItemSortingLayers.LastTrash);
-		for (int i = -48; i < ItemID.Count; i++) {
+		for (int i = -48; i < ItemLoader.ItemCount; i++) {
 			Item item = new Item();
 			item.netDefaults(i);
 			list2.Add(item);
@@ -1130,7 +_,7 @@
 
 		Item[] array = list2.ToArray();
 		_layerCount = list.Count;
-		_layerIndexForItemType = new int[ItemID.Count];
+		_layerIndexForItemType = new int[ItemLoader.ItemCount];
 		for (int j = 0; j < list.Count; j++) {
 			ItemSortingLayer itemSortingLayer = list[j];
 			List<int> list4 = itemSortingLayer.SortingMethod(itemSortingLayer, array, list3);
@@ -1161,7 +_,10 @@
 
 	private static int Descending(DamageTypeSortingLayerEntry x, DamageTypeSortingLayerEntry y)
 	{
+		/*
 		int num = y.Multiplier.CompareTo(x.Multiplier);
+		*/
+		int num = (y.Multiplier.Additive * y.Multiplier.Multiplicative).CompareTo(x.Multiplier.Additive * x.Multiplier.Multiplicative);
 		if (num == 0)
 			num = x.Index.CompareTo(y.Index);
 
@@ -1251,6 +_,9 @@
 
 				Item item3 = inv[_sort_itemsToSort[k]];
 				if (Item.CanStack(item2, item3) && item3.stack != item3.maxStack) {
+					ItemLoader.StackItems(item2, item3, out _);
+
+					/* #OnStackHook
 					int num2 = item3.stack;
 					if (num < num2)
 						num2 = num;
@@ -1258,6 +_,7 @@
 					item2.stack += num2;
 					item3.stack -= num2;
 					num -= num2;
+					*/
 					if (item3.stack == 0) {
 						inv[_sort_itemsToSort[k]] = new Item();
 						_sort_itemsToSort.Remove(_sort_itemsToSort[k]);
@@ -1557,6 +_,11 @@
 		for (int i = loopStartIndex; i < loopEndIndex; i++) {
 			Item item = inv[i];
 			if (item.stack >= 1 && item.type == itemToRefill.type && !item.favorited) {
+				if (!ItemLoader.TryStackItems(itemToRefill, inv[i], out int numTransfered))
+					continue;
+
+				num -= numTransfered;
+				/*
 				int num2 = item.stack;
 				if (num2 > num)
 					num2 = num;
@@ -1564,6 +_,8 @@
 				num -= num2;
 				itemToRefill.stack += num2;
 				item.stack -= num2;
+				*/
+
 				if (item.stack <= 0)
 					item.TurnToAir();
 
