--- src/Terraria/Terraria/UI/LegacyNetDiagnosticsUI.cs
+++ src/tModLoader/Terraria/UI/LegacyNetDiagnosticsUI.cs
@@ -12,7 +_,7 @@
 		public static int rxData = 0;
 		public static int txMsg = 0;
 		public static int rxMsg = 0;
-		private const int maxMsg = 142;
+		internal const int maxMsg = 142;
 		public static int[] rxMsgType = new int[142];
 		public static int[] rxDataType = new int[142];
 		public static int[] txMsgType = new int[142];
