--- src/TerrariaNetCore/Terraria/UI/NetDiagnosticsUI.cs
+++ src/tModLoader/Terraria/UI/NetDiagnosticsUI.cs
@@ -34,6 +_,7 @@
 			}
 		}
 
+		internal const int MaxMessages = 144; // Added by TML to check for parity when vanilla updates
 		private CounterForMessage[] _counterByMessageId = new CounterForMessage[144];
 		private Dictionary<int, CounterForMessage> _counterByModuleId = new Dictionary<int, CounterForMessage>();
 		private int _highestFoundReadBytes = 1;
