--- src/TerrariaNetCore/Terraria/UI/NetDiagnosticsUI.cs
+++ src/tModLoader/Terraria/UI/NetDiagnosticsUI.cs
@@ -65,7 +_,11 @@
 
 	public void CountReadMessage(int messageId, int messageLength)
 	{
+		// Check added by TML. Ignores TML MessageIDs and ModPackets.
+		if (messageId >= _counterByMessageId.Length)
+			return;
+
 		Interlocked.Add(ref bytesRecv, messageLength);
 		_counterByMessageId[messageId].CountReadMessage(messageLength);
 	}
 
