--- src/TerrariaNetCore/Terraria/UI/UIElement.cs
+++ src/tModLoader/Terraria/UI/UIElement.cs
@@ -8,7 +_,7 @@
 
 namespace Terraria.UI;
 
-public class UIElement : IComparable
+public partial class UIElement : IComparable
 {
 	public delegate void MouseEvent(UIMouseEvent evt, UIElement listeningElement);
 
@@ -167,13 +_,22 @@
 		if (overflowHidden) {
 			spriteBatch.End();
 			Rectangle clippingRectangle = GetClippingRectangle(spriteBatch);
+
+			/*
 			spriteBatch.GraphicsDevice.ScissorRectangle = clippingRectangle;
+			*/
+			Rectangle adjustedClippingRectangle = Rectangle.Intersect(clippingRectangle, spriteBatch.GraphicsDevice.ScissorRectangle);
+			spriteBatch.GraphicsDevice.ScissorRectangle = adjustedClippingRectangle;
+
 			spriteBatch.GraphicsDevice.RasterizerState = OverflowHiddenRasterizerState;
 			spriteBatch.Begin(SpriteSortMode.Deferred, BlendState.AlphaBlend, anisotropicClamp, DepthStencilState.None, OverflowHiddenRasterizerState, null, Main.UIScaleMatrix);
 		}
 
 		DrawChildren(spriteBatch);
 		if (overflowHidden) {
+			// TML: save a new rasterizer state snapshot to restore
+			rasterizerState = spriteBatch.GraphicsDevice.RasterizerState;
+
 			spriteBatch.End();
 			spriteBatch.GraphicsDevice.ScissorRectangle = scissorRectangle;
 			spriteBatch.GraphicsDevice.RasterizerState = rasterizerState;
