--- src/merged\Terraria.csproj
+++ src/Terraria\Terraria.csproj
@@ -26,9 +26,11 @@
     <Optimize>true</Optimize>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
-    <DefineConstants>CLIENT</DefineConstants>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <DefineConstants>DEBUG;CLIENT</DefineConstants>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants>CLIENT</DefineConstants>
   </PropertyGroup>
   <ItemGroup>
@@ -38,11 +40,20 @@
     <Reference Include="Microsoft.Xna.Framework.Graphics" />
     <Reference Include="System" />
     <Reference Include="Microsoft.Xna.Framework.Xact" />
-    <Reference Include="Steamworks.NET" />
-    <Reference Include="Newtonsoft.Json" />
+    <Reference Include="Steamworks.NET">
+      <HintPath>Terraria.Libraries\Steamworks.NET\Steamworks.NET.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Newtonsoft.Json">
+      <HintPath>Terraria.Libraries\JSON.NET\Newtonsoft.Json.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
     <Reference Include="System.Core" />
     <Reference Include="System.Drawing" />
-    <Reference Include="Ionic.Zip.CF" />
+    <Reference Include="Ionic.Zip.CF">
+      <HintPath>Terraria.Libraries\DotNetZip\Ionic.Zip.CF.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
     <Reference Include="WindowsBase" />
   </ItemGroup>
   <ItemGroup>

