--- src/merged\Terraria.GameContent.UI\WiresUI.cs
+++ src/Terraria\Terraria.GameContent.UI\WiresUI.cs
@@ -190,7 +_,11 @@
 							flag2 = WiresUI.Settings.ToolMode.HasFlag(WiresUI.Settings.MultiToolMode.Actuator);
 						}
 						Vector2 vector4 = vector2 + Vector2.UnitX * (45f * ((float)i - 1.5f));
-						int num = i ?? 3;
+						int num = i;
+						if (i == 0)
+						{
+							num = 3;
+						}
 						if (i == 3)
 						{
 							num = 0;
@@ -527,7 +_,11 @@
 					{
 						float scale2 = num2;
 						Vector2 vec = value + Vector2.UnitX * (45f * ((float)i - 1.5f));
-						int num3 = i ?? 3;
+						int num3 = i;
+						if (i == 0)
+						{
+							num3 = 3;
+						}
 						if (i == 1)
 						{
 							num3 = 2;

