--- src/Terraria\Terraria.Social.Steam\Lobby.cs
+++ src/tModLoader\Terraria.Social.Steam\Lobby.cs
@@ -24,9 +24,9 @@
 
 		public void Create(bool inviteOnly, CallResult<LobbyCreated_t>.APIDispatchDelegate callResult)
 		{
-			SteamAPICall_t hAPICall = SteamMatchmaking.CreateLobby(inviteOnly ? ELobbyType.k_ELobbyTypePrivate : ELobbyType.k_ELobbyTypeFriendsOnly, 256);
+			SteamAPICall_t steamAPICall_t = SteamMatchmaking.CreateLobby(inviteOnly ? ELobbyType.k_ELobbyTypePrivate : ELobbyType.k_ELobbyTypeFriendsOnly, 256);
 			this._lobbyCreatedExternalCallback = callResult;
-			this._lobbyCreated.Set(hAPICall, null);
+			this._lobbyCreated.Set(steamAPICall_t, null);
 			this.State = LobbyState.Creating;
 		}
 
@@ -48,8 +48,8 @@
 			}
 			this.State = LobbyState.Connecting;
 			this._lobbyEnterExternalCallback = callResult;
-			SteamAPICall_t hAPICall = SteamMatchmaking.JoinLobby(lobbyId);
-			this._lobbyEnter.Set(hAPICall, null);
+			SteamAPICall_t steamAPICall_t = SteamMatchmaking.JoinLobby(lobbyId);
+			this._lobbyEnter.Set(steamAPICall_t, null);
 		}
 
 		public byte[] GetMessage(int index)
@@ -125,7 +125,7 @@
 			}
 			this.Id = new CSteamID(result.m_ulSteamIDLobby);
 			this.Owner = SteamMatchmaking.GetLobbyOwner(this.Id);
-			this._lobbyEnterExternalCallback(result, failure);
+			this._lobbyEnterExternalCallback.Invoke(result, failure);
 		}
 
 		private void OnLobbyCreated(LobbyCreated_t result, bool failure)
@@ -144,7 +144,7 @@
 			}
 			this.Id = new CSteamID(result.m_ulSteamIDLobby);
 			this.Owner = SteamMatchmaking.GetLobbyOwner(this.Id);
-			this._lobbyCreatedExternalCallback(result, failure);
+			this._lobbyCreatedExternalCallback.Invoke(result, failure);
 		}
 	}
 }

