--- src/TerrariaNetCore/Terraria/GameContent/Creative/CreativeUI.cs
+++ src/tModLoader/Terraria/GameContent/Creative/CreativeUI.cs
@@ -9,13 +_,14 @@
 using Terraria.GameContent.UI.States;
 using Terraria.GameInput;
 using Terraria.Localization;
+using Terraria.ModLoader;
 using Terraria.Net;
 using Terraria.UI;
 using Terraria.UI.Gamepad;
 
 namespace Terraria.GameContent.Creative;
 
-public class CreativeUI
+public partial class CreativeUI
 {
 	public enum ItemSacrificeResult
 	{
@@ -240,6 +_,9 @@
 		int amountNeededTotal = 0;
 		int amountWeHave = 0;
 		amountWeSacrificed = 0;
+		if (!ItemLoader.CanResearch(item))
+			return ItemSacrificeResult.CannotSacrifice;
+
 		if (!Main.LocalPlayerCreativeTracker.ItemSacrifices.TryGetSacrificeNumbers(item.type, out amountWeHave, out amountNeededTotal))
 			return ItemSacrificeResult.CannotSacrifice;
 
@@ -256,6 +_,8 @@
 		NetManager.Instance.SendToServer(packet);
 		if (!Main.ServerSideCharacter)
 			Main.LocalPlayerCreativeTracker.ItemSacrifices.RegisterItemSacrifice(item.type, num2);
+
+		ItemLoader.OnResearched(item, flag);
 
 		item.stack -= num2;
 		if (item.stack <= 0)
