--- src/TerrariaNetCore/Terraria/GameContent/Creative/ItemFilters.cs
+++ src/tModLoader/Terraria/GameContent/Creative/ItemFilters.cs
@@ -7,6 +_,7 @@
 using Terraria.DataStructures;
 using Terraria.GameContent.UI.Elements;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent.Creative;
@@ -22,19 +_,43 @@
 		private int _unusedResearchLine;
 		private string _search;
 
+		// Added by TML.
+		private string[] _toolTipNames = new string[30];
+
 		public bool FitsFilter(Item entry)
 		{
 			if (_search == null)
 				return true;
 
+			// Have to reinitialize because ItemLoader.ModifyTooltips resizes the arrays
+			int tooltipCount = _tooltipMaxLines + entry.ToolTip?.Lines ?? 0; // Easy fix to #4772. A full fix would require rewrites of hooks.
+			_toolTipNames = new string[tooltipCount];
+			_toolTipLines = new string[tooltipCount];
+			_unusedColor = new Color[tooltipCount];
+
 			int numLines = 1;
 			float knockBack = entry.knockBack;
 			int stack = entry.stack;
 			entry.stack = 1;
-			Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedColor);
+			Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedColor, _toolTipNames, out _);
+
+			List<TooltipLine> modifiedTooltipLines;
+			try {
+				modifiedTooltipLines = ItemLoader.ModifyTooltips(entry, ref numLines, _toolTipNames, ref _toolTipLines, ref _unusedColor, ref _unusedYoyoLogo, -1);
+			}
+			catch {
+				// Fix erroneous tooltips preventing the filter from working entirely
+				return false;
+			}
+
 			entry.stack = stack;
+
+			/*
 			for (int i = 0; i < numLines; i++) {
 				if (_toolTipLines[i].IndexOf(_search, StringComparison.OrdinalIgnoreCase) != -1)
+			*/
+			foreach (var line in modifiedTooltipLines) {
+				if (line.Text.ToLower().IndexOf(_search, StringComparison.OrdinalIgnoreCase) != -1)
 					return true;
 			}
 
@@ -111,6 +_,7 @@
 
 	public class Tools : IItemEntryFilter, IEntryFilter<Item>
 	{
+		/* Replaced by ItemID.Sets.DuplicationMenuToolsFilter
 		private HashSet<int> _itemIdsThatAreAccepted = new HashSet<int> {
 			213,
 			5295,
@@ -170,6 +_,7 @@
 			114,
 			5667
 		};
+		*/
 
 		public bool FitsFilter(Item entry)
 		{
@@ -188,7 +_,10 @@
 			if (entry.tileWand != -1)
 				return true;
 
+			/* See the above comment
 			if (_itemIdsThatAreAccepted.Contains(entry.type))
+			*/
+			if (ItemID.Sets.DuplicationMenuToolsFilter[entry.type])
 				return true;
 
 			return false;
@@ -366,11 +_,12 @@
 	public class MiscFallback : IItemEntryFilter, IEntryFilter<Item>
 	{
 		private readonly List<IItemEntryFilter> otherFiltersToCheckAgainst;
-		private bool?[] _fitsFilterByItemType = new bool?[ItemID.Count];
+		private bool?[] _fitsFilterByItemType;
 
 		public MiscFallback(List<IItemEntryFilter> otherFiltersToCheckAgainst)
 		{
 			this.otherFiltersToCheckAgainst = otherFiltersToCheckAgainst;
+			_fitsFilterByItemType = new bool?[ItemLoader.ItemCount];
 		}
 
 		public bool FitsFilter(Item entry)
