--- src/Terraria/Terraria/GameContent/Creative/ItemFilters.cs
+++ src/tModLoader/Terraria/GameContent/Creative/ItemFilters.cs
@@ -19,6 +_,7 @@
 			private bool[] _unusedBadPrefixLines = new bool[30];
 			private int _unusedYoyoLogo;
 			private int _unusedResearchLine;
+			private string[] _toolTipNames = new string[30];
 			private string _search;
 
 			public bool FitsFilter(Item entry) {
@@ -27,7 +_,7 @@
 
 				int numLines = 1;
 				float knockBack = entry.knockBack;
-				Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedPrefixLine, _unusedBadPrefixLines);
+				Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedPrefixLine, _unusedBadPrefixLines, _toolTipNames);
 				for (int i = 0; i < numLines; i++) {
 					if (_toolTipLines[i].ToLower().IndexOf(_search, StringComparison.OrdinalIgnoreCase) != -1)
 						return true;
