--- src/TerrariaNetCore/Terraria/GameContent/Creative/ItemFilters.cs
+++ src/tModLoader/Terraria/GameContent/Creative/ItemFilters.cs
@@ -5,6 +_,7 @@
 using Terraria.DataStructures;
 using Terraria.GameContent.UI.Elements;
 using Terraria.ID;
+using Terraria.ModLoader;
 using Terraria.UI;
 
 namespace Terraria.GameContent.Creative;
@@ -21,16 +_,31 @@
 		private int _unusedResearchLine;
 		private string _search;
 
+		// Added by TML.
+		private string[] _toolTipNames = new string[30];
+
 		public bool FitsFilter(Item entry)
 		{
 			if (_search == null)
 				return true;
 
+			// Have to reinitialize because ItemLoader.ModifyTooltips resizes the arrays
+			_toolTipNames = new string[30];
+			_toolTipLines = new string[30];
+			_unusedPrefixLine = new bool[30];
+			_unusedBadPrefixLines = new bool[30];
+
 			int numLines = 1;
 			float knockBack = entry.knockBack;
-			Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedPrefixLine, _unusedBadPrefixLines);
+			Main.MouseText_DrawItemTooltip_GetLinesInfo(entry, ref _unusedYoyoLogo, ref _unusedResearchLine, knockBack, ref numLines, _toolTipLines, _unusedPrefixLine, _unusedBadPrefixLines, _toolTipNames, out _);
+			var modifiedTooltipLines = ItemLoader.ModifyTooltips(entry, ref numLines, _toolTipNames, ref _toolTipLines, ref _unusedPrefixLine, ref _unusedBadPrefixLines, ref _unusedYoyoLogo, out _, -1);
+
+			/*
 			for (int i = 0; i < numLines; i++) {
 				if (_toolTipLines[i].ToLower().IndexOf(_search, StringComparison.OrdinalIgnoreCase) != -1)
+			*/
+			foreach (var line in modifiedTooltipLines) {
+				if (line.Text.ToLower().IndexOf(_search, StringComparison.OrdinalIgnoreCase) != -1)
 					return true;
 			}
 
@@ -107,6 +_,7 @@
 
 	public class Tools : IItemEntryFilter, IEntryFilter<Item>
 	{
+		/* Replaced by ItemID.Sets.DuplicationMenuToolsFilter
 		private HashSet<int> _itemIdsThatAreAccepted = new HashSet<int> {
 			509,
 			850,
@@ -163,6 +_,7 @@
 			4049,
 			114
 		};
+		*/
 
 		public bool FitsFilter(Item entry)
 		{
@@ -181,7 +_,10 @@
 			if (entry.tileWand != -1)
 				return true;
 
+			/* See the above comment
 			if (_itemIdsThatAreAccepted.Contains(entry.type))
+			*/
+			if (ItemID.Sets.DuplicationMenuToolsFilter[entry.type])
 				return true;
 
 			return false;
@@ -354,7 +_,7 @@
 
 		public MiscFallback(List<IItemEntryFilter> otherFiltersToCheckAgainst)
 		{
-			short count = ItemID.Count;
+			int count = ItemLoader.ItemCount;
 			_fitsFilterByItemType = new bool[count];
 			for (int i = 1; i < count; i++) {
 				_fitsFilterByItemType[i] = true;
