--- src/TerrariaNetCore/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs
+++ src/tModLoader/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs
@@ -1,7 +_,8 @@
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Terraria.ID;
 
 namespace Terraria.GameContent.Creative;
@@ -9,11 +_,11 @@
 public class ItemsSacrificedUnlocksTracker : IPersistentPerWorldContent, IOnPlayerJoining
 {
 	public const int POSITIVE_SACRIFICE_COUNT_CAP = 9999;
-	private ConcurrentDictionary<string, int> _sacrificeCountByItemPersistentId;
+	internal Dictionary<string, int> _sacrificeCountByItemPersistentId;
-	private Dictionary<int, int> _sacrificesCountByItemIdCache;
 	private Dictionary<int, string> _unlockedByTeammate;
 	private HashSet<int> _newlyUnlocked;
 	public bool AnyNewUnlocksFromTeammates;
+	internal Dictionary<int, int> _sacrificesCountByItemIdCache;
 
 	public int LastEditId { get; private set; }
 
@@ -129,7 +_,14 @@
 	public void Save(BinaryWriter writer)
 	{
 		writer.Write(value: false);
+		/*
 		KeyValuePair<string, int>[] array = _sacrificeCountByItemPersistentId.ToArray();
+		*/
+		// Only include vanilla items in the dictionary.
+		Dictionary<string, int> dictionary = _sacrificeCountByItemPersistentId
+			.Where(pair => ContentSamples.ItemNetIdsByPersistentIds.TryGetValue(pair.Key, out int netID) && ContentSamples.ItemsByType.TryGetValue(netID, out Item realItem) && realItem.ModItem == null)
+			.ToArray();
+
 		writer.Write(array.Length);
 		KeyValuePair<string, int>[] array2 = array;
 		for (int i = 0; i < array2.Length; i++) {
