--- src/TerrariaNetCore/Terraria/GameContent/UI/Chat/ItemTagHandler.cs
+++ src/tModLoader/Terraria/GameContent/UI/Chat/ItemTagHandler.cs
@@ -1,6 +_,9 @@
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
 using Terraria.ID;
+using Terraria.ModLoader;
+using Terraria.ModLoader.IO;
 using Terraria.UI;
 using Terraria.UI.Chat;
 
@@ -31,6 +_,7 @@
 			Main.instance.MouseText(_item.Name, _item.rare, 0);
 		}
 
+		//TODO: Possibly allow modders to custom draw here
 		public override bool UniqueDraw(bool justCheckingString, out Vector2 size, SpriteBatch spriteBatch, Vector2 position = default(Vector2), Color color = default(Color), float scale = 1f)
 		{
 			if (Main.netMode != 2 && !Main.dedServ)
@@ -52,20 +_,30 @@
 	TextSnippet ITagHandler.Parse(string text, Color baseColor, string options)
 	{
 		Item item = new Item();
-		if (int.TryParse(text, out var result))
+		if (int.TryParse(text, out var result) && result < ItemLoader.ItemCount)
+			item.SetDefaults(result);
+
+		// Add support for [i:ModItem.FullName] ([i:ExampleMod/ExampleItem]). Coincidentally support [i:ItemID.FieldName] ([i:GoldBar])
+		if (ID.ItemID.Search.TryGetId(text, out result))
 			item.SetDefaults(result);
 
 		if (item.type <= 0)
 			return new TextSnippet(text);
 
 		item.stack = 1;
+		// Options happen here, we add MID (=ModItemData) options
 		if (options != null) {
+			// Don't know why all these options are here in vanilla,
+			// Since it only assumed one option (stack OR prefix, since prefixed items don't stack)
 			string[] array = options.Split(',');
 			for (int i = 0; i < array.Length; i++) {
 				if (array[i].Length == 0)
 					continue;
 
 				switch (array[i][0]) {
+					case 'd': // MID is present, we will override
+						item = ItemIO.FromBase64(array[i].Substring(1));
+						break;
 					case 's':
 					case 'x': {
 						if (int.TryParse(array[i].Substring(1), out var result3))
@@ -75,7 +_,7 @@
 					}
 					case 'p': {
 						if (int.TryParse(array[i].Substring(1), out var result2))
-							item.Prefix((byte)Utils.Clamp(result2, 0, PrefixID.Count));
+							item.Prefix((byte)Utils.Clamp(result2, 0, PrefixLoader.PrefixCount));
 
 						break;
 					}
@@ -86,10 +_,21 @@
 		return new ItemSnippet(item);
 	}
 
+	// We do not alter vanilla ways of doing things
+	// This can lead to trouble in future patches
 	public static string GenerateTag(Item I)
 	{
 		string text = "[i";
 		bool flag = false;
+		// Assuming we have modded data, simply write the item as base64.
+		// Do not write the other option, base64 holds all the info.
+		// Yes, doubling up on SaveGlobals isn't the best for performance, but oh well
+		if (ItemLoader.NeedsModSaving(I) || ItemIO.SaveGlobals(I) != null) {
+			text = text + "/d" + ItemIO.ToBase64(I);
+
+			goto End;
+		}
+
 		if (I.prefix != 0) {
 			text = text + (flag ? ",p" : "/p") + I.prefix;
 			flag = true;
@@ -99,6 +_,8 @@
 			text = text + (flag ? ",s" : "/s") + I.stack;
 			flag = true;
 		}
+
+		End:
 
 		return text + ":" + I.type + "]";
 	}
