--- src/TerrariaNetCore/Terraria/GameContent/UI/Chat/RemadeChatMonitor.cs
+++ src/tModLoader/Terraria/GameContent/UI/Chat/RemadeChatMonitor.cs
@@ -34,7 +_,9 @@
 
 	public void AddNewMessage(string text, Color color, int widthLimitInPixels = -1)
 	{
 		Trace.WriteLine("[chat] " + text);
+		if (Main.dedServ) // TML: added to prevent Main.NewText causing exception on server
+			return;
 		ChatMessageContainer chatMessageContainer = new ChatMessageContainer();
 		chatMessageContainer.SetContents(text, color, widthLimitInPixels);
 		_messages.Insert(0, chatMessageContainer);
