--- src/TerrariaNetCore/Terraria/Chat/Commands/EmojiCommand.cs
+++ src/tModLoader/Terraria/Chat/Commands/EmojiCommand.cs
@@ -2,6 +_,7 @@
 using System.Collections.Generic;
 using Terraria.GameContent.UI;
 using Terraria.Localization;
+using Terraria.ModLoader;
 
 namespace Terraria.Chat.Commands;
 
@@ -19,7 +_,7 @@
 	public void Initialize()
 	{
 		_byName.Clear();
-		for (int i = 0; i < EmoteID.Count; i++) {
+		for (int i = 0; i < EmoteBubbleLoader.EmoteBubbleCount; i++) {
 			LocalizedText emojiName = Lang.GetEmojiName(i);
 			if (emojiName != LocalizedText.Empty)
 				_byName[emojiName] = i;
@@ -28,10 +_,19 @@
 
 	public void PrepareAliases(ChatCommandProcessor commandProcessor)
 	{
+		Initialize(); // Added by TML, also sets up _byName for modded emotes
+
 		for (int i = 0; i < EmoteID.Count; i++) {
 			string name = EmoteID.Search.GetName(i);
 			commandProcessor.AddAlias(Language.GetText("EmojiCommand." + name), () => string.Format("{0} {1}", Language.GetTextValue("ChatCommand.Emoji_1"), Language.GetTextValue("EmojiName." + name)));
 		}
+
+		// Vanilla uses 2 keys per emote, EmojiName.X for the name and EmojiCommand.X = "/{$EmojiName.X}" for the command.
+		// TML added ChatCommandProcessor.MatchCommandPrefix so ModEmoteBubble.Command can serve as both name and command
+		for (int i = EmoteID.Count; i < EmoteBubbleLoader.EmoteBubbleCount; i++) {
+			LocalizedText command = Lang.GetEmojiName(i);
+			commandProcessor.AddAlias(command, () => string.Format("{0} {1}", Language.GetTextValue("ChatCommand.Emoji_1"), command.Value));
+		}
 	}
 
 	public void ProcessIncomingMessage(string text, byte clientId)
@@ -47,7 +_,7 @@
 
 		int result = -1;
 		if (int.TryParse(message.Text, out result)) {
-			if (result < 0 || result >= EmoteID.Count)
+			if (result < 0 || result >= EmoteBubbleLoader.EmoteBubbleCount)
 				return;
 		}
 		else {
